{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>.container { width:100% !important; }</style>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from IPython.core.display import display, HTML\n",
    "display(HTML(\"<style>.container { width:100% !important; }</style>\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import vcf\n",
    "\n",
    "%matplotlib inline\n",
    "import os\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib as mpl\n",
    "import matplotlib.ticker as ticker\n",
    "from itertools import compress\n",
    "from pylab import MaxNLocator\n",
    "import seaborn as sns; sns.set()\n",
    "from matplotlib.colors import LogNorm\n",
    "from matplotlib import gridspec\n",
    "import ast\n",
    "import itertools\n",
    "import seaborn as sns\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "\n",
    "import fastcluster\n",
    "from sklearn import cluster, datasets\n",
    "import scipy.cluster.hierarchy as hier\n",
    "from sklearn.cluster import KMeans\n",
    "import time\n",
    "import sys\n",
    "\n",
    "import Bio\n",
    "from Bio.Alphabet import IUPAC\n",
    "from Bio.Blast.Applications import NcbiblastnCommandline\n",
    "from Bio.Blast import NCBIXML\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.SeqFeature import SeqFeature, FeatureLocation\n",
    "from Bio import pairwise2\n",
    "from Bio import SeqIO\n",
    "from Bio.Graphics import GenomeDiagram\n",
    "from Bio.SeqUtils import GC\n",
    "\n",
    "from Bio.Align.Applications import MuscleCommandline\n",
    "from StringIO import StringIO\n",
    "from Bio import AlignIO\n",
    "from Bio.Align import AlignInfo\n",
    "from Bio.Seq import MutableSeq\n",
    "import itertools\n",
    "\n",
    "import networkx as nx\n",
    "import scipy\n",
    "from collections import Counter\n",
    "\n",
    "from pylab import plot, show, savefig, xlim, figure, hold, ylim, legend, boxplot, setp, axes\n",
    "import math\n",
    "\n",
    "#for exporting to Adobe Illustrator\n",
    "mpl.rcParams['pdf.fonttype'] = 42\n",
    "mpl.rcParams['ps.fonttype'] = 42"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load in Sample Annotation & both DataFrames from In-Host SNP analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Import Sample Annotation file for all patients that have been filtered by Lineage Call and Fixed SNP Distance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_annotation = pd.read_csv('/n/data1/hms/dbmi/farhat/Roger/inhost_TB_dynamics_project/CSV_files/sample_annotation_files/Longitudinal_fastq_path_names_and_JankyPipe_tags_filtered_final.csv' , sep = ',').set_index('patient_id')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fastq_files</th>\n",
       "      <th>population</th>\n",
       "      <th>run_ID</th>\n",
       "      <th>sample_ID</th>\n",
       "      <th>sample_order</th>\n",
       "      <th>tag</th>\n",
       "      <th>isolate_type</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>patient_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>P000183</th>\n",
       "      <td>/n/data1/hms/dbmi/farhat/fastq_db/walker/ERR03...</td>\n",
       "      <td>WALKER</td>\n",
       "      <td>ERR039337</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>ERR039337</td>\n",
       "      <td>longitudinal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P000183</th>\n",
       "      <td>/n/data1/hms/dbmi/farhat/fastq_db/walker/ERR03...</td>\n",
       "      <td>WALKER</td>\n",
       "      <td>ERR039338</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>ERR039338</td>\n",
       "      <td>longitudinal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1960</th>\n",
       "      <td>/n/data1/hms/dbmi/farhat/cetr_strains/good_wgs...</td>\n",
       "      <td>CETR</td>\n",
       "      <td>MMJA00000000</td>\n",
       "      <td>Peru5115</td>\n",
       "      <td>1</td>\n",
       "      <td>Peru5115</td>\n",
       "      <td>longitudinal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1960</th>\n",
       "      <td>/n/data1/hms/dbmi/farhat/cetr_strains/good_wgs...</td>\n",
       "      <td>CETR</td>\n",
       "      <td>MMPC00000000</td>\n",
       "      <td>Peru4668</td>\n",
       "      <td>2</td>\n",
       "      <td>Peru4668</td>\n",
       "      <td>longitudinal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2491</th>\n",
       "      <td>/n/data1/hms/dbmi/farhat/cetr_strains/good_wgs...</td>\n",
       "      <td>CETR</td>\n",
       "      <td>MMZT00000000</td>\n",
       "      <td>Peru4903</td>\n",
       "      <td>1</td>\n",
       "      <td>Peru4903</td>\n",
       "      <td>longitudinal</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                  fastq_files population  \\\n",
       "patient_id                                                                 \n",
       "P000183     /n/data1/hms/dbmi/farhat/fastq_db/walker/ERR03...     WALKER   \n",
       "P000183     /n/data1/hms/dbmi/farhat/fastq_db/walker/ERR03...     WALKER   \n",
       "1960        /n/data1/hms/dbmi/farhat/cetr_strains/good_wgs...       CETR   \n",
       "1960        /n/data1/hms/dbmi/farhat/cetr_strains/good_wgs...       CETR   \n",
       "2491        /n/data1/hms/dbmi/farhat/cetr_strains/good_wgs...       CETR   \n",
       "\n",
       "                  run_ID sample_ID  sample_order        tag  isolate_type  \n",
       "patient_id                                                                 \n",
       "P000183        ERR039337       NaN             1  ERR039337  longitudinal  \n",
       "P000183        ERR039338       NaN             2  ERR039338  longitudinal  \n",
       "1960        MMJA00000000  Peru5115             1   Peru5115  longitudinal  \n",
       "1960        MMPC00000000  Peru4668             2   Peru4668  longitudinal  \n",
       "2491        MMZT00000000  Peru4903             1   Peru4903  longitudinal  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_annotation.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Load in filtered Base Calls (SNPs) from in-host Base Call analysis (pairs of Base Calls that are different *in-host*, have a difference in Alternate Allele Frequency $\\ge 70\\%$)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Two** rows per SNP (1 for each isolate)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "within_host_Base_Call_variants_all_patients = pd.read_pickle('/n/data1/hms/dbmi/farhat/Roger/inhost_TB_dynamics_project/pickled_files/variant_calling/longitudinal_SNPs/longitudinal_SNP_variants_70_delta_in_alt_AF.pkl')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ref_base</th>\n",
       "      <th>alt_base</th>\n",
       "      <th>ref_position</th>\n",
       "      <th>quality</th>\n",
       "      <th>SNP_type</th>\n",
       "      <th>PASS_filter</th>\n",
       "      <th>INFO</th>\n",
       "      <th>alt_AF</th>\n",
       "      <th>depth</th>\n",
       "      <th>tag</th>\n",
       "      <th>population</th>\n",
       "      <th>patient_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_coord</th>\n",
       "      <th>gene_category</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>SNP_ftype</th>\n",
       "      <th>AA_change</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>C</td>\n",
       "      <td>Z</td>\n",
       "      <td>761139.0</td>\n",
       "      <td>3140.0</td>\n",
       "      <td>Ref_PASS</td>\n",
       "      <td>[]</td>\n",
       "      <td>{u'QP': [0, 100, 0, 0], u'AC': [0], u'BC': [0,...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>89.0</td>\n",
       "      <td>ERR369703</td>\n",
       "      <td>CASALI</td>\n",
       "      <td>P251</td>\n",
       "      <td>Rv0667</td>\n",
       "      <td>1333</td>\n",
       "      <td>Antibiotic Resistance</td>\n",
       "      <td>rpoB</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>761139.0</td>\n",
       "      <td>1266.0</td>\n",
       "      <td>Alt_PASS</td>\n",
       "      <td>[]</td>\n",
       "      <td>{u'QP': [0, 0, 100, 0], u'AC': [2], u'BC': [0,...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>36.0</td>\n",
       "      <td>ERR600658</td>\n",
       "      <td>CASALI</td>\n",
       "      <td>P251</td>\n",
       "      <td>Rv0667</td>\n",
       "      <td>1333</td>\n",
       "      <td>Antibiotic Resistance</td>\n",
       "      <td>rpoB</td>\n",
       "      <td>N</td>\n",
       "      <td>H445D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>T</td>\n",
       "      <td>Z</td>\n",
       "      <td>861378.0</td>\n",
       "      <td>3116.0</td>\n",
       "      <td>Ref_PASS</td>\n",
       "      <td>[]</td>\n",
       "      <td>{u'QP': [0, 0, 0, 100], u'AC': [0], u'BC': [0,...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>90.0</td>\n",
       "      <td>ERR369703</td>\n",
       "      <td>CASALI</td>\n",
       "      <td>P251</td>\n",
       "      <td>Rv0768</td>\n",
       "      <td>467</td>\n",
       "      <td>Non-Essential</td>\n",
       "      <td>aldA</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>861378.0</td>\n",
       "      <td>1167.0</td>\n",
       "      <td>Alt_PASS</td>\n",
       "      <td>[]</td>\n",
       "      <td>{u'QP': [0, 0, 100, 0], u'AC': [2], u'BC': [0,...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>34.0</td>\n",
       "      <td>ERR600658</td>\n",
       "      <td>CASALI</td>\n",
       "      <td>P251</td>\n",
       "      <td>Rv0768</td>\n",
       "      <td>467</td>\n",
       "      <td>Non-Essential</td>\n",
       "      <td>aldA</td>\n",
       "      <td>N</td>\n",
       "      <td>I156S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A</td>\n",
       "      <td>Z</td>\n",
       "      <td>1253207.0</td>\n",
       "      <td>2403.0</td>\n",
       "      <td>Ref_PASS</td>\n",
       "      <td>[]</td>\n",
       "      <td>{u'QP': [100, 0, 0, 0], u'AC': [0], u'BC': [66...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>66.0</td>\n",
       "      <td>ERR108130</td>\n",
       "      <td>BRYANT</td>\n",
       "      <td>2020E</td>\n",
       "      <td>Rv1129c</td>\n",
       "      <td>1328</td>\n",
       "      <td>Non-Essential</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  ref_base alt_base  ref_position  quality  SNP_type PASS_filter  \\\n",
       "0        C        Z      761139.0   3140.0  Ref_PASS          []   \n",
       "1        C        G      761139.0   1266.0  Alt_PASS          []   \n",
       "2        T        Z      861378.0   3116.0  Ref_PASS          []   \n",
       "3        T        G      861378.0   1167.0  Alt_PASS          []   \n",
       "4        A        Z     1253207.0   2403.0  Ref_PASS          []   \n",
       "\n",
       "                                                INFO  alt_AF  depth  \\\n",
       "0  {u'QP': [0, 100, 0, 0], u'AC': [0], u'BC': [0,...     0.0   89.0   \n",
       "1  {u'QP': [0, 0, 100, 0], u'AC': [2], u'BC': [0,...     1.0   36.0   \n",
       "2  {u'QP': [0, 0, 0, 100], u'AC': [0], u'BC': [0,...     0.0   90.0   \n",
       "3  {u'QP': [0, 0, 100, 0], u'AC': [2], u'BC': [0,...     1.0   34.0   \n",
       "4  {u'QP': [100, 0, 0, 0], u'AC': [0], u'BC': [66...     0.0   66.0   \n",
       "\n",
       "         tag population patient_id  gene_id gene_coord          gene_category  \\\n",
       "0  ERR369703     CASALI       P251   Rv0667       1333  Antibiotic Resistance   \n",
       "1  ERR600658     CASALI       P251   Rv0667       1333  Antibiotic Resistance   \n",
       "2  ERR369703     CASALI       P251   Rv0768        467          Non-Essential   \n",
       "3  ERR600658     CASALI       P251   Rv0768        467          Non-Essential   \n",
       "4  ERR108130     BRYANT      2020E  Rv1129c       1328          Non-Essential   \n",
       "\n",
       "  gene_symbol SNP_ftype AA_change  \n",
       "0        rpoB      None      None  \n",
       "1        rpoB         N     H445D  \n",
       "2        aldA      None      None  \n",
       "3        aldA         N     I156S  \n",
       "4         NaN      None      None  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "within_host_Base_Call_variants_all_patients.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(350, 18)"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.shape(within_host_Base_Call_variants_all_patients)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Load in filtered SNPs from within-host Base Call analysis (SNPs that are different *in-host*, have a difference in Alternate Allele Frequencies >= 70%)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**One** row per SNP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "SNP_variants_within_patients = pd.read_pickle('/n/data1/hms/dbmi/farhat/Roger/inhost_TB_dynamics_project/pickled_files/variant_calling/longitudinal_SNPs/SNPs_between_isolates_delta_70.pkl')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>population</th>\n",
       "      <th>patient_id</th>\n",
       "      <th>ref_position</th>\n",
       "      <th>ref_allele</th>\n",
       "      <th>alt_allele</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>genomic_coord</th>\n",
       "      <th>gene_category</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>alt_AF_diff</th>\n",
       "      <th>SNP_type</th>\n",
       "      <th>AA_change</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CASALI</td>\n",
       "      <td>P251</td>\n",
       "      <td>761139</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>Rv0667</td>\n",
       "      <td>1333</td>\n",
       "      <td>Antibiotic Resistance</td>\n",
       "      <td>rpoB</td>\n",
       "      <td>1.0</td>\n",
       "      <td>N</td>\n",
       "      <td>H445D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CASALI</td>\n",
       "      <td>P251</td>\n",
       "      <td>861378</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>Rv0768</td>\n",
       "      <td>467</td>\n",
       "      <td>Non-Essential</td>\n",
       "      <td>aldA</td>\n",
       "      <td>1.0</td>\n",
       "      <td>N</td>\n",
       "      <td>I156S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BRYANT</td>\n",
       "      <td>2020E</td>\n",
       "      <td>1253207</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>Rv1129c</td>\n",
       "      <td>1328</td>\n",
       "      <td>Non-Essential</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>N</td>\n",
       "      <td>I443T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BRYANT</td>\n",
       "      <td>2020E</td>\n",
       "      <td>2075138</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>Rv1830</td>\n",
       "      <td>298</td>\n",
       "      <td>Non-Essential</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>N</td>\n",
       "      <td>S100T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BRYANT</td>\n",
       "      <td>2020E</td>\n",
       "      <td>2694727</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>Rv2397c</td>\n",
       "      <td>238</td>\n",
       "      <td>Essential</td>\n",
       "      <td>cysA1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>N</td>\n",
       "      <td>F80L</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  population patient_id  ref_position ref_allele alt_allele  gene_id  \\\n",
       "0     CASALI       P251        761139          C          G   Rv0667   \n",
       "1     CASALI       P251        861378          T          G   Rv0768   \n",
       "2     BRYANT      2020E       1253207          A          G  Rv1129c   \n",
       "3     BRYANT      2020E       2075138          T          A   Rv1830   \n",
       "4     BRYANT      2020E       2694727          A          G  Rv2397c   \n",
       "\n",
       "  genomic_coord          gene_category gene_symbol  alt_AF_diff SNP_type  \\\n",
       "0          1333  Antibiotic Resistance        rpoB          1.0        N   \n",
       "1           467          Non-Essential        aldA          1.0        N   \n",
       "2          1328          Non-Essential         NaN          1.0        N   \n",
       "3           298          Non-Essential         NaN          1.0        N   \n",
       "4           238              Essential       cysA1          1.0        N   \n",
       "\n",
       "  AA_change  \n",
       "0     H445D  \n",
       "1     I156S  \n",
       "2     I443T  \n",
       "3     S100T  \n",
       "4      F80L  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SNP_variants_within_patients.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(175, 12)"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.shape(SNP_variants_within_patients)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Did the bacteria mutate toward the *Reference* Allele or toward the *Alternate* Allele?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "within_host_Base_Call_variants_all_patients_sample_order = pd.merge( within_host_Base_Call_variants_all_patients , sample_annotation.loc[: , ['sample_order' , 'tag']] , how = 'left' , on = 'tag')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ref_base</th>\n",
       "      <th>alt_base</th>\n",
       "      <th>ref_position</th>\n",
       "      <th>quality</th>\n",
       "      <th>SNP_type</th>\n",
       "      <th>PASS_filter</th>\n",
       "      <th>INFO</th>\n",
       "      <th>alt_AF</th>\n",
       "      <th>depth</th>\n",
       "      <th>tag</th>\n",
       "      <th>population</th>\n",
       "      <th>patient_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_coord</th>\n",
       "      <th>gene_category</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>SNP_ftype</th>\n",
       "      <th>AA_change</th>\n",
       "      <th>sample_order</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>C</td>\n",
       "      <td>Z</td>\n",
       "      <td>761139.0</td>\n",
       "      <td>3140.0</td>\n",
       "      <td>Ref_PASS</td>\n",
       "      <td>[]</td>\n",
       "      <td>{u'QP': [0, 100, 0, 0], u'AC': [0], u'BC': [0,...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>89.0</td>\n",
       "      <td>ERR369703</td>\n",
       "      <td>CASALI</td>\n",
       "      <td>P251</td>\n",
       "      <td>Rv0667</td>\n",
       "      <td>1333</td>\n",
       "      <td>Antibiotic Resistance</td>\n",
       "      <td>rpoB</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>761139.0</td>\n",
       "      <td>1266.0</td>\n",
       "      <td>Alt_PASS</td>\n",
       "      <td>[]</td>\n",
       "      <td>{u'QP': [0, 0, 100, 0], u'AC': [2], u'BC': [0,...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>36.0</td>\n",
       "      <td>ERR600658</td>\n",
       "      <td>CASALI</td>\n",
       "      <td>P251</td>\n",
       "      <td>Rv0667</td>\n",
       "      <td>1333</td>\n",
       "      <td>Antibiotic Resistance</td>\n",
       "      <td>rpoB</td>\n",
       "      <td>N</td>\n",
       "      <td>H445D</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>T</td>\n",
       "      <td>Z</td>\n",
       "      <td>861378.0</td>\n",
       "      <td>3116.0</td>\n",
       "      <td>Ref_PASS</td>\n",
       "      <td>[]</td>\n",
       "      <td>{u'QP': [0, 0, 0, 100], u'AC': [0], u'BC': [0,...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>90.0</td>\n",
       "      <td>ERR369703</td>\n",
       "      <td>CASALI</td>\n",
       "      <td>P251</td>\n",
       "      <td>Rv0768</td>\n",
       "      <td>467</td>\n",
       "      <td>Non-Essential</td>\n",
       "      <td>aldA</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>861378.0</td>\n",
       "      <td>1167.0</td>\n",
       "      <td>Alt_PASS</td>\n",
       "      <td>[]</td>\n",
       "      <td>{u'QP': [0, 0, 100, 0], u'AC': [2], u'BC': [0,...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>34.0</td>\n",
       "      <td>ERR600658</td>\n",
       "      <td>CASALI</td>\n",
       "      <td>P251</td>\n",
       "      <td>Rv0768</td>\n",
       "      <td>467</td>\n",
       "      <td>Non-Essential</td>\n",
       "      <td>aldA</td>\n",
       "      <td>N</td>\n",
       "      <td>I156S</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A</td>\n",
       "      <td>Z</td>\n",
       "      <td>1253207.0</td>\n",
       "      <td>2403.0</td>\n",
       "      <td>Ref_PASS</td>\n",
       "      <td>[]</td>\n",
       "      <td>{u'QP': [100, 0, 0, 0], u'AC': [0], u'BC': [66...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>66.0</td>\n",
       "      <td>ERR108130</td>\n",
       "      <td>BRYANT</td>\n",
       "      <td>2020E</td>\n",
       "      <td>Rv1129c</td>\n",
       "      <td>1328</td>\n",
       "      <td>Non-Essential</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  ref_base alt_base  ref_position  quality  SNP_type PASS_filter  \\\n",
       "0        C        Z      761139.0   3140.0  Ref_PASS          []   \n",
       "1        C        G      761139.0   1266.0  Alt_PASS          []   \n",
       "2        T        Z      861378.0   3116.0  Ref_PASS          []   \n",
       "3        T        G      861378.0   1167.0  Alt_PASS          []   \n",
       "4        A        Z     1253207.0   2403.0  Ref_PASS          []   \n",
       "\n",
       "                                                INFO  alt_AF  depth  \\\n",
       "0  {u'QP': [0, 100, 0, 0], u'AC': [0], u'BC': [0,...     0.0   89.0   \n",
       "1  {u'QP': [0, 0, 100, 0], u'AC': [2], u'BC': [0,...     1.0   36.0   \n",
       "2  {u'QP': [0, 0, 0, 100], u'AC': [0], u'BC': [0,...     0.0   90.0   \n",
       "3  {u'QP': [0, 0, 100, 0], u'AC': [2], u'BC': [0,...     1.0   34.0   \n",
       "4  {u'QP': [100, 0, 0, 0], u'AC': [0], u'BC': [66...     0.0   66.0   \n",
       "\n",
       "         tag population patient_id  gene_id gene_coord          gene_category  \\\n",
       "0  ERR369703     CASALI       P251   Rv0667       1333  Antibiotic Resistance   \n",
       "1  ERR600658     CASALI       P251   Rv0667       1333  Antibiotic Resistance   \n",
       "2  ERR369703     CASALI       P251   Rv0768        467          Non-Essential   \n",
       "3  ERR600658     CASALI       P251   Rv0768        467          Non-Essential   \n",
       "4  ERR108130     BRYANT      2020E  Rv1129c       1328          Non-Essential   \n",
       "\n",
       "  gene_symbol SNP_ftype AA_change  sample_order  \n",
       "0        rpoB      None      None             1  \n",
       "1        rpoB         N     H445D             2  \n",
       "2        aldA      None      None             1  \n",
       "3        aldA         N     I156S             2  \n",
       "4         NaN      None      None             1  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "within_host_Base_Call_variants_all_patients_sample_order.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(350, 19)"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.shape(within_host_Base_Call_variants_all_patients_sample_order)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "population_list = []\n",
    "patient_list = []\n",
    "ref_pos_list = []\n",
    "allele_1_list = [] #allele detected in abundance in 1st isolate\n",
    "allele_2_list = [] #allele detected in abundance in 2nd isolate\n",
    "alternate_AF_1 = []\n",
    "alternate_AF_2 = []\n",
    "change_in_alternate_AF = []\n",
    "\n",
    "#for each pair of Base Calls, store (alternate Allele Frequency of isolate_1) and (alternate Allele Frequency of isolate_2 - alternate Allele Frequency of isolate_1)\n",
    "for Base_Call_A_index , Base_Call_B_index in zip( range(0 , np.shape(within_host_Base_Call_variants_all_patients_sample_order)[0] , 2) , range(1 , np.shape(within_host_Base_Call_variants_all_patients_sample_order)[0] , 2) ):\n",
    "    \n",
    "    #find the Base Call index that corresponds to the earlier isolate\n",
    "    if within_host_Base_Call_variants_all_patients_sample_order.loc[Base_Call_A_index , 'sample_order'] == 1:\n",
    "        \n",
    "        Base_Call_isolate_1 = within_host_Base_Call_variants_all_patients_sample_order.loc[Base_Call_A_index , :]\n",
    "        Base_Call_isolate_2 = within_host_Base_Call_variants_all_patients_sample_order.loc[Base_Call_B_index , :] \n",
    "\n",
    "        #if the alternate allele frequency increased, then population is moving toward alternate base\n",
    "        if Base_Call_isolate_2.alt_AF > Base_Call_isolate_1.alt_AF:\n",
    "            allele_1 = Base_Call_isolate_1.ref_base #base that was present in the population\n",
    "            allele_2 = Base_Call_isolate_2.alt_base #base that population is moving toward\n",
    "            \n",
    "        #if the alternate allele frequency decreased, then popolation is moving toward reference base\n",
    "        elif Base_Call_isolate_2.alt_AF < Base_Call_isolate_1.alt_AF:\n",
    "            allele_1 = Base_Call_isolate_1.alt_base\n",
    "            allele_2 = Base_Call_isolate_2.ref_base\n",
    "            \n",
    "    elif within_host_Base_Call_variants_all_patients_sample_order.loc[Base_Call_B_index , 'sample_order'] == 1:\n",
    "        \n",
    "        Base_Call_isolate_1 = within_host_Base_Call_variants_all_patients_sample_order.loc[Base_Call_B_index , :]\n",
    "        Base_Call_isolate_2 = within_host_Base_Call_variants_all_patients_sample_order.loc[Base_Call_A_index , :] \n",
    "        \n",
    "        if Base_Call_isolate_2.alt_AF > Base_Call_isolate_1.alt_AF:\n",
    "            allele_1 = Base_Call_isolate_1.ref_base\n",
    "            allele_2 = Base_Call_isolate_2.alt_base\n",
    "            \n",
    "        elif Base_Call_isolate_2.alt_AF < Base_Call_isolate_1.alt_AF:\n",
    "            allele_1 = Base_Call_isolate_1.alt_base\n",
    "            allele_2 = Base_Call_isolate_2.ref_base\n",
    "            \n",
    "    #store the base that population is moving from\n",
    "    allele_1_list.append(allele_1)\n",
    "    \n",
    "    #store the base that population is moving toward\n",
    "    allele_2_list.append(allele_2)\n",
    "        \n",
    "    #store the Alterante Allele Frequency of the first serial isolate\n",
    "    alternate_AF_1.append(Base_Call_isolate_1.alt_AF)\n",
    "    \n",
    "    #stotre the Alternate Allele Frequency of the second serial isolate\n",
    "    alternate_AF_2.append(Base_Call_isolate_2.alt_AF)\n",
    "    \n",
    "    #store the difference of the Alternate Allele Frequencies of both serial isolates\n",
    "    change_in_alternate_AF.append( Base_Call_isolate_2.alt_AF - Base_Call_isolate_1.alt_AF )\n",
    "    \n",
    "    #store study ID, patient ID & reference positions\n",
    "    population_list.append(Base_Call_isolate_1.population)\n",
    "    patient_list.append(Base_Call_isolate_1.patient_id)\n",
    "    ref_pos_list.append(Base_Call_isolate_1.ref_position)\n",
    "    \n",
    "#create DataFrame with all Information\n",
    "change_in_AF_DF = pd.DataFrame()\n",
    "change_in_AF_DF['population'] = population_list\n",
    "change_in_AF_DF['patient_id'] = patient_list\n",
    "change_in_AF_DF['ref_position'] = ref_pos_list\n",
    "change_in_AF_DF['allele_1'] = allele_1_list\n",
    "change_in_AF_DF['allele_2'] = allele_2_list\n",
    "change_in_AF_DF['alt_AF_1'] = alternate_AF_1\n",
    "change_in_AF_DF['alt_AF_2'] = alternate_AF_2\n",
    "change_in_AF_DF['altAF2_minus_altAF1'] = change_in_alternate_AF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(175, 8)"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.shape(change_in_AF_DF)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>population</th>\n",
       "      <th>patient_id</th>\n",
       "      <th>ref_position</th>\n",
       "      <th>allele_1</th>\n",
       "      <th>allele_2</th>\n",
       "      <th>alt_AF_1</th>\n",
       "      <th>alt_AF_2</th>\n",
       "      <th>altAF2_minus_altAF1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CASALI</td>\n",
       "      <td>P251</td>\n",
       "      <td>761139.0</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CASALI</td>\n",
       "      <td>P251</td>\n",
       "      <td>861378.0</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BRYANT</td>\n",
       "      <td>2020E</td>\n",
       "      <td>1253207.0</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BRYANT</td>\n",
       "      <td>2020E</td>\n",
       "      <td>2075138.0</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BRYANT</td>\n",
       "      <td>2020E</td>\n",
       "      <td>2694727.0</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  population patient_id  ref_position allele_1 allele_2  alt_AF_1  alt_AF_2  \\\n",
       "0     CASALI       P251      761139.0        C        G       0.0       1.0   \n",
       "1     CASALI       P251      861378.0        T        G       0.0       1.0   \n",
       "2     BRYANT      2020E     1253207.0        A        G       0.0       1.0   \n",
       "3     BRYANT      2020E     2075138.0        T        A       0.0       1.0   \n",
       "4     BRYANT      2020E     2694727.0        A        G       0.0       1.0   \n",
       "\n",
       "   altAF2_minus_altAF1  \n",
       "0                  1.0  \n",
       "1                  1.0  \n",
       "2                  1.0  \n",
       "3                  1.0  \n",
       "4                  1.0  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "change_in_AF_DF.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Merge DataFrame above with DataFrame below to create a master DataFrame that contains all relevant information for the 175 in-host SNPs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>population</th>\n",
       "      <th>patient_id</th>\n",
       "      <th>ref_position</th>\n",
       "      <th>ref_allele</th>\n",
       "      <th>alt_allele</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>genomic_coord</th>\n",
       "      <th>gene_category</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>alt_AF_diff</th>\n",
       "      <th>SNP_type</th>\n",
       "      <th>AA_change</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CASALI</td>\n",
       "      <td>P251</td>\n",
       "      <td>761139</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>Rv0667</td>\n",
       "      <td>1333</td>\n",
       "      <td>Antibiotic Resistance</td>\n",
       "      <td>rpoB</td>\n",
       "      <td>1.0</td>\n",
       "      <td>N</td>\n",
       "      <td>H445D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CASALI</td>\n",
       "      <td>P251</td>\n",
       "      <td>861378</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>Rv0768</td>\n",
       "      <td>467</td>\n",
       "      <td>Non-Essential</td>\n",
       "      <td>aldA</td>\n",
       "      <td>1.0</td>\n",
       "      <td>N</td>\n",
       "      <td>I156S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BRYANT</td>\n",
       "      <td>2020E</td>\n",
       "      <td>1253207</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>Rv1129c</td>\n",
       "      <td>1328</td>\n",
       "      <td>Non-Essential</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>N</td>\n",
       "      <td>I443T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BRYANT</td>\n",
       "      <td>2020E</td>\n",
       "      <td>2075138</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>Rv1830</td>\n",
       "      <td>298</td>\n",
       "      <td>Non-Essential</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>N</td>\n",
       "      <td>S100T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BRYANT</td>\n",
       "      <td>2020E</td>\n",
       "      <td>2694727</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>Rv2397c</td>\n",
       "      <td>238</td>\n",
       "      <td>Essential</td>\n",
       "      <td>cysA1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>N</td>\n",
       "      <td>F80L</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  population patient_id  ref_position ref_allele alt_allele  gene_id  \\\n",
       "0     CASALI       P251        761139          C          G   Rv0667   \n",
       "1     CASALI       P251        861378          T          G   Rv0768   \n",
       "2     BRYANT      2020E       1253207          A          G  Rv1129c   \n",
       "3     BRYANT      2020E       2075138          T          A   Rv1830   \n",
       "4     BRYANT      2020E       2694727          A          G  Rv2397c   \n",
       "\n",
       "  genomic_coord          gene_category gene_symbol  alt_AF_diff SNP_type  \\\n",
       "0          1333  Antibiotic Resistance        rpoB          1.0        N   \n",
       "1           467          Non-Essential        aldA          1.0        N   \n",
       "2          1328          Non-Essential         NaN          1.0        N   \n",
       "3           298          Non-Essential         NaN          1.0        N   \n",
       "4           238              Essential       cysA1          1.0        N   \n",
       "\n",
       "  AA_change  \n",
       "0     H445D  \n",
       "1     I156S  \n",
       "2     I443T  \n",
       "3     S100T  \n",
       "4      F80L  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SNP_variants_within_patients.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "#iterate through 'SNP_variants_within_patients' and store corresponding SNP information from 'change_in_AF_DF' in lists below\n",
    "allele_1_list = []\n",
    "allele_2_list = []\n",
    "alternate_AF_1 = []\n",
    "alternate_AF_2 = []\n",
    "change_in_alternate_AF = []\n",
    "\n",
    "for patient_id , ref_position in zip(list(SNP_variants_within_patients.patient_id) , list(SNP_variants_within_patients.ref_position)):\n",
    "    \n",
    "    #create a filter to store the corresponding SNP from the 'change_in_AF_DF'\n",
    "    SNP_filter = [(patient_id_i == patient_id) and (ref_position_i == ref_position) for patient_id_i , ref_position_i in zip(list(change_in_AF_DF.patient_id) , list(change_in_AF_DF.ref_position))]\n",
    "    first_second_allele_info_for_SNP = change_in_AF_DF[SNP_filter]\n",
    "    \n",
    "    #store information for this SNP in lists\n",
    "    allele_1_list.append(change_in_AF_DF[SNP_filter].allele_1.values[0])\n",
    "    allele_2_list.append(change_in_AF_DF[SNP_filter].allele_2.values[0])\n",
    "    alternate_AF_1.append(change_in_AF_DF[SNP_filter].alt_AF_1.values[0])\n",
    "    alternate_AF_2.append(change_in_AF_DF[SNP_filter].alt_AF_2.values[0])\n",
    "    change_in_alternate_AF.append(change_in_AF_DF[SNP_filter].altAF2_minus_altAF1.values[0])\n",
    "    \n",
    "#store lists as columns in master DF\n",
    "SNP_variants_within_patients['allele_1'] = allele_1_list\n",
    "SNP_variants_within_patients['allele_2'] = allele_2_list\n",
    "SNP_variants_within_patients['alt_AF_1'] = alternate_AF_1\n",
    "SNP_variants_within_patients['alt_AF_2'] = alternate_AF_2\n",
    "SNP_variants_within_patients['altAF2_minus_altAF1'] = change_in_alternate_AF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>population</th>\n",
       "      <th>patient_id</th>\n",
       "      <th>ref_position</th>\n",
       "      <th>ref_allele</th>\n",
       "      <th>alt_allele</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>genomic_coord</th>\n",
       "      <th>gene_category</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>alt_AF_diff</th>\n",
       "      <th>SNP_type</th>\n",
       "      <th>AA_change</th>\n",
       "      <th>allele_1</th>\n",
       "      <th>allele_2</th>\n",
       "      <th>alt_AF_1</th>\n",
       "      <th>alt_AF_2</th>\n",
       "      <th>altAF2_minus_altAF1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CASALI</td>\n",
       "      <td>P251</td>\n",
       "      <td>761139</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>Rv0667</td>\n",
       "      <td>1333</td>\n",
       "      <td>Antibiotic Resistance</td>\n",
       "      <td>rpoB</td>\n",
       "      <td>1.0</td>\n",
       "      <td>N</td>\n",
       "      <td>H445D</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CASALI</td>\n",
       "      <td>P251</td>\n",
       "      <td>861378</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>Rv0768</td>\n",
       "      <td>467</td>\n",
       "      <td>Non-Essential</td>\n",
       "      <td>aldA</td>\n",
       "      <td>1.0</td>\n",
       "      <td>N</td>\n",
       "      <td>I156S</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BRYANT</td>\n",
       "      <td>2020E</td>\n",
       "      <td>1253207</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>Rv1129c</td>\n",
       "      <td>1328</td>\n",
       "      <td>Non-Essential</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>N</td>\n",
       "      <td>I443T</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BRYANT</td>\n",
       "      <td>2020E</td>\n",
       "      <td>2075138</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>Rv1830</td>\n",
       "      <td>298</td>\n",
       "      <td>Non-Essential</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>N</td>\n",
       "      <td>S100T</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BRYANT</td>\n",
       "      <td>2020E</td>\n",
       "      <td>2694727</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>Rv2397c</td>\n",
       "      <td>238</td>\n",
       "      <td>Essential</td>\n",
       "      <td>cysA1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>N</td>\n",
       "      <td>F80L</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  population patient_id  ref_position ref_allele alt_allele  gene_id  \\\n",
       "0     CASALI       P251        761139          C          G   Rv0667   \n",
       "1     CASALI       P251        861378          T          G   Rv0768   \n",
       "2     BRYANT      2020E       1253207          A          G  Rv1129c   \n",
       "3     BRYANT      2020E       2075138          T          A   Rv1830   \n",
       "4     BRYANT      2020E       2694727          A          G  Rv2397c   \n",
       "\n",
       "  genomic_coord          gene_category gene_symbol  alt_AF_diff SNP_type  \\\n",
       "0          1333  Antibiotic Resistance        rpoB          1.0        N   \n",
       "1           467          Non-Essential        aldA          1.0        N   \n",
       "2          1328          Non-Essential         NaN          1.0        N   \n",
       "3           298          Non-Essential         NaN          1.0        N   \n",
       "4           238              Essential       cysA1          1.0        N   \n",
       "\n",
       "  AA_change allele_1 allele_2  alt_AF_1  alt_AF_2  altAF2_minus_altAF1  \n",
       "0     H445D        C        G       0.0       1.0                  1.0  \n",
       "1     I156S        T        G       0.0       1.0                  1.0  \n",
       "2     I443T        A        G       0.0       1.0                  1.0  \n",
       "3     S100T        T        A       0.0       1.0                  1.0  \n",
       "4      F80L        A        G       0.0       1.0                  1.0  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SNP_variants_within_patients.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Store newly annotated DF for downstream analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "SNP_variants_within_patients.to_pickle('/n/data1/hms/dbmi/farhat/Roger/inhost_TB_dynamics_project/pickled_files/variant_calling/longitudinal_SNPs/SNPs_between_isolates_delta_70_intergenic_annotated_with_allele1_allele2_info.pkl')\n",
    "SNP_variants_within_patients.to_csv('/n/data1/hms/dbmi/farhat/Roger/inhost_TB_dynamics_project/CSV_files/variant_calling/longitudinal_SNPs/SNPs_between_isolates_delta_70_intergenic_annotated_with_allele1_allele2_info.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Reference Allele -> Alternate Allele"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Alternate Allele Frequency increasing: **Alternate allele rising** in frequency"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "142"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sum( SNP_variants_within_patients.altAF2_minus_altAF1 > 0.0 )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Reference Allele is the first base: **Reference allele falling** in frequency"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "142"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sum( SNP_variants_within_patients.ref_allele == SNP_variants_within_patients.allele_1 )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>population</th>\n",
       "      <th>patient_id</th>\n",
       "      <th>ref_position</th>\n",
       "      <th>ref_allele</th>\n",
       "      <th>alt_allele</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>genomic_coord</th>\n",
       "      <th>gene_category</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>alt_AF_diff</th>\n",
       "      <th>SNP_type</th>\n",
       "      <th>AA_change</th>\n",
       "      <th>allele_1</th>\n",
       "      <th>allele_2</th>\n",
       "      <th>alt_AF_1</th>\n",
       "      <th>alt_AF_2</th>\n",
       "      <th>altAF2_minus_altAF1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CASALI</td>\n",
       "      <td>P251</td>\n",
       "      <td>761139</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>Rv0667</td>\n",
       "      <td>1333</td>\n",
       "      <td>Antibiotic Resistance</td>\n",
       "      <td>rpoB</td>\n",
       "      <td>1.0</td>\n",
       "      <td>N</td>\n",
       "      <td>H445D</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CASALI</td>\n",
       "      <td>P251</td>\n",
       "      <td>861378</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>Rv0768</td>\n",
       "      <td>467</td>\n",
       "      <td>Non-Essential</td>\n",
       "      <td>aldA</td>\n",
       "      <td>1.0</td>\n",
       "      <td>N</td>\n",
       "      <td>I156S</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BRYANT</td>\n",
       "      <td>2020E</td>\n",
       "      <td>1253207</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>Rv1129c</td>\n",
       "      <td>1328</td>\n",
       "      <td>Non-Essential</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>N</td>\n",
       "      <td>I443T</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BRYANT</td>\n",
       "      <td>2020E</td>\n",
       "      <td>2075138</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>Rv1830</td>\n",
       "      <td>298</td>\n",
       "      <td>Non-Essential</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>N</td>\n",
       "      <td>S100T</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BRYANT</td>\n",
       "      <td>2020E</td>\n",
       "      <td>2694727</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>Rv2397c</td>\n",
       "      <td>238</td>\n",
       "      <td>Essential</td>\n",
       "      <td>cysA1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>N</td>\n",
       "      <td>F80L</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  population patient_id  ref_position ref_allele alt_allele  gene_id  \\\n",
       "0     CASALI       P251        761139          C          G   Rv0667   \n",
       "1     CASALI       P251        861378          T          G   Rv0768   \n",
       "2     BRYANT      2020E       1253207          A          G  Rv1129c   \n",
       "3     BRYANT      2020E       2075138          T          A   Rv1830   \n",
       "4     BRYANT      2020E       2694727          A          G  Rv2397c   \n",
       "\n",
       "  genomic_coord          gene_category gene_symbol  alt_AF_diff SNP_type  \\\n",
       "0          1333  Antibiotic Resistance        rpoB          1.0        N   \n",
       "1           467          Non-Essential        aldA          1.0        N   \n",
       "2          1328          Non-Essential         NaN          1.0        N   \n",
       "3           298          Non-Essential         NaN          1.0        N   \n",
       "4           238              Essential       cysA1          1.0        N   \n",
       "\n",
       "  AA_change allele_1 allele_2  alt_AF_1  alt_AF_2  altAF2_minus_altAF1  \n",
       "0     H445D        C        G       0.0       1.0                  1.0  \n",
       "1     I156S        T        G       0.0       1.0                  1.0  \n",
       "2     I443T        A        G       0.0       1.0                  1.0  \n",
       "3     S100T        T        A       0.0       1.0                  1.0  \n",
       "4      F80L        A        G       0.0       1.0                  1.0  "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SNP_variants_within_patients[SNP_variants_within_patients.altAF2_minus_altAF1 > 0.0].head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Alternate Allele -> Reference Allele"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Reference Allele Frequency increasing: **Reference allele rising** in frequency"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "33"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sum( SNP_variants_within_patients.altAF2_minus_altAF1 < 0.0 )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Alternate Allele is the first base: **Alternate allele falling** in frequency"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "33"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sum( SNP_variants_within_patients.ref_allele == SNP_variants_within_patients.allele_2 )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>population</th>\n",
       "      <th>patient_id</th>\n",
       "      <th>ref_position</th>\n",
       "      <th>ref_allele</th>\n",
       "      <th>alt_allele</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>genomic_coord</th>\n",
       "      <th>gene_category</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>alt_AF_diff</th>\n",
       "      <th>SNP_type</th>\n",
       "      <th>AA_change</th>\n",
       "      <th>allele_1</th>\n",
       "      <th>allele_2</th>\n",
       "      <th>alt_AF_1</th>\n",
       "      <th>alt_AF_2</th>\n",
       "      <th>altAF2_minus_altAF1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>CETR</td>\n",
       "      <td>2918</td>\n",
       "      <td>866109</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>Rv0773c</td>\n",
       "      <td>1281</td>\n",
       "      <td>Antigen</td>\n",
       "      <td>ggtA</td>\n",
       "      <td>0.86</td>\n",
       "      <td>S</td>\n",
       "      <td>L427L</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.86</td>\n",
       "      <td>0.00</td>\n",
       "      <td>-0.86</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>TRAUNER</td>\n",
       "      <td>P09</td>\n",
       "      <td>1013542</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>Rv0908</td>\n",
       "      <td>1812</td>\n",
       "      <td>Non-Essential</td>\n",
       "      <td>ctpE</td>\n",
       "      <td>1.00</td>\n",
       "      <td>S</td>\n",
       "      <td>S604S</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>-1.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>WALKER</td>\n",
       "      <td>P000155</td>\n",
       "      <td>2395250</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>Rv2134c_Rv2135c</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0.96</td>\n",
       "      <td>I</td>\n",
       "      <td>None</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.96</td>\n",
       "      <td>0.00</td>\n",
       "      <td>-0.96</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>WALKER</td>\n",
       "      <td>P000156</td>\n",
       "      <td>3270363</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>Rv2935</td>\n",
       "      <td>2627</td>\n",
       "      <td>Non-Essential</td>\n",
       "      <td>ppsE</td>\n",
       "      <td>0.98</td>\n",
       "      <td>N</td>\n",
       "      <td>A876E</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>0.98</td>\n",
       "      <td>0.00</td>\n",
       "      <td>-0.98</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>WALKER</td>\n",
       "      <td>P000095</td>\n",
       "      <td>2123169</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>Rv1872c_Rv1873</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0.94</td>\n",
       "      <td>I</td>\n",
       "      <td>None</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.06</td>\n",
       "      <td>-0.94</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   population patient_id  ref_position ref_allele alt_allele          gene_id  \\\n",
       "16       CETR       2918        866109          C          T          Rv0773c   \n",
       "20    TRAUNER        P09       1013542          C          G           Rv0908   \n",
       "28     WALKER    P000155       2395250          A          G  Rv2134c_Rv2135c   \n",
       "33     WALKER    P000156       3270363          C          A           Rv2935   \n",
       "35     WALKER    P000095       2123169          T          G   Rv1872c_Rv1873   \n",
       "\n",
       "   genomic_coord  gene_category gene_symbol  alt_AF_diff SNP_type AA_change  \\\n",
       "16          1281        Antigen        ggtA         0.86        S     L427L   \n",
       "20          1812  Non-Essential        ctpE         1.00        S     S604S   \n",
       "28          None           None        None         0.96        I      None   \n",
       "33          2627  Non-Essential        ppsE         0.98        N     A876E   \n",
       "35          None           None        None         0.94        I      None   \n",
       "\n",
       "   allele_1 allele_2  alt_AF_1  alt_AF_2  altAF2_minus_altAF1  \n",
       "16        T        C      0.86      0.00                -0.86  \n",
       "20        G        C      1.00      0.00                -1.00  \n",
       "28        G        A      0.96      0.00                -0.96  \n",
       "33        A        C      0.98      0.00                -0.98  \n",
       "35        G        T      1.00      0.06                -0.94  "
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SNP_variants_within_patients[SNP_variants_within_patients.altAF2_minus_altAF1 < 0.0].head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Spectrum of Mutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "first_base_list = []\n",
    "second_base_list = []\n",
    "beginning_alternate_AF = []\n",
    "ending_alternate_AF = []\n",
    "change_in_alternate_AF = []\n",
    "Gene_Category = []\n",
    "Gene_IDs = []\n",
    "\n",
    "#for each pair of Base Calls, store (alternate Allele Frequency of isolate_1) and (alternate Allele Frequency of isolate_2 - alternate Allele Frequency of isolate_1)\n",
    "for Base_Call_A_index , Base_Call_B_index in zip( range(0 , np.shape(within_host_Base_Call_variants_all_patients_sample_order)[0] , 2) , range(1 , np.shape(within_host_Base_Call_variants_all_patients_sample_order)[0] , 2) ):\n",
    "    \n",
    "    #find the Base Call index that corresponds to the earlier isolate\n",
    "    if within_host_Base_Call_variants_all_patients_sample_order.loc[Base_Call_A_index , 'sample_order'] == 1:\n",
    "        \n",
    "        Base_Call_isolate_1 = within_host_Base_Call_variants_all_patients_sample_order.loc[Base_Call_A_index , :]\n",
    "        Base_Call_isolate_2 = within_host_Base_Call_variants_all_patients_sample_order.loc[Base_Call_B_index , :] \n",
    "\n",
    "        #if the alternate allele frequency increased, then population is moving toward alternate base\n",
    "        if Base_Call_isolate_2.alt_AF > Base_Call_isolate_1.alt_AF:\n",
    "            first_base = Base_Call_isolate_1.ref_base #base that was present in the population\n",
    "            second_base = Base_Call_isolate_2.alt_base #base that population is moving toward\n",
    "            \n",
    "        #if the alternate allele frequency decreased, then popolation is moving toward reference base\n",
    "        elif Base_Call_isolate_2.alt_AF < Base_Call_isolate_1.alt_AF:\n",
    "            first_base = Base_Call_isolate_1.alt_base\n",
    "            second_base = Base_Call_isolate_2.ref_base\n",
    "            \n",
    "    elif within_host_Base_Call_variants_all_patients_sample_order.loc[Base_Call_B_index , 'sample_order'] == 1:\n",
    "        \n",
    "        Base_Call_isolate_1 = within_host_Base_Call_variants_all_patients_sample_order.loc[Base_Call_B_index , :]\n",
    "        Base_Call_isolate_2 = within_host_Base_Call_variants_all_patients_sample_order.loc[Base_Call_A_index , :] \n",
    "        \n",
    "        if Base_Call_isolate_2.alt_AF > Base_Call_isolate_1.alt_AF:\n",
    "            first_base = Base_Call_isolate_1.ref_base\n",
    "            second_base = Base_Call_isolate_2.alt_base\n",
    "            \n",
    "        elif Base_Call_isolate_2.alt_AF < Base_Call_isolate_1.alt_AF:\n",
    "            first_base = Base_Call_isolate_1.alt_base\n",
    "            second_base = Base_Call_isolate_2.ref_base\n",
    "            \n",
    "    #store the base that population is moving from\n",
    "    first_base_list.append(first_base)\n",
    "    \n",
    "    #store the base that population is moving toward\n",
    "    second_base_list.append(second_base)\n",
    "        \n",
    "    #store the Alterante Allele Frequency of the first serial isolate\n",
    "    beginning_alternate_AF.append(Base_Call_isolate_1.alt_AF)\n",
    "    \n",
    "    #stotre the Alternate Allele Frequency of the second serial isolate\n",
    "    ending_alternate_AF.append(Base_Call_isolate_2.alt_AF)\n",
    "    \n",
    "    #store the difference of the Alternate Allele Frequencies of both serial isolates\n",
    "    change_in_alternate_AF.append( abs(Base_Call_isolate_2.alt_AF - Base_Call_isolate_1.alt_AF) ) \n",
    "    \n",
    "    #store Gene Category & gene_id of the gene that Reference Position is located on\n",
    "    Gene_Category.append(Base_Call_isolate_1.gene_category)\n",
    "    Gene_IDs.append(Base_Call_isolate_1.gene_id)\n",
    "    \n",
    "#create DataFrame with all Information\n",
    "change_in_AF_DF = pd.DataFrame()\n",
    "change_in_AF_DF['first_base'] = first_base_list\n",
    "change_in_AF_DF['second_base'] = second_base_list\n",
    "change_in_AF_DF['beginning_alt_AF'] = beginning_alternate_AF\n",
    "change_in_AF_DF['ending_alt_AF'] = ending_alternate_AF\n",
    "change_in_AF_DF['change_alt_AF'] = change_in_alternate_AF\n",
    "change_in_AF_DF['Gene_Category'] = Gene_Category\n",
    "change_in_AF_DF['gene_id'] = Gene_IDs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>first_base</th>\n",
       "      <th>second_base</th>\n",
       "      <th>beginning_alt_AF</th>\n",
       "      <th>ending_alt_AF</th>\n",
       "      <th>change_alt_AF</th>\n",
       "      <th>Gene_Category</th>\n",
       "      <th>gene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Antibiotic Resistance</td>\n",
       "      <td>Rv0667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Non-Essential</td>\n",
       "      <td>Rv0768</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Non-Essential</td>\n",
       "      <td>Rv1129c</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Non-Essential</td>\n",
       "      <td>Rv1830</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Essential</td>\n",
       "      <td>Rv2397c</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  first_base second_base  beginning_alt_AF  ending_alt_AF  change_alt_AF  \\\n",
       "0          C           G               0.0            1.0            1.0   \n",
       "1          T           G               0.0            1.0            1.0   \n",
       "2          A           G               0.0            1.0            1.0   \n",
       "3          T           A               0.0            1.0            1.0   \n",
       "4          A           G               0.0            1.0            1.0   \n",
       "\n",
       "           Gene_Category  gene_id  \n",
       "0  Antibiotic Resistance   Rv0667  \n",
       "1          Non-Essential   Rv0768  \n",
       "2          Non-Essential  Rv1129c  \n",
       "3          Non-Essential   Rv1830  \n",
       "4              Essential  Rv2397c  "
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "change_in_AF_DF.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Organize Mutations by Type"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "#counts of mutations\n",
    "\n",
    "#Transversions\n",
    "AT_TA = 0.0\n",
    "AT_CG = 0.0\n",
    "GC_CG = 0.0\n",
    "GC_TA = 0.0\n",
    "\n",
    "#Transitions\n",
    "AT_GC = 0.0\n",
    "GC_AT = 0.0\n",
    "\n",
    "#iterate through mutations and categorize into appropriate category\n",
    "for SNP_i in change_in_AF_DF.index:\n",
    "    \n",
    "    SNP_info = change_in_AF_DF.loc[SNP_i , :]\n",
    "    \n",
    "    if ((SNP_info.first_base == 'A') and (SNP_info.second_base == 'T')) or ((SNP_info.first_base == 'T') and (SNP_info.second_base == 'A')):\n",
    "        AT_TA += 1.0\n",
    "        \n",
    "    elif ((SNP_info.first_base == 'A') and (SNP_info.second_base == 'C')) or ((SNP_info.first_base == 'T') and (SNP_info.second_base == 'G')):\n",
    "        AT_CG += 1.0\n",
    "        \n",
    "    elif ((SNP_info.first_base == 'G') and (SNP_info.second_base == 'C')) or ((SNP_info.first_base == 'C') and (SNP_info.second_base == 'G')):\n",
    "        GC_CG += 1.0\n",
    "        \n",
    "    elif ((SNP_info.first_base == 'G') and (SNP_info.second_base == 'T')) or ((SNP_info.first_base == 'C') and (SNP_info.second_base == 'A')):\n",
    "        GC_TA += 1.0\n",
    "        \n",
    "    elif ((SNP_info.first_base == 'A') and (SNP_info.second_base == 'G')) or ((SNP_info.first_base == 'T') and (SNP_info.second_base == 'C')):\n",
    "        AT_GC += 1.0\n",
    "        \n",
    "    elif ((SNP_info.first_base == 'G') and (SNP_info.second_base == 'A')) or ((SNP_info.first_base == 'C') and (SNP_info.second_base == 'T')):\n",
    "        GC_AT += 1.0\n",
    "        \n",
    "num_SNPs = float(AT_TA + AT_CG + GC_CG + GC_TA + AT_GC + GC_AT)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Plot Mutations by Type"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/rv76/anaconda2/lib/python2.7/site-packages/matplotlib/cbook/deprecation.py:107: MatplotlibDeprecationWarning: Passing one of 'on', 'true', 'off', 'false' as a boolean is deprecated; use an actual boolean (True/False) instead.\n",
      "  warnings.warn(message, mplDeprecation, stacklevel=1)\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYAAAAE4CAYAAABSXiSnAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMi4zLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvIxREBQAAIABJREFUeJzt3XtUVOX+P/D3yC06CmIqDChw0iWYGpiI4cRRQ8bDJccsBbFWlncjs9SVRyw10FDo4C1SzJOeo+AlyvGSIiZo3lHR9IhiwhGVGa9gpggK+/dHX+fnyMWtzN4J+/1aq7XYez97ns+z9jRv92WeUQmCIICIiBSnyZ9dABER/TkYAERECsUAICJSKAYAEZFCMQCIiBSKAUBEpFAMACIihWIAEBEpFAOAiEihGABERArFACAiUigGABGRQjX6AFCpVH92CURET6VGHwBERFQzBgARkUIxAIiIFIoBQESkUAwAIiKFYgAQESkUA4CISKEYAERECsUAICJSKAYAEZFCMQCIiBRKtgDIyspCYGAgNBoNFi1aVG37v//9bwQFBSE4OBgDBgxAfn6+advChQuh0WgQGBiI7OxsuUomImrUVIIgCFJ3UllZicDAQKSlpUGtViM0NBTJycno0KGDqc3NmzfRrFkzAMC2bduwYsUKrFq1Cvn5+Rg3bhw2b96MS5cuITIyEj///DOsrKxE9a1SqSDDEIlIAfz8/GAwGCTvR61W49ChQ5L3Yy15DwByc3Ph6ekJDw8PAIBOp0NGRoZZANz/8AeA27dvm2bxzMjIgE6ng52dHdzd3eHp6Ync3Fz4+fnJUToRkYnBYEBiYqLk/UyaNEnyPgCZAsBoNMLV1dW0rFarkZubW63d8uXLkZKSgoqKCqxdu9a070svvWS2r9FolL5oIqJGTpZ7ADVdgqlpnv5hw4Zh7969iImJwfz58x9rXyIiejyyBIBarUZxcbFp2WAwwNnZudb29y8RPcm+REQkjiwB4Ovri8LCQhQVFaGiogJ6vR5ardasTUFBgenv7du3469//SsAQKvVQq/Xo7y8HEVFRSgsLETXrl3lKJuIqFGT5R6AtbU14uLiEBUVhaqqKkRERMDLywsJCQnw8fGBVqvF8uXL8fPPP8Pa2hqOjo6YN28eAMDLywuvvfYa+vTpAysrK8yaNUv0E0BERFQ7WR4D/TPxMVAishQ3NzfZngK6ePGi5P3wm8BERArFACAiUigGABGRQjEAiIgUigFARKRQDAAiIoViABARKRQDgIhIoRgAREQKxQAgIlIoBgARkUIxAIiIFIoBQESkUAwAIiKFYgAQESkUA4CISKEYAERECsUAICJSKAYAEZFCMQCIiBSKAUBEpFAMACIihWIAEBEpFAOAiEihGABERArFACAiUigGABGRQjEAiIgUigFARKRQDAAiIoViABARKRQDgIhIoRgAREQKJVsAZGVlITAwEBqNBosWLaq2fcmSJejduzf69u2LwYMH48KFC6Ztbdu2RXBwMIKDgzFs2DC5SiYiatSs5eiksrISMTExSEtLg1qtRmhoKLRaLTp06GBq07lzZ2zZsgX29vZYsWIF4uLisHjxYgDAM888g8zMTDlKJSJSDFnOAHJzc+Hp6QkPDw/Y2tpCp9MhIyPDrI1Go4G9vT0AoFu3bjAYDHKURkSkWLIEgNFohKurq2lZrVbDaDTW2j4tLQ19+vQxLZeXlyMkJATh4eHYunWrpLUSESmFLJeABEGotk6lUtXYNj09HceOHUN6erpp3cGDB+Hi4oJz585h8ODB8Pb2hqenp1TlEhEpgqgzgCVLluDEiRMAgMOHD6N79+4ICAjAoUOHRHWiVqtRXFxsWjYYDHB2dq7WbteuXViwYAGWL18OOzs703oXFxcAgIeHBwICAky1EBHRkxMVAEuXLoW7uzsA4IsvvsCoUaMwfvx4zJgxQ1Qnvr6+KCwsRFFRESoqKqDX66HVas3anDhxAlOmTMG3336Lli1bmtaXlpaivLwcAHD9+nXk5OSY3TwmIqInI+oS0M2bN+Hg4IDff/8deXl5WLNmDaysrPD555+L68TaGnFxcYiKikJVVRUiIiLg5eWFhIQE+Pj4QKvVIjY2Frdu3cLo0aMBAG5ubli+fDnOnDmDKVOmQKVSQRAEREdHMwCIGgA/Pz/ZHuZQq9Wir0jQ/ycqAFxdXZGTk4P8/Hz06NEDVlZWuHnzJpo0EX8POSgoCEFBQWbrJk+ebPp7zZo1Ne7XvXt3/PTTT6L7IaKng8FgQGJioix9TZo0SZZ+GhtRATBt2jSMHj0aNjY2WLp0KQBg+/bt6Nq1q6TFERGRdEQFQFBQEI4cOWK2Ljw8HOHh4ZIURURE0qvzGs6VK1fw/vvvo2/fvpgwYQKuX79u2mZjYwMbGxvJCyQiImnUGQBTp07FlStX8NZbb6G4uBjTp0+Xqy4iIpJYnZeADhw4gJ9//hmOjo7o379/tUc3iYio4arzDKC8vByOjo4AgBYtWqCsrEyWooiISHp1ngHcu3cPa9asMU3lUF5ejtWrV5u1iYyMlK46IiKSTJ0B0LVrV6xbt8607Ovri++++860rFKpGABERA1UnQHw4Ic9ERE1LvxJSCIiharzDKBNmza1TtsM/HEJqKioyOJFERGR9OoMgN27d1dbJwgCtm7diq+++qrGKZ2JiKhhqDMAHv7RlezsbCQkJOC3335DbGwsBgwYIGVtREQkIVFzAR04cADx8fG4ePEiJkyYgIiICFhZWUldGxERSajOADh27Bjmzp2LvLw8fPDBBxg6dChsbW3lqo2IiCRUZwCEhYXByckJgwYNwtWrVzF//vxqbR6c05+IiBqOOgPgzTffhEqlQklJCUpKSqptr+sJISIierrVGQDz5s2Tqw4iIpJZnQFw+/ZtAMCzzz4L4I9HQFNTU3H69Gl069YNOp1O+gqJiEgSdX4TeOzYsfjxxx9Ny59//jlmz54No9GITz/9FIsXL5a8QCIikkadAfDLL78gODgYAFBRUYHU1FSkpKQgJSUFK1asQGpqqixFEhGR5dUZAGVlZabfAzh27Bisra2h0WgA/DFT6OXLl6WvkIiIJFFnADg7O+PkyZMAgF27dsHf39+07caNG/xOABFRA1bnTeAxY8YgKioKfn5+yM7OxjfffGPalp2djY4dO0peIBERSaPOABgyZAg8PT1x7NgxjBo1yuwMwN7eHh9//LHkBRIRkTQeORdQQEAAAgICqq3nD8QTETVs/EEYIiKFYgAQESkUA4CISKFEBcCnn35a4/rPPvvMosUQEZF8RAXA2rVra1yfnp5u0WKIiEg+dT4FtHr1agDAvXv3TH/fd+7cObRo0UK6yoiISFJ1BsB3330HALh7967pb+CP3wFo1aoVp4smImrARAXAnDlz8Mknn8hSEBERyUPUPYARI0bg1q1bAIDKykqsWbMG69atQ1VVleiOsrKyEBgYCI1Gg0WLFlXbvmTJEvTu3Rt9+/bF4MGDceHCBdO2tWvXQqPRQKPR1Ho/goiIHo+oAHjnnXdQWFgIAIiPj8fixYuRkpKCmTNniuqksrISMTExWLlyJbKysrB+/Xrk5+ebtencuTO2bNmC7du3IywsDHFxcQCAkpISJCUlYdOmTdi8eTOSkpJQWlr6OGMkIqIaiAqAgoICdOrUCQDw/fffY+XKlVi7di02bNggqpPc3Fx4enrCw8MDtra20Ol0yMjIMGuj0Whgb28PAOjWrRsMBgMAYOfOnQgMDISTkxOaN2+OwMBAZGdnix0fERHVQlQANGnSBBUVFcjLy4ODgwPc3Nzg6Ohouiz0KEajEa6urqZltVoNo9FYa/u0tDT06dPnifYlIiJxHjkZHAD06dMHY8aMQUlJCfr37w8AyM/Ph4uLi6hOBEGotk6lUtXYNj09HceOHTN9x6CmfYmIqP5EBUBiYiLWrVsHGxsbvPHGGwCA69evY+LEiaI6UavVKC4uNi0bDAY4OztXa7dr1y4sWLAA6enpsLOzM+27d+9es3179uwpql8iIqqdqEtAdnZ2eOuttzBo0CBcu3YNVVVV6NmzJ3Q6nahOfH19UVhYiKKiIlRUVECv11ebTvrEiROYMmUKvv32W7Rs2dK0vlevXti1axdKS0tRWlqKXbt2oVevXo8xRCIiqomoM4CbN29i2rRp2LBhA+7evQsbGxv0798fsbGxcHBweHQn1taIi4tDVFQUqqqqEBERAS8vLyQkJMDHxwdarRaxsbG4desWRo8eDQBwc3PD8uXL4eTkhAkTJiAsLAwA8NFHH8HJyakeQyYiIgBQCSIusk+YMAG3bt3ClClT0KZNG1y4cAFz5syBvb095s+fL0edT0ylUvE+AtGfwM3NDYmJibL0NWnSJFy8eFHyfuQak1zjEXUGkJ2djX379pke02zXrh2SkpJ4LZ6IqAETfQ/g2rVrZuuuX78OW1tbSYoiIiLpiToDGDJkCCIjIzFq1CjTJaClS5di6NChUtdHREQSERUAH374IVxcXPDDDz/g0qVLcHZ2xrhx4xAZGSl1fUREJBFRAaBSqRAZGckPfCKiRkRUAAB/zMnz3//+t9r0D5MnT7Z4UUREJD1RARATE4ONGzeiZ8+epieBgNqncyAioqefqABYv349tm3bBjc3N6nrISIimYh6DNTJyQmOjo5S10JERDKq9Qzg3Llzpr9Hjx6N6OhoREdHo1WrVmbtPDw8pKuOiIgkU2sAaDSaatMobN++3ayNSqXC+fPnpauOiIgkU2sAPPibvERE1PiIugfwoJycHCnqICIimT12ALz11ltS1EFERDJ77ADg1MpERI3DYwdAmzZtpKiDiIhk9tgBsGPHDinqICIimXEuICIiheJcQERECsW5gBTCz88PBoNB8n7UajUOHTokeT9EVH+iAoBzATV8BoNBth+zJqKGQVQAcC4gIqLGR1QA/OMf/wDAuYCIiBoTUQHAeYGIiBqfx/4eABERNQ61ngEMHToUq1atAgC8/vrrtT7y+f3330tTGRERSarWAHjzzTdNfw8ZMkSWYoiISD61BsDrr79u+nvw4MGyFENERPLhPQAiIoViABARKRQDgIhIoWoNgPDwcNPf//znP2UphoiI5FNrABQWFuLOnTsAgCVLlshWEBERyaPWp4C0Wi0CAwPRtm1b3LlzBwMHDqyxHb8HQETUMNUaAElJSTh48CDOnz+PY8eOITIysl4dZWVl4bPPPkNVVRWGDBmC6Ohos+379+/H9OnTkZeXh+TkZLNLUG3btoW3tzcAwM3NDcuXL69XLdTwcXprovqrcy4gf39/+Pv74+7du/X6LkBlZSViYmKQlpYGtVqN0NBQaLVadOjQwdTGzc0NSUlJWLx4cbX9n3nmGWRmZj5x/9T4cHprovoTNRlcZGQk9uzZg/T0dBiNRri4uGDgwIF45ZVXRHWSm5sLT09P09TROp0OGRkZZgHQtm1bAECTJnwwiYhIDqI+bVNTUzF27Fi0atUKISEhaN26NaKjo01zBT2K0WiEq6uraVmtVsNoNIousry8HCEhIQgPD8fWrVtF70dERLUTdQaQnJyMtLQ0dOrUybSuf//+GDlyJIYOHfrI/QVBqLbucX5P+ODBg3BxccG5c+cwePBgeHt7w9PTU/T+RERUnagzgJKSErPLNQDQrl07lJaWiupErVajuLjYtGwwGODs7Cy6SBcXFwB//PpYQEAATpw4IXpfIiKqmagA8Pf3x8yZM1FWVgYAuH37NuLi4uDn5yeqE19fXxQWFqKoqAgVFRXQ6/XQarWi9i0tLUV5eTkA4Pr168jJyakWRkRE9PhEXQKKj4/HuHHj4O3tjebNm6O0tBR+fn746quvxHVibY24uDhERUWhqqoKERER8PLyQkJCAnx8fKDVanH06FEMHz4cN27cQGZmJr788ktkZWXhzJkzmDJlClQqFQRBQHR0NAOAiMgCRAWAs7Mz0tPTUVxcjEuXLsHZ2dnspq4YQUFBCAoKMls3efJk09++vr44fPhwtf26d++On3766bH6IiKiRxMVAPe5uro+9gc/ERE9nfjQPRGRQjEAiIgUigFARKRQou8B/Pbbbzh79ixu3bpltl7sdBBERPR0ERUAa9asQUxMDP7yl7/A3t7etF6lUmHfvn2SFUdERNIRFQBz585FSkoKXn31VanrISIimYi6B3Dv3j306tVL6lqIiEhGogJg3LhxmD9/PqqqqqSuh4iIZCLqEtDSpUtx5coVJCcnw8nJyWxbTk6OJIUREZG0RAXAwoULpa6DiIhkJioAAgICpK6DiIhkJuoewN27d5GYmIiAgAA8//zzCAgIQGJiIioqKqSuj4iIJCLqDCAuLg5Hjx5FfHw83NzccPHiRcybNw83b97EzJkzpa6RiIgkICoANm3ahMzMTLRo0QIA0L59e3Tp0gXBwcEMACKiBqpecwHV9Fu/RETUMIgKgLCwMAwbNgzZ2dk4c+YMsrKy8N577yE8PFzq+oiISCKiLgFNmzYN8+fPR0xMDIxGI1xcXKDT6fDhhx9KXR8REUlEVADY2tpi8uTJZj/hSEREDVutAbB//368/PLLAIDdu3fX+gKcDpqIqGGqNQCmTp2KHTt2AAAmTZpUYxtOB01E1HDVGgD3P/yBP84GiIiocRH1FNC7775b4/oRI0ZYtBgiIpKPqADYu3dvjet5+YeIqOGq8ymghIQEAH/MBXT/7/uKiorg5uYmXWVERCSpOgOguLgYAFBVVWX6G/jj5q+rqys+/vhjaasjIiLJ1BkASUlJqKqqgp+fH958803Y2dnJVRcREUnskfcAmjRpgpkzZ/LDn4iokRF1E7hHjx44fPiw1LUQEZGMRE0F0aZNG7z99tvQarVwdXWFSqUybeP0EEREDZOoALhz5w769esHADAYDKb1DwYBERE1LKICICkpSeo6iIhIZqICAAAKCgqg1+thMBigVquh0+nw/PPPS1kbERFJSNRN4G3btiEkJAS//vornJyccPbsWYSGhmLbtm2iO8rKykJgYCA0Gg0WLVpUbfv+/fvRr18/uLu7Y9OmTWbb1q5dC41GA41Gg7Vr14ruk4iIaifqDGDOnDn417/+BY1GY1q3d+9eTJs2DVqt9pH7V1ZWIiYmBmlpaVCr1QgNDYVWq0WHDh1Mbdzc3JCUlITFixeb7VtSUoKkpCT8+OOPUKlUCAkJgVarRfPmzcWOkYiIaiDqDMBgMKBHjx5m6/z9/c1uCNclNzcXnp6e8PDwgK2tLXQ6HTIyMszatG3bFi+88AKaNDEvaefOnQgMDISTkxOaN2+OwMBAZGdni+qXiIhqJyoAXnjhBSxZssRsXUpKCjp16iSqE6PRCFdXV9OyWq2G0WiUfF8iIqqdqEtA8fHxeOedd7Bs2TK4urri4sWLePbZZ7F8+XJRnQiCUG2d2EdIa9qXiIjqT1QAtG/fHjt37sSRI0dMPwrftWtX2NjYiOpErVabTSZnMBjg7Owset8Hp6M2GAzo2bOnqH2JiKh2oh8DValUZv89fK2+Lr6+vigsLERRURFcXFyg1+vx1Vdfidq3V69eiI+PR2lpKQBg165d+Mc//iG6b6KGws/PT/R9tfpQq9U4dOiQ5P3Q009UAJw8eRLDhw9HRUUFXFxcYDAYYGdnh2+++UbUfQBra2vExcUhKioKVVVViIiIgJeXFxISEuDj4wOtVoujR49i+PDhuHHjBjIzM/Hll18iKysLTk5OmDBhAsLCwgAAH330EZycnOo3aqKnkMFgQGJiouT91PYb36Q8ogJg4sSJeOeddzB69GioVCoIgoCUlBRMnDgRW7duFdVRUFAQgoKCzNY9OI+Qr69vrRPORUZGIjIyUlQ/REQkjqjrOAUFBRg5cqTpxq1KpcKIESNQWFgoaXFERCQdUQHw6quvVvvWb2ZmZrV/0RMRUcMh6hJQZWUlxo0bhy5dusDV1RXFxcU4fvw4tFotxo8fb2q3YMECyQolIiLLEhUA3t7e8Pb2Ni136NABvXv3lqomIiKSgagA4I+/ExE1PqK/B7Bnzx6kp6ebvgg2cOBAvPLKK1LWRkREEhJ1Ezg1NRVjx45Fq1atEBISgtatWyM6OhqrVq2Suj4iIpKIqDOA5ORkpKWlmX3pq3///hg5ciSGDh0qWXFERCQdUWcAJSUlZnP3A0C7du1M0zMQEVHDIyoA/P39MXPmTJSVlQEAbt++jbi4OPj5+UlaHBERSUdUAMTHx+PUqVPw9vaGj48POnbsiJMnT2LOnDlS10dERBJ55D0AQRBw584drF69GpcvX8alS5fg7Oxs9iMtRETU8DwyAFQqFYKCgpCfnw9XV1d+8BMRNRKiLgF17twZBQUFUtdCREQyEvUYaEBAAIYOHYrBgwfD1dXV7OccOU0zEVHDJCoAcnJy0LZtW+zbt89svUqlYgAQETVQogLgu+++k7oOIiKSWZ0BUFZWhnnz5uH06dPo0qULoqOjYWdnJ1dtREQkoTpvAsfExGD79u1o3749Nm/ejNjYWLnqIiIiidUZAFlZWUhNTcW0adPwn//8B9u3b5erLiIiklidAXD79m04OzsDANzc3HDz5k1ZiiIiIunVeQ/g3r172LNnDwRBqHEZAH8TgIiogaozAFq2bImJEyealp2cnMyWVSpVtUdDiYioYagzAA4cOCBXHUREJDNRU0EQEVHjwwAgIlIoBgARkUIxAIiIFIoBQESkUAwAIiKFYgAQESkUA4CISKEYAERECsUAICJSKNkCICsrC4GBgdBoNFi0aFG17eXl5RgzZgw0Gg3Cw8Nx/vx5AMD58+fRrl07BAcHIzg4GJ988olcJRMRNWqifhKyviorKxETE4O0tDSo1WqEhoZCq9WiQ4cOpjZpaWlwdHTEnj17oNfrMWvWLCxevBgA4OHhgczMTDlKJSJSDFnOAHJzc+Hp6QkPDw/Y2tpCp9MhIyPDrM22bdswaNAgAEBYWBh2795tNu00ERFZliwBYDQa4erqalpWq9UwGo21trG2toaDgwNKSkoAAEVFRdBqtXjjjTc4QykRkYXIcgmopn/Jq1SqR7YBgNatW+PgwYNo0aIFfvnlF7z33nvIyspCs2bNJKmViEgpZDkDUKvVKC4uNi0bDAbTT03W1ObevXv47bff4OTkBDs7O7Ro0QIA8OKLL8LT0xMFBQVylE1E1KjJEgC+vr4oLCxEUVERKioqoNfrodVqzdpotVqsW7cOALB582ZoNBqoVCpcu3YNlZWVAIBz586hsLAQ7u7ucpRNRNSoyXIJyNraGnFxcYiKikJVVRUiIiLg5eWFhIQE+Pj4QKvVIjIyEuPHj4dGo0Hz5s2RnJwMANi/fz8SExNhZWUFKysrfPHFF3BycpKjbCKiRk2WAACAoKAgBAUFma2bPHmy6e9nnnkGKSkp1fYLCwtDWFiY5PURESkNvwlMRKRQDAAiIoViABARKRQDgIhIoRgAREQKxQAgIlIoBgARkUIxAIiIFIoBQESkUAwAIiKFYgAQESkUA4CISKEYAERECiXbbKANiZ+fHwwGg+T9qNVqHDp0SPJ+iIhqwgCogcFgQGJiouT9TJo0SfI+iIhqw0tAREQKxQAgIlIoBgARkUIxAIiIFIoBQESkUAwAIiKFYgAQESkUA4CISKEYAERECsUAICJSKAYAEZFCMQCIiBSKAUBEpFAMACIihWIAEBEpFAOAiEihGABERArFACAiUigGABGRQskWAFlZWQgMDIRGo8GiRYuqbS8vL8eYMWOg0WgQHh6O8+fPm7YtXLgQGo0GgYGByM7OlqtkIqJGTZYAqKysRExMDFauXImsrCysX78e+fn5Zm3S0tLg6OiIPXv2YOTIkZg1axYAID8/H3q9Hjt27MCqVaswdepUVFZWylE2EVGjJksA5ObmwtPTEx4eHrC1tYVOp0NGRoZZm23btmHQoEEAgLCwMOzevRuCICAjIwM6nQ52dnZwd3eHp6cncnNz5SibiKhRs5ajE6PRCFdXV9OyWq2u9iH+YBtra2s4ODigpKQERqMRL730ktm+RqPxsfpXqVSPXXNUVNRj7/MknqS2J9XYxtTYxgM0vjHJNR6g8Y3pScYjCMJjtZclAGoq6uHB1Va4mH0ft28iIpLpEpBarUZxcbFp2WAwwNnZudY29+7dw2+//QYnJydR+xIR0eOTJQB8fX1RWFiIoqIiVFRUQK/XQ6vVmrXRarVYt24dAGDz5s3QaDRQqVTQarXQ6/UoLy9HUVERCgsL0bVrVznKJiJq1GS5BGRtbY24uDhERUWhqqoKERER8PLyQkJCAnx8fKDVahEZGYnx48dDo9GgefPmSE5OBgB4eXnhtddeQ58+fWBlZYVZs2bByspKjrKJiBo1lcCL5EREisRvAhMRKRQDgIhIoRgAREQKxQAgIlIoBgARkULJ8hhoQ7NlyxaMGDECO3fuxN27dzF+/HgAQHFxMZo1a4ZmzZqhRYsWWLNmjdl+eXl5j2ybkpKC+Ph4HD16FA4ODk/1eO47e/YsZsyYgYKCAlhbW8Pb2xtxcXFo1aoVcnNzMWvWLBgMBjRt2hStW7fG1KlT0bFjR0nGcuXKFcyYMQNHjhyBo6MjbGxsMG7cOISEhCA3NxexsbG4cuUKVCoV/P39ERsbC3t7+xpfa8eOHUhISMDt27chCAL69u2Lzz77DACQnp6Or7/+GpWVlbCysoKvry8+/fRTODo6PpVjWrNmDb755hsAwJkzZ9CuXTs0adIEffr0wdSpUwEA7777Lq5evYqNGzdafAwPq+97rqCgADNmzMCvv/4KBwcHNG3aFJMmTcLLL78MoO5jJwVLvu+A6sdi/vz52LRpEwDg1KlT8Pb2BgBERkZi+PDhko0LAlUzatQoYcCAAUJiYqLZ+g8//FDYuHGjqNeorW1oaKgwYMAAYfXq1RapVYz6jKesrEzo2bOnkJGRYVq3e/duIS8vT7h8+bLg7+8vHDx40LTtwIEDwpYtWyw7gP9TVVUlhIeHCytWrDCtO3/+vLBs2TLh8uXLQvfu3YWcnBxT240bNwqXL1+u8bXy8vKEgIAA4cyZM4IgCMLdu3eFb7/9VhAEQdixY4fhvNAhAAAFIklEQVTQr18/obi4WBAEQbh3756QlpZmavu0juk+f39/4dq1a2brSktLhW7dugl/+9vfhHPnzll8HA+r73tOo9GYvefy8vJM/8/UdeykYOlj9Khj0b59e8sPohYMgIf8/vvvwksvvST8+uuvQmBgoNm2+gZAYWGh0KtXL2Hv3r3CkCFDLFZzXeo7nrS0NOGDDz6ocducOXOEhIQEi9X6KLt27RIGDhxY47a5c+cKc+fOFf1aH3zwgZCWllbjtgEDBgi7d+9+ohoflyXHdF9NAZCamipMmTJF+PLLL4UFCxY8Ua1i1fc9l5qaKowfP77W7XUdOylY+hg96ljIGQC8B/CQrVu3onfv3mjXrh2aN2+O48ePW+y1169fD51Ohx49euDs2bO4evWqxV67NvUdz6lTp/Diiy/WuC0/Px9dunSxRJmi5Ofno3PnzjVuO336dK111ta+ttrlHJclx1SX9evXY8CAAdDpdNDr9RZ5zdrU9z1X17ERs93SLH2M5DwWj8IAeMj9D2kA0Ol0WL9+vcVee8OGDdDpdGjSpAlCQkJM1/ykJOV4HhYeHo5evXpJei32QVOnTkXfvn0RGhoqWR95eXkIDg5Gz549ZfmfVYoxXblyBf/73//g7++Pdu3awcrKCqdOnbLY6z/M0u+54cOH49VXX8WIESMsUV691ecYyX0sHoU3gR9w/fp17N27F6dPn4ZKpUJlZSVUKhWmTZtW77nGT548icLCQgwZMgQAcPfuXbi7u2PYsGEWqLxmlhiPl5cX9u3bV+O2Dh064Pjx4+jXrx8AYNOmTdi0aRO2b99usTE83N+PP/5oWp49ezauX7+OkJAQ9O7dG7/88oupFjGvdfz4cXTq1KnWbRqNBh07dkRmZiZiYmJw584di43lwb4sNababNiwATdu3DDdQP3999+h1+tNNxotyVLvuf3795uWly1bhmPHjiE2NhZA3cdOCpY8RnIeCzF4BvCAzZs344033sDBgwdx4MABHDp0CO7u7jh48GC9X1uv1+Pjjz/GgQMHcODAARw5cgRGoxEXLlywQOU1s8R4BgwYgMOHD5t9qGdlZSEvLw/Dhg3D2rVrkZOTY9pWVlZm0TE86JVXXkF5eTlWrFhRrb93330X69atw5EjR0zb0tPTcfny5Rpfa+zYsVi4cCHOnj0LAKiqqsKSJUsAANHR0YiNjTWbhlyKD3/AsmOqzfr167Fy5UrTe2/Lli3YsGGDZQbwEEu95w4dOoRt27aZ1j34vqrr2EnBksdIzmMhBs8AHqDX6/H++++brQsNDcUPP/yAHj161Pu1V65cabbu73//e419WoolxmNvb48VK1Zg+vTpmD59OmxsbNCxY0d8/vnnaNWqFb7++mvMnj0bRqMRzz33HFq0aIGPPvpIiuFApVJh2bJlmDFjBr7++ms899xzsLe3x9SpU9GqVSskJycjNjYWV69eRZMmTdCjR49aT9NfeOEFzJgxA++//z7KysqgUqkQFBQEAAgKCsK1a9fw9ttvo7KyEg4ODvD29kbv3r2f6jHV5Pz58yguLka3bt1M69zd3dG0aVMcOXLE7Nf2LMGS77mZM2di+vTpaNmyJZo2bWp6lLSuYycFSx0juY+FGJwNlIhIoXgJiIhIoXgJ6Ak9+K3f++zs7GR5skcKjW089z34Ddn7unfvjtmzZ/9JFdVfYxlTY33PAQ3nGPESEBGRQvESEBGRQjEAiIgUigFARKRQDAAiIoViABARKdT/AwCZSvyXf1f6AAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 396x324 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#####################  set plot parameters  ####################\n",
    "plt.style.use('ggplot')\n",
    "plt.rcParams['lines.linewidth']=3\n",
    "plt.rcParams['axes.facecolor']='1.0'\n",
    "plt.rcParams['axes.grid']=True\n",
    "plt.rcParams['axes.edgecolor']='black'\n",
    "plt.rcParams['grid.color']= '1.0'\n",
    "plt.rcParams.update({'font.size': 12})\n",
    "plt.rcParams['xtick.color']='black'\n",
    "plt.rcParams['ytick.color']='black'\n",
    "\n",
    "#####################  Bar Plot of SNP types  ####################\n",
    "color = '0.65'\n",
    "\n",
    "fig , ax = plt.subplots()\n",
    "\n",
    "ax.bar([1] , [AT_TA/num_SNPs] ,   color = color , width = 0.5 , edgecolor ='k' , linewidth = 1.0)\n",
    "ax.bar([2] , [AT_CG/num_SNPs] ,   color = color , width = 0.5 , edgecolor ='k' , linewidth = 1.0)\n",
    "ax.bar([3] , [GC_CG/num_SNPs] ,   color = color , width = 0.5 , edgecolor ='k' , linewidth = 1.0)\n",
    "ax.bar([4] , [GC_TA/num_SNPs] ,   color = color , width = 0.5 , edgecolor ='k' , linewidth = 1.0)\n",
    "ax.bar([5] , [AT_GC/num_SNPs] ,   color = color , width = 0.5 , edgecolor ='k' , linewidth = 1.0)\n",
    "ax.bar([6] , [GC_AT/num_SNPs] ,   color = color , width = 0.5 , edgecolor ='k' , linewidth = 1.0)\n",
    "\n",
    "\n",
    "ax.yaxis.set_major_formatter(plt.ScalarFormatter())\n",
    "ax.tick_params(axis='y', which='major', labelsize=10 )\n",
    "\n",
    "ax.set_xlim([0.5 , 6.5])\n",
    "ax.set_ylabel('Proportion of in-host SNPs' , color = 'k' , fontsize = 12)\n",
    "#ax.set_xlabel('SNP type' , fontsize = 12 , color = 'k')\n",
    "\n",
    "#x-axis\n",
    "ax.set_xticks([1,2,3,4,5,6])\n",
    "ax.set_xticklabels([], rotation='horizontal', fontsize = 10)\n",
    "x_labels = ['AT_TA' , 'AT_CG' , 'GC_CG' , 'GC_TA' , 'AT_GC' , 'GC_AT']\n",
    "ax.set_xticklabels(x_labels, rotation='horizontal', fontsize = 10)\n",
    "\n",
    "\n",
    "ax.tick_params(\n",
    "    axis='x',          # changes apply to the x-axis\n",
    "    which='both',      # both major and minor ticks are affected\n",
    "    bottom='off',      # ticks along the bottom edge are off\n",
    "    top='off',         # ticks along the top edge are off\n",
    "    labelbottom='on') # labels along the bottom edge are off\n",
    "\n",
    "ax.tick_params(\n",
    "    axis='y',          # changes apply to the x-axis\n",
    "    which='both',      # both major and minor ticks are affected\n",
    "    left='off',      # ticks along the bottom edge are off\n",
    "    right='off',         # ticks along the top edge are off\n",
    "    labelleft='on') # labels along the bottom edge are off\n",
    "\n",
    "ax.spines['right'].set_visible(False)\n",
    "ax.spines['top'].set_visible(False)\n",
    "\n",
    "ax.tick_params(labelcolor = 'k')\n",
    "\n",
    "fig = plt.gcf()\n",
    "fig.set_size_inches(5.5, 4.5)\n",
    "fig.tight_layout()\n",
    "\n",
    "#output for Adobe Illustrator\n",
    "file_name = '/n/data1/hms/dbmi/farhat/Roger/inhost_TB_dynamics_project/figures/longitudinal_SNP_analysis/spectrum_of_mutations.pdf'\n",
    "plt.savefig(file_name, bbox_inches='tight', dpi = 300 , transparent = True)\n",
    "\n",
    "plt.show()\n",
    "##################################################################"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "### Compare $\\Delta AF$ between Syn, NSyn & Intergenic SNPs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>population</th>\n",
       "      <th>patient_id</th>\n",
       "      <th>ref_position</th>\n",
       "      <th>ref_allele</th>\n",
       "      <th>alt_allele</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>genomic_coord</th>\n",
       "      <th>gene_category</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>alt_AF_diff</th>\n",
       "      <th>SNP_type</th>\n",
       "      <th>AA_change</th>\n",
       "      <th>allele_1</th>\n",
       "      <th>allele_2</th>\n",
       "      <th>alt_AF_1</th>\n",
       "      <th>alt_AF_2</th>\n",
       "      <th>altAF2_minus_altAF1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CASALI</td>\n",
       "      <td>P251</td>\n",
       "      <td>761139</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>Rv0667</td>\n",
       "      <td>1333</td>\n",
       "      <td>Antibiotic Resistance</td>\n",
       "      <td>rpoB</td>\n",
       "      <td>1.0</td>\n",
       "      <td>N</td>\n",
       "      <td>H445D</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CASALI</td>\n",
       "      <td>P251</td>\n",
       "      <td>861378</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>Rv0768</td>\n",
       "      <td>467</td>\n",
       "      <td>Non-Essential</td>\n",
       "      <td>aldA</td>\n",
       "      <td>1.0</td>\n",
       "      <td>N</td>\n",
       "      <td>I156S</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BRYANT</td>\n",
       "      <td>2020E</td>\n",
       "      <td>1253207</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>Rv1129c</td>\n",
       "      <td>1328</td>\n",
       "      <td>Non-Essential</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>N</td>\n",
       "      <td>I443T</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BRYANT</td>\n",
       "      <td>2020E</td>\n",
       "      <td>2075138</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>Rv1830</td>\n",
       "      <td>298</td>\n",
       "      <td>Non-Essential</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>N</td>\n",
       "      <td>S100T</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BRYANT</td>\n",
       "      <td>2020E</td>\n",
       "      <td>2694727</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>Rv2397c</td>\n",
       "      <td>238</td>\n",
       "      <td>Essential</td>\n",
       "      <td>cysA1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>N</td>\n",
       "      <td>F80L</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  population patient_id  ref_position ref_allele alt_allele  gene_id  \\\n",
       "0     CASALI       P251        761139          C          G   Rv0667   \n",
       "1     CASALI       P251        861378          T          G   Rv0768   \n",
       "2     BRYANT      2020E       1253207          A          G  Rv1129c   \n",
       "3     BRYANT      2020E       2075138          T          A   Rv1830   \n",
       "4     BRYANT      2020E       2694727          A          G  Rv2397c   \n",
       "\n",
       "  genomic_coord          gene_category gene_symbol  alt_AF_diff SNP_type  \\\n",
       "0          1333  Antibiotic Resistance        rpoB          1.0        N   \n",
       "1           467          Non-Essential        aldA          1.0        N   \n",
       "2          1328          Non-Essential         NaN          1.0        N   \n",
       "3           298          Non-Essential         NaN          1.0        N   \n",
       "4           238              Essential       cysA1          1.0        N   \n",
       "\n",
       "  AA_change allele_1 allele_2  alt_AF_1  alt_AF_2  altAF2_minus_altAF1  \n",
       "0     H445D        C        G       0.0       1.0                  1.0  \n",
       "1     I156S        T        G       0.0       1.0                  1.0  \n",
       "2     I443T        A        G       0.0       1.0                  1.0  \n",
       "3     S100T        T        A       0.0       1.0                  1.0  \n",
       "4      F80L        A        G       0.0       1.0                  1.0  "
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SNP_variants_within_patients.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(175, 17)"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.shape(SNP_variants_within_patients)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### *Function* for Boxplots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pylab import plot, show, savefig, xlim, figure, hold, ylim, legend, boxplot, setp, axes\n",
    "\n",
    "linewidth_for_box = 3.0\n",
    "\n",
    "# function for setting the colors of the box plots pairs\n",
    "def setBoxColors(bp , color):\n",
    "    setp(bp['boxes'][0], color=color , linewidth = linewidth_for_box)\n",
    "    setp(bp['caps'][0], color=color , linewidth = linewidth_for_box)\n",
    "    setp(bp['caps'][1], color=color , linewidth = linewidth_for_box)\n",
    "    setp(bp['whiskers'][0], color=color , linewidth = linewidth_for_box)\n",
    "    setp(bp['whiskers'][1], color=color , linewidth = linewidth_for_box)\n",
    "    setp(bp['fliers'][0], color=color , alpha = 0.65)\n",
    "    setp(bp['medians'][0], color=color , linewidth = 3.0)\n",
    "    bp['fliers'][0].set_markerfacecolor(color)\n",
    "    bp['fliers'][0].set_markeredgewidth(0.5)\n",
    "    bp['fliers'][0].set_markersize(6.0)\n",
    "    bp['fliers'][0].set_markeredgecolor('white')\n",
    "    \n",
    "def values_to_variant_types(values):\n",
    "    SNP_vals = values.alt_AF_diff\n",
    "    return [SNP_vals]\n",
    "\n",
    "def color_bar(bar , color):\n",
    "    bar.set_color(color)\n",
    "    bar.set_edgecolor('black')\n",
    "    bar.set_linewidth(1.0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYYAAAHICAYAAABK0k35AAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMi4zLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvIxREBQAAIABJREFUeJzt3Xt4VNW9//HPkJyQmAlJIBcMYOKFqA0YL0AtlwIWQZCLUJ9CsBGLnlZEq9ZztCgCFuqhVI43DvQgogaTFo+ISgXUWqsRLFJLooRLQE0IpCQBhNwhQ9bvj/llyEpCCJjMTMj79Tx5htl7z+zv7DXMZ9a+rHEYY4wAAPj/Ovm6AACAfyEYAAAWggEAYCEYAAAWggEAYCEYAAAWggEAYCEYAAAWggEAYCEYAAAWggEAYCEYAAAWgqEBh8Ph6xIAwKcIBgCAhWAAAFgIBgCAhWAAAFgIBgCAhWAAAFgIBgCAhWAAAFjaRTAsWbJE/fr1U+fOnXXHHXc0u+zTTz+t7t27Kzw8XNOnT9fx48e9UyQAnCfaRTDExcVp9uzZmj59erPLvfvuu1q4cKE++OAD5eXl6euvv9bcuXO9VCUAnB8cxhjj6yJaavbs2dq/f79efvnlJudPnTpVCQkJevLJJyVJH3zwgW677TYdPHiwxetwOBw6102yeLE0b55UXn5OD+8QnE73NnroIS+vmMY5M181zs7F0pfzJBdtc1qBTqnvPOlK77RNu+gxtFROTo6Sk5M995OTk1VUVKTDhw97Zf187pxZebl7O3kdjXNmvmocQuHMXOXu7eQl51UwlJeXKzw83HO/7t9lZWVeWr9XVtPu+WQ70Tgt44vtRCi0jBe3U6DX1uQFTqdTpaWlnvt1/w4LC/N6Le1nB533+M3AtTROY/7SOFNpm0YyvN8251WPISkpSdnZ2Z772dnZio2NVbdu3XxYFQC0L+0iGFwul6qrq3Xy5EmdPHlS1dXVcrlcjZa7/fbb9eKLL2rHjh369ttvtWDBgjOe3goAsLWLYFiwYIFCQkK0cOFCvfrqqwoJCdGCBQu0b98+OZ1O7du3T5J000036eGHH9bw4cMVHx+v+Ph4PfHEEz6uHgDal3Z1uqo3fJfTVevvpmWrNubT7UPjNM+X26f+PnSOMTTmg+3TLnoMAADvIRgAABaCAQBgIRgAABaCAQBgOa+ufAZaxF+u8kVjPrjKF43RYwAAWAgGAICFXUnoeLjArTF/2b3GBW6NMYgeAMDXCAYAgIVgAABYCAYAgIVgAABYCAYAgIVgAABYCAYAgIUL3NqIv1wvBABnix4DAMBCMAAALOxKaiMMx9MYu9eA9oEeAwDAQjAAACwEAwDAQjAAACwEAwDAQjAAACwEAwDAQjAAACwEAwDAQjAAACwEAwDAQjAAACwEAwDAQjAAACwEAwDAQjAAACwEAwDAQjAAACwEAwDAQjAAACwEAwDAQjAAACwEAwDAQjAAACwEAwDAQjAAACwEAwDAQjAAACwEAwDAQjAAACwEAwDA4hfBcOTIEU2cOFGhoaGKj49XRkZGk8sdPXpU06ZNU0xMjGJiYjRv3jxrfkJCgkJCQuR0OuV0OjVy5EgvVA8A55dAXxcgSTNnzlRQUJCKioqUlZWlm2++WcnJyUpKSrKWe/DBB1VZWam8vDwVFxfrRz/6keLj4/Wzn/3Ms8y6des0YsQIb78EADhv+LzHUFFRoTVr1mj+/PlyOp0aPHiwxo8fr1WrVjVadt26dXr44Yd1wQUXKCEhQXfeeadWrlzpg6oB4Pzl82DIzc1VQECAEhMTPdOSk5OVk5PT5PLGGOvf27dvt+bfdtttio6O1siRI5Wdnd02RQPAecznwVBeXq7w8HBrWnh4uMrKyhote9NNN2nhwoUqKyvT3r17tXLlSlVWVnrmp6enKy8vT/n5+Ro+fLhGjRqlo0ePtvlrAIDzic+Dwel0qrS01JpWWlqqsLCwRss+99xzCgkJUe/evTVhwgSlpKSoZ8+envmDBg1SSEiILrjgAs2aNUsRERHKzMxs89cAAOcTnwdDYmKiXC6X9uzZ45mWnZ3d6MCzJHXt2lXp6ek6ePCgcnJyVFtbqwEDBpz2uR0Oh7XrCQBwZj4/Kyk0NFSTJk3SnDlztGLFCmVlZemtt97S5s2bGy371VdfKSIiQhEREXrvvfe0fPlyffTRR5Kkffv2qaCgQP3791dtba2ef/55HTp0SIMGDfL2SwKAds3nPQZJWrp0qaqqqhQTE6OUlBQtW7ZMSUlJyszMlNPp9Cz3+eefq2/fvgoLC9OsWbOUnp7u6VmUlZVpxowZioyMVI8ePbRx40Zt2LBB3bp189XLAoB2yWHY12L5LrufHI5T/2arNubT7UPjNM+X2yej3rqn0jaN+GD7+EWPAQDgPwgGAICFYAAAWAgGAICFYAAAWAgGAICFYAAAWAgGAIDF50NiAF5X/2IuAI3QYwBwSr0haNBxEQzoGOLifF2B/3M6pQa/o46OiV1J6BgOHPB1BTbGboIfo8cAALAQDAAAC8EAALAQDAAAC8EAALBwVhIA/5HBxYf+gB4DALQHgd67+JBgAOBbIVx8eEaBTqnvPK+tzmEMV9fU53A4dK6bhGuWmsf2qYeN4b/q786a2jHbhh4DAMBCMAAALAQDAMBCMAAALAQDAMBCMAAALAQDAMBCMAAALAQDAMBCMAAALAQDAMBCMAAALAQDAMBCMAAALAQDAMBCMAAALAQDAMBCMAAALAQDAMBCMAAALAQDAMBCMAAALAQDAMBCMAAALAQDAMBCMAAALAQDAMBCMAAALAQDAMBCMAAALAQDAMDSLoLhyJEjmjhxokJDQxUfH6+MjIwmlzt69KimTZummJgYxcTEaN68ed4tFADOA4G+LqAlZs6cqaCgIBUVFSkrK0s333yzkpOTlZSUZC334IMPqrKyUnl5eSouLtaPfvQjxcfH62c/+5mPKgeA9sdhjDG+LqI5FRUVioyM1Pbt25WYmChJSk1NVY8ePbRw4UJr2aioKG3YsEH9+/eXJD355JPasGGDMjMzW7w+h8Ohc90kDsepf/v3VvUNtk89bAz/lVGvbaZ2zLbx+11Jubm5CggI8ISCJCUnJysnJ6fJ5et/qBtjtH379javEQDOJ34fDOXl5QoPD7emhYeHq6ysrNGyN910kxYuXKiysjLt3btXK1euVGVlpbdKBYDzgt8Hg9PpVGlpqTWttLRUYWFhjZZ97rnnFBISot69e2vChAlKSUlRz549vVUqAJwX/D4YEhMT5XK5tGfPHs+07OzsRgeeJalr165KT0/XwYMHlZOTo9raWg0YMMCb5QJAu+f3ZyWFhoZq0qRJmjNnjlasWKGsrCy99dZb2rx5c6Nlv/rqK0VERCgiIkLvvfeeli9fro8++sgHVQNA++X3PQZJWrp0qaqqqhQTE6OUlBQtW7ZMSUlJyszMlNPp9Cz3+eefq2/fvgoLC9OsWbOUnp7eZM8CAHB6fn+6qrdxumrbYfvUw8bwX5yu2j56DAAA7yEYAAAWggEAYCEYAAAWggEAYCEYAAAWggEAYCEYAAAWggEAYCEYAAAWggEAYCEYAAAWggEAYCEYAAAWggEAYCEYAAAWggEAYCEYAAAWggEAYCEYAAAWggEAYCEYAAAWggEAYCEYAAAWggEAYCEYAAAWggEAYCEYAAAWggEAYCEYAAAWggEAYCEYAAAWggEAYCEYAACWZoOhX79+3qoDAOAnmg2G0tJS6/6IESPatBgAgO81GwwOh8O6X1hY2KbFAAB876yOMTQMCgDA+SewuZkVFRXq27evEhMT1bt3b504cULbt2/XFVdcocDAZh8KAGinmv10z8nJ8fxt375d8fHxGjt2rAICAnT55Zfrqquu0sKFC71VKwDAC5oNhvDwcA0cOFADBw70TDtx4oR27drlCQwAwPnlrPcHBQUF6aqrrlJUVJRiYmLaoiYAgA+dMRiqqqq0e/du7dy5Uzt37tSOHTu0c+dOHTt2TGFhYdq5c6c36gQAeEmzwfDDH/5Q+fn5CggI0MUXX6zExEQNGjRIOTk5Wrduna655hpv1dnucAIXgPaq2WBwOBwKCwvTb3/7W02YMMEzPS0tTb169Wrz4nB+cjp9XQGA5jR7HcNf//pX/epXv9Ls2bM1fvx4bdmyxVt1tUtxcb6uwP85ndK8eb6uAkBzmu0xBAQEaPr06br11lv19NNPKyUlRUOGDNHx48e9VV+7cuCAryuw1d+dZYzv6gDQvrToyucuXbpo7ty5+vDDDxUUFKTS0lI9/fTTqqqqauv6AABe5jDm7L9LfvbZZ/rNb36joqIibd26tS3q8hmHw6Fz2CR+iR6DH6Nx/FdGvbaZ2jHb5px+jyEsLEwDBgxQSUlJa9cDAPCxFl/gdvjwYa1du1avvfaa8vPzNXz4cJ08ebItawMA+ECzwVBTU6P33ntPr732mjZv3qzBgwfrnnvu0ahRo+RwOPTOO+94q04AgJc0uyvp6quv1gsvvKBRo0Zp69ateumll3TLLbcoJCSkVYfgPnLkiCZOnKjQ0FDFx8crIyOjyeWOHz+uu+++W7GxseratavGjRunA/VOBRo2bJiCg4PldDrldDp1+eWXt1qNANBRNBsMV155pfbs2aNt27YpJyenzQ7Kzpw5U0FBQSoqKlJ6erpmzJjR5AB9zz77rD799FN98cUXKiwsVEREhO677z5rmSVLlqi8vFzl5eXavXt3m9QLAOezZoPh9ddf17vvvqu4uDg9/PDD6tevn+bNm6fs7OxWK6CiokJr1qzR/Pnz5XQ6NXjwYI0fP16rVq1qtOw333yjUaNGKTY2VsHBwZoyZQojvAJAKzvjWUk9e/bUgw8+qE2bNmnZsmWqrKzUlClTNHz48FYpIDc3VwEBAUpMTPRMS05ObvID/84779SmTZtUWFioyspKpaena/To0dYys2bNUlRUlAYNGqS//e1vrVIj4LF4sRQW5j7d9Lv81fddnysszF0X0ErO6nTVAQMGaNGiRdq2bZsefvhhDR069DsXUF5ervDwcGtaeHi4ysrKGi2bmJioiy66SD169FCXLl20c+dOzZkzxzP/d7/7nb7++msdOHBAP//5zzVu3Dh99dVX37lGwGPxYqm83NdV2MrLCQa0qnO6jiE4OFi33HKLXn311e9cgNPpVGlpqTWttLRUYWFhjZadMWOGqqurdfjwYVVUVGjSpElWj+H73/++wsLC1LlzZ02bNk2DBg3S+vXrv3ONgMdDD/nfKIBOp7suoJWcUzC0psTERLlcLu3Zs8czLTs7W0lJSY2Wzc7O1h133KGuXbuqc+fOuu+++/TZZ5/p0KFDTT73+XQVM/zEQw9JZWXuq5X95a+sjGBAq/J5MISGhmrSpEmaM2eOKioqtGnTJr311ltKTU1ttGz//v2VlpamY8eOqaamRkuXLlVcXJyioqJ09OhRvfvuu6qurpbL5VJ6ero+/vhjjRo1ygevCgDaL58HgyQtXbpUVVVViomJUUpKipYtW6akpCRlZmbKWa/b/tRTTyk4OFi9e/dWdHS01q9fr7Vr10pyX4w3e/ZsRUdHKyoqSs8//7zefPNNrmUAgLN0ToPonc/Op91PjNMGnAMG0fOPHgMAwH8QDAAAC8EAALC0eNhtAK3k5Elp82YpP1+Kj5cGDpQCAnxdFeBBjwHwppMnpSVLpIkTpV/8wn27ZIl7OuAnCAbAmzZvlubPl+p+L72qyn1/82bf1gXUQzAA3pSffyoU6lRVSfv2+aYeoAkEA+BN8fFSSIg9LSREuugi39QDNIFgALxp4EDp8cdPhUNIiPv+wIG+rQuohyufG+DKZ7S5urOS9u1z9xQ4K8m/cOUzp6sCXhcQIA0Z4usqgNNiVxIAwEIwAAAsBAMAwEIwAAAsBAMAwMJZSYC3GSPl5UnFxVJMjJSQYJ9bDPgYwQB4kzHS+vXS6tWSyyUFBkqTJ0tjxhAO8BsEA+BNeXmnQkFy365eLX3ve9LFF/u0tHZv52Lpy3mSq7z1njPjO4Z1oFPqO0+68qFWKcdbOMYAeFNx8alQqONySSUlvqnnfLJrceuGQmtwlbvramcIBsCbYmLcu4/qCwyUoqN9U8/55IqH3N/Q/Umg011XO8NYSQ0wVhLaFMcY0A4QDA0QDGhzdWcllZS4ewqclQQ/QzA0QDAA6Og4xgAAsBAMAAALwQAAsBAMAAALVz4DQB1jpIo8qbpYCo6RQhM65BljBAMASO5QKFwv5a+WjEtyBErxk6W4jneNCbuSAEBy9xTqQkFy3+avdk/vYAgGAJDcu49Mg3GsjEuq7njjWBEMACC5jyk4GuxddwRKwR1vHCuCAQAk94Hm+MmnwqHuGENogi+r8gmGxGiAITGADsxzVlKJu6fQQc9KIhgaIBgAdHTsSgIAWAgGAICFYAAAWAgGAICFYAAAWBgrCfC2up/2LC6WYmL4aU/4HYIB8CZjpPXrpdWrJZdLCgyUJk+WxnS8gdrgv9iVBHhTXt6pUJDct6tXu6cDfoJgALypuPhUKNRxuaSSjjdQG/wXwQB4U0yMe/dRfYGBUnTHG6gN/otgALwpIcF9TKEuHOqOMSQk+LIqwMJYSQ0wVhLaXN1ZSSUl7p4CZyXBzxAMDRAMADo6diUBACwEAwDAQjAAACwEAwDA0i6C4ciRI5o4caJCQ0MVHx+vjIyMJpc7fvy47r77bsXGxqpr164aN26cDhw44OVqgTMwRvrmG2nLFvctZwbAz7SLYJg5c6aCgoJUVFSk9PR0zZgxQzk5OY2We/bZZ/Xpp5/qiy++UGFhoSIiInTffff5oGLgNOrGSpo7V3r2Wfft+vWEA/yK3wdDRUWF1qxZo/nz58vpdGrw4MEaP368Vq1a1WjZb775RqNGjVJsbKyCg4M1ZcqUJgME8BnGSkI74PfBkJubq4CAACUmJnqmJScnN/mBf+edd2rTpk0qLCxUZWWl0tPTNXr0aG+WCzSPsZLQDvj9sNvl5eUKDw+3poWHh6usrKzRsomJibrooovUo0cPBQQEqG/fvlqyZIm3SgXOrG6spPrhwFhJ8DN+32NwOp0qLS21ppWWliosLKzRsjNmzFB1dbUOHz6siooKTZo0iR4D/AtjJaEd8PshMSoqKhQZGamcnBz17t1bknT77bcrLi5OCxcutJbt06ePfvvb32rChAmSpKNHjyoyMlIlJSWKiopq0foYEgNtjrGS4Of8vscQGhqqSZMmac6cOaqoqNCmTZv01ltvKTU1tdGy/fv3V1pamo4dO6aamhotXbpUcXFxLQ4FwCscDunii6UBA9y3hAL8jN8HgyQtXbpUVVVViomJUUpKipYtW6akpCRlZmbK6XR6lnvqqacUHBys3r17Kzo6WuvXr9fatWt9WDkAtD9+vyvJ29iVBKCjaxc9BgCA9xAMAAALwQAAsBAMAAALwQAAsBAMAAALwQAAsBAMAAALwQAAsBAMAAALwQAAsBAMAAALwQAAsBAMAAALwQAAsBAMAAALwQAAsBAMAAALwQAAsBAMAAALwQAAsBAMAAALwQAAsBAMAAALwQAAsBAMAAALwQAAsBAMAABLoK8LADocY6S8PKm4WIqJkRISJIfD11UBHgQD4E3GSOvXS6tXSy6XFBgoTZ4sjRlDOMBvsCsJ8Ka8vFOhILlvV692Twf8BMEAeFNx8alQqONySSUlvqkHaALBAHhTTIx791F9gYFSdLRv6gGaQDAA3pSQ4D6mUBcOdccYEhJ8WRVgcRhjjK+L8CcOh0PnyyapfyzzPHlJ54e6s5JKStw9Bc5Kgp8hGBogGAB0dOxKAgBYCAYAgIVgAABYCAYAgIVgAABYCAYAgIVgAABYCAYAgIVgAABYCAYAgIVgAABYCAYAgIVgAABYCAYAgIVgAABYCAYAgMUvguHIkSOaOHGiQkNDFR8fr4yMjCaXGz16tJxOp+cvKChIffv29cxPSEhQSEiIZ/7IkSO99RIA4LwReOZF2t7MmTMVFBSkoqIiZWVl6eabb1ZycrKSkpKs5TZs2GDdHzZsmG644QZr2rp16zRixIg2rxk4Z7W10rZtUkGB1KuXdM01Uie/+I4GSPKDHkNFRYXWrFmj+fPny+l0avDgwRo/frxWrVrV7OPy8vKUmZmp1NRUL1UKtILaWmnFCmnqVOmBB9y3K1a4pwN+wufBkJubq4CAACUmJnqmJScnKycnp9nHpaWlaciQIbr44out6bfddpuio6M1cuRIZWdnt0nNwDnbtk1avFg6ftx9//hx9/1t23xbF1CPz4OhvLxc4eHh1rTw8HCVlZU1+7i0tDTdcccd1rT09HTl5eUpPz9fw4cP16hRo3T06NHWLhk4dwUFp0KhzvHj0v79vqkHaILPg8HpdKq0tNSaVlpaqrCwsNM+5pNPPtHBgwd16623WtMHDRqkkJAQXXDBBZo1a5YiIiKUmZnZJnUD56RXL6lzZ3ta585Sz56+qQdogs+DITExUS6XS3v27PFMy87ObnTgub5XXnlFkyZNktPpbPa5HQ6HjDGtVivwnV1zjfTQQ6fCoXNn9/1rrvFtXUA9DuMHn5xTpkyRw+HQihUrlJWVpTFjxmjz5s1NhkNVVZUuvPBCvfHGG9YZSfv27VNBQYH69++v2tpaPf/881q0aJF27dqlbt26tbiW8ylMHI5T/z5PXtL5oe6spP373T0FzkqCn/GLd+PSpUtVVVWlmJgYpaSkaNmyZUpKSlJmZmajXsGbb76p8PBwDR8+3JpeVlamGTNmKDIyUj169NDGjRu1YcOGswoFwCs6dZKuu06aMMF9SyjAz/hFj8Gf0GMA0NHxVQUAYCEYAAAWggEAYPGLsZKADsUYKS9PKi6WYmKkhAT7gBDgYwQD4E3GSOvXS6tXSy6XFBgoTZ4sjRlDOMBvsCsJ8Ka8vFOhILlvV692Twf8BMEAeFNx8alQqONySSUlvqkHaALBAHhTTIx791F9gYFSdLRv6gGaQDAA3pSQ4D6mUBcOdccYEhJ8WRVg4crnBrjyGW2u7qykkhJ3T4GzkuBnCIYGCAYAHR27kgAAFoIBAGAhGAAAFoIBAGBhSAzA206elDZvlvLzpfh4aeBAKSDA11UBHvQYAG86eVJaskSaOFH6xS/ct0uWuKcDfoLTVRvgdFW0qcxMdxhUVZ2aFhIirV0rDRniu7qAeugxAN6Un2+HguS+v2+fb+oBmkAwAN4UH+/uIdQXEiJddJFv6gGaQDAA3jRwoPT446fCISTEfX/gQN/WBdTDMYYGOMaANld3VtK+fe6eAmclwc8QDA0QDAA6OnYlAQAsBAMAwEIwAAAsBAMAwEIwAAAsDKIHeBuD6MHP0WMAvIlB9NAOEAx+aPFiKSzMfR3Cd/mr77s+V1iYuy58R5s3S/PnnxovqarKfX/zZt/WBdRDMPihxYul8nJfV2ErLycYWgWD6KEdIBj80EMPSU6nr6uwOZ3uuvAdMYge2gGGxGjgfBoSA36o7hhD3e6kukH07r2XA9DwGwRDAwQD2hyD6MHPEQwNEAwAOjqOMQAALAQDAMBCMAAALAQDAMDCWEmAt9XWStu2SQUFUq9e0jXXSJ34jgb/QTAA3lRbK61Y4b6M/PhxqXNn95WDd91FOMBvcLpqA5yuijb1+efS1KnuUKjTubOUkSFdd53v6gLq4SsK4E0FBXYoSO77+/f7ph6gCQQD4E29erl7CPV17iz17OmbeoAmEAyAN11zjfuYQl041B1juOYa39YF1MMxhgY4xoA2V3dW0v797p4CZyXBzxAMDRAMADo6vqYAACwEAwDAQjAAACwEAwDA0i6C4ciRI5o4caJCQ0MVHx+vjIyMJpcbPXq0nE6n5y8oKEh9+/b1crUA0L61i7GSZs6cqaCgIBUVFSkrK0s333yzkpOTlZSUZC23YcMG6/6wYcN0ww03eLNU4MyMkfLypOJiKSZGSkiQHA5fVwV4+P3pqhUVFYqMjNT27duVmJgoSUpNTVWPHj20cOHC0z4uLy9Pl156qfbu3auLL764xevjdFW0KWOk9eul1asll0sKDJQmT5bGjCEc4Df8fldSbm6uAgICPKEgScnJycrJyWn2cWlpaRoyZMhZhQLQ5vLyToWC5L5dvdo9HfATfh8M5eXlCg8Pt6aFh4errKys2celpaXpjjvuaMPKgHNQXHwqFOq4XFJJiW/qAZrg98HgdDpVWlpqTSstLVVYWNhpH/PJJ5/o4MGDuvXWW9u6PODsxMS4dx/VFxgoRUf7ph6gCX4fDImJiXK5XNqzZ49nWnZ2dqMDz/W98sormjRpkpxOpzdKBFouIcF9TKEuHOqOMSQk+LIqwOL3B58lacqUKXI4HFqxYoWysrI0ZswYbd68uclwqKqq0oUXXqg33njjnM5I4uAz2lzdWUklJe6eAmclwc/4fY9BkpYuXaqqqirFxMQoJSVFy5YtU1JSkjIzMxv1Ct58802Fh4dr+PDhPqoWOAOHQ7r4YmnAAPctoQA/0y56DN5EjwFAR9cuegwAAO8hGAAAFoIBAGAhGAAAFoIBAGAhGAAAlnYx7La3OTivHMB55mxOwycYGuAaBgAdHbuSAAAWggEAYCEYAAAWggEAYCEYAAAWggEAYCEYAAAWggEAYCEYAMALnE6nvv76a1+X0SIEQyv65JNPNHDgQIWHh6tr164aNGiQtm7d6uuyzhsJCQmKjY1VRUWFZ9qKFSs0bNiwVl9XTk6ORo4cqcjISEVEROi6667T+vXrW309OLOEhAT95S9/OeNyw4YN04oVK7xQ0bkpLy/XJZdc4usyWoRgaCWlpaUaO3as7rvvPh05ckQHDhzQ3Llz1blzZ1+Xdl5xuVx69tln23w948aN04033qiioiIVFxfrueeeU5cuXdp8vfCdkydP+roE/2HQKrZu3WrCw8MbTa+urjaRkZHmiy++8EwrKioywcHBpri42Hz44YemR48e5qmnnjLR0dGme/fuZuXKlZ5ljx49alJTU01UVJS56KKLzPz5883JkyeNMca89NJLZtCgQeahhx4yERERJiEhwaxfv94YY8xrr71mrr32WquWp56bpta/AAATfklEQVR6ykyYMMEYY8y0adPMjBkzzE033WRCQ0PNwIEDzb/+9S9z//33m4iICHP55Zebf/7zn57H7tixwwwdOtSEh4eb733ve+att97yzBs6dKh54YUXPPfr6jLGmNraWvPAAw+Y6Oho06VLF9O3b1/z5ZdfntM2jo+PN//1X/9lIiMjzbfffmuMMeaFF14wQ4cONcYYs2nTJtOvXz/TpUsX069fP7Np0yarxtmzZ5uBAwcap9NpbrzxRlNSUtLkekpKSowkzzoaSkpKMm+//bbn/okTJ0y3bt3Mtm3bzDfffGMkmZdfftn06tXLdOvWzSxYsMCzbHV1tbn//vvNhRdeaC688EJz//33m+rqamOMafa98Nlnn5mYmBhTU1Pjea7XX3/dJCcnG2OMmTt3rrn11lvNbbfdZpxOp+nTp4/ZvXu3efLJJ010dLTp2bOneffddz2PPXDggBk3bpyJjIw0l156qVm+fLln3rRp08xjjz3muV9XV52FCxeauLg443Q6TWJiovnLX/7S5HZqLfHx8eb9999v9v3+6KOPmk6dOpnOnTub0NBQM3PmTGOMMTt37jQjRowwkZGRJjEx0axevdp6nXfffbcZPXq0ueCCC8z7779vDh06ZMaOHWvCwsJMv379zGOPPeZ5L7fk+e655x4zZswY43Q6zYABA8zevXs98yWZPXv2GGOMqaysNL/61a/MRRddZLp06WIGDRpkKisr23Q7ng2CoZUcO3bMdO3a1dx+++1m/fr15siRI555M2bMMA8//LDn/jPPPGPGjh1rjHH/pwsICDCPP/64OXHihHnnnXdMSEiI5/Gpqalm/PjxprS01HzzzTemd+/eZsWKFcYY9wdwYGCgWb58uXG5XGbp0qXmwgsvNLW1tZ5A2rFjh2e9V199tXn99deNMe43cbdu3cw//vEPU1VVZYYPH24SEhLMK6+8Ylwul3nsscfMsGHDjDHuD75LL73U/Pa3vzXHjx83H3zwgXE6nWbXrl3GmOaDYePGjebaa6813377ramtrTU7duwwhYWF57SN6z4gJk6c6PngqguGw4cPm4iICJOWlmZqampMRkaGiYiIMIcOHfLUeMkll5jdu3ebyspKM3ToUPPII480uZ7a2lpz2WWXmZtvvtmsXbvWHDx40Jr/u9/9zvzkJz/x3H/zzTdNnz59jDHGEwx33XWXqaysNFlZWSYoKMjTDo8//rj5/ve/b4qKikxxcbH5wQ9+YGbPnm2MOfN74corr/R8EBpjzC233GKeeuopY4w7GDp37mw2btxoampqTGpqqklISDALFiwwJ06cMMuXLzcJCQmex/7whz80M2bMMFVVVWbbtm0mKirK8wHfXDDs2rXL9OzZ0xw4cMDzeut/+LWF+sFwuve7MY3fh+Xl5aZnz55m5cqVpqamxnz++eemW7duZvv27Z7X2aVLF/PJJ5+YkydPmqqqKjN58mQzefJkU1FRYXJyckzPnj097+WWPF9kZKTZsmWLqampMVOnTjWTJ0/21FM/GO655x4zdOhQs3//fuNyucymTZs8XxD8AcHQinbs2GGmTZtmevToYQICAsy4cePMwYMHzd///nfTs2dPzzf96667zvNN48MPPzTBwcHWN8Ho6Gjz6aefGpfLZYKCgkxOTo5n3h/+8AfPN+SXXnrJXHrppZ55FRUVRpL517/+ZYwx5u677zaPPvqoMcaY7du3m4iICM+bb9q0aeauu+7yPPa5554zV1xxhef+F1984ekBffzxxyY2NtZTvzHGTJkyxcydO9cY03wwfPDBB6Z3797m008/tR5/Luo+IL788kvTpUsXU1xc7AmGtLQ0079/f2v566+/3rz00kueGufPn++Z9z//8z9m1KhRp11XQUGBmTlzprnkkkuMw+EwQ4YMMbm5ucYY97dtp9Npjh07Zowx5sc//rH53e9+Z4w5FQwFBQWe5+rfv7/54x//aIwx5pJLLjHvvPOOZ97GjRtNfHy8Mab594Ix7m/qU6dONcYYc/jwYRMSEuIJ2blz55oRI0Z4Hvf222+b0NBQ43K5jDHGlJaWenpB+/btM506dTKlpaWe5X/961+badOmGWOaD4Y9e/aY6Oho8/7775sTJ06cdvu1pvrB0Nz7veH78E9/+pMZPHiw9Vw///nPzbx584wx7teZmprqmedyuUxgYKDnC48xxuoxtOT57rzzTs+8d955x1x++eWe+3XBcPLkSRMcHGyysrLObYN4AccYWtGVV16pl19+Wfv379f27dtVWFioBx54QN///vcVGhqqjz76SLt27dLevXs1fvx4z+O6deumwMBTI6BfcMEFKi8v16FDh3TixAnFx8d75sXHx+vAgQOe+927d7ceJ7kPcknStGnTlJGRIWOMVq1apZ/85CfWMY/Y2FjPv0NCQhrdr3uewsJC9erVS506nXq7NKzjdG644Qbde++9mjlzpmJjY/Xzn/9cpaWlZ3xcc/r06aOxY8dq4cKFnmmFhYXWdmqqxobbqu713X333XI6nXI6nXryySclST179tSSJUv01VdfKT8/X6Ghobr99tslSXFxcRo0aJDWrFmjo0ePasOGDbrtttusdZ9uXQ3rjI+PV2Fhoef+6d4LkvTTn/5U69atU3l5uV577TUNGTJEF154oWfZhu0XFRWlgIAAz33J/d4oLCxU165dFRYWdtptdTqXXXaZnnnmGc2bN08xMTGaMmWKVX9ba+793lB+fr62bNmiiIgIz196eroOHjzoWaZXr16ef5eUlMjlclnT6v+7Jc93unav79ChQ6qurtall156Ni/dqwiGNnLFFVfojjvu0Pbt2yW5P6RfffVVrVq1SrfeequCg4PP+BxRUVH6t3/7N+Xn53um7du3Tz169GhRDddff72CgoKUmZmpjIwMpaamntNriYuLU0FBgWpra5usIzQ0VJWVlZ559f+jSNIvf/lLff7558rJyVFubq5+//vfn1Md9T3xxBN64YUXPB9mcXFx1nZqWGNz/vCHP6i8vFzl5eV69NFHG83v1auXZs6c6WlL6VR7/t///Z9+8IMftLhNGta5b98+xcXFteixPXr00A9+8AOtXbtWq1at+k7teeTIEZWVlVl1tLQ9p06dqk8++UT5+flyOBx65JFHzqmO1tbwB7Z69eqloUOH6ujRo56/8vJyLVu2rMnHREdHKzAwUPv37/dMKygoOKvna4moqCgFBwfrq6++OtuX6DUEQyvZtWuXFi9e7HlTFRQU6I9//KOuv/56SVJqaqrWrl2rV1991fPN80wCAgL0k5/8RI899pjKysqUn5+v//7v/9ZPf/rTFtd1++23695771VgYKAGDx589i9M8vR4Fi1apJqaGv3tb3/TunXrNGXKFEnS1VdfrTfeeEOVlZXau3evXnzxRc9jt27dqi1btqimpkahoaEKDg72fIv9Li677DJNnjxZzz33nCRpzJgxys3NVUZGhlwul1avXq0dO3Zo7NixZ/3c3377rebOnau9e/eqtrZWhw4d0sqVKz1tKUm33HKL/vnPf+rZZ59tcXtKUkpKihYsWKCSkhIdOnRIv/nNb866PRctWqQvv/xSEydOPKvXVadXr14aOHCgZs2aperqan3xxRd68cUXPb2eq6++WuvXr9eRI0d08OBBPfPMM57H7t69W3/96191/PhxBQcHKyQkpFXaszXExsZa1wmMHTtWubm5WrVqlWpqalRTU6OtW7dq586dTT4+ICBAkyZN0rx581RZWaldu3YpLS3tnJ/vdDp16qTp06frV7/6lQoLC3Xy5El9+umnOn78+Lm98DZAMLSSsLAwbdmyxfMhev3116tPnz5avHixJPeuiWuvvVYOh0NDhgxp8fM+//zzCg0N1SWXXKLBgwdr6tSpmj59eosfn5qaqu3bt5/zt0tJCgoK0ttvv60NGzYoKipK99xzj9LS0nTFFVdIkh588EEFBQUpNjZW06ZNs3arlJaW6t///d8VGRmp+Ph4devWTf/xH/9xzrXUN2fOHM81Dd26ddOf//xnLV68WN26ddOiRYv05z//WVFRUef0evPy8jRixAh16dJFffr0UefOnfXyyy97lgkJCdGPf/xjffPNN5o0aVKLn3v27Nnq16+frrrqKvXt21fXXnutZs+e3eLHT5w4Ufn5+Zo4caJCQ0PP5mVZ/vjHPyovL09xcXGaOHGinnjiCd14442S3O+Z5ORkJSQkaOTIkZo8ebLnccePH9evf/1rRUVFqXv37iouLvbsfvO1+++/X6+//roiIyP1y1/+UmFhYXrvvff0pz/9SXFxcerevbseeeSRZj+AlyxZomPHjql79+5KTU1VSkqKZ/fruTzf6Tz11FPq27ev+vfvr65du+qRRx6xeuS+5jCG37L0lunTpysuLk4LFizw2jqrqqoUExOjf/7zn+rdu7fX1tsR/OY3v1Fubq5effVVr6730ksv1f/+7/9qxIgRXl1vR/TII4/o4MGDeuWVV3xdilfxm89ekpeXpzfeeEPbtm3z6nqXLVum/v37Ewqt7MiRI3rxxRe1atUqr653zZo1cjgcuuGGG7y63o5i165dOnHihPr27autW7fqxRdf9OurqdsKweAFjz/+uJ5++mnNmjVLF198sdfWm5CQIGOM3nzzTa+tsyN44YUX9MADDyg1NVU//OEPvbbeYcOGaceOHVq1apV1hhhaT1lZmVJSUlRYWKiYmBg99NBDmjBhgq/L8jp2JQEALHztAABYCAY/NGvWLOsUQV977rnn9Otf/9rXZfiEv7XFmUyaNEkbN270dRk+4W9t9fbbb3tO6W53fHnZNRorLi42cXFxbTag1vHjx83ll19uDYq2e/duM378eBMVFWUiIyPNyJEjrWEBqqqqTI8ePUxRUVGb1OSv2qot/vrXv5phw4aZLl26eIbDqG/27NmmT58+JiAgwDPsSJ0///nPZtCgQSY8PNzExsaau+66yxraYsuWLY0GT+wI2qqtFi1aZJKSkozT6TQJCQlm0aJF1vzm2soY94CL2dnZrVqTN9Bj8DMvv/yyxowZ4xnCoLX9/ve/V0xMjDXt6NGjGj9+vHbv3q2ioiINGDDAOuAWHBys0aNHWxf7dARt1RahoaGaPn36aa8Av+yyy7Ro0SLdfPPNjeYdO3ZMs2fPVmFhoXbu3Kn9+/frP//zPz3zBwwYoNLSUv3jH/9o1Zr9XVu1lTFGaWlp+vbbb7Vx40YtWbJEf/rTnzzzm2sryX1B4/Lly1u1Jq/wdTLBNnz4cLNq1SrP/TMNy302vv76a3PFFVeY9evXWz2Ghg4fPmwkeUYmNcaYV1991TPaakfRlm1hjDHvv/9+kz2GOrfddluT30LrW7NmjWdk1zp33XWXZ2C3jqKt26rOfffdZ+69995G00/XVp988ok1qm17QY/Bz3z55Ze6/PLLrWkHDx7UsWPHdODAAb344ouaOXOmvv32W0nSwoULrUG9Gv7Vd9999+nJJ58847eqjz/+WN27d1e3bt0806688kplZ2e30qtsH9qyLVrLxx9/rKSkJGsabeXW2m1ljFFmZmaj7d2cK6+8Unl5ed954Eiv83UywRYYGGh27tzpuX+moZhb6o033vAMM93wh1fqKygoMHFxcSYjI8Oanpubazp16nRW62zv2qot6nzXHsN7771nIiIizO7du63py5cvN8OHDz+nmtqrtm4rY4yZM2eOueqqq5r83YTTtdWJEyeMJJOfn3/O6/UFLnDzM5GRkdaol1LzQzG3REVFhR5++OEz/mZxSUmJRo4cqXvuuUcpKSnWvLKyMoWHh7d4neeDtmiL1vL3v/9dU6dO1euvv67ExERrXllZWZv1UPxVW7fVkiVLlJaWpszMzLP6ud66mtpbe7Aryc9cddVVys3NbfHyTz75pOe3BJr6k6Q9e/YoLy9PQ4YMUffu3TVp0iT961//Uvfu3ZWXlyfJPaLoyJEjNX78eD322GON1rNz504lJye3ymtsL9qiLVrDtm3bNH78eK1cuVI/+tGPGs2nrc7sbNpq5cqVWrhwoT744AP17NnzrOrauXOnEhIS2t3vhRMMfmbMmDH66KOPWrz8o48+6vktgab+JPcP2xQUFCgrK0tZWVlasWKFYmNjlZWVpV69eqm0tFSjRo3SoEGDrB+/qe+jjz7S6NGjW+U1thdt0RaSVFtbq+rqatXU1MgYo+rqap04ccIzv6amRtXV1aqtrZXL5VJ1dbXnh+q3b9+um266Sc8//7zGjRvXZB201Zm1tK3S09P16KOP6v3339cll1zS6HmaayupHbeFr/dlwVZSUmJ69OjhOR+7qeMBdT91eK4aPufLL79sJJkLLrjAhIaGev7q9ovWXcfQ8LePz3dt1RYffvihkWT91f1cqzHun4hsOL/uJ0rvuOMO43A4rHb63ve+53nsZ599Zq6++upze8HtWFu1VUJCggkMDLS29y9+8QvP/Obayhhj+vTp49c/4Xk6jJXkhx599FHFxMTogQce8HUpkty/CVFQUKBFixb5uhSv87e2OJMf//jHuvPOOzVmzBhfl+J1/tZW69at06pVq/Taa6/5upSzRjAAACwcYwAAWAgGAICFYAAAWAgGAICFYAAAWAgGAICFYAAAWAgGAIDl/wHirjCe0Z6SHQAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 378x468 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "fig = plt.figure(1, figsize=(5.25, 6.5))\n",
    "ax1 = axes()\n",
    "\n",
    "box_width = 0.7\n",
    "ind = [1,2,3]\n",
    "\n",
    "### Example in case we want to show means later ###\n",
    "##Syn_values = SNP_variants_within_patients[SNP_variants_within_patients.SNP_type == 'S']\n",
    "##meanlineprops = dict(linestyle='--' , color = 'b' , linewidth = linewidth_for_box)\n",
    "##bp = ax1.boxplot(values_to_variant_types(Syn_values) , positions = [ind[0]], widths = box_width , meanprops = meanlineprops, meanline = True, showmeans = True)\n",
    "##setBoxColors(bp , 'b')\n",
    "\n",
    "#Synonymous Mutations\n",
    "Syn_values = SNP_variants_within_patients[SNP_variants_within_patients.SNP_type == 'S']\n",
    "bp = ax1.boxplot(values_to_variant_types(Syn_values) , positions = [ind[0]], widths = box_width)\n",
    "setBoxColors(bp , 'blue')\n",
    "\n",
    "#Non-Synonymous Mutations\n",
    "NSyn_values = SNP_variants_within_patients[SNP_variants_within_patients.SNP_type == 'N']\n",
    "bp = ax1.boxplot(values_to_variant_types(NSyn_values) , positions = [ind[1]], widths = box_width)\n",
    "setBoxColors(bp , 'red')\n",
    "\n",
    "#Intergenic Mutations\n",
    "Int_values = SNP_variants_within_patients[SNP_variants_within_patients.SNP_type == 'I']\n",
    "bp = ax1.boxplot(values_to_variant_types(Int_values) , positions = [ind[2]], widths = box_width)\n",
    "setBoxColors(bp , 'orange')\n",
    "\n",
    "#y-labels\n",
    "ax1.set_ylim([0.69 , 1.01])\n",
    "ax1.set_yticks(np.arange(0.70 , 1.01 , .05))\n",
    "y_labels = np.round( np.arange(0.70 , 1.00 , .05) , 2)\n",
    "ax1.set_yticklabels(y_labels, rotation='horizontal', fontsize = 12)  \n",
    "ax1.tick_params(axis='y', which='major', labelsize=12, pad = 4)\n",
    "\n",
    "#x-labels\n",
    "ax1.set_xlim([0.5 , 3.5])\n",
    "ax1.set_xticks(ind)\n",
    "x_labels = ['Synonymous\\n(n=' + str(len(Syn_values)) + ')', 'Non-Synonymous\\n(n=' + str(len(NSyn_values)) + ')', 'Intergenic\\n(n=' + str(len(Int_values)) + ')']\n",
    "ax1.set_xticklabels(x_labels, rotation='horizontal', fontsize = 12)  \n",
    "\n",
    "ax1.spines['right'].set_visible(False)\n",
    "ax1.spines['top'].set_visible(False)\n",
    "\n",
    "#label y-axis\n",
    "fig.text(-0.013, 0.56, '$\\Delta AF$' ,  ha='center', va='center', rotation=90, fontsize = 12 , color = 'k')\n",
    "\n",
    "plt.tick_params(\n",
    "    axis='x',          # changes apply to the x-axis\n",
    "    which='both',      # both major and minor ticks are affected\n",
    "    bottom='off',      # ticks along the bottom edge are off\n",
    "    top='off',         # ticks along the top edge are off\n",
    "    labelbottom='on') # labels along the bottom edge are off\n",
    "\n",
    "plt.tick_params(\n",
    "    axis='y',          # changes apply to the x-axis\n",
    "    which='both',      # both major and minor ticks are affected\n",
    "    left='off',      # ticks along the bottom edge are off\n",
    "    right='off',         # ticks along the top edge are off\n",
    "    labelleft='on') # labels along the bottom edge are off\n",
    "\n",
    "\n",
    "fig = plt.gcf()\n",
    "fig.tight_layout()\n",
    "\n",
    "#output for Adobe Illustrator\n",
    "file_name = '/n/data1/hms/dbmi/farhat/Roger/inhost_TB_dynamics_project/figures/longitudinal_SNP_analysis/Syn_vs_NSyn_vs_change_in_AF.pdf'\n",
    "plt.savefig(file_name, bbox_inches='tight', dpi = 300 , transparent = True)\n",
    "\n",
    "plt.show()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
