{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>.container { width:100% !important; }</style>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from IPython.core.display import display, HTML\n",
    "display(HTML(\"<style>.container { width:100% !important; }</style>\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### This notebook was written to look for convergent evolution at the *pathway level* for in-host SNPs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import vcf\n",
    "\n",
    "%matplotlib inline\n",
    "import os\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib as mpl\n",
    "import matplotlib.ticker as ticker\n",
    "from pylab import plot, show, savefig, xlim, figure, hold, ylim, legend, boxplot, setp, axes\n",
    "from itertools import compress\n",
    "from pylab import MaxNLocator\n",
    "import seaborn as sns; sns.set()\n",
    "from matplotlib.colors import LogNorm\n",
    "from matplotlib import gridspec\n",
    "import ast\n",
    "import itertools\n",
    "import seaborn as sns\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "\n",
    "import fastcluster\n",
    "from sklearn import cluster, datasets\n",
    "import scipy.cluster.hierarchy as hier\n",
    "from sklearn.cluster import KMeans\n",
    "import time\n",
    "import sys\n",
    "\n",
    "import Bio\n",
    "from Bio.Alphabet import IUPAC\n",
    "from Bio.Blast.Applications import NcbiblastnCommandline\n",
    "from Bio.Blast import NCBIXML\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.SeqFeature import SeqFeature, FeatureLocation\n",
    "from Bio import pairwise2\n",
    "from Bio import SeqIO\n",
    "from Bio.Graphics import GenomeDiagram\n",
    "from Bio.SeqUtils import GC\n",
    "\n",
    "from Bio.Align.Applications import MuscleCommandline\n",
    "from StringIO import StringIO\n",
    "from Bio import AlignIO\n",
    "from Bio.Align import AlignInfo\n",
    "from Bio.Seq import MutableSeq\n",
    "import itertools\n",
    "\n",
    "import networkx as nx\n",
    "import scipy\n",
    "\n",
    "#for exporting to Adobe Illustrator\n",
    "mpl.rcParams['pdf.fonttype'] = 42\n",
    "mpl.rcParams['ps.fonttype'] = 42"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Import annotation from Mycobrowser"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "mycobrowser_table = pd.read_csv('/n/data1/hms/dbmi/farhat/Roger/inhost_TB_dynamics_project/CSV_files/H37Rv_info/Mycobrowser_Release3/Mycobacterium_tuberculosis_H37Rv_txt_v3.txt' , sep = '\\t')\n",
    "\n",
    "#drop unnecessary columns\n",
    "mycobrowser_table = mycobrowser_table.loc[: , ['Refseq_ID' , 'Feature' , 'Start' , 'Stop', 'Score' , 'Strand' , 'Frame' , 'Locus' , 'Name' , 'Function' , 'Product' , 'Comments' , 'UniProt_AC' , 'Functional_Category']]\n",
    "\n",
    "#organize by Start Ref Position\n",
    "mycobrowser_table.sort_values(by = ['Start'] , inplace = True)\n",
    "\n",
    "#subset to CDS regions\n",
    "mycobrowser_table = mycobrowser_table[mycobrowser_table.Feature == 'CDS']\n",
    "\n",
    "#reset index\n",
    "mycobrowser_table.reset_index(inplace = True , drop = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Refseq_ID</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>Stop</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>Locus</th>\n",
       "      <th>Name</th>\n",
       "      <th>Function</th>\n",
       "      <th>Product</th>\n",
       "      <th>Comments</th>\n",
       "      <th>UniProt_AC</th>\n",
       "      <th>Functional_Category</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_000962.3</td>\n",
       "      <td>CDS</td>\n",
       "      <td>1</td>\n",
       "      <td>1524</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Rv0001</td>\n",
       "      <td>dnaA</td>\n",
       "      <td>Plays an important role in the initiation and ...</td>\n",
       "      <td>Chromosomal replication initiator protein DnaA</td>\n",
       "      <td>Rv0001, (MT0001, MTV029.01, P49993), len: 507 ...</td>\n",
       "      <td>P9WNW3</td>\n",
       "      <td>information pathways</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_000962.3</td>\n",
       "      <td>CDS</td>\n",
       "      <td>2052</td>\n",
       "      <td>3260</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Rv0002</td>\n",
       "      <td>dnaN</td>\n",
       "      <td>DNA polymerase III is a complex, multichain en...</td>\n",
       "      <td>DNA polymerase III (beta chain) DnaN (DNA nucl...</td>\n",
       "      <td>Rv0002, (MTV029.02, MTCY10H4.0), len: 402 aa. ...</td>\n",
       "      <td>P9WNU1</td>\n",
       "      <td>information pathways</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_000962.3</td>\n",
       "      <td>CDS</td>\n",
       "      <td>3280</td>\n",
       "      <td>4437</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Rv0003</td>\n",
       "      <td>recF</td>\n",
       "      <td>The RECF protein is involved in DNA metabolism...</td>\n",
       "      <td>DNA replication and repair protein RecF (singl...</td>\n",
       "      <td>Rv0003, (MTCY10H4.01), len: 385 aa. RecF, DNA ...</td>\n",
       "      <td>P9WHI9</td>\n",
       "      <td>information pathways</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_000962.3</td>\n",
       "      <td>CDS</td>\n",
       "      <td>4434</td>\n",
       "      <td>4997</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Rv0004</td>\n",
       "      <td>Rv0004</td>\n",
       "      <td>Function unknown</td>\n",
       "      <td>Conserved hypothetical protein</td>\n",
       "      <td>Rv0004, (MTCY10H4.02), len: 187 aa. Conserved ...</td>\n",
       "      <td>P9WFL1</td>\n",
       "      <td>conserved hypotheticals</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_000962.3</td>\n",
       "      <td>CDS</td>\n",
       "      <td>5240</td>\n",
       "      <td>7267</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Rv0005</td>\n",
       "      <td>gyrB</td>\n",
       "      <td>DNA gyrase negatively supercoils closed circul...</td>\n",
       "      <td>DNA gyrase (subunit B) GyrB (DNA topoisomerase...</td>\n",
       "      <td>Rv0005, (MTCY10H4.03), len: 675 aa. GyrB, DNA ...</td>\n",
       "      <td>P9WG45</td>\n",
       "      <td>information pathways</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Refseq_ID Feature  Start  Stop Score Strand  Frame   Locus    Name  \\\n",
       "0  NC_000962.3     CDS      1  1524     .      +    0.0  Rv0001    dnaA   \n",
       "1  NC_000962.3     CDS   2052  3260     .      +    0.0  Rv0002    dnaN   \n",
       "2  NC_000962.3     CDS   3280  4437     .      +    0.0  Rv0003    recF   \n",
       "3  NC_000962.3     CDS   4434  4997     .      +    0.0  Rv0004  Rv0004   \n",
       "4  NC_000962.3     CDS   5240  7267     .      +    0.0  Rv0005    gyrB   \n",
       "\n",
       "                                            Function  \\\n",
       "0  Plays an important role in the initiation and ...   \n",
       "1  DNA polymerase III is a complex, multichain en...   \n",
       "2  The RECF protein is involved in DNA metabolism...   \n",
       "3                                   Function unknown   \n",
       "4  DNA gyrase negatively supercoils closed circul...   \n",
       "\n",
       "                                             Product  \\\n",
       "0     Chromosomal replication initiator protein DnaA   \n",
       "1  DNA polymerase III (beta chain) DnaN (DNA nucl...   \n",
       "2  DNA replication and repair protein RecF (singl...   \n",
       "3                     Conserved hypothetical protein   \n",
       "4  DNA gyrase (subunit B) GyrB (DNA topoisomerase...   \n",
       "\n",
       "                                            Comments UniProt_AC  \\\n",
       "0  Rv0001, (MT0001, MTV029.01, P49993), len: 507 ...     P9WNW3   \n",
       "1  Rv0002, (MTV029.02, MTCY10H4.0), len: 402 aa. ...     P9WNU1   \n",
       "2  Rv0003, (MTCY10H4.01), len: 385 aa. RecF, DNA ...     P9WHI9   \n",
       "3  Rv0004, (MTCY10H4.02), len: 187 aa. Conserved ...     P9WFL1   \n",
       "4  Rv0005, (MTCY10H4.03), len: 675 aa. GyrB, DNA ...     P9WG45   \n",
       "\n",
       "       Functional_Category  \n",
       "0     information pathways  \n",
       "1     information pathways  \n",
       "2     information pathways  \n",
       "3  conserved hypotheticals  \n",
       "4     information pathways  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mycobrowser_table.head(n=5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(4031, 14)"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.shape(mycobrowser_table)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Subset Mycobrowser table to genes that we kept for analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "#load CSV file for Gene Categories\n",
    "gene_categories = pd.read_csv('/n/data1/hms/dbmi/farhat/Roger/inhost_TB_dynamics_project/CSV_files/gene_categories/gene_categories.csv').set_index('name')\n",
    "\n",
    "#get list of all Genes to consider in Significance Testing\n",
    "H37Rv_tags_to_keep = []\n",
    "for gene_category in ['Essential' , 'Non-Essential' , 'Antigen' , 'Antibiotic Resistance' , 'PE/PPE']: #only genes included in analysis\n",
    "    H37Rv_tags_to_keep = H37Rv_tags_to_keep + list( gene_categories[gene_categories.Gene_Category == gene_category].gene_id )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3886"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(H37Rv_tags_to_keep)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "#subset to certain genes\n",
    "mycobrowser_table_filter = [(H37Rv_tag in H37Rv_tags_to_keep) for H37Rv_tag in list( mycobrowser_table.Locus )]\n",
    "mycobrowser_table = mycobrowser_table[mycobrowser_table_filter]\n",
    "\n",
    "#reset index\n",
    "mycobrowser_table.reset_index(inplace = True , drop = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(3883, 14)"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.shape(mycobrowser_table)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Create a list that holds each H37Rv tag and its genomic coordinates (+/- a fudge factor)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "H37Rv_coords_tag_map = []\n",
    "\n",
    "for H37Rv_region_i in mycobrowser_table.index:\n",
    "    \n",
    "    H37Rv_tag = mycobrowser_table.loc[H37Rv_region_i , 'Locus']\n",
    "    H37Rv_coords = np.sort( list( mycobrowser_table.loc[H37Rv_region_i , ['Start' , 'Stop']] ) ) \n",
    "    start_coord = H37Rv_coords[0]\n",
    "    stop_coord = H37Rv_coords[1]\n",
    "    \n",
    "    #structure: H37Rv tag, start coord +/- fudge factor, stop coord +/- fudge factor (20bp)\n",
    "    H37Rv_coords_tag_map.append( [H37Rv_tag , range(start_coord-20 , start_coord+21) , range(stop_coord-20 , stop_coord+21)] )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3883"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(H37Rv_coords_tag_map)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Test for some gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Rv0011c\n"
     ]
    }
   ],
   "source": [
    "print H37Rv_coords_tag_map[10][0] #H37Rv tag"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[13694, 13695, 13696, 13697, 13698, 13699, 13700, 13701, 13702, 13703, 13704, 13705, 13706, 13707, 13708, 13709, 13710, 13711, 13712, 13713, 13714, 13715, 13716, 13717, 13718, 13719, 13720, 13721, 13722, 13723, 13724, 13725, 13726, 13727, 13728, 13729, 13730, 13731, 13732, 13733, 13734]\n"
     ]
    }
   ],
   "source": [
    "print H37Rv_coords_tag_map[10][1] #20bp left/right of H37Rv start coordinate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[13975, 13976, 13977, 13978, 13979, 13980, 13981, 13982, 13983, 13984, 13985, 13986, 13987, 13988, 13989, 13990, 13991, 13992, 13993, 13994, 13995, 13996, 13997, 13998, 13999, 14000, 14001, 14002, 14003, 14004, 14005, 14006, 14007, 14008, 14009, 14010, 14011, 14012, 14013, 14014, 14015]\n"
     ]
    }
   ],
   "source": [
    "print H37Rv_coords_tag_map[10][2] #20bp left/right of H37Rv stop coordinate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Refseq_ID</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>Stop</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>Locus</th>\n",
       "      <th>Name</th>\n",
       "      <th>Function</th>\n",
       "      <th>Product</th>\n",
       "      <th>Comments</th>\n",
       "      <th>UniProt_AC</th>\n",
       "      <th>Functional_Category</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>NC_000962.3</td>\n",
       "      <td>CDS</td>\n",
       "      <td>13714</td>\n",
       "      <td>13995</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Rv0011c</td>\n",
       "      <td>Rv0011c</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Probable conserved transmembrane protein</td>\n",
       "      <td>Rv0011c, (MTCY10H4.11c), len: 93 aa. Probable ...</td>\n",
       "      <td>P9WP57</td>\n",
       "      <td>cell wall and cell processes</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      Refseq_ID Feature  Start   Stop Score Strand  Frame    Locus     Name  \\\n",
       "10  NC_000962.3     CDS  13714  13995     .      -    0.0  Rv0011c  Rv0011c   \n",
       "\n",
       "   Function                                   Product  \\\n",
       "10  Unknown  Probable conserved transmembrane protein   \n",
       "\n",
       "                                             Comments UniProt_AC  \\\n",
       "10  Rv0011c, (MTCY10H4.11c), len: 93 aa. Probable ...     P9WP57   \n",
       "\n",
       "             Functional_Category  \n",
       "10  cell wall and cell processes  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mycobrowser_table[mycobrowser_table.Locus == 'Rv0011c']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Import subsystem classification from SEED"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "SEED_annotation = pd.read_csv('//n/data1/hms/dbmi/farhat/Roger/inhost_TB_dynamics_project/CSV_files/pathway_analysis/SEED_H37Rv_83332_1_downloaded_4_1_19.tsv' , sep = '\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Feature ID</th>\n",
       "      <th>Type</th>\n",
       "      <th>Contig</th>\n",
       "      <th>Start</th>\n",
       "      <th>Stop</th>\n",
       "      <th>Frame</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Length (bp)</th>\n",
       "      <th>Function</th>\n",
       "      <th>Subsystem</th>\n",
       "      <th>NCBI GI</th>\n",
       "      <th>locus</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>fig|83332.1.pbs.3</td>\n",
       "      <td>PBS</td>\n",
       "      <td>NC_000962</td>\n",
       "      <td>2396081</td>\n",
       "      <td>2396096</td>\n",
       "      <td>2</td>\n",
       "      <td>+</td>\n",
       "      <td>16</td>\n",
       "      <td>NrdR_Proteobacteria</td>\n",
       "      <td>- none -</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>fig|83332.1.pbs.4</td>\n",
       "      <td>PBS</td>\n",
       "      <td>NC_000962</td>\n",
       "      <td>4265613</td>\n",
       "      <td>4265628</td>\n",
       "      <td>3</td>\n",
       "      <td>+</td>\n",
       "      <td>16</td>\n",
       "      <td>NrdR_Proteobacteria</td>\n",
       "      <td>- none -</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>fig|83332.1.peg.1</td>\n",
       "      <td>CDS</td>\n",
       "      <td>NC_000962</td>\n",
       "      <td>1</td>\n",
       "      <td>1524</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>1524</td>\n",
       "      <td>Chromosomal replication initiator protein DnaA</td>\n",
       "      <td>DNA replication cluster 1</td>\n",
       "      <td>gi|15607143</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>fig|83332.1.peg.2</td>\n",
       "      <td>CDS</td>\n",
       "      <td>NC_000962</td>\n",
       "      <td>2052</td>\n",
       "      <td>3260</td>\n",
       "      <td>3</td>\n",
       "      <td>+</td>\n",
       "      <td>1209</td>\n",
       "      <td>DNA polymerase III beta subunit (EC 2.7.7.7)</td>\n",
       "      <td>DNA replication cluster 1</td>\n",
       "      <td>gi|15607144</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>fig|83332.1.peg.3</td>\n",
       "      <td>CDS</td>\n",
       "      <td>NC_000962</td>\n",
       "      <td>3280</td>\n",
       "      <td>4437</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>1158</td>\n",
       "      <td>DNA recombination and repair protein RecF</td>\n",
       "      <td>DNA repair, bacterial RecFOR pathway; &lt;br&gt;DNA ...</td>\n",
       "      <td>gi|15607145</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          Feature ID Type     Contig    Start     Stop  Frame Strand  \\\n",
       "0  fig|83332.1.pbs.3  PBS  NC_000962  2396081  2396096      2      +   \n",
       "1  fig|83332.1.pbs.4  PBS  NC_000962  4265613  4265628      3      +   \n",
       "2  fig|83332.1.peg.1  CDS  NC_000962        1     1524      1      +   \n",
       "3  fig|83332.1.peg.2  CDS  NC_000962     2052     3260      3      +   \n",
       "4  fig|83332.1.peg.3  CDS  NC_000962     3280     4437      1      +   \n",
       "\n",
       "   Length (bp)                                        Function  \\\n",
       "0           16                             NrdR_Proteobacteria   \n",
       "1           16                             NrdR_Proteobacteria   \n",
       "2         1524  Chromosomal replication initiator protein DnaA   \n",
       "3         1209    DNA polymerase III beta subunit (EC 2.7.7.7)   \n",
       "4         1158       DNA recombination and repair protein RecF   \n",
       "\n",
       "                                           Subsystem      NCBI GI locus  \n",
       "0                                           - none -                     \n",
       "1                                           - none -                     \n",
       "2                          DNA replication cluster 1  gi|15607143        \n",
       "3                          DNA replication cluster 1  gi|15607144        \n",
       "4  DNA repair, bacterial RecFOR pathway; <br>DNA ...  gi|15607145        "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SEED_annotation.head(n=5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Find all unique subsystem classifications"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "subsystem_types = []\n",
    "\n",
    "#subset SEED annotation to genomic regions with at least 1 subsystem\n",
    "SEED_annotation_with_subsystems = SEED_annotation[SEED_annotation.Subsystem != '- none -']\n",
    "\n",
    "for region_with_subsytem_i in SEED_annotation_with_subsystems.Subsystem.index:\n",
    "    \n",
    "    #subset to subsystems for genomic region (CDS, RNA, etc.)\n",
    "    subsystems_for_region_i = SEED_annotation_with_subsystems.loc[region_with_subsytem_i , 'Subsystem']\n",
    "    \n",
    "    #break up subsystem list into different subsystems (if there is more than 1)\n",
    "    subsystems_for_region_i = subsystems_for_region_i.split(';')\n",
    "    \n",
    "    #append to list of all subsytems detected across all genomic regions\n",
    "    subsystem_types = subsystem_types + subsystems_for_region_i\n",
    "    \n",
    "#find all unique subsystems present in the H37Rv genome\n",
    "subsystem_types = list(set(subsystem_types))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "576"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(subsystem_types)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Assign each subsystem to any relevant loci from Mycobrowser table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "subsystem_gene_dict = {}\n",
    "\n",
    "for subsystem in subsystem_types:\n",
    "    \n",
    "    #create emtpy list for genes identified as part of subsystem\n",
    "    genes_in_subsystem = []\n",
    "    \n",
    "    #find all genomic regions from SEED annotation table assigned to that subsystem\n",
    "    for genomic_region_i in SEED_annotation.index:\n",
    "        \n",
    "        #if genomic region is part of subsytem\n",
    "        if subsystem in SEED_annotation.loc[genomic_region_i , 'Subsystem'].split(';'):\n",
    "            \n",
    "            #find coordinates and map to H37Rv tag using Mycobrowser table\n",
    "            SEED_region_coords = np.sort( list( SEED_annotation.loc[genomic_region_i , ['Start' , 'Stop']] ) )\n",
    "            start_coord = SEED_region_coords[0]\n",
    "            stop_coord = SEED_region_coords[1]\n",
    "            \n",
    "            #check to see if coordinates for SEED feature region match to a region from H37Rv Mycobrowser annotation\n",
    "            for H37Rv_locus_and_coords in H37Rv_coords_tag_map:\n",
    "                \n",
    "                #check to see if SEED region start/stop coordinates in in the same region as Mycobroswer annotation\n",
    "                if (start_coord in H37Rv_locus_and_coords[1]) and (stop_coord in H37Rv_locus_and_coords[2]):\n",
    "                    \n",
    "                    genes_in_subsystem.append(H37Rv_locus_and_coords[0])\n",
    "                \n",
    "    #store list of genes identified as being involved in subsystem in dictionary\n",
    "    subsystem_gene_dict[subsystem] = genes_in_subsystem"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "576"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(subsystem_gene_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "subsystem_gene_series = pd.Series( subsystem_gene_dict )\n",
    "\n",
    "#save to subsystem - H37Rv locus tag mappting to CSV\n",
    "subsystem_gene_series.to_csv('/n/data1/hms/dbmi/farhat/Roger/inhost_TB_dynamics_project/CSV_files/pathway_analysis/subsystem_H37Rv_tag_map.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       " <br>Acetyl-CoA fermentation to Butyrate                                     [Rv0468, Rv0860]\n",
       " <br>Adenosyl nucleosidases                                                          [Rv0091]\n",
       " <br>Arginine Biosynthesis -- gjo                                                    [Rv1653]\n",
       " <br>Arginine Biosynthesis extended         [Rv1202, Rv1652, Rv1653, Rv1654, Rv1655, Rv165...\n",
       " <br>Arginine Deiminase Pathway                                              [Rv1656, Rv1657]\n",
       "dtype: object"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "subsystem_gene_series.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load in *in-host* SNPs (with $\\Delta$AF $\\ge$ 70%)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "SNP_variants_within_patients = pd.read_pickle('/n/data1/hms/dbmi/farhat/Roger/inhost_TB_dynamics_project/pickled_files/variant_calling/longitudinal_SNPs/SNPs_between_isolates_delta_70.pkl')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>population</th>\n",
       "      <th>patient_id</th>\n",
       "      <th>ref_position</th>\n",
       "      <th>ref_allele</th>\n",
       "      <th>alt_allele</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>genomic_coord</th>\n",
       "      <th>gene_category</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>alt_AF_diff</th>\n",
       "      <th>SNP_type</th>\n",
       "      <th>AA_change</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CASALI</td>\n",
       "      <td>P251</td>\n",
       "      <td>761139</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>Rv0667</td>\n",
       "      <td>1333</td>\n",
       "      <td>Antibiotic Resistance</td>\n",
       "      <td>rpoB</td>\n",
       "      <td>1.0</td>\n",
       "      <td>N</td>\n",
       "      <td>H445D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CASALI</td>\n",
       "      <td>P251</td>\n",
       "      <td>861378</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>Rv0768</td>\n",
       "      <td>467</td>\n",
       "      <td>Non-Essential</td>\n",
       "      <td>aldA</td>\n",
       "      <td>1.0</td>\n",
       "      <td>N</td>\n",
       "      <td>I156S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BRYANT</td>\n",
       "      <td>2020E</td>\n",
       "      <td>1253207</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>Rv1129c</td>\n",
       "      <td>1328</td>\n",
       "      <td>Non-Essential</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>N</td>\n",
       "      <td>I443T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BRYANT</td>\n",
       "      <td>2020E</td>\n",
       "      <td>2075138</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>Rv1830</td>\n",
       "      <td>298</td>\n",
       "      <td>Non-Essential</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>N</td>\n",
       "      <td>S100T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BRYANT</td>\n",
       "      <td>2020E</td>\n",
       "      <td>2694727</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>Rv2397c</td>\n",
       "      <td>238</td>\n",
       "      <td>Essential</td>\n",
       "      <td>cysA1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>N</td>\n",
       "      <td>F80L</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  population patient_id  ref_position ref_allele alt_allele  gene_id  \\\n",
       "0     CASALI       P251        761139          C          G   Rv0667   \n",
       "1     CASALI       P251        861378          T          G   Rv0768   \n",
       "2     BRYANT      2020E       1253207          A          G  Rv1129c   \n",
       "3     BRYANT      2020E       2075138          T          A   Rv1830   \n",
       "4     BRYANT      2020E       2694727          A          G  Rv2397c   \n",
       "\n",
       "  genomic_coord          gene_category gene_symbol  alt_AF_diff SNP_type  \\\n",
       "0          1333  Antibiotic Resistance        rpoB          1.0        N   \n",
       "1           467          Non-Essential        aldA          1.0        N   \n",
       "2          1328          Non-Essential         NaN          1.0        N   \n",
       "3           298          Non-Essential         NaN          1.0        N   \n",
       "4           238              Essential       cysA1          1.0        N   \n",
       "\n",
       "  AA_change  \n",
       "0     H445D  \n",
       "1     I156S  \n",
       "2     I443T  \n",
       "3     S100T  \n",
       "4      F80L  "
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SNP_variants_within_patients.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(175, 12)"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.shape(SNP_variants_within_patients)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Subset to SNPs in gene regions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "SNP_variants_within_patients = SNP_variants_within_patients[SNP_variants_within_patients.SNP_type != 'I']\n",
    "SNP_variants_within_patients.reset_index(drop = True , inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>population</th>\n",
       "      <th>patient_id</th>\n",
       "      <th>ref_position</th>\n",
       "      <th>ref_allele</th>\n",
       "      <th>alt_allele</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>genomic_coord</th>\n",
       "      <th>gene_category</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>alt_AF_diff</th>\n",
       "      <th>SNP_type</th>\n",
       "      <th>AA_change</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CASALI</td>\n",
       "      <td>P251</td>\n",
       "      <td>761139</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>Rv0667</td>\n",
       "      <td>1333</td>\n",
       "      <td>Antibiotic Resistance</td>\n",
       "      <td>rpoB</td>\n",
       "      <td>1.0</td>\n",
       "      <td>N</td>\n",
       "      <td>H445D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CASALI</td>\n",
       "      <td>P251</td>\n",
       "      <td>861378</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>Rv0768</td>\n",
       "      <td>467</td>\n",
       "      <td>Non-Essential</td>\n",
       "      <td>aldA</td>\n",
       "      <td>1.0</td>\n",
       "      <td>N</td>\n",
       "      <td>I156S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BRYANT</td>\n",
       "      <td>2020E</td>\n",
       "      <td>1253207</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>Rv1129c</td>\n",
       "      <td>1328</td>\n",
       "      <td>Non-Essential</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>N</td>\n",
       "      <td>I443T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BRYANT</td>\n",
       "      <td>2020E</td>\n",
       "      <td>2075138</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>Rv1830</td>\n",
       "      <td>298</td>\n",
       "      <td>Non-Essential</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>N</td>\n",
       "      <td>S100T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BRYANT</td>\n",
       "      <td>2020E</td>\n",
       "      <td>2694727</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>Rv2397c</td>\n",
       "      <td>238</td>\n",
       "      <td>Essential</td>\n",
       "      <td>cysA1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>N</td>\n",
       "      <td>F80L</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  population patient_id  ref_position ref_allele alt_allele  gene_id  \\\n",
       "0     CASALI       P251        761139          C          G   Rv0667   \n",
       "1     CASALI       P251        861378          T          G   Rv0768   \n",
       "2     BRYANT      2020E       1253207          A          G  Rv1129c   \n",
       "3     BRYANT      2020E       2075138          T          A   Rv1830   \n",
       "4     BRYANT      2020E       2694727          A          G  Rv2397c   \n",
       "\n",
       "  genomic_coord          gene_category gene_symbol  alt_AF_diff SNP_type  \\\n",
       "0          1333  Antibiotic Resistance        rpoB          1.0        N   \n",
       "1           467          Non-Essential        aldA          1.0        N   \n",
       "2          1328          Non-Essential         NaN          1.0        N   \n",
       "3           298          Non-Essential         NaN          1.0        N   \n",
       "4           238              Essential       cysA1          1.0        N   \n",
       "\n",
       "  AA_change  \n",
       "0     H445D  \n",
       "1     I156S  \n",
       "2     I443T  \n",
       "3     S100T  \n",
       "4      F80L  "
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SNP_variants_within_patients.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Number of SNPs for each (subsystem) pathway"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "subsystem_SNP_count = {}\n",
    "\n",
    "for subsystem in subsystem_gene_series.index:\n",
    "    \n",
    "    SNPs_in_subsystem_genes = 0\n",
    "    \n",
    "    for gene_id in subsystem_gene_series[subsystem]:\n",
    "        \n",
    "        SNPs_in_subsystem_genes = SNPs_in_subsystem_genes + sum(SNP_variants_within_patients.gene_id == gene_id)\n",
    "    \n",
    "    subsystem_SNP_count[subsystem] = SNPs_in_subsystem_genes\n",
    "    \n",
    "#convert dict to series & sort\n",
    "subsystem_SNP_count = pd.Series( subsystem_SNP_count )\n",
    "subsystem_SNP_count.sort_values(ascending = False, inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       " <br>RNA polymerase bacterial                                                            9\n",
       "Mycobacterium virulence operon involved in DNA transcription                             9\n",
       "DNA gyrase subunits                                                                      4\n",
       " <br>Oxidative stress                                                                    4\n",
       " <br>DNA replication cluster 1                                                           4\n",
       " <br>DNA topoisomerases, Type II, ATP-dependent                                          4\n",
       "Oxidative stress                                                                         4\n",
       " <br>Protection from Reactive Oxygen Species                                             4\n",
       " <br>Resistance to fluoroquinolones                                                      4\n",
       "Biotin biosynthesis                                                                      3\n",
       "Ribosome LSU bacterial                                                                   3\n",
       "NAD and NADP cofactor biosynthesis global                                                2\n",
       "Glycerolipid and Glycerophospholipid Metabolism in Bacteria                              2\n",
       " <br>Translation initiation factors bacterial                                            2\n",
       "ESAT-6 proteins secretion system in Actinobacteria                                       2\n",
       "CBSS-164757.7.peg.5020                                                                   2\n",
       " <br>Coenzyme B12 biosynthesis                                                           2\n",
       " <br>Redox-dependent regulation of nucleus processes                                     2\n",
       "Cobalamin synthesis                                                                      2\n",
       " <br>cAMP signaling in bacteria                                                          2\n",
       "Mycobacterium virulence operon involved in protein synthesis (LSU ribosomal proteins)    2\n",
       "dtype: int64"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "subsystem_SNP_count[subsystem_SNP_count > 1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>population</th>\n",
       "      <th>patient_id</th>\n",
       "      <th>ref_position</th>\n",
       "      <th>ref_allele</th>\n",
       "      <th>alt_allele</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>genomic_coord</th>\n",
       "      <th>gene_category</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>alt_AF_diff</th>\n",
       "      <th>SNP_type</th>\n",
       "      <th>AA_change</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>CASALI</td>\n",
       "      <td>P233</td>\n",
       "      <td>4301121</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>Rv3826</td>\n",
       "      <td>1310</td>\n",
       "      <td>Non-Essential</td>\n",
       "      <td>fadD23</td>\n",
       "      <td>0.98</td>\n",
       "      <td>N</td>\n",
       "      <td>P437L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>GUERRA</td>\n",
       "      <td>KPS_79</td>\n",
       "      <td>485640</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>Rv0404</td>\n",
       "      <td>1664</td>\n",
       "      <td>Essential</td>\n",
       "      <td>fadD30</td>\n",
       "      <td>0.89</td>\n",
       "      <td>N</td>\n",
       "      <td>A555E</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>104</th>\n",
       "      <td>CETR</td>\n",
       "      <td>3451</td>\n",
       "      <td>3301183</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>Rv2950c</td>\n",
       "      <td>1273</td>\n",
       "      <td>Non-Essential</td>\n",
       "      <td>fadD29</td>\n",
       "      <td>0.73</td>\n",
       "      <td>N</td>\n",
       "      <td>V425F</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    population patient_id  ref_position ref_allele alt_allele  gene_id  \\\n",
       "11      CASALI       P233       4301121          C          T   Rv3826   \n",
       "69      GUERRA     KPS_79        485640          C          A   Rv0404   \n",
       "104       CETR       3451       3301183          C          A  Rv2950c   \n",
       "\n",
       "    genomic_coord  gene_category gene_symbol  alt_AF_diff SNP_type AA_change  \n",
       "11           1310  Non-Essential      fadD23         0.98        N     P437L  \n",
       "69           1664      Essential      fadD30         0.89        N     A555E  \n",
       "104          1273  Non-Essential      fadD29         0.73        N     V425F  "
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "subsystem_id = 'Biotin biosynthesis'\n",
    "SNP_variants_within_patients_subsystem_filter = [SNP_variants_within_patients.loc[SNP_i , 'gene_id'] in subsystem_gene_series[subsystem_id] for SNP_i in SNP_variants_within_patients.index ]\n",
    "SNP_variants_within_patients[SNP_variants_within_patients_subsystem_filter]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Number of patients with at least 1 in-host SNP in each (subsystem) pathway"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "subsystem_subjects_with_SNP_count = {}\n",
    "\n",
    "for subsystem in subsystem_gene_series.index:\n",
    "    \n",
    "    subjects_with_SNP_in_subsystem_genes = []\n",
    "    \n",
    "    for gene_id in subsystem_gene_series[subsystem]:\n",
    "        \n",
    "        subjects_with_SNP_in_subsystem_genes = subjects_with_SNP_in_subsystem_genes + list( SNP_variants_within_patients[SNP_variants_within_patients.gene_id == gene_id].patient_id ) \n",
    "    \n",
    "    subsystem_subjects_with_SNP_count[subsystem] = len(list(set(subjects_with_SNP_in_subsystem_genes)))\n",
    "    \n",
    "#convert dict to series & sort\n",
    "subsystem_subjects_with_SNP_count = pd.Series( subsystem_subjects_with_SNP_count )\n",
    "subsystem_subjects_with_SNP_count.sort_values(ascending = False, inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       " <br>RNA polymerase bacterial                                   7\n",
       "Mycobacterium virulence operon involved in DNA transcription    7\n",
       "DNA gyrase subunits                                             4\n",
       " <br>Oxidative stress                                           4\n",
       " <br>DNA replication cluster 1                                  4\n",
       " <br>DNA topoisomerases, Type II, ATP-dependent                 4\n",
       "Oxidative stress                                                4\n",
       " <br>Protection from Reactive Oxygen Species                    4\n",
       " <br>Resistance to fluoroquinolones                             4\n",
       "Biotin biosynthesis                                             3\n",
       "Ribosome LSU bacterial                                          3\n",
       "Glycerolipid and Glycerophospholipid Metabolism in Bacteria     2\n",
       "ESAT-6 proteins secretion system in Actinobacteria              2\n",
       " <br>Coenzyme B12 biosynthesis                                  2\n",
       "CBSS-164757.7.peg.5020                                          2\n",
       "Cobalamin synthesis                                             2\n",
       "dtype: int64"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "subsystem_subjects_with_SNP_count[subsystem_subjects_with_SNP_count > 1]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Display all *subsystems* pathways in which multiple individuals have at least 1 SNP along with associated *in-host* SNPs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/rv76/anaconda2/lib/python2.7/site-packages/ipykernel_launcher.py:13: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "  del sys.path[0]\n",
      "/home/rv76/anaconda2/lib/python2.7/site-packages/ipykernel_launcher.py:14: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "  \n"
     ]
    }
   ],
   "source": [
    "#subsystems with where multiple subjects have a SNP & count of individuals with at least 1 SNP\n",
    "parallel_mutation_subsytems = subsystem_subjects_with_SNP_count[subsystem_subjects_with_SNP_count > 1]\n",
    "\n",
    "#store subsystems showing evidence of parallel convergence & associated mutations\n",
    "parallel_mutation_subsytems_with_SNPs_DF = pd.DataFrame(columns = list(SNP_variants_within_patients.columns)+['pathway','Num Subjects w/ SNP in pathway'])\n",
    "\n",
    "for subsystem_id, num_individuals in zip(list(parallel_mutation_subsytems.index) , list(parallel_mutation_subsytems.values)):\n",
    "\n",
    "    SNP_variants_within_patients_subsystem_filter = [SNP_variants_within_patients.loc[SNP_i , 'gene_id'] in subsystem_gene_series[subsystem_id] for SNP_i in SNP_variants_within_patients.index ]\n",
    "    SNP_variants_within_patients_detected_in_subsystem = SNP_variants_within_patients[SNP_variants_within_patients_subsystem_filter]\n",
    "    \n",
    "    #store in DataFrame\n",
    "    SNP_variants_within_patients_detected_in_subsystem['pathway'] = [subsystem_id]*np.shape(SNP_variants_within_patients_detected_in_subsystem)[0] #column for subsystem ID\n",
    "    SNP_variants_within_patients_detected_in_subsystem['Num Subjects w/ SNP in pathway'] = [num_individuals]*np.shape(SNP_variants_within_patients_detected_in_subsystem)[0] #column for num subjects with SNP in pathway\n",
    "    parallel_mutation_subsytems_with_SNPs_DF = parallel_mutation_subsytems_with_SNPs_DF.append(SNP_variants_within_patients_detected_in_subsystem) #SNPs detected within pathway\n",
    "    \n",
    "#re-order columns\n",
    "column_order = ['pathway', 'Num Subjects w/ SNP in pathway','population', 'patient_id', 'ref_position', 'genomic_coord', 'ref_allele', 'alt_allele', 'gene_id', 'gene_category', 'gene_symbol', 'alt_AF_diff', 'SNP_type', 'AA_change']\n",
    "parallel_mutation_subsytems_with_SNPs_DF = parallel_mutation_subsytems_with_SNPs_DF.loc[: , column_order]\n",
    "\n",
    "#output as CSV\n",
    "parallel_mutation_subsytems_with_SNPs_DF.to_csv('/n/data1/hms/dbmi/farhat/Roger/inhost_TB_dynamics_project/CSV_files/pathway_analysis/subsystems_with_parallel_SNPs.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pathway</th>\n",
       "      <th>Num Subjects w/ SNP in pathway</th>\n",
       "      <th>population</th>\n",
       "      <th>patient_id</th>\n",
       "      <th>ref_position</th>\n",
       "      <th>genomic_coord</th>\n",
       "      <th>ref_allele</th>\n",
       "      <th>alt_allele</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_category</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>alt_AF_diff</th>\n",
       "      <th>SNP_type</th>\n",
       "      <th>AA_change</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>&lt;br&gt;RNA polymerase bacterial</td>\n",
       "      <td>7</td>\n",
       "      <td>CASALI</td>\n",
       "      <td>P251</td>\n",
       "      <td>761139</td>\n",
       "      <td>1333</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>Rv0667</td>\n",
       "      <td>Antibiotic Resistance</td>\n",
       "      <td>rpoB</td>\n",
       "      <td>1.00</td>\n",
       "      <td>N</td>\n",
       "      <td>H445D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>&lt;br&gt;RNA polymerase bacterial</td>\n",
       "      <td>7</td>\n",
       "      <td>CASALI</td>\n",
       "      <td>P052</td>\n",
       "      <td>760314</td>\n",
       "      <td>508</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>Rv0667</td>\n",
       "      <td>Antibiotic Resistance</td>\n",
       "      <td>rpoB</td>\n",
       "      <td>0.98</td>\n",
       "      <td>N</td>\n",
       "      <td>V170F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>&lt;br&gt;RNA polymerase bacterial</td>\n",
       "      <td>7</td>\n",
       "      <td>CASALI</td>\n",
       "      <td>P052</td>\n",
       "      <td>764819</td>\n",
       "      <td>1450</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>Rv0668</td>\n",
       "      <td>Antibiotic Resistance</td>\n",
       "      <td>rpoC</td>\n",
       "      <td>0.96</td>\n",
       "      <td>N</td>\n",
       "      <td>W484G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>&lt;br&gt;RNA polymerase bacterial</td>\n",
       "      <td>7</td>\n",
       "      <td>WALKER</td>\n",
       "      <td>P000059</td>\n",
       "      <td>760314</td>\n",
       "      <td>508</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>Rv0667</td>\n",
       "      <td>Antibiotic Resistance</td>\n",
       "      <td>rpoB</td>\n",
       "      <td>0.85</td>\n",
       "      <td>N</td>\n",
       "      <td>V170F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>&lt;br&gt;RNA polymerase bacterial</td>\n",
       "      <td>7</td>\n",
       "      <td>WALKER</td>\n",
       "      <td>P000059</td>\n",
       "      <td>765617</td>\n",
       "      <td>2248</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>Rv0668</td>\n",
       "      <td>Antibiotic Resistance</td>\n",
       "      <td>rpoC</td>\n",
       "      <td>0.98</td>\n",
       "      <td>N</td>\n",
       "      <td>E750Q</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>&lt;br&gt;RNA polymerase bacterial</td>\n",
       "      <td>7</td>\n",
       "      <td>WALKER</td>\n",
       "      <td>P000227</td>\n",
       "      <td>761277</td>\n",
       "      <td>1471</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>Rv0667</td>\n",
       "      <td>Antibiotic Resistance</td>\n",
       "      <td>rpoB</td>\n",
       "      <td>0.99</td>\n",
       "      <td>N</td>\n",
       "      <td>I491F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>105</th>\n",
       "      <td>&lt;br&gt;RNA polymerase bacterial</td>\n",
       "      <td>7</td>\n",
       "      <td>CETR</td>\n",
       "      <td>2688</td>\n",
       "      <td>767123</td>\n",
       "      <td>3754</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>Rv0668</td>\n",
       "      <td>Antibiotic Resistance</td>\n",
       "      <td>rpoC</td>\n",
       "      <td>1.00</td>\n",
       "      <td>N</td>\n",
       "      <td>V1252L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>122</th>\n",
       "      <td>&lt;br&gt;RNA polymerase bacterial</td>\n",
       "      <td>7</td>\n",
       "      <td>CETR</td>\n",
       "      <td>2511</td>\n",
       "      <td>761140</td>\n",
       "      <td>1334</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>Rv0667</td>\n",
       "      <td>Antibiotic Resistance</td>\n",
       "      <td>rpoB</td>\n",
       "      <td>1.00</td>\n",
       "      <td>N</td>\n",
       "      <td>H445P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129</th>\n",
       "      <td>&lt;br&gt;RNA polymerase bacterial</td>\n",
       "      <td>7</td>\n",
       "      <td>GUERRA</td>\n",
       "      <td>KPS_82</td>\n",
       "      <td>761139</td>\n",
       "      <td>1333</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>Rv0667</td>\n",
       "      <td>Antibiotic Resistance</td>\n",
       "      <td>rpoB</td>\n",
       "      <td>0.99</td>\n",
       "      <td>N</td>\n",
       "      <td>H445D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Mycobacterium virulence operon involved in DNA...</td>\n",
       "      <td>7</td>\n",
       "      <td>CASALI</td>\n",
       "      <td>P251</td>\n",
       "      <td>761139</td>\n",
       "      <td>1333</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>Rv0667</td>\n",
       "      <td>Antibiotic Resistance</td>\n",
       "      <td>rpoB</td>\n",
       "      <td>1.00</td>\n",
       "      <td>N</td>\n",
       "      <td>H445D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>Mycobacterium virulence operon involved in DNA...</td>\n",
       "      <td>7</td>\n",
       "      <td>CASALI</td>\n",
       "      <td>P052</td>\n",
       "      <td>760314</td>\n",
       "      <td>508</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>Rv0667</td>\n",
       "      <td>Antibiotic Resistance</td>\n",
       "      <td>rpoB</td>\n",
       "      <td>0.98</td>\n",
       "      <td>N</td>\n",
       "      <td>V170F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>Mycobacterium virulence operon involved in DNA...</td>\n",
       "      <td>7</td>\n",
       "      <td>CASALI</td>\n",
       "      <td>P052</td>\n",
       "      <td>764819</td>\n",
       "      <td>1450</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>Rv0668</td>\n",
       "      <td>Antibiotic Resistance</td>\n",
       "      <td>rpoC</td>\n",
       "      <td>0.96</td>\n",
       "      <td>N</td>\n",
       "      <td>W484G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>Mycobacterium virulence operon involved in DNA...</td>\n",
       "      <td>7</td>\n",
       "      <td>WALKER</td>\n",
       "      <td>P000059</td>\n",
       "      <td>760314</td>\n",
       "      <td>508</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>Rv0667</td>\n",
       "      <td>Antibiotic Resistance</td>\n",
       "      <td>rpoB</td>\n",
       "      <td>0.85</td>\n",
       "      <td>N</td>\n",
       "      <td>V170F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>Mycobacterium virulence operon involved in DNA...</td>\n",
       "      <td>7</td>\n",
       "      <td>WALKER</td>\n",
       "      <td>P000059</td>\n",
       "      <td>765617</td>\n",
       "      <td>2248</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>Rv0668</td>\n",
       "      <td>Antibiotic Resistance</td>\n",
       "      <td>rpoC</td>\n",
       "      <td>0.98</td>\n",
       "      <td>N</td>\n",
       "      <td>E750Q</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>Mycobacterium virulence operon involved in DNA...</td>\n",
       "      <td>7</td>\n",
       "      <td>WALKER</td>\n",
       "      <td>P000227</td>\n",
       "      <td>761277</td>\n",
       "      <td>1471</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>Rv0667</td>\n",
       "      <td>Antibiotic Resistance</td>\n",
       "      <td>rpoB</td>\n",
       "      <td>0.99</td>\n",
       "      <td>N</td>\n",
       "      <td>I491F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>105</th>\n",
       "      <td>Mycobacterium virulence operon involved in DNA...</td>\n",
       "      <td>7</td>\n",
       "      <td>CETR</td>\n",
       "      <td>2688</td>\n",
       "      <td>767123</td>\n",
       "      <td>3754</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>Rv0668</td>\n",
       "      <td>Antibiotic Resistance</td>\n",
       "      <td>rpoC</td>\n",
       "      <td>1.00</td>\n",
       "      <td>N</td>\n",
       "      <td>V1252L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>122</th>\n",
       "      <td>Mycobacterium virulence operon involved in DNA...</td>\n",
       "      <td>7</td>\n",
       "      <td>CETR</td>\n",
       "      <td>2511</td>\n",
       "      <td>761140</td>\n",
       "      <td>1334</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>Rv0667</td>\n",
       "      <td>Antibiotic Resistance</td>\n",
       "      <td>rpoB</td>\n",
       "      <td>1.00</td>\n",
       "      <td>N</td>\n",
       "      <td>H445P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129</th>\n",
       "      <td>Mycobacterium virulence operon involved in DNA...</td>\n",
       "      <td>7</td>\n",
       "      <td>GUERRA</td>\n",
       "      <td>KPS_82</td>\n",
       "      <td>761139</td>\n",
       "      <td>1333</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>Rv0667</td>\n",
       "      <td>Antibiotic Resistance</td>\n",
       "      <td>rpoB</td>\n",
       "      <td>0.99</td>\n",
       "      <td>N</td>\n",
       "      <td>H445D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>DNA gyrase subunits</td>\n",
       "      <td>4</td>\n",
       "      <td>WALKER</td>\n",
       "      <td>P000155</td>\n",
       "      <td>7521</td>\n",
       "      <td>220</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>Rv0006</td>\n",
       "      <td>Antibiotic Resistance</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>1.00</td>\n",
       "      <td>N</td>\n",
       "      <td>A74S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>DNA gyrase subunits</td>\n",
       "      <td>4</td>\n",
       "      <td>WALKER</td>\n",
       "      <td>P000227</td>\n",
       "      <td>7582</td>\n",
       "      <td>281</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>Rv0006</td>\n",
       "      <td>Antibiotic Resistance</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>1.00</td>\n",
       "      <td>N</td>\n",
       "      <td>D94G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100</th>\n",
       "      <td>DNA gyrase subunits</td>\n",
       "      <td>4</td>\n",
       "      <td>CETR</td>\n",
       "      <td>3451</td>\n",
       "      <td>9582</td>\n",
       "      <td>2281</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>Rv0006</td>\n",
       "      <td>Antibiotic Resistance</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>0.71</td>\n",
       "      <td>N</td>\n",
       "      <td>R761G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>130</th>\n",
       "      <td>DNA gyrase subunits</td>\n",
       "      <td>4</td>\n",
       "      <td>CETR</td>\n",
       "      <td>1960</td>\n",
       "      <td>7581</td>\n",
       "      <td>280</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>Rv0006</td>\n",
       "      <td>Antibiotic Resistance</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>0.99</td>\n",
       "      <td>N</td>\n",
       "      <td>D94N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>&lt;br&gt;Oxidative stress</td>\n",
       "      <td>4</td>\n",
       "      <td>GUERRA</td>\n",
       "      <td>KPS_37</td>\n",
       "      <td>2154154</td>\n",
       "      <td>1958</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>Rv1908c</td>\n",
       "      <td>Antibiotic Resistance</td>\n",
       "      <td>katG</td>\n",
       "      <td>1.00</td>\n",
       "      <td>N</td>\n",
       "      <td>L653Q</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>&lt;br&gt;Oxidative stress</td>\n",
       "      <td>4</td>\n",
       "      <td>WALKER</td>\n",
       "      <td>P000226</td>\n",
       "      <td>2154695</td>\n",
       "      <td>1417</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>Rv1908c</td>\n",
       "      <td>Antibiotic Resistance</td>\n",
       "      <td>katG</td>\n",
       "      <td>0.70</td>\n",
       "      <td>N</td>\n",
       "      <td>V473F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92</th>\n",
       "      <td>&lt;br&gt;Oxidative stress</td>\n",
       "      <td>4</td>\n",
       "      <td>GUERRA</td>\n",
       "      <td>KPS_24</td>\n",
       "      <td>2155637</td>\n",
       "      <td>475</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>Rv1908c</td>\n",
       "      <td>Antibiotic Resistance</td>\n",
       "      <td>katG</td>\n",
       "      <td>1.00</td>\n",
       "      <td>N</td>\n",
       "      <td>L159F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>126</th>\n",
       "      <td>&lt;br&gt;Oxidative stress</td>\n",
       "      <td>4</td>\n",
       "      <td>WALKER</td>\n",
       "      <td>P000189</td>\n",
       "      <td>2155573</td>\n",
       "      <td>539</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>Rv1908c</td>\n",
       "      <td>Antibiotic Resistance</td>\n",
       "      <td>katG</td>\n",
       "      <td>1.00</td>\n",
       "      <td>N</td>\n",
       "      <td>T180K</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>&lt;br&gt;DNA replication cluster 1</td>\n",
       "      <td>4</td>\n",
       "      <td>WALKER</td>\n",
       "      <td>P000155</td>\n",
       "      <td>7521</td>\n",
       "      <td>220</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>Rv0006</td>\n",
       "      <td>Antibiotic Resistance</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>1.00</td>\n",
       "      <td>N</td>\n",
       "      <td>A74S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>&lt;br&gt;DNA replication cluster 1</td>\n",
       "      <td>4</td>\n",
       "      <td>WALKER</td>\n",
       "      <td>P000227</td>\n",
       "      <td>7582</td>\n",
       "      <td>281</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>Rv0006</td>\n",
       "      <td>Antibiotic Resistance</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>1.00</td>\n",
       "      <td>N</td>\n",
       "      <td>D94G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100</th>\n",
       "      <td>&lt;br&gt;DNA replication cluster 1</td>\n",
       "      <td>4</td>\n",
       "      <td>CETR</td>\n",
       "      <td>3451</td>\n",
       "      <td>9582</td>\n",
       "      <td>2281</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>Rv0006</td>\n",
       "      <td>Antibiotic Resistance</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>0.71</td>\n",
       "      <td>N</td>\n",
       "      <td>R761G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>130</th>\n",
       "      <td>&lt;br&gt;DNA replication cluster 1</td>\n",
       "      <td>4</td>\n",
       "      <td>CETR</td>\n",
       "      <td>1960</td>\n",
       "      <td>7581</td>\n",
       "      <td>280</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>Rv0006</td>\n",
       "      <td>Antibiotic Resistance</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>0.99</td>\n",
       "      <td>N</td>\n",
       "      <td>D94N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>&lt;br&gt;DNA topoisomerases, Type II, ATP-dependent</td>\n",
       "      <td>4</td>\n",
       "      <td>WALKER</td>\n",
       "      <td>P000155</td>\n",
       "      <td>7521</td>\n",
       "      <td>220</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>Rv0006</td>\n",
       "      <td>Antibiotic Resistance</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>1.00</td>\n",
       "      <td>N</td>\n",
       "      <td>A74S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>&lt;br&gt;DNA topoisomerases, Type II, ATP-dependent</td>\n",
       "      <td>4</td>\n",
       "      <td>WALKER</td>\n",
       "      <td>P000227</td>\n",
       "      <td>7582</td>\n",
       "      <td>281</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>Rv0006</td>\n",
       "      <td>Antibiotic Resistance</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>1.00</td>\n",
       "      <td>N</td>\n",
       "      <td>D94G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100</th>\n",
       "      <td>&lt;br&gt;DNA topoisomerases, Type II, ATP-dependent</td>\n",
       "      <td>4</td>\n",
       "      <td>CETR</td>\n",
       "      <td>3451</td>\n",
       "      <td>9582</td>\n",
       "      <td>2281</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>Rv0006</td>\n",
       "      <td>Antibiotic Resistance</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>0.71</td>\n",
       "      <td>N</td>\n",
       "      <td>R761G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>130</th>\n",
       "      <td>&lt;br&gt;DNA topoisomerases, Type II, ATP-dependent</td>\n",
       "      <td>4</td>\n",
       "      <td>CETR</td>\n",
       "      <td>1960</td>\n",
       "      <td>7581</td>\n",
       "      <td>280</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>Rv0006</td>\n",
       "      <td>Antibiotic Resistance</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>0.99</td>\n",
       "      <td>N</td>\n",
       "      <td>D94N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Oxidative stress</td>\n",
       "      <td>4</td>\n",
       "      <td>GUERRA</td>\n",
       "      <td>KPS_37</td>\n",
       "      <td>2154154</td>\n",
       "      <td>1958</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>Rv1908c</td>\n",
       "      <td>Antibiotic Resistance</td>\n",
       "      <td>katG</td>\n",
       "      <td>1.00</td>\n",
       "      <td>N</td>\n",
       "      <td>L653Q</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>Oxidative stress</td>\n",
       "      <td>4</td>\n",
       "      <td>WALKER</td>\n",
       "      <td>P000226</td>\n",
       "      <td>2154695</td>\n",
       "      <td>1417</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>Rv1908c</td>\n",
       "      <td>Antibiotic Resistance</td>\n",
       "      <td>katG</td>\n",
       "      <td>0.70</td>\n",
       "      <td>N</td>\n",
       "      <td>V473F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92</th>\n",
       "      <td>Oxidative stress</td>\n",
       "      <td>4</td>\n",
       "      <td>GUERRA</td>\n",
       "      <td>KPS_24</td>\n",
       "      <td>2155637</td>\n",
       "      <td>475</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>Rv1908c</td>\n",
       "      <td>Antibiotic Resistance</td>\n",
       "      <td>katG</td>\n",
       "      <td>1.00</td>\n",
       "      <td>N</td>\n",
       "      <td>L159F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>126</th>\n",
       "      <td>Oxidative stress</td>\n",
       "      <td>4</td>\n",
       "      <td>WALKER</td>\n",
       "      <td>P000189</td>\n",
       "      <td>2155573</td>\n",
       "      <td>539</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>Rv1908c</td>\n",
       "      <td>Antibiotic Resistance</td>\n",
       "      <td>katG</td>\n",
       "      <td>1.00</td>\n",
       "      <td>N</td>\n",
       "      <td>T180K</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>&lt;br&gt;Protection from Reactive Oxygen Species</td>\n",
       "      <td>4</td>\n",
       "      <td>GUERRA</td>\n",
       "      <td>KPS_37</td>\n",
       "      <td>2154154</td>\n",
       "      <td>1958</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>Rv1908c</td>\n",
       "      <td>Antibiotic Resistance</td>\n",
       "      <td>katG</td>\n",
       "      <td>1.00</td>\n",
       "      <td>N</td>\n",
       "      <td>L653Q</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>&lt;br&gt;Protection from Reactive Oxygen Species</td>\n",
       "      <td>4</td>\n",
       "      <td>WALKER</td>\n",
       "      <td>P000226</td>\n",
       "      <td>2154695</td>\n",
       "      <td>1417</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>Rv1908c</td>\n",
       "      <td>Antibiotic Resistance</td>\n",
       "      <td>katG</td>\n",
       "      <td>0.70</td>\n",
       "      <td>N</td>\n",
       "      <td>V473F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92</th>\n",
       "      <td>&lt;br&gt;Protection from Reactive Oxygen Species</td>\n",
       "      <td>4</td>\n",
       "      <td>GUERRA</td>\n",
       "      <td>KPS_24</td>\n",
       "      <td>2155637</td>\n",
       "      <td>475</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>Rv1908c</td>\n",
       "      <td>Antibiotic Resistance</td>\n",
       "      <td>katG</td>\n",
       "      <td>1.00</td>\n",
       "      <td>N</td>\n",
       "      <td>L159F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>126</th>\n",
       "      <td>&lt;br&gt;Protection from Reactive Oxygen Species</td>\n",
       "      <td>4</td>\n",
       "      <td>WALKER</td>\n",
       "      <td>P000189</td>\n",
       "      <td>2155573</td>\n",
       "      <td>539</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>Rv1908c</td>\n",
       "      <td>Antibiotic Resistance</td>\n",
       "      <td>katG</td>\n",
       "      <td>1.00</td>\n",
       "      <td>N</td>\n",
       "      <td>T180K</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>&lt;br&gt;Resistance to fluoroquinolones</td>\n",
       "      <td>4</td>\n",
       "      <td>WALKER</td>\n",
       "      <td>P000155</td>\n",
       "      <td>7521</td>\n",
       "      <td>220</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>Rv0006</td>\n",
       "      <td>Antibiotic Resistance</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>1.00</td>\n",
       "      <td>N</td>\n",
       "      <td>A74S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>&lt;br&gt;Resistance to fluoroquinolones</td>\n",
       "      <td>4</td>\n",
       "      <td>WALKER</td>\n",
       "      <td>P000227</td>\n",
       "      <td>7582</td>\n",
       "      <td>281</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>Rv0006</td>\n",
       "      <td>Antibiotic Resistance</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>1.00</td>\n",
       "      <td>N</td>\n",
       "      <td>D94G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100</th>\n",
       "      <td>&lt;br&gt;Resistance to fluoroquinolones</td>\n",
       "      <td>4</td>\n",
       "      <td>CETR</td>\n",
       "      <td>3451</td>\n",
       "      <td>9582</td>\n",
       "      <td>2281</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>Rv0006</td>\n",
       "      <td>Antibiotic Resistance</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>0.71</td>\n",
       "      <td>N</td>\n",
       "      <td>R761G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>130</th>\n",
       "      <td>&lt;br&gt;Resistance to fluoroquinolones</td>\n",
       "      <td>4</td>\n",
       "      <td>CETR</td>\n",
       "      <td>1960</td>\n",
       "      <td>7581</td>\n",
       "      <td>280</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>Rv0006</td>\n",
       "      <td>Antibiotic Resistance</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>0.99</td>\n",
       "      <td>N</td>\n",
       "      <td>D94N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Biotin biosynthesis</td>\n",
       "      <td>3</td>\n",
       "      <td>CASALI</td>\n",
       "      <td>P233</td>\n",
       "      <td>4301121</td>\n",
       "      <td>1310</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>Rv3826</td>\n",
       "      <td>Non-Essential</td>\n",
       "      <td>fadD23</td>\n",
       "      <td>0.98</td>\n",
       "      <td>N</td>\n",
       "      <td>P437L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>Biotin biosynthesis</td>\n",
       "      <td>3</td>\n",
       "      <td>GUERRA</td>\n",
       "      <td>KPS_79</td>\n",
       "      <td>485640</td>\n",
       "      <td>1664</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>Rv0404</td>\n",
       "      <td>Essential</td>\n",
       "      <td>fadD30</td>\n",
       "      <td>0.89</td>\n",
       "      <td>N</td>\n",
       "      <td>A555E</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>104</th>\n",
       "      <td>Biotin biosynthesis</td>\n",
       "      <td>3</td>\n",
       "      <td>CETR</td>\n",
       "      <td>3451</td>\n",
       "      <td>3301183</td>\n",
       "      <td>1273</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>Rv2950c</td>\n",
       "      <td>Non-Essential</td>\n",
       "      <td>fadD29</td>\n",
       "      <td>0.73</td>\n",
       "      <td>N</td>\n",
       "      <td>V425F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Ribosome LSU bacterial</td>\n",
       "      <td>3</td>\n",
       "      <td>WALKER</td>\n",
       "      <td>P000089</td>\n",
       "      <td>1134561</td>\n",
       "      <td>8</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>Rv1015c</td>\n",
       "      <td>Non-Essential</td>\n",
       "      <td>rplY</td>\n",
       "      <td>1.00</td>\n",
       "      <td>N</td>\n",
       "      <td>K3R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85</th>\n",
       "      <td>Ribosome LSU bacterial</td>\n",
       "      <td>3</td>\n",
       "      <td>BRYANT</td>\n",
       "      <td>2027W</td>\n",
       "      <td>123993</td>\n",
       "      <td>272</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>Rv0105c</td>\n",
       "      <td>Non-Essential</td>\n",
       "      <td>rpmB1</td>\n",
       "      <td>1.00</td>\n",
       "      <td>N</td>\n",
       "      <td>A91V</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>145</th>\n",
       "      <td>Ribosome LSU bacterial</td>\n",
       "      <td>3</td>\n",
       "      <td>CASALI</td>\n",
       "      <td>P305</td>\n",
       "      <td>812422</td>\n",
       "      <td>364</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>Rv0716</td>\n",
       "      <td>Essential</td>\n",
       "      <td>rplE</td>\n",
       "      <td>0.74</td>\n",
       "      <td>N</td>\n",
       "      <td>R122C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Glycerolipid and Glycerophospholipid Metabolis...</td>\n",
       "      <td>2</td>\n",
       "      <td>CASALI</td>\n",
       "      <td>P251</td>\n",
       "      <td>861378</td>\n",
       "      <td>467</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>Rv0768</td>\n",
       "      <td>Non-Essential</td>\n",
       "      <td>aldA</td>\n",
       "      <td>1.00</td>\n",
       "      <td>N</td>\n",
       "      <td>I156S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>Glycerolipid and Glycerophospholipid Metabolis...</td>\n",
       "      <td>2</td>\n",
       "      <td>GUERRA</td>\n",
       "      <td>KPS_23</td>\n",
       "      <td>3330756</td>\n",
       "      <td>606</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>Rv2974c</td>\n",
       "      <td>Non-Essential</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.00</td>\n",
       "      <td>N</td>\n",
       "      <td>Y202*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>ESAT-6 proteins secretion system in Actinobact...</td>\n",
       "      <td>2</td>\n",
       "      <td>WALKER</td>\n",
       "      <td>P000226</td>\n",
       "      <td>4355378</td>\n",
       "      <td>372</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>Rv3877</td>\n",
       "      <td>Essential</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.00</td>\n",
       "      <td>S</td>\n",
       "      <td>I124I</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78</th>\n",
       "      <td>ESAT-6 proteins secretion system in Actinobact...</td>\n",
       "      <td>2</td>\n",
       "      <td>CETR</td>\n",
       "      <td>1948</td>\n",
       "      <td>4346950</td>\n",
       "      <td>470</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>Rv3870</td>\n",
       "      <td>Essential</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.00</td>\n",
       "      <td>N</td>\n",
       "      <td>A157E</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>&lt;br&gt;Coenzyme B12 biosynthesis</td>\n",
       "      <td>2</td>\n",
       "      <td>TRAUNER</td>\n",
       "      <td>P09</td>\n",
       "      <td>2328028</td>\n",
       "      <td>198</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>Rv2070c</td>\n",
       "      <td>Non-Essential</td>\n",
       "      <td>cobK</td>\n",
       "      <td>1.00</td>\n",
       "      <td>S</td>\n",
       "      <td>A66A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>&lt;br&gt;Coenzyme B12 biosynthesis</td>\n",
       "      <td>2</td>\n",
       "      <td>CETR</td>\n",
       "      <td>3673</td>\n",
       "      <td>2322892</td>\n",
       "      <td>341</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>Rv2065</td>\n",
       "      <td>Non-Essential</td>\n",
       "      <td>cobH</td>\n",
       "      <td>1.00</td>\n",
       "      <td>N</td>\n",
       "      <td>R114Q</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>CBSS-164757.7.peg.5020</td>\n",
       "      <td>2</td>\n",
       "      <td>CETR</td>\n",
       "      <td>2307</td>\n",
       "      <td>3994622</td>\n",
       "      <td>1938</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>Rv3554</td>\n",
       "      <td>Non-Essential</td>\n",
       "      <td>fdxB</td>\n",
       "      <td>0.98</td>\n",
       "      <td>S</td>\n",
       "      <td>E646E</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>125</th>\n",
       "      <td>CBSS-164757.7.peg.5020</td>\n",
       "      <td>2</td>\n",
       "      <td>WALKER</td>\n",
       "      <td>P000183</td>\n",
       "      <td>1339741</td>\n",
       "      <td>393</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>Rv1196</td>\n",
       "      <td>PE/PPE</td>\n",
       "      <td>PPE18</td>\n",
       "      <td>0.97</td>\n",
       "      <td>S</td>\n",
       "      <td>T131T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>Cobalamin synthesis</td>\n",
       "      <td>2</td>\n",
       "      <td>TRAUNER</td>\n",
       "      <td>P09</td>\n",
       "      <td>2328028</td>\n",
       "      <td>198</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>Rv2070c</td>\n",
       "      <td>Non-Essential</td>\n",
       "      <td>cobK</td>\n",
       "      <td>1.00</td>\n",
       "      <td>S</td>\n",
       "      <td>A66A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>Cobalamin synthesis</td>\n",
       "      <td>2</td>\n",
       "      <td>CETR</td>\n",
       "      <td>3673</td>\n",
       "      <td>2322892</td>\n",
       "      <td>341</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>Rv2065</td>\n",
       "      <td>Non-Essential</td>\n",
       "      <td>cobH</td>\n",
       "      <td>1.00</td>\n",
       "      <td>N</td>\n",
       "      <td>R114Q</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                               pathway  \\\n",
       "0                         <br>RNA polymerase bacterial   \n",
       "40                        <br>RNA polymerase bacterial   \n",
       "41                        <br>RNA polymerase bacterial   \n",
       "48                        <br>RNA polymerase bacterial   \n",
       "49                        <br>RNA polymerase bacterial   \n",
       "56                        <br>RNA polymerase bacterial   \n",
       "105                       <br>RNA polymerase bacterial   \n",
       "122                       <br>RNA polymerase bacterial   \n",
       "129                       <br>RNA polymerase bacterial   \n",
       "0    Mycobacterium virulence operon involved in DNA...   \n",
       "40   Mycobacterium virulence operon involved in DNA...   \n",
       "41   Mycobacterium virulence operon involved in DNA...   \n",
       "48   Mycobacterium virulence operon involved in DNA...   \n",
       "49   Mycobacterium virulence operon involved in DNA...   \n",
       "56   Mycobacterium virulence operon involved in DNA...   \n",
       "105  Mycobacterium virulence operon involved in DNA...   \n",
       "122  Mycobacterium virulence operon involved in DNA...   \n",
       "129  Mycobacterium virulence operon involved in DNA...   \n",
       "23                                 DNA gyrase subunits   \n",
       "54                                 DNA gyrase subunits   \n",
       "100                                DNA gyrase subunits   \n",
       "130                                DNA gyrase subunits   \n",
       "14                                <br>Oxidative stress   \n",
       "63                                <br>Oxidative stress   \n",
       "92                                <br>Oxidative stress   \n",
       "126                               <br>Oxidative stress   \n",
       "23                       <br>DNA replication cluster 1   \n",
       "54                       <br>DNA replication cluster 1   \n",
       "100                      <br>DNA replication cluster 1   \n",
       "130                      <br>DNA replication cluster 1   \n",
       "23      <br>DNA topoisomerases, Type II, ATP-dependent   \n",
       "54      <br>DNA topoisomerases, Type II, ATP-dependent   \n",
       "100     <br>DNA topoisomerases, Type II, ATP-dependent   \n",
       "130     <br>DNA topoisomerases, Type II, ATP-dependent   \n",
       "14                                    Oxidative stress   \n",
       "63                                    Oxidative stress   \n",
       "92                                    Oxidative stress   \n",
       "126                                   Oxidative stress   \n",
       "14         <br>Protection from Reactive Oxygen Species   \n",
       "63         <br>Protection from Reactive Oxygen Species   \n",
       "92         <br>Protection from Reactive Oxygen Species   \n",
       "126        <br>Protection from Reactive Oxygen Species   \n",
       "23                  <br>Resistance to fluoroquinolones   \n",
       "54                  <br>Resistance to fluoroquinolones   \n",
       "100                 <br>Resistance to fluoroquinolones   \n",
       "130                 <br>Resistance to fluoroquinolones   \n",
       "11                                 Biotin biosynthesis   \n",
       "69                                 Biotin biosynthesis   \n",
       "104                                Biotin biosynthesis   \n",
       "9                               Ribosome LSU bacterial   \n",
       "85                              Ribosome LSU bacterial   \n",
       "145                             Ribosome LSU bacterial   \n",
       "1    Glycerolipid and Glycerophospholipid Metabolis...   \n",
       "33   Glycerolipid and Glycerophospholipid Metabolis...   \n",
       "66   ESAT-6 proteins secretion system in Actinobact...   \n",
       "78   ESAT-6 proteins secretion system in Actinobact...   \n",
       "20                       <br>Coenzyme B12 biosynthesis   \n",
       "35                       <br>Coenzyme B12 biosynthesis   \n",
       "46                              CBSS-164757.7.peg.5020   \n",
       "125                             CBSS-164757.7.peg.5020   \n",
       "20                                 Cobalamin synthesis   \n",
       "35                                 Cobalamin synthesis   \n",
       "\n",
       "    Num Subjects w/ SNP in pathway population patient_id ref_position  \\\n",
       "0                                7     CASALI       P251       761139   \n",
       "40                               7     CASALI       P052       760314   \n",
       "41                               7     CASALI       P052       764819   \n",
       "48                               7     WALKER    P000059       760314   \n",
       "49                               7     WALKER    P000059       765617   \n",
       "56                               7     WALKER    P000227       761277   \n",
       "105                              7       CETR       2688       767123   \n",
       "122                              7       CETR       2511       761140   \n",
       "129                              7     GUERRA     KPS_82       761139   \n",
       "0                                7     CASALI       P251       761139   \n",
       "40                               7     CASALI       P052       760314   \n",
       "41                               7     CASALI       P052       764819   \n",
       "48                               7     WALKER    P000059       760314   \n",
       "49                               7     WALKER    P000059       765617   \n",
       "56                               7     WALKER    P000227       761277   \n",
       "105                              7       CETR       2688       767123   \n",
       "122                              7       CETR       2511       761140   \n",
       "129                              7     GUERRA     KPS_82       761139   \n",
       "23                               4     WALKER    P000155         7521   \n",
       "54                               4     WALKER    P000227         7582   \n",
       "100                              4       CETR       3451         9582   \n",
       "130                              4       CETR       1960         7581   \n",
       "14                               4     GUERRA     KPS_37      2154154   \n",
       "63                               4     WALKER    P000226      2154695   \n",
       "92                               4     GUERRA     KPS_24      2155637   \n",
       "126                              4     WALKER    P000189      2155573   \n",
       "23                               4     WALKER    P000155         7521   \n",
       "54                               4     WALKER    P000227         7582   \n",
       "100                              4       CETR       3451         9582   \n",
       "130                              4       CETR       1960         7581   \n",
       "23                               4     WALKER    P000155         7521   \n",
       "54                               4     WALKER    P000227         7582   \n",
       "100                              4       CETR       3451         9582   \n",
       "130                              4       CETR       1960         7581   \n",
       "14                               4     GUERRA     KPS_37      2154154   \n",
       "63                               4     WALKER    P000226      2154695   \n",
       "92                               4     GUERRA     KPS_24      2155637   \n",
       "126                              4     WALKER    P000189      2155573   \n",
       "14                               4     GUERRA     KPS_37      2154154   \n",
       "63                               4     WALKER    P000226      2154695   \n",
       "92                               4     GUERRA     KPS_24      2155637   \n",
       "126                              4     WALKER    P000189      2155573   \n",
       "23                               4     WALKER    P000155         7521   \n",
       "54                               4     WALKER    P000227         7582   \n",
       "100                              4       CETR       3451         9582   \n",
       "130                              4       CETR       1960         7581   \n",
       "11                               3     CASALI       P233      4301121   \n",
       "69                               3     GUERRA     KPS_79       485640   \n",
       "104                              3       CETR       3451      3301183   \n",
       "9                                3     WALKER    P000089      1134561   \n",
       "85                               3     BRYANT      2027W       123993   \n",
       "145                              3     CASALI       P305       812422   \n",
       "1                                2     CASALI       P251       861378   \n",
       "33                               2     GUERRA     KPS_23      3330756   \n",
       "66                               2     WALKER    P000226      4355378   \n",
       "78                               2       CETR       1948      4346950   \n",
       "20                               2    TRAUNER        P09      2328028   \n",
       "35                               2       CETR       3673      2322892   \n",
       "46                               2       CETR       2307      3994622   \n",
       "125                              2     WALKER    P000183      1339741   \n",
       "20                               2    TRAUNER        P09      2328028   \n",
       "35                               2       CETR       3673      2322892   \n",
       "\n",
       "    genomic_coord ref_allele alt_allele  gene_id          gene_category  \\\n",
       "0            1333          C          G   Rv0667  Antibiotic Resistance   \n",
       "40            508          G          T   Rv0667  Antibiotic Resistance   \n",
       "41           1450          T          G   Rv0668  Antibiotic Resistance   \n",
       "48            508          G          T   Rv0667  Antibiotic Resistance   \n",
       "49           2248          G          C   Rv0668  Antibiotic Resistance   \n",
       "56           1471          A          T   Rv0667  Antibiotic Resistance   \n",
       "105          3754          G          T   Rv0668  Antibiotic Resistance   \n",
       "122          1334          A          C   Rv0667  Antibiotic Resistance   \n",
       "129          1333          C          G   Rv0667  Antibiotic Resistance   \n",
       "0            1333          C          G   Rv0667  Antibiotic Resistance   \n",
       "40            508          G          T   Rv0667  Antibiotic Resistance   \n",
       "41           1450          T          G   Rv0668  Antibiotic Resistance   \n",
       "48            508          G          T   Rv0667  Antibiotic Resistance   \n",
       "49           2248          G          C   Rv0668  Antibiotic Resistance   \n",
       "56           1471          A          T   Rv0667  Antibiotic Resistance   \n",
       "105          3754          G          T   Rv0668  Antibiotic Resistance   \n",
       "122          1334          A          C   Rv0667  Antibiotic Resistance   \n",
       "129          1333          C          G   Rv0667  Antibiotic Resistance   \n",
       "23            220          G          T   Rv0006  Antibiotic Resistance   \n",
       "54            281          A          G   Rv0006  Antibiotic Resistance   \n",
       "100          2281          A          G   Rv0006  Antibiotic Resistance   \n",
       "130           280          G          A   Rv0006  Antibiotic Resistance   \n",
       "14           1958          A          T  Rv1908c  Antibiotic Resistance   \n",
       "63           1417          C          A  Rv1908c  Antibiotic Resistance   \n",
       "92            475          G          A  Rv1908c  Antibiotic Resistance   \n",
       "126           539          G          T  Rv1908c  Antibiotic Resistance   \n",
       "23            220          G          T   Rv0006  Antibiotic Resistance   \n",
       "54            281          A          G   Rv0006  Antibiotic Resistance   \n",
       "100          2281          A          G   Rv0006  Antibiotic Resistance   \n",
       "130           280          G          A   Rv0006  Antibiotic Resistance   \n",
       "23            220          G          T   Rv0006  Antibiotic Resistance   \n",
       "54            281          A          G   Rv0006  Antibiotic Resistance   \n",
       "100          2281          A          G   Rv0006  Antibiotic Resistance   \n",
       "130           280          G          A   Rv0006  Antibiotic Resistance   \n",
       "14           1958          A          T  Rv1908c  Antibiotic Resistance   \n",
       "63           1417          C          A  Rv1908c  Antibiotic Resistance   \n",
       "92            475          G          A  Rv1908c  Antibiotic Resistance   \n",
       "126           539          G          T  Rv1908c  Antibiotic Resistance   \n",
       "14           1958          A          T  Rv1908c  Antibiotic Resistance   \n",
       "63           1417          C          A  Rv1908c  Antibiotic Resistance   \n",
       "92            475          G          A  Rv1908c  Antibiotic Resistance   \n",
       "126           539          G          T  Rv1908c  Antibiotic Resistance   \n",
       "23            220          G          T   Rv0006  Antibiotic Resistance   \n",
       "54            281          A          G   Rv0006  Antibiotic Resistance   \n",
       "100          2281          A          G   Rv0006  Antibiotic Resistance   \n",
       "130           280          G          A   Rv0006  Antibiotic Resistance   \n",
       "11           1310          C          T   Rv3826          Non-Essential   \n",
       "69           1664          C          A   Rv0404              Essential   \n",
       "104          1273          C          A  Rv2950c          Non-Essential   \n",
       "9               8          T          C  Rv1015c          Non-Essential   \n",
       "85            272          G          A  Rv0105c          Non-Essential   \n",
       "145           364          C          T   Rv0716              Essential   \n",
       "1             467          T          G   Rv0768          Non-Essential   \n",
       "33            606          G          C  Rv2974c          Non-Essential   \n",
       "66            372          C          T   Rv3877              Essential   \n",
       "78            470          C          A   Rv3870              Essential   \n",
       "20            198          G          C  Rv2070c          Non-Essential   \n",
       "35            341          G          A   Rv2065          Non-Essential   \n",
       "46           1938          G          A   Rv3554          Non-Essential   \n",
       "125           393          C          G   Rv1196                 PE/PPE   \n",
       "20            198          G          C  Rv2070c          Non-Essential   \n",
       "35            341          G          A   Rv2065          Non-Essential   \n",
       "\n",
       "    gene_symbol  alt_AF_diff SNP_type AA_change  \n",
       "0          rpoB         1.00        N     H445D  \n",
       "40         rpoB         0.98        N     V170F  \n",
       "41         rpoC         0.96        N     W484G  \n",
       "48         rpoB         0.85        N     V170F  \n",
       "49         rpoC         0.98        N     E750Q  \n",
       "56         rpoB         0.99        N     I491F  \n",
       "105        rpoC         1.00        N    V1252L  \n",
       "122        rpoB         1.00        N     H445P  \n",
       "129        rpoB         0.99        N     H445D  \n",
       "0          rpoB         1.00        N     H445D  \n",
       "40         rpoB         0.98        N     V170F  \n",
       "41         rpoC         0.96        N     W484G  \n",
       "48         rpoB         0.85        N     V170F  \n",
       "49         rpoC         0.98        N     E750Q  \n",
       "56         rpoB         0.99        N     I491F  \n",
       "105        rpoC         1.00        N    V1252L  \n",
       "122        rpoB         1.00        N     H445P  \n",
       "129        rpoB         0.99        N     H445D  \n",
       "23         gyrA         1.00        N      A74S  \n",
       "54         gyrA         1.00        N      D94G  \n",
       "100        gyrA         0.71        N     R761G  \n",
       "130        gyrA         0.99        N      D94N  \n",
       "14         katG         1.00        N     L653Q  \n",
       "63         katG         0.70        N     V473F  \n",
       "92         katG         1.00        N     L159F  \n",
       "126        katG         1.00        N     T180K  \n",
       "23         gyrA         1.00        N      A74S  \n",
       "54         gyrA         1.00        N      D94G  \n",
       "100        gyrA         0.71        N     R761G  \n",
       "130        gyrA         0.99        N      D94N  \n",
       "23         gyrA         1.00        N      A74S  \n",
       "54         gyrA         1.00        N      D94G  \n",
       "100        gyrA         0.71        N     R761G  \n",
       "130        gyrA         0.99        N      D94N  \n",
       "14         katG         1.00        N     L653Q  \n",
       "63         katG         0.70        N     V473F  \n",
       "92         katG         1.00        N     L159F  \n",
       "126        katG         1.00        N     T180K  \n",
       "14         katG         1.00        N     L653Q  \n",
       "63         katG         0.70        N     V473F  \n",
       "92         katG         1.00        N     L159F  \n",
       "126        katG         1.00        N     T180K  \n",
       "23         gyrA         1.00        N      A74S  \n",
       "54         gyrA         1.00        N      D94G  \n",
       "100        gyrA         0.71        N     R761G  \n",
       "130        gyrA         0.99        N      D94N  \n",
       "11       fadD23         0.98        N     P437L  \n",
       "69       fadD30         0.89        N     A555E  \n",
       "104      fadD29         0.73        N     V425F  \n",
       "9          rplY         1.00        N       K3R  \n",
       "85        rpmB1         1.00        N      A91V  \n",
       "145        rplE         0.74        N     R122C  \n",
       "1          aldA         1.00        N     I156S  \n",
       "33          NaN         1.00        N     Y202*  \n",
       "66          NaN         1.00        S     I124I  \n",
       "78          NaN         1.00        N     A157E  \n",
       "20         cobK         1.00        S      A66A  \n",
       "35         cobH         1.00        N     R114Q  \n",
       "46         fdxB         0.98        S     E646E  \n",
       "125       PPE18         0.97        S     T131T  \n",
       "20         cobK         1.00        S      A66A  \n",
       "35         cobH         1.00        N     R114Q  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "with pd.option_context('display.max_rows', None, 'display.max_columns', None):\n",
    "    display(parallel_mutation_subsytems_with_SNPs_DF)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
