{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>.container { width:100% !important; }</style>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from IPython.core.display import display, HTML\n",
    "display(HTML(\"<style>.container { width:100% !important; }</style>\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##########################################################################################################################################################################################################################################"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## *Functions* for retrieving SNPs between each pair of *longitudinal* samples (SNPs with $>= 25$% $\\Delta$ AF)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##########################################################################################################################################################################################################################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import vcf\n",
    "\n",
    "%matplotlib inline\n",
    "import os\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib as mpl\n",
    "import matplotlib.ticker as ticker\n",
    "from pylab import plot, show, savefig, xlim, figure, hold, ylim, legend, boxplot, setp, axes\n",
    "from itertools import compress\n",
    "from pylab import MaxNLocator\n",
    "import seaborn as sns; sns.set()\n",
    "from matplotlib.colors import LogNorm\n",
    "from matplotlib import gridspec\n",
    "from matplotlib.gridspec import GridSpec\n",
    "import ast\n",
    "import itertools\n",
    "import seaborn as sns\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "\n",
    "import fastcluster\n",
    "from sklearn import cluster, datasets\n",
    "import scipy.cluster.hierarchy as hier\n",
    "from sklearn.cluster import KMeans\n",
    "import time\n",
    "import sys\n",
    "\n",
    "import Bio\n",
    "from Bio.Alphabet import IUPAC\n",
    "from Bio.Blast.Applications import NcbiblastnCommandline\n",
    "from Bio.Blast import NCBIXML\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.SeqFeature import SeqFeature, FeatureLocation\n",
    "from Bio import pairwise2\n",
    "from Bio import SeqIO\n",
    "from Bio.Graphics import GenomeDiagram\n",
    "from Bio.SeqUtils import GC\n",
    "\n",
    "from Bio.Align.Applications import MuscleCommandline\n",
    "from StringIO import StringIO\n",
    "from Bio import AlignIO\n",
    "from Bio.Align import AlignInfo\n",
    "from Bio.Seq import MutableSeq\n",
    "import itertools\n",
    "\n",
    "import networkx as nx\n",
    "import scipy\n",
    "import pickle\n",
    "\n",
    "#for exporting to Adobe Illustrator\n",
    "mpl.rcParams['pdf.fonttype'] = 42\n",
    "mpl.rcParams['ps.fonttype'] = 42"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Decide on a threshold for difference in Alternate Allele Frequencies to call SNPs between two isolates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "alt_AF_diff_threshold = 0.25 #x% "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load regions to exclude from analysis per EBR score across H37Rv (dropping sites with EBR score < 0.8)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('/n/data1/hms/dbmi/farhat/Roger/inhost_TB_dynamics_project/pickled_files/variant_calling/H37Rv_sites_to_drop.pkl', 'rb') as f:\n",
    "    H37Rv_positions_to_drop = pickle.load(f)\n",
    "    \n",
    "#convert to a set (faster to query)\n",
    "H37Rv_positions_to_drop = set(H37Rv_positions_to_drop)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### *Cell* to annotate SNPs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Important Packages\n",
    "################################################################################################################################################################################################\n",
    "import os\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import sys\n",
    "import pickle\n",
    "\n",
    "import Bio\n",
    "from Bio.Alphabet import IUPAC\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.SeqFeature import SeqFeature, FeatureLocation\n",
    "from Bio import SeqIO\n",
    "from StringIO import StringIO\n",
    "from Bio import AlignIO\n",
    "from Bio.Align import AlignInfo\n",
    "from Bio.Seq import MutableSeq\n",
    "################################################################################################################################################################################################\n",
    "\n",
    "\n",
    "# Relevant Information for H37Rv sequence SNP functional annotation\n",
    "################################################################################################################################################################################################\n",
    "####### Collect all DNA and Amino Acid sequences corresponding to genes on H37Rv #######\n",
    "#load reference genome and reference annotation\n",
    "reference_genome = '/n/data1/hms/dbmi/farhat/bin/work-horse/bin/h37rv.fasta'\n",
    "for reference_genome in SeqIO.parse(reference_genome, \"fasta\"):\n",
    "    reference_genome.seq.alphabet = IUPAC.unambiguous_dna\n",
    "\n",
    "reference_genome_annotation = pd.read_csv('/n/data1/hms/dbmi/farhat/Roger/inhost_TB_dynamics_project/H37Rv/h37rv_genome_summary.txt', '\\t').set_index('name')\n",
    "\n",
    "####### Function to translate coding DNA sequences ####### \n",
    "def translate(gene_id, sequence):\n",
    "\n",
    "    #find which strand the gene is located on and translate\n",
    "    strand = reference_genome_annotation.loc[gene_id, 'strand']\n",
    "    if strand == '+':\n",
    "        protein_sequence = sequence.translate(table=\"Bacterial\", cds=False)\n",
    "    elif strand == '-':\n",
    "        protein_sequence = sequence.reverse_complement().translate(table=\"Bacterial\", cds=False)\n",
    "\n",
    "    return protein_sequence\n",
    "\n",
    "####### Load in dictionaries for SNP annotation #######\n",
    "with open('/n/data1/hms/dbmi/farhat/Roger/inhost_TB_dynamics_project/pickled_files/dicts_for_SNP_annotation/H37Rv_gene_seq_records.pickle', 'rb') as handle:\n",
    "    ref_gene_sequences_records = pickle.load(handle)\n",
    "    \n",
    "with open('/n/data1/hms/dbmi/farhat/Roger/inhost_TB_dynamics_project/pickled_files/dicts_for_SNP_annotation/H37Rv_protein_seq_records.pickle', 'rb') as handle:\n",
    "    ref_protein_sequences_records = pickle.load(handle)\n",
    "    \n",
    "with open('/n/data1/hms/dbmi/farhat/Roger/inhost_TB_dynamics_project/pickled_files/dicts_for_SNP_annotation/H37Rv_coord_gene_mapping.pickle', 'rb') as handle:\n",
    "    ReferencePosition_Gene_mapping = pickle.load(handle)\n",
    "    \n",
    "####### get Gene Categories #######\n",
    "gene_categories = pd.read_csv('/n/data1/hms/dbmi/farhat/Roger/inhost_TB_dynamics_project/CSV_files/gene_categories/gene_categories.csv').set_index('name')\n",
    "gene_categories_dict = dict([gene_id , gene_category] for gene_id, gene_category in zip(list(gene_categories.gene_id) , list(gene_categories.Gene_Category)))\n",
    "\n",
    "####### get Gene Symbols #######\n",
    "gene_symbol_dict = dict([gene_id , gene_symbol] for gene_id, gene_symbol in zip(list(reference_genome_annotation.symbol.index) , list( reference_genome_annotation.symbol )))\n",
    "################################################################################################################################################################################################\n",
    "\n",
    "\n",
    "# Function to annotate Intergenic SNPs\n",
    "################################################################################################################################################################################################\n",
    "def find_flanking_genes_for_intergenic_region(intergenic_ref_pos): \n",
    "\n",
    "    #this function finds the genes flagging an intergenic region given a reference position\n",
    "\n",
    "    #find gene immediately in the 5' direction\n",
    "    for i in range(0 , 100000):\n",
    "\n",
    "        #move toward 5' direction\n",
    "        if ReferencePosition_Gene_mapping[intergenic_ref_pos - i] != []:\n",
    "\n",
    "            gene_to_left = ReferencePosition_Gene_mapping[intergenic_ref_pos - i][0]\n",
    "            break\n",
    "\n",
    "    #find gene immediately in the 3' direction       \n",
    "    for i in range(0 , 100000):\n",
    "\n",
    "        #move toward 3' direction\n",
    "        try:\n",
    "            if ReferencePosition_Gene_mapping[intergenic_ref_pos + i] != []:\n",
    "\n",
    "                gene_to_right = ReferencePosition_Gene_mapping[intergenic_ref_pos + i][0]\n",
    "                break\n",
    "        \n",
    "        #KeyError means we have hit the 'end' of the chromosome, the intergenic region at then end of H37Rv in 5' > 3' orientation \n",
    "        #since TB chromosome is circular the gene to the 'right' is Rv0001    \n",
    "        except KeyError:\n",
    "            \n",
    "            gene_to_right = 'Rv0001'\n",
    "            break\n",
    "    \n",
    "    return gene_to_left + '_' + gene_to_right\n",
    "################################################################################################################################################################################################\n",
    "\n",
    "\n",
    "# Function to determine whether SNPs are Synonymous or Non-Synonymous; Returns gene coordinate, codon position, AA changes, Gene Category & Symbol\n",
    "################################################################################################################################################################################################\n",
    "def SNP_annotate(ref_seq_position , alt_allele_i):\n",
    "    \n",
    "    '''\n",
    "    This function takes as input a reference position on H37Rv located within a \n",
    "    gene and an alternate allele and returns whether the base change \n",
    "    would correspond to a different Amino Acid sequence that results \n",
    "    from translating the DNA sequence into an AA sequence.\n",
    "    \n",
    "    '''\n",
    "    gene_intergenic_id_list = []\n",
    "    genomic_coord_list = []\n",
    "    gene_category_list = []\n",
    "    gene_symbol_list = []\n",
    "    Syn_NSyn_list = []\n",
    "    AA_change_list = []\n",
    "    \n",
    "    #get the Reference Allele from the complete H37Rv reference genome, indexing starts from 0\n",
    "    ref_allele_i = reference_genome.seq[int(ref_seq_position) - 1] \n",
    "    \n",
    "    #find the gene that SNP occurs on; check list corresponding to H37Rv coordinate to see if there are any genes associated with RefPosition\n",
    "    if len(ReferencePosition_Gene_mapping[ref_seq_position]) > 0:\n",
    "\n",
    "        #iterate through all genes that ReferencePosition is mapped to (i.e. SNP might correspond to 2 genes)\n",
    "        for gene_intergenic_id in ReferencePosition_Gene_mapping[ref_seq_position]:\n",
    "\n",
    "            #find genomic coordinate of SNP relative to gene (subtract 1 since reference seq starts counting at 1)\n",
    "            gene_relative_coord = (ref_seq_position - 1) - min( reference_genome_annotation.loc[gene_intergenic_id , 'chromStart'] , reference_genome_annotation.loc[gene_intergenic_id , 'chromEnd'] )\n",
    "            \n",
    "            #find the genomic coordinate (relative to the gene, in the 5' to 3' direction)\n",
    "            strand = reference_genome_annotation.loc[gene_intergenic_id, 'strand']\n",
    "            if strand == '+':\n",
    "                 genomic_5_to_3_coord = (ref_seq_position) - reference_genome_annotation.loc[gene_intergenic_id , 'chromStart']\n",
    "\n",
    "            elif strand == '-':\n",
    "                 genomic_5_to_3_coord = (reference_genome_annotation.loc[gene_intergenic_id , 'chromEnd']) - (ref_seq_position-1)\n",
    "                    \n",
    "            #find gene category (if one exists)\n",
    "            try:\n",
    "                gene_category_i = gene_categories_dict[gene_intergenic_id]\n",
    "            except KeyError:\n",
    "                gene_category_i = 'None'\n",
    "            \n",
    "            #find gene symbol (if one exists)\n",
    "            try:\n",
    "                gene_symbol_i = gene_symbol_dict[gene_intergenic_id]\n",
    "            except KeyError:\n",
    "                gene_symbol_i = 'None'\n",
    "            \n",
    "            #alternate allele is an actual base\n",
    "            if alt_allele_i in ['A','C','G','T']:\n",
    "\n",
    "                #translate into protein sequence with the SNP in place if not InDel or intergenic region\n",
    "                SNP_change = alt_allele_i\n",
    "\n",
    "                #ALTERNATE allele (is it Syn or NSyn?)\n",
    "                #get sequence from dictionary of sequences (and convert to mutable object)\n",
    "                test_gene_sequence = ref_gene_sequences_records[gene_intergenic_id].seq.tomutable()\n",
    "\n",
    "                #change reference gene sequence by the SNP in the query sequence\n",
    "                test_gene_sequence[int(gene_relative_coord)] = SNP_change\n",
    "\n",
    "                #convert back immutable object\n",
    "                test_gene_sequence = test_gene_sequence.toseq()\n",
    "\n",
    "                #translate sequence into amino acid seq\n",
    "                test_protein_sequence = translate(gene_intergenic_id , test_gene_sequence)\n",
    "\n",
    "                #store the H37Rv AA seq to compare against\n",
    "                H37Rv_AA_sequence = ref_protein_sequences_records[gene_intergenic_id].seq\n",
    "\n",
    "                #get the codon number where the SNP occurs within\n",
    "                ## take the genomic coordinate (relative to the gene, in the 5' to 3' direction), divide by 3, then take the ceiling of this number (will be fraction if SNP occurs in 1st or 2nd position on codon)\n",
    "                strand = reference_genome_annotation.loc[gene_intergenic_id, 'strand']\n",
    "                if strand == '+':\n",
    "                     genomic_5_to_3_coord = (ref_seq_position) - reference_genome_annotation.loc[gene_intergenic_id , 'chromStart']\n",
    "\n",
    "                elif strand == '-':\n",
    "                     genomic_5_to_3_coord = (reference_genome_annotation.loc[gene_intergenic_id , 'chromEnd']) - (ref_seq_position-1)\n",
    "\n",
    "                codon_coord = int(np.ceil( float( genomic_5_to_3_coord) / 3.0 ))\n",
    "\n",
    "                #compare to AA seq of original gene\n",
    "                if test_protein_sequence == H37Rv_AA_sequence:\n",
    "\n",
    "                    SNP_type = 'S'\n",
    "\n",
    "                    #get the AA before & after\n",
    "                    AA_change = H37Rv_AA_sequence[codon_coord-1] + str(codon_coord) + test_protein_sequence[codon_coord-1]\n",
    "\n",
    "                else:\n",
    "                    SNP_type = 'N'\n",
    "\n",
    "                    #get the AA before & after\n",
    "                    AA_change = H37Rv_AA_sequence[codon_coord-1] + str(codon_coord) + test_protein_sequence[codon_coord-1]\n",
    "                    \n",
    "            #alternate allele is a dummy (Base Call completely supports the Reference Allele)       \n",
    "            else:\n",
    "                \n",
    "                SNP_type = 'None'\n",
    "                AA_change = 'None'\n",
    "\n",
    "            #store relevant info in lists    \n",
    "            gene_intergenic_id_list.append(gene_intergenic_id)\n",
    "            genomic_coord_list.append(genomic_5_to_3_coord)\n",
    "            gene_category_list.append(gene_category_i)\n",
    "            gene_symbol_list.append(gene_symbol_i)\n",
    "            Syn_NSyn_list.append(SNP_type)\n",
    "            AA_change_list.append(AA_change)\n",
    "    \n",
    "    #if no gene in H37Rv corresponds to the Reference Position for SNP, then SNP must be intergenic\n",
    "    else:\n",
    "        \n",
    "        gene_intergenic_id = find_flanking_genes_for_intergenic_region(ref_seq_position)\n",
    "        genomic_5_to_3_coord = 'None'\n",
    "        gene_category_i = 'None'\n",
    "        gene_symbol_i = 'None'\n",
    "        SNP_type = 'I'\n",
    "        AA_change = 'None'\n",
    "        \n",
    "        #store relevant info in lists    \n",
    "        gene_intergenic_id_list.append(gene_intergenic_id)\n",
    "        genomic_coord_list.append(genomic_5_to_3_coord)\n",
    "        gene_category_list.append(gene_category_i)\n",
    "        gene_symbol_list.append(gene_symbol_i)\n",
    "        Syn_NSyn_list.append(SNP_type)\n",
    "        AA_change_list.append(AA_change)\n",
    "    \n",
    "    #if there is only a single gene associated with this SNP, just return the individual elememts\n",
    "    if len(gene_intergenic_id_list) == 1:\n",
    "        return [ref_allele_i , gene_intergenic_id , genomic_5_to_3_coord , gene_category_i , gene_symbol_i , SNP_type , AA_change]\n",
    "    \n",
    "    #else if there are two genes associated with this SNP, return elements for each SNP annotation in a list\n",
    "    elif len(gene_intergenic_id_list) > 1:\n",
    "        return [ref_allele_i , gene_intergenic_id_list , genomic_coord_list , gene_category_list , gene_symbol_list , Syn_NSyn_list , AA_change_list]\n",
    "################################################################################################################################################################################################"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### *Function* to get SNPs between paired isolates (filtered for $\\Delta AF$, MGE and low EBR score regions)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_filtered_SNPs_between_isolates(isolate_pair_ID , alt_AF_diff_threshold):\n",
    "\n",
    "    '''\n",
    "    This function only the fixed SNP variants that occur between a given isolate pair \n",
    "    by loading in the pickled DataFrame for isolate pair and comparing alternate allele frequencies called in each isolate.\n",
    "    (Differing Base Calls that have an Alternate Allele Frequencies >= x% different). \n",
    "    This function also drops regions flagged as Mobile Genetic Elements & Regions of poor Illumina mapping / variant calling\n",
    "    per Empirical Base Recall (EBR) scores across H37Rv.\n",
    "    '''\n",
    "\n",
    "    ################################################################################\n",
    "    ### get SNPs between pair of isolates\n",
    "    ################################################################################\n",
    "\n",
    "    population = sample_annotation.loc[isolate_pair_ID , 'population'][0]\n",
    "\n",
    "    #load in the differing Base Calls for the isolate pair from pickle\n",
    "    different_base_calls_between_isolates = pd.read_pickle(SNP_variant_dir + population + '_' + isolate_pair_ID + '/base_calls_different_between_isolates.pkl')\n",
    "\n",
    "    ################################################################################\n",
    "    ### Drop SNPs with change in AF < x%\n",
    "    ################################################################################\n",
    "\n",
    "    #FILTER out paired Base Calls that have a difference in Alternate Allele Frequency of less than x%\n",
    "    alt_AF_isolate_A = different_base_calls_between_isolates.loc[range(0 , np.shape(different_base_calls_between_isolates)[0] , 2) , 'alt_AF']\n",
    "    alt_AF_isolate_B = different_base_calls_between_isolates.loc[range(1 , np.shape(different_base_calls_between_isolates)[0] , 2) , 'alt_AF']\n",
    "    alt_AF_diff_btwn_paired_isolates = abs(alt_AF_isolate_A.values - alt_AF_isolate_B.values)\n",
    "\n",
    "    isolate_A_Base_Call_indices_small_change_alt_AF = list(alt_AF_isolate_A[alt_AF_diff_btwn_paired_isolates < alt_AF_diff_threshold].index)\n",
    "    isolate_B_Base_Call_indices_small_change_alt_AF = list(alt_AF_isolate_B[alt_AF_diff_btwn_paired_isolates < alt_AF_diff_threshold].index)\n",
    "    Base_Call_Indices_SMALL_Alt_AF_Diff = isolate_A_Base_Call_indices_small_change_alt_AF + isolate_B_Base_Call_indices_small_change_alt_AF\n",
    "\n",
    "    #drop paired Base Calls w/ corresponding change in Alterante Allele Frequency < x%\n",
    "    different_base_calls_between_isolates.drop(Base_Call_Indices_SMALL_Alt_AF_Diff , axis = 0 , inplace = True)\n",
    "\n",
    "    #reset index of filtered SNP DataFrame\n",
    "    different_base_calls_between_isolates.reset_index(inplace = True, drop = True)\n",
    "\n",
    "    ################################################################################\n",
    "    ### Drop SNPs with change in regions with low EBR scores\n",
    "    ################################################################################\n",
    "\n",
    "    #Drop Base Calls in H37Rv sites with low EBR score (make sure there is at least 1 SNP)\n",
    "    if np.shape(different_base_calls_between_isolates)[0] > 0:\n",
    "\n",
    "        #create a boolean filter for SNPs to keep\n",
    "        SNPs_to_keep_filter = [SNP_i_ref_pos not in H37Rv_positions_to_drop for SNP_i_ref_pos in different_base_calls_between_isolates.ref_position]\n",
    "\n",
    "        #filter out SNPs in H37Rv sites with low EBR scores and reset index\n",
    "        different_base_calls_between_isolates = different_base_calls_between_isolates[SNPs_to_keep_filter]\n",
    "        different_base_calls_between_isolates.reset_index(inplace = True, drop = True)\n",
    "\n",
    "    ################################################################################\n",
    "    ### Annotate SNPs & Drop SNPs in MGE regions\n",
    "    ################################################################################\n",
    "\n",
    "    gene_id_list = []\n",
    "    gene_coord_list = []\n",
    "    gene_category_list = []\n",
    "    gene_symbol_list = []\n",
    "    SNP_ftype_list = []\n",
    "    AA_change_list = []\n",
    "\n",
    "    #Annotate Filtered Base Calls (make sure there is at least 1 SNP)\n",
    "    if np.shape(different_base_calls_between_isolates)[0] > 0:\n",
    "\n",
    "        for ref_position_i , alt_base_i in zip(list(different_base_calls_between_isolates.ref_position) , list(different_base_calls_between_isolates.alt_base)):\n",
    "\n",
    "            #annotate SNP\n",
    "            gene_id_i , gene_coord_i , gene_category_i , gene_symbol_i , SNP_ftype_i , AA_change_i = SNP_annotate(ref_position_i , alt_base_i)[1:]\n",
    "\n",
    "            gene_id_list.append(gene_id_i)\n",
    "            gene_coord_list.append(gene_coord_i)\n",
    "            gene_category_list.append(gene_category_i)\n",
    "            gene_symbol_list.append(gene_symbol_i)\n",
    "            SNP_ftype_list.append(SNP_ftype_i)\n",
    "            AA_change_list.append(AA_change_i)\n",
    "\n",
    "        #create columns to store SNP annotation info\n",
    "        different_base_calls_between_isolates['gene_id'] = gene_id_list\n",
    "        different_base_calls_between_isolates['gene_coord'] = gene_coord_list\n",
    "        different_base_calls_between_isolates['gene_category'] = gene_category_list\n",
    "        different_base_calls_between_isolates['gene_symbol'] = gene_symbol_list\n",
    "        different_base_calls_between_isolates['SNP_ftype'] = SNP_ftype_list\n",
    "        different_base_calls_between_isolates['AA_change'] = AA_change_list\n",
    "\n",
    "        #FILTER out Base Calls in MGE regions (Mobile Genentic Elements)\n",
    "        SNPs_to_drop_filter = [] #True if SNP is located within an MGE region\n",
    "\n",
    "        for gene_id_i in list(different_base_calls_between_isolates.gene_category):\n",
    "\n",
    "            #only 1 or 0 genes associated with this SNP\n",
    "            if (type(gene_id_i) == str) and (gene_id_i == 'Mobile Genetic Element'):\n",
    "\n",
    "                SNPs_to_drop_filter.append(True)\n",
    "\n",
    "            #two genes associated with this SNP\n",
    "            elif (type(gene_id_i) == list) and ('Mobile Genetic Element' in gene_id_i):\n",
    "\n",
    "                SNPs_to_drop_filter.append(True)\n",
    "\n",
    "            #SNP not in an MGE region so don't drop\n",
    "            else:\n",
    "\n",
    "                SNPs_to_drop_filter.append(False)\n",
    "\n",
    "        #create a boolean filter for SNPs to keep\n",
    "        SNPs_to_keep_filter = [not MGE_SNP for MGE_SNP in SNPs_to_drop_filter]\n",
    "\n",
    "        #filter out SNPs in MGE regions and reset index\n",
    "        different_base_calls_between_isolates = different_base_calls_between_isolates[SNPs_to_keep_filter]\n",
    "        different_base_calls_between_isolates.reset_index(inplace = True, drop = True)\n",
    "\n",
    "    #No SNPs detected between this pair of isolates (empty DataFrame)\n",
    "    else:\n",
    "\n",
    "        different_base_calls_between_isolates['gene_id'] = \"\"\n",
    "        different_base_calls_between_isolates['gene_coord'] = \"\"\n",
    "        different_base_calls_between_isolates['gene_category'] = \"\"\n",
    "        different_base_calls_between_isolates['gene_symbol'] = \"\"\n",
    "        different_base_calls_between_isolates['SNP_ftype'] = \"\"\n",
    "        different_base_calls_between_isolates['AA_change'] = \"\"\n",
    "    \n",
    "    return different_base_calls_between_isolates"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##########################################################################################################################################################################################################################################"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Longitudinal Sample Pairs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##########################################################################################################################################################################################################################################"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "#### Import Sample Annotation file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_annotation = pd.read_csv('/n/data1/hms/dbmi/farhat/Roger/inhost_TB_dynamics_project/CSV_files/sample_annotation_files/Longitudinal_fastq_path_names_and_JankyPipe_tags_filtered_final.csv' , sep  = ',').set_index('patient_id')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "SNP_variant_dir = '/n/data1/hms/dbmi/farhat/Roger/inhost_TB_dynamics_project/pickled_files/variant_calling/longitudinal_SNPs/all_SNPs_between_longitudinal_pairs/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fastq_files</th>\n",
       "      <th>population</th>\n",
       "      <th>run_ID</th>\n",
       "      <th>sample_ID</th>\n",
       "      <th>sample_order</th>\n",
       "      <th>tag</th>\n",
       "      <th>isolate_type</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>patient_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>P000183</th>\n",
       "      <td>/n/data1/hms/dbmi/farhat/fastq_db/walker/ERR03...</td>\n",
       "      <td>WALKER</td>\n",
       "      <td>ERR039337</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>ERR039337</td>\n",
       "      <td>longitudinal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P000183</th>\n",
       "      <td>/n/data1/hms/dbmi/farhat/fastq_db/walker/ERR03...</td>\n",
       "      <td>WALKER</td>\n",
       "      <td>ERR039338</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>ERR039338</td>\n",
       "      <td>longitudinal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1960</th>\n",
       "      <td>/n/data1/hms/dbmi/farhat/cetr_strains/good_wgs...</td>\n",
       "      <td>CETR</td>\n",
       "      <td>MMJA00000000</td>\n",
       "      <td>Peru5115</td>\n",
       "      <td>1</td>\n",
       "      <td>Peru5115</td>\n",
       "      <td>longitudinal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1960</th>\n",
       "      <td>/n/data1/hms/dbmi/farhat/cetr_strains/good_wgs...</td>\n",
       "      <td>CETR</td>\n",
       "      <td>MMPC00000000</td>\n",
       "      <td>Peru4668</td>\n",
       "      <td>2</td>\n",
       "      <td>Peru4668</td>\n",
       "      <td>longitudinal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2491</th>\n",
       "      <td>/n/data1/hms/dbmi/farhat/cetr_strains/good_wgs...</td>\n",
       "      <td>CETR</td>\n",
       "      <td>MMZT00000000</td>\n",
       "      <td>Peru4903</td>\n",
       "      <td>1</td>\n",
       "      <td>Peru4903</td>\n",
       "      <td>longitudinal</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                  fastq_files population  \\\n",
       "patient_id                                                                 \n",
       "P000183     /n/data1/hms/dbmi/farhat/fastq_db/walker/ERR03...     WALKER   \n",
       "P000183     /n/data1/hms/dbmi/farhat/fastq_db/walker/ERR03...     WALKER   \n",
       "1960        /n/data1/hms/dbmi/farhat/cetr_strains/good_wgs...       CETR   \n",
       "1960        /n/data1/hms/dbmi/farhat/cetr_strains/good_wgs...       CETR   \n",
       "2491        /n/data1/hms/dbmi/farhat/cetr_strains/good_wgs...       CETR   \n",
       "\n",
       "                  run_ID sample_ID  sample_order        tag  isolate_type  \n",
       "patient_id                                                                 \n",
       "P000183        ERR039337       NaN             1  ERR039337  longitudinal  \n",
       "P000183        ERR039338       NaN             2  ERR039338  longitudinal  \n",
       "1960        MMJA00000000  Peru5115             1   Peru5115  longitudinal  \n",
       "1960        MMPC00000000  Peru4668             2   Peru4668  longitudinal  \n",
       "2491        MMZT00000000  Peru4903             1   Peru4903  longitudinal  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_annotation.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "200\n"
     ]
    }
   ],
   "source": [
    "num_isolate_pair_IDs = np.shape(sample_annotation)[0] / 2\n",
    "print num_isolate_pair_IDs "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "isolate_pair_ID_list = list(set(sample_annotation.index))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Call function to collect SNPs passing Difference in Alternate Allele Frequency Threshold"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "5\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/rv76/anaconda2/lib/python2.7/site-packages/Bio/Seq.py:2715: BiopythonWarning: Partial codon, len(sequence) not a multiple of three. Explicitly trim the sequence or add trailing N before translation. This may become an error in future.\n",
      "  BiopythonWarning)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10\n",
      "15\n",
      "20\n",
      "25\n",
      "30\n",
      "35\n",
      "40\n",
      "45\n",
      "50\n",
      "55\n",
      "60\n",
      "65\n",
      "70\n",
      "75\n",
      "80\n",
      "85\n",
      "90\n",
      "95\n",
      "100\n",
      "105\n",
      "110\n",
      "115\n",
      "120\n",
      "125\n",
      "130\n",
      "135\n",
      "140\n",
      "145\n",
      "150\n",
      "155\n",
      "160\n",
      "165\n",
      "170\n",
      "175\n",
      "180\n",
      "185\n",
      "190\n",
      "195\n",
      "200\n"
     ]
    }
   ],
   "source": [
    "Base_Call_variants_btwn_isolates_big_change_in_alt_AF = []\n",
    "\n",
    "isolate_pair_index = 0\n",
    "\n",
    "#iterate through isolate pairs, collect all SNP variants arising between each pair of isolates\n",
    "for isolate_pair_ID in isolate_pair_ID_list:\n",
    "    \n",
    "    #retrieve filtered paired Base Calls with a change in Alternate Allele Frequency > threshold\n",
    "    Base_Call_variants_btwn_isolates_big_change_in_alt_AF_pair_i = get_filtered_SNPs_between_isolates(isolate_pair_ID , alt_AF_diff_threshold)\n",
    "    \n",
    "    #store relevant Base Call info in list of DataFrames (1 for each isolate pair)\n",
    "    Base_Call_variants_btwn_isolates_big_change_in_alt_AF.append(Base_Call_variants_btwn_isolates_big_change_in_alt_AF_pair_i)\n",
    "    isolate_pair_index += 1\n",
    "    \n",
    "    if isolate_pair_index % 5 == 0:\n",
    "        print isolate_pair_index\n",
    "        \n",
    "#concatenate DataFrames for each subject into 1 DataFrame\n",
    "Base_Call_variants_btwn_isolates_big_change_in_alt_AF = pd.concat(Base_Call_variants_btwn_isolates_big_change_in_alt_AF , axis = 0)\n",
    "\n",
    "Base_Call_variants_btwn_isolates_big_change_in_alt_AF.reset_index(inplace = True , drop = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### *Filter*: Drop paired Base Calls if both Base Calls in a pair support *different* Alternate Alleles"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "#list that stores the indices of paired Base Calls with DIFFERENT Alternate Alleles\n",
    "Base_Calls_to_Drop = []\n",
    "\n",
    "#iterate through each PAIR of corresponding Base Calls from paired isolates\n",
    "for isolate_A_Base_Call_i , isolate_B_Base_Call_i in zip(range(0 , np.shape(Base_Call_variants_btwn_isolates_big_change_in_alt_AF)[0] , 2) , range(1 , np.shape(Base_Call_variants_btwn_isolates_big_change_in_alt_AF)[0] , 2) ):\n",
    "    \n",
    "    #pull info that both Base Calls should have in COMMON\n",
    "    isolate_A_Base_Call_info = list( Base_Call_variants_btwn_isolates_big_change_in_alt_AF.loc[isolate_A_Base_Call_i , ['ref_base','ref_position','gene_id','genomic_coord','population','patient_id']] ) \n",
    "    isolate_B_Base_Call_info = list( Base_Call_variants_btwn_isolates_big_change_in_alt_AF.loc[isolate_B_Base_Call_i , ['ref_base','ref_position','gene_id','genomic_coord','population','patient_id']] ) \n",
    "    \n",
    "    #make sure Base Calls Match with respect to Reference Base, Reference Position, gene ID, Genomic Coordinate, Gene Category, Symbol, Population & Patient ID\n",
    "    if isolate_A_Base_Call_info == isolate_B_Base_Call_info:\n",
    "        \n",
    "        #pull alternate Allele for each of the paired isolates\n",
    "        isolate_A_Alt_Base = Base_Call_variants_btwn_isolates_big_change_in_alt_AF.loc[isolate_A_Base_Call_i , 'alt_base']\n",
    "        isolate_B_Alt_Base = Base_Call_variants_btwn_isolates_big_change_in_alt_AF.loc[isolate_B_Base_Call_i , 'alt_base']\n",
    "        \n",
    "        #Check to see if there is a 'Z' in the pair of Alternate Alleles, if so one of the Base Calls supported the Reference Base (so there was no Alternate Allele)\n",
    "        if (isolate_A_Alt_Base == 'Z') or (isolate_B_Alt_Base == 'Z'):\n",
    "            pass\n",
    "        \n",
    "        #if neither Alternate Allele is a 'Z', then check to see that the Alternate Allele Bases Match\n",
    "        elif isolate_A_Alt_Base == isolate_B_Alt_Base:\n",
    "            pass\n",
    "        \n",
    "        #if the Alternate Alleles DON'T match and both Base Calls supported Alternate Alleles (not the Reference), then we can't compare the Allele Frequencies of these Alternate Alleles (since they're different)\n",
    "        else:\n",
    "            Base_Calls_to_Drop = Base_Calls_to_Drop + [isolate_A_Base_Call_i , isolate_B_Base_Call_i]\n",
    "        \n",
    "    #print indices of Base Calls and see what went wrong if the paired Base Calls have different information that the Calls should have in Common (Ref Position, Ref Base, Gene ID, Patient ID, etc.)\n",
    "    else:\n",
    "        print (isolate_A_Base_Call_i , isolate_B_Base_Call_i)\n",
    "        \n",
    "#Drop Paired Base Calls that supported different Alternate Alleles\n",
    "Base_Call_variants_btwn_isolates_big_change_in_alt_AF.drop(Base_Calls_to_Drop , axis = 0 , inplace = True)\n",
    "\n",
    "#reset index\n",
    "Base_Call_variants_btwn_isolates_big_change_in_alt_AF.reset_index(inplace = True, drop = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ref_base</th>\n",
       "      <th>alt_base</th>\n",
       "      <th>ref_position</th>\n",
       "      <th>quality</th>\n",
       "      <th>SNP_type</th>\n",
       "      <th>PASS_filter</th>\n",
       "      <th>INFO</th>\n",
       "      <th>alt_AF</th>\n",
       "      <th>depth</th>\n",
       "      <th>tag</th>\n",
       "      <th>population</th>\n",
       "      <th>patient_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_coord</th>\n",
       "      <th>gene_category</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>SNP_ftype</th>\n",
       "      <th>AA_change</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>C</td>\n",
       "      <td>Z</td>\n",
       "      <td>761139</td>\n",
       "      <td>3140</td>\n",
       "      <td>Ref_PASS</td>\n",
       "      <td>[]</td>\n",
       "      <td>{u'QP': [0, 100, 0, 0], u'AC': [0], u'AF': [0....</td>\n",
       "      <td>0.00</td>\n",
       "      <td>89</td>\n",
       "      <td>ERR369703</td>\n",
       "      <td>CASALI</td>\n",
       "      <td>P251</td>\n",
       "      <td>Rv0667</td>\n",
       "      <td>1333</td>\n",
       "      <td>Antibiotic Resistance</td>\n",
       "      <td>rpoB</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>761139</td>\n",
       "      <td>1266</td>\n",
       "      <td>Alt_PASS</td>\n",
       "      <td>[]</td>\n",
       "      <td>{u'QP': [0, 0, 100, 0], u'AC': [2], u'AF': [1....</td>\n",
       "      <td>1.00</td>\n",
       "      <td>36</td>\n",
       "      <td>ERR600658</td>\n",
       "      <td>CASALI</td>\n",
       "      <td>P251</td>\n",
       "      <td>Rv0667</td>\n",
       "      <td>1333</td>\n",
       "      <td>Antibiotic Resistance</td>\n",
       "      <td>rpoB</td>\n",
       "      <td>N</td>\n",
       "      <td>H445D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>T</td>\n",
       "      <td>Z</td>\n",
       "      <td>861378</td>\n",
       "      <td>3116</td>\n",
       "      <td>Ref_PASS</td>\n",
       "      <td>[]</td>\n",
       "      <td>{u'QP': [0, 0, 0, 100], u'AC': [0], u'AF': [0....</td>\n",
       "      <td>0.00</td>\n",
       "      <td>90</td>\n",
       "      <td>ERR369703</td>\n",
       "      <td>CASALI</td>\n",
       "      <td>P251</td>\n",
       "      <td>Rv0768</td>\n",
       "      <td>467</td>\n",
       "      <td>Non-Essential</td>\n",
       "      <td>aldA</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>861378</td>\n",
       "      <td>1167</td>\n",
       "      <td>Alt_PASS</td>\n",
       "      <td>[]</td>\n",
       "      <td>{u'QP': [0, 0, 100, 0], u'AC': [2], u'AF': [1....</td>\n",
       "      <td>1.00</td>\n",
       "      <td>34</td>\n",
       "      <td>ERR600658</td>\n",
       "      <td>CASALI</td>\n",
       "      <td>P251</td>\n",
       "      <td>Rv0768</td>\n",
       "      <td>467</td>\n",
       "      <td>Non-Essential</td>\n",
       "      <td>aldA</td>\n",
       "      <td>N</td>\n",
       "      <td>I156S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>4402310</td>\n",
       "      <td>2360</td>\n",
       "      <td>Amb</td>\n",
       "      <td>[Amb]</td>\n",
       "      <td>{u'QP': [48, 0, 52, 0], u'AC': [1], u'AF': [0....</td>\n",
       "      <td>0.52</td>\n",
       "      <td>71</td>\n",
       "      <td>ERR369703</td>\n",
       "      <td>CASALI</td>\n",
       "      <td>P251</td>\n",
       "      <td>Rv3913</td>\n",
       "      <td>583</td>\n",
       "      <td>Essential</td>\n",
       "      <td>trxB2</td>\n",
       "      <td>N</td>\n",
       "      <td>I195V</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>4402310</td>\n",
       "      <td>1340</td>\n",
       "      <td>Alt_PASS</td>\n",
       "      <td>[]</td>\n",
       "      <td>{u'QP': [0, 0, 100, 0], u'AC': [2], u'AF': [1....</td>\n",
       "      <td>1.00</td>\n",
       "      <td>38</td>\n",
       "      <td>ERR600658</td>\n",
       "      <td>CASALI</td>\n",
       "      <td>P251</td>\n",
       "      <td>Rv3913</td>\n",
       "      <td>583</td>\n",
       "      <td>Essential</td>\n",
       "      <td>trxB2</td>\n",
       "      <td>N</td>\n",
       "      <td>I195V</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>656329</td>\n",
       "      <td>7836</td>\n",
       "      <td>Ref_PASS</td>\n",
       "      <td>[]</td>\n",
       "      <td>{u'QP': [91, 9, 0, 0], u'AC': [0], u'AF': [0.0...</td>\n",
       "      <td>0.09</td>\n",
       "      <td>306</td>\n",
       "      <td>Peru5140</td>\n",
       "      <td>CETR</td>\n",
       "      <td>2968</td>\n",
       "      <td>Rv0565c</td>\n",
       "      <td>1142</td>\n",
       "      <td>Non-Essential</td>\n",
       "      <td>NaN</td>\n",
       "      <td>N</td>\n",
       "      <td>V381G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>656329</td>\n",
       "      <td>5372</td>\n",
       "      <td>Amb</td>\n",
       "      <td>[Amb]</td>\n",
       "      <td>{u'QP': [41, 59, 0, 0], u'AC': [1], u'AF': [0....</td>\n",
       "      <td>0.59</td>\n",
       "      <td>219</td>\n",
       "      <td>Peru5139</td>\n",
       "      <td>CETR</td>\n",
       "      <td>2968</td>\n",
       "      <td>Rv0565c</td>\n",
       "      <td>1142</td>\n",
       "      <td>Non-Essential</td>\n",
       "      <td>NaN</td>\n",
       "      <td>N</td>\n",
       "      <td>V381G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>C</td>\n",
       "      <td>Z</td>\n",
       "      <td>656956</td>\n",
       "      <td>15111</td>\n",
       "      <td>Ref_PASS</td>\n",
       "      <td>[]</td>\n",
       "      <td>{u'QP': [0, 100, 0, 0], u'AC': [0], u'AF': [0....</td>\n",
       "      <td>0.00</td>\n",
       "      <td>389</td>\n",
       "      <td>Peru5140</td>\n",
       "      <td>CETR</td>\n",
       "      <td>2968</td>\n",
       "      <td>Rv0565c</td>\n",
       "      <td>515</td>\n",
       "      <td>Non-Essential</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>656956</td>\n",
       "      <td>3512</td>\n",
       "      <td>Amb</td>\n",
       "      <td>[Amb]</td>\n",
       "      <td>{u'QP': [0, 65, 0, 34], u'AC': [1], u'AF': [0....</td>\n",
       "      <td>0.34</td>\n",
       "      <td>235</td>\n",
       "      <td>Peru5139</td>\n",
       "      <td>CETR</td>\n",
       "      <td>2968</td>\n",
       "      <td>Rv0565c</td>\n",
       "      <td>515</td>\n",
       "      <td>Non-Essential</td>\n",
       "      <td>NaN</td>\n",
       "      <td>N</td>\n",
       "      <td>W172*</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  ref_base alt_base  ref_position  quality  SNP_type PASS_filter  \\\n",
       "0        C        Z        761139     3140  Ref_PASS          []   \n",
       "1        C        G        761139     1266  Alt_PASS          []   \n",
       "2        T        Z        861378     3116  Ref_PASS          []   \n",
       "3        T        G        861378     1167  Alt_PASS          []   \n",
       "4        A        G       4402310     2360       Amb       [Amb]   \n",
       "5        A        G       4402310     1340  Alt_PASS          []   \n",
       "6        A        C        656329     7836  Ref_PASS          []   \n",
       "7        A        C        656329     5372       Amb       [Amb]   \n",
       "8        C        Z        656956    15111  Ref_PASS          []   \n",
       "9        C        T        656956     3512       Amb       [Amb]   \n",
       "\n",
       "                                                INFO  alt_AF  depth  \\\n",
       "0  {u'QP': [0, 100, 0, 0], u'AC': [0], u'AF': [0....    0.00     89   \n",
       "1  {u'QP': [0, 0, 100, 0], u'AC': [2], u'AF': [1....    1.00     36   \n",
       "2  {u'QP': [0, 0, 0, 100], u'AC': [0], u'AF': [0....    0.00     90   \n",
       "3  {u'QP': [0, 0, 100, 0], u'AC': [2], u'AF': [1....    1.00     34   \n",
       "4  {u'QP': [48, 0, 52, 0], u'AC': [1], u'AF': [0....    0.52     71   \n",
       "5  {u'QP': [0, 0, 100, 0], u'AC': [2], u'AF': [1....    1.00     38   \n",
       "6  {u'QP': [91, 9, 0, 0], u'AC': [0], u'AF': [0.0...    0.09    306   \n",
       "7  {u'QP': [41, 59, 0, 0], u'AC': [1], u'AF': [0....    0.59    219   \n",
       "8  {u'QP': [0, 100, 0, 0], u'AC': [0], u'AF': [0....    0.00    389   \n",
       "9  {u'QP': [0, 65, 0, 34], u'AC': [1], u'AF': [0....    0.34    235   \n",
       "\n",
       "         tag population patient_id  gene_id gene_coord          gene_category  \\\n",
       "0  ERR369703     CASALI       P251   Rv0667       1333  Antibiotic Resistance   \n",
       "1  ERR600658     CASALI       P251   Rv0667       1333  Antibiotic Resistance   \n",
       "2  ERR369703     CASALI       P251   Rv0768        467          Non-Essential   \n",
       "3  ERR600658     CASALI       P251   Rv0768        467          Non-Essential   \n",
       "4  ERR369703     CASALI       P251   Rv3913        583              Essential   \n",
       "5  ERR600658     CASALI       P251   Rv3913        583              Essential   \n",
       "6   Peru5140       CETR       2968  Rv0565c       1142          Non-Essential   \n",
       "7   Peru5139       CETR       2968  Rv0565c       1142          Non-Essential   \n",
       "8   Peru5140       CETR       2968  Rv0565c        515          Non-Essential   \n",
       "9   Peru5139       CETR       2968  Rv0565c        515          Non-Essential   \n",
       "\n",
       "  gene_symbol SNP_ftype AA_change  \n",
       "0        rpoB      None      None  \n",
       "1        rpoB         N     H445D  \n",
       "2        aldA      None      None  \n",
       "3        aldA         N     I156S  \n",
       "4       trxB2         N     I195V  \n",
       "5       trxB2         N     I195V  \n",
       "6         NaN         N     V381G  \n",
       "7         NaN         N     V381G  \n",
       "8         NaN      None      None  \n",
       "9         NaN         N     W172*  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Base_Call_variants_btwn_isolates_big_change_in_alt_AF.head(n = 10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(3198, 18)"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.shape(Base_Call_variants_btwn_isolates_big_change_in_alt_AF)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Pickle DataFrame for Downstream analyses (Alternate Allele Frequency 1 vs. Alternate Allele Frequency 2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "Base_Call_variants_btwn_isolates_big_change_in_alt_AF.to_pickle('/n/data1/hms/dbmi/farhat/Roger/inhost_TB_dynamics_project/pickled_files/variant_calling/longitudinal_SNPs/longitudinal_SNP_variants_25_delta_in_alt_AF.pkl')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "### Re-Shape Filtered DataFrame (Paired Base Calls across all isolate pairs) to store one entry per SNP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "SNP_variants_between_paired_isolates = pd.DataFrame()\n",
    "\n",
    "#common information to both Base Calls (can just look at isolate A)\n",
    "population_dict = {}\n",
    "patient_id_dict = {}\n",
    "ref_position_dict = {}\n",
    "ref_allele_dict = {}\n",
    "gene_id_dict  = {}\n",
    "genomic_coord_dict = {}\n",
    "gene_category_dict = {}\n",
    "gene_symbol_dict = {}\n",
    "\n",
    "#look at info for both Base Calls\n",
    "alt_allele_dict = {}\n",
    "alt_AF_diff_dict = {}\n",
    "SNP_type_dict = {}\n",
    "AA_change_dict = {}\n",
    "\n",
    "SNP_index = 0\n",
    "#iterate through indices for isolate A (store common information for isolate pair A & B came from and Base Call), calculate different in Alternate Allele Frequencies, store Syn or NSyn info\n",
    "for even_index in range(0 , np.shape(Base_Call_variants_btwn_isolates_big_change_in_alt_AF)[0] , 2):\n",
    "    \n",
    "    #Base Call info for isolate A\n",
    "    Base_Call_info_isolate_A = Base_Call_variants_btwn_isolates_big_change_in_alt_AF.loc[even_index , :]\n",
    "    #Base Call info for isolate B\n",
    "    Base_Call_info_isolate_B = Base_Call_variants_btwn_isolates_big_change_in_alt_AF.loc[even_index+1 , :]\n",
    "    \n",
    "    population_dict[SNP_index] = Base_Call_info_isolate_A.population\n",
    "    patient_id_dict[SNP_index] = Base_Call_info_isolate_A.patient_id\n",
    "    ref_position_dict[SNP_index] = Base_Call_info_isolate_A.ref_position\n",
    "    ref_allele_dict[SNP_index] = Base_Call_info_isolate_A.ref_base\n",
    "    gene_id_dict[SNP_index] = Base_Call_info_isolate_A.gene_id\n",
    "    genomic_coord_dict[SNP_index] = Base_Call_info_isolate_A.gene_coord\n",
    "    gene_category_dict[SNP_index] = Base_Call_info_isolate_A.gene_category\n",
    "    gene_symbol_dict[SNP_index] = Base_Call_info_isolate_A.gene_symbol\n",
    "    \n",
    "    #get alternate allele\n",
    "    alt_allele_calls = [Base_Call_info_isolate_A.alt_base , Base_Call_info_isolate_B.alt_base]\n",
    "    try:\n",
    "        alt_allele_calls.remove('Z')\n",
    "    except ValueError:\n",
    "        pass\n",
    "    alt_allele_dict[SNP_index] = alt_allele_calls[0]\n",
    "\n",
    "    #get difference in Alternate Allele Frequencies\n",
    "    alt_AF_diff_dict[SNP_index] = abs(Base_Call_info_isolate_A.alt_AF - Base_Call_info_isolate_B.alt_AF)\n",
    "    \n",
    "    #get type of SNP \n",
    "    if 'S' in [Base_Call_info_isolate_A.SNP_ftype , Base_Call_info_isolate_B.SNP_ftype]:\n",
    "        SNP_type_dict[SNP_index] = 'S'\n",
    "        \n",
    "    elif 'N' in [Base_Call_info_isolate_A.SNP_ftype , Base_Call_info_isolate_B.SNP_ftype]:\n",
    "        SNP_type_dict[SNP_index] = 'N'\n",
    "        \n",
    "    elif 'I' in [Base_Call_info_isolate_A.SNP_ftype , Base_Call_info_isolate_B.SNP_ftype]:\n",
    "        SNP_type_dict[SNP_index] = 'I'\n",
    "        \n",
    "    #get AA change\n",
    "    AA_change_calls = [Base_Call_info_isolate_A.AA_change , Base_Call_info_isolate_B.AA_change]\n",
    "    try:\n",
    "        AA_change_calls.remove('None')\n",
    "    except ValueError:\n",
    "        pass\n",
    "    AA_change_dict[SNP_index] = AA_change_calls[0]\n",
    "        \n",
    "    SNP_index += 1\n",
    "    \n",
    "#convert dictionaries into series\n",
    "population = pd.Series(population_dict)\n",
    "patient_id = pd.Series(patient_id_dict)\n",
    "ref_position = pd.Series(ref_position_dict)\n",
    "ref_allele = pd.Series(ref_allele_dict)\n",
    "alt_allele = pd.Series(alt_allele_dict)\n",
    "gene_id = pd.Series(gene_id_dict)\n",
    "genomic_coord = pd.Series(genomic_coord_dict)\n",
    "gene_category = pd.Series(gene_category_dict)\n",
    "gene_symbol = pd.Series(gene_symbol_dict)\n",
    "alt_AF_diff = pd.Series(alt_AF_diff_dict)\n",
    "SNP_type = pd.Series(SNP_type_dict)\n",
    "AA_change = pd.Series(AA_change_dict)\n",
    "        \n",
    "#create DataFrame\n",
    "SNP_variants_between_paired_isolates['population'] = population\n",
    "SNP_variants_between_paired_isolates['patient_id'] = patient_id\n",
    "SNP_variants_between_paired_isolates['ref_position'] = ref_position \n",
    "SNP_variants_between_paired_isolates['ref_allele'] = ref_allele\n",
    "SNP_variants_between_paired_isolates['alt_allele'] = alt_allele\n",
    "SNP_variants_between_paired_isolates['gene_id'] = gene_id\n",
    "SNP_variants_between_paired_isolates['genomic_coord'] = genomic_coord\n",
    "SNP_variants_between_paired_isolates['gene_category'] = gene_category\n",
    "SNP_variants_between_paired_isolates['gene_symbol'] = gene_symbol\n",
    "SNP_variants_between_paired_isolates['alt_AF_diff'] = alt_AF_diff\n",
    "SNP_variants_between_paired_isolates['SNP_type'] = SNP_type\n",
    "SNP_variants_between_paired_isolates['AA_change'] = AA_change"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>population</th>\n",
       "      <th>patient_id</th>\n",
       "      <th>ref_position</th>\n",
       "      <th>ref_allele</th>\n",
       "      <th>alt_allele</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>genomic_coord</th>\n",
       "      <th>gene_category</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>alt_AF_diff</th>\n",
       "      <th>SNP_type</th>\n",
       "      <th>AA_change</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CASALI</td>\n",
       "      <td>P251</td>\n",
       "      <td>761139</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>Rv0667</td>\n",
       "      <td>1333</td>\n",
       "      <td>Antibiotic Resistance</td>\n",
       "      <td>rpoB</td>\n",
       "      <td>1.00</td>\n",
       "      <td>N</td>\n",
       "      <td>H445D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CASALI</td>\n",
       "      <td>P251</td>\n",
       "      <td>861378</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>Rv0768</td>\n",
       "      <td>467</td>\n",
       "      <td>Non-Essential</td>\n",
       "      <td>aldA</td>\n",
       "      <td>1.00</td>\n",
       "      <td>N</td>\n",
       "      <td>I156S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CASALI</td>\n",
       "      <td>P251</td>\n",
       "      <td>4402310</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>Rv3913</td>\n",
       "      <td>583</td>\n",
       "      <td>Essential</td>\n",
       "      <td>trxB2</td>\n",
       "      <td>0.48</td>\n",
       "      <td>N</td>\n",
       "      <td>I195V</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CETR</td>\n",
       "      <td>2968</td>\n",
       "      <td>656329</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>Rv0565c</td>\n",
       "      <td>1142</td>\n",
       "      <td>Non-Essential</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.50</td>\n",
       "      <td>N</td>\n",
       "      <td>V381G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CETR</td>\n",
       "      <td>2968</td>\n",
       "      <td>656956</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>Rv0565c</td>\n",
       "      <td>515</td>\n",
       "      <td>Non-Essential</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.34</td>\n",
       "      <td>N</td>\n",
       "      <td>W172*</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  population patient_id  ref_position ref_allele alt_allele  gene_id  \\\n",
       "0     CASALI       P251        761139          C          G   Rv0667   \n",
       "1     CASALI       P251        861378          T          G   Rv0768   \n",
       "2     CASALI       P251       4402310          A          G   Rv3913   \n",
       "3       CETR       2968        656329          A          C  Rv0565c   \n",
       "4       CETR       2968        656956          C          T  Rv0565c   \n",
       "\n",
       "  genomic_coord          gene_category gene_symbol  alt_AF_diff SNP_type  \\\n",
       "0          1333  Antibiotic Resistance        rpoB         1.00        N   \n",
       "1           467          Non-Essential        aldA         1.00        N   \n",
       "2           583              Essential       trxB2         0.48        N   \n",
       "3          1142          Non-Essential         NaN         0.50        N   \n",
       "4           515          Non-Essential         NaN         0.34        N   \n",
       "\n",
       "  AA_change  \n",
       "0     H445D  \n",
       "1     I156S  \n",
       "2     I195V  \n",
       "3     V381G  \n",
       "4     W172*  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SNP_variants_between_paired_isolates.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1599, 12)"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.shape(SNP_variants_between_paired_isolates)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Save DataFrame as CSV"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "SNP_variants_between_paired_isolates.to_csv('/n/data1/hms/dbmi/farhat/Roger/inhost_TB_dynamics_project/CSV_files/variant_calling/longitudinal_SNPs/SNPs_between_isolates_delta_25.csv' , sep = ',')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Pickle DataFrame for Downstream analyses"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "SNP_variants_between_paired_isolates.to_pickle('/n/data1/hms/dbmi/farhat/Roger/inhost_TB_dynamics_project/pickled_files/variant_calling/longitudinal_SNPs/SNPs_between_isolates_delta_25.pkl')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
