{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>.container { width:100% !important; }</style>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from IPython.core.display import display, HTML\n",
    "display(HTML(\"<style>.container { width:100% !important; }</style>\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### This notebook was created to analyze the timing between samples for three different categories of longitudinal samples (patients), those with: *mixed* infection (defined by F2 mixed measure), *re-infection* (defined by fSNP distance) and *persistent*/*clonal* infection"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "import vcf\n",
    "\n",
    "%matplotlib inline\n",
    "import os\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib as mpl\n",
    "import matplotlib.ticker as ticker\n",
    "from pylab import plot, show, savefig, xlim, figure, hold, ylim, legend, boxplot, setp, axes\n",
    "from itertools import compress\n",
    "from pylab import MaxNLocator\n",
    "import seaborn as sns; sns.set()\n",
    "from matplotlib.colors import LogNorm\n",
    "from matplotlib import gridspec\n",
    "from matplotlib.gridspec import GridSpec\n",
    "import ast\n",
    "import itertools\n",
    "import seaborn as sns\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "\n",
    "import fastcluster\n",
    "from sklearn import cluster, datasets\n",
    "import scipy.cluster.hierarchy as hier\n",
    "from sklearn.cluster import KMeans\n",
    "import time\n",
    "import sys\n",
    "\n",
    "import Bio\n",
    "from Bio.Alphabet import IUPAC\n",
    "from Bio.Blast.Applications import NcbiblastnCommandline\n",
    "from Bio.Blast import NCBIXML\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.SeqFeature import SeqFeature, FeatureLocation\n",
    "from Bio import pairwise2\n",
    "from Bio import SeqIO\n",
    "from Bio.Graphics import GenomeDiagram\n",
    "from Bio.SeqUtils import GC\n",
    "\n",
    "from Bio.Align.Applications import MuscleCommandline\n",
    "from StringIO import StringIO\n",
    "from Bio import AlignIO\n",
    "from Bio.Align import AlignInfo\n",
    "from Bio.Seq import MutableSeq\n",
    "import itertools\n",
    "\n",
    "import networkx as nx\n",
    "import scipy\n",
    "\n",
    "#for exporting to Adobe Illustrator\n",
    "mpl.rcParams['pdf.fonttype'] = 42\n",
    "mpl.rcParams['ps.fonttype'] = 42"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "######################################################################################################################################################################################################################################################################"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Sample Annotation for all (unfiltered) Longitudinal Isolate pairs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "######################################################################################################################################################################################################################################################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_annotation_longitudinal = pd.read_csv('/n/data1/hms/dbmi/farhat/Roger/inhost_TB_dynamics_project/CSV_files/sample_annotation_files/cetr_casali_walker_trauner_witney_xu_guerra_bryant_fastq_path_names_and_JankyPipe_tags.csv' , sep  = ',').set_index('patient_id')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(614, 7)"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.shape(sample_annotation_longitudinal)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>fastq_files</th>\n",
       "      <th>population</th>\n",
       "      <th>run_ID</th>\n",
       "      <th>sample_ID</th>\n",
       "      <th>sample_order</th>\n",
       "      <th>tag</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>patient_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>I0005973-8</th>\n",
       "      <td>0</td>\n",
       "      <td>/n/data1/hms/dbmi/farhat/cetr_strains/good_wgs...</td>\n",
       "      <td>CETR</td>\n",
       "      <td>MQFF00000000</td>\n",
       "      <td>Peru3062</td>\n",
       "      <td>1</td>\n",
       "      <td>Peru3062</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>I0005973-8</th>\n",
       "      <td>1</td>\n",
       "      <td>/n/data1/hms/dbmi/farhat/cetr_strains/good_wgs...</td>\n",
       "      <td>CETR</td>\n",
       "      <td>MQKH00000000</td>\n",
       "      <td>Peru3315</td>\n",
       "      <td>2</td>\n",
       "      <td>Peru3315</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>I0005229-5</th>\n",
       "      <td>2</td>\n",
       "      <td>/n/data1/hms/dbmi/farhat/cetr_strains/good_wgs...</td>\n",
       "      <td>CETR</td>\n",
       "      <td>MQAO00000000</td>\n",
       "      <td>Peru2908</td>\n",
       "      <td>1</td>\n",
       "      <td>Peru2908</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>I0005229-5</th>\n",
       "      <td>3</td>\n",
       "      <td>/n/data1/hms/dbmi/farhat/cetr_strains/good_wgs...</td>\n",
       "      <td>CETR</td>\n",
       "      <td>MQKD00000000</td>\n",
       "      <td>Peru3278</td>\n",
       "      <td>2</td>\n",
       "      <td>Peru3278</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>I0005235-2</th>\n",
       "      <td>4</td>\n",
       "      <td>/n/data1/hms/dbmi/farhat/cetr_strains/good_wgs...</td>\n",
       "      <td>CETR</td>\n",
       "      <td>MQBA00000000</td>\n",
       "      <td>Peru2921</td>\n",
       "      <td>1</td>\n",
       "      <td>Peru2921</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            Unnamed: 0                                        fastq_files  \\\n",
       "patient_id                                                                  \n",
       "I0005973-8           0  /n/data1/hms/dbmi/farhat/cetr_strains/good_wgs...   \n",
       "I0005973-8           1  /n/data1/hms/dbmi/farhat/cetr_strains/good_wgs...   \n",
       "I0005229-5           2  /n/data1/hms/dbmi/farhat/cetr_strains/good_wgs...   \n",
       "I0005229-5           3  /n/data1/hms/dbmi/farhat/cetr_strains/good_wgs...   \n",
       "I0005235-2           4  /n/data1/hms/dbmi/farhat/cetr_strains/good_wgs...   \n",
       "\n",
       "           population        run_ID sample_ID  sample_order       tag  \n",
       "patient_id                                                             \n",
       "I0005973-8       CETR  MQFF00000000  Peru3062             1  Peru3062  \n",
       "I0005973-8       CETR  MQKH00000000  Peru3315             2  Peru3315  \n",
       "I0005229-5       CETR  MQAO00000000  Peru2908             1  Peru2908  \n",
       "I0005229-5       CETR  MQKD00000000  Peru3278             2  Peru3278  \n",
       "I0005235-2       CETR  MQBA00000000  Peru2921             1  Peru2921  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_annotation_longitudinal.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### F2 Mixed Measure contaminated samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "patients_contaminated_by_F2 = ['3162L', 'I0004290-8', '2459', 'KPS_45', 'KPS_40', 'KPS_43', 'KPS_42', '3145', 'KPS_83', 'KPS_8', '2014W', '2', '1674', '1671', 'KPS_32', 'KPS_78', '1117N', '3641', '21', '1646', 'I0001485-7', 'I0002074-8', '3443', 'KPS_65', 'KPS_63', 'KPS_69', '1669', '1667', 'KPS_20', '4098U', 'KPS_25']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "31"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(patients_contaminated_by_F2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_annotation_F2_contaminated = sample_annotation_longitudinal.loc[patients_contaminated_by_F2 , :]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(62, 7)"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.shape(sample_annotation_F2_contaminated)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>fastq_files</th>\n",
       "      <th>population</th>\n",
       "      <th>run_ID</th>\n",
       "      <th>sample_ID</th>\n",
       "      <th>sample_order</th>\n",
       "      <th>tag</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>patient_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3162L</th>\n",
       "      <td>560</td>\n",
       "      <td>/n/data1/hms/dbmi/farhat/fastq_db/bryant/ERR17...</td>\n",
       "      <td>BRYANT</td>\n",
       "      <td>ERR171166</td>\n",
       "      <td>401391.0</td>\n",
       "      <td>1</td>\n",
       "      <td>ERR171166</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3162L</th>\n",
       "      <td>561</td>\n",
       "      <td>/n/data1/hms/dbmi/farhat/fastq_db/bryant/ERR17...</td>\n",
       "      <td>BRYANT</td>\n",
       "      <td>ERR171167</td>\n",
       "      <td>402713.0</td>\n",
       "      <td>2</td>\n",
       "      <td>ERR171167</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>I0004290-8</th>\n",
       "      <td>10</td>\n",
       "      <td>/n/data1/hms/dbmi/farhat/cetr_strains/good_wgs...</td>\n",
       "      <td>CETR</td>\n",
       "      <td>MQDN00000000</td>\n",
       "      <td>Peru3011</td>\n",
       "      <td>1</td>\n",
       "      <td>Peru3011</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>I0004290-8</th>\n",
       "      <td>11</td>\n",
       "      <td>/n/data1/hms/dbmi/farhat/cetr_strains/good_wgs...</td>\n",
       "      <td>CETR</td>\n",
       "      <td>MQIU00000000</td>\n",
       "      <td>Peru3365</td>\n",
       "      <td>2</td>\n",
       "      <td>Peru3365</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2459</th>\n",
       "      <td>110</td>\n",
       "      <td>/n/data1/hms/dbmi/farhat/cetr_strains/good_wgs...</td>\n",
       "      <td>CETR</td>\n",
       "      <td>MMYC00000000</td>\n",
       "      <td>Peru4946</td>\n",
       "      <td>1</td>\n",
       "      <td>Peru4946</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            Unnamed: 0                                        fastq_files  \\\n",
       "patient_id                                                                  \n",
       "3162L              560  /n/data1/hms/dbmi/farhat/fastq_db/bryant/ERR17...   \n",
       "3162L              561  /n/data1/hms/dbmi/farhat/fastq_db/bryant/ERR17...   \n",
       "I0004290-8          10  /n/data1/hms/dbmi/farhat/cetr_strains/good_wgs...   \n",
       "I0004290-8          11  /n/data1/hms/dbmi/farhat/cetr_strains/good_wgs...   \n",
       "2459               110  /n/data1/hms/dbmi/farhat/cetr_strains/good_wgs...   \n",
       "\n",
       "           population        run_ID sample_ID  sample_order        tag  \n",
       "patient_id                                                              \n",
       "3162L          BRYANT     ERR171166  401391.0             1  ERR171166  \n",
       "3162L          BRYANT     ERR171167  402713.0             2  ERR171167  \n",
       "I0004290-8       CETR  MQDN00000000  Peru3011             1   Peru3011  \n",
       "I0004290-8       CETR  MQIU00000000  Peru3365             2   Peru3365  \n",
       "2459             CETR  MMYC00000000  Peru4946             1   Peru4946  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_annotation_F2_contaminated.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Export Sample Annotation to CSV for modifying by inclusion of times between isolate collection"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/rv76/anaconda2/lib/python2.7/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "  \"\"\"Entry point for launching an IPython kernel.\n",
      "/home/rv76/anaconda2/lib/python2.7/site-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "  \n",
      "/home/rv76/anaconda2/lib/python2.7/site-packages/ipykernel_launcher.py:3: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "  This is separate from the ipykernel package so we can avoid doing imports until\n"
     ]
    }
   ],
   "source": [
    "sample_annotation_F2_contaminated.drop(labels = ['fastq_files' , 'Unnamed: 0'] , axis = 1 , inplace = True)\n",
    "sample_annotation_F2_contaminated['patient_id'] = list(sample_annotation_F2_contaminated.index)\n",
    "sample_annotation_F2_contaminated.sort_values(by = ['population' , 'patient_id' , 'sample_order'] , inplace = True)\n",
    "sample_annotation_F2_contaminated.reset_index(inplace = True , drop = True)\n",
    "sample_annotation_F2_contaminated.to_csv('/n/data1/hms/dbmi/farhat/Roger/inhost_TB_dynamics_project/CSV_files/sample_annotation_files/contaminated_samples/F2_contaminated_longitudinal_sample_annotation.csv' , sep  = ',')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>population</th>\n",
       "      <th>run_ID</th>\n",
       "      <th>sample_ID</th>\n",
       "      <th>sample_order</th>\n",
       "      <th>tag</th>\n",
       "      <th>patient_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BRYANT</td>\n",
       "      <td>ERR171155</td>\n",
       "      <td>1500034.0</td>\n",
       "      <td>1</td>\n",
       "      <td>ERR171155</td>\n",
       "      <td>1117N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BRYANT</td>\n",
       "      <td>ERR171154</td>\n",
       "      <td>1500283.0</td>\n",
       "      <td>2</td>\n",
       "      <td>ERR171154</td>\n",
       "      <td>1117N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BRYANT</td>\n",
       "      <td>ERR108127</td>\n",
       "      <td>300239.0</td>\n",
       "      <td>1</td>\n",
       "      <td>ERR108127</td>\n",
       "      <td>2014W</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BRYANT</td>\n",
       "      <td>ERR108128</td>\n",
       "      <td>301725.0</td>\n",
       "      <td>2</td>\n",
       "      <td>ERR108128</td>\n",
       "      <td>2014W</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BRYANT</td>\n",
       "      <td>ERR171166</td>\n",
       "      <td>401391.0</td>\n",
       "      <td>1</td>\n",
       "      <td>ERR171166</td>\n",
       "      <td>3162L</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  population     run_ID  sample_ID  sample_order        tag patient_id\n",
       "0     BRYANT  ERR171155  1500034.0             1  ERR171155      1117N\n",
       "1     BRYANT  ERR171154  1500283.0             2  ERR171154      1117N\n",
       "2     BRYANT  ERR108127   300239.0             1  ERR108127      2014W\n",
       "3     BRYANT  ERR108128   301725.0             2  ERR108128      2014W\n",
       "4     BRYANT  ERR171166   401391.0             1  ERR171166      3162L"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_annotation_F2_contaminated.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### fSNP Threshold  contaminated samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "patients_contaminated_by_fSNP = ['2718', 'KPS_46', 'KPS_41', '3103J', '4085U', 'KPS_80', 'I0004304-7', '3439', 'KPS_11', '3430', 'KPS_2', '1941', 'KPS_56', '1682', 'I0000221-7', '4', '1678', 'KPS_36', 'KPS_30', 'KPS_31', '4078X', 'B', 'KPS_59', 'KPS_70', 'KPS_73', 'KPS_52', 'KPS_50', 'I0001031-9', '3725', 'KPS_18', '1620', '15', 'I0004240-3', 'I0004172-8', '2492', '1657', 'KPS_67', '3100U', 'KPS_29', 'KPS_64', '3018C', '2047T', 'I0003710-6', 'P000035']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "44"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(patients_contaminated_by_fSNP)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_annotation_fSNP_contaminated = sample_annotation_longitudinal.loc[patients_contaminated_by_fSNP , :]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(88, 7)"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.shape(sample_annotation_fSNP_contaminated)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>fastq_files</th>\n",
       "      <th>population</th>\n",
       "      <th>run_ID</th>\n",
       "      <th>sample_ID</th>\n",
       "      <th>sample_order</th>\n",
       "      <th>tag</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>patient_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2718</th>\n",
       "      <td>124</td>\n",
       "      <td>/n/data1/hms/dbmi/farhat/cetr_strains/good_wgs...</td>\n",
       "      <td>CETR</td>\n",
       "      <td>MRDO00000000</td>\n",
       "      <td>Peru5134</td>\n",
       "      <td>1</td>\n",
       "      <td>Peru5134</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2718</th>\n",
       "      <td>125</td>\n",
       "      <td>/n/data1/hms/dbmi/farhat/cetr_strains/good_wgs...</td>\n",
       "      <td>CETR</td>\n",
       "      <td>MMGA00000000</td>\n",
       "      <td>Peru4741</td>\n",
       "      <td>2</td>\n",
       "      <td>Peru4741</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KPS_46</th>\n",
       "      <td>450</td>\n",
       "      <td>/n/data1/hms/dbmi/farhat/fastq_db/guerra/ERR17...</td>\n",
       "      <td>GUERRA</td>\n",
       "      <td>ERR176649</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>ERR176649</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KPS_46</th>\n",
       "      <td>451</td>\n",
       "      <td>/n/data1/hms/dbmi/farhat/fastq_db/guerra/ERR19...</td>\n",
       "      <td>GUERRA</td>\n",
       "      <td>ERR190334</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>ERR190334</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KPS_41</th>\n",
       "      <td>440</td>\n",
       "      <td>/n/data1/hms/dbmi/farhat/fastq_db/guerra/ERR17...</td>\n",
       "      <td>GUERRA</td>\n",
       "      <td>ERR176764</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>ERR176764</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            Unnamed: 0                                        fastq_files  \\\n",
       "patient_id                                                                  \n",
       "2718               124  /n/data1/hms/dbmi/farhat/cetr_strains/good_wgs...   \n",
       "2718               125  /n/data1/hms/dbmi/farhat/cetr_strains/good_wgs...   \n",
       "KPS_46             450  /n/data1/hms/dbmi/farhat/fastq_db/guerra/ERR17...   \n",
       "KPS_46             451  /n/data1/hms/dbmi/farhat/fastq_db/guerra/ERR19...   \n",
       "KPS_41             440  /n/data1/hms/dbmi/farhat/fastq_db/guerra/ERR17...   \n",
       "\n",
       "           population        run_ID sample_ID  sample_order        tag  \n",
       "patient_id                                                              \n",
       "2718             CETR  MRDO00000000  Peru5134             1   Peru5134  \n",
       "2718             CETR  MMGA00000000  Peru4741             2   Peru4741  \n",
       "KPS_46         GUERRA     ERR176649       NaN             1  ERR176649  \n",
       "KPS_46         GUERRA     ERR190334       NaN             2  ERR190334  \n",
       "KPS_41         GUERRA     ERR176764       NaN             1  ERR176764  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_annotation_fSNP_contaminated.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Export Sample Annotation to CSV for modifying by inclusion of times between isolate collection"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/rv76/anaconda2/lib/python2.7/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "  \"\"\"Entry point for launching an IPython kernel.\n",
      "/home/rv76/anaconda2/lib/python2.7/site-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "  \n",
      "/home/rv76/anaconda2/lib/python2.7/site-packages/ipykernel_launcher.py:3: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "  This is separate from the ipykernel package so we can avoid doing imports until\n"
     ]
    }
   ],
   "source": [
    "sample_annotation_fSNP_contaminated.drop(labels = ['fastq_files' , 'Unnamed: 0'] , axis = 1 , inplace = True)\n",
    "sample_annotation_fSNP_contaminated['patient_id'] = list(sample_annotation_fSNP_contaminated.index)\n",
    "sample_annotation_fSNP_contaminated.sort_values(by = ['population' , 'patient_id' , 'sample_order'] , inplace = True)\n",
    "sample_annotation_fSNP_contaminated.reset_index(inplace = True , drop = True)\n",
    "sample_annotation_fSNP_contaminated.to_csv('/n/data1/hms/dbmi/farhat/Roger/inhost_TB_dynamics_project/CSV_files/sample_annotation_files/contaminated_samples/fSNP_contaminated_longitudinal_sample_annotation.csv' , sep  = ',')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>population</th>\n",
       "      <th>run_ID</th>\n",
       "      <th>sample_ID</th>\n",
       "      <th>sample_order</th>\n",
       "      <th>tag</th>\n",
       "      <th>patient_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BRYANT</td>\n",
       "      <td>ERR966621</td>\n",
       "      <td>401063.0</td>\n",
       "      <td>1</td>\n",
       "      <td>ERR966621</td>\n",
       "      <td>2047T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BRYANT</td>\n",
       "      <td>ERR108140</td>\n",
       "      <td>402106.0</td>\n",
       "      <td>2</td>\n",
       "      <td>ERR108140</td>\n",
       "      <td>2047T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BRYANT</td>\n",
       "      <td>ERR108143</td>\n",
       "      <td>300191.0</td>\n",
       "      <td>1</td>\n",
       "      <td>ERR108143</td>\n",
       "      <td>3018C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BRYANT</td>\n",
       "      <td>ERR108144</td>\n",
       "      <td>302235.0</td>\n",
       "      <td>2</td>\n",
       "      <td>ERR108144</td>\n",
       "      <td>3018C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BRYANT</td>\n",
       "      <td>ERR108152</td>\n",
       "      <td>300659.0</td>\n",
       "      <td>1</td>\n",
       "      <td>ERR108152</td>\n",
       "      <td>3100U</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  population     run_ID sample_ID  sample_order        tag patient_id\n",
       "0     BRYANT  ERR966621  401063.0             1  ERR966621      2047T\n",
       "1     BRYANT  ERR108140  402106.0             2  ERR108140      2047T\n",
       "2     BRYANT  ERR108143  300191.0             1  ERR108143      3018C\n",
       "3     BRYANT  ERR108144  302235.0             2  ERR108144      3018C\n",
       "4     BRYANT  ERR108152  300659.0             1  ERR108152      3100U"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_annotation_fSNP_contaminated.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "######################################################################################################################################################################################################################################################################"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Import Annotation Files with Dates and get *time between isolate collection* for different groups of subjects"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "######################################################################################################################################################################################################################################################################"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### A: F2 Mixed Measure (*mixed*)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "time_between_serial_isolate_collection = pd.read_csv('/n/data1/hms/dbmi/farhat/Roger/inhost_TB_dynamics_project/CSV_files/sample_annotation_files/contaminated_samples/F2_contaminated_longitudinal_sample_annotation_with_collection_times.csv').set_index('population')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>run_ID</th>\n",
       "      <th>sample_ID</th>\n",
       "      <th>sample_order</th>\n",
       "      <th>tag</th>\n",
       "      <th>patient_id</th>\n",
       "      <th>date_information</th>\n",
       "      <th>tube_label (cetr)</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>population</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>BRYANT</th>\n",
       "      <td>ERR171155</td>\n",
       "      <td>1500034</td>\n",
       "      <td>1</td>\n",
       "      <td>ERR171155</td>\n",
       "      <td>1117N</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BRYANT</th>\n",
       "      <td>ERR171154</td>\n",
       "      <td>1500283</td>\n",
       "      <td>2</td>\n",
       "      <td>ERR171154</td>\n",
       "      <td>1117N</td>\n",
       "      <td>17</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BRYANT</th>\n",
       "      <td>ERR108127</td>\n",
       "      <td>300239</td>\n",
       "      <td>1</td>\n",
       "      <td>ERR108127</td>\n",
       "      <td>2014W</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BRYANT</th>\n",
       "      <td>ERR108128</td>\n",
       "      <td>301725</td>\n",
       "      <td>2</td>\n",
       "      <td>ERR108128</td>\n",
       "      <td>2014W</td>\n",
       "      <td>36</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BRYANT</th>\n",
       "      <td>ERR171166</td>\n",
       "      <td>401391</td>\n",
       "      <td>1</td>\n",
       "      <td>ERR171166</td>\n",
       "      <td>3162L</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               run_ID sample_ID  sample_order        tag patient_id  \\\n",
       "population                                                            \n",
       "BRYANT      ERR171155   1500034             1  ERR171155      1117N   \n",
       "BRYANT      ERR171154   1500283             2  ERR171154      1117N   \n",
       "BRYANT      ERR108127    300239             1  ERR108127      2014W   \n",
       "BRYANT      ERR108128    301725             2  ERR108128      2014W   \n",
       "BRYANT      ERR171166    401391             1  ERR171166      3162L   \n",
       "\n",
       "           date_information tube_label (cetr)  \n",
       "population                                     \n",
       "BRYANT                    0               NaN  \n",
       "BRYANT                   17               NaN  \n",
       "BRYANT                    0               NaN  \n",
       "BRYANT                   36               NaN  \n",
       "BRYANT                    0               NaN  "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "time_between_serial_isolate_collection.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(62, 7)"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.shape(time_between_serial_isolate_collection)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Retain only patients that for which we have date Information for both samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "#get dates for each sample in pair\n",
    "dates_for_first_sample = list( time_between_serial_isolate_collection.iloc[range(0 , np.shape(time_between_serial_isolate_collection)[0] , 2) , 5] )\n",
    "dates_for_second_sample = list( time_between_serial_isolate_collection.iloc[range(1 , np.shape(time_between_serial_isolate_collection)[0] , 2) , 5] )\n",
    "\n",
    "#get a boolean to check that there are dates for both samples\n",
    "dates_for_both_samples = [ ( (isinstance(date_1, basestring)) and (isinstance(date_2, basestring)) ) for date_1 , date_2 in zip(dates_for_first_sample , dates_for_second_sample) ]\n",
    "\n",
    "#get list of patients to include that have dates for sample retreival for both samples\n",
    "patients_to_keep = list( time_between_serial_isolate_collection[time_between_serial_isolate_collection.sample_order == 1][dates_for_both_samples].patient_id )\n",
    "\n",
    "#filter out patients with 1 or 2 missing dates\n",
    "time_between_serial_isolate_collection_filter = [patient in patients_to_keep for patient in list(time_between_serial_isolate_collection.patient_id)]\n",
    "time_between_serial_isolate_dropped_filter = [not dates_both_isolates for dates_both_isolates in time_between_serial_isolate_collection_filter]\n",
    "\n",
    "time_between_serial_isolate_collection_dropped = time_between_serial_isolate_collection[time_between_serial_isolate_dropped_filter]\n",
    "time_between_serial_isolate_collection = time_between_serial_isolate_collection[time_between_serial_isolate_collection_filter]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(62, 7)"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.shape(time_between_serial_isolate_collection)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(0, 7)"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.shape(time_between_serial_isolate_collection_dropped)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Retain only patients that have different dates of collection for both samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "#get dates for each sample in pair\n",
    "dates_for_first_sample = list( time_between_serial_isolate_collection.iloc[range(0 , np.shape(time_between_serial_isolate_collection)[0] , 2) , 5] )\n",
    "dates_for_second_sample = list( time_between_serial_isolate_collection.iloc[range(1 , np.shape(time_between_serial_isolate_collection)[0] , 2) , 5] )\n",
    "\n",
    "#get a boolean to check that there are dates for both samples\n",
    "dates_different_for_both_samples = [ ( date_1 != date_2 ) for date_1 , date_2 in zip(dates_for_first_sample , dates_for_second_sample) ]\n",
    "\n",
    "#get list of patients to include that have dates for sample retreival for both samples\n",
    "patients_to_keep = list( time_between_serial_isolate_collection[time_between_serial_isolate_collection.sample_order == 1][dates_different_for_both_samples].patient_id )\n",
    "\n",
    "#filter out patients with 1 or 2 missing dates\n",
    "time_between_serial_isolate_collection_filter = [patient in patients_to_keep for patient in list(time_between_serial_isolate_collection.patient_id)]\n",
    "time_between_serial_isolate_dropped_filter = [not dates_both_isolates for dates_both_isolates in time_between_serial_isolate_collection_filter]\n",
    "\n",
    "time_between_serial_isolate_collection_dropped = time_between_serial_isolate_collection[time_between_serial_isolate_dropped_filter]\n",
    "time_between_serial_isolate_collection = time_between_serial_isolate_collection[time_between_serial_isolate_collection_filter]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(62, 7)"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.shape(time_between_serial_isolate_collection)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(0, 7)"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.shape(time_between_serial_isolate_collection_dropped )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>run_ID</th>\n",
       "      <th>sample_ID</th>\n",
       "      <th>sample_order</th>\n",
       "      <th>tag</th>\n",
       "      <th>patient_id</th>\n",
       "      <th>date_information</th>\n",
       "      <th>tube_label (cetr)</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>population</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [run_ID, sample_ID, sample_order, tag, patient_id, date_information, tube_label (cetr)]\n",
       "Index: []"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "time_between_serial_isolate_collection_dropped "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Find distance between all pairs of times"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "pairwise_isolate_time = pd.DataFrame(index = patients_to_keep , columns = ['population' , 'time_between_collection'])\n",
    "\n",
    "for patient_id in patients_to_keep:\n",
    "    \n",
    "    population = time_between_serial_isolate_collection[time_between_serial_isolate_collection.patient_id == patient_id].index[0]\n",
    "    \n",
    "    #append info to dataframe\n",
    "    pairwise_isolate_time.loc[patient_id , 'population'] = population"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>population</th>\n",
       "      <th>time_between_collection</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1117N</th>\n",
       "      <td>BRYANT</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2014W</th>\n",
       "      <td>BRYANT</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3162L</th>\n",
       "      <td>BRYANT</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4098U</th>\n",
       "      <td>BRYANT</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1646</th>\n",
       "      <td>CETR</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      population time_between_collection\n",
       "1117N     BRYANT                     NaN\n",
       "2014W     BRYANT                     NaN\n",
       "3162L     BRYANT                     NaN\n",
       "4098U     BRYANT                     NaN\n",
       "1646        CETR                     NaN"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pairwise_isolate_time.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "for patient_id in pairwise_isolate_time.index:\n",
    "    \n",
    "    population = pairwise_isolate_time.loc[patient_id , 'population']\n",
    "    patient_id = patient_id\n",
    "    \n",
    "    if population == 'BRYANT':\n",
    "        \n",
    "        bryant_pair = time_between_serial_isolate_collection[time_between_serial_isolate_collection.patient_id == patient_id]\n",
    "        \n",
    "        first_date = bryant_pair[bryant_pair.sample_order == 1].date_information.values[0]\n",
    "        second_date = bryant_pair[bryant_pair.sample_order == 2].date_information.values[0]\n",
    "\n",
    "        num_days_between = (int(second_date) - int(first_date))*7 #dates are in # of weeks apart\n",
    "    \n",
    "    if population == 'CASALI':\n",
    "        \n",
    "        casali_pair = time_between_serial_isolate_collection[time_between_serial_isolate_collection.patient_id == patient_id]\n",
    "\n",
    "        first_date = casali_pair[casali_pair.sample_order == 1].date_information.values[0]\n",
    "        day = first_date.split('/')[0]\n",
    "        month = first_date.split('/')[1]\n",
    "        year = '20' + first_date.split('/')[2]\n",
    "        first_date = pd.to_datetime(pd.DataFrame({'year':[year] , 'month':[month] , 'day':[day]}))\n",
    "\n",
    "        second_date = casali_pair[casali_pair.sample_order == 2].date_information.values[0]\n",
    "        day = second_date.split('/')[0]\n",
    "        month = second_date.split('/')[1]\n",
    "        year = '20' + second_date.split('/')[2]\n",
    "        second_date = pd.to_datetime(pd.DataFrame({'year':[year] , 'month':[month] , 'day':[day]}))\n",
    "\n",
    "        num_days_between = (second_date - first_date).dt.days[0]\n",
    "        \n",
    "    elif (population == 'CETR') or (population == 'GUERRA') or (population == 'WITNEY') or (population == 'XU'):\n",
    "        \n",
    "        serial_pair = time_between_serial_isolate_collection[time_between_serial_isolate_collection.patient_id == patient_id]\n",
    "\n",
    "        first_date = serial_pair[serial_pair.sample_order == 1].date_information.values[0]\n",
    "        day = first_date.split('/')[1]\n",
    "        month = first_date.split('/')[0]\n",
    "        if int(first_date.split('/')[2]) in range(50,100): #some dates in 1990's others in 2000's\n",
    "            year = '19' + first_date.split('/')[2]\n",
    "        else:\n",
    "            year = '20' + first_date.split('/')[2]\n",
    "\n",
    "        first_date = pd.to_datetime(pd.DataFrame({'year':[year] , 'month':[month] , 'day':[day]}))\n",
    "\n",
    "\n",
    "        second_date = serial_pair[serial_pair.sample_order == 2].date_information.values[0]\n",
    "        day = second_date.split('/')[1]\n",
    "        month = second_date.split('/')[0]\n",
    "        if int(second_date.split('/')[2]) in range(50,100):\n",
    "            year = '19' + second_date.split('/')[2]\n",
    "        else:\n",
    "            year = '20' + second_date.split('/')[2]\n",
    "\n",
    "        second_date = pd.to_datetime(pd.DataFrame({'year':[year] , 'month':[month] , 'day':[day]}))\n",
    "\n",
    "        num_days_between = (second_date - first_date).dt.days[0]\n",
    "        \n",
    "    elif (population == 'TRAUNER') or (population == 'BRYANT'):\n",
    "        \n",
    "        serial_pair = time_between_serial_isolate_collection[time_between_serial_isolate_collection.patient_id == patient_id]\n",
    "\n",
    "        first_date = serial_pair[serial_pair.sample_order == 1].date_information.values[0]\n",
    "        second_date = serial_pair[serial_pair.sample_order == 2].date_information.values[0]\n",
    "\n",
    "        num_days_between = (int(second_date) - int(first_date))*7 #dates are in # of weeks apart\n",
    "        \n",
    "    elif population == 'WALKER':\n",
    "        \n",
    "        walker_pair = time_between_serial_isolate_collection[time_between_serial_isolate_collection.patient_id == patient_id]\n",
    "\n",
    "        first_date = walker_pair[walker_pair.sample_order == 1].date_information.values[0]\n",
    "        day = '1'\n",
    "        month = '1'\n",
    "        year = first_date\n",
    "        first_date = pd.to_datetime(pd.DataFrame({'year':[year] , 'month':[month] , 'day':[day]}))\n",
    "\n",
    "        second_date = walker_pair[walker_pair.sample_order == 2].date_information.values[0]\n",
    "        day = '1'\n",
    "        month = '1'\n",
    "        year = second_date\n",
    "        second_date = pd.to_datetime(pd.DataFrame({'year':[year] , 'month':[month] , 'day':[day]}))\n",
    "\n",
    "        num_days_between = (second_date - first_date).dt.days[0]\n",
    "        \n",
    "    #append distance between isolate collection\n",
    "    pairwise_isolate_time.loc[patient_id , 'time_between_collection'] = num_days_between"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>population</th>\n",
       "      <th>time_between_collection</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1117N</th>\n",
       "      <td>BRYANT</td>\n",
       "      <td>119</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2014W</th>\n",
       "      <td>BRYANT</td>\n",
       "      <td>252</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3162L</th>\n",
       "      <td>BRYANT</td>\n",
       "      <td>252</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4098U</th>\n",
       "      <td>BRYANT</td>\n",
       "      <td>336</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1646</th>\n",
       "      <td>CETR</td>\n",
       "      <td>374</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      population time_between_collection\n",
       "1117N     BRYANT                     119\n",
       "2014W     BRYANT                     252\n",
       "3162L     BRYANT                     252\n",
       "4098U     BRYANT                     336\n",
       "1646        CETR                     374"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pairwise_isolate_time.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "647.1935483870968"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.mean( pairwise_isolate_time.time_between_collection )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "F2_contaminated_pairwise_isolate_times = pairwise_isolate_time.time_between_collection"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### B: fSNP Measure  (*re-infection*)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "time_between_serial_isolate_collection = pd.read_csv('/n/data1/hms/dbmi/farhat/Roger/inhost_TB_dynamics_project/CSV_files/sample_annotation_files/contaminated_samples/fSNP_contaminated_longitudinal_sample_annotation_with_collection_times.csv').set_index('population')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>run_ID</th>\n",
       "      <th>sample_ID</th>\n",
       "      <th>sample_order</th>\n",
       "      <th>tag</th>\n",
       "      <th>patient_id</th>\n",
       "      <th>date_information</th>\n",
       "      <th>tube_label (cetr)</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>population</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>BRYANT</th>\n",
       "      <td>ERR966621</td>\n",
       "      <td>401063</td>\n",
       "      <td>1</td>\n",
       "      <td>ERR966621</td>\n",
       "      <td>2047T</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BRYANT</th>\n",
       "      <td>ERR108140</td>\n",
       "      <td>402106</td>\n",
       "      <td>2</td>\n",
       "      <td>ERR108140</td>\n",
       "      <td>2047T</td>\n",
       "      <td>26</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BRYANT</th>\n",
       "      <td>ERR108143</td>\n",
       "      <td>300191</td>\n",
       "      <td>1</td>\n",
       "      <td>ERR108143</td>\n",
       "      <td>3018C</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BRYANT</th>\n",
       "      <td>ERR108144</td>\n",
       "      <td>302235</td>\n",
       "      <td>2</td>\n",
       "      <td>ERR108144</td>\n",
       "      <td>3018C</td>\n",
       "      <td>48</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BRYANT</th>\n",
       "      <td>ERR108152</td>\n",
       "      <td>300659</td>\n",
       "      <td>1</td>\n",
       "      <td>ERR108152</td>\n",
       "      <td>3100U</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               run_ID sample_ID  sample_order        tag patient_id  \\\n",
       "population                                                            \n",
       "BRYANT      ERR966621    401063             1  ERR966621      2047T   \n",
       "BRYANT      ERR108140    402106             2  ERR108140      2047T   \n",
       "BRYANT      ERR108143    300191             1  ERR108143      3018C   \n",
       "BRYANT      ERR108144    302235             2  ERR108144      3018C   \n",
       "BRYANT      ERR108152    300659             1  ERR108152      3100U   \n",
       "\n",
       "           date_information tube_label (cetr)  \n",
       "population                                     \n",
       "BRYANT                    0               NaN  \n",
       "BRYANT                   26               NaN  \n",
       "BRYANT                    0               NaN  \n",
       "BRYANT                   48               NaN  \n",
       "BRYANT                    0               NaN  "
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "time_between_serial_isolate_collection.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(88, 7)"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.shape(time_between_serial_isolate_collection)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Retain only patients that for which we have date Information for both samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "#get dates for each sample in pair\n",
    "dates_for_first_sample = list( time_between_serial_isolate_collection.iloc[range(0 , np.shape(time_between_serial_isolate_collection)[0] , 2) , 5] )\n",
    "dates_for_second_sample = list( time_between_serial_isolate_collection.iloc[range(1 , np.shape(time_between_serial_isolate_collection)[0] , 2) , 5] )\n",
    "\n",
    "#get a boolean to check that there are dates for both samples\n",
    "dates_for_both_samples = [ ( (isinstance(date_1, basestring)) and (isinstance(date_2, basestring)) ) for date_1 , date_2 in zip(dates_for_first_sample , dates_for_second_sample) ]\n",
    "\n",
    "#get list of patients to include that have dates for sample retreival for both samples\n",
    "patients_to_keep = list( time_between_serial_isolate_collection[time_between_serial_isolate_collection.sample_order == 1][dates_for_both_samples].patient_id )\n",
    "\n",
    "#filter out patients with 1 or 2 missing dates\n",
    "time_between_serial_isolate_collection_filter = [patient in patients_to_keep for patient in list(time_between_serial_isolate_collection.patient_id)]\n",
    "time_between_serial_isolate_dropped_filter = [not dates_both_isolates for dates_both_isolates in time_between_serial_isolate_collection_filter]\n",
    "\n",
    "time_between_serial_isolate_collection_dropped = time_between_serial_isolate_collection[time_between_serial_isolate_dropped_filter]\n",
    "time_between_serial_isolate_collection = time_between_serial_isolate_collection[time_between_serial_isolate_collection_filter]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(86, 7)"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.shape(time_between_serial_isolate_collection)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(2, 7)"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.shape(time_between_serial_isolate_collection_dropped )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>run_ID</th>\n",
       "      <th>sample_ID</th>\n",
       "      <th>sample_order</th>\n",
       "      <th>tag</th>\n",
       "      <th>patient_id</th>\n",
       "      <th>date_information</th>\n",
       "      <th>tube_label (cetr)</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>population</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>BRYANT</th>\n",
       "      <td>ERR171136</td>\n",
       "      <td>200470</td>\n",
       "      <td>1</td>\n",
       "      <td>ERR171136</td>\n",
       "      <td>4085U</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BRYANT</th>\n",
       "      <td>ERR171137</td>\n",
       "      <td>201113</td>\n",
       "      <td>2</td>\n",
       "      <td>ERR171137</td>\n",
       "      <td>4085U</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               run_ID sample_ID  sample_order        tag patient_id  \\\n",
       "population                                                            \n",
       "BRYANT      ERR171136    200470             1  ERR171136      4085U   \n",
       "BRYANT      ERR171137    201113             2  ERR171137      4085U   \n",
       "\n",
       "           date_information tube_label (cetr)  \n",
       "population                                     \n",
       "BRYANT                    0               NaN  \n",
       "BRYANT                  NaN               NaN  "
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "time_between_serial_isolate_collection_dropped "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Dropped 1 patient from Bryant et. al. dataset that didn't have date information both both isolates "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Retain only patients that have different dates of collection for both samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "#get dates for each sample in pair\n",
    "dates_for_first_sample = list( time_between_serial_isolate_collection.iloc[range(0 , np.shape(time_between_serial_isolate_collection)[0] , 2) , 5] )\n",
    "dates_for_second_sample = list( time_between_serial_isolate_collection.iloc[range(1 , np.shape(time_between_serial_isolate_collection)[0] , 2) , 5] )\n",
    "\n",
    "#get a boolean to check that there are dates for both samples\n",
    "dates_different_for_both_samples = [ ( date_1 != date_2 ) for date_1 , date_2 in zip(dates_for_first_sample , dates_for_second_sample) ]\n",
    "\n",
    "#get list of patients to include that have dates for sample retreival for both samples\n",
    "patients_to_keep = list( time_between_serial_isolate_collection[time_between_serial_isolate_collection.sample_order == 1][dates_different_for_both_samples].patient_id )\n",
    "\n",
    "#filter out patients with 1 or 2 missing dates\n",
    "time_between_serial_isolate_collection_filter = [patient in patients_to_keep for patient in list(time_between_serial_isolate_collection.patient_id)]\n",
    "time_between_serial_isolate_dropped_filter = [not dates_both_isolates for dates_both_isolates in time_between_serial_isolate_collection_filter]\n",
    "\n",
    "time_between_serial_isolate_collection_dropped = time_between_serial_isolate_collection[time_between_serial_isolate_dropped_filter]\n",
    "time_between_serial_isolate_collection = time_between_serial_isolate_collection[time_between_serial_isolate_collection_filter]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(86, 7)"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.shape(time_between_serial_isolate_collection)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(0, 7)"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.shape(time_between_serial_isolate_collection_dropped )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>run_ID</th>\n",
       "      <th>sample_ID</th>\n",
       "      <th>sample_order</th>\n",
       "      <th>tag</th>\n",
       "      <th>patient_id</th>\n",
       "      <th>date_information</th>\n",
       "      <th>tube_label (cetr)</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>population</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [run_ID, sample_ID, sample_order, tag, patient_id, date_information, tube_label (cetr)]\n",
       "Index: []"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "time_between_serial_isolate_collection_dropped "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Find distance between all pairs of times"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [],
   "source": [
    "pairwise_isolate_time = pd.DataFrame(index = patients_to_keep , columns = ['population' , 'time_between_collection'])\n",
    "\n",
    "for patient_id in patients_to_keep:\n",
    "    \n",
    "    population = time_between_serial_isolate_collection[time_between_serial_isolate_collection.patient_id == patient_id].index[0]\n",
    "    \n",
    "    #append info to dataframe\n",
    "    pairwise_isolate_time.loc[patient_id , 'population'] = population"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>population</th>\n",
       "      <th>time_between_collection</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2047T</th>\n",
       "      <td>BRYANT</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3018C</th>\n",
       "      <td>BRYANT</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3100U</th>\n",
       "      <td>BRYANT</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3103J</th>\n",
       "      <td>BRYANT</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4078X</th>\n",
       "      <td>BRYANT</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      population time_between_collection\n",
       "2047T     BRYANT                     NaN\n",
       "3018C     BRYANT                     NaN\n",
       "3100U     BRYANT                     NaN\n",
       "3103J     BRYANT                     NaN\n",
       "4078X     BRYANT                     NaN"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pairwise_isolate_time.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "for patient_id in pairwise_isolate_time.index:\n",
    "    \n",
    "    population = pairwise_isolate_time.loc[patient_id , 'population']\n",
    "    patient_id = patient_id\n",
    "    \n",
    "    if population == 'BRYANT':\n",
    "        \n",
    "        bryant_pair = time_between_serial_isolate_collection[time_between_serial_isolate_collection.patient_id == patient_id]\n",
    "        \n",
    "        first_date = bryant_pair[bryant_pair.sample_order == 1].date_information.values[0]\n",
    "        second_date = bryant_pair[bryant_pair.sample_order == 2].date_information.values[0]\n",
    "\n",
    "        num_days_between = (int(second_date) - int(first_date))*7 #dates are in # of weeks apart\n",
    "    \n",
    "    if population == 'CASALI':\n",
    "        \n",
    "        casali_pair = time_between_serial_isolate_collection[time_between_serial_isolate_collection.patient_id == patient_id]\n",
    "\n",
    "        first_date = casali_pair[casali_pair.sample_order == 1].date_information.values[0]\n",
    "        day = first_date.split('/')[0]\n",
    "        month = first_date.split('/')[1]\n",
    "        year = '20' + first_date.split('/')[2]\n",
    "        first_date = pd.to_datetime(pd.DataFrame({'year':[year] , 'month':[month] , 'day':[day]}))\n",
    "\n",
    "        second_date = casali_pair[casali_pair.sample_order == 2].date_information.values[0]\n",
    "        day = second_date.split('/')[0]\n",
    "        month = second_date.split('/')[1]\n",
    "        year = '20' + second_date.split('/')[2]\n",
    "        second_date = pd.to_datetime(pd.DataFrame({'year':[year] , 'month':[month] , 'day':[day]}))\n",
    "\n",
    "        num_days_between = (second_date - first_date).dt.days[0]\n",
    "        \n",
    "    elif (population == 'CETR') or (population == 'GUERRA') or (population == 'WITNEY') or (population == 'XU'):\n",
    "        \n",
    "        serial_pair = time_between_serial_isolate_collection[time_between_serial_isolate_collection.patient_id == patient_id]\n",
    "\n",
    "        first_date = serial_pair[serial_pair.sample_order == 1].date_information.values[0]\n",
    "        day = first_date.split('/')[1]\n",
    "        month = first_date.split('/')[0]\n",
    "        if int(first_date.split('/')[2]) in range(50,100): #some dates in 1990's others in 2000's\n",
    "            year = '19' + first_date.split('/')[2]\n",
    "        else:\n",
    "            year = '20' + first_date.split('/')[2]\n",
    "\n",
    "        first_date = pd.to_datetime(pd.DataFrame({'year':[year] , 'month':[month] , 'day':[day]}))\n",
    "\n",
    "\n",
    "        second_date = serial_pair[serial_pair.sample_order == 2].date_information.values[0]\n",
    "        day = second_date.split('/')[1]\n",
    "        month = second_date.split('/')[0]\n",
    "        if int(second_date.split('/')[2]) in range(50,100):\n",
    "            year = '19' + second_date.split('/')[2]\n",
    "        else:\n",
    "            year = '20' + second_date.split('/')[2]\n",
    "\n",
    "        second_date = pd.to_datetime(pd.DataFrame({'year':[year] , 'month':[month] , 'day':[day]}))\n",
    "\n",
    "        num_days_between = (second_date - first_date).dt.days[0]\n",
    "        \n",
    "    elif (population == 'TRAUNER') or (population == 'BRYANT'):\n",
    "        \n",
    "        serial_pair = time_between_serial_isolate_collection[time_between_serial_isolate_collection.patient_id == patient_id]\n",
    "\n",
    "        first_date = serial_pair[serial_pair.sample_order == 1].date_information.values[0]\n",
    "        second_date = serial_pair[serial_pair.sample_order == 2].date_information.values[0]\n",
    "\n",
    "        num_days_between = (int(second_date) - int(first_date))*7 #dates are in # of weeks apart\n",
    "        \n",
    "    elif population == 'WALKER':\n",
    "        \n",
    "        walker_pair = time_between_serial_isolate_collection[time_between_serial_isolate_collection.patient_id == patient_id]\n",
    "\n",
    "        first_date = walker_pair[walker_pair.sample_order == 1].date_information.values[0]\n",
    "        day = '1'\n",
    "        month = '1'\n",
    "        year = first_date\n",
    "        first_date = pd.to_datetime(pd.DataFrame({'year':[year] , 'month':[month] , 'day':[day]}))\n",
    "\n",
    "        second_date = walker_pair[walker_pair.sample_order == 2].date_information.values[0]\n",
    "        day = '1'\n",
    "        month = '1'\n",
    "        year = second_date\n",
    "        second_date = pd.to_datetime(pd.DataFrame({'year':[year] , 'month':[month] , 'day':[day]}))\n",
    "\n",
    "        num_days_between = (second_date - first_date).dt.days[0]\n",
    "        \n",
    "    #append distance between isolate collection\n",
    "    pairwise_isolate_time.loc[patient_id , 'time_between_collection'] = num_days_between"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>population</th>\n",
       "      <th>time_between_collection</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2047T</th>\n",
       "      <td>BRYANT</td>\n",
       "      <td>182</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3018C</th>\n",
       "      <td>BRYANT</td>\n",
       "      <td>336</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3100U</th>\n",
       "      <td>BRYANT</td>\n",
       "      <td>420</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3103J</th>\n",
       "      <td>BRYANT</td>\n",
       "      <td>182</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4078X</th>\n",
       "      <td>BRYANT</td>\n",
       "      <td>420</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      population time_between_collection\n",
       "2047T     BRYANT                     182\n",
       "3018C     BRYANT                     336\n",
       "3100U     BRYANT                     420\n",
       "3103J     BRYANT                     182\n",
       "4078X     BRYANT                     420"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pairwise_isolate_time.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [],
   "source": [
    "fSNP_contaminated_pairwise_isolate_times = pairwise_isolate_time.time_between_collection"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### C: filtered longitudinal isolate pairs  (*persistent*)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [],
   "source": [
    "time_between_serial_isolate_collection = pd.read_csv('/n/data1/hms/dbmi/farhat/Roger/inhost_TB_dynamics_project/CSV_files/sample_annotation_files/Longitudinal_sample_annotation_for_temporal_analysis_with_dates.csv').set_index('population')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>run_ID</th>\n",
       "      <th>sample_ID</th>\n",
       "      <th>sample_order</th>\n",
       "      <th>tag</th>\n",
       "      <th>isolate_type</th>\n",
       "      <th>patient_id</th>\n",
       "      <th>date_information</th>\n",
       "      <th>tube_label (cetr)</th>\n",
       "      <th>tag.1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>population</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>BRYANT</th>\n",
       "      <td>0</td>\n",
       "      <td>ERR108130</td>\n",
       "      <td>400044</td>\n",
       "      <td>1</td>\n",
       "      <td>ERR108130</td>\n",
       "      <td>longitudinal</td>\n",
       "      <td>2020E</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BRYANT</th>\n",
       "      <td>1</td>\n",
       "      <td>ERR108129</td>\n",
       "      <td>401206</td>\n",
       "      <td>2</td>\n",
       "      <td>ERR108129</td>\n",
       "      <td>longitudinal</td>\n",
       "      <td>2020E</td>\n",
       "      <td>36</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BRYANT</th>\n",
       "      <td>2</td>\n",
       "      <td>ERR108131</td>\n",
       "      <td>400673</td>\n",
       "      <td>1</td>\n",
       "      <td>ERR108131</td>\n",
       "      <td>longitudinal</td>\n",
       "      <td>2027W</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BRYANT</th>\n",
       "      <td>3</td>\n",
       "      <td>ERR108132</td>\n",
       "      <td>402568</td>\n",
       "      <td>2</td>\n",
       "      <td>ERR108132</td>\n",
       "      <td>longitudinal</td>\n",
       "      <td>2027W</td>\n",
       "      <td>48</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BRYANT</th>\n",
       "      <td>4</td>\n",
       "      <td>ERR108133</td>\n",
       "      <td>400711</td>\n",
       "      <td>1</td>\n",
       "      <td>ERR108133</td>\n",
       "      <td>longitudinal</td>\n",
       "      <td>2028K</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            Unnamed: 0     run_ID sample_ID  sample_order        tag  \\\n",
       "population                                                             \n",
       "BRYANT               0  ERR108130    400044             1  ERR108130   \n",
       "BRYANT               1  ERR108129    401206             2  ERR108129   \n",
       "BRYANT               2  ERR108131    400673             1  ERR108131   \n",
       "BRYANT               3  ERR108132    402568             2  ERR108132   \n",
       "BRYANT               4  ERR108133    400711             1  ERR108133   \n",
       "\n",
       "            isolate_type patient_id date_information tube_label (cetr)  tag.1  \n",
       "population                                                                     \n",
       "BRYANT      longitudinal      2020E                0               NaN    NaN  \n",
       "BRYANT      longitudinal      2020E               36               NaN    NaN  \n",
       "BRYANT      longitudinal      2027W                0               NaN    NaN  \n",
       "BRYANT      longitudinal      2027W               48               NaN    NaN  \n",
       "BRYANT      longitudinal      2028K                0               NaN    NaN  "
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "time_between_serial_isolate_collection.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(400, 10)"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.shape(time_between_serial_isolate_collection)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Retain only patients that for which we have date Information for both samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [],
   "source": [
    "#get dates for each sample in pair\n",
    "dates_for_first_sample = list( time_between_serial_isolate_collection.iloc[range(0 , np.shape(time_between_serial_isolate_collection)[0] , 2) , 7] )\n",
    "dates_for_second_sample = list( time_between_serial_isolate_collection.iloc[range(1 , np.shape(time_between_serial_isolate_collection)[0] , 2) , 7] )\n",
    "\n",
    "#get a boolean to check that there are dates for both samples\n",
    "dates_for_both_samples = [ ( (isinstance(date_1, basestring)) and (isinstance(date_2, basestring)) ) for date_1 , date_2 in zip(dates_for_first_sample , dates_for_second_sample) ]\n",
    "\n",
    "#get list of patients to include that have dates for sample retreival for both samples\n",
    "patients_to_keep = list( time_between_serial_isolate_collection[time_between_serial_isolate_collection.sample_order == 1][dates_for_both_samples].patient_id )\n",
    "\n",
    "#filter out patients with 1 or 2 missing dates\n",
    "time_between_serial_isolate_collection_filter = [patient in patients_to_keep for patient in list(time_between_serial_isolate_collection.patient_id)]\n",
    "time_between_serial_isolate_dropped_filter = [not dates_both_isolates for dates_both_isolates in time_between_serial_isolate_collection_filter]\n",
    "\n",
    "time_between_serial_isolate_collection_dropped = time_between_serial_isolate_collection[time_between_serial_isolate_dropped_filter]\n",
    "time_between_serial_isolate_collection = time_between_serial_isolate_collection[time_between_serial_isolate_collection_filter]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(396, 10)"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.shape(time_between_serial_isolate_collection)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(4, 10)"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.shape(time_between_serial_isolate_collection_dropped )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>run_ID</th>\n",
       "      <th>sample_ID</th>\n",
       "      <th>sample_order</th>\n",
       "      <th>tag</th>\n",
       "      <th>isolate_type</th>\n",
       "      <th>patient_id</th>\n",
       "      <th>date_information</th>\n",
       "      <th>tube_label (cetr)</th>\n",
       "      <th>tag.1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>population</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>WITNEY</th>\n",
       "      <td>380</td>\n",
       "      <td>ERR1750925</td>\n",
       "      <td>SAMEA23779168</td>\n",
       "      <td>1</td>\n",
       "      <td>ERR1750925</td>\n",
       "      <td>longitudinal</td>\n",
       "      <td>27</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WITNEY</th>\n",
       "      <td>381</td>\n",
       "      <td>ERR1750926</td>\n",
       "      <td>SAMEA23779918</td>\n",
       "      <td>2</td>\n",
       "      <td>ERR1750926</td>\n",
       "      <td>longitudinal</td>\n",
       "      <td>27</td>\n",
       "      <td>1/25/11</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WITNEY</th>\n",
       "      <td>382</td>\n",
       "      <td>ERR1750927</td>\n",
       "      <td>SAMEA23780668</td>\n",
       "      <td>1</td>\n",
       "      <td>ERR1750927</td>\n",
       "      <td>longitudinal</td>\n",
       "      <td>28</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WITNEY</th>\n",
       "      <td>383</td>\n",
       "      <td>ERR1750928</td>\n",
       "      <td>SAMEA23781418</td>\n",
       "      <td>2</td>\n",
       "      <td>ERR1750928</td>\n",
       "      <td>longitudinal</td>\n",
       "      <td>28</td>\n",
       "      <td>8/3/11</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            Unnamed: 0      run_ID      sample_ID  sample_order         tag  \\\n",
       "population                                                                    \n",
       "WITNEY             380  ERR1750925  SAMEA23779168             1  ERR1750925   \n",
       "WITNEY             381  ERR1750926  SAMEA23779918             2  ERR1750926   \n",
       "WITNEY             382  ERR1750927  SAMEA23780668             1  ERR1750927   \n",
       "WITNEY             383  ERR1750928  SAMEA23781418             2  ERR1750928   \n",
       "\n",
       "            isolate_type patient_id date_information tube_label (cetr)  tag.1  \n",
       "population                                                                     \n",
       "WITNEY      longitudinal         27              NaN               NaN    NaN  \n",
       "WITNEY      longitudinal         27          1/25/11               NaN    NaN  \n",
       "WITNEY      longitudinal         28              NaN               NaN    NaN  \n",
       "WITNEY      longitudinal         28           8/3/11               NaN    NaN  "
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "time_between_serial_isolate_collection_dropped "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Retain only patients that have different dates of collection for both samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [],
   "source": [
    "#get dates for each sample in pair\n",
    "dates_for_first_sample = list( time_between_serial_isolate_collection.iloc[range(0 , np.shape(time_between_serial_isolate_collection)[0] , 2) , 7] )\n",
    "dates_for_second_sample = list( time_between_serial_isolate_collection.iloc[range(1 , np.shape(time_between_serial_isolate_collection)[0] , 2) , 7] )\n",
    "\n",
    "#get a boolean to check that there are dates for both samples\n",
    "dates_different_for_both_samples = [ ( date_1 != date_2 ) for date_1 , date_2 in zip(dates_for_first_sample , dates_for_second_sample) ]\n",
    "\n",
    "#get list of patients to include that have dates for sample retreival for both samples\n",
    "patients_to_keep = list( time_between_serial_isolate_collection[time_between_serial_isolate_collection.sample_order == 1][dates_different_for_both_samples].patient_id )\n",
    "\n",
    "#filter out patients with 1 or 2 missing dates\n",
    "time_between_serial_isolate_collection_filter = [patient in patients_to_keep for patient in list(time_between_serial_isolate_collection.patient_id)]\n",
    "time_between_serial_isolate_dropped_filter = [not dates_both_isolates for dates_both_isolates in time_between_serial_isolate_collection_filter]\n",
    "\n",
    "time_between_serial_isolate_collection_dropped = time_between_serial_isolate_collection[time_between_serial_isolate_dropped_filter]\n",
    "time_between_serial_isolate_collection = time_between_serial_isolate_collection[time_between_serial_isolate_collection_filter]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(390, 10)"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.shape(time_between_serial_isolate_collection)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(6, 10)"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.shape(time_between_serial_isolate_collection_dropped )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>run_ID</th>\n",
       "      <th>sample_ID</th>\n",
       "      <th>sample_order</th>\n",
       "      <th>tag</th>\n",
       "      <th>isolate_type</th>\n",
       "      <th>patient_id</th>\n",
       "      <th>date_information</th>\n",
       "      <th>tube_label (cetr)</th>\n",
       "      <th>tag.1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>population</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>WALKER</th>\n",
       "      <td>310</td>\n",
       "      <td>ERR046918</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>ERR046918</td>\n",
       "      <td>longitudinal</td>\n",
       "      <td>P000059</td>\n",
       "      <td>2007</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WALKER</th>\n",
       "      <td>311</td>\n",
       "      <td>ERR046933</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>ERR046933</td>\n",
       "      <td>longitudinal</td>\n",
       "      <td>P000059</td>\n",
       "      <td>2007</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WALKER</th>\n",
       "      <td>342</td>\n",
       "      <td>ERR047010</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>ERR047010</td>\n",
       "      <td>longitudinal</td>\n",
       "      <td>P000259</td>\n",
       "      <td>2008</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WALKER</th>\n",
       "      <td>343</td>\n",
       "      <td>ERR047012</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>ERR047012</td>\n",
       "      <td>longitudinal</td>\n",
       "      <td>P000259</td>\n",
       "      <td>2008</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WALKER</th>\n",
       "      <td>346</td>\n",
       "      <td>ERR047007</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>ERR047007</td>\n",
       "      <td>longitudinal</td>\n",
       "      <td>P000267</td>\n",
       "      <td>2007</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WALKER</th>\n",
       "      <td>347</td>\n",
       "      <td>ERR072030</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>ERR072030</td>\n",
       "      <td>longitudinal</td>\n",
       "      <td>P000267</td>\n",
       "      <td>2007</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            Unnamed: 0     run_ID sample_ID  sample_order        tag  \\\n",
       "population                                                             \n",
       "WALKER             310  ERR046918       NaN             1  ERR046918   \n",
       "WALKER             311  ERR046933       NaN             2  ERR046933   \n",
       "WALKER             342  ERR047010       NaN             1  ERR047010   \n",
       "WALKER             343  ERR047012       NaN             2  ERR047012   \n",
       "WALKER             346  ERR047007       NaN             1  ERR047007   \n",
       "WALKER             347  ERR072030       NaN             2  ERR072030   \n",
       "\n",
       "            isolate_type patient_id date_information tube_label (cetr)  tag.1  \n",
       "population                                                                     \n",
       "WALKER      longitudinal    P000059             2007               NaN    NaN  \n",
       "WALKER      longitudinal    P000059             2007               NaN    NaN  \n",
       "WALKER      longitudinal    P000259             2008               NaN    NaN  \n",
       "WALKER      longitudinal    P000259             2008               NaN    NaN  \n",
       "WALKER      longitudinal    P000267             2007               NaN    NaN  \n",
       "WALKER      longitudinal    P000267             2007               NaN    NaN  "
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "time_between_serial_isolate_collection_dropped "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Find distance between all pairs of times"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [],
   "source": [
    "pairwise_isolate_time = pd.DataFrame(index = patients_to_keep , columns = ['population' , 'time_between_collection'])\n",
    "\n",
    "for patient_id in patients_to_keep:\n",
    "    \n",
    "    population = time_between_serial_isolate_collection[time_between_serial_isolate_collection.patient_id == patient_id].index[0]\n",
    "    \n",
    "    #append info to dataframe\n",
    "    pairwise_isolate_time.loc[patient_id , 'population'] = population"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>population</th>\n",
       "      <th>time_between_collection</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2020E</th>\n",
       "      <td>BRYANT</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2027W</th>\n",
       "      <td>BRYANT</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2028K</th>\n",
       "      <td>BRYANT</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2041Q</th>\n",
       "      <td>BRYANT</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2043R</th>\n",
       "      <td>BRYANT</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      population time_between_collection\n",
       "2020E     BRYANT                     NaN\n",
       "2027W     BRYANT                     NaN\n",
       "2028K     BRYANT                     NaN\n",
       "2041Q     BRYANT                     NaN\n",
       "2043R     BRYANT                     NaN"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pairwise_isolate_time.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [],
   "source": [
    "for patient_id in pairwise_isolate_time.index:\n",
    "    \n",
    "    population = pairwise_isolate_time.loc[patient_id , 'population']\n",
    "    patient_id = patient_id\n",
    "    \n",
    "    if population == 'BRYANT':\n",
    "        \n",
    "        bryant_pair = time_between_serial_isolate_collection[time_between_serial_isolate_collection.patient_id == patient_id]\n",
    "        \n",
    "        first_date = bryant_pair[bryant_pair.sample_order == 1].date_information.values[0]\n",
    "        second_date = bryant_pair[bryant_pair.sample_order == 2].date_information.values[0]\n",
    "\n",
    "        num_days_between = (int(second_date) - int(first_date))*7 #dates are in # of weeks apart\n",
    "    \n",
    "    if population == 'CASALI':\n",
    "        \n",
    "        casali_pair = time_between_serial_isolate_collection[time_between_serial_isolate_collection.patient_id == patient_id]\n",
    "\n",
    "        first_date = casali_pair[casali_pair.sample_order == 1].date_information.values[0]\n",
    "        day = first_date.split('/')[0]\n",
    "        month = first_date.split('/')[1]\n",
    "        year = '20' + first_date.split('/')[2]\n",
    "        first_date = pd.to_datetime(pd.DataFrame({'year':[year] , 'month':[month] , 'day':[day]}))\n",
    "\n",
    "        second_date = casali_pair[casali_pair.sample_order == 2].date_information.values[0]\n",
    "        day = second_date.split('/')[0]\n",
    "        month = second_date.split('/')[1]\n",
    "        year = '20' + second_date.split('/')[2]\n",
    "        second_date = pd.to_datetime(pd.DataFrame({'year':[year] , 'month':[month] , 'day':[day]}))\n",
    "\n",
    "        num_days_between = (second_date - first_date).dt.days[0]\n",
    "        \n",
    "    elif (population == 'CETR') or (population == 'GUERRA') or (population == 'WITNEY') or (population == 'XU'):\n",
    "        \n",
    "        serial_pair = time_between_serial_isolate_collection[time_between_serial_isolate_collection.patient_id == patient_id]\n",
    "\n",
    "        first_date = serial_pair[serial_pair.sample_order == 1].date_information.values[0]\n",
    "        day = first_date.split('/')[1]\n",
    "        month = first_date.split('/')[0]\n",
    "        if int(first_date.split('/')[2]) in range(50,100): #some dates in 1990's others in 2000's\n",
    "            year = '19' + first_date.split('/')[2]\n",
    "        else:\n",
    "            year = '20' + first_date.split('/')[2]\n",
    "\n",
    "        first_date = pd.to_datetime(pd.DataFrame({'year':[year] , 'month':[month] , 'day':[day]}))\n",
    "\n",
    "\n",
    "        second_date = serial_pair[serial_pair.sample_order == 2].date_information.values[0]\n",
    "        day = second_date.split('/')[1]\n",
    "        month = second_date.split('/')[0]\n",
    "        if int(second_date.split('/')[2]) in range(50,100):\n",
    "            year = '19' + second_date.split('/')[2]\n",
    "        else:\n",
    "            year = '20' + second_date.split('/')[2]\n",
    "\n",
    "        second_date = pd.to_datetime(pd.DataFrame({'year':[year] , 'month':[month] , 'day':[day]}))\n",
    "\n",
    "        num_days_between = (second_date - first_date).dt.days[0]\n",
    "        \n",
    "    elif (population == 'TRAUNER') or (population == 'BRYANT'):\n",
    "        \n",
    "        serial_pair = time_between_serial_isolate_collection[time_between_serial_isolate_collection.patient_id == patient_id]\n",
    "\n",
    "        first_date = serial_pair[serial_pair.sample_order == 1].date_information.values[0]\n",
    "        second_date = serial_pair[serial_pair.sample_order == 2].date_information.values[0]\n",
    "\n",
    "        num_days_between = (int(second_date) - int(first_date))*7 #dates are in # of weeks apart\n",
    "        \n",
    "    elif population == 'WALKER':\n",
    "        \n",
    "        walker_pair = time_between_serial_isolate_collection[time_between_serial_isolate_collection.patient_id == patient_id]\n",
    "\n",
    "        first_date = walker_pair[walker_pair.sample_order == 1].date_information.values[0]\n",
    "        day = '1'\n",
    "        month = '1'\n",
    "        year = first_date\n",
    "        first_date = pd.to_datetime(pd.DataFrame({'year':[year] , 'month':[month] , 'day':[day]}))\n",
    "\n",
    "        second_date = walker_pair[walker_pair.sample_order == 2].date_information.values[0]\n",
    "        day = '1'\n",
    "        month = '1'\n",
    "        year = second_date\n",
    "        second_date = pd.to_datetime(pd.DataFrame({'year':[year] , 'month':[month] , 'day':[day]}))\n",
    "\n",
    "        num_days_between = (second_date - first_date).dt.days[0]\n",
    "        \n",
    "    #append distance between isolate collection\n",
    "    pairwise_isolate_time.loc[patient_id , 'time_between_collection'] = num_days_between"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>population</th>\n",
       "      <th>time_between_collection</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2020E</th>\n",
       "      <td>BRYANT</td>\n",
       "      <td>252</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2027W</th>\n",
       "      <td>BRYANT</td>\n",
       "      <td>336</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2028K</th>\n",
       "      <td>BRYANT</td>\n",
       "      <td>182</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2041Q</th>\n",
       "      <td>BRYANT</td>\n",
       "      <td>182</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2043R</th>\n",
       "      <td>BRYANT</td>\n",
       "      <td>252</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      population time_between_collection\n",
       "2020E     BRYANT                     252\n",
       "2027W     BRYANT                     336\n",
       "2028K     BRYANT                     182\n",
       "2041Q     BRYANT                     182\n",
       "2043R     BRYANT                     252"
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pairwise_isolate_time.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [],
   "source": [
    "filtered_persistent_pairwise_isolate_times = pairwise_isolate_time.time_between_collection"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "######################################################################################################################################################################################################################################################################"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Compare F2 (mixed) samples and fSNP (re-infection) samples with filtered (persistent) samples"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "######################################################################################################################################################################################################################################################################"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Using 1.7 months as the *duration* threshold for treatment treshold"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [],
   "source": [
    "F2_contaminated_pairwise_isolate_times_months = F2_contaminated_pairwise_isolate_times/30.0\n",
    "fSNP_contaminated_pairwise_isolate_times_months = fSNP_contaminated_pairwise_isolate_times/30.0\n",
    "filtered_persistent_pairwise_isolate_times_months = filtered_persistent_pairwise_isolate_times/30.0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(30,)"
      ]
     },
     "execution_count": 80,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.shape(F2_contaminated_pairwise_isolate_times_months[F2_contaminated_pairwise_isolate_times_months >= 1.7])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(39,)"
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.shape(fSNP_contaminated_pairwise_isolate_times_months[fSNP_contaminated_pairwise_isolate_times_months >= 1.7])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(178,)"
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.shape(filtered_persistent_pairwise_isolate_times_months[filtered_persistent_pairwise_isolate_times_months >= 1.7])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Mixed (F2 filter) isolate pairs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "647.1935483870968"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.mean( F2_contaminated_pairwise_isolate_times )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "336.0"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.median( F2_contaminated_pairwise_isolate_times )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3"
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(F2_contaminated_pairwise_isolate_times[ F2_contaminated_pairwise_isolate_times < 60 ])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "28"
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(F2_contaminated_pairwise_isolate_times[ F2_contaminated_pairwise_isolate_times >= 60 ])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Re-infection (fSNP filter) isolate pairs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1019.1395348837209"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.mean( fSNP_contaminated_pairwise_isolate_times )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "538"
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.median( fSNP_contaminated_pairwise_isolate_times )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "8"
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(fSNP_contaminated_pairwise_isolate_times[ fSNP_contaminated_pairwise_isolate_times < 60 ])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "35"
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(fSNP_contaminated_pairwise_isolate_times[ fSNP_contaminated_pairwise_isolate_times >= 60 ])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Persistent (passed all filters) isolate pairs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "427.20512820512823"
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.mean( filtered_persistent_pairwise_isolate_times )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "281"
      ]
     },
     "execution_count": 79,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.median( filtered_persistent_pairwise_isolate_times )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "28"
      ]
     },
     "execution_count": 80,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(filtered_persistent_pairwise_isolate_times[ filtered_persistent_pairwise_isolate_times < 60 ])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "167"
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(filtered_persistent_pairwise_isolate_times[ filtered_persistent_pairwise_isolate_times >= 60 ])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Total Number of Isolate Pairs & Number of Isolate Pairs with collections times $\\ge$ 2 months/60 days"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "269"
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "3 + 28 + 8 + 35 + 28 + 167 #total number of pairs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "230"
      ]
     },
     "execution_count": 83,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "28 + 35 + 167 #total number of isolate pairs with collection times >= 2 months"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Plot *time between isolate collection* for *mixed* group and *re-infection* group"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.style.use('ggplot')\n",
    "plt.rcParams['lines.linewidth']=0\n",
    "plt.rcParams['axes.facecolor']='1.0'\n",
    "plt.rcParams['xtick.color']='black'\n",
    "plt.rcParams['axes.grid']=False\n",
    "plt.rcParams['axes.edgecolor']='black'\n",
    "plt.rcParams['grid.color']= '1.0'\n",
    "plt.rcParams.update({'font.size': 14})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pylab import plot, show, savefig, xlim, figure, hold, ylim, legend, boxplot, setp, axes\n",
    "\n",
    "linewidth_for_box = 3.0\n",
    "\n",
    "# function for setting the colors of the box plots pairs\n",
    "def setBoxColors(bp , color):\n",
    "    setp(bp['boxes'][0], color=color , linewidth = linewidth_for_box)\n",
    "    setp(bp['caps'][0], color=color , linewidth = linewidth_for_box)\n",
    "    setp(bp['caps'][1], color=color , linewidth = linewidth_for_box)\n",
    "    setp(bp['whiskers'][0], color=color , linewidth = linewidth_for_box)\n",
    "    setp(bp['whiskers'][1], color=color , linewidth = linewidth_for_box)\n",
    "    setp(bp['fliers'][0], color=color , alpha = 0.65)\n",
    "    setp(bp['medians'][0], color=color , linewidth = 3.0)\n",
    "    bp['fliers'][0].set_markerfacecolor(color)\n",
    "    bp['fliers'][0].set_markeredgewidth(0.5)\n",
    "    bp['fliers'][0].set_markersize(6.0)\n",
    "    bp['fliers'][0].set_markeredgecolor('white')\n",
    "    \n",
    "def values_to_variant_types(values):\n",
    "    SNP_vals = values.alt_AF_diff\n",
    "    return [SNP_vals]\n",
    "\n",
    "def color_bar(bar , color):\n",
    "    bar.set_color(color)\n",
    "    bar.set_edgecolor('k')\n",
    "    bar.set_linewidth(1.0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/rv76/anaconda2/lib/python2.7/site-packages/matplotlib/cbook/deprecation.py:107: MatplotlibDeprecationWarning: Passing one of 'on', 'true', 'off', 'false' as a boolean is deprecated; use an actual boolean (True/False) instead.\n",
      "  warnings.warn(message, mplDeprecation, stacklevel=1)\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAATsAAAGgCAYAAADVfcBHAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMi4zLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvIxREBQAAIABJREFUeJzs3XtcVHX+P/DXjCMexuGmiNymAS1FvKGoSCpSaKWlljcktPbrbRPLTLc0zVX7lXdW01UrRV03MbxmdtvKXW+sIpYCAl5KGWDkMii3gTkOl8/vD5aTIwJHmYEZ5v18PHhscz7nDG9m2lfnc87nfD4SxhgDIYS0ctKWLoAQQpoDhR0hxCZQ2BFCbAKFHSHEJlDYEUJsAoUdIcQmUNgRQmwChR0hxCZQ2BFCbAKFHSHEJlDYEUJsAoUdIcQmUNiJIJFIWroEQkgTUdgRQmyCrKULIISYBmMMGRkZyM/Ph5ubG3x8fKhXch8KO0JaAcYYvvvuO8TFxaGyshIymQzh4eEYPXo0Bd7/UDeWkFYgIyNDCDoAqKysRFxcHDIyMlq2MAtCYUdIK5Cfny8EXa3KykpotdoWqsjyUNgR0gq4ublBJjO+KiWTydCpU6cWqsjyUNgR0gr4+PggPDxcCLzaa3Y+Pj4tW5gFkdCCO42TSCSgj4lYutq7sVqtFp06daK7sQ+gsBOBwo4Q60fdWEKITaCwI4TYBAo7QohNoLAjhNgECjtCiE2gsCOE2AQKO0KITaCwI4TYBIua4mnFihW4ceMGpNKaDO7QoQM++eQTpKam4sMPP4SdnZ2w74wZMxAaGgoA0Ol02L59O5KTk+Hg4IBXX30VQ4cOFfY9e/YsYmNjUVpait69eyMqKgoKhaJZ/zZCSMuyqLADgOnTpyMsLKzOdhcXF3z66acPPWbnzp2QyWTYsWMHMjIysHr1aqhUKiiVSmRlZeHzzz/H4sWL0aVLF3z22WfYuXMn5s+fb+4/hRBiQay+G8vzPBISEhAeHg6O4+Dn54cBAwbg9OnTAIAzZ84gMDAQ/v7+4DgO4eHhSEhIgF6vb+HKCSHNyeLO7GJjYxEbGwtPT09MmTIFPXv2BAAUFxdj1qxZsLOzw8CBAzFlyhRwHIecnBxIpVJ4enoK76FSqZCWlgYAyM7ORrdu3YQ2d3d3yGQy5OTkoEuXLs37xxFCWoxFhV1kZCS8vb0hk8kQHx+PtWvXYt26dfDy8sL69evh6emJgoICbN26FXv37sXs2bPB8zzkcrnR+8jlcvA8DwD1ttOZHSG2xaK6sU899RTs7e3Rtm1bhIaGonv37rh06RKcnZ3h7e0NqVQKNzc3REZG4vz58wAAjuPqBJderwfHcQ2229vbN88fRQixCBYVdg+qb2ql+7d7eHigqqoKOTk5QrtarYZSqQQAeHt7Q61WC215eXmoqKiAh4eHmasnhFgSiwm7srIyXL58GQaDAVVVVThz5gzS09MREBCA1NRUFBQUgDGGgoICxMbGYuDAgQBqztyCgoIQFxcHnudx9epVJCYmIiQkBAAwbNgw/PLLL0hPTwfP84iLi0NQUBCd2RFiYyxm8s6SkhKsXr0aGo0GUqkUXl5eCA8PR58+ffDNN9/g+PHjKCsrg4ODAwYOHIiIiAghsHQ6HbZt24aUlBQoFApERkbWGWe3b98+6HS6xxpnR5N3EmL9LCbsLBmFHSHWz2K6sYQQYk4UdoQQm0BhRwixCRR2hBCbQGFHCLEJFHaEEJtAYUcIsQkUdoQQm0BhRwixCRR2hBCbQGFHCLEJFHaEEJtAYUcIsQmipmUvLCzEp59+irS0NJSVlRm1HTlyxCyFEUKIKYkKuzfffBP37t3DmDFjaNJLQohVEjWfXffu3ZGcnIx27do1R00Wh+azI8T6ibpm16NHD6M1HgghxNqI6sYOGTIEkZGRCA8Ph5ubm1HblClTzFIYIYSYkqhu7MSJEx9+sESCgwcPmrwoS0PdWEKsH61BIQKFHSHWT1Q3FgCKiorw008/ITc3F+7u7hg5ciScnZ1NWsyKFStw48YNSKU1lxI7dOiATz75BEDNCmGxsbEoLS2ts0KYTqfD9u3bkZycDAcHB7z66qt1Vher71hCWgvGGDIyMpCfnw83Nzf4+PhAIpG0dFkWQ9SZ3cWLF/H666+ja9eu8Pb2hkajwW+//YZ//OMfGDBggMmKWbFiBYYNG4awsDCj7VlZWVi6dCkWL16MLl264LPPPgNjDPPnzwcAbNq0CYwxzJkzBxkZGVi9ejU++ugjKJXKRo8Vg87siKVjjOG7775DXFwcKisrIZPJEB4ejtGjR1Pg/Y+ou7ErVqzAqlWr8PXXX2Pbtm04duwYVq1ahb/+9a/mrg8AcObMGQQGBsLf3x8cxyE8PBwJCQnQ6/XgeR4JCQkIDw8Hx3Hw8/PDgAEDcPr06UaPJaS1yMjIEIIOACorKxEXF4eMjIyWLcyCiAq7mzdvYsyYMUbbXnrpJbN8kLGxsZgxYwaWLVuG1NRUAEB2djZUKpWwj7u7O2QyGXJycpCTkwOpVApPT0+hXaVSISsrq9FjCWkt8vPzhaCrVVlZCa1W20IVWR5R1+x8fX1x7NgxvPLKK8K2b775xihETCEyMhLe3t6QyWSIj4/H2rVrsW7dOvA8D7lcbrSvXC6HXq+HVCp9aBvP8wDQ4LGEtBZubm6QyWRGgSeTydCpU6cWrMqyiAq7FStW4E9/+hNiYmLg7e2N7Oxs3Lp1C3v27DFpMU899ZTwz6GhoYiPj8elS5fAcVydcNLr9bC3t4dEInloG8dxANDgsYS0Fj4+PggPD69zzc7Hx6elS7MYosJu4MCBiI+Px4kTJ5CXl4eRI0fi2WefhYuLi1mLq70x4O3tDbVaLWzPy8tDRUUFPDw8IJFIUFVVhZycHHh4eAAA1Go1lEolADR4LCGthUQiwejRo+Hv7w+tVotOnTrR3dgHiJ7iydnZGRMmTEBUVBQmTJhg8qArKyvD5cuXYTAYUFVVhTNnziA9PR0BAQEYNmwYfvnlF6Snp4PnecTFxSEoKAj29vbgOA5BQUGIi4sDz/O4evUqEhMTERISAgANHktIayKRSODr64tBgwbB19eXgu4B9Q49iYyMxL59+wAAr7zySr0fnKmmeCopKcHq1auh0WgglUrh5eWF8PBw9OnTB0DNWLl9+/ZBp9M9dJzdtm3bkJKSAoVCgcjIyDrj7Oo7VgwaekKI9as37I4ePSrckDhw4EC9bzB58mTzVGZBKOwIsX6iBhX/+uuv6N+/f53tly5dQr9+/cxSmCWhsCPE+omez+7atWt1tvfs2VMYC9eaUdgRYv0avBtbXV0NxpjRT62MjAy0adPG7AUSQogpNBh2TzzxhHBj4oknnjBqk0qlmDdvnvkqI4QQE2qwG5udnQ3GGCZMmGB011UikaBDhw42M3yDurGEWD9R1+x4nkebNm3Qtm1bYVtFRQWqq6ttYl0KCjtCrJ+oQcWvvvoqkpOTjbYlJycjMjLSLEURQoipiTqz69GjB9LS0owGFldXV6NXr15IS0sza4GWgM7sCLF+os7sHB0d60wVo9VqbeaaHSHE+okKu9GjR2Pu3Lm4evUq9Ho90tPT8fbbb9eZ444QQiyV6BsUH374IeLi4nDv3j1hxt9ly5YJUym1ZtSNJcT6PdLqYowx3L17Fx06dLCpGRUo7AixfqJXF7tx4wa++eYbFBQU4OOPP8Zvv/0Gg8EAf39/c9ZHCCEmIeqa3fHjxzF+/Hjk5ubi0KFDAGrmn1u5cqVZiyOEEFMR1Y0dPnw4tm7dil69esHf3x9paWmoqKhA//79kZKS0hx1tijqxhJi/USd2RUUFKBnz54AIFyrk0gkNnXdjhBi3USFXZ8+fYTua61jx44hICDALEURQoipierG/vbbb4iIiMATTzyBX3/9FcHBwbh58yZiY2PRpUuX5qizRVE3lhDrJ3roiV6vx08//QSNRgNPT0+MGDEC7du3N3d9FoHCjhDr90jj7GwVhR0h1q/ecXYNrSh2P1OtLlYrJycHf/nLXxAUFIR58+YhNTUVH374Iezs7IR9ZsyYgdDQUAA1K4tt374dycnJcHBwwKuvvlpnZbHY2FiUlpY+1spihJDWod6wi4iIaM46BDExMejatavRNhcXF3z66acP3X/nzp2QyWTYsWMHMjIysHr1aqhUKiiVSmRlZeHzzz/H4sWL0aVLF3z22WfYuXMn5s+f3xx/CiHEgtQbdi2xRGJ8fDzkcjm6deuG3NzcRvfneR4JCQmIjo4Gx3Hw8/PDgAEDcPr0aURGRuLMmTMIDAwUnvIIDw/HO++8A71eTzO2EGJj6g27L7/8UtQbTJkyxSSFlJeX48CBA1i2bBn+/e9/G7UVFxdj1qxZsLOzw8CBAzFlyhRwHIecnBxIpVJ4enoK+6pUKmGOvezsbHTr1k1oc3d3h0wmQ05Ojk3cRSaE/KHesHtwXN3DSCQSk4VdXFwcnnnmGbi6uhpt9/Lywvr16+Hp6YmCggJs3boVe/fuxezZs8HzPORyudH+crkcPM8DQL3ter3eJDUTQqxHk8LOVDIyMpCSkoJ169bVaXN2doazszMAwM3NDZGRkVizZg1mz54NjuPqBJderxemnaqvnbqwhNge0bOeFBUV4aeffkJubi7c3d0xcuRIIYSaKjU1FVqtFnPmzAFQc0ZWXV2NRYsWYe3atUb73j8MxMPDA1VVVcjJyYGHhwcAQK1WQ6lUAgC8vb2hVquFY/Py8lBRUSHsSwixHaIeF7t48SKGDBmCf/7zn0hPT8cXX3yBIUOG4OLFiyYpYsSIEdiyZQvWr1+P9evXY+TIkejfvz+WLl2K1NRUFBQUgDGGgoICxMbGYuDAgQBqztyCgoIQFxcHnudx9epVJCYmIiQkBAAwbNgw/PLLL0hPTwfP84iLi0NQUBCd2RFig0QNKn7ppZcwa9YsjBs3Tth27NgxfPbZZ/juu+9MXtSBAweQm5uLefPm4ZtvvsHx48dRVlYGBwcHDBw4EBEREUJg6XQ6bNu2DSkpKVAoFIiMjKwzzm7fvn3Q6XSPPc6OBhUTYv1EhZ2/vz+uXLkCqfSPE8Gqqir07t2bVhcjhFgFUd1YX19fHDt2zGjbN998A5VKZZaiCCHE1ESd2SUmJuJPf/oTfH194e3tjezsbNy6dQt79uwRrp+1ZnRmR4j1Ez0RQFFREU6cOIG8vDx07twZzz77LFxcXMxdn0WgsCPE+jUYdnq9Hmq1Gn5+fnXarl69Ch8fH1pKkRBiFRq8Zrd9+3bs37//oW1xcXHYvn27WYoihBBTa/DMLjQ0FPv373/oINycnBxERETg5MmT5qzPItCZHSHWr8Ezu9zc3HqfNvDw8BA1MwkhhFiCBsNOLpdDo9E8tE2j0dCTCIQQq9Fg2D377LNYs2bNQ9vWrVuHsLAwsxRFCCGm1uA1u/z8fIwdOxYKhQKjR4+Gm5sb8vPz8f3330On0+HYsWNwc3NrznpbBF2zI8T6NTrOrqioCJ999hnOnj2LwsJCuLi4YNiwYZg9e7bJZj2xdBR2hFg/Wl1MBAo7QqyfqGdjCSHE2lHYEUJsAoUdIcQmUNgRQmyCqLC7d+8e1qxZg+DgYGFSgFOnTmH37t1mLY4QQkxFVNitWLEC165dw5YtWyCRSAAA3bp1w969e81aHCGEmIqo1cV++OEHxMfHQy6XC1Oz07OxrQNjQEYGkJ8PuLkBPj7A//57RkirIirs2rZti8rKSqNtd+7csZnJO1srxoDvvgPi4oDKSkAmA8LDgdGjKfBI6yOqG/vSSy9h/vz5yMzMBFCz/urSpUsxduxYsxZHzCsj44+gA2r+Ny6uZjshrY2osFu8eDG8vb0RFhaG4uJiDB06FJ07d8Y777xj8oJycnIQGRmJzZs3C9vOnj2LqKgoTJs2DevWrYNOpxPadDod1q9fj2nTpiEqKgpnz541er+GjrV1+fl/BF2tykpAq22ZeggxJ1FhZ2dnhw8//BA3btxAUlISrl+/jpUrV6Jdu3YmLygmJgZdu3YVXmdlZeHzzz/Hm2++iR07dqBdu3bYuXOn0L5z507IZDLs2LED8+bNw44dO5CVlSXqWFvn5lbTdb2fTAZ06tQy9RBiTqLCrmfPnsI/d+zYUbgj26dPH5MWU3sTpFevXsK2M2fOIDAwEP7+/uA4DuHh4UhISIBerwfP80hISEB4eDg4joOfnx8GDBiA06dPN3osqbkZER7+R+DVXrPz8WnJqggxD1E3KCoqKh66raqqymSFlJeX48CBA1i2bBn+/e9/C9uzs7PRrVs34bW7uztkMhlycnIgkUgglUrh6ekptKtUKmHh7oaO7dKli8lqt1YSSc3NCH//mq5rp050N5a0Xg2G3SuvvAKJRIJ79+5h/PjxRm05OTkYMGCAyQqJi4vDM888A1dXV6PtPM9DLpcbbZPL5dDr9ZBKpQ9t43m+0WNJDYkE8PWt+SGkNWsw7CIiIsAYQ1JSEqZMmSJsl0gk6NSpE4YMGWKSIjIyMpCSkoJ169bVaeM4rk446fV62NvbQyKRPLStdnnHho4lhNiWBsNu8uTJAIDAwEA8+eSTZisiNTUVWq0Wc+bMAVBzRlZdXY1Fixahb9++UKvVwr55eXmoqKiAh4cHJBIJqqqqkJOTIywMpFaroVQqAQDe3t71HksIsS2irtk9+eST0Gq1uHTpEgoLC40msrz/jO9xjRgxwugs8euvv4ZWq8WsWbNQXFyMDz74AOnp6fD19UVcXByCgoKEs7OgoCDExcXhjTfeQEZGBhITE/HRRx8BAIYNG9bgsYQQ2yFqpuIffvgBb731Fnx9fXH9+nV069YN165dw8CBA3Ho0CGTF3XgwAHk5uZi3rx5AGrGyu3btw86nQ69e/dGVFQUFAoFgJpxdtu2bUNKSgoUCgUiIyMxdOhQ4b0aOlYsmqmYEOsnKuyeffZZvPPOOxgzZgz8/f2RlpaGuLg4XLt2DX/961+bo84WRWFHiPUTFXbdu3fHtWvXAEAIu+rqagQEBCA5OdnsRbY0CjtCrJ+oQcWurq7Q/u8ZIm9vb1y8eBEZGRkmHWdHCCHmJCrsXn31VVy4cAEAMGvWLEyePBkjR47E66+/btbiCCHEVB5rKUWNRoPy8nI89dRT5qjJ4lA3lhDrR+vGikBhR4j1q3ec3YABA4QH/huSmJho0oIIIcQc6g27LVu2NGcdhBBiVtSNFYG6sYRYP1F3YysqKrBhwwYEBwejS5cuCA4OxoYNG2AwGMxdHyGEmISoZ2M/+ugjXL58GWvWrIGXlxc0Gg02bdqE0tJSrFy50tw1EkJIk4nqxgYGBuKnn35Chw4dhG13797FyJEj8csvv5i1QEtA3VhCrJ+oM7v6UAAQYjkYY8jIyEB+fj7c3Nzg4+MjakSFrRAVdi+++CL+9Kc/YcGCBfDy8kJ2djY++eQTvPTSS+aujxAiAmMM3333HeLi4lBZWQmZTIbw8HCMHj2aAu9/RHVjDQYDPvnkE3z11VfIzc2Fu7s7xo0bh7ffftssK4xZGurGEkt369YtLF++3Ggxe5lMhpUrV8KX5twHIPLMzs7ODu+++y7effddc9dDCHkM+fn5RkEHAJWVldBqtRR2/yNq6El8fDwyMzMB1Hyob7/9NhYsWID8/HyzFkcIEcfNzQ2yBxYBlslk6ESLAAtEhd2SJUvQpk0bAMDKlStRWVkJiUSC9957z6zFEULE8fHxQXh4uBB4tdfsfGgRYMEjTd5ZWVmJvn374sKFC2jbti369++PK1euNEedLYqu2RFrUHs3VqvVolOnTnQ39gGirtkpFApotVpcvXoV3bp1Q/v27WEwGOpcIyCEtByJRAJfX1+6RlcPUWE3ffp0jB49GhUVFcITE4mJiWZdXpEQQkxJ9EQAv//+O9q0aSNcA/j9999hMBjQo0cPkxWzefNmXLlyBffu3YOzszPGjh2LsLAw5Ofn48033zQa5jJu3DhMnDgRQM2zuzt27EBCQgLs7Owwbtw4ozGAKSkpiImJQUFBAZ566ilERUU90oVb6sYS0gowC5KZmckMBgNjjLHs7Gw2c+ZM9vvvv7O8vDw2adIkVllZ+dDj9u3bx5YtW8ZKS0tZVlYWmzlzJrt06RJjjLHi4mL22muvsf/+97/s3r17bO/evWzJkiWPVJeFfUyEkMfQYDf2lVdeafQC55EjR0wWvEqlUvhniUQCiUSC3NzcRrvLp06dEtaDVSgUCAsLw8mTJxEQEIALFy5AqVQiODgYADBp0iTMmDEDGo0GXl5eJqudEGLZGgy7iIiI5qpDsHPnTpw8eRIGgwG+vr7o378/SkpKAABRUVGQSCTo06cPpk6dCkdHR+h0OhQWFkKlUgnv4ePjI8ygnJWVZdTGcRzc3d2RlZVFYUeIDWkw7CZPntxcdQhmzpyJ6dOn4/r160hNTYVMJoOjoyNWr14NHx8flJaWIiYmBlu2bMHSpUvB8zwAQC6XC+8hl8uF7TzPw9HR0eh33N9OCLENogYVNzepVAo/Pz/cuXMHP/74IziOQ9euXdGmTRs4OztjxowZSEpKQnl5OTiOAwDo9Xrh+Pu3cxxn1PZgOyHENlhk2NWqrq5GXl5eg/soFAq4uLhArVYL29RqtXD9T6lUGrXxPI+8vDyj64OEkNbPYsKuuLgY8fHx4Hke1dXVuHz5MuLj49GrVy/cuHEDt2/fRnV1NUpLS7F792707NlT6LqGhITg8OHD0Ol00Gg0OHHiBEJDQwEAgwYNQmZmJs6fPw+DwYBDhw5BpVLR9TpCbIzFLLhTUlKC6OhoqNVqMMbg6uqKUaNGYcSIETh79iz279+PkpIS2NvbCzconJ2dATQ+zi45ORm7du2CVqsVxtm5ubmJro3G2RFi/UTPZ3fgwAGkpqairKzMqG3z5s1mK85SUNgRYv1EPS42f/58pKWlYeTIkTRlDCHEKok6s/P398e5c+fg5OTUHDVZHDqzI8T6ibpB4enpSWvEEkKsmqhu7MSJE/F///d/mDlzJlxdXY3ahg4dapbCCCHElER1YwcPHvzwgyUSnDt3zuRFWRrqxhJi/Sxm6Iklo7AjxPqJHlRcUVGBhIQEHDt2DEDNI1fl5eVmK4wQQkxJ1Jldeno6/u///g92dnbIycnBjRs3cOLECRw8eBCffvppc9TZoujMjhDrJ+rM7v3338df/vIXnD59Gm3btgUABAcHC9MoEUKIpRMVdtevX8eECRMAQJjMk6ZJIoRYE1Fh5+3tjeTkZKNtly5dojUpCSFWQ9Q4u3fffRevvfYapk2bBoPBgC1btuCf//wn1q9fb+76CCHEJEQPPUlJSUFsbCyys7Ph6emJyMhI9OnTx9z1WQS6QUGI9aNxdiJQ2BFi/URds7t37x7WrFmD4OBg+Pn5AahZ0Wv37t1mLY4QQkxFVNitWLEC165dw5YtW4S7sd26dcPevXvNWhwhhJiKqBsUP/zwA+Lj4yGXyyGV1uSjh4cHcnNzzVocIYSYiqgzu7Zt26KystJo2507d+Di4mKWogghxNREhd1LL72E+fPnIzMzEwCQl5eHpUuXYuzYsWYtjhBCTEVU2C1evBje3t4ICwtDcXExhg4dis6dO2PBggXmro8QQkzikYee3LlzBx06dBBuVJjS5s2bceXKFdy7dw/Ozs4YO3YswsLCANSM84uJiUFBQYGwQljtehiNrS7W0LFi0NATQloBJsLy5cvZv/71L1ZYWChm98eWmZnJDAYDY4yx7OxsNnPmTPb777+z4uJi9tprr7H//ve/7N69e2zv3r1syZIlwnH79u1jy5YtY6WlpSwrK4vNnDmTXbp0iTHGGj1WDJEfEyHEgonqxsrlcnz22WcIDAzEiBEjsGzZMnz77be4c+eOSYNXqVQKs6pIJBJIJBLk5ubiwoULUCqVCA4Ohp2dHSZNmoSMjAxoNBoANWP+JkyYAIVCIXS3T548CQCNHksIsQ2ihp689957AGoGF//66684ceIEFi5ciLKyMmRlZZm0oJ07d+LkyZMwGAzw9fVF//79sX//fqhUKmEfjuPg7u6OrKwsODk5obCw0Kjdx8dHmH4qKyur3mO9vLxMWjshxHKJCruysjIkJibi3LlzOHfuHDQaDYYPH47g4GCTFzRz5kxMnz4d169fR2pqKmQyGXieh6Ojo9F+tVNM1U4zJZfL67QBaPBYQojtEBV2/v7+8Pb2xowZM/C3v/0NTz75pFmLkkql8PPzw+nTp/Hjjz+C4zjo9XqjfcrLy8FxHDiOAwDo9XrY2dkZtQFo8FhCiO0Qdc3unXfegUqlwpYtW/DXv/4VmzdvRmJiIioqKsxaXHV1NfLy8qBUKqFWq4XtPM8L2xUKBVxcXIza1Wo1lEolADR4LCHEdogKu/nz5yM2NhYXL17Ee++9h7KyMkybNg3+/v4mK6S4uBjx8fHgeR7V1dW4fPky4uPj0atXLwwaNAiZmZk4f/48DAYDDh06BJVKJVxzCwkJweHDh6HT6aDRaHDixAmEhoYCQKPHEkJsg6hxdoWFhTh//jzOnTuH//73v7h58yZ69+6NwYMH4/333zdJISUlJYiOjoZarQZjDK6urhg1ahRGjBgBAEhOTsauXbug1WqFsXJubm4AGh9n19CxYtA4O0Ksn6iw69q1K/r164egoCAMHjwYAwYMgL29fXPUZxEo7AixfqLCLi8vD507d66zPT8//5HOkKwVhR0h1k/UNbuQkJCHbn/mmWdMWgwhtiY6OhoODg7CIHpL/HFwcEB0dHRLf1RNJirsHnZWU1paapbnYwmxJdHR0dDpdC1dRoN0Ol2rCLsGx9kNGDAAEokEPM9j4MCBRm2FhYV4+eWXzVocIa3dwoULsWLFCosOPIVCgYULF7Z0GU3W4DW7c+fOgTGGadOm4YsvvvjjIIkErq6uZh9cbCnomh2xFvf3tujfWWOiblDo9Xqbuvv6IAo7Yi0o7Oon6pqdVCql1cUIIVZNVNgtX76cVhcjhFg1URMB/Otf/6LVxQghVo1WFyOE2ARaXYwQYhNodTFCiE2wqNXFLBUNPSHWgoae1E/UDQoAuHnzJo73jGjsAAAgAElEQVQfPy5MCjBmzBh06dLFnLURQojJiOrGHj16FM8//zzS09Mhl8tx9epVvPDCCzh69Ki56yOEEJMQ1Y0NDg7Gxo0bMXjwYGFbQkIC5s2bh4SEBLMWaAmoG0usBXVj6yfqzE6n0yEwMNBoW//+/VFeXm6WogghxNREhd3s2bOxZs0aYflBvV6PtWvXYvbs2WYtjhBCTKXebmzt9E5AzemwVquFRCKBk5MTiouLwRiDm5ubsBh1a0bdWGItqBtbv3rvxm7ZsqU56yCEELN65HF25lJRUYGdO3ciJSUFOp0O7u7uiIiIQL9+/ZCfn48333wT7dq1E/YfN24cJk6cKBzb0OpiKSkpiImJQUFBgbC6WKdOnUTXRmd2xFrQmV39RI+zM7eqqip07NgRK1asgKurKy5duoSNGzdiw4YNwj579uxBmzZt6hx78OBB5ObmYuvWrSgqKsLKlSvh7e2NgIAAlJSUYMOGDXjjjTcQGBiIuLg4bNq0CR9//HFz/nmEkBYm6gZFc+A4DpMnT4abmxukUikCAwPh5uaGmzdvNnrsqVOnMGHCBCgUCuGxtpMnTwIALly4AKVSieDgYNjZ2WHSpEnIyMiARqMx819ECLEkFnNm96CioiLk5ORAqVQK26KioiCRSNCnTx9MnToVjo6O0Ol0KCwshEqlEvbz8fERbpxkZWUZtXEcB3d3d2RlZcHLy6v5/iBCSIt6rDM7vV4Pg8Fg6loElZWV2LJlC4YPHw4vLy84Ojpi9erV2LZtG9asWQO9Xi/cQKkdDiOXy4Xj5XK5sJ3neaO2B9sJIbZBVNh9+OGHuHTpEgDg559/Rq9evdCjRw/8+OOPJi+ouroaf//73yGTyTB9+nQANWdjXbt2RZs2beDs7IwZM2YgKSkJ5eXl4DgOQE0A17p/O8dxRm0PthNCbIPoZ2Nr157YtGkTNm/ejN27d2Pt2rUmLYYxhk8//RTFxcVYuHAhZLLGe9kKhQIuLi5Qq9XCNrVaLXR/lUqlURvP88jLyzPqHhNCWj9RYVe7utjdu3ehVqvx4osvIiQkBNnZ2SYtZseOHdBoNFi0aBHs7OyE7Tdu3MDt27dRXV2N0tJS7N69Gz179hS6pyEhITh8+DB0Oh00Gg1OnDiB0NBQAMCgQYOQmZmJ8+fPw2Aw4NChQ1CpVHS9jhAbI2qc3ejRozFz5kxkZGTg999/x9atW3H37l0888wzSEpKMkkhWq0Wc+fORdu2bYV1LoCaR9UkEgn279+PkpIS2NvbCzconJ2dATQ+zi45ORm7du2CVqsVxtm5ubmJro3G2RFrQePs6icq7C5duoTly5fDzs4OGzZsgI+PD44cOYL//Oc/NvGkBYUdsRYUdvWzmCcoLBmFHbEWFHb1E3XN7rnnnsOOHTtQUFBg7noIIcQsRIXd22+/jYSEBAwePBhTp07F0aNH6wznIIQQS/ZI3djCwkIcP34cR44cwbVr1zBq1CiMHz8eQ4cONWeNLY66scRaUDe2fo98zU6v1+O7777D9u3bkZ2djY4dO0IikWDVqlUICQkxV50tisKOWAsKu/qJCrvq6mqcPn0ahw8fxs8//4zAwEBMmDABL7zwAuzt7fHtt99i6dKluHz5cnPU3Owo7Ii1oLCrn6iwCwgIQIcOHTBx4kSMHz8e7u7udfaZOHEiDh06ZJYiW1prDjvGgIwMID8fcHMDfHwAG1oSuNWhsKufqLBLSkpC3759m6Mei9Raw44x4LvvgLg4oLISkMmA8HBg9GgKPGtFYVe/R7pmp9PpcPfuXaMP8f7pk1qr1hp2t24By5fXBF0tmQxYuRLw9W25usjjo7Crn6j57K5fv44333wTaWlpwv/xaz/UrKwssxZIzCc/3zjogJrXWi2FHWl9RI2zW7JkCZ5++mmkpqbCwcEBaWlpmDp1KjZt2mTu+ogZubnVnMndTyYDHmF5DkKshqhurL+/P5KSktC2bVv06NED6enpKC8vx7PPPovz5883R50tqrV2Y+maXetD3dj6ierGtmvXDpWVlWjbti06dOgAjUYDJycnFBYWmrs+YkYSCTBqFODuDmRlAUol0K8fBR1pnUSF3aBBg/D1118jPDwcL774IqZOnQo7OzsMGTLE3PURM2IM+P57OrMjtuGRn6Corq7GkSNHUFZWhkmTJtVZ36E1aq3dWLob2/pQN7Z+j7y6mFQqFRanJtaN7sYSW9Lo3djr16/jz3/+M/r16wdfX1/069cPb7zxBq5fv94c9REzoruxxJY0GHY3b97EmDFjwPM8Fi9ejN27d2PRokUoLy/HmDFj8NtvvzVXncQMfHxqrtHVBl7tNTsfn5asihDzaPCa3YIFC8BxHFatWlWn7YMPPkBZWRk2btxo1gItQWu9Zgf88WysVltzRkfPxlo3umZXvwbD7umnn8aXX36JJ554ok5bVlYWJk2aROPsCLEgFHb1a/AGxZ07d+pdX9XLywt37941WSEVFRXYuXMnUlJSoNPp4O7ujoiICPTr1w8AkJKSgpiYGBQUFAgrhHX638WlxlYXa+hYQohtaPQGhaSePo1UKq237XFUVVWhY8eOWLFiBfbs2YPw8HBs3LgR+fn5KCkpwYYNGxAeHo5du3ahS5cuRo+qHTx4ELm5udi6dSuWL1+OY8eOCXPrNXYsIcQ2NHhmp9frMX78+Ie2McbA87zJCuE4DpMnTxZeBwYGws3NDTdv3oROp4NSqURwcDAAYNKkSZgxYwY0Gg28vLxw6tQpREVFQaFQQKFQICwsDCdPnkRAQAAuXLjQ4LGEENvQYNht2LChwYMjIiJMWsz9ioqKkJOTA6VSiR9//NFoKimO4+Du7o6srCzhsbX72318fJCYmAig5tpifcdS2BFiOxoMu/vPtJpTZWUltmzZguHDh8PLyws8z8PR0dFoH7lcDp7nhbPL+5/kqG0D0OCxhBDbIWqKp+ZUXV2Nv//975DJZJg+fTqAmrOxB5duLC8vB8dx4DgOAIzaa9saO5YQYjssKuwYY/j0009RXFyMhQsXQva/0a5KpRJqtVrYj+d55OXlQalUQqFQwMXFxahdrVYLd5EbOpYQYjssKux27NgBjUaDRYsWwc7OTtg+aNAgZGZm4vz58zAYDDh06BBUKpVwzS0kJASHDx+GTqeDRqPBiRMnEBoaKupYQohtqHdQ8UsvvYRvvvkGAPC3v/0NCxYsMGshWq0Wc+fORdu2bSGV/pHBs2fPxrBhw5CcnIxdu3ZBq9UKY+Xc3NwAND7OrqFjxaBBxcRa0KDi+tUbdj179sQvv/wCjuPQvXt3XLt2rblrsxgUdsRaUNjVr967sc899xyGDRsGpVIJnufrHW935MgRsxVHCCGmUm/Ybdy4ERcuXEBWVhaSkpIwZcqU5qyLEEJMStRMxV9++aVNhx11Y4m1oG5s/URPyx4fH4/Dhw8jNzcX7u7uGD9+PIYOHWru+iwChR2xFhR29RM19CQ2NhZz5sxBp06dMGrUKLi5ueHNN9/Evn37zF0fIYSYhKgzu6FDh+Kzzz5Dz549hW1paWmYNWsW4uPjzVqgJaAzO2It6MyufqLO7AoLC9GtWzejbV27dkVRUZFZiiKEEFMTFXaDBg3CypUrhWdMy8vL8dFHH2HAgAFmLY4QQkxFVDc2Ly8PUVFRuHjxIpydnVFUVIQBAwZg69atcHd3b446WxR1Y4m1oG5s/R5pkezbt28jLy8PnTt3hqenpznrsiitOexqF9zJz69ZWpEW3LFuFHb1e6Sws1WtNewYA777DoiLq1kcu3YpxdGjKfCsFYVd/Sxq1hPSvDIy/gg6oOZ/4+JqthPS2lDY2bD8/D+CrlZlZc0asoS0No2GXXV1Nc6ePQuDwdAc9ZBm5OZW03W9n0xWs1g2Ia1No2EnlUoxffp0o8k0Sevg41Nzja428Gqv2fn4tGRVhJhHgwvu1AoKCsIvv/yCwMBAc9dDmpFEUnMzwt+/puvaqRPdjSWtl6iw8/b2xrRp0/Dcc8/B09PT6I7Pu+++a7biiPlJJICvb80PIa2ZqLDjeR7PP/88ACAnJ8esBRFCiDnQODsRWus4O9L60Di7+ok6swOAGzdu4JtvvkFBQQE+/vhj/PbbbzAYDPD39zdnfcTM6AkKYitEndkdP34cS5YswejRo/HVV1/h2rVrSEpKwqpVqxAXF2eSQn744QecPHkSmZmZGDJkCObOnQsAyM/Px5tvvol27doJ+44bNw4TJ04E0PjKYikpKYiJiUFBQYGwslinRxxb0VrP7OgJitaHzuzqJ+rMbsOGDdi/fz969eqF48ePAwD8/f2RlpZmskJcXFwwfvx4JCUlPXRM3549e9CmTZs62w8ePIjc3Fxs3boVRUVFWLlyJby9vREQEICSkhJs2LABb7zxBgIDAxEXF4dNmzbh448/Nlnd1qy+Jyj8/emGBWl9RD1BUVBQIEzcWftfDolEYvRfkaYKCgrCoEGD4ODg8EjHnTp1ChMmTIBCoYC3tzfCwsJw8uRJAMCFCxegVCoRHBwMOzs7TJo0CRkZGdBoNCar25rRExSPLjo6Gg4ODsK//5b2c7+WrqW+HwcHB0RHRzf7dycq7Pr06YNDhw4ZbTt27BgCAgLMUtTDREVF4Y033sC2bdtQUlICANDpdCgsLIRKpRL28/HxQVZWFgAgKyvLqI3jOLi7uwvtto6eoHh0K1asgE6na+kyrJpOp8OKFSua/feK6sb+v//3/xAREYEvv/wS5eXlePXVV3Hz5k3Exsaauz44Ojpi9erV8PHxQWlpKWJiYrBlyxYsXboUPM8DAORyubC/XC4XtvM8D0dHR6P3u7/d1tU+QfHgNTt6gqJ+FHSm0RKfo6iwe/LJJ3H69Gn89NNPGDFiBDw9PTFixAi0b9/e3PWB4zh07doVAODs7IwZM2Zg9uzZKC8vB8dxAAC9Xi88znb/do7jhNmVa93fbuvoCYqmoRsAj86Ul74eleihJ/b29hg4cCCeeOIJdO7cuVmCrjEKhQIuLi5Qq9Xo06cPAECtVkOpVAIAlEolTp06JezP8zzy8vKEdkJPUBDbIeqanUajwSuvvILBgwfjtddew+DBg/Hyyy8jOzvbZIVUVVXBYDCguroa1dXVMBgMqKqqwo0bN3D79m1UV1ejtLQUu3fvRs+ePYWua0hICA4fPgydTgeNRoMTJ04gNDQUQM3aGZmZmTh//jwMBgMOHToElUoFLy8vk9VNCLEOosbZTZw4Ef7+/li8eDHkcjnKysqwbt06pKam1rlx8bgOHDhQ570mTpwIT09P7N+/HyUlJbC3t0efPn0wdepUODs7A2h8nF1ycjJ27doFrVYrjLNzc3N7pNpa6zg7gAYVPyoax9Y0Lfn5iQq77t2748qVK2jbtq2wzWAwoFevXrh+/bpZC7QErTXsaFDxo6Owa5qW/PxEdWP79++Py5cvG21LSkqiKZ+sHE3LTmxJvTco1q9fL/yzSqXCtGnTEBYWBk9PT9y+fRv//ve/8fLLLzdLkcQ8GhpUTDcsSGtTb9jdvn3b6PWoUaMA1DxNYWdnhxdeeAH37t0zb3XErGoHFd8feDSomLRWNMWTCHTNjtSia3ZNY/E3KICagbu3bt1CWVmZ0faBAweapTBL0lrDDvjjbiwNKhaHwq5pLD7sDh48iA8++ABt27Y1evpAIpEgMTHRrAVagtYcduTRUNg1jcWHXUBAADZv3oyQkJDmqMniUNiRWhR2TWPxQ0/atm2L4OBgc9dCCCFmIyrs3n33XaxcuRJ37941dz2EEGIWorqxFy9eRFRUlNHKYowxSCQSm5gbjrqxpBZ1Y5vG4q/ZDRkyBOPGjcPYsWPrTI/kYwOTn1HYkVoUdk1j8WHn7++P1NTUFp2LqiVR2JFaFHZNY/E3KCZPnmyy2U0IIaQliDqze/nll3H58mUolco6yxAeOXLEbMVZCjqzI7XozK5pLL4be+DAgXrbJk+ebNKCLBGFHalFYdc0Fh92to7CjtSisGualvz8RK1B8eWXX9bbNmXKFJMVQwgh5iIq7B68OaHVaqFWqzFgwAAKO0KIVXissANqzvZu3Lhh8oIIIcQcHvuaXXV1NXr37o3U1FRT12Rx6JodqUXX7JrG4q/ZVVdXG73W6/U4fPgwHB0dzVIUIYSYmqiwe+KJJ+o8PeHu7m60TkVT/fDDDzh58iQyMzMxZMgQzJ07V2hLSUlBTEwMCgoKhOUQa8f7NbaUYkPHEkJsh6hu7IOLYcvlcnTo0MGkhSQkJEAikSApKQkGg0EIu5KSErz11lt44403EBgYiLi4OFy9ehUff/wxACA2NhZXr17Fe++9h6KiIqxcuRJz585FQEBAo8eKRd1YUou6sU1j8Y+LeXt7G/2YOugAICgoCIMGDYKDg4PR9gsXLkCpVCI4OBh2dnaYNGkSMjIyoNFoAACnTp3ChAkToFAo4O3tjbCwMJw8eVLUsYQQ29FgN3bixIkNPvwvkUgafLrCFLKysqBSqYTXHMfB3d0dWVlZcHJyQmFhoVG7j4+PMFV8Q8d6eXmZtW5CiGVpMOwmTJjw0O25ubmIiYmBXq83S1H343m+zo0QuVwOnufB87zw+sG2xo4lhNiWBsMuIiLC6PXdu3fx97//HbGxsRg7dizmz59v1uKAmrOxB0O1vLwcHMcJc+vp9XrY2dkZtTV2LCHEtoi6ZldaWop169Zh6NCh0Gq1+OGHH7Bu3Tp4enqauz4olUqo1WrhNc/zyMvLg1KphEKhgIuLi1G7Wq2GUqls9FhCiG1pMOz0ej22bNmC4OBg3LhxA0ePHsWWLVvMMjtxVVUVDAYDqqurUV1dDYPBgKqqKgwaNAiZmZk4f/48DAYDDh06BJVKJVxzCwkJweHDh6HT6aDRaHDixAmEhoYCQKPHEkJsR4NDTwICAlBVVYU5c+agT58+D91n6NChJinkwIEDdR5LmzhxIiZPnozk5GTs2rULWq1WGCvn5uYGoPFxdg0dKxYNPSG1aOhJ01jsFE9BQUGN3o09d+6cWQqzJBR2pBaFXdNYbNiRGhR2pBaFXdNY/KBiYlmiowEHB0AisdwfB4eaOgmxFHRmJ4Klndl5egL3LeFrsTw8gNu3W7oK06Izu6ahMzvySBYuBBSKlq6iYQpFTZ2EWAo6sxPB0s7sTO3+e1Ct+M80CTqzaxo6syOEEDOjsCOE2AQKO0KITaCwI4TYBAo7QohNoLAjhNgECjtCiE2gsCOE2AQKO0KITaCwI4TYBAo7QohNoLAjhNgECjtCiE2gsCOE2AQKO0KITWhwkWxLsmLFCty4cQNSaU0+d+jQAZ988gkA4OzZs4iNjUVpaSl69+6NqKgoKP43u6VOp8P27duRnJwMBwcHvPrqqyZbEU2s6GhgxQpAp2vWX/tYGlhfqcUpFDWfo6VMCtrQYlTEAjErsXz5cvbzzz/X2Z6ZmcmmTZvGUlNTmV6vZ5s2bWIbN24U2jdu3Mj+9re/Mb1ez9LT09lrr73GMjMzH+l3N/VjUigYq5kWk36a+qNQNOmraDIA9GOin+Zm9d3YM2fOIDAwEP7+/uA4DuHh4UhISIBerwfP80hISEB4eDg4joOfnx8GDBiA06dPN2uN1nBGZy3osySPy2q6sQAQGxuL2NhYeHp6YsqUKejZsyeys7PRrVs3YR93d3fIZDLk5ORAIpFAKpXC09NTaFepVEhLS2uJ8gGApj1/TJbYY2T0ZT6yluz6W03YRUZGwtvbGzKZDPHx8Vi7di3WrVsHnuchl8uN9pXL5dDr9ZBKpQ9t43m+OUsnhFgAqwm7p556Svjn0NBQxMfH49KlS+A4Dnq93mhfvV4Pe3t7SCSSh7ZxHNcsNRNCLIfVXrOrXfHL29sbarVa2J6Xl4eKigp4eHjAw8MDVVVVyLlvkVW1Wg2lUtkSJRNCWpBVhF1ZWRkuX74Mg8GAqqoqnDlzBunp6QgICMCwYcPwyy+/ID09HTzPIy4uDkFBQbC3twfHcQgKCkJcXBx4nsfVq1eRmJiIkJCQlv6TCCHNzCrWjS0pKcHq1auh0WgglUrh5eWF8PBw9OnTB0DNOLt9+/ZBp9M9dJzdtm3bkJKSAoVCgcjIyEceZ9fUdWNpXdams5TPkNaNbZqW/PysIuxaGoVdy7OUz5DCrmlokWxCCDEzCjtCiE2wmqEnrYUlDo4lxBbQmR0hxCZQ2BFCbAJ1Y5sZ3cB7PNT9J01FZ3aEEJtAYUcIsQkUdoQQm0BhRwixCRR2hBCbQGFHCLEJNPSkmdEQCkJaBp3ZEavyv5m7CHlkFHbN4L71fkgT1K4bS8jjoPnsRGjqfHaWzlLmirMGtDC26dB8doSQVk/RAtcjKOwIeQSedE2iyRQKBVa0wPUI6saKQN1YYi1o2vj60ZkdIcQm2ETY6XQ6rF+/HtOmTUNUVBTOnj3b0iU1SXQ04OBQc0Zmip/7meo9HRxq6iTEUthE2O3cuRMymQw7duzAvHnzsGPHDmRlZbV0WY8tOhrQ6Vq6iobpdBR2YkRHR8PBwQESicQkP/cz1Xs6ODgguhV8ma0+7HieR0JCAsLDw8FxHPz8/DBgwACcPn26pUt7bAsXWv7gWoWipk7SsOjoaOgs/L9cOp2Ows4a5OTkQCqVGt1FU6lUVn1mt3AhUFpaczPBUn9KSynsxFi4cGGLDMN4FAqFAgtbwZfZ6p+N5XkecrncaJtcLgfP8y1UESF/WLhwYasIEmvQ6s/sOI6DXq832qbX68FxXAtVRAhpCa0+7Dw8PFBVVYWcnBxhm1qthlKpbMGqCCHNrdWHHcdxCAoKQlxcHHiex9WrV5GYmIiQkJCWLo0Q0oxs4gkKnU6Hbdu2ISUlBQqFApGRkRg6dKjo4+nhb0Is06PEl02EHSGEtPpuLCGEABR2hBAbQWFHCLEJFHaEEJtAYUcIsQkUdoQQm0BhRwixCRR2hBCbQGFHyGO4du0a+vXrBwcHB2zevLmlyyEiUNhZCB8fH9jb20OhUAg/t2/fxvXr1zFu3Dh06tQJHTp0wPPPP49r1641a20nT56Et7d3s/5OS7du3TqEhoaitLQU8+bNQ3Z2NiZMmABXV1c4OTmhd+/e2LNnDwAgIyMDEokEL774otF7TJ06VVhl6+TJk5BKpVAoFHBwcED37t2xe/fuZv2bfHx88PPPPzfr72xOFHYW5Pjx49DpdMKPp6cnioqKMHbsWFy7dg15eXkYNGgQxo0b19Kl2jy1Wo2ePXsKr6dNmwalUgm1Wo07d+5g79696Ny5s9Ex58+fR3x8fL3v6enpCZ1Oh5KSEqxduxazZs1CWlqa2f4Gm8OIRVCpVOynn35qdL87d+4wAKygoOCh7eXl5WzBggXsiSeeYI6OjmzIkCGsvLycMcbYsWPHmL+/P3NycmLDhw9naWlpRr9//fr1rHfv3szR0ZFNnjyZ6fV6ptPpGMdxTCKRsPbt27P27dszjUbDEhIS2ODBg5mTkxNzd3dnc+fOZffu3RPeDwDbunUre/LJJ5lCoWAffPAB++2339jgwYOZg4MDmzRpktH+1uSZZ55hUqmUtWvXjrVv355du3aNtW/fnl26dOmh+9+6dYsBYGvWrGGhoaHC9sjISLZ8+XLGGGP/+c9/mJeXl9Fxrq6u7ODBgw99z6+++or17duXOTg4sC5durDvv/+eMcaYRqNhY8aMYS4uLqxr167s888/F45Zvnw5mzRpEps2bRpTKBTM39+fJSYmMsYYmzp1KpNIJIzjONa+fXu2du1axhhjEydOZJ07d2aOjo5s2LBh7MqVK8L7vf7662zOnDnshRdeYO3bt2dPP/00y8nJYW+//TZzdnZm3bt3Z7/++usjfrrmQ2FnIcSG3dGjR5m7u3u97VFRUWz48OEsOzubVVZWsvj4eMbzPLt27RqTy+Xsxx9/ZAaDga1du5Z17dpVCByVSsUGDhzINBoNu3PnDvPz82Pbt29njD38/4gXL15k586dYxUVFezWrVvMz8+Pbdy4UWgHwMaMGcOKi4vZlStXmJ2dHXv22WfZ77//zoqKiliPHj3Ynj17HuejsgjDhw9nO3bsEF6HhYWxp59+mu3fv5+p1WqjfWvDrrS0lHl6egrfc31hV1VVxY4cOcJkMhm7evVqnd+dkJDAHB0d2Y8//siqqqpYdnY2S09PZ4wxFhISwubMmcP0ej27dOkSc3V1ZT///DNjrCbs2rVrx7799ltWWVnJFi9ezIKCgoT3fdi/gzExMaykpITxPM/efvtt1rdvX6Ht9ddfZx07dmQXL15ker2ePfPMM8zHx4f94x//YJWVlWzp0qVG4d7SKOwshEqlYu3bt2dOTk7MycmJjRs3rs4+WVlZzNPTk8XGxj70PaqqqhjHcezy5ct12j788EM2adIko309PT3Zf/7zH+H3//Of/xTa3333XfbnP/+ZMfbwsHvQxo0b2csvvyy8BsDOnj0rvO7fvz9bs2aN8HrBggXs7bffbvA9LdmDYXf37l22aNEi5u/vz6RSKevbty+7cOECY+yPsKuoqGBbt24VAubBsJNIJMzJyYm5uLiwvn37sv379z/0d8+ePZvNnz+/zvbMzEwmlUpZSUmJsG3x4sXs9ddfZ4zVhF1YWJjQlpqayjiOE1439h/cwsJCBoAVFRUxxmrCbubMmUL75s2bmZ+fn/A6OTmZOTk51ft+zY2u2VmQr776CkVFRSgqKsJXX31l1KbVavHcc88hKioKERERDz2+oKAAPM+ja9euddpu374NlUolvJZKpVAqldBoNMI2d3d34Z/lcnmDq15dv34dL730Etzd3eHo6IglS5agoKDAaJ/7r1nZ29vXeW3pq2o9ChcXF6xZswapqanIy8tDQEAAXn755Trzrc2aNQt5eXk4fvx4nfeovUZ79+5dXL58GVOmTHno78rKyqr3O+7QoQMcHByEbSqVqtGAkzcAAAc+SURBVMHvmOd5VFZWPvT3VFVVYfHixejatSscHR3h4+MDAEbfszV9xxR2VqCwsBDPPfccxo4di6VLl9a7n6urKziOw++//16nzdPTE2q1WnjNGENWVha8vLwa/f0Pm7x0zpw58PPzw40bN1BSUoJVq1Y90kSKrZmrqyv+8pe/4Pbt27h7965RW9u2bbF8+XIsW7bssT8vpVJZ73d89+5dlJaWCtsyMzNFfcdA3e85NjYWx44dw88//4zi4mJkZGQAeLQJMy0JhZ2FKykpwfPPP48hQ4ZgzZo1De4rlUoxffp0LFiwALdv30ZVVRXOnTuHe/fuYfLkyfj2229x4sQJVFRUIDo6Gu3atcPTTz/daA2dO3fGnTt3UFxcLGwrLS2Fo6MjFAoFrl69iu3btzf5b7VmixYtwpUrV1BZWYnS0lJs374dTz75JDp27Fhn32nTpuHevXv44YcfHut3zZgxA7t378aJEydQXV0NjUaDq1evQqlU4umnn8b7778PnueRnJyMmJgYREZGinrfzp074+bNm8Lr0tJStGvXDh07dkR5eTmWLFnyWPVaCgo7C3f06FEkJiZi9+7dRmPwMjMzH7r/hg0b0Lt3bwwcOBAdOnTAokWLUF1dje7du+OLL77AW2+9BVdXVxw/fhzHjx+HnZ1dozX4+fkhIiICXbp0gbOzM27fvo0NGzYgNjYWDg4OmDVrFsLDwx/7b8zMzGzwb7IG5eXleOWVV+Ds7IwuXbpArVbj66+/fui+bdq0wcqVK+uc9Yk1aNAg7N69G++88w6cnJwwfPhw4ax9//79yMjIgKenJ1555RWsXLkSI0eOFPW+77//Pj766CM4Oztjw4YNeO2116BSqeDl5QV/f38MHjz4seqtNWrUKKxatapJ79EUNC07IcQm0JkdIcQmUNgRQmwChR0hxCZQ2BFCbAKFnRV7//33sWnTppYuQ/D111/XOxCWNMzSvsvNmzdj8eLFLV2GabXg0xukCfLz85mnp6fwkL+pbNy4kfn6+jIHBwfm4eHB5s+fzyoqKoT2Dz74gPXq1Yu1adNGeNTpfj179mRJSUkmram1M9d3WevevXuse/fuRo/8abVa9vTTT7MOHTowJycnNnjwYKPH+/R6PfPy8mJ5eXlmqakl0JmdldqzZw9Gjx4Ne3t7k77vmDFj8Ouvv6KkpARXrlxBUlKS0eSUTz75JNatW1dnbrZaERER+Pzzz01aU2tnru+y1vr16+Hm5ma0TaFQYNeuXdBqtSgsLMSiRYswZswY4dExjuMwatQo7N271yw1tQQKOyv1/fffY/jw4cLr2gk2o6Oj4ebmBg8Pj8ea/LFr165wdnYGUPNYkFQqxW+//Sa0v/766xg1apTR85f3Cw0NxbfffvvIv9eWmeu7BIBbt27hiy++wPvvv2+0neM4dO/eHVKpFIwxtGnTBoWFhUYDnVvbd0lhZ6VSUlLQvXt3o225ubkoLi6GRqNBTEwM5s6di8LCQgDAmjVr4OzsXO/P/WJjY+Ho6AhXV1ckJSXhz3/+s+i6evTogYyMDJSUlDT9j7QR5vwu33rrLaxatares8Y+ffqA4ziMHTsWM2fONDoD7NGjB5KSkkz817aglu5Hk8cjk8mEOcwYq5kiiOM4o+trnTp1YufOnXvs33H9+nX2wQcfsJycnDpt909PdD+DwcAA1JnTjdTPXN/lkSNH2PPPPy+8Z33TdOn1ehYbG1tnfsHr168zqVT6SL/TktGZnZVycXExmt0CADp27AiZTCa8bmyapsY89dRT6NmzJ6KiokQfU1vTg2cYpH7m+C7Lysrw3nvvYcuWLY3uy3EcIiIisGbNGqMzudLSUjg5OYn+nZaOws5K9enTB9evXxe9/6pVq4wmEnjwpz6VlZUPnU6oPunp6fDx8YGjo6PoY2ydOb7LGzduICMjA8OGDYO7uzvGjx+PnJwcuLu7C1M1PaiiosJo1pP09HT07du3SX+bJaGws1KjR4/GqVOnRO+/ZMkSo8V8HvyptXPnTuTn5wMA0tLSsHr1aoSFhQntFRUV4Hke1dXVqKysBM/zqKqqEtpPnTqFUaNGmeAv/P/t3TGKhDAYBeBXi42NFoN4AbG18yp6BAsrb6CtrWBh5ykULCxzBxvFAwRsMsWyMrvMwjLs4MzmfWAVMD/88CAhJPp4Ri9938c8zxBCQAiBuq7hOA6EEHBdF9M0YRxH7PsOKSWKosC6rgjD8Jjn3/Xy7HU0PWbbNnW5XI6zWff2ZH77rsWtJEmUbdvKMAzleZ7KskxJKY/xOI4VgC9f0zTHuO/7d6+Fp589q5e3vv+z73sVBIEyTVNZlqWiKFLDMBzjn+fslmV5eM5Xwyue3lie57BtG2manl0KgI+nINu2Rdd1Z5fydl6tl1VVYZ5nlGV5dil/hmFHRFrgnh0RaYFhR0RaYNgRkRYYdkSkBYYdEWmBYUdEWmDYEZEWGHZEpIUrdDuaOMd2ZYUAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 306x432 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "fig = plt.figure(1, figsize=(4.25, 6.0))\n",
    "ax1 = axes()\n",
    "\n",
    "box_width = 0.7\n",
    "ind = [1,2]\n",
    "\n",
    "### Example in case we want to show means later ###\n",
    "##Syn_values = SNP_variants_within_patients[SNP_variants_within_patients.SNP_type == 'S']\n",
    "##meanlineprops = dict(linestyle='--' , color = 'b' , linewidth = linewidth_for_box)\n",
    "##bp = ax1.boxplot(values_to_variant_types(Syn_values) , positions = [ind[0]], widths = box_width , meanprops = meanlineprops, meanline = True, showmeans = True)\n",
    "##setBoxColors(bp , 'b')\n",
    "\n",
    "#times between isolate collection for F2 contaminated samples\n",
    "bp = ax1.boxplot(F2_contaminated_pairwise_isolate_times.values.astype(float) , positions = [ind[0]], widths = box_width)\n",
    "setBoxColors(bp , 'blue')\n",
    "\n",
    "#times between isolate collection for fSNP contaminated samples\n",
    "bp = ax1.boxplot(fSNP_contaminated_pairwise_isolate_times.values.astype(float) , positions = [ind[1]], widths = box_width)\n",
    "setBoxColors(bp , 'black')\n",
    "\n",
    "max_time_btwn_collection = max(np.max(F2_contaminated_pairwise_isolate_times.values) , np.max(fSNP_contaminated_pairwise_isolate_times.values) )\n",
    "\n",
    "#y-labels\n",
    "ax1.set_ylim([-75 , max_time_btwn_collection + 75])\n",
    "ax1.set_yticks(range(0 , max_time_btwn_collection , 500))\n",
    "y_labels = range(0 , max_time_btwn_collection , 500)\n",
    "ax1.set_yticklabels(y_labels, rotation='horizontal', fontsize = 12)  \n",
    "ax1.tick_params(axis='y', which='major', labelsize=12, pad = 4)\n",
    "\n",
    "#x-labels\n",
    "ax1.set_xlim([0.5 , 2.5])\n",
    "ax1.set_xticks(ind)\n",
    "x_labels = ['F2 contam.\\n(n=' + str(len(F2_contaminated_pairwise_isolate_times)) + ')', 'fSNP contam.\\n(n=' + str(len(fSNP_contaminated_pairwise_isolate_times)) + ')']\n",
    "ax1.set_xticklabels(x_labels, rotation='horizontal', fontsize = 12)  \n",
    "\n",
    "ax1.spines['right'].set_visible(False)\n",
    "ax1.spines['top'].set_visible(False)\n",
    "\n",
    "#label y-axis\n",
    "fig.text(-0.013, 0.56, 'Number of Days between Isolate Collection',  ha='center', va='center', rotation=90, fontsize = 12 , color = 'k')\n",
    "\n",
    "plt.tick_params(\n",
    "    axis='x',          # changes apply to the x-axis\n",
    "    which='both',      # both major and minor ticks are affected\n",
    "    bottom='off',      # ticks along the bottom edge are off\n",
    "    top='off',         # ticks along the top edge are off\n",
    "    labelbottom='on') # labels along the bottom edge are off\n",
    "\n",
    "plt.tick_params(\n",
    "    axis='y',          # changes apply to the x-axis\n",
    "    which='both',      # both major and minor ticks are affected\n",
    "    left='off',      # ticks along the bottom edge are off\n",
    "    right='off',         # ticks along the top edge are off\n",
    "    labelleft='on') # labels along the bottom edge are off\n",
    "\n",
    "fig = plt.gcf()\n",
    "fig.tight_layout()\n",
    "\n",
    "#output for Adobe Illustrator\n",
    "#file_name = '/home/rv76/Farhat_Lab/Figures/Final_5/Sample_Summary/times_between_serial_isolate_collection_AR_SNPs_vs_no_AR_SNPs.pdf'\n",
    "#plt.savefig(file_name, bbox_inches='tight', dpi = 300 , transparent = True)\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [],
   "source": [
    "#compute Mann-Whitnes U test\n",
    "non_parametric_test_pval = scipy.stats.mannwhitneyu(F2_contaminated_pairwise_isolate_times.values , fSNP_contaminated_pairwise_isolate_times.values )[1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.21993464470036528"
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "non_parametric_test_pval"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
