{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>.container { width:100% !important; }</style>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from IPython.core.display import display, HTML\n",
    "display(HTML(\"<style>.container { width:100% !important; }</style>\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### This notebook was created to find all of the High Quality SNV calls for all of the isolates in a DataBase with ~33k VCF files from processed FASTQ files that were downloaded from NCBI"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import vcf\n",
    "\n",
    "%matplotlib inline\n",
    "import os\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib as mpl\n",
    "import matplotlib.ticker as ticker\n",
    "from itertools import compress\n",
    "from pylab import MaxNLocator\n",
    "import seaborn as sns; sns.set()\n",
    "from matplotlib.colors import LogNorm\n",
    "from matplotlib import gridspec\n",
    "import ast\n",
    "import itertools\n",
    "import seaborn as sns\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "\n",
    "import fastcluster\n",
    "from sklearn import cluster, datasets\n",
    "import scipy.cluster.hierarchy as hier\n",
    "from sklearn.cluster import KMeans\n",
    "import time\n",
    "import sys\n",
    "\n",
    "import Bio\n",
    "from Bio.Alphabet import IUPAC\n",
    "from Bio.Blast.Applications import NcbiblastnCommandline\n",
    "from Bio.Blast import NCBIXML\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.SeqFeature import SeqFeature, FeatureLocation\n",
    "from Bio import pairwise2\n",
    "from Bio import SeqIO\n",
    "from Bio.Graphics import GenomeDiagram\n",
    "from Bio.SeqUtils import GC\n",
    "\n",
    "from Bio.Align.Applications import MuscleCommandline\n",
    "from StringIO import StringIO\n",
    "from Bio import AlignIO\n",
    "from Bio.Align import AlignInfo\n",
    "from Bio.Seq import MutableSeq\n",
    "import itertools\n",
    "import gzip\n",
    "\n",
    "import networkx as nx\n",
    "import scipy\n",
    "import pickle"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "####################################################################################################################################################################################"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## [1] This section of the notebook loads all of the Lineage Calls that were made on the isolates in rollingDB as well as the mean coverage / isolates for all of the isolates in rollingDB. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "####################################################################################################################################################################################"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### We runs through the pilon-generated VCF for each isolate (N = 33,875) and scrapes the High Quality Single Nucleotide Variant calls made against H37Rv. The DataFrame containing these high quality SNV calls (with the Reference Position, Reference Allele and Alterante Allele) for each isolate is then stored in a directory containing all of the DFs for all isolates. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Side-Note: Some isolate have ~20x coverage but we require 25 reads at a given position to make a SNP call"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Scrape Luca's DB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.0500090232801\n",
      "0.10001804656\n",
      "0.15002706984\n",
      "0.20003609312\n",
      "0.2500451164\n",
      "0.30005413968\n",
      "0.35006316296\n",
      "0.400072186241\n",
      "0.450081209521\n",
      "0.500090232801\n",
      "0.550099256081\n",
      "0.600108279361\n",
      "0.650117302641\n"
     ]
    }
   ],
   "source": [
    "#genomic data directory\n",
    "rolling_DB_dir = '/n/data1/hms/dbmi/farhat/rollingDB/genomic_data/'\n",
    "\n",
    "#get all folders (each folder corresponds to a different sequenced isolate)\n",
    "isolate_directories = os.listdir(rolling_DB_dir)\n",
    "\n",
    "#iterate through each sequenced isolate\n",
    "isolate_i = 0\n",
    "\n",
    "for isolate_ID in isolate_directories:\n",
    "\n",
    "    #directory that stores files for each sequenced isolate\n",
    "    directory_for_sequenced_isolate = rolling_DB_dir + isolate_ID\n",
    "\n",
    "    #check to see if megapipe successfully ran on sequenced isolate\n",
    "    try:\n",
    "        #existence of a PILON directory and corresponding VCF file\n",
    "        if ( 'pilon' in os.listdir(directory_for_sequenced_isolate) ):\n",
    "            \n",
    "            #existence of a REDUCED VCF and existence of FULL VCF file in relevent directories \n",
    "            if ( 'vcf' in list( itertools.chain.from_iterable( [filename.split('.') for filename in os.listdir(directory_for_sequenced_isolate + '/pilon/')] ) ) ) and ( isolate_ID + '_full' in list( itertools.chain.from_iterable( [filename.split('.') for filename in os.listdir(directory_for_sequenced_isolate + '/pilon/')] ) ) ):\n",
    "                \n",
    "                #we have a valid VCF file so megapipe ran successfully, let's keep the variant call information for this sequenced isolate\n",
    "                \n",
    "                #PILON-generated VCF DATA\n",
    "                ########################################################################################################################\n",
    "                \n",
    "                #dictionaries to construct DataFrame from later that hold information about each high-quality SNP found for a specific isolate from rollingDB\n",
    "                ref_dict = {} #reference allele\n",
    "                alt_dict = {} #alternate allele\n",
    "                alt_AF_dict = {} #frequency of alternate allele in population (0.90 - 1.0)\n",
    "                \n",
    "                #load in VCF file if one was found\n",
    "                vcf_reader = vcf.Reader( open(directory_for_sequenced_isolate + '/pilon/' + isolate_ID + '.vcf' , 'r') )\n",
    "\n",
    "                #iterate through each Variant Call \n",
    "                for record in vcf_reader:\n",
    "                    \n",
    "                    #check to see if the call is a PASS by Pilon\n",
    "                    if record.FILTER == []:\n",
    "\n",
    "                        #check to see if variant is SNP (length of Reference Allele is 1 base ; there's only 1 alternate allele ; length of alternate allele is 1 base)\n",
    "                        if (len(record.REF) == 1) and (len(record.ALT) == 1) and (len(str(record.ALT[0])) == 1):\n",
    "\n",
    "                            ##### Retrieve Relevant information for filtering quality of Base Call #####\n",
    "                            # Mean Base Quality @ locus\n",
    "                            BQ = record.INFO['BQ']\n",
    "                            # Mean Mapping Quality @ locus\n",
    "                            MQ = record.INFO['MQ']\n",
    "                            # Number of Reads w/ Deletion \n",
    "                            DC = record.INFO['DC']\n",
    "                            # Number of Reads w/ Insertion\n",
    "                            IC = record.INFO['IC']\n",
    "                            # Depth of Valid Reads in Pileup\n",
    "                            VD = record.INFO['DP']\n",
    "\n",
    "                            ### Filtering Criteria\n",
    "                            #---> Mean Base Quality > 20\n",
    "                            #---> Mean Mapping Quality > 30\n",
    "                            #---> No Reads Supporting Insertions\n",
    "                            #---> No Reads Supporting Deletions\n",
    "                            #---> Number of High Quality Reads >= 25\n",
    "                            if (BQ > 20) and (MQ > 30) and (DC == 0) and (IC == 0) and (VD >= 25): #SNP passed filtering criteria!\n",
    "\n",
    "                                # Filtering Criteria for mutant allele frequency calculation (range: 0.90 - 1.0)\n",
    "                                ref_allele = str( record.REF )\n",
    "                                alt_allele = str( record.ALT[0] )\n",
    "                                ref_pos = int( record.POS )\n",
    "                                \n",
    "                                # Calculate Alternate Allele Frequency\n",
    "                                ########################################################################\n",
    "                                #position-base dictionaries (order of output in Pilon)\n",
    "                                base_order = {0:'A' , 1:'C' , 2:'G', 3:'T'}\n",
    "                                base_order_r = {'A':0 , 'C':1 , 'G':2 , 'T':3}\n",
    "\n",
    "                                # Quality-Percentage - Percentage of As, Cs, Gs, Ts weighted by Q & MQ at locus\n",
    "                                QP = record.INFO['QP']\n",
    "\n",
    "                                if (sum(np.array(QP) == 0) == 3): #1 Allele supported by Reads (Alternate), store information for this Base Call (3 bases have NO reads supporting)\n",
    "\n",
    "                                    #create a Boolean list for the Bases that have at least 1% of support from reads\n",
    "                                    supported_bases_boolean = list( np.array( QP ) > 0 )\n",
    "\n",
    "                                    #positions of Bases [A,C,G,T] with at least 1% of support from reads\n",
    "                                    supported_bases_i = [i for i,x in enumerate(supported_bases_boolean) if x == True] #returns list of length \n",
    "                                    supported_base = base_order[supported_bases_i[0]]\n",
    "\n",
    "                                    #if the Reference Allele is supported\n",
    "                                    if supported_base == ref_allele:\n",
    "                                        alt_allele_frequency = 0.0 #no alternate allele\n",
    "\n",
    "                                    #if an Alternate Allele is supported:\n",
    "                                    if supported_base != ref_allele:\n",
    "                                        alt_allele_frequency = 1.0 #alternate allele\n",
    "\n",
    "                                elif (sum(np.array(QP) == 0) == 2): #2 Alleles supported, make sure 1 is the Reference & confirm the Alternate (2 bases have NO reads supporting)\n",
    "\n",
    "                                    #position of the Reference Allele in [A,C,G,T]\n",
    "                                    ref_allele_base_i = base_order_r[ref_allele] \n",
    "\n",
    "                                    #create a Boolean list for the Bases that have at least 1% of support from reads\n",
    "                                    supported_bases_boolean = list( np.array( QP ) > 0 )\n",
    "\n",
    "                                    #positions of Bases [A,C,G,T] with at least 1% of support from reads\n",
    "                                    supported_bases_i = [i for i,x in enumerate(supported_bases_boolean) if x == True] #returns list of length 2\n",
    "\n",
    "                                    #check that Reference Allele is supported by at least 1% of the reads\n",
    "                                    if ref_allele_base_i in supported_bases_i:\n",
    "\n",
    "                                        #find the Alternate Allele that was supported by at least 1% of the reads\n",
    "                                        for supported_base_i in supported_bases_i:\n",
    "\n",
    "                                            if supported_base_i != ref_allele_base_i: #then must be alternate allele\n",
    "\n",
    "                                                alt_allele = base_order[supported_base_i] #Alternate Allele\n",
    "                                                alt_allele_frequency = float(QP[supported_base_i]) / 100.0 #weighted percentage of reads supporting Alternate Allele\n",
    "                                                break\n",
    "                                ########################################################################\n",
    "                                \n",
    "                                # use only consensus SNPs (where alternate allele frequency >= 90%)\n",
    "                                if alt_allele_frequency >= 0.90:\n",
    "                                \n",
    "                                    # Store good-quality SNV in SNV dictionaries\n",
    "                                    ref_dict[ref_pos] = ref_allele\n",
    "                                    alt_dict[ref_pos] = alt_allele\n",
    "\n",
    "                #construct DataFrame from dictionaries\n",
    "                SNV_DF = pd.DataFrame( index = pd.Series( ref_dict.keys() ) )\n",
    "                SNV_DF['ref'] = pd.Series(ref_dict)\n",
    "                SNV_DF['alt'] = pd.Series(alt_dict)\n",
    "\n",
    "                #make reference-position a column and re-index, sort according to reference position\n",
    "                SNV_DF['pos'] = SNV_DF.index\n",
    "                SNV_DF.sort_values(by = 'pos' , ascending = True , inplace = True)\n",
    "                SNV_DF.reset_index(inplace = True , drop = True)\n",
    "\n",
    "                #pickle DataFrame holding High Quality SNVs from isolate\n",
    "                SNV_DF.to_pickle('/n/data1/hms/dbmi/farhat/Roger/inhost_TB_dynamics_project/rolling_DB_scrape/all_HighQuality_SNVs_found_in_rollingDB/' + isolate_ID + '_SNVs.pkl')\n",
    "                ########################################################################################################################\n",
    "                \n",
    "                #keep track of progress   \n",
    "                isolate_i += 1\n",
    "                if isolate_i % np.ceil(0.05*len(isolate_directories)) == 0:\n",
    "                    print float(isolate_i) / float(len(isolate_directories))\n",
    "        \n",
    "    except OSError: #hit some file that is not another directory with genomic data \n",
    "        continue"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "####################################################################################################################################################################################"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## [2] This section of the notebook constructs a script that iterates over *subsets* of the HQ SNV DataFrames for each isolate and creates a 'catalogue' of all of the HQ SNVs found in rollingDB. All of the relevent information is stored in dictionaries before being transformed into a DataFrame with all of the unique SNVs (indexed by Reference Position) and the number of isolates that contain a SNV at that position."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "####################################################################################################################################################################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import vcf\n",
    "import os\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from pylab import MaxNLocator\n",
    "import ast\n",
    "import itertools\n",
    "import time\n",
    "import sys\n",
    "import itertools\n",
    "import gzip\n",
    "import scipy"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### We are going to split this up into several jobs, each iterating through N isolates. Each job will produce a DF with all SNVs detected in any of the N isolates & corresponding alleles.  Once all jobs have finished, we will merge all DFs together to create one master DF of all SNVs detected in any of the 11,034 isolates in rollingDB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#INPUTS\n",
    "first_isolate_index = int(sys.argv[1]) #INPUT - first isolate index\n",
    "last_isolate_index = int(sys.argv[2]) #INPUT - last isolate index\n",
    "job_number = str(sys.argv[3]) #INPUT - job number"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#list of pickles DataFrame files for SNVs for each isolate in rollingDB\n",
    "SNV_DF_list = os.listdir('/n/data1/hms/dbmi/farhat/Roger/inhost_TB_dynamics_project/rolling_DB_scrape/all_HighQuality_SNVs_found_in_rollingDB')\n",
    "\n",
    "#subset to only the pickled DFs for the isolates that will be searched for SNVs in this job\n",
    "SNV_DF_list = SNV_DF_list[first_isolate_index:last_isolate_index]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def store_SNV_info_from_SNV(SNV_DF_row):\n",
    "\n",
    "    ref_allele , alt_allele , ref_pos = list(SNV_DF_row)\n",
    "\n",
    "    #SNV is not documented at this Reference Position\n",
    "    if ref_pos not in ref_dict.keys():\n",
    "\n",
    "        #store a new row for SNV at this reference position; ref allele, alt allele 1, nan, nan, 1 isolate has polymorphism here (so far)\n",
    "        ref_dict[ref_pos] = ref_allele\n",
    "        alt1_dict[ref_pos] = alt_allele\n",
    "        alt2_dict[ref_pos] = np.nan\n",
    "        alt3_dict[ref_pos] = np.nan\n",
    "        num_isolates_dict[ref_pos] = 1\n",
    "\n",
    "    #a SNV is already documented at this Reference Position\n",
    "    else:\n",
    "\n",
    "        #check to see if specific allele is documented, if not add it\n",
    "        if alt_allele not in [alt1_dict[ref_pos] , alt2_dict[ref_pos] , alt3_dict[ref_pos]]:\n",
    "\n",
    "            #need to document presence of new allele in isolate pool (either going to be alt2 or alt3)\n",
    "            if ( type(alt2_dict[ref_pos])==float ) and ( np.isnan(alt2_dict[ref_pos]) ): #alt2 has not been found yet\n",
    "\n",
    "                alt2_dict[ref_pos] = alt_allele\n",
    "\n",
    "            elif ( type(alt3_dict[ref_pos])==float ) and ( np.isnan(alt3_dict[ref_pos]) ): #alt3 has not been found yet\n",
    "\n",
    "                alt3_dict[ref_pos] = alt_allele\n",
    "\n",
    "        #increment # of isolates that have a polymorphism here\n",
    "        num_isolates_dict[ref_pos] = num_isolates_dict[ref_pos] + 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#create dictionaries to store SNV info for each isolate\n",
    "ref_dict = {}\n",
    "alt1_dict = {}\n",
    "alt2_dict = {}\n",
    "alt3_dict = {}\n",
    "num_isolates_dict = {}\n",
    "\n",
    "#iterate through all DataFrames for each isolate in rollingDB\n",
    "for pickled_SNV_DF in SNV_DF_list:\n",
    "    \n",
    "    #load SNV DF for isolates from pickled object\n",
    "    SNV_DF = pd.read_pickle('/n/data1/hms/dbmi/farhat/Roger/inhost_TB_dynamics_project/rolling_DB_scrape/all_HighQuality_SNVs_found_in_rollingDB/' + pickled_SNV_DF)\n",
    "\n",
    "    #extract important info from SNV df\n",
    "    _ = SNV_DF.apply(lambda row: store_SNV_info_from_SNV(row) , axis = 1 )  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "#construct DataFrame from dictionaries\n",
    "SNV_DF = pd.DataFrame( index = pd.Series( ref_dict.keys() ) )\n",
    "SNV_DF['ref'] = pd.Series(ref_dict)\n",
    "SNV_DF['alt1'] = pd.Series(alt1_dict)\n",
    "SNV_DF['alt2'] = pd.Series(alt2_dict)\n",
    "SNV_DF['alt3'] = pd.Series(alt3_dict)\n",
    "SNV_DF['num_isolates'] = pd.Series(num_isolates_dict)\n",
    "\n",
    "#make reference-position a column and re-index\n",
    "SNV_DF['pos'] = SNV_DF.index\n",
    "\n",
    "#sort according to reference position\n",
    "SNV_DF.sort_values(by = 'pos' , ascending = True , inplace = True)\n",
    "\n",
    "SNV_DF.reset_index(inplace = True , drop = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Output Pickle files for downstream analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "SNV_DF.to_pickle( '/n/data1/hms/dbmi/farhat/Roger/inhost_TB_dynamics_project/rolling_DB_scrape/all_HighQuality_SNVs_found_in_rollingDB_catalogues/seperate_HQ_SNV_DFs/HighQuality_SNVs_found_in_rollingDB_catalogue_' + job_number + '.pkl' )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "####################################################################################################################################################################################"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## [3] This section of the notebook launches the script that was created in the previous section of the notebook."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "####################################################################################################################################################################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "from slurmpy import Slurm\n",
    "import numpy as np\n",
    "import itertools\n",
    "import shutil"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "#list of all SNV DFs for each isolate that was run through MegaPipe in rollingDB\n",
    "SNV_DF_list = os.listdir('/n/data1/hms/dbmi/farhat/Roger/inhost_TB_dynamics_project/rolling_DB_scrape/all_HighQuality_SNVs_found_in_rollingDB')\n",
    "\n",
    "isolates_per_job = 500 #number of isolates each job will catalogue\n",
    "njobs = int( np.ceil( float( len( SNV_DF_list ) ) / float(isolates_per_job) ) ) #number of jobs required if we split for every 500 isolates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "68\n"
     ]
    }
   ],
   "source": [
    "print njobs #number of jobs we will submit"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3299559\n",
      "submitted: Submitted batch job 3299560\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SNV_1 : 3299559\n",
      "SNV_2 : 3299560\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3299561\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SNV_3 : 3299561\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3299563\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SNV_4 : 3299563\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3299568\n",
      "submitted: Submitted batch job 3299578\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SNV_5 : 3299568\n",
      "SNV_6 : 3299578\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3299580\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SNV_7 : 3299580\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3299584\n",
      "submitted: Submitted batch job 3299592\n",
      "submitted: Submitted batch job 3299596\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SNV_8 : 3299584\n",
      "SNV_9 : 3299592\n",
      "SNV_10 : 3299596\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3299597\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SNV_11 : 3299597\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3299611\n",
      "submitted: Submitted batch job 3299613\n",
      "submitted: Submitted batch job 3299614\n",
      "submitted: Submitted batch job 3299615\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SNV_12 : 3299611\n",
      "SNV_13 : 3299613\n",
      "SNV_14 : 3299614\n",
      "SNV_15 : 3299615\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3299616\n",
      "submitted: Submitted batch job 3299617\n",
      "submitted: Submitted batch job 3299618\n",
      "submitted: Submitted batch job 3299619\n",
      "submitted: Submitted batch job 3299620\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SNV_16 : 3299616\n",
      "SNV_17 : 3299617\n",
      "SNV_18 : 3299618\n",
      "SNV_19 : 3299619\n",
      "SNV_20 : 3299620\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3299621\n",
      "submitted: Submitted batch job 3299623\n",
      "submitted: Submitted batch job 3299624\n",
      "submitted: Submitted batch job 3299626\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SNV_21 : 3299621\n",
      "SNV_22 : 3299623\n",
      "SNV_23 : 3299624\n",
      "SNV_24 : 3299626\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3299627\n",
      "submitted: Submitted batch job 3299628\n",
      "submitted: Submitted batch job 3299629\n",
      "submitted: Submitted batch job 3299630\n",
      "submitted: Submitted batch job 3299631\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SNV_25 : 3299627\n",
      "SNV_26 : 3299628\n",
      "SNV_27 : 3299629\n",
      "SNV_28 : 3299630\n",
      "SNV_29 : 3299631\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3299632\n",
      "submitted: Submitted batch job 3299633\n",
      "submitted: Submitted batch job 3299634\n",
      "submitted: Submitted batch job 3299635\n",
      "submitted: Submitted batch job 3299636\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SNV_30 : 3299632\n",
      "SNV_31 : 3299633\n",
      "SNV_32 : 3299634\n",
      "SNV_33 : 3299635\n",
      "SNV_34 : 3299636\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3299637\n",
      "submitted: Submitted batch job 3299638\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SNV_35 : 3299637\n",
      "SNV_36 : 3299638\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3299639\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SNV_37 : 3299639\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3299640\n",
      "submitted: Submitted batch job 3299641\n",
      "submitted: Submitted batch job 3299642\n",
      "submitted: Submitted batch job 3299643\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SNV_38 : 3299640\n",
      "SNV_39 : 3299641\n",
      "SNV_40 : 3299642\n",
      "SNV_41 : 3299643\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3299644\n",
      "submitted: Submitted batch job 3299652\n",
      "submitted: Submitted batch job 3299654\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SNV_42 : 3299644\n",
      "SNV_43 : 3299652\n",
      "SNV_44 : 3299654\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3299655\n",
      "submitted: Submitted batch job 3299656\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SNV_45 : 3299655\n",
      "SNV_46 : 3299656\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3299657\n",
      "submitted: Submitted batch job 3299661\n",
      "submitted: Submitted batch job 3299662\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SNV_47 : 3299657\n",
      "SNV_48 : 3299661\n",
      "SNV_49 : 3299662\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3299663\n",
      "submitted: Submitted batch job 3299664\n",
      "submitted: Submitted batch job 3299666\n",
      "submitted: Submitted batch job 3299667\n",
      "submitted: Submitted batch job 3299668\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SNV_50 : 3299663\n",
      "SNV_51 : 3299664\n",
      "SNV_52 : 3299666\n",
      "SNV_53 : 3299667\n",
      "SNV_54 : 3299668\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3299669\n",
      "submitted: Submitted batch job 3299670\n",
      "submitted: Submitted batch job 3299671\n",
      "submitted: Submitted batch job 3299672\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SNV_55 : 3299669\n",
      "SNV_56 : 3299670\n",
      "SNV_57 : 3299671\n",
      "SNV_58 : 3299672\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3299674\n",
      "submitted: Submitted batch job 3299676\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SNV_59 : 3299674\n",
      "SNV_60 : 3299676\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3299677\n",
      "submitted: Submitted batch job 3299678\n",
      "submitted: Submitted batch job 3299679\n",
      "submitted: Submitted batch job 3299680\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SNV_61 : 3299677\n",
      "SNV_62 : 3299678\n",
      "SNV_63 : 3299679\n",
      "SNV_64 : 3299680\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3299681\n",
      "submitted: Submitted batch job 3299682\n",
      "submitted: Submitted batch job 3299683\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SNV_65 : 3299681\n",
      "SNV_66 : 3299682\n",
      "SNV_67 : 3299683\n",
      "SNV_68 : 3299684\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3299684\n"
     ]
    }
   ],
   "source": [
    "first_isolate_index = 0\n",
    "max_isolate_index = int( len( SNV_DF_list ) )\n",
    "\n",
    "#run each job using a difference set of indices for isolates that we will catalogue SNVs from\n",
    "for job_number in np.arange(1 , njobs+1):\n",
    "\n",
    "    last_isolate_index = first_isolate_index + isolates_per_job\n",
    "    if last_isolate_index > max_isolate_index: #watch out for the last job\n",
    "        last_isolate_index = max_isolate_index\n",
    "    \n",
    "    rollingDB_SNV_cataloging_job = 'python /n/data1/hms/dbmi/farhat/Roger/inhost_TB_dynamics_project/python_scripts/rollingDB_HQ_SNV_catalogue_scrape.py ' + str(first_isolate_index) + ' ' + str(last_isolate_index) + ' ' + str(job_number)\n",
    "    \n",
    "    #directory where you want output + error files\n",
    "    os.chdir('/n/data1/hms/dbmi/farhat/Roger/inhost_TB_dynamics_project/rolling_DB_scrape/all_HighQuality_SNVs_found_in_rollingDB_catalogues/SNV_catalogue_scrape_jobs/')\n",
    "\n",
    "    job_name = 'SNV_' + str(job_number)\n",
    "\n",
    "    s = Slurm(job_name , {'partition':'short' , 'n':'1' , 't':'0-01:00:00' , 'mem-per-cpu':'4G' , 'mail-type':'FAIL' , 'mail-user':'roger_vargas@g.harvard.edu'})\n",
    "\n",
    "    #submits the job\n",
    "    job_id = s.run(rollingDB_SNV_cataloging_job)\n",
    "\n",
    "    print job_name  + ' : ' +  str(job_id)\n",
    "    \n",
    "    #setup isolate indices for next job\n",
    "    first_isolate_index = last_isolate_index"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "####################################################################################################################################################################################"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## [4] The purpose of this section of the notebook is generate the master catalogue of (high-quality) SNV calls made in rollingDB by merging the catalogues of SNVs found in seperate batches of isolates."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "####################################################################################################################################################################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import ast\n",
    "import itertools\n",
    "import time\n",
    "import sys\n",
    "import itertools"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "#INPUTS\n",
    "njobs = 68 #number of seperate SNV catalogues for subsets of isolates in rollingDB"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Load in SNV catalogue(s) (High Quality SNV calls across all isolates in RollingDB)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We're going to iterate through each DataFrame (catalogue of SNVs detected within each batch of isolates) and create a master DF of all SNVs detected within all the isolates in rollingDB."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "def convert_SNV_DF_rows_to_DICT_elments(row):\n",
    "    \n",
    "    '''\n",
    "    This function converts the row of a SNV catalogue DF to a\n",
    "    list with two elements: (1) The Reference Position of an SNV detected within a subset of isolates\n",
    "    and (2) another list containing the Reference Allele, all Alternate Alleles, and Number of Isolates\n",
    "    with an SNV at this Reference Position\n",
    "    '''\n",
    "    \n",
    "    #create set of alternate alleles present for this SNV\n",
    "    alt_alleles_SNV_i = set( [row['alt1'] , row['alt2'] , row['alt3']] )\n",
    "    #purge nan values\n",
    "    alt_alleles_SNV_i = {x for x in alt_alleles_SNV_i if x==x}\n",
    "        \n",
    "    return [ row['pos'] , [row['ref'] , alt_alleles_SNV_i , row['num_isolates']] ]\n",
    "\n",
    "def convert_SNV_DF_to_DICT(SNV_DF):\n",
    "    \n",
    "    '''\n",
    "    This function takes an SNV DF catalogue for a subset of isolates and converts it into \n",
    "    an SNV DICT catalogue using the function above by applying it to each row.\n",
    "    '''\n",
    "    \n",
    "    SNV_dict = SNV_DF.apply( convert_SNV_DF_rows_to_DICT_elments , axis = 1 )\n",
    "    return dict(list( SNV_dict.values ))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Convert SNV catalogue(s) DFs to SNV catalogue DICTs and store SNP catalogue info across jobs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1\n",
      "2\n",
      "3\n",
      "4\n",
      "5\n",
      "6\n",
      "7\n",
      "8\n",
      "9\n",
      "10\n",
      "11\n",
      "12\n",
      "13\n",
      "14\n",
      "15\n",
      "16\n",
      "17\n",
      "18\n",
      "19\n",
      "20\n",
      "21\n",
      "22\n",
      "23\n",
      "24\n",
      "25\n",
      "26\n",
      "27\n",
      "28\n",
      "29\n",
      "30\n",
      "31\n",
      "32\n",
      "33\n",
      "34\n",
      "35\n",
      "36\n",
      "37\n",
      "38\n",
      "39\n",
      "40\n",
      "41\n",
      "42\n",
      "43\n",
      "44\n",
      "45\n",
      "46\n",
      "47\n",
      "48\n",
      "49\n",
      "50\n",
      "51\n",
      "52\n",
      "53\n",
      "54\n",
      "55\n",
      "56\n",
      "57\n",
      "58\n",
      "59\n",
      "60\n",
      "61\n",
      "62\n",
      "63\n",
      "64\n",
      "65\n",
      "66\n",
      "67\n",
      "68\n"
     ]
    }
   ],
   "source": [
    "#master set of all reference positions detected within rollingDB (faster to query a set)\n",
    "SNV_set_Ref_Position = set([])\n",
    "\n",
    "#master dictionaries of all SNVs detected within rollingDB\n",
    "SNV_dict_Ref_Allele = {} #keys - reference position; values - ref_allele (string)\n",
    "SNV_dict_Alt_Allele = {} #keys - reference position; values - alt_alleles (set)\n",
    "SNV_dict_Num_Isolates = {} #keys - reference position; values - num_isolates (int)\n",
    "\n",
    "for job_num in np.arange(1 , njobs+1):\n",
    "    \n",
    "    #load the SNV DataFrame holding SNVs from the batch of isolates\n",
    "    SNV_DF_subset = pd.read_pickle( '/n/data1/hms/dbmi/farhat/Roger/inhost_TB_dynamics_project/rolling_DB_scrape/all_HighQuality_SNVs_found_in_rollingDB_catalogues/seperate_HQ_SNV_DFs/HighQuality_SNVs_found_in_rollingDB_catalogue_' + str(job_num) + '.pkl' )\n",
    "    \n",
    "    #convert SNV DataFrame to an SNV Dict\n",
    "    SNV_dict_subset = convert_SNV_DF_to_DICT(SNV_DF_subset)\n",
    "    \n",
    "    #iterate through each SNV in Dict\n",
    "    for ref_pos in SNV_dict_subset.keys():\n",
    "        \n",
    "        #update master SNV catalogue dict with SNV info\n",
    "        ref , alt_alleles_SNV_i , num_isolates = SNV_dict_subset[ref_pos]\n",
    "    \n",
    "        #[A] reference position not stored in master SNV dictionaries\n",
    "        ##if ref_pos not in SNV_dict_Ref_Allele.keys():\n",
    "        if ref_pos not in SNV_set_Ref_Position:\n",
    "            \n",
    "            SNV_set_Ref_Position.add(ref_pos)\n",
    "            \n",
    "            SNV_dict_Ref_Allele[ref_pos] = ref\n",
    "            SNV_dict_Alt_Allele[ref_pos] = alt_alleles_SNV_i\n",
    "            SNV_dict_Num_Isolates[ref_pos] = num_isolates\n",
    "        \n",
    "        #[B] reference position already stored in master SNV dictionaries\n",
    "        else:\n",
    "            \n",
    "            #update SNV info with new SNV info (ref_allele stays the same)\n",
    "            SNV_dict_Alt_Allele[ref_pos] = SNV_dict_Alt_Allele[ref_pos].union(alt_alleles_SNV_i) #store any new alleles detected\n",
    "            SNV_dict_Num_Isolates[ref_pos] = SNV_dict_Num_Isolates[ref_pos] + num_isolates #add number of isolates\n",
    "            \n",
    "    #keep track of progress\n",
    "    print job_num"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Convert master SNV dict to SNV df "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "#convert dictionaries into pandas series objects\n",
    "SNV_series_Ref_Allele = pd.Series(SNV_dict_Ref_Allele , name = 'ref')\n",
    "SNV_series_Alt_Allele = pd.Series(SNV_dict_Alt_Allele , name = 'alt')\n",
    "SNV_series_Num_Isolates = pd.Series(SNV_dict_Num_Isolates , name = 'num_isolates')\n",
    "\n",
    "#construct DataFrame from Series objects\n",
    "SNV_DF = pd.DataFrame([SNV_series_Ref_Allele , SNV_series_Alt_Allele , SNV_series_Num_Isolates]).T\n",
    "\n",
    "#convert Alternate Allele(s) column to list\n",
    "SNV_DF['alt'] = SNV_DF.alt.apply(list)\n",
    "\n",
    "#create a column for reference position\n",
    "SNV_DF['pos'] = list(SNV_DF.index)\n",
    "\n",
    "#re-index DataFrame\n",
    "SNV_DF.reset_index(inplace = True , drop = True)\n",
    "\n",
    "#store master SNV DataFrame as a pickled object\n",
    "SNV_DF.to_pickle( '/n/data1/hms/dbmi/farhat/Roger/inhost_TB_dynamics_project/rolling_DB_scrape/all_HighQuality_SNVs_found_in_rollingDB_catalogues/all_HighQuality_SNVs_found_in_rollingDB_catalogue.pkl' )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Use to load pickle"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "SNV_DF = pd.read_pickle('/n/data1/hms/dbmi/farhat/Roger/inhost_TB_dynamics_project/rolling_DB_scrape/all_HighQuality_SNVs_found_in_rollingDB_catalogues/all_HighQuality_SNVs_found_in_rollingDB_catalogue.pkl')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>num_isolates</th>\n",
       "      <th>pos</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A</td>\n",
       "      <td>[C]</td>\n",
       "      <td>2</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>G</td>\n",
       "      <td>[C]</td>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A</td>\n",
       "      <td>[C]</td>\n",
       "      <td>8</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>G</td>\n",
       "      <td>[C]</td>\n",
       "      <td>1</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>G</td>\n",
       "      <td>[A]</td>\n",
       "      <td>1</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  ref  alt num_isolates  pos\n",
       "0   A  [C]            2    8\n",
       "1   G  [C]            1   10\n",
       "2   A  [C]            8   11\n",
       "3   G  [C]            1   16\n",
       "4   G  [A]            1   23"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SNV_DF.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>num_isolates</th>\n",
       "      <th>pos</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>878239</th>\n",
       "      <td>C</td>\n",
       "      <td>[G]</td>\n",
       "      <td>2</td>\n",
       "      <td>4411502</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>878240</th>\n",
       "      <td>C</td>\n",
       "      <td>[A]</td>\n",
       "      <td>1</td>\n",
       "      <td>4411511</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>878241</th>\n",
       "      <td>A</td>\n",
       "      <td>[G]</td>\n",
       "      <td>1</td>\n",
       "      <td>4411516</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>878242</th>\n",
       "      <td>C</td>\n",
       "      <td>[T]</td>\n",
       "      <td>4</td>\n",
       "      <td>4411518</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>878243</th>\n",
       "      <td>G</td>\n",
       "      <td>[A]</td>\n",
       "      <td>3</td>\n",
       "      <td>4411522</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       ref  alt num_isolates      pos\n",
       "878239   C  [G]            2  4411502\n",
       "878240   C  [A]            1  4411511\n",
       "878241   A  [G]            1  4411516\n",
       "878242   C  [T]            4  4411518\n",
       "878243   G  [A]            3  4411522"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SNV_DF.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(878244, 4)"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.shape(SNV_DF)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "####################################################################################################################################################################################"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## [5] The purpose of this section of the notebook is to analyze the catalogue of (high-quality) SNV calls made in rollingDB. For population genetic analyses, we will retain only SNVs that are present in at least 2 isolates. We then construct a skeleton genotype DF (rows = Reference Position , columns = isolates) for all of the remaining SNVs."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "####################################################################################################################################################################################"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Construct a series for the isolate annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "HQ_SNV_per_isolate_files = os.listdir('/n/data1/hms/dbmi/farhat/Roger/inhost_TB_dynamics_project/rolling_DB_scrape/all_HighQuality_SNVs_found_in_rollingDB/')\n",
    "isolate_ids_list = [isolate_id.split('_SNV')[0] for isolate_id in HQ_SNV_per_isolate_files]\n",
    "isolate_id_series = pd.Series(isolate_ids_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    SAMEA3558733\n",
       "1    SAMN05575260\n",
       "2    SAMN03648641\n",
       "3    SAMN03647419\n",
       "4    SAMEA3671418\n",
       "dtype: object"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "isolate_id_series.head() #index = column indices of genotype matrix, values isolate IDs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(33875,)"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.shape(isolate_id_series)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Remove 2 bad isolates from isolate annotation"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Remove:** Delete ID for isolate **IDR1400002533** since Full GZipped VCF looks corrupt and remove from list accordingly!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "isolate_id_series = isolate_id_series[isolate_id_series != 'IDR1400002533']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Remove:** Delete ID for **Peru4994** since F2 calculations keeps failing for this isolate (lineage-defining SNP depths df is empty) **Maybe don't need to remove this isolate?**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "isolate_id_series = isolate_id_series[isolate_id_series != 'Peru4994']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "isolate_id_series.reset_index(inplace = True, drop = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    SAMEA3558733\n",
       "1    SAMN05575260\n",
       "2    SAMN03648641\n",
       "3    SAMN03647419\n",
       "4    SAMEA3671418\n",
       "dtype: object"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "isolate_id_series.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "33873"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(isolate_id_series)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "isolate_id_series.to_pickle( '/n/data1/hms/dbmi/farhat/Roger/inhost_TB_dynamics_project/rolling_DB_scrape/Genotypes/genotypes_isolate_annotation.pkl' )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load in catologue of SNVs detected within rollingDB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "#SNV data for all isoaltes in rollingDB \n",
    "SNV_DF = pd.read_pickle( '/n/data1/hms/dbmi/farhat/Roger/inhost_TB_dynamics_project/rolling_DB_scrape/all_HighQuality_SNVs_found_in_rollingDB_catalogues/all_HighQuality_SNVs_found_in_rollingDB_catalogue.pkl' )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>num_isolates</th>\n",
       "      <th>pos</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A</td>\n",
       "      <td>[C]</td>\n",
       "      <td>2</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>G</td>\n",
       "      <td>[C]</td>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A</td>\n",
       "      <td>[C]</td>\n",
       "      <td>8</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>G</td>\n",
       "      <td>[C]</td>\n",
       "      <td>1</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>G</td>\n",
       "      <td>[A]</td>\n",
       "      <td>1</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  ref  alt num_isolates  pos\n",
       "0   A  [C]            2    8\n",
       "1   G  [C]            1   10\n",
       "2   A  [C]            8   11\n",
       "3   G  [C]            1   16\n",
       "4   G  [A]            1   23"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SNV_DF.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(878244, 4)"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.shape(SNV_DF)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### How many SNPs are not biallelic?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "61918"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#SNPs that are not biallelic will have more than 1 alternate allele\n",
    "np.sum( SNV_DF.alt.apply(len) > 1 )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>num_isolates</th>\n",
       "      <th>pos</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>A</td>\n",
       "      <td>[C, G]</td>\n",
       "      <td>10</td>\n",
       "      <td>31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>G</td>\n",
       "      <td>[A, T]</td>\n",
       "      <td>3</td>\n",
       "      <td>67</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>C</td>\n",
       "      <td>[A, T]</td>\n",
       "      <td>15</td>\n",
       "      <td>108</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>C</td>\n",
       "      <td>[A, T]</td>\n",
       "      <td>2</td>\n",
       "      <td>116</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>T</td>\n",
       "      <td>[C, G]</td>\n",
       "      <td>4</td>\n",
       "      <td>147</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>G</td>\n",
       "      <td>[A, C]</td>\n",
       "      <td>2</td>\n",
       "      <td>186</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>G</td>\n",
       "      <td>[C, T]</td>\n",
       "      <td>3</td>\n",
       "      <td>195</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>G</td>\n",
       "      <td>[A, C]</td>\n",
       "      <td>4</td>\n",
       "      <td>203</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>G</td>\n",
       "      <td>[A, C]</td>\n",
       "      <td>2</td>\n",
       "      <td>234</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>G</td>\n",
       "      <td>[A, T]</td>\n",
       "      <td>2</td>\n",
       "      <td>236</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   ref     alt num_isolates  pos\n",
       "7    A  [C, G]           10   31\n",
       "12   G  [A, T]            3   67\n",
       "27   C  [A, T]           15  108\n",
       "29   C  [A, T]            2  116\n",
       "34   T  [C, G]            4  147\n",
       "43   G  [A, C]            2  186\n",
       "46   G  [C, T]            3  195\n",
       "49   G  [A, C]            4  203\n",
       "53   G  [A, C]            2  234\n",
       "54   G  [A, T]            2  236"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SNV_DF[SNV_DF.alt.apply(len) > 1].head(n = 10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.style.use('ggplot')\n",
    "plt.rcParams['lines.linewidth']=3\n",
    "plt.rcParams['axes.facecolor']='1.0'\n",
    "plt.rcParams['xtick.color']='black'\n",
    "plt.rcParams['axes.grid']=True\n",
    "plt.rcParams['axes.edgecolor']='black'\n",
    "plt.rcParams['grid.color']= '1.0'\n",
    "plt.rcParams.update({'font.size': 12})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAhMAAAFxCAYAAADJdyroAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMi4zLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvIxREBQAAIABJREFUeJzt3XtcFPX+P/DXQqiwoYIUCRLCcicVETEzUjNR8JYBapmmUVZmRwStQ3Y3IQ0zTtYpM/FnJ40UkWMiYV5TUSGJEIJQRDRviXiEhVV3md8fPtivK7CsDHuD1/Px4PFwP8PM5z07IK/9zGdmJIIgCCAiIiJqIwtjF0BERETmjWGCiIiIRGGYICIiIlEYJoiIiEgUhgkiIiIShWGCiIiIRGGYILM2ZMgQODs749ChQ8YuBSqVCjExMfD19YWzszM++OADvfcZExMDZ2dnrFixQu99mQpTOua6OHToEJydnTFkyBBjl9KuGo9DVlZWs8tXrFgBZ2dnxMTEGLgyMgaGCRKt8T+VwMBA1NfXAwCOHz8OZ2dnODs7G7k6w9m+fTs2bdoES0tLPP/883j44YebfE/jHxZfX18jVGiYP2yNfdz5NXr0aL312Zrjx4/j6aefhr+/P2QyGYYMGYJZs2bhf//7X5Oa33nnHfV677zzjsYfxDNnzmjsk4uLCwYNGoSYmBhcunSp2b579+6N6OhoTJs2rU21R0ZGNvt+tvRH3FQEBgYiOjoaw4cPN3YpZAD3GLsA6jguXryI9evX46WXXjJ2KaLcvHkTVlZWd71eeXk5AODxxx/HkiVL2rsss9OlSxfMmDFD/bp3795GqaO+vh4zZszApUuXEBoaivvvvx9nzpzBL7/8Arlcjh49emh8/3fffYe5c+figQce0Lrdxn3bsWMHNm3ahFOnTiEjI6PJ97m5ubXLKNXDDz8Mf39/9WtXV1fR29SnkSNHYuTIkcYugwyEIxPUbiQSCb744gv16MSdGj9RnTlzBkDTYdDU1FQ4OzvjiSeewHvvvQdPT0+MGDECx48fx/Lly+Hj44OhQ4di3759TbZdWFiI0aNHw8vLC9HR0bhy5Yp62dGjRxEZGQk/Pz8EBgYiNjZWvfz2T5rr169HYGAgnn766WbrP3PmDObMmYOBAwfCz88PkZGROHbsmHpfPv74YwBAWloanJ2dkZqa2up7JggCEhMTERQUBDc3NwQEBOCZZ55R11dVVYWFCxdi8ODB8Pb2xvjx47Fnz54Wt7d//36EhobCx8cHrq6uCA4ORlJSEoBbn76joqIAAGfPntUYOaqvr0dCQgKGDRsGDw8PjBkzRuOT7/79+zF27Fh4eHjAx8cHY8aMQWZmptZ969atGz744AP11yuvvALgVlibNm0aAgIC0LdvX/j6+mLWrFn466+/1OueO3cO8+fPR3BwMNzd3TF8+HD89ttvGts/fvw4wsLC4OHhgRkzZuDq1avN1nHixAlcunQJvr6+SElJwbJly7Bhwwb8/vvvuO+++zS+VyKRQKFQYNWqVVr3DQDi4+Px0UcfISEhAQCQl5enHum43Z2jQTdu3MCiRYsQEBAANzc3BAUFYdasWa32N3bsWI33s3F06+bNm/jss8/w2GOPwcPDA8OHD8fq1avR0NAAoPnTDXf+LjaOLq5atQqhoaHNvqfr1q1DUFAQ+vXrh3//+9+t1tvS7/eTTz6Jd999F76+vhg0aBC2bNnS6rbI9DFMULsZP348Ll++jHXr1onaTklJCfLz8+Hp6YmysjJERUVh+/btCAwMRGVlJWJjY5us88knn6Bfv36wt7dHVlYW3njjDfW2pk6dit9//x0jRoyAu7s7UlNT8dJLL+HOO8kvW7YMI0eORFBQUJPt19XVYcqUKdi+fTvc3d0xbNgw5OTkYMqUKaioqEBgYCAGDhwIAPD09ER0dDS8vLxa3ddffvkFq1atgqWlJaZNm4YhQ4agpKQEcrkcDQ0NmDVrFjZu3Ah7e3uEhoaisLAQM2fORG5ubrPbO3/+POzt7TFx4kRERERALpdj5cqVyMjIQO/evREeHg4AuPfeexEdHY3o6GgAQFxcHD7//HPY2toiPDwc586dwwsvvKCel7BgwQIUFxcjPDwc4eHhsLCwQGlpqdZ9UygUeOedd9Rf6enpAICGhgZcunQJw4cPxzPPPIMHH3wQO3fuxOuvvw7gVrCZMmUKNm/ejK5duyIiIgI9evTAxYsXNba/fPlyeHl5oWvXrti9ezdWr17dbB0ODg6wsLDAH3/8gUmTJiEhIQH79u2DlZVVkxEoW1tbDB8+HBs2bMC5c+e07h8AXL9+XX0sunXrBqlU2uo6mzdvxoYNG2Bvb49p06ahf//+LR7P22VlZWm8n9XV1QBu/dx+9NFHkMvlmDRpEq5cuYL3338fn3/+eavbvNPKlSvh6+vb5D09dOgQFi9ejIsXL2L48OFIS0vT6f1pTm5uLvLy8hAQEIALFy7gjTfeQE1NTZu2RaaDpzmo3UycOBGlpaX497//jUGDBrV5OzY2Nvj++++Rn5+PqKgoXLt2DRkZGXBycoK3tzcuXLiAqqoq9OrVS73OokWL8MILL6CoqAihoaHIzMyEXC7H+vXrcePGDfj7+8PBwQEODg44duwYDh06hJMnT6Jr167qbXz11Vd49NFHm63p559/RmVlJVxdXbF582ZIJBJER0cjKysLGzduRHx8PI4dO4b8/HwEBAToPKytVCoBAH379sWECRPg5eWFXr16QRAEFBQU4NixY5BKpUhPT4eNjQ3s7e2xZs0arFu3DoMHD26yvaioKDg4OKCwsBDV1dVwdXXF1atXcfDgQUyaNAmzZ89GZmYmevbsqa6xqqoKGRkZsLCwQFBQECwsLODt7Y2cnBx8++23eOSRR6BUKtG1a1f1qIe7u3uTMHanGzdu4JtvvtGobfLkyejatSvWrFmDnTt34u+//4aPjw+OHz+OnJwcNDQ0YNeuXTh16hQcHR2RnZ0Na2trALc+gd8uLi4Or7zyCpKSkrBy5UocP3682Tp69+6Nd999FwkJCcjLy0NeXh4+//xz9O3bFxs3bsSDDz7YZLsTJ07Ev/71L3Tp0qXF/fPz89N4vXDhQtxzT+v/pTbuh4+PD5566il4enrC1ta21fUOHz6Mw4cPq1+/+OKL6NmzJ/7f//t/AIBVq1Zh6NCh+Omnn/D8888jJSUFr732WqvbvXMfmntPG0cPpkyZghUrVqC6uhoDBw5Uj37cjZ49eyItLQ2WlpaQyWSoq6tDeXk5BgwYcNfbItPBMEHtxsLCAgsWLMArr7yClJSUVr9fpVI12+7i4gJra2t0795d3SaTyWBpaal+XVdXpxEmPD09AQAeHh7qtvPnz6uHcfPz85Gfn6/Rz6lTp+Dj46N+3dwf50Znz55V1yGRSDT6un14/m4NHz4czz33HNLS0tSnIAICArB27Vp17U5OTrCxsdHos7GeO/3zn//Ed99916S9qqqqxRoa+2loaGhy3CoqKgDc+vS7ZMkS9XwYOzs7LF26FJMmTWpxu927d8cff/zRpP3IkSOIiopqcvyvX7+O2tpadT0+Pj7qIAGgyShC4/yBxp+Turq6Fmt54YUXMG3aNBw6dAiHDh3Chg0bUFFRga+//rrJ/JZBgwbh8ccfR2pqKkaNGtXiNmfMmAFra2s4OjoiJCREYz6DNlFRUcjJycFPP/2EjIwMSCQShISE4JtvvlEf5+a89957ePHFFzXaLl++rN7vxt8BmUwG4NYcphs3bjTZTku/d0DL7+mFCxc0tm1nZwc7O7sWJ51q4+npiW7dugG49cGhpqYGcrn8rrdDpoWnOahdTZgwAb6+vti2bVuTZY1/GBqHNFsaJrewaPpjeXuQaE5ZWRmAW+fHG/Xu3RsuLi4AgDlz5uCvv/5Sf+Xk5DS5uuD2UYo79enTB8CtSZaNn8hPnjwJAKKuWFGpVFi6dCn++OMPHDx4EJGRkfjtt9+wYcMGde3nzp1Tz0Np7LOxnjs1vu8rV67EmTNnMHPmTABQ19z4Pt7+ibKxny5duqCwsFD9HlVUVGDNmjUAbk2mO3jwIAoLC7F69WpUV1dj2bJlbdrn7du3Q6VSYeTIkThx4gR+/PFH9TJBENT1lJSUaMy/aRzFadQ4CtAY7lpSXV2NoqIi3HvvvQgNDcV7772nvrKitra22XUWLlyIGzduaL1iIj4+Hu+++y5efvllnYMEcOsYfPnllygtLcW+ffsQEhKC/fv3tzoHpTm9evVSB5DGn/3GicCOjo7o0qWLennjvpaUlLS4vZbe08bJqI0/f9XV1erTLHfr9tGb1o4dmQ+OTFC7kkgkWLBgAebMmdNk2UMPPYTc3Fy89dZbkMlk+Omnn9qt348//hjFxcXqc/xhYWGQSqWYPn06NmzYgG+++QaVlZWwt7dHWVkZ8vLyWvx035wnnngCLi4uqKioQFRUFOzs7LBjxw5069atzZf8Abcm7S1YsACDBg1Cz549kZeXBwDo0aMHBgwYgIEDByI/Px+TJ0+Gl5eX+pPsc8891+z2HBwccO3aNaxduxb79+/Hjh07NJY7OTkBuDVqs3DhQri5ueHVV1/FhAkTsG3bNowfPx4hISGorq7G0aNHMWPGDMTFxWHMmDFwcXGBk5OT+lz57SNHzWmcM9HI2toa8fHx6kmP+fn5eOuttzSG7gFg1KhRcHNzw6lTpzBmzBg8/PDDOHnyJObMmYMxY8bcxbt7y8WLFxEaGop+/fqpR6K2b98OAHjssceaXWfAgAF44okn8PPPP991f63ZunUrvvjiC/Tv3x9SqVT9x72197M5EokEM2fOxJdffolXX30VI0eORHZ2NgCoJ3U2Bp3du3fj/fffx+7du++6nyeffBIbN27EDz/8gOvXr6OkpKRJuKPOjSMT1O7Cw8Ob/aT24YcfwtfXF0VFRTh//jymTp3abn3GxsaisLAQVVVVCA0NxfLlywHc+o/0+++/x5AhQ3D48GFkZGSgtrYWr7766l1t38bGBqmpqQgPD8eJEydw4MABDB06FKmpqXBzc2tz3Q888ADc3Nxw4MABbNiwQX0Z4/Tp02FhYYF169Zh6tSpuHz5MrKysvDQQw8hJSUFwcHBzW4vKSkJHh4eKCsrQ21tLZ599lmN5S4uLnj55ZfRvXt3bNy4EWlpaQBuzbyfN28eJBIJNm3ahLy8PAwaNAgjRowAAISEhODkyZPYvHkzcnNzMXToUPVVIi1pnDPR+LV+/XoAwOzZszF27Fhcv34dhw8fxj/+8Q+N9aytrZGamoqIiAjU19dj8+bNuHz5MhwdHdvyFsPR0REzZ87E9evXkZWVhW3btsHZ2Rnvv/8+Jk+e3OJ6CxcubFN/rZHJZLC3t8fu3bvx/fffw8rKCvPnz2/zfTjeeOMNLFq0CNbW1khPT0fPnj3x9ttvq3/GH3vsMURHR6Nbt27IysrC7Nmz77qPRx99FEuWLMH999+PPXv2YNy4cZ3qHjLUOonQ2iwqIiIiIi04MkFERESiMEwQERGRKAwTREREJArDBBEREYnCMEFERESidPgwwZuiEBER6VeHDxNERESkXwwTREREJArDBBEREYnCMEFERESimEWYOHToEKZMmYLIyMgmDy4iIiIi4zL5p4YqFAp89dVX+M9//oMuXboYuxwiIiK6g8mPTOTl5aFbt26YNWsWoqOjcenSJWOXRERERLcxWJhISUlBWFgY3NzcEBMTo7Gsuroa0dHR8PDwQHBwMNLT09XLLl++jIqKCqxbtw7PPPMMVqxYYaiSiYiISAcGCxOOjo6YP38+pk6d2mTZ4sWLYWVlhYKCAqxatQrx8fEoLS0FAHTv3h2DBw9Gly5d8Oijj6KsrMxQJRMREZEODBYmwsPDMXbsWNjZ2Wm019XVITMzE4sWLYJUKkVwcDBGjx6NtLQ0AEBAQADKysogCAKKiorg6upqqJKJiIhIB0afgFleXg5LS0vIZDJ1m7+/P3JycgAA9vb2CAsLQ0REBCwsLHiag4iIyMQYPUzI5XLY2tpqtNna2kIul6tfz5o1C7NmzTJwZURERKQLo1/NIZVKUVNTo9FWU1MDqVRqpIq0U6lUbVpGRETUURl9ZMLd3R0qlQrl5eVwd3cHABQXF8Pb29vIlTXP0tISL774IsrLyzXa3d3d8fXXXxupKiIiIuMxWJhQKpVQKpVoaGiASqWCQqHAPffcAxsbG4SFhSEpKQlJSUkoKipCdnY2MjIyDFXaXSsvL0dJSYmxyyAiIjIJBjvNkZycDJlMhlWrVmHLli2QyWRITk4GACQkJEChUKB///6YO3cuEhMTTXZkgoiIiDRJBEEQjF2EPkkkErT3Lo4aNarJyISPjw927drVrv0QERGZA6NPwCQiIiLzxjBBREREojBMEBERkSgME0RERCQKwwQRERGJwjBBREREojBMEBERkSgME0RERCQKwwQRERGJwjBBREREojBMEBERkSgME0RERCQKwwQRERGJwjBBREREojBMEBERkSgME0RERCQKwwQRERGJwjBBREREojBMEBERkSgME0RERCQKwwQRERGJwjBBREREojBMEBERkSgME0RERCQKwwQRERGJwjBBREREojBMEBERkSgME0RERCQKwwQRERGJwjBBREREojBMEBERkSgME0RERCSKyYeJM2fOoF+/foiMjERkZCSqqqqMXRIRERHd5h5jF6CLhx9+GF9//bWxyyAiIqJmmPzIBADk5eVh8uTJSExMhCAIxi6HiIiIbmOwMJGSkoKwsDC4ubkhJiZGY1l1dTWio6Ph4eGB4OBgpKenq5fdf//9OHjwILZs2YKqqipkZmYaqmQiIiLSgcHChKOjI+bPn4+pU6c2WbZ48WJYWVmhoKAAq1atQnx8PEpLSwEAXbt2hY2NDSQSCcLCwlBcXGyokomIiEgHBgsT4eHhGDt2LOzs7DTa6+rqkJmZiUWLFkEqlSI4OBijR49GWloaAKCmpkb9vUePHkXfvn0NVTIRERHpwOhzJsrLy2FpaQmZTKZu8/f3V49M5OTkYOzYsZg8eTLOnz+PyZMnG6tUIiIiaobRr+aQy+WwtbXVaLO1tYVcLgcAhIaGIjQ01BilERERkQ6MPjIhlUo1TmUAt05tSKVSI1VEREREd8PoYcLd3R0qlQrl5eXqtuLiYnh7exuxKiIiItKVwcKEUqmEQqFAQ0MDVCoVFAoFlEolbGxsEBYWhqSkJNTV1SE3NxfZ2dmIiIgwVGlEREQkgsHCRHJyMmQyGVatWoUtW7ZAJpMhOTkZAJCQkACFQoH+/ftj7ty5SExM5MgEERGRmZAIHfyWkhKJpN3vmjlq1CiUlJRotPn4+GDXrl3t2g8REZE5MPqcCSIiIjJvDBNEREQkCsMEERERicIwQURERKIwTBAREZEoDBNEREQkCsMEERERicIwQURERKIwTBAREZEoDBNEREQkCsMEERERicIwQURERKIwTBAREZEoDBNEREQkCsMEERERicIwQURERKIwTBAREZEoDBNEREQkCsMEERERicIwQURERKIwTBAREZEoDBNEREQkCsMEERERicIwQURERKIwTBAREZEoDBNEREQkCsMEERERicIwQURERKIwTBAREZEoDBNEREQkCsMEERERicIwQURERKIwTBAREZEoOoWJrKwsfPrppxAEARkZGRg2bBjGjRuHoqIifdentnXrVvTr189g/REREZFudAoTn376KXbu3AkAeOutt3D69GkUFBTggw8+0GtxjRoaGrB9+3Y4OTkZpD8iIiLSnU5h4vTp0/D19UV5eTmqq6uxevVqPP744ygsLNR3fQCA9PR0jBs3DhYWPCtDRERkanT666xUKqFUKlFaWgqJRIJHHnkETk5OuHHjhs4dpaSkICwsDG5uboiJidFYVl1djejoaHh4eCA4OBjp6enqZSqVCtu2bcPEiRN17ouIiIgM5x5dvsnDwwNbt27Fjh070LdvX/Ts2RN//fUXHB0dde7I0dER8+fPx969e6FQKDSWLV68GFZWVigoKEBRURFmzpwJPz8/eHt7Iy0tDRMmTOCoBBERkYnS6S90fHw8evXqBWtra7z99tu4ceMGzp49i1GjRuncUXh4OMaOHQs7OzuN9rq6OmRmZmLRokWQSqUIDg7G6NGjkZaWBgAoKyvD5s2bMX36dJw6dQpvv/32XeweERER6ZtOIxOPPfYYfv31V422PXv2tEsB5eXlsLS0hEwmU7f5+/sjJycHwK1Ri0ZhYWFYsmRJu/RLRERE7UPryIRSqcTy5csxduxYTJ8+Hbm5ue1egFwuh62trUabra0t5HJ5k+/dsWNHu/dPRERE4mgdmfjXv/6Fzz77DIIgAACOHDmCw4cPw8HBod0KkEqlqKmp0WirqamBVCpttz6IiIhIf7SOTGzZsgVdunTB7NmzERwcjOvXr2P79u3tWoC7uztUKhXKy8vVbcXFxfD29m7XfoiIiEg/tIaJc+fOYcKECViyZAm++eYbCIKAc+fOtakjpVIJhUKBhoYGqFQqKBQKKJVK2NjYICwsDElJSairq0Nubi6ys7MRERHRpn6IiIjIsLSGiRs3bqBnz54AoL4K4/r1623qKDk5GTKZDKtWrcKWLVsgk8mQnJwMAEhISIBCoUD//v0xd+5cJCYmcmSCiIjITEiExgkRzejTpw969eqFBx54AABQVFSk8VoikSArK8swlbaRRCKBll1sk1GjRqGkpESjzcfHB7t27WrXfoiIiMxBq5eGVlVVoaqqqtnXEolEf5URERGRWdAaJj755BND1UFERERmSmuYmDJliqHqICIiIjOlNUxs2rSp1Q1ERUW1WzFERERkfrSGiQULFmidFyGRSBgmiIiIOjmtYcLX17dJmGhoaEBpaWm7XyFBRERE5klrmNi5c6fG6//+979YuXIlBEHAfffdh7lz5+q1OCIiIjJ9Oj01dMeOHVixYgVKS0thb2+Pt956C8899xysra31XR8RERGZOK1hIjs7GytWrEBxcTHs7OwQHx+P2bNnM0QQERGRmtYw8fzzz0MikUAikcDPzw/Hjh3DsWPH1MslEgm+/vprvRdJREREpqvV0xyCIEAQBBw4cKDJMt4Bk4iIiLSGidjYWEPVQURERGaKYYKIiIhE0elqjkYqlQo7d+5EbW0thgwZAhcXF33VRURERGZCa5h45513sHfvXmzbtg09evTAc889h3379gEAunbtipSUFISEhBikUCIiIjJNFtoW7t+/H3Z2dujRowcKCwuxd+9eCIIAX19fKBQKrFy50lB1EhERkYnSGiYuXLgALy8vAFCPSEyePBnZ2dkIDQ1FaWmp/iskIiIik6Y1TKhUKiiVSgDAr7/+ColEgpEjRwIAHBwccOPGDf1XSERERCZN65wJNzc3ZGZmonv37ti3bx8kEgmGDRsGAKioqMB9991nkCKJiIjIdGkdmZg3bx7q6+vxzTff4MaNG3jyySfh6OiIS5cuITc3FwMHDjRUnURERGSitI5MTJw4EW5ubsjJyYGjoyMmTJgAALh27RreffddBAUFGaRIIiIiMl2t3meiX79+6Nevn0abh4cHPDw89FYUERERmQ+tpzmIiIiIWsMwQURERKIwTBAREZEoLYaJDz/8EDt37gQAbNq0CQUFBQYrioiIiMxHi2Hiyy+/xIEDBwAACxYswNatWw1WFBEREZmPFq/m6N69O9LT0/G///0PwK3bad/5SHKJRIIVK1bot0IiIiIyaS2GidDQUGzevBmbN2+GRCLBn3/+iT///FPjexgm/o+DgwNUKhUsLS2bXa5tGRERkTlrMUx88sknGDVqFE6ePImkpCQMHDhQ/VwOasrW1haWlpZ48cUXUV5errHM3d0dX3/9tZEqIyIi0q8Ww4SFhYX6jpcNDQ0IDAzEiBEjDFWX2SovL0dJSYmxyyAiIjKYVu+ACQCxsbGorKzEsmXLcPbsWbi4uGDq1KlwdXXVd31ERERk4nQKE/n5+Zg6dSrq6+shCAIAYM2aNUhNTeXDvoiIiDo5nW5alZCQgLq6OowZMwavvvoqxo4di7q6OiQmJuq7Pvz999+YOHEiIiIiEBUVhYsXL+q9TyIiItKdTiMTxcXFCA8Px+rVq9Vtc+bMwcGDB/VWWCN7e3ts3boVFhYWSE1NxcaNGxETE6P3fomIiEg3OoWJrl274sqVKxptV65cQdeuXfVS1O1uv5xSLpfD29tb730SERGR7nQ6zREUFIQjR45g2LBhmDlzJh599FEcOXIEgwcP1rmjlJQUhIWFwc3NrcnIQnV1NaKjo+Hh4YHg4GCkp6drLD9+/DjGjx+PlJSUJo9DJyIiIuPSaWTinXfeQWFhIU6fPo3Tp08DAFxcXPDWW2/p3JGjoyPmz5+PvXv3QqFQaCxbvHgxrKysUFBQgKKiIsycORN+fn7qUYiHHnoIP/74I/773//is88+w7Jly3Tul4iIiPRLpzDRp08f7N69Gz///DPOnj2LPn364IknnoC1tbXOHYWHhwMACgoKcP78eXV7XV0dMjMzsWvXLkilUgQHB2P06NFIS0vDm2++ievXr6tPp3Tv3v2u+iQiIiL90ylMAIC1tbX6Jlbtqby8HJaWlpDJZOo2f39/5OTkAAAKCwuRkJAACwsLdOvWjbfvJiIiMjE6hwl9kcvlsLW11WiztbWFXC4HcGu+xpYtW4xRGhEREelApwmY+iSVSlFTU6PRVlNTA6lUaqSKzIdKpbqrdiIiIn0w+siEu7s7VCoVysvL4e7uDuDWfS14CWjrmnuwGB8qRkREhtbqyIRSqcSUKVOwfPlyUR0plUooFAo0NDRApVJBoVBAqVTCxsYGYWFhSEpKQl1dHXJzc5GdnY2IiAhR/XUWjQ8Wa/y684mlRERE+tZqmLjnnntw8uRJ/P3336I6Sk5Ohkwmw6pVq7BlyxbIZDIkJycDuHW7boVCgf79+2Pu3LlITEzkyAQREZGZ0Ok0R0xMDJYvX46nnnoKQUFBsLKyuuuO4uLiEBcX1+wyOzs7rF279q63SURERManU5iIj4+HRCLBlClTNNolEgkqKyv1UhhFn5MGAAAecklEQVQRERGZB50nYDY+ery1NiIiIupcdAoThw8f1ncdREREZKZ0vp02AJw8eRIVFRUYNWqUXosiIiIi86HTTauqq6sxdepUjBgxArNnz8aZM2fw4IMPir5clIiIiMyfTmHiww8/xMGDB9GlSxcIggAXFxcEBgbi559/1nd9REREZOJ0ChN79+7F0KFD8eyzz6rbvLy8eCUHERER6RYmFAoFHB0dNdquXLnSpvtNEBERUceiU5jw9fXFzz//jPz8fADAkiVLsHPnTvj5+em1OCIiIjJ9OoWJ119/HTdu3MCxY8cgCAK++uorWFhYtHhHSyIiIuo8dLo0NDg4GJmZmVi/fj3Onj0LFxcXTJ8+nSMTOnJwcIBKpYKlpWWzy7UtIyIiMnU63wHTx8cHb7/9Ns6fP4/evXvD2tpan3V1KLa2ts0+LhzgI8OJiMj86RQmrl69itdffx07duxQt4WFhWHZsmWws7PTW3EdTePjwomIiDoSncLEwoULkZWVpdG2Y8cONDQ0YM2aNXopjG7hKRAiIjJ1OoWJX375BQ8++CDWrFkDDw8PlJWV4YUXXsAvv/yi7/o6vZZOj4SEhOC9994zTlFERES30SlMuLq6on///uoJl/7+/njkkUdw/PhxvRZHtzR3esTNzc1I1RAREWlqMUzc/qTQyMhIrFy5Ev7+/pDJZDhx4gQyMzMRGxtrkCI7stau9CAiIjJ1LYaJyMhISCQSjbZ33nlH/W9BELBkyRK8+OKL+quuE9B2pQdPZRARkTloMUw4Ozs3CROkPzyVQURE5qrFMHHkyBFD1kFERERmSuebVgG3Hu5VX1+v0ebs7NyuBREREZF50SlM7N+/H3Fxcbhw4YJGu0Qi4WPITQxv3U1ERIamU5j45z//ifPnzzdpFwSh3QsicXjrbiIiMjSdwkR1dTVGjBiBr776ClKpVN81UTvoqLfu5qgLEZHp0SlMzJ49Gzt27MD58+chk8l4lQcZDUddiIhMj05hIjw8HOvWrcPIkSM12jlngoyho466EBGZK53CxGuvvYZr1641aeecCSIiItIpTPz1118ICAjA4sWL0b17d33XRERERGZEpzAxffp0/Pbbbxg0aBC6dOmi75qIiIjIjOgUJg4cOIDS0lL0798fffr0Uc+Yl0gkyMrK0muBREREZNp0ChN//PEHAKC2tlZj4huv6iAiIiKdwsQnn3yi7zqIiIjITOkUJqZMmaLvOoiIiMhM6RQmYmNjm22XSCRYsWJFuxbUnLy8PLz//vvo0qULHB0dkZycDCsrK733S0RERK3TKUz88MMPkEgk6vtKNP7bUGGiT58++OGHH2BtbY2PPvoIP/30E8aPH6/3fkk3vMW1+eExI6L2dNcjEyqVCiUlJcjOzsa0adP0VtjtHnjgAfW/77nnHlhYWBikX9JNe9/imn/M9I+3JSei9tTm0xyvv/56s08S1SYlJQU//PADSkpKMGnSJHz66afqZdXV1Vi4cCH27dsHe3t7xMfHY/LkyRrrV1ZWYs+ePfjHP/5xV/2S/rXnLa5b+kMXEhKC9957r136IN6WnIjaj853wLxdTU0NTp48ieLi4rvqzNHREfPnz8fevXuhUCg0li1evBhWVlYoKChAUVERZs6cCT8/P3h7e6v7jImJQXJyMm+c1Qk094fOzc3NSNUQEZE2OoWJhx9+uNl2Hx+fu+osPDwcAFBQUKAxqlFXV4fMzEzs2rULUqkUwcHBGD16NNLS0vDmm29CqVRi7ty5iI2NhYeHx131SURERPql0+QDQRA0vrp164agoCCsXLmyXYooLy+HpaUlZDKZus3f3x+lpaUAgK1btyI/Px+ffvopIiMjkZGR0S79kiaVStWmZURE1LnpNDJx9uxZvRYhl8tha2ur0WZrawu5XA4AiIyMRGRkpF5rIMNOyuMkSyLqDDrLlVM6hQl9k0qlqKmp0WirqamBVCo1UkWdl6Em5XGSJRF1Bp3lyimtYcLFxUXryhKJBJWVlaKLcHd3h0qlQnl5Odzd3QEAxcXF6smX1DFxkiURdQad4coprXMm7pwr0dzX3VAqlVAoFGhoaIBKpYJCoYBSqYSNjQ3CwsKQlJSEuro65ObmIjs7GxEREaJ2joiIiPRP68jETz/9pPG6rq4OKSkp2LZtGwRBgL+//111lpycrPHQsC1btiA2NhZxcXFISEhAXFwc+vfvDzs7OyQmJnJkgoiIyAxoDRONYUGhUGDdunX48ssvcfnyZfj4+CA2NlZ9qaeu4uLiEBcX1+wyOzs7rF279q62R0RERManNUxcv34d69evxxdffIHLly/D09MTS5YswYQJEwxVHxERkUnrSFdltJXWMPHII4/g0qVLkEgkmDBhAiZOnAgLCwtkZ2ervyc0NFTvRRIREZkqXp3WSpi4ePGi+gmh27Ztw7Zt2zSWt9fVHEREROass1+dpjVMODs7QyKRGKoWIiIiMkNaw8SRI0cMVQcRERGZKZ2ezUFERGROWnqeEJ8zpB8mcTttIiKi9tTcpMiOdgtrU8IwQdTOOsuDfYhMXWe4jbWpYJggamed5cE+RESNGCZIbxwcHDrtJ3F+IiKizoRhgvTG1ta22U/pnelGLkREnQHDBOndnZ/SO9ONXIiIOgNeGkpEpIW2Swl5mSHRLRyZ6EQ68xwGorbihFrTZc7/n2n7/9gc94thohNpaQ4DwHkMRNpwQq1pMuf/z1r6/9hcQyrDRCfU2R9IQ0Qdh7n/f9ZRgirnTBAREZEoDBPUqsZze2ReeMyIyFB4moNaxbkW5onHjIgMhWGCdGbu5yY7Ix4zIjIEnuYgIiJqBU8baseRCSIiolbwtKF2DBNEREQ64GnDlvE0BxEREYnCMEFERESiMEwQEZFRaZvcqFQqDVgJtRXnTBARkVG1NrmREx9NH8MEdQitPRHVHJ/CR9SZaJvcyImPpo9hgjoEbXfpNNen8JHhMGwSicMwQR1KR3kCHxkW7yFAJA7DBBEROJROJAav5iAiIiJRGCaIiIhIFJMPE9euXcO4cePg6enJc+FEREQmyOTDhLW1NdavX49x48YZuxQiURovX20Jn0pIZFz8HWw7k5+AaWVlhV69ehm7DCLRzOHyVd6vgzozXtXTdgYLEykpKfjhhx9QUlKCSZMm4dNPP1Uvq66uxsKFC7Fv3z7Y29sjPj4ekydPNlRpRAZlypevmkPgIdInXtXTNgYLE46Ojpg/fz727t0LhUKhsWzx4sWwsrJCQUEBioqKMHPmTPj5+cHb29tQ5RGZJX2NFJhy4CHzxFGtjs1gYSI8PBwAUFBQgPPnz6vb6+rqkJmZiV27dkEqlSI4OBijR49GWloa3nzzTUOVR2SWOCxL5oI/qx2b0edMlJeXw9LSEjKZTN3m7++PnJwc9esZM2agqKgIJ0+exLPPPoupU6cao1Qik8RhWTIX/FntuIweJuRyOWxtbTXabG1tIZfL1a+//fZbQ5dFREREOjL6paFSqRQ1NTUabTU1NZBKpUaqiIiIiO6G0cOEu7s7VCqVxnm04uJiTr4kIiIyEwYLE0qlEgqFAg0NDVCpVFAoFFAqlbCxsUFYWBiSkpJQV1eH3NxcZGdnIyIiwlClETWLN5giItKNweZMJCcn45NPPlG/3rJlC2JjYxEXF4eEhATExcWhf//+sLOzQ2JiIkcmyOh4vwUiIt0YLEzExcUhLi6u2WV2dnZYu3atoUoh0hnvt0BE1Dqjz5kgMjetPWODiEwTf3f1x+iXhhKZG223nOYNeIhMF3939YdhgqiNeAMeIvPE3932x9McRCaAjycnc9LSzyN/TjsvjkwQmQA+rZPMSXM/q/w57dwYJohMCK8eIXPBn1W6HU9zEBERkSgME0Qmjpez3R1t75VSqTRgJYbD+TZkbDzNQWTiWppPwUvZmtfapX8d8X3kfBsyNoYJIjNx5zlqXsrWMm2X/nXU95FzGMiYeJqDiIiIRGGYICIyA6Y+94Fzezo3nuYgIjIDpn4baN6qunNjmCAiMhPmcBtoc6iR2h9PcxAREZEoDBNEREQkCsMEERERicIwQZ0eZ6BTW7R29YK2u20a6mfOHGqkjoETMKnT490lqS10uXrB2HelNIcaqWNgmCBCx70rIunf3dxt01jMoUYybzzNQURERKIwTBAREZEoDBNEREQkCsMEERERicIwQURERKIwTFCHx6cZGpe2976ty7TdH4GIDI+XhlKHx6cZGldL731r9zJoy/0ReDyJjINhgjoNPs3QeNp6L4O7vT8CjyeRcfA0BxEREYnCMEFERESiMEwQERGRKAwTREREJArDBBEREYliFmHio48+wuTJk/Hiiy+ivr7e2OUQERHRbUw+TJSUlOD06dNIT09HSEgIvv/+e2OXRERERLcx+TBx5MgRjBw5EgDw+OOP4+jRo0auiIiIiG5nsDCRkpKCsLAwuLm5ISYmRmNZdXU1oqOj4eHhgeDgYKSnp6uX/e9//0P37t0B3LqT4dWrVw1VMhEREenAYHfAdHR0xPz587F3714oFAqNZYsXL4aVlRUKCgpQVFSEmTNnws/PD97e3ujRoweuXbsGAKipqUHPnj0NVTIRERHpwGAjE+Hh4Rg7dizs7Ow02uvq6pCZmYlFixZBKpUiODgYo0ePRlpaGgAgODgY+/btAwDs3bsXgwcPNlTJREREpAOjz5koLy+HpaUlZDKZus3f3x+lpaUAAF9fX/Tp0weTJ0/Gvn37MG3aNGOVStSptfb01bt9kief5mq6eGzobhn9QV9yuRy2trYabba2tpDL5erX8fHxhi6LiO6gy9NX71ym7SmefJqr6eKxobtl9DAhlUpRU1Oj0VZTUwOpVGqkiohIm7t5kqcuT/Hk0z9NF48N6cropznc3d2hUqk00m9xcTG8vb2NWBURERHpymBhQqlUQqFQoKGhASqVCgqFAkqlEjY2NggLC0NSUhLq6uqQm5uL7OxsREREGKo0IiIiEsFgYSI5ORkymQyrVq3Cli1bIJPJkJycDABISEiAQqFA//79MXfuXCQmJnJkgoiIyEwYbM5EXFwc4uLiml1mZ2eHtWvXGqoUIiIiakdGnzNBRERE5o1hgoiIiERhmCAiIiJRGCaIiIhIFIYJIiIiEoVhgoiIiESRCIIgGLsIfZJIJMYugYiIyOj0+efe6M/m0LcOnpWIiIiMjqc5iIiISBSGCSIiIhKFYYKIiIhEYZggIiIiURgmiIiISBSGCSIiIhKFYYKIiIhEYZggIiIiURgmiIiISBSGCR1VV1cjOjoaHh4eCA4ORnp6urFLarPIyEi4u7vD09MTnp6eCAkJUS9LT09HcHAwPDw88Pzzz6O6ulq9rLX3QNu6hpSSkoKwsDC4ubkhJiZGY9kvv/yCxx57DDKZDJGRkTh79qx62fXr1xEbGwtvb28EBATgq6++ard1DaGl/T5z5gycnZ3Vx9vT0xMrV67UuXZT3u/r168jLi4OwcHB8PLyQmhoKHbv3t0utZvqfmvb5458rF977TUMHDgQ3t7eePTRR7Fhw4Z2qdsc99kkj7NAOnnllVeEl156SaitrRWOHDkieHt7CyUlJcYuq00iIiKE7777rkl7SUmJ4OnpKeTk5Ai1tbXC3LlzhZdfflm9XNt70Nq6hrR9+3Zhx44dwhtvvCHMnz9f3V5VVSV4e3sL//3vf4X6+nrhgw8+EMaNG6denpCQIDz55JNCdXW18OeffwoDBgwQdu/eLXpdQ2lpvysrKwUnJyfh5s2bza5nzvstl8uFpKQkobKyUlCpVEJ2drbg6ekpVFZWdtjjrW2fO/KxLikpERQKhSAIglBWViYMGDBAKCgo6LDHWds+m+JxZpjQgVwuF1xdXYUTJ06o2+bNmycsXbrUiFW1XUthIiEhQZg7d6769alTpwRXV1ehpqam1fdA27rG8tFHH2n8Uf3222+FCRMmqF/L5XLB3d1dKCsrEwRBEAIDA4W9e/eqly9btkwdiMSsa2h37ndr//F0lP1uNGrUKOHHH3/sNMdbEP5vnzvLsS4rKxMCAgKEjIyMTnOcb99nUzzOPM2hg/LyclhaWkImk6nb/P39UVpaasSqxElMTMRDDz2ESZMm4dChQwCAP//8E35+furv6du3L6ysrFBeXt7qe6BtXVNxZ402NjZwdXVFaWkprl69igsXLmgs9/Pzw59//il6XVMxZMgQDBo0CAsWLMCVK1cAoMPt999//43y8nJ4e3t3muN9+z436qjHOj4+HjKZDMOHD8f999+PUaNGdfjj3Nw+NzKl48wwoQO5XA5bW1uNNltbW8jlciNVJM6bb76JnJwc/Prrr5g+fTpmzZqFioqKZveze/fuqK2tbfU90LauqZDL5ejevbtGm62trXr/Gl83ur1+Mesam729PTIzM3HkyBFkZWWhtrYW8+bNA4AOtd83b97EvHnzEBkZCQ8Pj05xvO/c545+rBMTE/Hnn38iPT0dYWFh6NKlS4c/zs3tsykeZ4YJHUilUtTU1Gi01dTUQCqVGqkicQIDA3Hvvfeia9eumDJlCgYPHozdu3dDKpU2+aGpqanBvffe2+p7oG1dU9HcPtTW1qr3r/F1o9vrF7OusUmlUgwYMAD33HMP7rvvPixduhT79u3TOH7mvt8NDQ34xz/+gS5dumDp0qUAOv7xbmmfO/qxtrS0RHBwMM6fP4/169d3+OMMNL/PpnacGSZ04O7uDpVKpTFkX1xcrDGsaM4kEgkEQYCXlxeKi4vV7adPn8aNGzfg7u7e6nugbV1TcWeNdXV1qKiogLe3N3r27AlHR0eN5cXFxfDy8hK9rqmRSCQAAEEQOsR+C4KAuLg4/P3331i9ejWsrKxE127q+93SPt+pox3r26lUKpw+fbpDH+c7Ne7znUziON/lHJBO6+WXXxZeeeUVQS6XC0ePHjXbqzmuXr0q7NmzR6ivrxdu3rwppKWlCTKZTCgrKxNKSkoELy8v4fDhw4JcLhfmzZunMfFG23vQ2rqGdPPmTaG+vl5ISEgQ5s2bp97Xy5cvC97e3sKPP/4o1NfXCx9++KHGLOalS5cKTz31lFBdXa2e7NQ4i1nMusbe719//VUoKysTVCqVUFVVJbz00ktCREREh9nv119/XRg3bpxQW1ur0d6Rj3dL+9xRj/Xff/8tbN26VaitrRWUSqWwZ88eQSaTCVlZWR32OGvbZ1M8zgwTOrpy5Yowe/ZsQSaTCUFBQcKWLVuMXVKbXL58WQgLCxM8PT0FHx8fYdy4ccK+ffvUy7ds2SIEBQUJMplMmDVrlnDlyhX1stbeA23rGlJSUpLg5OSk8ZWUlCQIgiDs27dPCAkJEdzd3YWIiAihsrJSvZ5CoRAWLFggeHl5Cf379xe+/PJLje2KWdcQWtrv9PR0YciQIYJMJhMCAgKE1157Tbh48aLOtZvyfp85c0ZwcnIS3NzcBA8PD/VXWlqa6NpNdb+17XNHPdaXL18WnnrqKcHHx0fw8vISHn/8ceE///lPu9RtjvtsisdZIgiC0NYhFyIiIiLOmSAiIiJRGCaIiIhIFIYJIiIiEoVhgoiIiERhmCAiIiJRGCaIiIhIFIYJog5uyJAhcHZ2Vj/QzZhUKhViYmLg6+sLZ2dnfPDBB3rvMyYmBs7OzlixYoXe+yLqrBgmiAyg8Q96YGAg6uvrAQDHjx+Hs7MznJ2djVyd4Wzfvh2bNm2CpaUlnn/+eTz88MNNvufQoUNwdnaGr6+vESr8v/6HDBlilP6JzBHDBJEBXbx4EevXrzd2GaLdvHmzTes1Ptvl8ccfx5IlSxAaGtqeZRGRkTBMEBmQRCLBF198oR6duFPjSMWZM2cAACtWrICzszNiYmIAAKmpqXB2dsYTTzyB9957D56enhgxYgSOHz+O5cuXw8fHB0OHDsW+ffuabLuwsBCjR4+Gl5cXoqOjceXKFfWyo0ePIjIyEn5+fggMDERsbKx6+ZkzZ9R1rV+/HoGBgXj66aebrf/MmTOYM2cOBg4cCD8/P0RGRuLYsWPqffn4448BAGlpaXB2dkZqamqr75kgCEhMTERQUBDc3NwQEBCAZ555Rl1fVVUVFi5ciMGDB8Pb2xvjx4/Hnj17Wtze/v37ERoaCh8fH7i6uiI4OBhJSUkAbo1KREVFAQDOnj2rMXJUX1+PhIQEDBs2DB4eHhgzZgyysrI0tjt27Fh4eHjAx8cHY8aMQWZmZqv7R9QRMEwQGdD48eNx+fJlrFu3TtR2SkpKkJ+fD09PT5SVlSEqKgrbt29HYGAgKisrERsb22SdTz75BP369YO9vT2ysrLwxhtvqLc1depU/P777xgxYgTc3d2RmpqKl156CXfebX/ZsmUYOXIkgoKCmmy/rq4OU6ZMwfbt2+Hu7o5hw4YhJycHU6ZMQUVFBQIDAzFw4EAAgKenJ6Kjo3V6GuEvv/yCVatWwdLSEtOmTcOQIUNQUlICuVyOhoYGzJo1Cxs3boS9vT1CQ0NRWFiImTNnIjc3t9ntnT9/Hvb29pg4cSIiIiIgl8uxcuVKZGRkoHfv3ggPDwcA3HvvvYiOjkZ0dDQAIC4uDp9//jlsbW0RHh6Oc+fO4YUXXlDPRVmwYAGKi4sRHh6O8PBwWFhYoLS0tNX9I+oI7jF2AUSdycSJE1FaWop///vfGDRoUJu3Y2Njg++//x75+fmIiorCtWvXkJGRAScnJ3h7e+PChQuoqqpCr1691OssWrQIL7zwAoqKihAaGorMzEzI5XKsX78eN27cgL+/PxwcHODg4IBjx47h0KFDOHnyJLp27arexldffYVHH3202Zp+/vlnVFZWwtXVFZs3b4ZEIkF0dDSysrKwceNGxMfH49ixY8jPz0dAQIDOky+VSiUAoG/fvpgwYQK8vLzQq1cvCIKAgoICHDt2DFKpFOnp6bCxsYG9vT3WrFmDdevWYfDgwU22FxUVBQcHBxQWFqK6uhqurq64evUqDh48iEmTJmH27NnIzMxEz5491TVWVVUhIyMDFhYWCAoKgoWFBby9vZGTk4Nvv/0WjzzyCJRKJbp27aoe9XB3d28Sxog6KoYJIgOysLDAggUL8MorryAlJaXV71epVM22u7i4wNraGt27d1e3yWQyWFpaql/X1dVphAlPT08AgIeHh7rt/Pnz6lMq+fn5yM/P1+jn1KlT8PHxUb9u7o9zo7Nnz6rrkEgkGn399ddfWvZSu+HDh+O5555DWlqa+hREQEAA1q5dq67dyckJNjY2Gn021nOnf/7zn/juu++atFdVVbVYQ2M/DQ0NTY5bRUUFgFujNkuWLMFLL70EALCzs8PSpUsxadIkXXeVyGzxNAeRgU2YMAG+vr7Ytm1bk2XW1tYAgJqaGgBocZjcwqLpr+7tQaI5ZWVlAIATJ06o23r37g0XFxcAwJw5c/DXX3+pv3JycjB69GiNbdw+SnGnPn36ALg1ybLxE/nJkycBQNQVKyqVCkuXLsUff/yBgwcPIjIyEr/99hs2bNigrv3cuXPqeSiNfTbWc6fG933lypU4c+YMZs6cCQDqmhvfx4aGBvU6jf106dIFhYWF6veooqICa9asAQCMHDkSBw8eRGFhIVavXo3q6mosW7aszftNZE44MkFkYBKJBAsWLMCcOXOaLHvooYeQm5uLt956CzKZDD/99FO79fvxxx+juLhYfY4/LCwMUqkU06dPx4YNG/DNN9+gsrIS9vb2KCsrQ15eXouf7pvzxBNPwMXFBRUVFYiKioKdnR127NiBbt26Ydq0aW2uOy8vDwsWLMCgQYPQs2dP5OXlAQB69OiBAQMGYODAgcjPz8fkyZPh5eWFjIwMSCQSPPfcc81uz8HBAdeuXcPatWuxf/9+7NixQ2O5k5MTgFujNgsXLoSbmxteffVVTJgwAdu2bcP48eMREhKC6upqHD16FDNmzEBcXBzGjBkDFxcXODk54dy5cwCgMXJE1JFxZILICMLDw+Hv79+k/cMPP4Svry+Kiopw/vx5TJ06td36jI2NRWFhIaqqqhAaGorly5cDAPz9/fH9999jyJAhOHz4MDIyMlBbW4tXX331rrZvY2OD1NRUhIeH48SJEzhw4ACGDh2K1NRUuLm5tbnuBx54AG5ubjhw4AA2bNiA+vp6zJgxA9OnT4eFhQXWrVuHqVOn4vLly8jKysJDDz2ElJQUBAcHN7u9pKQkeHh4oKysDLW1tXj22Wc1lru4uODll19G9+7dsXHjRqSlpQG4dTXKvHnzIJFIsGnTJuTl5WHQoEEYMWIEACAkJAQnT57E5s2bkZubi6FDh6qvEiHq6CQCZwgRERGRCByZICIiIlEYJoiIiEgUhgkiIiIShWGCiIiIRGGYICIiIlEYJoiIiEgUhgkiIiIShWGCiIiIRGGYICIiIlH+P0MyunPdWQsCAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 540x378 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "fig , ax = plt.subplots()\n",
    "\n",
    "n , bins , patches = plt.hist(SNV_DF.num_isolates.astype(float) , bins = 60 , color = 'k' , rwidth = 1.0 , edgecolor='white', linewidth=1.35)\n",
    "\n",
    "plt.title('Number of Isolates Each SNP is Found in', fontweight = 'bold' , fontsize = 12, color = 'k')\n",
    "plt.xlabel('Number of Isolates', fontweight = 'bold' , fontsize = 12, color = 'k')\n",
    "plt.ylabel('Number of SNPs', fontweight = 'bold' , fontsize = 12, color = 'k')\n",
    "\n",
    "ax.spines['right'].set_visible(False)\n",
    "ax.spines['top'].set_visible(False)\n",
    "\n",
    "ax.tick_params(labelcolor = 'k')\n",
    "ax.set_yscale(\"log\", nonposy='clip')\n",
    "\n",
    "fig = plt.gcf()\n",
    "fig.set_size_inches(7.5, 5.25)\n",
    "fig.tight_layout()\n",
    "\n",
    "plt.tick_params(axis='y', which='major', labelsize=12 , labelcolor = 'k')\n",
    "plt.tick_params(axis='x', which='major', labelsize=12 , labelcolor = 'k')\n",
    "\n",
    "#file_name = '/home/rv76/Farhat_Lab/Figures/Final_5/Scraping_rollingDB/Mean_Coverage_Distribution_across_all_isolates.pdf'\n",
    "#plt.savefig(file_name, bbox_inches='tight', dpi = 300 , transparent = True)\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### How many SNP sites have at least *1* isolate with an alternate allele?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(878244, 4)"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.shape(SNV_DF[SNV_DF.num_isolates >= 1])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### How many SNP sites have at least *2* isolates with an alternate allele?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(417738, 4)"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.shape(SNV_DF[SNV_DF.num_isolates >= 2])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### How many SNP sites have at least *3* isolates with an alternate allele?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(254265, 4)"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.shape(SNV_DF[SNV_DF.num_isolates >= 3])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### How many SNP sites have at least *4* isolates with an alternate allele?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(187622, 4)"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.shape(SNV_DF[SNV_DF.num_isolates >= 4])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### How many SNP sites have at least *5* isolates with an alternate allele?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(146874, 4)"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.shape(SNV_DF[SNV_DF.num_isolates >= 5])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Keep only SNVs in which the corresponding *Reference Position* has a SNV in $\\ge$ 5 isolates"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "SNVs founds in only 1-4 isolate(s) are dropped"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "SNV_DF = SNV_DF[SNV_DF.num_isolates >= 5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(146874, 4)"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.shape(SNV_DF)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>num_isolates</th>\n",
       "      <th>pos</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A</td>\n",
       "      <td>[C]</td>\n",
       "      <td>8</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>C</td>\n",
       "      <td>[G]</td>\n",
       "      <td>19</td>\n",
       "      <td>29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>A</td>\n",
       "      <td>[C, G]</td>\n",
       "      <td>10</td>\n",
       "      <td>31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>C</td>\n",
       "      <td>[T]</td>\n",
       "      <td>5</td>\n",
       "      <td>48</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>G</td>\n",
       "      <td>[C]</td>\n",
       "      <td>5</td>\n",
       "      <td>64</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   ref     alt num_isolates  pos\n",
       "2    A     [C]            8   11\n",
       "6    C     [G]           19   29\n",
       "7    A  [C, G]           10   31\n",
       "10   C     [T]            5   48\n",
       "11   G     [C]            5   64"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SNV_DF.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Re-Structure DataFrame to include only the Reference Position, Reference Allele and Alternate Allele(s)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "SNV_annotation_DF = pd.DataFrame(index = range( 0 , np.shape(SNV_DF)[0] ) ) #index for each SNV\n",
    "SNV_annotation_DF['pos'] = list(SNV_DF.pos) #H37Rv Reference Position\n",
    "SNV_annotation_DF['ref'] = list(SNV_DF.ref) #reference allele\n",
    "SNV_annotation_DF['alt'] = list(SNV_DF.alt) #alterante allele(s) list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(146874, 3)"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.shape(SNV_annotation_DF)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>11</td>\n",
       "      <td>A</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>29</td>\n",
       "      <td>C</td>\n",
       "      <td>[G]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>31</td>\n",
       "      <td>A</td>\n",
       "      <td>[C, G]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>48</td>\n",
       "      <td>C</td>\n",
       "      <td>[T]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>64</td>\n",
       "      <td>G</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   pos ref     alt\n",
       "0   11   A     [C]\n",
       "1   29   C     [G]\n",
       "2   31   A  [C, G]\n",
       "3   48   C     [T]\n",
       "4   64   G     [C]"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SNV_annotation_DF.head() #index = row indices of genotype matrix , 0 supports ref, 1 supports alt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### This is the DataFrame which we'll use to extract SNVs of interest when iterating over all of the VCF files in rollingDB"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Output Annotation for SNVs of interest in Genotypes folder"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "SNV_annotation_DF.to_pickle('/n/data1/hms/dbmi/farhat/Roger/inhost_TB_dynamics_project/rolling_DB_scrape/Genotypes/genotypes_SNP_annotation.pkl')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Create a text file with a single column of all Reference Positions that we're interested in based off of DF above. We will use this text file to subset the *full* VCF files to only the positions that Reference Positions that have a SNV in at least 2 isolates."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "SNV_ref_positions = ['NC_000962.3\\t' + ref_pos for ref_pos in list( SNV_annotation_DF.pos.astype(str) ) ]\n",
    "\n",
    "#save each element as a new row in a text file\n",
    "with open('/n/data1/hms/dbmi/farhat/Roger/inhost_TB_dynamics_project/rolling_DB_scrape/Genotypes/ref_positions_to_select.txt', 'w') as f:\n",
    "    for item in SNV_ref_positions:\n",
    "        f.write(\"%s\\n\" % item)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "####################################################################################################################################################################################"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## [6] This section of the notebook submits jobs to subset Full VCFs to the 146,874 reference positions that have a SNV in $\\ge 5$ isolates in rollingDB."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "####################################################################################################################################################################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "from slurmpy import Slurm\n",
    "import numpy as np\n",
    "import itertools\n",
    "import shutil"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Every isolate in the pool of **33,873** isolates in rollingDB had a reduced VCF file and a corresponding Full VCF file. We will subset each of these Full VCF files and create subset VCF files containing information for only the **146,874** reference positions that we're interested in."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [],
   "source": [
    "isolate_id_series = pd.read_pickle( '/n/data1/hms/dbmi/farhat/Roger/inhost_TB_dynamics_project/rolling_DB_scrape/Genotypes/genotypes_isolate_annotation.pkl' )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    SAMEA3558733\n",
       "1    SAMN05575260\n",
       "2    SAMN03648641\n",
       "3    SAMN03647419\n",
       "4    SAMEA3671418\n",
       "dtype: object"
      ]
     },
     "execution_count": 91,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "isolate_id_series.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "33873"
      ]
     },
     "execution_count": 92,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(isolate_id_series)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Create directory for each isolate to store output for genotyping"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [],
   "source": [
    "for tag in isolate_id_series.values:\n",
    "\n",
    "    #create directory to store subset VCF and genotypes for each isolate\n",
    "    isolate_genotype_dir = '/n/data1/hms/dbmi/farhat/Roger/inhost_TB_dynamics_project/rolling_DB_scrape/Genotypes/Genotypes_per_isolate/' + tag\n",
    "    if os.path.exists(isolate_genotype_dir):\n",
    "        shutil.rmtree(isolate_genotype_dir)\n",
    "        os.makedirs(isolate_genotype_dir)\n",
    "    elif not os.path.exists(isolate_genotype_dir):\n",
    "        os.makedirs(isolate_genotype_dir)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Subset VCF to reference positions of interest for isolates that had Full VCF files "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create BASH scripts to extract VCFs of N isolates per job"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [],
   "source": [
    "N_isolates_per_job = 70\n",
    "\n",
    "njobs = int( np.ceil( float( len( isolate_id_series ) ) / float(N_isolates_per_job) ) ) #number of jobs required if we split for every N isolates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "484\n"
     ]
    }
   ],
   "source": [
    "print njobs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [],
   "source": [
    "bash_scripts = []\n",
    "\n",
    "#genomic data directory\n",
    "rolling_DB_dir = '/n/data1/hms/dbmi/farhat/rollingDB/genomic_data/'\n",
    "\n",
    "#split jobs up into tasks of N\n",
    "for first_isolate_i_for_job in np.arange(0 , (njobs)*N_isolates_per_job , N_isolates_per_job):\n",
    "\n",
    "    last_isolate_i_for_job = first_isolate_i_for_job + N_isolates_per_job\n",
    "    \n",
    "    commands_list = []\n",
    "    commands_list.append( 'module load bcftools/1.9' ) #only need to load this once per job\n",
    "\n",
    "    for isolate_i in range(first_isolate_i_for_job , last_isolate_i_for_job): \n",
    "\n",
    "        try:\n",
    "            tag = isolate_id_series[isolate_i]\n",
    "            zipped_VCF = rolling_DB_dir + tag + '/pilon/' + tag + '_full.vcf.gz'\n",
    "            isolate_genotypes_dir = '/n/data1/hms/dbmi/farhat/Roger/inhost_TB_dynamics_project/rolling_DB_scrape/Genotypes/Genotypes_per_isolate/' + tag\n",
    "\n",
    "            commands_list.append( 'cd {}'.format(isolate_genotypes_dir) )\n",
    "            commands_list.append( 'time bcftools view {} -O b -o current.bcf'.format(zipped_VCF) )\n",
    "            commands_list.append( 'bcftools index current.bcf' )\n",
    "            commands_list.append( 'bcftools view current.bcf --regions-file /n/data1/hms/dbmi/farhat/Roger/inhost_TB_dynamics_project/rolling_DB_scrape/Genotypes/ref_positions_to_select.txt -o {0}_ref_positions_for_genotyping.vcf -O v'.format(tag) )\n",
    "            commands_list.append( 'rm current.bcf.csi current.bcf' )\n",
    "            \n",
    "        except KeyError: #applies only to the last job if it has < N isolates\n",
    "            continue\n",
    "        \n",
    "    bash_scripts.append(commands_list)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Submit each job to O2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3318878\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_1\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3318879\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_2\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3318880\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_3\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3318881\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_4\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3318882\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_5\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3318883\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_6\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3318884\n",
      "submitted: Submitted batch job 3318885\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_7\n",
      "VCF_8\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3318886\n",
      "submitted: Submitted batch job 3318887\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_9\n",
      "VCF_10\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3318888\n",
      "submitted: Submitted batch job 3318889\n",
      "submitted: Submitted batch job 3318890\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_11\n",
      "VCF_12\n",
      "VCF_13\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3318891\n",
      "submitted: Submitted batch job 3318892\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_14\n",
      "VCF_15\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3318893\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_16\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3318894\n",
      "submitted: Submitted batch job 3318895\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_17\n",
      "VCF_18\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3318896\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_19\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3318897\n",
      "submitted: Submitted batch job 3318898\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_20\n",
      "VCF_21\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3318899\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_22\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3318900\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_23\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3318901\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_24\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3318902\n",
      "submitted: Submitted batch job 3318903\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_25\n",
      "VCF_26\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3318904\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_27\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3318905\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_28\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3318906\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_29\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3318907\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_30\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3318909\n",
      "submitted: Submitted batch job 3318910\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_31\n",
      "VCF_32\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3318911\n",
      "submitted: Submitted batch job 3318912\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_33\n",
      "VCF_34\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3318913\n",
      "submitted: Submitted batch job 3318914\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_35\n",
      "VCF_36\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3318915\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_37\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3318916\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_38\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3318917\n",
      "submitted: Submitted batch job 3318918\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_39\n",
      "VCF_40\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3318919\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_41\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3318920\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_42\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3318921\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_43\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3318922\n",
      "submitted: Submitted batch job 3318923\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_44\n",
      "VCF_45\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3318924\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_46\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3318925\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_47\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3318926\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_48\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3318927\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_49\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3318928\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_50\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3318929\n",
      "submitted: Submitted batch job 3318930\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_51\n",
      "VCF_52\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3318931\n",
      "submitted: Submitted batch job 3318932\n",
      "submitted: Submitted batch job 3318933\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_53\n",
      "VCF_54\n",
      "VCF_55\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3318934\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_56\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3318935\n",
      "submitted: Submitted batch job 3318936\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_57\n",
      "VCF_58\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3318937\n",
      "submitted: Submitted batch job 3318938\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_59\n",
      "VCF_60\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3318939\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_61\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3318940\n",
      "submitted: Submitted batch job 3318941\n",
      "submitted: Submitted batch job 3318942\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_62\n",
      "VCF_63\n",
      "VCF_64\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3318943\n",
      "submitted: Submitted batch job 3318944\n",
      "submitted: Submitted batch job 3318945\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_65\n",
      "VCF_66\n",
      "VCF_67\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3318946\n",
      "submitted: Submitted batch job 3318947\n",
      "submitted: Submitted batch job 3318948\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_68\n",
      "VCF_69\n",
      "VCF_70\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3318949\n",
      "submitted: Submitted batch job 3318950\n",
      "submitted: Submitted batch job 3318951\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_71\n",
      "VCF_72\n",
      "VCF_73\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3318952\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_74\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3318953\n",
      "submitted: Submitted batch job 3318954\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_75\n",
      "VCF_76\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3318955\n",
      "submitted: Submitted batch job 3318956\n",
      "submitted: Submitted batch job 3318957\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_77\n",
      "VCF_78\n",
      "VCF_79\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3318958\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_80\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3318959\n",
      "submitted: Submitted batch job 3318960\n",
      "submitted: Submitted batch job 3318961\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_81\n",
      "VCF_82\n",
      "VCF_83\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3318962\n",
      "submitted: Submitted batch job 3318963\n",
      "submitted: Submitted batch job 3318964\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_84\n",
      "VCF_85\n",
      "VCF_86\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3318965\n",
      "submitted: Submitted batch job 3318966\n",
      "submitted: Submitted batch job 3318967\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_87\n",
      "VCF_88\n",
      "VCF_89\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3318968\n",
      "submitted: Submitted batch job 3318969\n",
      "submitted: Submitted batch job 3318970\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_90\n",
      "VCF_91\n",
      "VCF_92\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3318971\n",
      "submitted: Submitted batch job 3318972\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_93\n",
      "VCF_94\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3318973\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_95\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3318974\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_96\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3318975\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_97\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3318976\n",
      "submitted: Submitted batch job 3318977\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_98\n",
      "VCF_99\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3318978\n",
      "submitted: Submitted batch job 3318979\n",
      "submitted: Submitted batch job 3318980\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_100\n",
      "VCF_101\n",
      "VCF_102\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3318981\n",
      "submitted: Submitted batch job 3318982\n",
      "submitted: Submitted batch job 3318983\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_103\n",
      "VCF_104\n",
      "VCF_105\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3318984\n",
      "submitted: Submitted batch job 3318985\n",
      "submitted: Submitted batch job 3318986\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_106\n",
      "VCF_107\n",
      "VCF_108\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3318987\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_109\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3318988\n",
      "submitted: Submitted batch job 3318989\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_110\n",
      "VCF_111\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3318990\n",
      "submitted: Submitted batch job 3318991\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_112\n",
      "VCF_113\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3318992\n",
      "submitted: Submitted batch job 3318993\n",
      "submitted: Submitted batch job 3318994\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_114\n",
      "VCF_115\n",
      "VCF_116\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3318995\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_117\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3318996\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_118\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3318997\n",
      "submitted: Submitted batch job 3318998\n",
      "submitted: Submitted batch job 3318999\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_119\n",
      "VCF_120\n",
      "VCF_121\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319000\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_122\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319001\n",
      "submitted: Submitted batch job 3319002\n",
      "submitted: Submitted batch job 3319003\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_123\n",
      "VCF_124\n",
      "VCF_125\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319004\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_126\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319005\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_127\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319006\n",
      "submitted: Submitted batch job 3319007\n",
      "submitted: Submitted batch job 3319008\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_128\n",
      "VCF_129\n",
      "VCF_130\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319009\n",
      "submitted: Submitted batch job 3319010\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_131\n",
      "VCF_132\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319011\n",
      "submitted: Submitted batch job 3319012\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_133\n",
      "VCF_134\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319013\n",
      "submitted: Submitted batch job 3319014\n",
      "submitted: Submitted batch job 3319015\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_135\n",
      "VCF_136\n",
      "VCF_137\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319016\n",
      "submitted: Submitted batch job 3319017\n",
      "submitted: Submitted batch job 3319018\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_138\n",
      "VCF_139\n",
      "VCF_140\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319019\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_141\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319020\n",
      "submitted: Submitted batch job 3319021\n",
      "submitted: Submitted batch job 3319022\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_142\n",
      "VCF_143\n",
      "VCF_144\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319023\n",
      "submitted: Submitted batch job 3319024\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_145\n",
      "VCF_146\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319025\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_147\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319026\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_148\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319027\n",
      "submitted: Submitted batch job 3319028\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_149\n",
      "VCF_150\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319029\n",
      "submitted: Submitted batch job 3319030\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_151\n",
      "VCF_152\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319031\n",
      "submitted: Submitted batch job 3319032\n",
      "submitted: Submitted batch job 3319033\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_153\n",
      "VCF_154\n",
      "VCF_155\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319034\n",
      "submitted: Submitted batch job 3319035\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_156\n",
      "VCF_157\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319036\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_158\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319037\n",
      "submitted: Submitted batch job 3319038\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_159\n",
      "VCF_160\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319039\n",
      "submitted: Submitted batch job 3319040\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_161\n",
      "VCF_162\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319041\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_163\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319042\n",
      "submitted: Submitted batch job 3319043\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_164\n",
      "VCF_165\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319044\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_166\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319045\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_167\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319046\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_168\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319047\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_169\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319048\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_170\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319049\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_171\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319050\n",
      "submitted: Submitted batch job 3319051\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_172\n",
      "VCF_173\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319052\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_174\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319053\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_175\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319054\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_176\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319055\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_177\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319056\n",
      "submitted: Submitted batch job 3319057\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_178\n",
      "VCF_179\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319058\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_180\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319059\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_181\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319060\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_182\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319061\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_183\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319062\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_184\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319063\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_185\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319064\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_186\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319065\n",
      "submitted: Submitted batch job 3319066\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_187\n",
      "VCF_188\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319067\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_189\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319068\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_190\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319069\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_191\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319070\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_192\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319071\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_193\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319072\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_194\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319073\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_195\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319074\n",
      "submitted: Submitted batch job 3319075\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_196\n",
      "VCF_197\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319076\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_198\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319077\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_199\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319078\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_200\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319079\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_201\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319080\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_202\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319081\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_203\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319082\n",
      "submitted: Submitted batch job 3319083\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_204\n",
      "VCF_205\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319084\n",
      "submitted: Submitted batch job 3319085\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_206\n",
      "VCF_207\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319086\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_208\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319087\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_209\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319088\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_210\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319089\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_211\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319090\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_212\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319091\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_213\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319092\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_214\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319093\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_215\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319094\n",
      "submitted: Submitted batch job 3319095\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_216\n",
      "VCF_217\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319096\n",
      "submitted: Submitted batch job 3319097\n",
      "submitted: Submitted batch job 3319098\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_218\n",
      "VCF_219\n",
      "VCF_220\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319099\n",
      "submitted: Submitted batch job 3319100\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_221\n",
      "VCF_222\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319101\n",
      "submitted: Submitted batch job 3319102\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_223\n",
      "VCF_224\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319103\n",
      "submitted: Submitted batch job 3319104\n",
      "submitted: Submitted batch job 3319105\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_225\n",
      "VCF_226\n",
      "VCF_227\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319106\n",
      "submitted: Submitted batch job 3319107\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_228\n",
      "VCF_229\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319108\n",
      "submitted: Submitted batch job 3319109\n",
      "submitted: Submitted batch job 3319110\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_230\n",
      "VCF_231\n",
      "VCF_232\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319111\n",
      "submitted: Submitted batch job 3319112\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_233\n",
      "VCF_234\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319113\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_235\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319114\n",
      "submitted: Submitted batch job 3319115\n",
      "submitted: Submitted batch job 3319116\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_236\n",
      "VCF_237\n",
      "VCF_238\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319117\n",
      "submitted: Submitted batch job 3319118\n",
      "submitted: Submitted batch job 3319119\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_239\n",
      "VCF_240\n",
      "VCF_241\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319120\n",
      "submitted: Submitted batch job 3319121\n",
      "submitted: Submitted batch job 3319122\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_242\n",
      "VCF_243\n",
      "VCF_244\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319123\n",
      "submitted: Submitted batch job 3319124\n",
      "submitted: Submitted batch job 3319125\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_245\n",
      "VCF_246\n",
      "VCF_247\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319126\n",
      "submitted: Submitted batch job 3319127\n",
      "submitted: Submitted batch job 3319128\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_248\n",
      "VCF_249\n",
      "VCF_250\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319129\n",
      "submitted: Submitted batch job 3319130\n",
      "submitted: Submitted batch job 3319131\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_251\n",
      "VCF_252\n",
      "VCF_253\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319132\n",
      "submitted: Submitted batch job 3319133\n",
      "submitted: Submitted batch job 3319134\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_254\n",
      "VCF_255\n",
      "VCF_256\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319135\n",
      "submitted: Submitted batch job 3319136\n",
      "submitted: Submitted batch job 3319137\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_257\n",
      "VCF_258\n",
      "VCF_259\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319138\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_260\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319139\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_261\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319140\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_262\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319141\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_263\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319142\n",
      "submitted: Submitted batch job 3319143\n",
      "submitted: Submitted batch job 3319144\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_264\n",
      "VCF_265\n",
      "VCF_266\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319145\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_267\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319146\n",
      "submitted: Submitted batch job 3319147\n",
      "submitted: Submitted batch job 3319148\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_268\n",
      "VCF_269\n",
      "VCF_270\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319149\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_271\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319150\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_272\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319151\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_273\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319152\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_274\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319154\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_275\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319164\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_276\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319165\n",
      "submitted: Submitted batch job 3319166\n",
      "submitted: Submitted batch job 3319167\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_277\n",
      "VCF_278\n",
      "VCF_279\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319168\n",
      "submitted: Submitted batch job 3319169\n",
      "submitted: Submitted batch job 3319170\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_280\n",
      "VCF_281\n",
      "VCF_282\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319171\n",
      "submitted: Submitted batch job 3319172\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_283\n",
      "VCF_284\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319173\n",
      "submitted: Submitted batch job 3319174\n",
      "submitted: Submitted batch job 3319175\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_285\n",
      "VCF_286\n",
      "VCF_287\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319176\n",
      "submitted: Submitted batch job 3319177\n",
      "submitted: Submitted batch job 3319178\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_288\n",
      "VCF_289\n",
      "VCF_290\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319179\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_291\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319180\n",
      "submitted: Submitted batch job 3319181\n",
      "submitted: Submitted batch job 3319182\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_292\n",
      "VCF_293\n",
      "VCF_294\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319183\n",
      "submitted: Submitted batch job 3319184\n",
      "submitted: Submitted batch job 3319185\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_295\n",
      "VCF_296\n",
      "VCF_297\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319186\n",
      "submitted: Submitted batch job 3319187\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_298\n",
      "VCF_299\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319188\n",
      "submitted: Submitted batch job 3319189\n",
      "submitted: Submitted batch job 3319190\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_300\n",
      "VCF_301\n",
      "VCF_302\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319191\n",
      "submitted: Submitted batch job 3319192\n",
      "submitted: Submitted batch job 3319193\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_303\n",
      "VCF_304\n",
      "VCF_305\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319194\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_306\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319195\n",
      "submitted: Submitted batch job 3319196\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_307\n",
      "VCF_308\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319197\n",
      "submitted: Submitted batch job 3319198\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_309\n",
      "VCF_310\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319199\n",
      "submitted: Submitted batch job 3319200\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_311\n",
      "VCF_312\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319201\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_313\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319202\n",
      "submitted: Submitted batch job 3319203\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_314\n",
      "VCF_315\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319204\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_316\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319205\n",
      "submitted: Submitted batch job 3319206\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_317\n",
      "VCF_318\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319207\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_319\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319208\n",
      "submitted: Submitted batch job 3319209\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_320\n",
      "VCF_321\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319210\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_322\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319211\n",
      "submitted: Submitted batch job 3319212\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_323\n",
      "VCF_324\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319213\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_325\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319214\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_326\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319215\n",
      "submitted: Submitted batch job 3319216\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_327\n",
      "VCF_328\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319217\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_329\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319218\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_330\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319219\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_331\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319220\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_332\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319221\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_333\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319222\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_334\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319223\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_335\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319224\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_336\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319225\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_337\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319226\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_338\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319227\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_339\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319228\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_340\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319229\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_341\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319230\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_342\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319231\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_343\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319232\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_344\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319233\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_345\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319234\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_346\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319235\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_347\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319236\n",
      "submitted: Submitted batch job 3319237\n",
      "submitted: Submitted batch job 3319238\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_348\n",
      "VCF_349\n",
      "VCF_350\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319239\n",
      "submitted: Submitted batch job 3319240\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_351\n",
      "VCF_352\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319241\n",
      "submitted: Submitted batch job 3319242\n",
      "submitted: Submitted batch job 3319243\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_353\n",
      "VCF_354\n",
      "VCF_355\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319244\n",
      "submitted: Submitted batch job 3319245\n",
      "submitted: Submitted batch job 3319246\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_356\n",
      "VCF_357\n",
      "VCF_358\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319247\n",
      "submitted: Submitted batch job 3319248\n",
      "submitted: Submitted batch job 3319249\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_359\n",
      "VCF_360\n",
      "VCF_361\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319250\n",
      "submitted: Submitted batch job 3319251\n",
      "submitted: Submitted batch job 3319252\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_362\n",
      "VCF_363\n",
      "VCF_364\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319253\n",
      "submitted: Submitted batch job 3319254\n",
      "submitted: Submitted batch job 3319255\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_365\n",
      "VCF_366\n",
      "VCF_367\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319256\n",
      "submitted: Submitted batch job 3319257\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_368\n",
      "VCF_369\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319258\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_370\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319259\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_371\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319260\n",
      "submitted: Submitted batch job 3319261\n",
      "submitted: Submitted batch job 3319262\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_372\n",
      "VCF_373\n",
      "VCF_374\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319263\n",
      "submitted: Submitted batch job 3319264\n",
      "submitted: Submitted batch job 3319265\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_375\n",
      "VCF_376\n",
      "VCF_377\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319266\n",
      "submitted: Submitted batch job 3319267\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_378\n",
      "VCF_379\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319268\n",
      "submitted: Submitted batch job 3319269\n",
      "submitted: Submitted batch job 3319270\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_380\n",
      "VCF_381\n",
      "VCF_382\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319271\n",
      "submitted: Submitted batch job 3319272\n",
      "submitted: Submitted batch job 3319273\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_383\n",
      "VCF_384\n",
      "VCF_385\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319274\n",
      "submitted: Submitted batch job 3319275\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_386\n",
      "VCF_387\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319276\n",
      "submitted: Submitted batch job 3319277\n",
      "submitted: Submitted batch job 3319278\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_388\n",
      "VCF_389\n",
      "VCF_390\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319279\n",
      "submitted: Submitted batch job 3319280\n",
      "submitted: Submitted batch job 3319281\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_391\n",
      "VCF_392\n",
      "VCF_393\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319282\n",
      "submitted: Submitted batch job 3319283\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_394\n",
      "VCF_395\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319284\n",
      "submitted: Submitted batch job 3319285\n",
      "submitted: Submitted batch job 3319286\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_396\n",
      "VCF_397\n",
      "VCF_398\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319287\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_399\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319288\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_400\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319289\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_401\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319290\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_402\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319291\n",
      "submitted: Submitted batch job 3319292\n",
      "submitted: Submitted batch job 3319293\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_403\n",
      "VCF_404\n",
      "VCF_405\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319294\n",
      "submitted: Submitted batch job 3319295\n",
      "submitted: Submitted batch job 3319296\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_406\n",
      "VCF_407\n",
      "VCF_408\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319297\n",
      "submitted: Submitted batch job 3319298\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_409\n",
      "VCF_410\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319299\n",
      "submitted: Submitted batch job 3319300\n",
      "submitted: Submitted batch job 3319301\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_411\n",
      "VCF_412\n",
      "VCF_413\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319302\n",
      "submitted: Submitted batch job 3319303\n",
      "submitted: Submitted batch job 3319304\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_414\n",
      "VCF_415\n",
      "VCF_416\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319305\n",
      "submitted: Submitted batch job 3319306\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_417\n",
      "VCF_418\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319308\n",
      "submitted: Submitted batch job 3319309\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_419\n",
      "VCF_420\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319310\n",
      "submitted: Submitted batch job 3319311\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_421\n",
      "VCF_422\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319312\n",
      "submitted: Submitted batch job 3319313\n",
      "submitted: Submitted batch job 3319314\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_423\n",
      "VCF_424\n",
      "VCF_425\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319315\n",
      "submitted: Submitted batch job 3319316\n",
      "submitted: Submitted batch job 3319317\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_426\n",
      "VCF_427\n",
      "VCF_428\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319318\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_429\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319319\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_430\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319320\n",
      "submitted: Submitted batch job 3319321\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_431\n",
      "VCF_432\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319322\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_433\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319323\n",
      "submitted: Submitted batch job 3319324\n",
      "submitted: Submitted batch job 3319325\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_434\n",
      "VCF_435\n",
      "VCF_436\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319326\n",
      "submitted: Submitted batch job 3319327\n",
      "submitted: Submitted batch job 3319328\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_437\n",
      "VCF_438\n",
      "VCF_439\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319329\n",
      "submitted: Submitted batch job 3319330\n",
      "submitted: Submitted batch job 3319331\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_440\n",
      "VCF_441\n",
      "VCF_442\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319332\n",
      "submitted: Submitted batch job 3319333\n",
      "submitted: Submitted batch job 3319334\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_443\n",
      "VCF_444\n",
      "VCF_445\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319335\n",
      "submitted: Submitted batch job 3319336\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_446\n",
      "VCF_447\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319337\n",
      "submitted: Submitted batch job 3319338\n",
      "submitted: Submitted batch job 3319339\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_448\n",
      "VCF_449\n",
      "VCF_450\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319340\n",
      "submitted: Submitted batch job 3319341\n",
      "submitted: Submitted batch job 3319342\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_451\n",
      "VCF_452\n",
      "VCF_453\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319343\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_454\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319344\n",
      "submitted: Submitted batch job 3319345\n",
      "submitted: Submitted batch job 3319346\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_455\n",
      "VCF_456\n",
      "VCF_457\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319347\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_458\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319348\n",
      "submitted: Submitted batch job 3319349\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_459\n",
      "VCF_460\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319350\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_461\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319351\n",
      "submitted: Submitted batch job 3319352\n",
      "submitted: Submitted batch job 3319353\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_462\n",
      "VCF_463\n",
      "VCF_464\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319354\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_465\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319355\n",
      "submitted: Submitted batch job 3319356\n",
      "submitted: Submitted batch job 3319357\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_466\n",
      "VCF_467\n",
      "VCF_468\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319358\n",
      "submitted: Submitted batch job 3319359\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_469\n",
      "VCF_470\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319360\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_471\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319361\n",
      "submitted: Submitted batch job 3319362\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_472\n",
      "VCF_473\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319363\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_474\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319364\n",
      "submitted: Submitted batch job 3319365\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_475\n",
      "VCF_476\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319366\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_477\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319367\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_478\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319368\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_479\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319369\n",
      "submitted: Submitted batch job 3319370\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_480\n",
      "VCF_481\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319371\n",
      "submitted: Submitted batch job 3319372\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_482\n",
      "VCF_483\n",
      "VCF_484\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3319373\n"
     ]
    }
   ],
   "source": [
    "job_num = 1\n",
    "for job_i_commands_list in bash_scripts:\n",
    "    \n",
    "    #append all commands in a single string to be submitted as a job\n",
    "    VCF_extract_job = ''\n",
    "    for command_i in job_i_commands_list:\n",
    "        VCF_extract_job  = VCF_extract_job  + '\\n' + command_i\n",
    "    \n",
    "    #directory where you want output + error files\n",
    "    os.chdir('/n/data1/hms/dbmi/farhat/Roger/inhost_TB_dynamics_project/rolling_DB_scrape/Genotypes/O2_SLURM_logs/VCF_refpos_for_genotyping_extraction/')\n",
    "\n",
    "    job_name = 'VCF_' + str(job_num)\n",
    "\n",
    "    s = Slurm(job_name , {'partition':'short' , 'n':'1' , 't':'0-12:00:00' , 'mem-per-cpu':'2G' , 'mail-type':'FAIL' , 'mail-user':'roger_vargas@g.harvard.edu'})\n",
    "\n",
    "    #submits the job\n",
    "    job_id = s.run(VCF_extract_job)\n",
    "\n",
    "    print job_name\n",
    "    job_num += 1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "####################################################################################################################################################################################"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Run any remaining isolates through VCF subsetting procedure that don't have a corresponding subset VCF"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "####################################################################################################################################################################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "from slurmpy import Slurm\n",
    "import numpy as np\n",
    "import itertools\n",
    "import shutil"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "isolate_id_series = pd.read_pickle( '/n/data1/hms/dbmi/farhat/Roger/inhost_TB_dynamics_project/rolling_DB_scrape/Genotypes/genotypes_isolate_annotation.pkl' )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "__Check Progress of Jobs:__ Isolates that need to have Full VCF >> Subset VCF for Reference Positions of Interest"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "isolate_ids_to_run_VCF_extraction = []\n",
    "isolate_ids_completed = []\n",
    "\n",
    "for tag in isolate_id_series.values:\n",
    "\n",
    "    #create directory to store subset VCF and genotypes for each isolate\n",
    "    isolate_genotype_dir = '/n/data1/hms/dbmi/farhat/Roger/inhost_TB_dynamics_project/rolling_DB_scrape/Genotypes/Genotypes_per_isolate/' + tag\n",
    "    \n",
    "    #current.bcf file\n",
    "    current_bcf_path = isolate_genotype_dir + '/current.bcf'\n",
    "    \n",
    "    #current.bcf.csi file\n",
    "    current_bcf_csi_path = isolate_genotype_dir + '/current.bcf.csi'\n",
    "    \n",
    "    #subset VCF file\n",
    "    subset_VCF_path = isolate_genotype_dir + '/' + tag + '_ref_positions_for_genotyping.vcf'\n",
    "    \n",
    "    #check for the existence of a VCF file, if TRUE job has either started or completed\n",
    "    if os.path.exists(subset_VCF_path):\n",
    "        \n",
    "        #check to see if intermediary files exist, if they do then VCF subset task hasn't completed\n",
    "        if os.path.exists(current_bcf_path) or os.path.exists(current_bcf_csi_path):\n",
    "            \n",
    "            #remove files to start task over\n",
    "            ## os.remove(current_bcf_path)\n",
    "            ## os.remove(current_bcf_csi_path)\n",
    "            ## os.remove(subset_VCF_path)\n",
    "            \n",
    "            isolate_ids_to_run_VCF_extraction.append(tag)\n",
    "            \n",
    "        #VCF exists, intermediary files gone so task finished\n",
    "        else: \n",
    "            \n",
    "            isolate_ids_completed.append(tag)\n",
    "            \n",
    "    #VCF file doesn't exist, task never started\n",
    "    else:\n",
    "        \n",
    "        isolate_ids_to_run_VCF_extraction.append(tag)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "33873"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(isolate_ids_completed)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(isolate_ids_to_run_VCF_extraction)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Subset VCF to reference positions of interest for isolates that had Full VCF files "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "isolate_id_series = pd.Series(isolate_ids_to_run_VCF_extraction)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create BASH scripts to extract VCFs of N isolates per job"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "N_isolates_per_job = 65\n",
    "\n",
    "njobs = int( np.ceil( float( len( isolate_id_series ) ) / float(N_isolates_per_job) ) ) #number of jobs required if we split for every N isolates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "114\n"
     ]
    }
   ],
   "source": [
    "print njobs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "bash_scripts = []\n",
    "\n",
    "#genomic data directory\n",
    "rolling_DB_dir = '/n/data1/hms/dbmi/farhat/rollingDB/genomic_data/'\n",
    "\n",
    "#split jobs up into tasks of N\n",
    "for first_isolate_i_for_job in np.arange(0 , (njobs)*N_isolates_per_job , N_isolates_per_job):\n",
    "\n",
    "    last_isolate_i_for_job = first_isolate_i_for_job + N_isolates_per_job\n",
    "    \n",
    "    commands_list = []\n",
    "    commands_list.append( 'module load bcftools/1.9' ) #only need to load this once per job\n",
    "\n",
    "    for isolate_i in range(first_isolate_i_for_job , last_isolate_i_for_job): \n",
    "\n",
    "        try:\n",
    "            tag = isolate_id_series[isolate_i]\n",
    "            zipped_VCF = rolling_DB_dir + tag + '/pilon/' + tag + '_full.vcf.gz'\n",
    "            isolate_genotypes_dir = '/n/data1/hms/dbmi/farhat/Roger/inhost_TB_dynamics_project/rolling_DB_scrape/Genotypes/Genotypes_per_isolate/' + tag\n",
    "\n",
    "            commands_list.append( 'cd {}'.format(isolate_genotypes_dir) )\n",
    "            commands_list.append( 'time bcftools view {} -O b -o current.bcf'.format(zipped_VCF) )\n",
    "            commands_list.append( 'bcftools index current.bcf' )\n",
    "            commands_list.append( 'bcftools view current.bcf --regions-file /n/data1/hms/dbmi/farhat/Roger/inhost_TB_dynamics_project/rolling_DB_scrape/Genotypes/ref_positions_to_select.txt -o {0}_ref_positions_for_genotyping.vcf -O v'.format(tag) )\n",
    "            commands_list.append( 'rm current.bcf.csi current.bcf' )\n",
    "            \n",
    "        except KeyError: #applies only to the last job if it has < N isolates\n",
    "            continue\n",
    "        \n",
    "    bash_scripts.append(commands_list)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Submit each job to O2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3589690\n",
      "submitted: Submitted batch job 3589692\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_1\n",
      "VCF_2\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3589694\n",
      "submitted: Submitted batch job 3589695\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_3\n",
      "VCF_4\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3589696\n",
      "submitted: Submitted batch job 3589697\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_5\n",
      "VCF_6\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3589698\n",
      "submitted: Submitted batch job 3589701\n",
      "submitted: Submitted batch job 3589702\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_7\n",
      "VCF_8\n",
      "VCF_9\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3589703\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_10\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3589704\n",
      "submitted: Submitted batch job 3589706\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_11\n",
      "VCF_12\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3589708\n",
      "submitted: Submitted batch job 3589710\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_13\n",
      "VCF_14\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3589713\n",
      "submitted: Submitted batch job 3589715\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_15\n",
      "VCF_16\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3589718\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_17\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3589720\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_18\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3589724\n",
      "submitted: Submitted batch job 3589725\n",
      "submitted: Submitted batch job 3589726\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_19\n",
      "VCF_20\n",
      "VCF_21\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3589728\n",
      "submitted: Submitted batch job 3589731\n",
      "submitted: Submitted batch job 3589732\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_22\n",
      "VCF_23\n",
      "VCF_24\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3589733\n",
      "submitted: Submitted batch job 3589734\n",
      "submitted: Submitted batch job 3589735\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_25\n",
      "VCF_26\n",
      "VCF_27\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3589736\n",
      "submitted: Submitted batch job 3589737\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_28\n",
      "VCF_29\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3589739\n",
      "submitted: Submitted batch job 3589741\n",
      "submitted: Submitted batch job 3589742\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_30\n",
      "VCF_31\n",
      "VCF_32\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3589744\n",
      "submitted: Submitted batch job 3589745\n",
      "submitted: Submitted batch job 3589746\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_33\n",
      "VCF_34\n",
      "VCF_35\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3589747\n",
      "submitted: Submitted batch job 3589748\n",
      "submitted: Submitted batch job 3589749\n",
      "submitted: Submitted batch job 3589751\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_36\n",
      "VCF_37\n",
      "VCF_38\n",
      "VCF_39\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3589752\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_40\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3589753\n",
      "submitted: Submitted batch job 3589756\n",
      "submitted: Submitted batch job 3589757\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_41\n",
      "VCF_42\n",
      "VCF_43\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3589758\n",
      "submitted: Submitted batch job 3589760\n",
      "submitted: Submitted batch job 3589762\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_44\n",
      "VCF_45\n",
      "VCF_46\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3589764\n",
      "submitted: Submitted batch job 3589766\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_47\n",
      "VCF_48\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3589767\n",
      "submitted: Submitted batch job 3589769\n",
      "submitted: Submitted batch job 3589771\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_49\n",
      "VCF_50\n",
      "VCF_51\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3589773\n",
      "submitted: Submitted batch job 3589775\n",
      "submitted: Submitted batch job 3589777\n",
      "submitted: Submitted batch job 3589778\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_52\n",
      "VCF_53\n",
      "VCF_54\n",
      "VCF_55\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3589779\n",
      "submitted: Submitted batch job 3589780\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_56\n",
      "VCF_57\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3589782\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_58\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3589783\n",
      "submitted: Submitted batch job 3589786\n",
      "submitted: Submitted batch job 3589787\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_59\n",
      "VCF_60\n",
      "VCF_61\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3589790\n",
      "submitted: Submitted batch job 3589793\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_62\n",
      "VCF_63\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3589794\n",
      "submitted: Submitted batch job 3589795\n",
      "submitted: Submitted batch job 3589796\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_64\n",
      "VCF_65\n",
      "VCF_66\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3589797\n",
      "submitted: Submitted batch job 3589798\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_67\n",
      "VCF_68\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3589799\n",
      "submitted: Submitted batch job 3589802\n",
      "submitted: Submitted batch job 3589803\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_69\n",
      "VCF_70\n",
      "VCF_71\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3589804\n",
      "submitted: Submitted batch job 3589805\n",
      "submitted: Submitted batch job 3589806\n",
      "submitted: Submitted batch job 3589808\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_72\n",
      "VCF_73\n",
      "VCF_74\n",
      "VCF_75\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3589809\n",
      "submitted: Submitted batch job 3589810\n",
      "submitted: Submitted batch job 3589811\n",
      "submitted: Submitted batch job 3589812\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_76\n",
      "VCF_77\n",
      "VCF_78\n",
      "VCF_79\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3589813\n",
      "submitted: Submitted batch job 3589814\n",
      "submitted: Submitted batch job 3589816\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_80\n",
      "VCF_81\n",
      "VCF_82\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3589817\n",
      "submitted: Submitted batch job 3589820\n",
      "submitted: Submitted batch job 3589821\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_83\n",
      "VCF_84\n",
      "VCF_85\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3589822\n",
      "submitted: Submitted batch job 3589823\n",
      "submitted: Submitted batch job 3589826\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_86\n",
      "VCF_87\n",
      "VCF_88\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3589827\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_89\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3589830\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_90\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3589833\n",
      "submitted: Submitted batch job 3589834\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_91\n",
      "VCF_92\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3589837\n",
      "submitted: Submitted batch job 3589838\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_93\n",
      "VCF_94\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3589840\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_95\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3589841\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_96\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3589842\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_97\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3589845\n",
      "submitted: Submitted batch job 3589848\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_98\n",
      "VCF_99\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3589849\n",
      "submitted: Submitted batch job 3589850\n",
      "submitted: Submitted batch job 3589851\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_100\n",
      "VCF_101\n",
      "VCF_102\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3589852\n",
      "submitted: Submitted batch job 3589853\n",
      "submitted: Submitted batch job 3589855\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_103\n",
      "VCF_104\n",
      "VCF_105\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3589856\n",
      "submitted: Submitted batch job 3589857\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_106\n",
      "VCF_107\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3589859\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_108\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3589862\n",
      "submitted: Submitted batch job 3589863\n",
      "submitted: Submitted batch job 3589864\n",
      "submitted: Submitted batch job 3589865\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF_109\n",
      "VCF_110\n",
      "VCF_111\n",
      "VCF_112\n",
      "VCF_113\n",
      "VCF_114\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3589866\n",
      "submitted: Submitted batch job 3589867\n"
     ]
    }
   ],
   "source": [
    "job_num = 1\n",
    "for job_i_commands_list in bash_scripts:\n",
    "    \n",
    "    #append all commands in a single string to be submitted as a job\n",
    "    VCF_extract_job = ''\n",
    "    for command_i in job_i_commands_list:\n",
    "        VCF_extract_job  = VCF_extract_job  + '\\n' + command_i\n",
    "    \n",
    "    #directory where you want output + error files\n",
    "    os.chdir('/n/data1/hms/dbmi/farhat/Roger/inhost_TB_dynamics_project/rolling_DB_scrape/Genotypes/O2_SLURM_logs/VCF_refpos_for_genotyping_extraction/')\n",
    "\n",
    "    job_name = 'VCF_' + str(job_num)\n",
    "\n",
    "    s = Slurm(job_name , {'partition':'short' , 'n':'1' , 't':'0-12:00:00' , 'mem-per-cpu':'2G' , 'mail-type':'FAIL' , 'mail-user':'roger_vargas@g.harvard.edu'})\n",
    "\n",
    "    #submits the job\n",
    "    job_id = s.run(VCF_extract_job)\n",
    "\n",
    "    print job_name\n",
    "    job_num += 1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "####################################################################################################################################################################################"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## [7] This section of the notebook prepares the script that iterates over all of the VCF files in rollingDB and outputs a genotype vector."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "####################################################################################################################################################################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "import vcf\n",
    "import os\n",
    "import pandas as pd\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "isolate_ID = 'Peru4988' #INPUT - isolate ID"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Load SNV-Index annotation (for indexing genotypes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "SNV_annotation_DF = pd.read_pickle( '/n/data1/hms/dbmi/farhat/Roger/inhost_TB_dynamics_project/rolling_DB_scrape/Genotypes/genotypes_SNP_annotation.pkl' )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>11</td>\n",
       "      <td>A</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>29</td>\n",
       "      <td>C</td>\n",
       "      <td>[G]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>31</td>\n",
       "      <td>A</td>\n",
       "      <td>[C, G]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>48</td>\n",
       "      <td>C</td>\n",
       "      <td>[T]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>64</td>\n",
       "      <td>G</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   pos ref     alt\n",
       "0   11   A     [C]\n",
       "1   29   C     [G]\n",
       "2   31   A  [C, G]\n",
       "3   48   C     [T]\n",
       "4   64   G     [C]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SNV_annotation_DF.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Convert SNV annotation into dictionaries with keys = ROW indices of genotype array"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "#keys: H37Rv Reference Position , values: Genotype Row Index\n",
    "ref_pos_dict = dict( zip( list(SNV_annotation_DF.pos) , list(SNV_annotation_DF.index) ) )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Construct 'skeleton' Numpy Array that will store the supported base for all Reference Positions of interest for a single isolate in rollingDB\n",
    "\n",
    "- **0** for _A_\n",
    "- **1** for _C_\n",
    "- **2** for _G_\n",
    "- **3** for _T_\n",
    "- **9** for Bad Quality SNV calls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "#number of Reference Positions we're interested in (where High Quality SNV calls were made)\n",
    "number_of_SNVs = len(ref_pos_dict.keys()) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "146874"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "number_of_SNVs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### In constructing the Genotype Vector, we're going to iterate through all Reference Positions of interest and find the supported base from the VCF file for each position"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "#create an empty numpy array to store genotypes\n",
    "genotypes_array = np.full(number_of_SNVs , 9 , dtype = int)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([9, 9, 9, ..., 9, 9, 9])"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genotypes_array"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(146874,)"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.shape(genotypes_array)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Iterate through the subset VCF file corresponding to the isolate in rollingDB and fill in genotypes vector"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "#genomic data directory\n",
    "VCF_genotypes_dir = '/n/data1/hms/dbmi/farhat/Roger/inhost_TB_dynamics_project/rolling_DB_scrape/Genotypes/Genotypes_per_isolate/'\n",
    "\n",
    "#directory that stores files for each sequenced isolate\n",
    "directory_for_sequenced_isolate = VCF_genotypes_dir + isolate_ID + '/'\n",
    "\n",
    "#position-base dictionaries (order of output in Pilon)\n",
    "base_order = {0:'A' , 1:'C' , 2:'G', 3:'T'}\n",
    "base_order_r = {'A':0 , 'C':1 , 'G':2 , 'T':3}\n",
    "\n",
    "#load in VCF file if one was found\n",
    "vcf_reader = vcf.Reader( open(directory_for_sequenced_isolate + isolate_ID + '_ref_positions_for_genotyping.vcf' , 'r') )\n",
    "\n",
    "#iterate through each Variant Call \n",
    "for record in vcf_reader:\n",
    "    \n",
    "    #check to see if variant is SNP (length of Reference Allele is 1 base and there's only 1 alternate allele and length of alternate allele is 1 base is alternate allele exists)\n",
    "    if (len(record.REF) == 1) and (len(record.ALT) == 1) and ( (record.ALT[0] == None) or (len(str(record.ALT[0])) == 1) ):\n",
    "\n",
    "        #check to see if the call is a PASS by Pilon\n",
    "        if record.FILTER == []:\n",
    "\n",
    "            ##### Retrieve Relevant information for filtering quality of Base Call #####\n",
    "            # Mean Base Quality @ locus\n",
    "            BQ = record.INFO['BQ']\n",
    "            # Mean Mapping Quality @ locus\n",
    "            MQ = record.INFO['MQ']\n",
    "            # Number of Reads w/ Deletion \n",
    "            DC = record.INFO['DC']\n",
    "            # Number of Reads w/ Insertion\n",
    "            IC = record.INFO['IC']\n",
    "            # Depth of Valid Reads in Pileup\n",
    "            VD = record.INFO['DP']\n",
    "\n",
    "            ### Filtering Criteria\n",
    "            #---> Mean Base Quality > 20\n",
    "            #---> Mean Mapping Quality > 30\n",
    "            #---> No Reads Supporting Insertions\n",
    "            #---> No Reads Supporting Deletions\n",
    "            #---> Number of High Quality Reads >= 25\n",
    "            if (BQ > 20) and (MQ > 30) and (DC == 0) and (IC == 0) and (VD >= 25): \n",
    "\n",
    "                # Quality-Percentage - Percentage of As, Cs, Gs, Ts weighted by Q & MQ at locus\n",
    "                QP = record.INFO['QP']\n",
    "\n",
    "                if (sum(np.array(QP) >= 90) == 1): #1 Allele is supported by at least 90% of the reads, store information for this Base Call\n",
    "\n",
    "                    #create a Boolean list for the Bases that have at least 90% of support from reads\n",
    "                    supported_bases_boolean = list( np.array( QP ) >= 90 )\n",
    "\n",
    "                    #position of Base [A,C,G,T] with at least 90% of support from reads\n",
    "                    supported_bases_i = [i for i,x in enumerate(supported_bases_boolean) if x == True] #returns list of length \n",
    "                    supported_base = base_order[supported_bases_i[0]]\n",
    "\n",
    "                    #write an exception in case a SNV (with a reference position that we're not interested in) 'sneaks' past the subsetting and filtering\n",
    "                    try:\n",
    "                        #store a '0', '1', '2', or '3' in Genotype Vector (Reference Position Index) referencing the Base with at least 90% support from the reads\n",
    "                        genotypes_array[ref_pos_dict[record.POS]] = base_order_r[supported_base]\n",
    "                    except KeyError:\n",
    "                        pass"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Pickle Genotypes Array for downstream analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([0, 1, 0, ..., 2, 2, 1])"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genotypes_array"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.save(directory_for_sequenced_isolate + '/' + isolate_ID + '_genotypes' , genotypes_array , allow_pickle = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "####################################################################################################################################################################################"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## [8] This section of the notebook launches the script that was created in the previous section of the notebook and submits jobs for batches of isolates to make genotypes arrays for each isolate."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "####################################################################################################################################################################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "from slurmpy import Slurm\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "isolate_id_series = pd.read_pickle( '/n/data1/hms/dbmi/farhat/Roger/inhost_TB_dynamics_project/rolling_DB_scrape/Genotypes/genotypes_isolate_annotation.pkl' )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(33873,)"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.shape(isolate_id_series)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Extract genotypes for each isolate from VCF file containing a subset of Reference Positions from the Full VCF files"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create BASH scripts to extract genotypes of N isolates per job"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "N_isolates_per_job = 720\n",
    "\n",
    "njobs = int( np.ceil( float( len( isolate_id_series ) ) / float(N_isolates_per_job) ) ) #number of jobs required if we split for every N isolates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "48\n"
     ]
    }
   ],
   "source": [
    "print njobs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "bash_scripts = []\n",
    "\n",
    "#split jobs up into tasks of N\n",
    "for first_isolate_i_for_job in np.arange(0 , (njobs)*N_isolates_per_job , N_isolates_per_job):\n",
    "\n",
    "    last_isolate_i_for_job = first_isolate_i_for_job + N_isolates_per_job\n",
    "    \n",
    "    commands_list = []\n",
    "\n",
    "    for isolate_i in range(first_isolate_i_for_job , last_isolate_i_for_job): \n",
    "\n",
    "        try:\n",
    "            tag = isolate_id_series[isolate_i]\n",
    "\n",
    "            commands_list.append( 'python /n/data1/hms/dbmi/farhat/Roger/inhost_TB_dynamics_project/python_scripts/rollingDB_extract_HQ_SNV_genotypes.py {}'.format(tag) )\n",
    "            \n",
    "        except KeyError: #applies only to the last job if it has < N isolates\n",
    "            continue\n",
    "        \n",
    "    bash_scripts.append(commands_list)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Submit each job to O2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3793777\n",
      "submitted: Submitted batch job 3793778\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Geno1 : 3793777\n",
      "Geno2 : 3793778\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3793780\n",
      "submitted: Submitted batch job 3793781\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Geno3 : 3793780\n",
      "Geno4 : 3793781\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3793782\n",
      "submitted: Submitted batch job 3793783\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Geno5 : 3793782\n",
      "Geno6 : 3793783\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3793784\n",
      "submitted: Submitted batch job 3793785\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Geno7 : 3793784\n",
      "Geno8 : 3793785\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3793786\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Geno9 : 3793786\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3793789\n",
      "submitted: Submitted batch job 3793790\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Geno10 : 3793789\n",
      "Geno11 : 3793790\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3793791\n",
      "submitted: Submitted batch job 3793792\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Geno12 : 3793791\n",
      "Geno13 : 3793792\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3793793\n",
      "submitted: Submitted batch job 3793794\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Geno14 : 3793793\n",
      "Geno15 : 3793794\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3793796\n",
      "submitted: Submitted batch job 3793797\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Geno16 : 3793796\n",
      "Geno17 : 3793797\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3793799\n",
      "submitted: Submitted batch job 3793800\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Geno18 : 3793799\n",
      "Geno19 : 3793800\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3793801\n",
      "submitted: Submitted batch job 3793803\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Geno20 : 3793801\n",
      "Geno21 : 3793803\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3793804\n",
      "submitted: Submitted batch job 3793805\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Geno22 : 3793804\n",
      "Geno23 : 3793805\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3793807\n",
      "submitted: Submitted batch job 3793809\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Geno24 : 3793807\n",
      "Geno25 : 3793809\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3793810\n",
      "submitted: Submitted batch job 3793811\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Geno26 : 3793810\n",
      "Geno27 : 3793811\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3793812\n",
      "submitted: Submitted batch job 3793814\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Geno28 : 3793812\n",
      "Geno29 : 3793814\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3793815\n",
      "submitted: Submitted batch job 3793816\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Geno30 : 3793815\n",
      "Geno31 : 3793816\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3793817\n",
      "submitted: Submitted batch job 3793818\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Geno32 : 3793817\n",
      "Geno33 : 3793818\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3793819\n",
      "submitted: Submitted batch job 3793821\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Geno34 : 3793819\n",
      "Geno35 : 3793821\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3793822\n",
      "submitted: Submitted batch job 3793823\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Geno36 : 3793822\n",
      "Geno37 : 3793823\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3793824\n",
      "submitted: Submitted batch job 3793826\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Geno38 : 3793824\n",
      "Geno39 : 3793826\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3793827\n",
      "submitted: Submitted batch job 3793828\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Geno40 : 3793827\n",
      "Geno41 : 3793828\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3793829\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Geno42 : 3793829\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3793832\n",
      "submitted: Submitted batch job 3793833\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Geno43 : 3793832\n",
      "Geno44 : 3793833\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3793835\n",
      "submitted: Submitted batch job 3793836\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Geno45 : 3793835\n",
      "Geno46 : 3793836\n",
      "Geno47 : 3793837\n",
      "Geno48 : 3793838\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 3793837\n",
      "submitted: Submitted batch job 3793838\n"
     ]
    }
   ],
   "source": [
    "job_num = 1\n",
    "for job_i_commands_list in bash_scripts:\n",
    "    \n",
    "    #append all commands in a single string to be submitted as a job\n",
    "    genotypes_extract_job = ''\n",
    "    for command_i in job_i_commands_list:\n",
    "        genotypes_extract_job  = genotypes_extract_job  + '\\n' + command_i\n",
    "    \n",
    "    #directory where you want output + error files\n",
    "    os.chdir('/n/data1/hms/dbmi/farhat/Roger/inhost_TB_dynamics_project/rolling_DB_scrape/Genotypes/O2_SLURM_logs/make_genotypes_arrays_from_VCF/')\n",
    "\n",
    "    job_name = 'Geno' + str(job_num)\n",
    "\n",
    "    s = Slurm(job_name , {'partition':'short' , 'n':'1' , 't':'0-6:00:00' , 'mem-per-cpu':'2G' , 'mail-type':'FAIL' , 'mail-user':'roger_vargas@g.harvard.edu'})\n",
    "\n",
    "    #submits the job\n",
    "    job_id = s.run(genotypes_extract_job)\n",
    "\n",
    "    print job_name  + ' : ' +  str(job_id)\n",
    "    job_num += 1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "####################################################################################################################################################################################"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## [9] Inspect the distribution of missing values after Genotype Matrix is constructed (from genotype vectors) and filter out SNPs & Isolates"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "####################################################################################################################################################################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "import os\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib as mpl\n",
    "import matplotlib.ticker as ticker\n",
    "from itertools import compress\n",
    "from matplotlib.colors import LogNorm\n",
    "from matplotlib import gridspec\n",
    "import pickle"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Make sure all isolates have a corresponding _genotypes_ array"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "isolate_id_series = pd.read_pickle( '/n/data1/hms/dbmi/farhat/Roger/inhost_TB_dynamics_project/rolling_DB_scrape/Genotypes/genotypes_isolate_annotation.pkl' )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(33873,)"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.shape(isolate_id_series)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "__Check for Genotypes Arrays:__ Get isolate IDs that have a corresponding genotypes array"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "isolate_ids_with_genotypes = []\n",
    "isolate_ids_without_genotypes = []\n",
    "\n",
    "for tag in isolate_id_series.values:\n",
    "\n",
    "    #create directory to store subset VCF and genotypes for each isolate\n",
    "    isolate_genotype_dir = '/n/data1/hms/dbmi/farhat/Roger/inhost_TB_dynamics_project/rolling_DB_scrape/Genotypes/Genotypes_per_isolate/' + tag\n",
    "    \n",
    "    #genotypes_array\n",
    "    genotypes_array_pickle_file = isolate_genotype_dir + '/' + tag + '_genotypes.npy'\n",
    "    \n",
    "    #check for the existence of a genotypes file for this isolate\n",
    "    if os.path.exists(genotypes_array_pickle_file):\n",
    "            \n",
    "        isolate_ids_with_genotypes.append(tag)\n",
    "            \n",
    "    else:\n",
    "        \n",
    "        isolate_ids_without_genotypes.append(tag)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "33873"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(isolate_ids_with_genotypes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(isolate_ids_without_genotypes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "del isolate_ids_with_genotypes #clear variable from memory"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Concatenate genotypes arrays into a genotypes matrix for all isolates in rollingDB"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Store the __genotypes array__ for each isolate into a list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.0\n",
      "0.0295220382015\n",
      "0.059044076403\n",
      "0.0885661146046\n",
      "0.118088152806\n",
      "0.147610191008\n",
      "0.177132229209\n",
      "0.206654267411\n",
      "0.236176305612\n",
      "0.265698343814\n",
      "0.295220382015\n",
      "0.324742420217\n",
      "0.354264458418\n",
      "0.38378649662\n",
      "0.413308534821\n",
      "0.442830573023\n",
      "0.472352611224\n",
      "0.501874649426\n",
      "0.531396687627\n",
      "0.560918725829\n",
      "0.59044076403\n",
      "0.619962802232\n",
      "0.649484840433\n",
      "0.679006878635\n",
      "0.708528916836\n",
      "0.738050955038\n",
      "0.767572993239\n",
      "0.797095031441\n",
      "0.826617069642\n",
      "0.856139107844\n",
      "0.885661146046\n",
      "0.915183184247\n",
      "0.944705222449\n",
      "0.97422726065\n"
     ]
    }
   ],
   "source": [
    "genotypes_directory = '/n/data1/hms/dbmi/farhat/Roger/inhost_TB_dynamics_project/rolling_DB_scrape/Genotypes/Genotypes_per_isolate/'\n",
    "genotype_arrays_per_isolate_list = []\n",
    "\n",
    "#iterate through the genotypes arrays for each isolate\n",
    "isolate_i = 0\n",
    "for isolate_ID in isolate_id_series:\n",
    "    \n",
    "    #load genotypes array\n",
    "    genotypes_array = np.load(genotypes_directory + isolate_ID + '/' + isolate_ID + '_genotypes.npy')\n",
    "    \n",
    "    #store genotypes array for this isolate in a list with all other genotypes arrays\n",
    "    genotype_arrays_per_isolate_list.append(genotypes_array)\n",
    "    \n",
    "    if isolate_i % 1000 == 0:\n",
    "        print float(isolate_i) / float(len(isolate_id_series))\n",
    "    isolate_i += 1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Stack genotype arrays for all isolates to create one __genotypes matrix__ for all isolates "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "genotypes_matrix = np.stack(genotype_arrays_per_isolate_list , axis = -1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(146874, 33873)"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.shape(genotypes_matrix)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[0, 9, 9, ..., 0, 0, 0],\n",
       "       [1, 9, 9, ..., 1, 1, 1],\n",
       "       [0, 9, 0, ..., 0, 0, 0],\n",
       "       ...,\n",
       "       [2, 9, 9, ..., 2, 2, 2],\n",
       "       [2, 9, 9, ..., 2, 2, 2],\n",
       "       [1, 9, 9, ..., 1, 1, 1]])"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genotypes_matrix"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Load __SNP annotation file__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "SNP_annotation_DF = pd.read_pickle('/n/data1/hms/dbmi/farhat/Roger/inhost_TB_dynamics_project/rolling_DB_scrape/Genotypes/genotypes_SNP_annotation.pkl')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>11</td>\n",
       "      <td>A</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>29</td>\n",
       "      <td>C</td>\n",
       "      <td>[G]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>31</td>\n",
       "      <td>A</td>\n",
       "      <td>[C, G]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>48</td>\n",
       "      <td>C</td>\n",
       "      <td>[T]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>64</td>\n",
       "      <td>G</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   pos ref     alt\n",
       "0   11   A     [C]\n",
       "1   29   C     [G]\n",
       "2   31   A  [C, G]\n",
       "3   48   C     [T]\n",
       "4   64   G     [C]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SNP_annotation_DF.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(146874, 3)"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.shape(SNP_annotation_DF)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "del genotype_arrays_per_isolate_list #clear variable from memory"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## [1] Drop SNP sites that correspond that (1) are located in Mobile Genetic Element (MGE) regions and/or (2) have low Empirical Base Recall (EBR) Scores "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### (Filter 1)  *FILTER* out SNPs with low EBR scores"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Load list of low EBR score Reference positions to Drop (IL mapping/variant calling: sites with empirical score < 0.80)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('/n/data1/hms/dbmi/farhat/Roger/inhost_TB_dynamics_project/rolling_DB_scrape/Genotypes/H37Rv_sites_with_low_EBR_to_drop.pkl', 'rb') as f:\n",
    "    H37Rv_positions_to_drop = pickle.load(f)\n",
    "    \n",
    "#convert to set, faster to check\n",
    "H37Rv_positions_to_drop = set(H37Rv_positions_to_drop)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Filter out sites with low EBR scores"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Drop unique SNPs from the __GENOTYPE MATRIX__ and __SNP ANNOTATION FILE__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "#construct boolean filter\n",
    "SNPs_to_keep_filter = []\n",
    "for H37Rv_pos_i in SNP_annotation_DF.pos:\n",
    "    \n",
    "    if H37Rv_pos_i not in H37Rv_positions_to_drop:\n",
    "        \n",
    "        SNPs_to_keep_filter.append(True)\n",
    "        \n",
    "    elif H37Rv_pos_i in H37Rv_positions_to_drop:\n",
    "        \n",
    "        SNPs_to_keep_filter.append(False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "#filter Genotypes Matrix\n",
    "genotypes_matrix = genotypes_matrix[SNPs_to_keep_filter , :]\n",
    "\n",
    "#filter SNP annotation file\n",
    "SNP_annotation_DF = SNP_annotation_DF[SNPs_to_keep_filter]\n",
    "SNP_annotation_DF.reset_index(drop = True , inplace = True) #re-index new filtered SNP annotation DF (so new index matches indexing of genotypes matrix rows)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2348"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(SNPs_to_keep_filter) - sum(SNPs_to_keep_filter) #number of sites dropped by this filter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(144526, 33873)"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.shape(genotypes_matrix)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>11</td>\n",
       "      <td>A</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>29</td>\n",
       "      <td>C</td>\n",
       "      <td>[G]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>31</td>\n",
       "      <td>A</td>\n",
       "      <td>[C, G]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>48</td>\n",
       "      <td>C</td>\n",
       "      <td>[T]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>64</td>\n",
       "      <td>G</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   pos ref     alt\n",
       "0   11   A     [C]\n",
       "1   29   C     [G]\n",
       "2   31   A  [C, G]\n",
       "3   48   C     [T]\n",
       "4   64   G     [C]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SNP_annotation_DF.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(144526, 3)"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.shape(SNP_annotation_DF) #Annotation for Rows of Genotypes Matrix"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### (Filter 2) *FILTER* out SNPs in MGE regions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Collect list of MGE Reference Positions to Drop"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "#csv files with category for each gene\n",
    "gene_categories_df = pd.read_csv('/n/data1/hms/dbmi/farhat/Roger/inhost_TB_dynamics_project/CSV_files/gene_categories/gene_categories.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>Gene_Category</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Rv0001</td>\n",
       "      <td>Rv0001</td>\n",
       "      <td>Essential</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Rv0002</td>\n",
       "      <td>Rv0002</td>\n",
       "      <td>Non-Essential</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Rv0003</td>\n",
       "      <td>Rv0003</td>\n",
       "      <td>Non-Essential</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Rv0004</td>\n",
       "      <td>Rv0004</td>\n",
       "      <td>Non-Essential</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Rv0005</td>\n",
       "      <td>Rv0005</td>\n",
       "      <td>Antibiotic Resistance</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     name gene_id          Gene_Category\n",
       "0  Rv0001  Rv0001              Essential\n",
       "1  Rv0002  Rv0002          Non-Essential\n",
       "2  Rv0003  Rv0003          Non-Essential\n",
       "3  Rv0004  Rv0004          Non-Essential\n",
       "4  Rv0005  Rv0005  Antibiotic Resistance"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_categories_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "#reference genome annotation\n",
    "reference_genome_annotation = pd.read_csv('/n/data1/hms/dbmi/farhat/Roger/inhost_TB_dynamics_project/H37Rv/h37rv_genome_summary.txt', '\\t').set_index('name')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Collect positions for **MGEs**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "H37Rv_positions_MGE_genes = []\n",
    "\n",
    "for gene_id in gene_categories_df[gene_categories_df.Gene_Category == 'Mobile Genetic Element'].gene_id:\n",
    "    \n",
    "    #get reference positions tht corresond to each gene\n",
    "    gene_id_info = reference_genome_annotation.loc[gene_id, :]\n",
    "    gene_id = gene_id_info.name\n",
    "\n",
    "    chrom_start = gene_id_info.chromStart\n",
    "    chrom_end = gene_id_info.chromEnd\n",
    "\n",
    "    #find the position of the first base relative to H37Rv in 5' -> 3'\n",
    "    H37Rv_start = min(chrom_start , chrom_end)\n",
    "    H37Rv_end = max(chrom_start , chrom_end)\n",
    "    \n",
    "    #store all corresponding H37Rv Reference Positions to gene_id in list of reference positions to scan\n",
    "    H37Rv_positions_MGE_genes.append([H37Rv_start+1 , H37Rv_end+1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "108"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(H37Rv_positions_MGE_genes)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### FILTER out sites in MGE regions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "#construct boolean filter\n",
    "SNPs_to_keep_filter = []\n",
    "for H37Rv_pos_i in SNP_annotation_DF.pos:\n",
    "\n",
    "    bad_pos_i = False\n",
    "\n",
    "    for problem_region in H37Rv_positions_MGE_genes:\n",
    "\n",
    "        if (H37Rv_pos_i > problem_region[0]) and (H37Rv_pos_i < problem_region[1]):\n",
    "            bad_pos_i = True\n",
    "            break\n",
    "\n",
    "    if bad_pos_i == False:\n",
    "        SNPs_to_keep_filter.append(True)\n",
    "    else:\n",
    "        SNPs_to_keep_filter.append(False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "#filter Genotypes Matrix\n",
    "genotypes_matrix = genotypes_matrix[SNPs_to_keep_filter , :]\n",
    "\n",
    "#filter SNP annotation file\n",
    "SNP_annotation_DF = SNP_annotation_DF[SNPs_to_keep_filter]\n",
    "SNP_annotation_DF.reset_index(drop = True , inplace = True) #re-index new filtered SNP annotation DF (so new index matches indexing of genotypes matrix rows)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1509"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(SNPs_to_keep_filter) - sum(SNPs_to_keep_filter) #number of sites dropped by this filter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(143017, 33873)"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.shape(genotypes_matrix)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>11</td>\n",
       "      <td>A</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>29</td>\n",
       "      <td>C</td>\n",
       "      <td>[G]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>31</td>\n",
       "      <td>A</td>\n",
       "      <td>[C, G]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>48</td>\n",
       "      <td>C</td>\n",
       "      <td>[T]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>64</td>\n",
       "      <td>G</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   pos ref     alt\n",
       "0   11   A     [C]\n",
       "1   29   C     [G]\n",
       "2   31   A  [C, G]\n",
       "3   48   C     [T]\n",
       "4   64   G     [C]"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SNP_annotation_DF.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(143017, 3)"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.shape(SNP_annotation_DF) #Annotation for Rows of Genotypes Matrix"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### (Filter 3) *FILTER* out SNPs that are missing in $> 25\\%$ of isolates"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Investigate the Number of Missing Calls for each Unique SNP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.style.use('ggplot')\n",
    "plt.rcParams['lines.linewidth']=3\n",
    "plt.rcParams['axes.facecolor']='1.0'\n",
    "plt.rcParams['xtick.color']='black'\n",
    "plt.rcParams['axes.grid']=True\n",
    "plt.rcParams['axes.edgecolor']='black'\n",
    "plt.rcParams['grid.color']= '1.0'\n",
    "plt.rcParams.update({'font.size': 12})"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Look for SNPs with an abundance of missing values, drop SNPs that have missing values $> 25\\%$ of the isolates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [],
   "source": [
    "missing_values = genotypes_matrix == 9\n",
    "missing_values_per_SNP = missing_values.sum(axis = 1) #will range from 0 - 33,873"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [],
   "source": [
    "#impose a threshold for maximum number of missing values per SNP (across all isolates)\n",
    "max_proportion_missing = 0.25\n",
    "number_of_isolates = float(np.shape(genotypes_matrix)[1])\n",
    "\n",
    "max_number_values_missing_per_SNP = np.floor( max_proportion_missing * number_of_isolates )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "8468.0\n"
     ]
    }
   ],
   "source": [
    "print max_number_values_missing_per_SNP #that we're going to allow"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAcgAAAFuCAYAAAAF5IgZAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMi4zLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvIxREBQAAIABJREFUeJzt3XtUVOX+P/D3CAIj4KgIJKSgJiqoaFJ4wVuhqCgKaF7SJD1L8VgH65xuR49aclIrf2ay5Fip5fUcC1Mr08SFaFgWKhxFU1CRwEsqJhcZdIbn94eH+Ypshj3MFeb9Wou1ZD8zz/7sB5wPz2U/WyGEECAiIqIamlk7ACIiIlvEBElERCSBCZKIiEgCEyQREZEEJkgiIiIJTJBEREQSbD5B5ufnw9PTE0OHDsXQoUNx48YNa4dERER2wNHaAcgxZMgQfPnll9YOg4iI7IjN9yABICMjA4MGDcLf//53cF8DIiKyBIslyKSkJISEhMDZ2RlxcXE1yoqLixEdHQ1XV1f4+flh27ZturJ27dohLy8Phw8fxu+//46dO3daKmQiIrJjFkuQPj4+WLhwIWbOnFmrbN68eXBycsL169exdetWzJ07Fzk5OQAAZ2dnuLq6QqFQICYmBtnZ2ZYKmYiI7JjFEmRMTAzGjx8PDw+PGsfLy8uRkpKCpUuXws3NDWFhYYiKisLmzZsBACUlJbrXHjlyBE888US951IoFKYNnoiI7I7V5yDPnz8PBwcHBAQE6I4FBwfrepDp6eno27cvBg0ahKKiIkydOtVaoRIRkR2x+irWsrIyqFSqGsdUKhVKS0sBAGPHjsXYsWPrrSc1NRWpqalmiZGIiOyP1ROkm5tbjWFU4MGwqru7u0H1hIeHIzw8HACwYsUKk8VHRET2yeoJMiAgABqNBrm5uejSpQsAIDs7G0FBQQbXlZmZiePHj5s6RCIiskMWm4PUaDRQq9XQarXQarVQq9XQaDRwdXVFTEwMFi1ahPLycmRkZGD37t2YPn26wecICQnBnDlzzBA9ERHZG4slyMTERCiVSixfvhxbtmyBUqlEYmIiAGDt2rWoqKiAl5cXpkyZguTk5Ab1IImIiExFIZrQ1jTVQ6zx8fHccYeIiIzSpBJkNYVCwQRJRERGsfp9kERERLaICZKIiEiC1W/zMCXe5kFERKbCOUgiIiIJHGIlIiKSwATZQGq1ukFlRETUOHAOsoFcXFzqfKwWh3eJiBo/zkEaeR4pTbBJiYjsDodYiYiIJDBBEhERSeAcJBERkQTOQRp5HilNsEmJiOwOh1iJiIgkMEESERFJYIIkIiKSwARJREQkgatY9VCr1XBxcTFZfURE1HhwFauMuqQIIbiKlYioCeMQKxERkQQmSCIiIglMkERERBKYIImIiCQwQRIREUngbR5EREQSeJuHjLqk8DYPIqKmjUOsREREEpggiYiIJDBBEhERSWCCJCIiksAESUREJIEJkoiISAITJBERkQQmSCIiIglMkGagVquNKiciIuvjVnNm4OLiUucuOwB32iEiagy41ZyMuqTUt9UcEyQRUePGIVYiIiIJTJBEREQSmCCJiIgkMEESERFJYIIkIiKSwARJREQkgQmSiIhIAhMkERGRBCZIIiIiCY0mQW7fvh2enp7WDoOIiOxEo0iQVVVV+PLLL9G+fXtrh0JERHaiUSTIbdu2YcKECWjWrFGES0RETYDFMk5SUhJCQkLg7OyMuLi4GmXFxcWIjo6Gq6sr/Pz8sG3bNl2ZVqvFjh07MGnSJEuFSkREZLnHXfn4+GDhwoXYv38/KioqapTNmzcPTk5OuH79OrKyshAZGYng4GAEBQVhy5YteO6559h7JCIii7JY1omJicH48ePh4eFR43h5eTlSUlKwdOlSuLm5ISwsDFFRUdi8eTMA4MyZM9i0aRNGjhyJ3Nxc/OUvf7FUyEREZMes/sDk8+fPw8HBAQEBAbpjwcHBSE9PBwCsWLFCdzwkJAQfffSRZD2pqalITU01b7BERGQ3rJ4gy8rKoFKpahxTqVQoLS2t9drMzMw66wkPD0d4eDiAmkmViIioIaw+sefm5oaSkpIax0pKSuDu7m6liIiIiGygBxkQEACNRoPc3Fx06dIFAJCdnY2goCCD68rMzMTx48dNHSIREdkhi/UgNRoN1Go1tFottFot1Go1NBoNXF1dERMTg0WLFqG8vBwZGRnYvXs3pk+fbvA5QkJCMGfOHDNET0RE9sZiCTIxMRFKpRLLly/Hli1boFQqkZiYCABYu3YtKioq4OXlhSlTpiA5OblBPUgiIiJTUQghhLWDMJXqIdb4+HiY6rIUCoXkcSFEg8qqy4mIyLY1qQRZTaFQMEESEZFRrL6K1R6p1eoGlRERkeVYfRWrPXJxcdHb+yQiIutrUgmSt3kQEZGpcA5SRl1SjJ2DZA+SiMi2cQ6SiIhIAodYiYiIJHCIVUZdUjjESkTUtHGIlYiISIKsBHn06FFs3rwZQggcPnwYw4cPx3PPPYeCggJzx0dERGQVsoZYBwwYgNLSUpw6dQodOnRAYWEhFAoFxowZg927d1siToNwiJWIiIwlqwd57tw5PPnkk7h06RIKCwuxevVqDBw4EBkZGeaOzyCZmZlYt26dtcMgIqImQFaCrKiogLOzM86cOQOFQoHnnnsOPXv2RHl5ubnjMwgfd0VERKYi6zaPjh07Yvv27di3bx98fHzg7e2NK1euwNvb29zxERERWYWsHuSCBQtQWVmJK1euYPHixbh37x4OHTqE0NBQc8dHRERkFbLvg1Sr1dBqtXB1dTV3TEbjIh0iIjJWvT3I9evXIzY2FvPnz8fVq1ctEZNd46OwiIhsg945yKSkJCQkJOi+37lzJ/Ly8tCyZUuzB9YQTWGrOT4Ki4jINugdYu3RowfOnj2LyMhInDt3Dnl5efj4448xa9YsS8ZosMY+xMoESURkfXp7kBcvXkRsbCx27NiBoqIitG/fHpcuXbJUbERERFajdw5SrVbj8ccfBwD4+voCAO7evWv+qIiIiKys3vsgv/76a+Tn50t+r1AokJKSYq7YiIiIrEbvHGSzZvoXuSoUCmi1WpMHZSzOQRIRkbH09iAXL15sqTiIiIhsSpN6YHL1bR7x8fHsQRIRkVH0JsjDhw/XW8HgwYNNGpApcIiViIiMVe8cpL4PeoVCAY1GY5bAjMEESURExtI7B9myZUvJD+uSkhJ+WFuBWq2Gi4uLwWVERGQ4vctU//jjD9y+fVv3tX//fvTr1w9CCLi4uOAvf/mLpeIk/N82dFJfTI5ERKYl63FXJ06cwNixY9G/f38cOnQI8+bNQ15eHlatWmXu+IiIiKxC7xBrdnY2Fi9ejK+//hrNmzfHnDlz8Pe//123uw4REVFTpXeRjoODw4MXKRQYN24c/Pz8ar5ZocDKlSvNG2EDNOVFOlzAQ0RkGbJ30nn0g7n6w5o76RhWZsx7mSCJiCxH7xDrjBkzLBUHERGRTWlSO+lUYw+SiIiMVe/TPB6VlZWF0tJS9O7dG+7u7uaIqcGqt5ojIiIylt4e5LJly3Do0CH85z//QatWrfCnP/0JGzduBAB4enri22+/Rd++fS0WrFzsQRIRkbH0Jsjg4GA4ODjgxIkTOHv2LIKCggAArVq1wh9//IGIiAh89913FgtWLiZIIiIylt6NAn777Tf06tULALBv3z4AQGRkJIqLizFs2DAOZxIRUZOlN0Gq1Wo4OzsDAH788UcoFApER0cDAAICAlBWVmb+CImIiKxA7yIdf39/7Nq1C4GBgboe5LBhwwAA+fn58PT0NH+EJAs3MiciMi29Pch58+bhxo0bePXVV1FWVoYRI0agY8eOuHPnDo4cOWKTC3TsFTcyJyIyLb09yHnz5sHX1xdpaWlo164dXnrpJQDApUuXMHfuXIwaNcoiQRIREVkaNwqQUZcUW1zFyhWuRESmI+txV0RERPaGCZKIiEiCwVvNWdr169cRHR2N5s2bw8HBAVu3bkW7du2sHRYRETVxNj8HqdVqoVAo0KxZM3z22WcoLCzEwoUL9b6Hc5C1y4iIyDCyh1gPHz6MqVOnIjQ0FEVFRXjnnXfw888/mzM2AA8e2lz9XMrS0lLddndERETmJCtB7t69G8888wz+/e9/IzMzE97e3khOTsZ7770n+0RJSUkICQmBs7Mz4uLiapQVFxcjOjoarq6u8PPzw7Zt22qUZ2VlITQ0FElJSXjyySdln5OIiKihZCXId955B23atMGQIUMAAI6OjggLC8NPP/0k+0Q+Pj5YuHAhZs6cWats3rx5cHJywvXr17F161bMnTsXOTk5uvLevXvj2LFjWLp0KZYtWyb7nPSAWq1uUBkRkT2TlSDPnTuH0aNHo3fv3rpjbdu2RXFxsewTxcTEYPz48fDw8KhxvLy8HCkpKVi6dCnc3NwQFhaGqKgobN68GQBQWVmpe61KpUKLFi1kn5Me4C47RESGk7WKtV27dsjJyUH//v0BPBgSPXDgANq3b290AOfPn4eDgwMCAgJ0x4KDg5Geng4AOHHiBN544w04ODjAxcUFGzZskKwnNTUVqampRsdDREQEyEyQsbGxeO+995CVlQUA8PX1xb179/Dmm28aHUBZWRlUKlWNYyqVCqWlpQCA/v374/Dhw/XWEx4ejvDwcADAihUrjI6LiIjsm6wh1sWLF2PSpEnQarUQQqCyshITJ07EggULjA7Azc0NJSUlNY6VlJTA3d3d6LqJiIgaSlYPUqlUYvv27VizZg3y8/Ph7++Ptm3bmiSAgIAAaDQa5ObmokuXLgCA7OzsBt3OkZmZyYc4ExGRScjaKEDfEOfgwYNlnUij0UCj0eDtt99GYWEhPvnkEzg6OsLR0RGTJ0+GQqHAp59+iqysLIwePRpHjx5t8D2P3CjAsDIiIqpNVoJs1qyZ5AesQqGARqORdaIlS5bg7bffrnFs8eLFWLJkCYqLizFz5kwcOHAAHh4eWL58OaZOnSrzEmpjgjSsjIiIapOVIP39/XUfsFqtFteuXYNGo4Gfnx8uXbpk9iDlqh5ijY+PZ4I0oIyIiGpr0F6st27dQmRkJCIiImr1Cm0Be5CGlRERUW0NetyVh4cHQkJCsHHjRlPHQ0REZBNkrWJ95plnanxfUlKCrKysWrviWBtXsRIRkanIXqRT640KBVasWIG//e1vZgnMGBxiNayMiIhqk9WDfHQo1dXVFcHBwbr7FomIiJoam39gckOwB2lYGRER1SarBxkTE1PvaxQKBVJSUowOyBicgzScWq2u84ke+sqIiJo6gzYKePilUt9rtVrzRGkg9iBNV0ZEZK9k9SBfeuklJCcnY9y4cQgICMC5c+ewZ88ezJ071+ZWshIREZmCrB7kmDFj4Obmhn//+9+6Y5MmTcLdu3fx9ddfmzXAhmAP0nRlRET2SlYPMi0tDT169EBlZSWcnZ2hVquRn5+PnJwcc8dHRERkFbISZM+ePfHLL7/A29sbjz/+OAoLC1FaWoqnn37a3PEZhIt0iIjIVGQNsZ45cwZjx46tsTF5p06dsGfPHgQGBpo1wIbgEKvpyoiI7JXs+yC1Wi1+/PFHXLlyBb6+vujXrx8cHBzMHV+DMEGaroyIyF7VmSALCgrQsmVLtGrVCgUFBXVW0KFDB7MF11BMkKYrIyKyV3UmSAcHB8yfPx8rV640yQOTLYkJ0nRlRET2qs7HXQkhanxAVn//8FdVVZVFgiT7ISoqGlRGRGRqda5ifTj5NZZEyFWsplXfVnPm2IpOoVQCdfRoFezREpEFcbNyGXVJsZch1vquwyzqOmfT+1UlIhtW5xDrw/773/9iyJAhaNmyJRwcHHRfjo6ybqMkMg21us4iDr8SkanJ6kEGBwfj1KlTkmW2OPzKHqT5y6rLzUJfD5K9SyKyEFldwLy8PAQFBWHNmjVo1aqV3g9NIiKipkBWgnz22WfRvHlzDB061MzhUGPCZ0kSUVMmK0H6+vri448/RlRUFHr37l1j7nHRokVmC45sm4uLC++hJKImS/YDk3Vv+N8HYvX8lK08JBn4v9s84uPjOQdp5jI5720wzkESkQ2QlSDj4uLq/CDcuHGjyYMyFhfpmL9MznvrUu/wKxMkEdkA3gcpoy4pTJD6yysqKvQmQb2JlQmSiGyA3jnIw4cP1zrWokULdO7cGa1btzZbUNT4cX6SiBo7vT3IujYpB4CJEyfik08+gbu7u9mCayj2IM1fZs5zsgdJRLZAbw+yZcuWtT7I7t27h4qKCnzxxRfw9fXFypUrzRogERGRNTRoDvLkyZMYNmwYPD09kZuba464jMIepPnLzHlO9iCJyBbI2ov1UX369EFwcDCuXLli6niIiIhsQoMS5KlTp5CdnQ0fHx9Tx0NERGQT9M5BtmnTptax6jlIAHjxxRfNExWRodRqoI7bSkRFxYPnTBIRGUBvgvzjjz/qLBs/fjzefvttkwdkDD4w2Y65uPBBy0RkUnoX6aSnp9c6plQq0blzZ3h4eJg1MGNwkY75y8x5zgYv0rHwAh59PVP2WokaP+6kI6MuKbaYWJggTZ8g6010dZ2zooJDvkSNnKyneRCZlFrdaHbTUSiVDUu8HPIlavQatIqVyCjVyUPqixpE/G/hXEPLiag29iCJmgC9PV2w10rUEHX2IF9//XV88803AIBNmzYhMzPTYkERmZRarbfY4r0rPfGwp0dkO+pMkB988AEOHjwI4MHzILdv326xoIhMSt+QrkLxf70vSw356omHi3eIbEedQ6wqlQpbt27F7du3AQD79+/X/buaQqHA+vXrzRshERGRFdR5m0dcXBw2bdr04EV13DahUCig1WrNG2ED8DYP85cZW69ZbuVoSJk1zqmvzJjbQ+q7RiIySJ09yA0bNiAyMhLnzp3DokWLEBoaipEjR1oyNiL7w9tDiGxGnQmyWbNmmDhxIgBAq9WiX79+iIiIsFhgRERE1iR7J51Lly5h/fr1uHz5Mvz9/fHiiy+iU6dO5o4PAPDjjz/i1VdfhZOTE3x8fLBp0yY0b968ztdziNX8ZcbWazNDmtY4p7l2BNJ3jdzZh8hgshLkzz//jGeffRZ3797VJR5XV1ccPHgQTz/9tNmDvHLlClq3bg2lUokFCxagT58+mDBhQp2vZ4I0f5mx9dpM0rHGOa2RIPmwaSKDydpJ580330R5eTnGjRuHN998E+PHj0d5eTneeustc8cHAPDx8YHyf3/hOjo6olkzbgBERETmJSvTZGVlITY2Fjt37sS7776LnTt3IjY2FidPnjToZElJSQgJCYGzszPi4uJqlBUXFyM6Ohqurq7w8/PDtm3bar3/0qVL+O677zBmzBiDzktERGQoWVvNubi44ObNmzWO3bx5Ey51zGnUxcfHBwsXLsT+/ft1D12uNm/ePDg5OeH69evIyspCZGQkgoODERQUBAAoKSnBjBkzsHnzZjg5ORl0XiIiIkPJSpADBw7Ezp070aVLF3Tt2hXnz5/HhQsXEBsba9DJYmJiADx4sHFhYaHueHl5OVJSUnD69Gm4ubkhLCwMUVFR2Lx5M5YvXw6NRoMpU6ZgyZIl6Nq1q0HnJCIiaghZCfKDDz7A8ePHceHCBVy4cAEA4O/vj/fee88kQZw/fx4ODg4ICAjQHQsODtY9sHn79u04duwY3nnnHbzzzjuYO3cuJk2aVKOO1NRUpKammiQeIpukVte5EpWITE9WgvTz88Pp06fx7bffIj8/H/7+/oiMjESLFi1MEkRZWRlUKlWNYyqVCqWlpQCA6dOnY/r06XrrCA8PR3h4OABgxYoVJomLyKbo2USAK1GJTE/2465atGih2zjA1Nzc3FBSUlLjWElJCdzd3c1yPiKSR989krx/kpo6m3geZEBAADQaDXJzc9GlSxcAQHZ2tm6BjlyZmZk4fvy4OUIkskv6njPJre+oqbPoDYUajQZqtRparRZarRZqtRoajQaurq6IiYnBokWLUF5ejoyMDOzevbveYdVHhYSEYM6cOWaKnoiI7IlFE2RiYiKUSiWWL1+OLVu2QKlUIjExEQCwdu1aVFRUwMvLC1OmTEFycrLBPUgiaoB6HihNZK/q3WpOo9EgIiICAwYMwNKlSy0VV4NUD7HGx8dzqzkzlxlbr81s32aNczalayRqwurtQTo6OuLXX3/FtWvXLBGPUTjESkREpiJriHXRokXYtWsX0tPTcf/+fXPHREREZHWynubRrFkzyeEyhUIBjUZjlsAagkOsHGI1uMwa52xK10jUhMlOkHWpqqoyaUCmwMddmb/M2HrtInnYSpk5z0nUhMm6D/LSpUvmjoOImhBuMEBNgeyt5oAHe6bm5eVh9OjRZg2KiBo3bjBATYGsRTrFxcUIDw9H9+7dERUVhfz8fDg6OuIf//iHueMzSGZmJtatW2ftMIiIqAmQNQc5a9YsbNy4Ec7Ozrh37x60Wi3CwsJQXl5u8EOTLYFzkOYvM7Zeu5ifs5Uyc55TH85dUiMnqwe5b98+DB06FPHx8bpjgYGBuHjxotkCIyIisiZZCbKiogLt2rWrcezmzZtwcnIyS1BERETWJitB9urVC9988w2OHTsGAHjttdfw9ddfIzg42KzBERERWYusOcgffvgB4eHhuHfvnu5Y8+bNcfDgQYSFhZk1QENwowDOQRpcZo1zNqVr1IdzkNTIyUqQAHD69GkkJyfj8uXL8Pf3x+zZs9GrVy9zx9cgXKRj/jJj67WL5GErZeY8pz51va+iAnBxka6S90iSDZGdIAHg7t27KCoqgq+vL1q0aGHOuIzCBGn+MmPrtYvkYStl5jynPtydhxo5WXOQt2/fxsSJE+Hu7o5u3brB3d0dEydORHFxsbnjIyIisgpZO+nMmjULu3btqnFs586d0Gq12Llzp1kCIyIbp1bXOVRK1BTI6kGmpqaiY8eOyMrKglqtxsmTJ+Hn54fU1FRzx2cQ7qRDZEEuLg+GSqW+iJoAWXOQffr0Qd++ffHpp5/qjs2aNQsnT57EiRMnzBpgQ3AO0vxlxtZrF/NztlJma/FwDpIaiTqHWA8fPqz79wsvvIC3334bvXv3Rrdu3XD27FmkpKRgyZIlloiRiIjI4ursQdb1kORqQgg0a9bMph6YXI09SPOXGVuvzfRYrHFOXiN7kNQo1NmD7NChg94PRyIioqaszgSZn59vwTCIiIhsi6zbPKrdunUL5eXlNY516NDBpAERERHZAlkJ8sCBA5g1axaKiopqHFcoFDY5B0lERGQsWQkyPj4ehYWFtY6baiGMqVRvVk5ERGQsWfdBtmrVCv3798cXX3wBNzc3S8RlFK5iNX+ZsfXazKpJa5yT18hVrNQoyNpJ5+WXX0ZBQQGKiopsrtdIRERkDrISZGxsLK5cuYLAwEA4OjrCwcEBDg4OcHQ0aI0PERFRoyErwz3//PO4c+dOrePsTRIRUVMlK0EWFBTgqaeewnvvvYdWrVqZOyYislf1PCGED1QmS5KVIGfPno2ff/4Z/fv3h5OTk7ljIiJ7Vf2EkDooOGpFFiQrQR48eBCnT5+Gp6cn/P394eDgAODBCk/eVkFERE2RrNs8mjWTXsujUCig1WpNHpSxeJuH+cuMrddmbiuwxjl5jcZdB5GFyOpBbty40dxxEBER2RRZCXLGjBnmjsMkuJMOERGZiqwh1pkzZ0q/WaHA+vXrTR6UsTjEav4yY+u1uWE7W4mH18ghVrIZsucgH0461f/mHCQTJBNkIyiztXiYIKmRkDXEunjxYt2/tVotTp06hT179tTZsyQiImrsDE6Q1WbPnl3r8VdERERNheyddB5WUlKCc+fOITs72yxBERERWZusBNmxY0fJ4z179jRpMERERLZC1tM8hBA1vpRKJQYMGIDPPvvMzOERERFZh6weZFVVlbnjICIyG32bnHMDdKoLH+hIRE2eQqms8/YRboBOddGbIKs3Ja+LQqGARqMxaUBERES2QO8c5KNzj1JfREQWo1bXWSQqKiwYCJmLvp+jpX/GenuQJ0+erPF9WVkZkpKSsGPHDggh0Lt3b7MGBwB37tzB8OHDcebMGfz000/o0aOH2c9JRDZKz/MiOVTaNNjScLjeHmRwcDCCg4PRtWtXpKamIjY2Fv/5z3/Qo0cPfPnllxbZGLxFixb49ttvMWHCBLOfi4gaMT29y4a+j71S+6a3B1lZWYnk5GSsWLECv//+O7p37441a9Zg4sSJlooPzZs3h6enp8XOR0SNlJ7epd49XNkrpTro7UF27twZf/3rX3Hjxg0899xzSExMhLOzM/bs2aP7kispKQkhISFwdnZGXFxcjbLi4mJER0fD1dUVfn5+2LZtW4MuhojIpNi7tGt6e5BXrlzRPRljx44d2LFjR41yQ1ax+vj4YOHChdi/fz8qHvnFmjdvHpycnHD9+nVkZWUhMjISwcHBCAoKMvByiIhMiL1Lu6Y3QXbo0EHv444MERMTA+DBQ40LCwt1x8vLy5GSkoLTp0/Dzc0NYWFhiIqKwubNm7F8+XKTnJuIiMhQehNkfn6+2QM4f/48HBwcEBAQoDsWHByM9PR03fejR49GVlYWzp07hzlz5tQaogWA1NRUpKammj1eIiIyTmPZvcjqO+mUlZVBpVLVOKZSqVBaWqr7fu/evfXWEx4ejvDwcADAihUrTBskERGZjL5bOWzpodiyNis3Jzc3N5SUlNQ4VlJSAnd3dytFREREZAM9yICAAGg0GuTm5qJLly4AgOzs7AYt0MnMzLTIvZlERNT0WawHqdFooFarodVqodVqoVarodFo4OrqipiYGCxatAjl5eXIyMjA7t27MX36dIPPERISgjlz5pgheiIisjcWS5CJiYlQKpVYvnw5tmzZAqVSicTERADA2rVrUVFRAS8vL0yZMgXJycm8xYOIiKxKIZrQjuPVQ6zx8fEm20i9rttchBANKjPmvbZUZmy9eifoLVlmjXPyGpvOdVDDNYI2b1IJslr15gamqksKEyQTZKMps7V4mtJ1UMM1gja3+ipWIiIiW2T1VaymxFWsRERkKhxilVGXFA6xcoi10ZTZWjxN6Tqo4RpBm3OIlYiISAITJBERkQR1kr2zAAAaFUlEQVTOQRIREUngHKSMuqRwDpJzkI2mzNbiaUrXQQ3XCNqcQ6xEREQSmCCJiIgkMEESERFJ4CIdIqKGUKsBFxfJIlFR8eChwNSocZGOjLqkcJEOF+k0mjJbi8deroP0awTtyiFWIiIiCUyQREREEpggiYiIJHCRDhGRqXEBT5PARToy6pLCRTpcpNNoymwtHnu/jqb3kdswjaDtOMRKREQkgQmSiIhIAhMkERGRBCZIIiIiCUyQREREEpggiYiIJPA+SCKiRkLfPZS8v9L0eB+kjLqk8D5I3gfZaMpsLR57vw5jP5ts5B5Bo/E+SCIiosaJCZKIiEgCEyQREZEEJkgiIiIJTJBEREQSmCCJiIgkMEESERFJYIIkIiKSwJ10iIgsSa0GXFzqLLalHXGM2bnHlq6jobiTjoy6pHAnHe6k02jKbC0ee78Oc16jORhzHaaulzvpEBERWR8TJBERkQQmSCIiIglMkERERBKYIImIiCQwQRIREUlggiQiIpLABElERCSBCZKIiEgCEyQREZGERpEgFyxYgEGDBmHChAm4e/eutcMhIiI7YPMJ8vTp07hw4QKOHDmC8PBwbNiwwdohERGRHbD5BHnkyBGMGjUKADBq1Cj88MMPVo6IiIjsgcUSZFJSEkJCQuDs7Iy4uLgaZcXFxYiOjoarqyv8/Pywbds2Xdnt27ehUqkAACqVCsXFxZYKmYiI7JjFngfp4+ODhQsXYv/+/aioqKhRNm/ePDg5OeH69evIyspCZGQkgoODERQUhNatW+POnTsAgDt37qBNmzaWCpmIiOyYxXqQMTExGD9+PDw8PGocLy8vR0pKCpYuXQo3NzeEhYUhKioKmzdvBgCEhYVh//79AID9+/dj4MCBlgqZiIjsmMV6kHU5f/48HBwcEBAQoDsWHByM9PR0AEDPnj3h5+eHQYMGwcvLC5s2bZKsJzU1FampqRaJmYioqRAVFVAolaavWK0GXFxMX68FWT1BlpWV6eYYq6lUKpSWluq+X7ZsWb31hIeHIzw8HACwYsUK0wZJRNREKZRKQKGQLhSi4RW7uJinXguyeoJ0c3NDSUlJjWMlJSVwd3c3uK7MzEwcP37cVKEREZEds/ptHgEBAdBoNMjNzdUdy87ORlBQkMF1hYSEYM6cOaYMj4iI7JTFEqRGo4FarYZWq4VWq4VarYZGo4GrqytiYmKwaNEilJeXIyMjA7t378b06dMtFRoREVEtFkuQiYmJUCqVWL58ObZs2QKlUonExEQAwNq1a1FRUQEvLy9MmTIFycnJDepBZmZmYt26daYOnYiI7JBCiEYyW2oAhUIBU12Woo5JZiFEg8qMea8tlRlbr97Je0uWWeOcvEb7vg5zXmND2dp12MjiHqvPQRIREdkiJkgiIiIJVr/Nw5R4mwcREZkK5yBl1CWFc5Ccg2w0ZbYWj71fB+cgOQdJRETUmDFBEhERSeAcJBERkYQmOwdJRERUFzmpr0n1IKsZm/NDQkKQmZlpomiaHraPfmwf/dg++rF99LNk+3AOkoiISAITJBERkQSHJUuWLLF2ELaob9++1g7BprF99GP76Mf20Y/to5+l2qdJLtIhIiIyFodYiYiIJDBBEhERSbDbBFlcXIzo6Gi4urrCz88P27Ztk3ydEAJvvPEGPDw84OHhgddff91k+7zaMrnt8/7776NHjx5wd3dHx44d8f7771s4UuuQ2z7V7t27h27duuHxxx+3UITWY0jbnDhxAoMHD4abmxu8vb2xevVqC0ZqHXLbp7KyEvHx8fD29kabNm0wduxYFBUVWThay0tKSkJISAicnZ0RFxen97WrVq3CY489BpVKhZkzZ6KystK0wQg7NXnyZPHcc8+J0tJSceTIEdGyZUtx+vTpWq/717/+JQICAsRvv/0mCgsLRffu3UVycrIVIrYsue2zYsUKcfz4cXH//n3x66+/ig4dOojt27dbIWLLkts+1RITE8WgQYOEr6+vBaO0Drltc+PGDeHp6Sm2bNki1Gq1KCkpEWfOnLFCxJZlyP+tXr16iWvXromKigoxbdo0ER0dbYWILSslJUV89dVXIj4+XsyYMaPO1+3bt094eXmJ06dPi+LiYjFkyBDxxhtvmDQWu0yQZWVlonnz5uLcuXO6Y9OmTZNs3P79+4t169bpvv/0009FaGioReK0FkPa51Evv/yyeOmll8wZntUZ2j4XL14U3bp1E3v37m3yCdKQtnnrrbfEtGnTLBme1RnSPvHx8eK1117Tff/NN9+IgIAAi8RpCxYsWKA3QU6ZMkW89dZbuu9TU1OFt7e3SWOwyyHW8+fPw8HBAQEBAbpjwcHByMnJqfXanJwcBAcH1/u6psSQ9nmYEAJHjhxBUFCQuUO0KkPb5+WXX8a7774LpVJpqRCtxpC2+emnn9CmTRsMGDAAXl5eGDt2LAoKCiwZrsUZ0j6zZs1CRkYGrly5grt372Lr1q0YNWqUJcO1aVKfzdevX8etW7dMdg67TJBlZWVQqVQ1jqlUKpSWltb7WpVKhbKysiY9D2lI+zxsyZIlqKqqwosvvmjO8KzOkPb56quvoNFoEB0dbanwrMqQtiksLMTnn3+O1atXo6CgAB07dsSUKVMsFapVGNI+AQEB6NChA3x9fdGyZUucPXsWixYtslSoNk/qsxlAvZ9ThrDLBOnm5oaSkpIax0pKSuDu7l7va0tKSuDm5takN0Q3pH2qJSUlYdOmTfj222/h7Oxs7hCtSm77lJeX4/XXX8eaNWssGZ5VGfK7o1QqER0djaeeegouLi5YvHgxjh49ijt37lgqXIszpH3mzp0LtVqNW7duoby8HDExMexBPkTqsxmA3s8pQ9llggwICIBGo0Fubq7uWHZ2tuTQYFBQELKzs+t9XVNiSPsAwIYNG7B8+XIcPHjQLlZpym2f3Nxc5OfnY9CgQXjssccQExODq1ev4rHHHkN+fr6Fo7YMQ353evXqVeMPzep/N+XRGUPaJzs7G3FxcWjTpg2cnZ3x8ssv4+eff8bNmzctGbLNkvps9vb2hoeHh+lOYtIZzUZk0qRJYvLkyaKsrEz88MMPda4kS05OFt26dROFhYWiqKhIBAYG2sUqVrnts2XLFuHt7W0Xqw8fJqd97t+/L65evar7SklJEe3atRNXr14VGo3GSpGbn9zfnYMHD4pWrVqJkydPinv37on58+eLsLAwK0RsWXLbJy4uTsTExIg//vhD3Lt3T/zzn/8UPj4+VojYsu7fvy8qKirEm2++KaZNmyYqKirE/fv3a73uu+++E97e3iInJ0cUFxeLYcOGcRWrqdy6dUuMGzdOtGjRQrRv315s3bpVCCHE4cOHhaurq+51VVVV4rXXXhOtW7cWrVu3Fq+99pqoqqqyVtgWI7d9/P39haOjo3B1ddV9zZkzx1phW4zc9nlYWlpak1/FKoRhbbN27Vrh4+MjWrVqJcaMGSMKCgqsEbJFyW2fmzdviqlTpwpPT0+hUqnEwIEDxbFjx6wVtsUsXrxYAKjxtXjxYnH58mXh6uoqLl++rHvtypUrhZeXl3B3dxdxcXFCrVabNBbuxUpERCTBLucgiYiI6sMESUREJIEJkoiISAITJBERkQQmSCIiIglMkERERBKYIMlqPvvsMygUCgwdOtSqcfz2228YNmwYlEolFAoFvvnmG7Of09/fHwqFAocOHTL7ucxhyZIlUCgU9T6vb+jQoVAoFPjss8/MFotCoYBCoTB6d6K4uDgoFAosWbLEJHGZkyXalZggG53qD9bqr7Zt2yIiIgKZmZnWDk0vqYQQGBiIhIQETJgwwXqBAVi2bBkOHTqEJ554AgkJCejUqVOt11QnhPHjx1shQvkJyVL69euHhIQEjBgxAgBw6NAhKBQK+Pv7G133w7/jGRkZuuNHjhzRHX/4PAkJCUhISEDLli2NOu+IESOQkJCAfv36GVVPXYQQ8PPzk0xsX3zxBRQKBby9vaHRaMxyfjKco7UDoIYZM2YMOnbsiPT0dHz//ff45Zdf8Ouvv8LLy6vWa+/fv4/mzZtbIUr953766afx9NNPWzii2s6fPw8AeOWVVzBz5kwrR9M4jBw5EiNHjjT7eZKTkzFw4EAAwL/+9S/J13z44YcmOdfUqVMxdepUk9QlRaFQYOrUqVi+fDm2b99e44+dbdu2AQAmT54MR0d+LNsMk+7LQ2bn5+cnAIivvvpKCPHgqez433ZMu3btEpcuXdJ9n5ycLNq1ayeGDh0qhBAiOztbRERECA8PD9G2bVsxZswY8euvv9aqe/ny5aJHjx7Czc1NjB8/Xty8eVP3mvT0dDFo0CChUqlEu3btxNSpU0VRUZGuvPrcq1atEv7+/qJjx466eh/+2rhxo9i4caMAIIYMGWJw/WvWrBFdunQRbm5u4vnnnxeVlZV1tpm+OocMGVIrtkuXLtWqo3r7q3HjxgkhhCguLhYTJkwQHh4ewtnZWfj7+4vZs2frXi+3rdPS0oQQQrz//vviiSeeEC1atBBOTk6iV69e4osvvqhx7oe/qtvs8uXLYtKkScLHx0eoVCoxfPhwcerUKd15Vq1aJTp16iScnZ1F27ZtxZAhQ2rEUe33338XAETLli2FVqsVt27dEgqFQvcA2oKCAgFAeHp6iqqqKl1MM2bMEGlpabXiq/5oqW7fRYsWiWHDhgmlUikGDBgg8vPz6/x5VbdN69athbOzs7hx44b4/fffhbOzs2jdurUAIPz8/HSvf/Tnpu+at27dKrp37y5cXFxE69atRb9+/cSRI0eEEELMmDFDt63Zw+0eGxsrpk+fLlxdXUXnzp3FgQMHdOfOyckRoaGhQqlUitGjR4uXXnqpxu/Jo3JycgQA4eDgIK5fvy6EEOKPP/4Qzs7OAoD45ZdfhBBCvPrqq8LPz084OzsLpVIpQkNDdb8rD7frxo0bJb+v/pk83E6nTp0So0ePFp6enqJt27YiJiZGt21bVVWVeOutt8Tjjz8unJychLe3txgxYkSN//v2iEOsjVhVVRXS09N137dt27ZG+YIFCzBq1CgMGDAAV69exZAhQ7B//37069cPffr0wTfffIOhQ4fi9u3bNd6XmJiIp556Cm3btsWuXbswe/ZsAMB///tfhIeH44cffsDIkSPh5+eHbdu2ISIiAvfv369Rx9///ncMHjwYI0aMwMyZM3WPoImNjUVCQgICAwNrXY8h9S9evBgDBgyARqPB1q1bsXnzZsk2qq/OCRMmwNfXFwAwfPhw2UN1K1euxJdffokuXbrgxRdfRPfu3XH06FEAMKitq126dAk9e/ZEXFwcxo0bh5ycHEybNg35+fno168fQkNDAQDdu3fXDUvfvXsXzzzzDHbs2IFevXph+PDhOHToEIYNG4abN28iLy8Pr7zyCkpKShAXF4fhw4ejoKAAV69erXV+T09PBAYGoqSkBKdPn8bRo0chhMD169eRl5enG+ocPHhwrUe9Pf7444iNjQXw4FFD1UOeD3v33XfRrl07tG3bFkePHsXChQvrbeMZM2agsrISGzZswIYNG1BZWVnvELO+a66oqEBcXBwuX76M559/HpGRkSgpKcGFCxf01pmSkoIrV66gR48euHDhgm6UQaPRICoqCseOHUNgYCCUSiWSk5P11hUYGIjevXtDq9Xiiy++AAB8+eWXqKysRNeuXRESEgLgwe9DaGgoZs2ahWHDhuHYsWOYOHFig591eO3aNQwePBgHDhxAWFgYQkNDsXPnTkRERKCyshIHDx7EsmXL4ODggFmzZmHw4ME4deqUSZ+t2ChZO0OTYaR6YwDE2LFjhVarrdGDPHjwoO59K1asEAB0vUkhhOjdu7cAINatW1ej7g8//FAIIURWVpaurtLSUjF37lwBQMTFxQkhhLh3757w8vISAMT+/fuFEP/31/z69esl4374r+BHe5CG1L9jxw4hhBAvvPCCACDmzZsn2V5y6nz0r28pj/YgX3/9dQFAJCQkiGPHjonS0lLdEzoMaevq9igrKxOff/65WLhwoZg/f7547LHHBADdRtYP99iq7dixQwAQvr6+IiEhQSQkJIjOnTvrRg/OnDkjAIhevXqJ/fv3i99++00IIep8kkh1WyUnJ4s333xT+Pv7CycnJ7Fx40Zdz+ijjz6SjEeqx/Jw2/75z38WQgixYcMGAUAEBQXV2dYPj5IEBgaKzp07i06dOonAwEDx1Vdf6e1B6rvm0tJS0axZM+Hr6yu+/vprceHChRrtUVcPMigoSFRVVYmLFy/qznXjxg1x5MgRAUC4ubmJsrIyIYQQUVFRenuQQgjxwQcfCABi4MCBQgghnn32WQFAJCYm6l5z69YtsW7dOvHWW2+JhIQE0aJFCwFAZGRk1GhXuT3I9957TwAQ3bt31/2ueHp6CgDiu+++E3v37hUAxLPPPivS0tLE9evXRVVVldBqtXVehz1gD7KRGjNmDObPn4+lS5di79692L17N5o1q/njrJ67AaBb4de9e3fdsW7dugEALl++XON91a+pLgeAoqKiWnU0b95ct6Dl0ToePrdchtTfp08fAECrVq0APHi6uLF1GmL+/PmIiIjA2rVrERoailatWuGFF15AVVWVQW0NAPfu3UO/fv0wY8YMJCYm4sMPP8S1a9cAADdu3KgzhurzFBUVYfXq1Vi9erWuN5SXl4fu3bvj7bffRlFRESIiItC+fXt069YNZ8+elayvejXx0aNHkZGRgWHDhuHJJ59ERkaGrnfc0BXHcn9ej4qPj8eFCxdw8eJFzJ07t97X67tmNzc3JCcnQwiBsWPHonPnzmjfvj2OHDmit87evXtDoVDoYq+Ov6ioCADQvn17uLq6AoDkyMijpk6dimbNmuHo0aM4duyYboHT888/DwC4desWevTogTlz5mDZsmVYvXo17t69C0D/78PDtFptje+rf1fOnj2r+12prisvLw8jRozAn//8Z/z8888YNmwYvL29ERoaiuvXr8s6X1PFBNlIzZo1C6tWrcLChQsxatSoWsNeAODs7Kz7d/Wqv19//VV37Ny5cwAAPz+/Gu+r/gB9+LW+vr616rh//z4uXrwoWcfD5wYABwcHAA+GhetiSP3VCxmkrruhdRqiTZs22LdvH0pLS3UPvN22bRsyMjIMamsAOHPmDE6fPg0HBwfk5uaiqqpK90Er/vewHan2qz5P3759UVVVBfHg8XW4ffs2FixYAK1WiwULFuDmzZu4fPky3njjDZw7dw6rVq2SvKYhQ4YAAA4fPoxffvkFAwcOxMCBA5Gamors7Gx4eHigR48eku+t7+cr9+f1qBdeeAEtWrSAq6srpk+fXu/r67vmGTNmoKioCFeuXMHq1atRWFiIpUuX6q2zrtirh+YLCwtRUVEBoObPvC7t2rXDM888AyEEpk2bBq1Wi7CwMN3P88iRI7h69So8PT1x7do1VFZW6pJz9e/Do6oTdElJCQDg9OnTNcqr646JidH9ngghcPXqVcyaNQtarRZJSUn4448/kJeXhxdeeAG//PILPv3003qvpynjcik7MW3aNLz77rtIS0tDVFQU7t27h5MnT8Lb27vWbRb/+Mc/kJ2djbS0NABAdHQ03NzcMHv2bHzyySf4/PPPUVFRgcuXL+P3339HUFBQvT2L9u3b4+LFi1i0aBH27NmDv/71r7VeY0z9dTFHnQCwfPly7NmzBz179oSTk5PuL3SVSmVQWwMP5o6bNWsGrVaLV199FXfv3q3xxHngQfsBwHfffYeXX34ZQ4cOxejRo9GpUyccP34cAwcORK9evVBQUIBDhw5h79698Pf3R2hoKAYPHgwvLy/dPOLDPaGHeXt7o1u3broP+YEDB6JNmzZYuXIlAOn5x0fjKywsxJ/+9Cd06dIFb7zxhoGtWptKpcLhw4d1/67Pb7/9pveavb29MXToUPj4+ODUqVM1ygzVr18/PPHEE8jLy8PQoUPRoUMH7NmzR9Z7p02bhtTUVOTl5em+r+bt7Q3gQW/xlVdewcWLF+vtcffp0wd79+7F//t//w8FBQW1Etvzzz+Pd999Vzfv6O/vjwsXLiA9PR25ubnIz89HXFwc+vfvjzZt2tT7u2Iv2IO0Ez4+PkhLS8OIESOQkZGBzMxMREZGIi0tDW3atKnx2iVLluDEiRO4ceMGoqKi8PHHHwN4MNT0/fffo3///ti7dy8uXbqEyZMnY9++fXByctJ7/iVLluCJJ57Ajz/+iNWrV0sO3RhTf13MUScAPPnkk3B0dMSuXbuwadMmeHt746OPPkKvXr0MamvgwSKXNWvWwNvbG+np6ejbty8GDBhQ4zUTJ05EREQEysvLkZSUhLS0NLi6uuLgwYOYMmUKCgoK8Pnnn+PcuXOYNm0aunbtipYtW+Lpp59GRkYGPvnkE1y5cgWTJ0/Wu0Cmuhfp4eGBrl271hgqry6T4u/vj7/97W9QqVRYv359nYumGqJv377o27evrNfWd83Dhw/HiRMnsH79euTk5CAyMlL3B4ChHB0dsXv3boSGhuLUqVO4e/cuZs2aBaD2CMqjYmNj0aJFCwCAk5MTJk6cqCvr378/FixYgNatW+PAgQOYMmWKrrdal1dffRUjR47EzZs3kZaWhldeeaVGuY+PD9LT0zFmzBhkZWVhy5YtKCoqwrx589C2bVv4+vqiS5cuOHjwID755BPcvXsX8fHxugV69ooPTCYdf39/XL58GWlpaVbf3YaoMbhz506Nnm1ERAS+//57LFy4sN6hW7J9HGIlImqgJUuW4MKFCwgJCUFWVha+//57uLq6csOJJoJDrEREDRQcHIycnBz885//xI8//ojRo0fj8OHD6Nixo7VDIxPgECsREZEE9iCJiIgkMEESERFJYIIkIiKSwARJREQkgQmSiIhIAhMkERGRhP8PYVgdORWCYBAAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 468x378 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "fig , ax = plt.subplots()\n",
    "\n",
    "n , bins , patches = plt.hist( missing_values_per_SNP / number_of_isolates , bins = 50 , color = 'k' , rwidth = 1.0 , edgecolor='white', linewidth=1.0)\n",
    "plt.hist( missing_values_per_SNP[missing_values_per_SNP > max_number_values_missing_per_SNP] / number_of_isolates , bins = bins , color = 'red' , rwidth = 1.0 , edgecolor='white', linewidth=1.0)\n",
    "\n",
    "#plt.title('Mean Coverage across sequenced isolates in rollingDB', fontweight = 'bold' , fontsize = 12, color = 'k')\n",
    "plt.xlabel('Proportion of Isolates with Missing Values', fontweight = 'bold' , fontsize = 12, color = 'k')\n",
    "plt.ylabel('Number of Unique SNPs', fontweight = 'bold' , fontsize = 12, color = 'k')\n",
    "\n",
    "ax.spines['right'].set_visible(False)\n",
    "ax.spines['top'].set_visible(False)\n",
    "\n",
    "ax.tick_params(labelcolor = 'k')\n",
    "ax.set_yscale(\"log\", nonposy='clip')\n",
    "\n",
    "fig = plt.gcf()\n",
    "fig.set_size_inches(6.5, 5.25)\n",
    "fig.tight_layout()\n",
    "\n",
    "plt.tick_params(axis='y', which='major', labelsize=12 , labelcolor = 'k')\n",
    "plt.tick_params(axis='x', which='major', labelsize=12 , labelcolor = 'k')\n",
    "\n",
    "#file_name = '/home/rv76/Farhat_Lab/Figures/Final_5/Scraping_rollingDB/Mean_Coverage_Distribution_across_all_isolates.pdf'\n",
    "#plt.savefig(file_name, bbox_inches='tight', dpi = 300 , transparent = True)\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of Isolates = 33873.0\n"
     ]
    }
   ],
   "source": [
    "print 'Number of Isolates = ' + str( number_of_isolates )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Range and Average Number of Missing Values per SNP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "166"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.min(missing_values_per_SNP)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1681.338931735388"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.mean(missing_values_per_SNP)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "33805"
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.max(missing_values_per_SNP) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SNP with most missing values is missing from 99.7992501402% of isolates\n"
     ]
    }
   ],
   "source": [
    "print 'SNP with most missing values is missing from ' + str( float(np.max(missing_values_per_SNP) ) / number_of_isolates * 100) + '% of isolates' "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Drop SNPs that are flagged as 'missing values' in more than 25% of isolates in rollingDB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3220\n"
     ]
    }
   ],
   "source": [
    "print sum( missing_values_per_SNP > max_number_values_missing_per_SNP )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Percentage of SNPs with many missing values that we're going to drop from analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2.251480593216191\n"
     ]
    }
   ],
   "source": [
    "number_of_unique_SNPs = float(np.shape(genotypes_matrix)[0])\n",
    "print 100.0 * ( sum( missing_values_per_SNP > max_number_values_missing_per_SNP ) / number_of_unique_SNPs )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### FILTER out sites with missing calls in $> 25\\%$ of isolates"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Drop unique SNPs with a high abundance of missing values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(143017, 33873)"
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.shape( genotypes_matrix )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Drop unique SNPs from the __GENOTYPE MATRIX__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "143017"
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len( missing_values_per_SNP <= max_number_values_missing_per_SNP )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [],
   "source": [
    "SNPs_to_keep_filter = missing_values_per_SNP <= max_number_values_missing_per_SNP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "139797"
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sum(SNPs_to_keep_filter)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [],
   "source": [
    "genotypes_matrix = genotypes_matrix[SNPs_to_keep_filter , :]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(139797, 33873)"
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.shape(genotypes_matrix) #Genotype Matrix"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Drop unique SNPs with a high abundance of missing values from the __SNP ANNOTATION FILE__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [],
   "source": [
    "SNP_annotation_DF = SNP_annotation_DF[SNPs_to_keep_filter]\n",
    "\n",
    "#re-index new filtered SNP annotation DF (so new index matches indexing of genotypes matrix rows)\n",
    "SNP_annotation_DF.reset_index(drop = True , inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>29</td>\n",
       "      <td>C</td>\n",
       "      <td>[G]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>31</td>\n",
       "      <td>A</td>\n",
       "      <td>[C, G]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>48</td>\n",
       "      <td>C</td>\n",
       "      <td>[T]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>64</td>\n",
       "      <td>G</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>71</td>\n",
       "      <td>C</td>\n",
       "      <td>[T]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   pos ref     alt\n",
       "0   29   C     [G]\n",
       "1   31   A  [C, G]\n",
       "2   48   C     [T]\n",
       "3   64   G     [C]\n",
       "4   71   C     [T]"
      ]
     },
     "execution_count": 79,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SNP_annotation_DF.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(139797, 3)"
      ]
     },
     "execution_count": 80,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.shape(SNP_annotation_DF) #Annotation for Rows of Genotypes Matrix"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## [9.2] Drop low-quality isolates"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### (Filter 4) *FILTER* out isolates that have missing base calls $> 25\\%$ of SNP sites"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Investigate the Number of Missing Calls for each Sequenced Isolate"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Look for isolates with an abundance of missing values, drop isolates that have missing values > 25% of SNP calls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {},
   "outputs": [],
   "source": [
    "missing_values = genotypes_matrix == 9\n",
    "missing_values_per_isolate = missing_values.sum(axis = 0) #will range from 0 - 139,797"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {},
   "outputs": [],
   "source": [
    "#impose a threshold for maximum number of missing values per isolate (across all SNPs)\n",
    "max_proportion_missing = 0.25\n",
    "number_of_SNPs = float(np.shape(genotypes_matrix)[0])\n",
    "\n",
    "max_number_values_missing_per_isolate = np.floor( max_proportion_missing * number_of_SNPs )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "34949.0\n"
     ]
    }
   ],
   "source": [
    "print max_number_values_missing_per_isolate #that we're going to allow"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAhAAAAFuCAYAAADQwCGlAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMi4zLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvIxREBQAAIABJREFUeJzt3XlYVfW+x/HPBmQQFBWUnBIzyamwKzmkx6FMTc1yOqVpkp7Ssq4Nt+nmURxKG84py/J6SjNNOw2WTZaJmZaahSUq5SyaaBZg4QDqht/9w8u+krBZS9kTvl/Pw/PAWovv+u5f5P7u37QcxhgjAAAAG4J8nQAAAAg8FBAAAMA2CggAAGAbBQQAALCNAgIAANhGAQEAAGyjgAAAALZRQAAAANsoIAAAgG2VqoBIS0vT7Nmz5XA4fJ0KAACVmqMybmXtcDhUCV8WAAB+o1L1QAAAAO+ggAAAALZRQAAAANsoIAAAgG0UEAAAwDYKCAAAYBsFBAAAsC3E1wlUpLS0NG3YsMHXaQAAUOmxkRQAALCNIQwAAGAbBQQAALCNAgIAANhGAWFRQUFBhVwDAEBlwCRKm3HdqYRNCQBAqeiBAAAAtlFAAAAA2yggAACAbRQQAADANgoIAABgGwUEAACwjYdpAQAA29gHwmZcdyphUwIAUCqGMAAAgG0UEAAAwDYKCAAAYBsFBAAAsI0CAgAA2EYBAQAAbKOAAAAAtlFAAAAA2yggAACAbRQQAADANgoIAABgGwUEAACwjQICAADYRgEBAABso4AAAAC2BUwB8eabb6p27dq+TgMAAChACoiioiK9++67atiwoa9TAQAACpACYtGiRRo0aJCCggIiXQAAKj2vvSPPnDlTSUlJCgsLU3Jycolzubm56t+/vyIjI9WoUSMtWrTIda6wsFBvv/22br75Zm+lCgAAyhHirRvVq1dP48eP17Jly5Sfn1/i3NixYxUaGqpDhw5p48aN6tOnjxITE9WyZUu98cYb+utf/0rvAwAAfsRr78oDBgzQTTfdpJiYmBLHjx07psWLF2vKlCmKiopSp06d1K9fPy1YsECS9OOPP2r+/Pnq1auXduzYof/8z//0VsoAAKAMXuuBKMv27dsVHByshIQE17HExEStWrVKkvTUU0+5jiclJemFF14oNU5qaqpSU1M9mywAAJDkBwXE0aNHFR0dXeJYdHS0jhw5cta1aWlpZcbp3r27unfvLqlk0QEAACqezycWREVFKS8vr8SxvLw8VatWzUcZAQCA8vi8ByIhIUFOp1M7duxQ06ZNJUnp6elq2bKl7VhpaWnasGFDRacIAAD+xGs9EE6nUwUFBSosLFRhYaEKCgrkdDoVGRmpAQMGaMKECTp27JjWrFmjDz74QMOHD7d9j6SkJI0ePdoD2QMAgDN5rYCYOnWqIiIiNH36dL3xxhuKiIjQ1KlTJUkvv/yy8vPzVadOHQ0ZMkSzZs06px4IAADgHQ5jjPF1EhXN4XDIEy/L4XC4PV8JmxIAgFL5fA5ERWIOBAAA3kEPhM247lTCpgQAoFQ+X8YJAAACDwUEAACwjTkQAADANuZA2IzrTiVsSgAASsUQBgAAsI0CAgAA2EYBAQAAbGMSJQAAsI1JlDbjulMJmxIAgFIxhAEAAGyjgAAAALZRQAAAANsoIAAAgG2swgAAALaxCsNmXHcqYVMCAFAqhjAAAIBtFBAAAMA2CggAAGAbBQQAALCNAgIAANjGMk4AAGAbyzhtxnWnEjYlAAClYggDAADYRgEBAABso4AAAAC2UUAAAADbKCAAAIBtFBAAAMA2CggAAGAbBQQAALCNnSgrUEFBgcLDw8/5PAAAgYKdKG3GdccY4/aaStjUAIALFEMYAADANgoIAABgGwUEAACwjQICAADYRgEBAABso4AAAAC2UUAAAADbKCAAAIBtlgqIJUuWaOrUqTLG6K233lLTpk3Vtm1bpaenezo/AADghyztRNmmTRsFBwdr/fr1qlOnjnJyciRJ11xzjVJTUz2epF3sRAkAgGdZ6oHYtWuXrrjiCu3YsUM5OTl699131bt3b58+dwIAAPiOpYdpOZ1OOZ1ObdmyRQ6HQ926ddPy5ct18uRJT+dni68fpgUAwIXC0hBGUlKSNm3apPDwcNWtW1fbtm1Tnz59tH37du3YscMbedrCEAYAAJ5laQhj2rRpqlOnjqpWrapnnnlGJ0+eVGZmpvr06ePp/AAAgB/icd4247pDDwQA4EJheR+I1atXa+jQoWrXrp2ysrI0efJkffvtt57MDQAA+ClLPRAffPCBBg4cqKKiIjkcDp04cUINGzZUx44d9e6773ojT1vogQAAwLMs9UBMnjxZtWrVUpcuXSRJISEh6tSpk7755huPJgcAAPyTpQJi27Zt6t27t1q3bu06Fhsbq9zcXI8lBgAA/JelAqJu3brKyMhQYWGhJCk3N1fLly9Xw4YNPZocAADwT5YKiIEDB2rDhg2aNWuWJKl+/fras2ePBg0a5NHkAACAf7I0iTI/P18jR47UW2+95Tr217/+VXPmzFFkZKRHEzwXTKIEAMCzbO0DkZ2drczMTMXHxys2NtaTeZ0XCggAADzL0hBGcHCwHnzwQcXGxiopKUmxsbF6+umnlZiY6On8AACAH7JUQBhjzvr0vGPHDm3ZssUjSQEAAP/mdgjjkksukSRlZmaqevXqqlWrliSpqKhI+/fvV3R0tHJycryTqQ0MYQAA4FluH+edmZkp6fQbZ15envLy8kqc/9vf/uaxxAAAgP9yW0CsXLlSxhhdc801GjRokMaOHSvp9JyIhg0bqlGjRh5P8NChQ+rfv7+qVKmi4OBgLVy4UHXr1vX4fQEAQNksrcJYtWqVGjRooCZNmngjpxIKCwvlcDgUFBSkefPmaf/+/Ro/frzb3/HXIYz8/HyFh4e7jVFQUFDuNQAA+JrbHohiXbp00dtvv61nnnlGv/32m+vN2eFwaPHixR5NMDg42PX9kSNH1LJlS4/ez5PCw8MtFSEAAPg7SwVESkqKpkyZ4vr5zALCqpkzZ2revHnavHmzhgwZonnz5rnO5ebmatSoUfr8888VGxuradOmaejQoa7zGzdu1OjRo/X777/r888/t3xPAADgGZaWcc6bN081atRQmzZtJEl33XWXIiIidN9991m+Ub169TR+/HiNHDnyrHNjx45VaGioDh06pIULF+quu+5SRkaG63zr1q21fv16TZkyRdOmTbN8TwAA4BmWeiAOHDig2267TdWrV9eGDRv00ksv6Y8//tCuXbss32jAgAGSpLS0NO3fv991/NixY1q8eLG2bNmiqKgoderUSf369dOCBQs0ffp0nThxQmFhYZKk6OhoVa1a1c7rAwAAHmCpgKhataqqVKmimJgYSdKCBQu0detW/fTTT+edwPbt2xUcHKyEhATXscTERK1atUqS9P333+uRRx5RcHCwwsPDNXfu3FLjpKamKjU19bzzAQAA5bNUQFxyySXavXu3+vXrJ2OMkpOTJUnNmjU77wSOHj2q6OjoEseio6N15MgRSVKHDh20evXqcuN0795d3bt3lyQ99dRT550XAAAom6UC4umnn1ZWVpZ69eqlO+64Q2+//bbi4uI0e/bs804gKirqrA2q8vLyVK1atfOODQAAPMNSAVH8yV6SZs+eXSGFQ7GEhAQ5nU7t2LFDTZs2lSSlp6ef03LNtLQ0bdiwocJyAwAApStzI6nJkyeX/8sOh/7+979bupHT6ZTT6dSkSZO0f/9+vfLKKwoJCVFISIhuueUWORwOvfrqq9q4caN69+6ttWvXnvOeD/66kVR554uvAQDA35VZQAQFBVl6MywsLLR0o5SUFE2aNKnEsYkTJyolJUW5ubkaOXKkli9frpiYGE2fPr3EPhB2UUAAAOBZZRYQycnJljaKeu211yo8qfNFAQEAgGeVOQfizJ0iAwVzIAAA8A5LD9OSpHXr1umll17S3r17FR8fr7vvvlsdOnTwdH7nhB4IAAA8y1IBsWzZMvXt27fEfIegoCB99NFHuv766z2a4LmggAAAwLMsFRAdOnTQhg0bdM8996hZs2baunWrZs6cqTZt2mjdunXeyNMWCggAADzL0j4QP/74o4YMGaJ//vOfrmM5OTlasmSJxxIDAAD+y1IBUaNGDWVkZKigoEDh4eEqKChQRkaGatSo4en8bGESJQAA3mHpcd7XXXedvv/+e8XFxalVq1aKi4vTDz/8oOuuu87T+dmSlJSk0aNH+zoNAAAqPUtzILKzs3XDDTdo/fr1rmPt2rXTRx99pNjYWI8meC4CeQ5Efn6+wsPDyzxf3AsEAIAvWV7GaYxRWlqaMjMzFR8fr6SkJEsbTflCIBcQVmIAAOBrlgqIw4cP68iRI7r44ot16NAhvfrqq6pZs6buuOMOValSxRt52kIBAQCAZ1maRDlkyBBlZGTo559/Vr9+/ZSWlibp9OqMmTNnejRBO5hECQCAd1jqgahbt666dOmiF198UXFxcbrtttu0Zs0a5efna//+/d7I0xZ6IAAA8CxLqzAOHz6s2NhY/fjjj3I4HJo4caJ69uyp7OxsT+cHAAD8kKUhjDp16mjp0qXavn27qlWrpsaNGys3N9fv9oEAAADeYakHYtiwYcrMzFRqaqqGDx8u6fTDtRITEz2aHAAA8E+Wl3GuWLFCp06dUo8ePVRYWKi1a9eqfv36uvTSSz2do2XFkyjHjBnDHAgAADzIcgERSJhECQCAZ7mdA1GrVi23v+xwOJSTk1OhCcE9KztRslslAMDT3PZABAW5nyLhcDhUWFhY4Umdr8reA8EjwQEAvua2B2LlypXeygMAAAQQ5kDYjOsOPRAAgAuFpWWcAAAAZ6KAAAAAtlnaiTJQ8DAtAAC8o8w5EA8//LA6d+6svn37av78+WrRooWSkpK8nd85YQ4EcyAAAJ5V5hDGs88+qxUrVkiSkpOT9eabb3otKQAA4N/KHMKIjo7WwoULdfjwYUnSsmXLXN8XczgcmjNnjmczhG3lbSTFRlMAgPNV5hBGcnKy5s+ff/qiMoYE2EiqJH8awmA7bACAJ5XZAzF37lz16dNH27Zt04QJE9SuXTv16tXLm7kBAAA/VWYBERQUpMGDB0uSCgsL1b59e/Xs2dNriQEAAP9leSfKPXv2aM6cOdq7d6/i4+N1++2365JLLvF0fueEIQyGMAAAnmWpgPj222917bXX6vjx4643n8jISK1YsUJt27b1eJJ2UUBQQAAAPMvSRlKPPvqojh07pptuuknNmjXT1q1btWTJEj322GOupZ7+gI2kAADwDks9ELVq1dK1116rd955x3Vs8ODBWrFihXJzcz2a4LmgB4IeCACAZ1l6FkZ4eLiys7NLHMvOzmYvAQAALlCWhjA6duyo9957T02bNtVll12m7du3a9euXRo4cKCn84MHWNlIis2mAADuWBrC2Lt3r7p166bMzEzXsfj4eH3xxReKj4/3YHrnhiEMnqcBAPAsy8s4jx8/rk8++USZmZmKj49Xnz59VLVqVU/nd04oICggAACeZbmACCQUEBQQAADPsjSJEheegoKC8zoPAKjcLE2ixIUnPDycpaAAgDLRAwEAAGwrt4BwOp269tpr9fe//90b+QAAgABQbgEREhKirVu36pdffvFGPggQzJEAgAubpSGMCRMmaMmSJVq1apVOnTrl6ZwQAIrnSJT1xSZUAFC5WVrGGRQUVOqEOofDIafT6ZHEzkXxw7TGjBnDMk4fx2CSJQBUbpYLiLIUFRVVaEIVgX0gfB+DAgIAKjdLyzj37Nnj6TwAAEAAsbUT5fbt27Vz50717t3bkzmdN3ogfB8jPz+fB3YBQCVmaRJlbm6uunfvrubNm6tfv37KzMxUSEgISztRpvImWTLREgACm6UC4qGHHtIXX3yh0NBQGWMUHx+v9u3b6+OPP/Z0fgAAwA9ZKiA+++wzde3aVWPGjHEda9GihXbv3u2xxAAAgP+yVEDk5+erbt26JY5lZ2crNDTUI0kBAAD/ZqmAuOKKK/Txxx9r/fr1kk4PaXz00UdKTEz0aHKo3NjNEgACl6VVGF9//bW6d++ukydPuo5VqVJFK1asUKdOnTya4LlgFYbvY1TUPQAA/snyMs4tW7Zo1qxZ2rt3r+Lj43XnnXfqiiuu8HR+54QCwvcxKCAAoHKztQ/E8ePHlZWVpfr166tq1aqezOu8UED4PkZF3IO9JADAf1maA3H48GENHjxY1apVU7NmzVStWjUNHjxYubm5ns4PFzD2kgAA/2VpK+tRo0ZpyZIlJY699957Kiws1HvvveeRxAAAgP+y1AORmpqqxo0ba+PGjSooKNAPP/ygRo0aKTU11dP5AW6xkgMAfMNSD0STJk3Upk0b16TJxMREdevWTT/88INHkwPKUzzMURYmYgKAZ5RZQKxevdr1/W233aZJkyapdevWatasmX766SctXrxYKSkp3sgRAAD4mTJXYQQFBZX7yS4oKEhOp9NjyRVbt26dHnjgAYWGhqpevXqaP3++qlSpUub1rMLwfQx/ybPo+HE5IiLcx8jPL/caAEBJZRYQ8fHx5f7jLUl79uyp8KT+7MCBA6pZs6YiIiL0+OOP68orr9SgQYPKvJ4Cwvcx/ClPlfd3zDAHANhW5hBGZmamF9Nwr169eq7vQ0JCFBRkae4nAADwEFvvxDk5Odq3b1+JLztmzpyppKQkhYWFKTk5ucS53Nxc9e/fX5GRkWrUqJEWLVp01u/v2bNHn376qfr27WvrvgAAoGJZWoWxfPlyjRo1SllZWSWOOxwOW3Mg6tWrp/Hjx2vZsmXKz88vcW7s2LEKDQ3VoUOHtHHjRvXp00eJiYlq2bKlJCkvL08jRozQggULeAooKlZBgeRmQyrmSADA2SxtZd2kSZMy5zoUFRXZvun48eO1f/9+zZs3T5J07Ngx1axZU1u2bFFCQoIkafjw4apfv76mT58up9OpG2+8UQ8++KCuueaacuMzB8L3MfwpT0tzINxdwxwJADiLpSGMnJwc9ezZU3l5eSoqKirxVRG2b9+u4OBgV/Egnd5rIiMjQ5L05ptvav369Zo8ebK6du2qt95666wYqampevTRR/Xoo49WSE4AAKBsloYw7r33Xr333nvKyspSQkKCpdUZdhw9elTR0dEljkVHR+vIkSOSTvdGDB8+3G2M7t27q3v37pKkp556qkLzAwAAJVnqgRg4cKAOHDigFi1aKCQkRMHBwQoODlZIiKX6o1xRUVHKy8srcSwvL0/VqlWrkPgAAKBiWaoAbr31Vv3xxx9nHa+oeQYJCQlyOp3asWOHmjZtKklKT093TaC0Ki0tTRs2bKiQnAAAQNksTaKsVq2aWrRooaefflo1atQocS4xMdHyzZxOp5xOpyZNmqT9+/frlVdeUUhIiEJCQnTLLbfI4XDo1Vdf1caNG9W7d2+tXbvWdhEhMYnSH2L4U57nPYkyP9/tKg2JlRoALjyWCogHH3xQ3377rVasWHFeSyhTUlI0adKkEscmTpyolJQU5ebmauTIkVq+fLliYmI0ffp0DR069JzuQwHh+xj+lGeFrMJgN0sAKMFSAdG6dWtt2bJFkZGRio+PV3Bw8Olfdjj8csiAAsL3MfwpTwoIAKh4luZAbNq0SZJ05MgRbd682XW8oldjnC/mQMBn2IwKwAXGUg/E66+/Xua5ESNGVGhCFYEeCN/H8Kc8vdYDcR7zKCgwAAQaSwVEoKGA8H0Mf8rTLwoIdrsEUMlYGsIYOXJkqccdDofmzJlToQkBAAD/Z6kHIigoqMSn+uLvHQ6HCgsLPZ6kVcVzIMaMGUMPBD0QrvP0QABAxbNUQJy59LKwsFCbN2/Whx9+qJEjR+qVV17xaILngiEM38fwpzwDooBgrwkAAeac50DceeedysrK0ieffFLROZ03Cgjfx/CnPAOigGCpKIAAY2kOxL59+0r8nJeXp23btik9Pd0jSQEoBUtFAfgRSwVE48aNSz1++eWXV2gyANwID3fbS+GghwKAF1kqIP48HFC1alVdeeWVevHFFz2S1LliI6kLUEGBR4arAADusQ+EzbjuBNK8gMo0B8Ln8xcqIkZF3aMc5Q1zVKZhkAvptQK+YKkHAkAAKGeOhKTTb5gXyDDIhfRaAV9wW0AUPzSrLA6HQ06ns0ITAnCOypkjIckrKzmsfLLn0z8Q+NwWEJVwdAOAh5X3yV/i0z9QGbgtIH744YcSPx89elQzZ87U22+/LWOMWrdu7dHkAHiZhWEQeg8ASOUUEImJiZKkgoICvfTSS3rmmWf066+/6vLLL9fEiRM1YMAAryRpFaswgPNkYRjEG70HTIAE/J/bVRgnTpzQrFmz9NRTT+nXX39V8+bNNXHiRA0ePNibOdrGKgzfx2AVhh/eoyJjlMfTMaz+/83zRwCPcVtANGjQQAcPHpTD4dDgwYN18803KygoqMQ1/fr183iSdlFA+D4GBYQf3qMiY5SHAgKo9NwWEMVP4Szzl/10FQYFhO9jUED44T0qMkZ5KCCASs/tHIiLL7643H/AAVxgLEy0BFD5uS0gMjMzvZQGgIBR3kRLPtkDF4Sg8i8BAAAoqVJtZc0yTgAAvIOHadmM644/TSz0dQwmUfrhPfwpRn5++fMoPD2J0kIO7DcBlK1S9UAACBD+MI+iAjbN4rkfuJBRQAAIPH6yEoTnfuBCRgEBIPD4yZNHgQsZqzAAAIBtFBAAAMA2CggAAGAbBQQAALCtUk2iZCMpAJVReUtBWSoKX2AjKZtx3fGnzZV8HYONpPzwHpUpRmV78qgVvt43A/gThjAAwJMKCtyeNvn5fnGPCskDF5RKNYQBAH6nnD0rHBXxxl3ePSz0trDhFeyiBwIAylLOJ/sKUfzm7+4L8EMUEABQlvLe3CsTbwy1oFKhgAAAlFsslTehWaLIuNAwBwIAUL4KeHopKhd6IAAAgG0UEAAAwDYKCAAAYBsFBAAAsI0CAgAA2EYBAQAAbKtUyzh5GicAAN7B0zhtxnXHn55Q6esYPI3TD+9RmWIESp7+EsOfnl6KSoMhDAAAYBsFBACgYnjhseJsl+0/KtUcCACAD1XUY8XLiwG/QA8EAACwjQICAADYRgEBAABso4AAAAC2UUAAAADbKCAAAIBtFBAAAMA2CggAAGAbBQQAALCNAgIAcEFhu+yKwVbWAIALiiMigu2yK4Df90D88ccfatu2raKiorRlyxZfpwMA8KVyHtgl0YPgLX5fQFStWlWffPKJBg0a5OtUAAC+VvzALjdfjogIX2d5QfD7AqJKlSqqXbu2r9MAAABn8FoBMXPmTCUlJSksLEzJycklzuXm5qp///6KjIxUo0aNtGjRIm+lBQAAzoHXJlHWq1dP48eP17Jly5T/p/GpsWPHKjQ0VIcOHdLGjRvVp08fJSYmqmXLlt5KDwAA2OC1HogBAwbopptuUkxMTInjx44d0+LFizVlyhRFRUWpU6dO6tevnxYsWOCt1AAAgE0+X8a5fft2BQcHKyEhwXUsMTFRq1atcv3cu3dvbdy4Udu2bdPo0aPPGgKRpNTUVKWmpnojZQCAPysoOD3ZEh7l8wLi6NGjio6OLnEsOjpaR44ccf28dOnScuN0795d3bt3lyQ99dRTFZskACBwFK/UKAv7PFQIn6/CiIqKUl5eXoljeXl5qlatmo8yAgAA5fF5D0RCQoKcTqd27Nihpk2bSpLS09PPaQJlWlqaNmzYUNEpAgCAP/FaD4TT6VRBQYEKCwtVWFiogoICOZ1ORUZGasCAAZowYYKOHTumNWvW6IMPPtDw4cNt3yMpKUmjR4/2QPYAAOBMXisgpk6dqoiICE2fPl1vvPGGIiIiNHXqVEnSyy+/rPz8fNWpU0dDhgzRrFmzWMIJAIAfcxhT+WaTOBwOeeJlOdxNypFkjHF7TXnnK1MMb+ZZ7mSpcmKUe403YgRKnv4SI1Dy9JcYgZKnv8SofG+LHuHzORAViTkQAAB4Bz0QNuO640+fyn0dgx4IP7xHZYoRKHn6S4xAydNfYlS+t0WP8PkyTgAAEHgoIAAAgG0UEAAAwDYmUQIAANuYRGkzrjv+NLHQ1zGYROmH96hMMQIlT3+JESh5+kuMyve26BEMYQAAANsoIAAAgG0UEAAAwDYmUQIAcKaCAik83O0lJj9fjoiIcz5fGTCJ0mZcd/xpYqGvYzCJ0g/vUZliBEqe/hIjUPL0lxgVdY9KjiEMAABgGwUEAACwjQICAADYxiRKAAAqWgVMxPR3TKK0Gdcdf5pY6OsYTKL0w3tUphiBkqe/xAiUPP0lhjfzDGAMYQAAANsoIAAAgG0UEAAAwDYKCAAAYBsFBAAAsI0CAgAA2MY+EAAA+CEr+0T4ci8J9oGwGdcdf9obwdcx2AfCD+9RmWIESp7+EiNQ8vSXGP60D4Qf7yXBEAYAALCNAgIAANhGAQEAAGyjgAAAALZRQAAAANsoIAAAgG0UEAAAwDY2kgIAALaxkZTNuO740+ZKvo7BRlJ+eI/KFCNQ8vSXGIGSp7/EYCMpSxjCAAAAtlFAAAAA2yggAACAbZV2DgQAADg3VkqDSrUKo5gnaqKkpCSlpaVVeNwLGW1asWjPikebVizas+L5sk0ZwgAAALZRQAAAANuCU1JSUnydRKBo06aNr1OodGjTikV7VjzatGLRnhXPV21aKSdRAgAAz2IIAwAA2EYBAQAAbKOAOENubq769++vyMhINWrUSIsWLSr1OmOMHnnkEcXExCgmJkYPP/ywR5aOBjqr7fnMM8+oVatWqlatmho3bqxnnnnGy5kGDqttWuzkyZNq1qyZGjRo4KUMA4ud9vz+++/VuXNnRUVFKS4uTjNmzPBipoHBanueOHFCY8aMUVxcnGrVqqUbbrhBWVlZXs42MMycOVNJSUkKCwtTcnKy22ufe+45XXTRRYqOjtbIkSN14sQJj+ZGAXGGsWPHKjQ0VIcOHdLChQt11113KSMj46zr/vWvf2nJkiVKT0/Xpk2b9PHHH2v27Nk+yNi/WW1PY4zmz5+vw4cP67PPPtPMmTP173//2wcZ+z+rbVrsmWeeUZ2gj2v+AAAUZ0lEQVQ6dbyYYWCx2p7Z2dnq1auXRo8erZycHO3cuVM9evTwQcb+zWp7zpgxQ+vWrdOmTZt04MAB1ahRQ/fee68PMvZ/9erV0/jx4zVy5Ei31y1btkzTp0/XihUrlJmZqd27d2vixImeTc7AGGPM0aNHTZUqVcy2bdtcx4YNG2YeeeSRs67t0KGDmT17tuvnV1991bRr184reQYKO+35Z/fee6+55557PJleQLLbprt37zbNmjUzS5cuNfXr1/dWmgHDTns+9thjZtiwYd5ML+DYac8xY8aYhx56yPXzxx9/bBISErySZ6B6/PHHzYgRI8o8P2TIEPPYY4+5fk5NTTVxcXEezYkeiP+zfft2BQcHKyEhwXUsMTGx1Oo5IyNDiYmJ5V53IbPTnmcyxuirr75Sy5YtPZ1iwLHbpvfee6+efPJJRUREeCvFgGKnPb/55hvVqlVLV199terUqaMbbrhB+/bt82a6fs9Oe44aNUpr1qzRgQMHdPz4cS1cuFDXX3+9N9OtdEp7Xzp06JBycnI8dk8KiP9z9OhRRUdHlzgWHR2tI0eOlHttdHS0jh49yjyIM9hpzzOlpKSoqKhIt99+uyfTC0h22vT999+X0+lU//79vZVewLHTnvv379frr7+uGTNmaN++fWrcuLGGDBnirVQDgp32TEhI0MUXX6z69eurevXq+umnnzRhwgRvpVoplfa+JKncf3PPBwXE/4mKilJeXl6JY3l5eapWrVq51+bl5SkqKoqHeJ3BTnsWmzlzpubPn69PPvlEYWFhnk4x4Fht02PHjunhhx/Wiy++6M30Ao6dv9GIiAj1799fV111lcLDwzVx4kStXbtWf/zxh7fS9Xt22vOuu+5SQUGBcnJydOzYMQ0YMIAeiPNU2vuSJLf/5p4vCoj/k5CQIKfTqR07driOpaenl9qV3rJlS6Wnp5d73YXMTntK0ty5c10TgFgxUDqrbbpjxw5lZmbqL3/5iy666CINGDBABw8e1EUXXaTMzEwvZ+2/7PyNXnHFFSU+IBR/T6/j/7PTnunp6UpOTlatWrUUFhame++9V99++62ys7O9mXKlUtr7UlxcnGJiYjx3U4/OsAgwN998s7nlllvM0aNHzddff22qV69utmzZctZ1s2bNMs2aNTP79+83WVlZpkWLFmbWrFk+yNi/WW3PN954w8TFxZkff/zRB1kGFitteurUKXPw4EHX1+LFi03dunXNwYMHjdPp9FHm/snq3+iKFStMjRo1zA8//GBOnjxp7rvvPtOpUycfZOzfrLZncnKyGTBggPn999/NyZMnzRNPPGHq1avng4z936lTp0x+fr559NFHzbBhw0x+fr45derUWdd9+umnJi4uzmRkZJjc3FzTrVs3S5PWzwcFxBlycnLMjTfeaKpWrWoaNmxoFi5caIwxZvXq1SYyMtJ1XVFRkXnooYdMzZo1Tc2aNc1DDz1kioqKfJW237LanvHx8SYkJMRERka6vkaPHu2rtP2a1TY908qVK1mFUQY77fnyyy+bevXqmRo1api+ffuaffv2+SJlv2a1PbOzs83QoUNN7dq1TXR0tOnYsaNZv369r9L2axMnTjSSSnxNnDjR7N2710RGRpq9e/e6rv3HP/5h6tSpY6pVq2aSk5NNQUGBR3PjWRgAAMA25kAAAADbKCAAAIBtFBAAAMA2CggAAGAbBQQAALCNAgIAANhGAQHYMG/ePDkcDnXt2tWnefz888/q1q2bIiIi5HA49PHHH/s0H09KSUmRw+FQcnKy2+u6du0qh8OhefPmeSwXh8Mhh8Nx3jt6Jicny+FwKCUlpULy8iRvtCsCEwUEPC4+Pt71D6/D4VBsbKx69uyptLQ0X6fmVnHeX375petYixYtNG7cOA0aNMh3iUmaNm2avvzyS1166aUaN26cLrnkkrOuKSoq0qRJk9SkSROFhYUpJiZGV111lebMmeO6pvjNoWrVqjp48KAk6ffffz/rjbL4Da/4q1q1amrTpo3+/e9/e/y1tm/fXuPGjVOPHj0kSV9++aUcDofi4+PPO/aZf5tr1qxxHf/qq69cx8+8z7hx4zRu3DhVr179vO7bo0cPjRs3Tu3btz+vOGUxxqhRo0alvvG/8847cjgciouLk9Pp9Mj9cWEI8XUCuHD07dtXjRs31qpVq/T555/ru+++09atW1WnTp2zrj116pSqVKnigyzd37tt27Zq27atlzM62/bt2yVJ999/v0aOHFnqNTNmzFBKSoouvvhi3X777crLy9OGDRu0bt06jRo1qsS1+fn5mjZtml544QW3923durW6dOmirVu3atmyZRoyZIhiYmJ03XXXVcwLK0WvXr3Uq1cvj8UvNmvWLHXs2FGS9D//8z+lXvP8889XyL2GDh2qoUOHVkis0jgcDg0dOlTTp0/Xm2++WaL3ZtGiRZKkW265RSEhvAXgPHh0n0vAGNOoUSMjybz//vvGGGN+++0315asS5YsMXv27HH9PGvWLFO3bl3TtWtXY4wx6enppmfPniYmJsbExsaavn37mq1bt54Ve/r06aZVq1YmKirK3HTTTSY7O9t1zapVq8xf/vIXEx0dberWrWuGDh1qsrKyXOeL7/3cc8+Z+Ph407hxY1fcM79ee+0189prrxlJpkuXLrbjv/jii6Zp06YmKirK3HrrrebEiRNltpm7mF26dDkrtz179pwV44YbbjCSzHvvvVfieE5Ojuv74lgOh8OEhYWZ/fv3m8OHD58Vd8SIEUaSGTdunOt3W7VqZSSZ++67zxhjzHPPPWcuueQSExYWZmJjY02XLl1K/Lcq9uuvvxpJpnr16qawsNDk5OQYh8Nh4uLijDHG7Nu3z0gytWvXNkVFRa6tfEeMGGFWrlx51msv/mes+LVMmDDBdOvWzURERJirr77aZGZmltnOxf+da9asacLCwsxvv/1mfv31VxMWFmZq1qxpJJlGjRqd9d+yuF3cveaFCxea5s2bm/DwcFOzZk3Tvn1789VXX5Voz4kTJxpj/n+74oEDB5rhw4ebyMhI06RJE7N8+XLXvTMyMky7du1MRESE6d27t7nnnnuMJHPjjTeW+toyMjKMJBMcHGwOHTpkjDHm999/N2FhYUaS+e6774wxxjzwwAOmUaNGJiwszERERJh27dqZlStXnvU38tprr5X6c/F/kzPbafPmzaZ3796mdu3aJjY21gwYMMC15XJRUZF57LHHTIMGDUxoaKiJi4szPXr0KPH/LAIDQxjwqqKiIq1atcr1c2xsbInzjz/+uK6//npdffXVOnjwoLp06aJly5apffv2uvLKK/Xxxx+ra9euOnz4cInfmzp1qq666irFxsZqyZIluvPOOyVJmzZtUvfu3fX111+rV69eatSokRYtWqSePXvq1KlTJWL893//tzp37qwePXpo5MiRrsfgDhw4UOPGjVOLFi3Oej124k+cOFFXX321nE6nFi5cqAULFpTaRuXFHDRokOrXry9Juu6668rsUq9bt64k6Y477tDw4cP18ssvKzMzU7Vq1Trr2sGDB+vEiRN68sknS83pz7Zu3aoDBw5IOv3fcOfOnbr//vuVl5en5ORkXXfdddq3b59rWORMtWvXVosWLZSXl6ctW7Zo7dq1Msbo0KFD2rlzp2sooXPnziWegClJDRo00MCBAyWdfkxx8ZDCmZ588knVrVtXsbGxWrt2rcaPH1/u6xkxYoROnDihuXPnau7cuTpx4kS5cy7cveb8/HwlJydr7969uvXWW9WnTx/l5eVp165dbmMuXrxYBw4cUKtWrbRr1y5X75LT6VS/fv20fv16tWjRQhEREZo1a5bbWC1atFDr1q1VWFiod955R5L07rvv6sSJE7rsssuUlJQkSdqzZ4/atWunUaNGqVu3blq/fr0GDx6sI0eOlNtupfnll1/UuXNnLV++XJ06dVK7du303nvvqWfPnjpx4oRWrFihadOmKTg4WKNGjVLnzp21efPmc74ffMjXFQwqv9I+zUsyN9xwgyksLCzRA7FixQrX7z311FNGkqs3whhjWrdubSSZ2bNnl4j9/PPPG2OM2bhxoyvWkSNHzF133WUkmeTkZGOMMSdPnjR16tQxksyyZcuMMf//qXLOnDml5n3mp7E/90DYif/2228bY4y57bbbjCQzduzYUtvLSsw/fwoszc8//2ySkpJKtHlQUJDrU++ZcZ577jnTs2dPExoaajZt2lRmD8Sfv+Lj482vv/5qfvzxRyPJXHHFFWbZsmXm559/NsaYMp/+WfwaZ82aZR599FETHx9vQkNDzWuvveb6ZP3CCy8YY0yJHghjSv/Ee+Zrufvuu40xxsydO9dIMi1btiyzjc7sHWvRooVp0qSJueSSS0yLFi3M+++/77YHwt1rPnLkiAkKCjL169c3H330kdm1a1eJ9iirB6Jly5amqKjI7N6923Wv3377zXz11VdGkomKijJHjx41xhjTr18/tz0Qxhjz7LPPGkmmY8eOxhhjrr32WiPJTJ061XVNTk6OmT17tnnsscfMuHHjTNWqVY0ks2bNmhLtarUH4umnnzaSTPPmzc24cePMuHHjTO3atY0k8+mnn5qlS5caSebaa681K1euNIcOHTJFRUWmsLCwzNcB/0QPBLymb9++uu+++zRlyhQtXbpUH3zwgYKCSv4JFo9BS3JN4GvevLnrWLNmzSRJe/fuLfF7xdcUn5ekrKyss2JUqVLFNeHwzzHOvLdVduJfeeWVkqQaNWpIko4ePXreMd1p0KCBvvvuO23atEkvvviiOnbsqKKiIk2ePFnZ2dlnXT958mSdPHlSTzzxRJkxW7durXHjxmn8+PGaP3++fvzxR9WuXVvNmzfXpEmTlJWVpZ49e6phw4Zq1qyZfvrpp1LjFK9iWbt2rdasWaNu3brpP/7jP7RmzRqtXbu2xDV2WW3nPxszZox27dql3bt366677ir3enevOSoqSrNmzZIxRjfccIOaNGmihg0b6quvvnIbs3Xr1nI4HK7ci/PPysqSJDVs2FCRkZGSVGqP2J8NHTpUQUFBWrt2rdavX++agHrrrbdKknJyctSqVSuNHj1a06ZN04wZM3T8+HFJ0m+//VZufEkqLCws8XPx3+9PP/2kGTNmaMaMGa5YO3fuVI8ePXT33Xfr22+/Vbdu3RQXF6d27drp0KFDlu4H/0EBAa8ZNWqUnnvuOY0fP17XX3/9Wd3TkhQWFub6vnj2+9atW13Htm3bJklq1KhRid8rfqM689r69eufFePUqVPavXt3qTHOvLckBQcHSzo97FIWO/GLJ6yV9rrPNaY733zzjY4fP67LL79c99xzj2uppzFGx44dO+v6tm3bqk+fPnr77bfLjNmlSxc9//zzmjJlioYPH66IiAhJp99EHn/8cWVnZ2vv3r165JFHtG3bNj333HNlxpGk1atX67vvvlPHjh3VsWNHpaamKj09XTExMWrVqlWpv1vefxer7fxnt912m6pWrarIyEgNHz683OvLe80jRoxQVlaWDhw4oBkzZmj//v2aMmWK25hl5V48ZLV//37l5+dLKvm3Xpa6devqmmuukTFGw4YNU2FhoTp16uT6G/vqq6908OBB1a5dW7/88otOnDjhKl5MGQ9qLi5g8vLyJElbtmwpcb449oABA2SMcX0dPHhQo0aNUmFhoWbOnKnff/9dO3fu1G233abvvvtOr776armvB/6FKbjwW8OGDdOTTz6plStXql+/fjp58qR++OEHxcXFnbWM8u9//7vS09O1cuVKSVL//v0VFRWlO++8U6+88opef/115efna+/evfr111/VsmXLcj/hNmzYULt379aECRP04Ycf6sEHHzzrmvOJX5aKivnss8/qiy++UMeOHdWgQQNt2LBBkpSQkKCLL7641N+ZNGmSPvnkE9s5//zzz2rXrp06d+6sOnXquOYxnPlJ+kxxcXFq1qyZ602wY8eOqlWrlv7xj39IKn3+Q7GGDRtKOv1m+re//U1NmzbVI488YjvnP4uOjtbq1atd35envNccFxenrl27ql69etq8eXOJc3a1b99el156qXbu3KmuXbvq4osv1ocffmjpd4cNG6bU1FTt3LnT9XOxuLg4Sad7G+6//37t3r273B6bK6+8UkuXLtU///lP7du376w3/ltvvVVPPvmka95DfHy8du3apVWrVmnHjh3KzMxUcnKyOnTooFq1apX7twL/RQ8E/Fa9evW0cuVK9ejRQ2vWrFFaWpr69OmjlStXnjURMCUlRd9//71+++039evXT//6178kne4S/vzzz9WhQwctXbpUe/bs0S233KLPPvtMoaGhbu+fkpKiSy+9VOvWrdOMGTNK7WI9n/hlqaiYN998s6666ipt2LBBc+fO1b59+9S/f3999NFHZb45t2nTRv369bOdc/Xq1dW2bVutWbNGr7zyig4cOKBbbrnF7QTG4l6ImJgYXXbZZSWGkIrPlSY+Pl7/9V//pejoaM2ZM6fMyajnok2bNmrTpo2la8t7zdddd52+//57zZkzRxkZGerTp4+rQLIrJCREH3zwgdq1a6fNmzfr+PHjrqW4f+45+7OBAweqatWqkqTQ0FANHjzYda5Dhw56/PHHVbNmTS1fvlxDhgxx9XaU5YEHHlCvXr2UnZ2tlStX6v777y9xvl69elq1apX69u2rjRs36o033lBWVpbGjh2r2NhY1a9fX02bNtWKFSv0yiuv6Pjx4xozZoxr4jMCh8OU1U8FBID4+Hjt3btXK1eu9PnukIAn/fHHHyV6Rnr27KnPP/9c48ePL3doBPAEhjAAIACkpKRo165dSkpK0saNG/X5558rMjKyzI3EAE9jCAMAAkBiYqIyMjL0xBNPaN26derdu7dWr16txo0b+zo1XKAYwgAAALbRAwEAAGyjgAAAALZRQAAAANsoIAAAgG0UEAAAwDYKCAAAYNv/AtP18Rcu34BsAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 540x378 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "fig , ax = plt.subplots()\n",
    "\n",
    "n , bins , patches = plt.hist( missing_values_per_isolate / number_of_SNPs , bins = 50 , color = 'k' , rwidth = 1.0 , edgecolor='white', linewidth=1.0)\n",
    "plt.hist( missing_values_per_isolate[missing_values_per_isolate > max_number_values_missing_per_isolate] / number_of_SNPs , bins = bins , color = 'red' , rwidth = 1.0 , edgecolor='white', linewidth=1.0)\n",
    "\n",
    "\n",
    "#plt.title('Mean Coverage across sequenced isolates in rollingDB', fontweight = 'bold' , fontsize = 12, color = 'k')\n",
    "plt.xlabel('Proportion of SNPs with Missing Values', fontweight = 'bold' , fontsize = 12, color = 'k')\n",
    "plt.ylabel('Number of Isolates', fontweight = 'bold' , fontsize = 12, color = 'k')\n",
    "\n",
    "ax.spines['right'].set_visible(False)\n",
    "ax.spines['top'].set_visible(False)\n",
    "\n",
    "ax.tick_params(labelcolor = 'k')\n",
    "ax.set_yscale(\"log\", nonposy='clip')\n",
    "\n",
    "fig = plt.gcf()\n",
    "fig.set_size_inches(7.5, 5.25)\n",
    "fig.tight_layout()\n",
    "\n",
    "plt.tick_params(axis='y', which='major', labelsize=12 , labelcolor = 'k')\n",
    "plt.tick_params(axis='x', which='major', labelsize=12 , labelcolor = 'k')\n",
    "\n",
    "#file_name = '/home/rv76/Farhat_Lab/Figures/Final_5/Scraping_rollingDB/Mean_Coverage_Distribution_across_all_isolates.pdf'\n",
    "#plt.savefig(file_name, bbox_inches='tight', dpi = 300 , transparent = True)\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of unique SNPs = 139797.0\n"
     ]
    }
   ],
   "source": [
    "print 'Number of unique SNPs = ' + str( number_of_SNPs )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Range and Average Number of Missing Values per unique SNP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "54"
      ]
     },
     "execution_count": 104,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.min(missing_values_per_isolate)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "5819.3169781241695"
      ]
     },
     "execution_count": 105,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.mean(missing_values_per_isolate)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "139797"
      ]
     },
     "execution_count": 106,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.max(missing_values_per_isolate) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Isolate with most missing values has 100.0% missing values\n"
     ]
    }
   ],
   "source": [
    "print 'Isolate with most missing values has ' + str( float(np.max(missing_values_per_isolate) ) / number_of_SNPs * 100) + '% missing values' "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Drop isolates that are flagged as 'missing values' in more than 25% of isolates in rollingDB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1518\n"
     ]
    }
   ],
   "source": [
    "print sum( missing_values_per_isolate > max_number_values_missing_per_isolate )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Percentage of isolates with many missing values that we're going to drop from analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4.481445398990346\n"
     ]
    }
   ],
   "source": [
    "number_of_unique_isolates = float(np.shape(genotypes_matrix)[1])\n",
    "print 100.0 * ( sum( missing_values_per_isolate > max_number_values_missing_per_isolate ) / number_of_unique_isolates )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### FILTER out low-quality isolates with $> 25\\%$ missing SNP calls"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Drop unique isolates with a high abundance of missing values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(139797, 33873)"
      ]
     },
     "execution_count": 110,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.shape( genotypes_matrix )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Drop isolates from the __GENOTYPE MATRIX__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "33873"
      ]
     },
     "execution_count": 111,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len( missing_values_per_isolate <= max_number_values_missing_per_isolate )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {},
   "outputs": [],
   "source": [
    "isolates_to_keep_filter = missing_values_per_isolate <= max_number_values_missing_per_isolate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "32355"
      ]
     },
     "execution_count": 113,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sum(isolates_to_keep_filter)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "metadata": {},
   "outputs": [],
   "source": [
    "genotypes_matrix = genotypes_matrix[: , isolates_to_keep_filter]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(139797, 32355)"
      ]
     },
     "execution_count": 115,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.shape(genotypes_matrix) #Genotype Matrix"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Drop isolates with a high abundance of missing values from __ISOLATE ANNOTATION FILE__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "metadata": {},
   "outputs": [],
   "source": [
    "isolate_annotation_series = pd.read_pickle('/n/data1/hms/dbmi/farhat/Roger/inhost_TB_dynamics_project/rolling_DB_scrape/Genotypes/genotypes_isolate_annotation.pkl')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    SAMEA3558733\n",
       "1    SAMN05575260\n",
       "2    SAMN03648641\n",
       "3    SAMN03647419\n",
       "4    SAMEA3671418\n",
       "dtype: object"
      ]
     },
     "execution_count": 117,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "isolate_annotation_series.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(33873,)"
      ]
     },
     "execution_count": 118,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.shape(isolate_annotation_series) #Annotation for Columns of Genotypes Matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "metadata": {},
   "outputs": [],
   "source": [
    "isolate_annotation_series = isolate_annotation_series[isolates_to_keep_filter]\n",
    "\n",
    "#re-index new filtered isolate annotation series (so new index matches indexing of genotypes matrix columns)\n",
    "isolate_annotation_series.reset_index(drop = True , inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    SAMEA3558733\n",
       "1    SAMN03648641\n",
       "2    SAMN03647419\n",
       "3    SAMEA3671418\n",
       "4    SAMN07659096\n",
       "dtype: object"
      ]
     },
     "execution_count": 120,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "isolate_annotation_series.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(32355,)"
      ]
     },
     "execution_count": 121,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.shape(isolate_annotation_series) #Annotation for Columns of Genotypes Matrix"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Save annotation files and genotypes matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "metadata": {},
   "outputs": [],
   "source": [
    "#save isolate annotation file (columns of Genotype Matrix)\n",
    "isolate_annotation_series.to_pickle('/n/data1/hms/dbmi/farhat/Roger/inhost_TB_dynamics_project/rolling_DB_scrape/Genotypes_Filtered/genotypes_isolate_annotation.pkl')\n",
    "\n",
    "#save SNP annotation file (rows of Genotype Matrix)\n",
    "SNP_annotation_DF.to_pickle('/n/data1/hms/dbmi/farhat/Roger/inhost_TB_dynamics_project/rolling_DB_scrape/Genotypes_Filtered/genotypes_SNP_annotation.pkl')\n",
    "\n",
    "#save Genotypes Matrix\n",
    "np.save('/n/data1/hms/dbmi/farhat/Roger/inhost_TB_dynamics_project/rolling_DB_scrape/Genotypes_Filtered/genotypes_matrix' , genotypes_matrix , allow_pickle = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "####################################################################################################################################################################################"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## [9] Annotate SNPs w/ H37Rv Locus Info"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "####################################################################################################################################################################################"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "#### Retrieve information for H37Rv Reference Genome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/rv76/anaconda2/lib/python2.7/site-packages/ipykernel_launcher.py:18: DeprecationWarning: \n",
      ".ix is deprecated. Please use\n",
      ".loc for label based indexing or\n",
      ".iloc for positional indexing\n",
      "\n",
      "See the documentation here:\n",
      "http://pandas.pydata.org/pandas-docs/stable/indexing.html#ix-indexer-is-deprecated\n"
     ]
    }
   ],
   "source": [
    "####################################################################################################################################################################################\n",
    "#load reference genome and reference annotation\n",
    "reference_genome = '/home/rv76/Farhat_Lab/Reference_Seqs/H37Rv/h37rv.fasta'\n",
    "for reference_genome in SeqIO.parse(reference_genome, \"fasta\"):\n",
    "    reference_genome.seq.alphabet = IUPAC.unambiguous_dna\n",
    "\n",
    "reference_genome_annotation = pd.read_csv('/n/data1/hms/dbmi/farhat/Roger/inhost_TB_dynamics_project/H37Rv/h37rv_genome_summary.txt', '\\t').set_index('name')\n",
    "\n",
    "####### Make a dictionary of all H37Rv reference positions and corresponding genes #######\n",
    "ReferencePosition_Gene_mapping = {} #keys: H37Rv Reference Positions , values: gene_ids (may be multiple genes for some Reference Positions)\n",
    "\n",
    "#store a list corresponding to every Reference Position (to store all genes that map to the Reference Position)\n",
    "for H37Rv_RefPos in range(0 , len(reference_genome.seq) + 1):\n",
    "    ReferencePosition_Gene_mapping[H37Rv_RefPos] = []\n",
    "\n",
    "for gene_id_index in range(0 , len(reference_genome_annotation.index)):\n",
    "\n",
    "    gene_id_info = reference_genome_annotation.ix[gene_id_index , :]\n",
    "    gene_id = gene_id_info.name\n",
    "\n",
    "    chrom_start = gene_id_info.chromStart\n",
    "    chrom_end = gene_id_info.chromEnd\n",
    "\n",
    "    #find the position of the first base relative to H37Rv in 5' -> 3'\n",
    "    H37Rv_start = min(chrom_start , chrom_end)\n",
    "    H37Rv_end = max(chrom_start , chrom_end)\n",
    "\n",
    "    #store all corresponding H37Rv Reference Positions to gene_id in dictionary\n",
    "    for H37Rv_RefPos in range(H37Rv_start+1 , H37Rv_end+1):\n",
    "\n",
    "        ReferencePosition_Gene_mapping[H37Rv_RefPos].append(gene_id) #append gene_id to list already in dict\n",
    "####################################################################################################################################################################################"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Annotate SNPs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To annotate SNPs, we get gene_ids that the SNPs 'land in' and the gene-relative coordinate."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>29</td>\n",
       "      <td>C</td>\n",
       "      <td>[G]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>31</td>\n",
       "      <td>A</td>\n",
       "      <td>[C, G]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>48</td>\n",
       "      <td>C</td>\n",
       "      <td>[T]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>64</td>\n",
       "      <td>G</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>71</td>\n",
       "      <td>C</td>\n",
       "      <td>[T]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   pos ref     alt\n",
       "0   29   C     [G]\n",
       "1   31   A  [C, G]\n",
       "2   48   C     [T]\n",
       "3   64   G     [C]\n",
       "4   71   C     [T]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#load SNP annotation file (rows of Genotype Matrix)\n",
    "SNP_annotation_DF = pd.read_pickle('/n/data1/hms/dbmi/farhat/Roger/inhost_TB_dynamics_project/rolling_DB_scrape/Genotypes_Filtered/genotypes_SNP_annotation.pkl')\n",
    "\n",
    "SNP_annotation_DF.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(139797, 3)"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.shape(SNP_annotation_DF)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Gene Categories"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "#get Gene Categories\n",
    "gene_categories = pd.read_csv('/n/data1/hms/dbmi/farhat/Roger/inhost_TB_dynamics_project/CSV_files/gene_categories/gene_categories.csv').set_index('name')\n",
    "\n",
    "#create dictionary, keys: H37Rv locus, values: gene category\n",
    "gene_categories_map = dict( zip( list(gene_categories.gene_id) , list(gene_categories.Gene_Category) ) ) "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Gene Names"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Parse through Mycobrowser (tuberculist) to collect gene names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "mycobrowser_table = pd.read_csv('/n/data1/hms/dbmi/farhat/Roger/inhost_TB_dynamics_project/H37Rv/Mycobrowser_Release3/Mycobacterium_tuberculosis_H37Rv_txt_v3.txt' , sep = '\\t')\n",
    "\n",
    "#retreive columns for locus tag & gene name\n",
    "mycobrowser_table = mycobrowser_table.loc[: , ['Locus' , 'Name']]\n",
    "\n",
    "#create dictionary, keys: H37Rv loucs, values: gene category\n",
    "gene_names_map = dict( zip( list(mycobrowser_table.Locus) , list(mycobrowser_table.Name) ) ) "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Iterate through each SNP to annotate"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Some SNPs are present in more than 1 gene, have to account for this"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.0\n",
      "0.200004291938\n",
      "0.400008583875\n",
      "0.600012875813\n",
      "0.80001716775\n"
     ]
    }
   ],
   "source": [
    "#create dictionaries for each columsn of information\n",
    "locus_tag_dict = {} #H37Rv id\n",
    "genomic_coord_dict = {} #coordinate on gene, 5' -> 3'\n",
    "gene_categ_dict = {} #gene Category\n",
    "gene_name_dict = {} #gene Name\n",
    "\n",
    "for SNP_index in SNP_annotation_DF.index:\n",
    "\n",
    "    #reference position, ref allele and alt allele for SNP\n",
    "    SNP_info = SNP_annotation_DF.loc[SNP_index , :]\n",
    "\n",
    "    #find the gene that SNP occurs on\n",
    "    ref_seq_position = SNP_info.pos\n",
    "\n",
    "    #check list corresponding to H37Rv coordinate to see if there are any genes associated with RefPosition\n",
    "    if len(ReferencePosition_Gene_mapping[ref_seq_position]) > 0:\n",
    "        \n",
    "        #create lists to store annotation information for SNP and associated gene(s)\n",
    "        gene_intergenic_id_list = []\n",
    "        gene_relative_coord_list = []\n",
    "        gene_categ_list = []\n",
    "        gene_name_list = []\n",
    "\n",
    "        #iterate through all genes that ReferencePosition is mapped to (i.e. SNP might correspond to 2 genes)\n",
    "        for gene_intergenic_id in ReferencePosition_Gene_mapping[ref_seq_position]:\n",
    "\n",
    "            #find genomic coordinate of SNP relative to gene (subtract 1 since reference seq starts counting at 1)\n",
    "            gene_relative_coord = (ref_seq_position - 1) - min( reference_genome_annotation.loc[gene_intergenic_id , 'chromStart'] , reference_genome_annotation.loc[gene_intergenic_id , 'chromEnd'] )\n",
    "\n",
    "            gene_intergenic_id_list.append( gene_intergenic_id )\n",
    "            gene_relative_coord_list.append( gene_relative_coord )\n",
    "            \n",
    "            #use gene id to get category and names from mappings\n",
    "            if gene_intergenic_id in gene_categories_map.keys():\n",
    "                gene_categ_list.append( gene_categories_map[gene_intergenic_id] )\n",
    "            else:\n",
    "                gene_categ_list.append( 'None' )\n",
    "                \n",
    "            if gene_intergenic_id in gene_names_map.keys():\n",
    "                gene_name_list.append( gene_names_map[gene_intergenic_id] ) \n",
    "            else:\n",
    "                gene_name_list.append( 'None' )\n",
    "\n",
    "        #store all information for SNP in relevant dictionaries\n",
    "        locus_tag_dict[SNP_index] = gene_intergenic_id_list\n",
    "        genomic_coord_dict[SNP_index] = gene_relative_coord_list\n",
    "        gene_categ_dict[SNP_index] = gene_categ_list\n",
    "        gene_name_dict[SNP_index] = gene_name_list\n",
    "            \n",
    "    #H37Rv RefSeq position was not found in dictionary so Reference position must be intergenic    \n",
    "    elif len(ReferencePosition_Gene_mapping[ref_seq_position]) == 0:\n",
    "\n",
    "        #store all information for SNP in relevant dictionaries\n",
    "        locus_tag_dict[SNP_index] = ['intergenic']\n",
    "        genomic_coord_dict[SNP_index] = [np.nan]\n",
    "        gene_categ_dict[SNP_index] = ['None']\n",
    "        gene_name_dict[SNP_index] = ['None']\n",
    "\n",
    "    #keep track of progress of loop\n",
    "    num_SNPs = float( np.shape(SNP_annotation_DF)[0] )\n",
    "    if SNP_index % np.ceil(0.20 * num_SNPs) == 0:\n",
    "        print float(SNP_index) / num_SNPs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Update SNP annotation DF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "#append columns for info from annotation\n",
    "SNP_annotation_DF['gene_id'] = pd.Series(locus_tag_dict)\n",
    "SNP_annotation_DF['gene_pos'] = pd.Series(genomic_coord_dict)\n",
    "SNP_annotation_DF['gene_category'] = pd.Series(gene_categ_dict)\n",
    "SNP_annotation_DF['gene_name'] = pd.Series(gene_name_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(139797, 7)"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.shape( SNP_annotation_DF )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_pos</th>\n",
       "      <th>gene_category</th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>29</td>\n",
       "      <td>C</td>\n",
       "      <td>[G]</td>\n",
       "      <td>[Rv0001]</td>\n",
       "      <td>[28]</td>\n",
       "      <td>[Essential]</td>\n",
       "      <td>[dnaA]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>31</td>\n",
       "      <td>A</td>\n",
       "      <td>[C, G]</td>\n",
       "      <td>[Rv0001]</td>\n",
       "      <td>[30]</td>\n",
       "      <td>[Essential]</td>\n",
       "      <td>[dnaA]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>48</td>\n",
       "      <td>C</td>\n",
       "      <td>[T]</td>\n",
       "      <td>[Rv0001]</td>\n",
       "      <td>[47]</td>\n",
       "      <td>[Essential]</td>\n",
       "      <td>[dnaA]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>64</td>\n",
       "      <td>G</td>\n",
       "      <td>[C]</td>\n",
       "      <td>[Rv0001]</td>\n",
       "      <td>[63]</td>\n",
       "      <td>[Essential]</td>\n",
       "      <td>[dnaA]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>71</td>\n",
       "      <td>C</td>\n",
       "      <td>[T]</td>\n",
       "      <td>[Rv0001]</td>\n",
       "      <td>[70]</td>\n",
       "      <td>[Essential]</td>\n",
       "      <td>[dnaA]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   pos ref     alt   gene_id gene_pos gene_category gene_name\n",
       "0   29   C     [G]  [Rv0001]     [28]   [Essential]    [dnaA]\n",
       "1   31   A  [C, G]  [Rv0001]     [30]   [Essential]    [dnaA]\n",
       "2   48   C     [T]  [Rv0001]     [47]   [Essential]    [dnaA]\n",
       "3   64   G     [C]  [Rv0001]     [63]   [Essential]    [dnaA]\n",
       "4   71   C     [T]  [Rv0001]     [70]   [Essential]    [dnaA]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SNP_annotation_DF.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_pos</th>\n",
       "      <th>gene_category</th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>139792</th>\n",
       "      <td>4411459</td>\n",
       "      <td>C</td>\n",
       "      <td>[T, G]</td>\n",
       "      <td>[intergenic]</td>\n",
       "      <td>[nan]</td>\n",
       "      <td>[None]</td>\n",
       "      <td>[None]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>139793</th>\n",
       "      <td>4411467</td>\n",
       "      <td>C</td>\n",
       "      <td>[T]</td>\n",
       "      <td>[intergenic]</td>\n",
       "      <td>[nan]</td>\n",
       "      <td>[None]</td>\n",
       "      <td>[None]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>139794</th>\n",
       "      <td>4411475</td>\n",
       "      <td>G</td>\n",
       "      <td>[A]</td>\n",
       "      <td>[intergenic]</td>\n",
       "      <td>[nan]</td>\n",
       "      <td>[None]</td>\n",
       "      <td>[None]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>139795</th>\n",
       "      <td>4411485</td>\n",
       "      <td>G</td>\n",
       "      <td>[A, C]</td>\n",
       "      <td>[intergenic]</td>\n",
       "      <td>[nan]</td>\n",
       "      <td>[None]</td>\n",
       "      <td>[None]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>139796</th>\n",
       "      <td>4411501</td>\n",
       "      <td>C</td>\n",
       "      <td>[T]</td>\n",
       "      <td>[intergenic]</td>\n",
       "      <td>[nan]</td>\n",
       "      <td>[None]</td>\n",
       "      <td>[None]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            pos ref     alt       gene_id gene_pos gene_category gene_name\n",
       "139792  4411459   C  [T, G]  [intergenic]    [nan]        [None]    [None]\n",
       "139793  4411467   C     [T]  [intergenic]    [nan]        [None]    [None]\n",
       "139794  4411475   G     [A]  [intergenic]    [nan]        [None]    [None]\n",
       "139795  4411485   G  [A, C]  [intergenic]    [nan]        [None]    [None]\n",
       "139796  4411501   C     [T]  [intergenic]    [nan]        [None]    [None]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SNP_annotation_DF.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "#update SNP annotation file (rows of Genotype Matrix) with extra SNP information\n",
    "SNP_annotation_DF.to_pickle('/n/data1/hms/dbmi/farhat/Roger/inhost_TB_dynamics_project/rolling_DB_scrape/Genotypes_Filtered/genotypes_SNP_annotation_with_gene_info.pkl')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "How to investigate SNPs that appeared on more than 1 gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_pos</th>\n",
       "      <th>gene_category</th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3337</th>\n",
       "      <td>60400</td>\n",
       "      <td>C</td>\n",
       "      <td>[T]</td>\n",
       "      <td>[Rv0057, Rv0058]</td>\n",
       "      <td>[504, 4]</td>\n",
       "      <td>[Non-Essential, Essential]</td>\n",
       "      <td>[Rv0057, dnaB]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3338</th>\n",
       "      <td>60414</td>\n",
       "      <td>C</td>\n",
       "      <td>[T]</td>\n",
       "      <td>[Rv0057, Rv0058]</td>\n",
       "      <td>[518, 18]</td>\n",
       "      <td>[Non-Essential, Essential]</td>\n",
       "      <td>[Rv0057, dnaB]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4089</th>\n",
       "      <td>77620</td>\n",
       "      <td>C</td>\n",
       "      <td>[G]</td>\n",
       "      <td>[Rv0069c, Rv0070c]</td>\n",
       "      <td>[1383, 1]</td>\n",
       "      <td>[Non-Essential, Non-Essential]</td>\n",
       "      <td>[sdaA, glyA2]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4751</th>\n",
       "      <td>95414</td>\n",
       "      <td>G</td>\n",
       "      <td>[A, C]</td>\n",
       "      <td>[Rv0086, Rv0087]</td>\n",
       "      <td>[1463, 0]</td>\n",
       "      <td>[Essential, Non-Essential]</td>\n",
       "      <td>[hycQ, hycE]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4752</th>\n",
       "      <td>95415</td>\n",
       "      <td>T</td>\n",
       "      <td>[C]</td>\n",
       "      <td>[Rv0086, Rv0087]</td>\n",
       "      <td>[1464, 1]</td>\n",
       "      <td>[Essential, Non-Essential]</td>\n",
       "      <td>[hycQ, hycE]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        pos ref     alt             gene_id   gene_pos  \\\n",
       "3337  60400   C     [T]    [Rv0057, Rv0058]   [504, 4]   \n",
       "3338  60414   C     [T]    [Rv0057, Rv0058]  [518, 18]   \n",
       "4089  77620   C     [G]  [Rv0069c, Rv0070c]  [1383, 1]   \n",
       "4751  95414   G  [A, C]    [Rv0086, Rv0087]  [1463, 0]   \n",
       "4752  95415   T     [C]    [Rv0086, Rv0087]  [1464, 1]   \n",
       "\n",
       "                       gene_category       gene_name  \n",
       "3337      [Non-Essential, Essential]  [Rv0057, dnaB]  \n",
       "3338      [Non-Essential, Essential]  [Rv0057, dnaB]  \n",
       "4089  [Non-Essential, Non-Essential]   [sdaA, glyA2]  \n",
       "4751      [Essential, Non-Essential]    [hycQ, hycE]  \n",
       "4752      [Essential, Non-Essential]    [hycQ, hycE]  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SNP_annotation_DF[ SNP_annotation_DF.gene_name.apply(len) > 1 ].head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "How to investigate SNPs that appeared within a particular gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_pos</th>\n",
       "      <th>gene_category</th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>27629</th>\n",
       "      <td>759814</td>\n",
       "      <td>A</td>\n",
       "      <td>[G]</td>\n",
       "      <td>[Rv0667]</td>\n",
       "      <td>[7]</td>\n",
       "      <td>[Antibiotic Resistance]</td>\n",
       "      <td>[rpoB]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27630</th>\n",
       "      <td>759818</td>\n",
       "      <td>C</td>\n",
       "      <td>[T]</td>\n",
       "      <td>[Rv0667]</td>\n",
       "      <td>[11]</td>\n",
       "      <td>[Antibiotic Resistance]</td>\n",
       "      <td>[rpoB]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27631</th>\n",
       "      <td>759841</td>\n",
       "      <td>G</td>\n",
       "      <td>[C]</td>\n",
       "      <td>[Rv0667]</td>\n",
       "      <td>[34]</td>\n",
       "      <td>[Antibiotic Resistance]</td>\n",
       "      <td>[rpoB]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27632</th>\n",
       "      <td>759852</td>\n",
       "      <td>A</td>\n",
       "      <td>[T]</td>\n",
       "      <td>[Rv0667]</td>\n",
       "      <td>[45]</td>\n",
       "      <td>[Antibiotic Resistance]</td>\n",
       "      <td>[rpoB]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27633</th>\n",
       "      <td>759862</td>\n",
       "      <td>A</td>\n",
       "      <td>[G]</td>\n",
       "      <td>[Rv0667]</td>\n",
       "      <td>[55]</td>\n",
       "      <td>[Antibiotic Resistance]</td>\n",
       "      <td>[rpoB]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          pos ref  alt   gene_id gene_pos            gene_category gene_name\n",
       "27629  759814   A  [G]  [Rv0667]      [7]  [Antibiotic Resistance]    [rpoB]\n",
       "27630  759818   C  [T]  [Rv0667]     [11]  [Antibiotic Resistance]    [rpoB]\n",
       "27631  759841   G  [C]  [Rv0667]     [34]  [Antibiotic Resistance]    [rpoB]\n",
       "27632  759852   A  [T]  [Rv0667]     [45]  [Antibiotic Resistance]    [rpoB]\n",
       "27633  759862   A  [G]  [Rv0667]     [55]  [Antibiotic Resistance]    [rpoB]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SNP_annotation_DF[['Rv0667' in gene_list for gene_list in SNP_annotation_DF.gene_id]].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
