{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>.container { width:100% !important; }</style>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from IPython.core.display import display, HTML\n",
    "display(HTML(\"<style>.container { width:100% !important; }</style>\"))\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### This notebook is used to create the script (& submit the jobs for) mapping the CDS regions from each Reference Genome (downloaded from NCBI) against the CDS regions of H37Rv. The goal is to assess the quality of the mappings across genes on the entire genome to determine which Reference Genomes to use in our simulations."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "################################################################################################################################################################################################################################################"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## [1] Create the Script to map the CDS regions from a given Reference Genome against the CDS regions of H37Rv "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "################################################################################################################################################################################################################################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#Choose a specific reference genome, pull sequence & annotation; compare to H37Rv (indexed: 0 - 59)\n",
    "ref_genome_i = 8 #H37Rv (fasta & annotation table) as downloaded from NCBI"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "#import necessary packages\n",
    "import vcf\n",
    "import os\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "from itertools import compress\n",
    "import ast\n",
    "import itertools\n",
    "import time\n",
    "\n",
    "import Bio\n",
    "from Bio.Alphabet import IUPAC\n",
    "from Bio.Blast.Applications import NcbiblastnCommandline\n",
    "from Bio.Blast import NCBIXML\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.SeqFeature import SeqFeature, FeatureLocation\n",
    "from Bio import pairwise2\n",
    "from Bio import SeqIO\n",
    "from Bio.Graphics import GenomeDiagram\n",
    "from Bio.SeqUtils import GC\n",
    "\n",
    "from Bio.Align.Applications import MuscleCommandline\n",
    "from StringIO import StringIO\n",
    "from Bio import AlignIO\n",
    "from Bio.Align import AlignInfo\n",
    "from Bio.Seq import MutableSeq\n",
    "from Bio import pairwise2\n",
    "from shutil import copy\n",
    "import subprocess\n",
    "import pickle"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "##########################################################################################\n",
    "##########################################################################################\n",
    "'''\n",
    "This piece of code constructs a mapping of the CDS regions of H37Rv to the CDS \n",
    "regions of the chosen reference sequence (ref_genome_i) using the \n",
    "feature annotation tables for both sequences.\n",
    "'''\n",
    "\n",
    "# check to see if the reference genomes have a corresponding feature annotation table\n",
    "\n",
    "#annotation for reference sequences\n",
    "ref_seqs_annot = pd.read_csv('/n/data1/hms/dbmi/farhat/Roger/inhost_TB_dynamics_project/simulated_reads/reference_sequences/strains_typed/final_table.tsv' ,  sep = '\\t')\n",
    "\n",
    "#directory where annotation files are stored\n",
    "ref_strain_genome_annot_files_directory = '/n/data1/hms/dbmi/farhat/Roger/inhost_TB_dynamics_project/simulated_reads/reference_sequences/strains_typed/reference_sequence_annotation_tables/'\n",
    "reference_strain_feature_counts_tables = os.listdir(ref_strain_genome_annot_files_directory)\n",
    "\n",
    "reference_genome_strain_has_feature_table = []\n",
    "reference_genome_strain_feature_table_names = []\n",
    "\n",
    "for reference_genome_strain in ref_seqs_annot.Strain:\n",
    "\n",
    "    #split the reference id to get back relevant info to find the filename\n",
    "    genbank , reference_id_1 , reference_id_2 , genomic_word = reference_genome_strain.split('_')\n",
    "    folder_1 = reference_id_1[0:3]\n",
    "    folder_2 = reference_id_1[3:6]\n",
    "    folder_3 = reference_id_1[6:9]\n",
    "    folder_4 = 'GCF_' + reference_id_1 + '_' + reference_id_2\n",
    "\n",
    "    reference_strain_annotation_filename = folder_4 + '_feature_table.txt.gz'\n",
    "\n",
    "    if reference_strain_annotation_filename in reference_strain_feature_counts_tables:\n",
    "        reference_genome_strain_has_feature_table.append('YES')\n",
    "        reference_genome_strain_feature_table_names.append(reference_strain_annotation_filename)\n",
    "\n",
    "    else:\n",
    "        reference_genome_strain_has_feature_table.append('NO')\n",
    "        reference_genome_strain_feature_table_names.append('No Table')\n",
    "\n",
    "ref_seqs_annot.loc[: , 'Has_Feature_Table'] = reference_genome_strain_has_feature_table\n",
    "ref_seqs_annot.loc[: , 'Feature_Table_filename'] = reference_genome_strain_feature_table_names\n",
    "\n",
    "#drop strains that don't have a corresponding feature table\n",
    "ref_seqs_annot = ref_seqs_annot[ref_seqs_annot.Has_Feature_Table == 'YES']\n",
    "ref_seqs_annot.reset_index(drop = True , inplace = True)\n",
    "\n",
    "#############################################################################\n",
    "\n",
    "#PULL H37Rv sequence & annotation information\n",
    "\n",
    "#H37Rv GENOME SEQUENCE\n",
    "H37Rv_genome = '/n/data1/hms/dbmi/farhat/Roger/inhost_TB_dynamics_project/H37Rv/h37rv.fasta'\n",
    "for H37Rv_seq in SeqIO.parse(H37Rv_genome, \"fasta\"):\n",
    "    H37Rv_seq.seq.alphabet = IUPAC.unambiguous_dna\n",
    "\n",
    "#H37Rv REFERENCE ANNOTATION\n",
    "H37Rv_annotation = pd.read_csv('/n/data1/hms/dbmi/farhat/Roger/inhost_TB_dynamics_project/H37Rv/h37rv_genome_summary.txt', '\\t').set_index('name')\n",
    "\n",
    "#filter out annotation for t-RNAs and other misc RNAs\n",
    "RNAs_filter = []\n",
    "for CDS_i in H37Rv_annotation.index:\n",
    "    if ('tRNA' in H37Rv_annotation.loc[CDS_i , 'description'].split(' ')) or ('Rvn' in CDS_i):\n",
    "        RNAs_filter.append(True)\n",
    "        \n",
    "    else:\n",
    "        RNAs_filter.append(False)\n",
    "\n",
    "non_RNAs_filter = [not RNA for RNA in RNAs_filter]\n",
    "\n",
    "H37Rv_annotation_no_RNA = H37Rv_annotation[non_RNAs_filter]\n",
    "\n",
    "#drop unneccessary columns\n",
    "H37Rv_annotation_no_RNA = H37Rv_annotation_no_RNA.loc[: , ['symbol' , 'length' , 'chromStart' , 'chromEnd' , 'strand' , 'description']]\n",
    "\n",
    "\n",
    "#STORE CORRESPONDING SEQUENCES FROM H37Rv IN GENOME ANNOTATION\n",
    "for gene_id in H37Rv_annotation_no_RNA.index:\n",
    "\n",
    "    start_site = min( (H37Rv_annotation_no_RNA.loc[gene_id , 'chromStart']) , H37Rv_annotation_no_RNA.loc[gene_id , 'chromEnd'] ) #starts counting from 1\n",
    "    end_site = max( (H37Rv_annotation_no_RNA.loc[gene_id , 'chromStart']) , H37Rv_annotation_no_RNA.loc[gene_id , 'chromEnd'] )\n",
    "\n",
    "    seq_i = H37Rv_seq.seq[start_site:end_site] \n",
    "\n",
    "    #store sequence\n",
    "    H37Rv_annotation_no_RNA.loc[gene_id , 'sequence'] = seq_i\n",
    "\n",
    "#############################################################################\n",
    "#PULL REFERENCE sequence & annotation information\n",
    "\n",
    "#specify specific reference sequence\n",
    "ref_genome_strain = list(ref_seqs_annot.Strain)[ref_genome_i]\n",
    "ref_genome_strain_annotation_filename = list(ref_seqs_annot.Feature_Table_filename)[ref_genome_i]\n",
    "\n",
    "#directory where the fasta files for the reference sequences are stored\n",
    "ref_seq_directory = '/n/data1/hms/dbmi/farhat/Roger/inhost_TB_dynamics_project/simulated_reads/reference_sequences/strains_typed/'\n",
    "#directory where annotation files are stored\n",
    "ref_strain_genome_annot_files_directory = '/n/data1/hms/dbmi/farhat/Roger/inhost_TB_dynamics_project/simulated_reads/reference_sequences/strains_typed/reference_sequence_annotation_tables/'\n",
    "\n",
    "#REFERENCE GENOME SEQUENCE\n",
    "#get full path for the fasta file corrresponding to the reference sequence\n",
    "ref_genome_fasta = ref_seq_directory + ref_genome_strain  + '.fasta'\n",
    "\n",
    "#parse the fasta file corresponding to the reference genome\n",
    "for reference_genome in SeqIO.parse(ref_genome_fasta, \"fasta\"):\n",
    "    reference_genome.seq.alphabet = IUPAC.unambiguous_dna\n",
    "\n",
    "#REFERENCE GENOME FEATURE TABLE\n",
    "ref_genome_annot = pd.read_csv(ref_strain_genome_annot_files_directory + ref_genome_strain_annotation_filename , compression = 'gzip' , header = 0 , sep = '\\t')\n",
    "\n",
    "#filter for rows that code for 'genes'\n",
    "ref_genome_annot = ref_genome_annot[ref_genome_annot.ix[: , 0] == 'gene']\n",
    "\n",
    "#filter for rows that are 'protein_coding' or 'pseudogene'\n",
    "protein_coding_or_pseudogene_filter = [protein_coding or pseudogene for protein_coding , pseudogene in zip(list(ref_genome_annot.ix[: , 1] == 'protein_coding') , list(ref_genome_annot.ix[: , 1] == 'pseudogene'))]\n",
    "ref_genome_annot = ref_genome_annot[protein_coding_or_pseudogene_filter]\n",
    "\n",
    "#filter out for buggy first rows that contain the start/end of the whole Reference Genome Sequence\n",
    "ref_genome_annot = ref_genome_annot[ (ref_genome_annot.end - ref_genome_annot.start) < 1000000 ]\n",
    "\n",
    "#drop unnecessary columns\n",
    "ref_genome_annot = ref_genome_annot.loc[: , [ 'start' , 'end' , 'strand' , 'symbol' , 'feature_interval_length']]\n",
    "\n",
    "ref_genome_annot.reset_index(inplace = True , drop = True)\n",
    "\n",
    "#STORE CORRESPONDING SEQUENCES FROM REFERENCE GENOME IN GENOME ANNOTATION\n",
    "for CDS_i in ref_genome_annot.index:\n",
    "\n",
    "    start_site = min( (ref_genome_annot.loc[CDS_i , 'start'] - 1) , ref_genome_annot.loc[CDS_i , 'end'] ) #starts counting from 1\n",
    "    end_site = max( (ref_genome_annot.loc[CDS_i , 'start'] - 1) , ref_genome_annot.loc[CDS_i , 'end'] )\n",
    "\n",
    "    seq_i = reference_genome.seq[start_site:end_site] \n",
    "\n",
    "    #store sequence\n",
    "    ref_genome_annot.loc[CDS_i , 'sequence'] = seq_i\n",
    "    \n",
    "#############################################################################\n",
    "#############################################################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "#############################################################################\n",
    "#############################################################################\n",
    "'''\n",
    "This piece of code compares the given Reference Genome with H37Rv & attempts \n",
    "to map ALL Relevant Coding Regions between both Reference Genomes for ALL Genes.\n",
    "'''\n",
    "\n",
    "#define window size for CDS length comparison (has to be an even number)\n",
    "window_size = 10\n",
    "\n",
    "num_genes = len(H37Rv_annotation_no_RNA.index) #number of genes after excluding RNAs\n",
    "\n",
    "#WRAP AROUND CHROMOSOME FOR FIRST/LAST FEW GENES\n",
    "#in order to 'wrap around' chromosome, append last 5 genes to beginning of DF and first 5 genes to end of DF to create sliding window for all genes\n",
    "last_few_genes_annot = H37Rv_annotation_no_RNA.ix[num_genes - (window_size / 2) : , :] #last few genes\n",
    "first_few_genes_annot = H37Rv_annotation_no_RNA.ix[0 : (window_size / 2) , :] #first few genes\n",
    "H37Rv_annotation_no_RNA_fudged_for_circular_chrom = last_few_genes_annot.append(H37Rv_annotation_no_RNA.append(first_few_genes_annot))\n",
    "\n",
    "#H37Rv-ref_genome coding mapping\n",
    "#create dataframe to hold mapping\n",
    "H37Rv_RefGeome_gene_mapping = pd.DataFrame(columns = ['symbol' , 'H37Rv_start' , 'H37Rv_end' , 'RefGenome_start' , 'RefGenome_end' , 'strand' , 'Highest_window_score' , 'second_Highest_window_score' , 'window_index'])\n",
    "\n",
    "#look across all genes in H37Rv\n",
    "all_genes = (H37Rv_annotation_no_RNA_fudged_for_circular_chrom.index) \n",
    "\n",
    "for gene_id in all_genes[ (window_size / 2) : - (window_size / 2) ]: #skip first several rows & last several rows (which are redundant genes created for sliding windows) \n",
    "\n",
    "    #find index in H37Rv annotation dataframe\n",
    "    row_index = list(H37Rv_annotation_no_RNA_fudged_for_circular_chrom.ix[(window_size / 2) : - (window_size / 2)].index).index(gene_id) + 5\n",
    "\n",
    "    #take the 5 CDS regions before & after into account\n",
    "    H37Rv_annotation_no_RNA_neighborhood = H37Rv_annotation_no_RNA_fudged_for_circular_chrom.ix[ row_index - (window_size / 2) : row_index + ((window_size / 2) + 1) , :]\n",
    "    \n",
    "    #look at gene lengths of the 11 consective genes & look for similar gene length profile in reference genome\n",
    "    H37Rv_annotation_no_RNA_neighborhood_lengths = H37Rv_annotation_no_RNA_neighborhood.length\n",
    "\n",
    "    #compare the neighborhood of gene lengths with every set of consecutive 11 gene lengths in reference genome annotation\n",
    "    ## key: window_index , value: score of gene_length_window_match\n",
    "    gene_length_match_scores = {}\n",
    "\n",
    "    #create a sliding window of gene lengths to find the appropriate coding region on the reference genome for novel genes\n",
    "    first_window_index = window_size / 2\n",
    "    last_window_index = np.shape(ref_genome_annot)[0] - (window_size / 2)\n",
    "\n",
    "    for gene_lengths_window_i in range(first_window_index , last_window_index):\n",
    "\n",
    "        #take the 5 CDS regions before & after into account\n",
    "        ref_genome_annot_neighborhood = ref_genome_annot.ix[ gene_lengths_window_i - (window_size / 2) : gene_lengths_window_i + (window_size / 2) , :]\n",
    "\n",
    "        #look at gene lengths of the 11 consecutive genes & look for similar gene length profile in H37Rv\n",
    "        ref_genome_annot_neighborhood_lengths =  ref_genome_annot_neighborhood.feature_interval_length\n",
    "        \n",
    "        #compare window of gene lengths between reference genome and H37Rv\n",
    "        length_match_score = len( set(list(ref_genome_annot_neighborhood_lengths)).intersection(set(list(H37Rv_annotation_no_RNA_neighborhood_lengths))) ) #rough match method\n",
    "\n",
    "        #store window index and corresponding match score\n",
    "        gene_length_match_scores[gene_lengths_window_i] = length_match_score\n",
    "\n",
    "    #find the max match score and corresponding window index\n",
    "    gene_length_match_scores = pd.Series(gene_length_match_scores)\n",
    "    \n",
    "    gene_length_match_scores_genome_order = gene_length_match_scores\n",
    "    \n",
    "    gene_length_match_scores = gene_length_match_scores.sort_values(ascending = False)\n",
    "    \n",
    "    window_index_top_match_score = list(gene_length_match_scores)[0] #top score for sliding window alignments\n",
    "    \n",
    "    #store window indice(s) that have the top match score\n",
    "    window_indices_with_top_score = list(gene_length_match_scores[gene_length_match_scores == window_index_top_match_score].index)\n",
    "    \n",
    "    gene_alignment_scores_in_top_window = {} #key: sub-window indices , valus: score from pariwise alignment\n",
    "    \n",
    "    #iterate through all highest scoring windows (or neighborhoods)\n",
    "    for window_match_index in window_indices_with_top_score:\n",
    "        \n",
    "        #align gene_i to every CDS within the 'sliding window' of the highest window scoring CDS region(s)\n",
    "        first_chosen_window_index = window_match_index - (window_size / 2)\n",
    "        last_chosen_window_index = window_match_index + (window_size / 2)\n",
    "\n",
    "        for window_index in range(first_chosen_window_index , last_chosen_window_index + 1):\n",
    "\n",
    "            #find the gene sequence corresponding to window index\n",
    "            corresponding_gene_ref_genome = ref_genome_annot.loc[window_index , :]\n",
    "\n",
    "            #check to see if sequences are identical by aligning them\n",
    "            H37Rv_coding_seq = H37Rv_annotation_no_RNA.loc[gene_id , 'sequence']\n",
    "            ref_genome_coding_seq = corresponding_gene_ref_genome.sequence\n",
    "\n",
    "            #Identical characters are given 1 point, -1/2 point is deducted for each non-identical character, 2 points are deduced when opening a gap, 2 point is deducted for extending it\n",
    "            alignment_score = pairwise2.align.globalms(H37Rv_coding_seq, ref_genome_coding_seq , 2 , -1 , -2 , -2 , score_only = True)\n",
    "\n",
    "            #store in dictionary\n",
    "            gene_alignment_scores_in_top_window[window_index] = alignment_score\n",
    "            \n",
    "    #choose window index with highest alignment score as match (this time from genes within top/chosen window)\n",
    "    top_window_score_gene_alignments_results = pd.Series(gene_alignment_scores_in_top_window)\n",
    "    top_window_score_gene_alignments_results = top_window_score_gene_alignments_results.sort_values(ascending = False)\n",
    "    \n",
    "    window_match_index = list(top_window_score_gene_alignments_results.index)[0] #take the index corresponding to the top alignment score\n",
    "    window_match_score = list(top_window_score_gene_alignments_results)[0] #actual raw score from top alignment\n",
    "    \n",
    "    #store relevant information\n",
    "    corresponding_gene_ref_genome = ref_genome_annot.loc[window_match_index , :]\n",
    "    \n",
    "    H37Rv_gene_length = H37Rv_annotation_no_RNA.loc[gene_id , 'length']\n",
    "    RefGenome_gene_length = corresponding_gene_ref_genome.feature_interval_length\n",
    "    \n",
    "    #if window match score positive and gene lengths match, then likely a correct match also ensure 'found' gene has same length as gene in H37Rv\n",
    "    if (window_match_score > 0) and (H37Rv_gene_length == RefGenome_gene_length): \n",
    "\n",
    "        symbol = H37Rv_annotation_no_RNA.loc[gene_id , 'symbol']\n",
    "        strand = H37Rv_annotation_no_RNA.loc[gene_id , 'strand']\n",
    "        H37Rv_start = H37Rv_annotation_no_RNA.loc[gene_id , 'chromStart']\n",
    "        H37Rv_end = H37Rv_annotation_no_RNA.loc[gene_id , 'chromEnd']\n",
    "        RefGenome_start = corresponding_gene_ref_genome.start\n",
    "        RefGenome_end = corresponding_gene_ref_genome.end\n",
    "\n",
    "        Highest_window_score = list(gene_length_match_scores)[0]\n",
    "        second_Highest_window_score = list(gene_length_match_scores)[1]\n",
    "\n",
    "        #start counting from 0!\n",
    "        H37Rv_RefGeome_gene_mapping.loc[gene_id , :] = [symbol , H37Rv_start , H37Rv_end , RefGenome_start - 1 , RefGenome_end , strand , Highest_window_score , second_Highest_window_score , window_match_index]\n",
    "        \n",
    "    else: #if window match score negative, then likely no gene/CDS analogue was found\n",
    "        \n",
    "        H37Rv_RefGeome_gene_mapping.loc[gene_id , :] = [np.nan]*9\n",
    "    \n",
    "##########################################################################################\n",
    "##########################################################################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "##########################################################################################\n",
    "##########################################################################################\n",
    "'''\n",
    "This piece of code stores all of the Mappings in a CSV file.\n",
    "'''\n",
    "\n",
    "#calculate the differences between the highest score and the second highest score\n",
    "H37Rv_RefGeome_gene_mapping['Score_Diff'] =  H37Rv_RefGeome_gene_mapping.Highest_window_score - H37Rv_RefGeome_gene_mapping.second_Highest_window_score\n",
    "\n",
    "#save output file\n",
    "directory_for_tables = '/n/data1/hms/dbmi/farhat/Roger/inhost_TB_dynamics_project/simulated_reads/reference_sequences/strains_typed/genome_plasticity_window_analysis_csv_files_10_gene_windows_rough_match/'\n",
    "H37Rv_RefGeome_gene_mapping.to_csv(directory_for_tables + ref_genome_strain + '_H37Rv_mapping_window_analysis.csv' , sep = ',')\n",
    "\n",
    "##########################################################################################\n",
    "##########################################################################################"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "################################################################################################################################################################################################################################################"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## [2] Submit the above as a job for each seperate Reference Genome"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "################################################################################################################################################################################################################################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "from slurmpy import Slurm\n",
    "import os\n",
    "import pandas as pd\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 2968686\n",
      "submitted: Submitted batch job 2968688\n",
      "submitted: Submitted batch job 2968689\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0 : 2968686\n",
      "1 : 2968688\n",
      "2 : 2968689\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 2968690\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3 : 2968690\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 2968693\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4 : 2968693\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 2968694\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "5 : 2968694\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 2968696\n",
      "submitted: Submitted batch job 2968697\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "6 : 2968696\n",
      "7 : 2968697\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 2968699\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "8 : 2968699\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 2968700\n",
      "submitted: Submitted batch job 2968701\n",
      "submitted: Submitted batch job 2968702\n",
      "submitted: Submitted batch job 2968703\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "9 : 2968700\n",
      "10 : 2968701\n",
      "11 : 2968702\n",
      "12 : 2968703\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 2968704\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "13 : 2968704\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 2968706\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "14 : 2968706\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 2968707\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "15 : 2968707\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 2968708\n",
      "submitted: Submitted batch job 2968710\n",
      "submitted: Submitted batch job 2968711\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "16 : 2968708\n",
      "17 : 2968710\n",
      "18 : 2968711\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 2968712\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "19 : 2968712\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 2968714\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "20 : 2968714\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 2968715\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "21 : 2968715\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 2968717\n",
      "submitted: Submitted batch job 2968718\n",
      "submitted: Submitted batch job 2968719\n",
      "submitted: Submitted batch job 2968720\n",
      "submitted: Submitted batch job 2968721\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "22 : 2968717\n",
      "23 : 2968718\n",
      "24 : 2968719\n",
      "25 : 2968720\n",
      "26 : 2968721\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 2968724\n",
      "submitted: Submitted batch job 2968725\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "27 : 2968724\n",
      "28 : 2968725\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 2968726\n",
      "submitted: Submitted batch job 2968727\n",
      "submitted: Submitted batch job 2968728\n",
      "submitted: Submitted batch job 2968729\n",
      "submitted: Submitted batch job 2968730\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "29 : 2968726\n",
      "30 : 2968727\n",
      "31 : 2968728\n",
      "32 : 2968729\n",
      "33 : 2968730\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 2968731\n",
      "submitted: Submitted batch job 2968733\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "34 : 2968731\n",
      "35 : 2968733\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 2968734\n",
      "submitted: Submitted batch job 2968735\n",
      "submitted: Submitted batch job 2968736\n",
      "submitted: Submitted batch job 2968737\n",
      "submitted: Submitted batch job 2968738\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "36 : 2968734\n",
      "37 : 2968735\n",
      "38 : 2968736\n",
      "39 : 2968737\n",
      "40 : 2968738\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 2968739\n",
      "submitted: Submitted batch job 2968740\n",
      "submitted: Submitted batch job 2968742\n",
      "submitted: Submitted batch job 2968743\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "41 : 2968739\n",
      "42 : 2968740\n",
      "43 : 2968742\n",
      "44 : 2968743\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 2968744\n",
      "submitted: Submitted batch job 2968745\n",
      "submitted: Submitted batch job 2968746\n",
      "submitted: Submitted batch job 2968747\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "45 : 2968744\n",
      "46 : 2968745\n",
      "47 : 2968746\n",
      "48 : 2968747\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 2968749\n",
      "submitted: Submitted batch job 2968750\n",
      "submitted: Submitted batch job 2968751\n",
      "submitted: Submitted batch job 2968752\n",
      "submitted: Submitted batch job 2968753\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "49 : 2968749\n",
      "50 : 2968750\n",
      "51 : 2968751\n",
      "52 : 2968752\n",
      "53 : 2968753\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 2968754\n",
      "submitted: Submitted batch job 2968755\n",
      "submitted: Submitted batch job 2968756\n",
      "submitted: Submitted batch job 2968757\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "54 : 2968754\n",
      "55 : 2968755\n",
      "56 : 2968756\n",
      "57 : 2968757\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 2968758\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "58 : 2968758\n",
      "59 : 2968760\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 2968760\n"
     ]
    }
   ],
   "source": [
    "#Choose a specific reference genome, pull sequence & annotation; compare to H37Rv (indexed: 0 - 59)\n",
    "for ref_genome_i in range(0 , 60):\n",
    "\n",
    "    ref_genome_i = str(ref_genome_i)\n",
    "\n",
    "    #command to run script\n",
    "    run_script = \"python /n/data1/hms/dbmi/farhat/Roger/inhost_TB_dynamics_project/python_scripts/Mtb_RefGenome_H37Rv_gene_mapping_and_window_analysis_rough_match.py \" + ref_genome_i\n",
    "\n",
    "    #submit job to O2\n",
    "    #directory where you want output + error files\n",
    "    os.chdir(\"/n/data1/hms/dbmi/farhat/Roger/inhost_TB_dynamics_project/simulated_reads/O2_jobs_output/Mtb_RefGenome_H37Rv_mapping_and_window_analysis\")\n",
    "\n",
    "    #submit job to O2\n",
    "    job_name = ref_genome_i\n",
    "\n",
    "    s = Slurm(job_name , {'partition':'short' , 'n':'1' , 't':'0-12:00:00' , 'mem-per-cpu':'8G' , 'mail-type':'FAIL' , 'mail-user':'roger_vargas@g.harvard.edu'})\n",
    "\n",
    "    #submits the job\n",
    "    job_id = s.run(run_script)\n",
    "\n",
    "    print job_name  + ' : ' +  str(job_id)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "################################################################################################################################################################################################################################################"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "## [3] Analyze the output (CSV files) from the jobs above"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "################################################################################################################################################################################################################################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "#import necessary packages\n",
    "import vcf\n",
    "import os\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "from itertools import compress\n",
    "import ast\n",
    "import itertools\n",
    "\n",
    "import time\n",
    "import sys\n",
    "\n",
    "import Bio\n",
    "from Bio.Alphabet import IUPAC\n",
    "from Bio.Blast.Applications import NcbiblastnCommandline\n",
    "from Bio.Blast import NCBIXML\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.SeqFeature import SeqFeature, FeatureLocation\n",
    "from Bio import pairwise2\n",
    "from Bio import SeqIO\n",
    "from Bio.Graphics import GenomeDiagram\n",
    "from Bio.SeqUtils import GC\n",
    "\n",
    "from Bio.Align.Applications import MuscleCommandline\n",
    "from StringIO import StringIO\n",
    "from Bio import AlignIO\n",
    "from Bio.Align import AlignInfo\n",
    "from Bio.Seq import MutableSeq\n",
    "from Bio import pairwise2\n",
    "from shutil import copy\n",
    "import subprocess\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import fastcluster\n",
    "from sklearn import cluster, datasets\n",
    "import scipy.cluster.hierarchy as hier\n",
    "import matplotlib as mpl\n",
    "\n",
    "#for exporting to Adobe Illustrator\n",
    "mpl.rcParams['pdf.fonttype'] = 42\n",
    "mpl.rcParams['ps.fonttype'] = 42"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Retrieve a Table of the Reference Genomes that had Corresponding Feature Annotation Tables and were Mapped to H37Rv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# check to see if the reference genomes have a corresponding feature annotation table\n",
    "\n",
    "#annotation for reference sequences\n",
    "ref_seqs_annot = pd.read_csv('/n/data1/hms/dbmi/farhat/Roger/inhost_TB_dynamics_project/simulated_reads/reference_sequences/strains_typed/final_table.tsv' ,  sep = '\\t')\n",
    "\n",
    "#directory where annotation files are stored\n",
    "ref_strain_genome_annot_files_directory = '/n/data1/hms/dbmi/farhat/Roger/inhost_TB_dynamics_project/simulated_reads/reference_sequences/strains_typed/reference_sequence_annotation_tables/'\n",
    "reference_strain_feature_counts_tables = os.listdir(ref_strain_genome_annot_files_directory)\n",
    "\n",
    "reference_genome_strain_has_feature_table = []\n",
    "reference_genome_strain_feature_table_names = []\n",
    "\n",
    "for reference_genome_strain in ref_seqs_annot.Strain:\n",
    "\n",
    "    #split the reference id to get back relevant info to find the filename\n",
    "    genbank , reference_id_1 , reference_id_2 , genomic_word = reference_genome_strain.split('_')\n",
    "    folder_1 = reference_id_1[0:3]\n",
    "    folder_2 = reference_id_1[3:6]\n",
    "    folder_3 = reference_id_1[6:9]\n",
    "    folder_4 = 'GCF_' + reference_id_1 + '_' + reference_id_2\n",
    "\n",
    "    reference_strain_annotation_filename = folder_4 + '_feature_table.txt.gz'\n",
    "\n",
    "    if reference_strain_annotation_filename in reference_strain_feature_counts_tables:\n",
    "        reference_genome_strain_has_feature_table.append('YES')\n",
    "        reference_genome_strain_feature_table_names.append(reference_strain_annotation_filename)\n",
    "\n",
    "    else:\n",
    "        reference_genome_strain_has_feature_table.append('NO')\n",
    "        reference_genome_strain_feature_table_names.append('No Table')\n",
    "\n",
    "ref_seqs_annot.loc[: , 'Has_Feature_Table'] = reference_genome_strain_has_feature_table\n",
    "ref_seqs_annot.loc[: , 'Feature_Table_filename'] = reference_genome_strain_feature_table_names\n",
    "\n",
    "#drop strains that don't have a corresponding feature table\n",
    "ref_seqs_annot = ref_seqs_annot[ref_seqs_annot.Has_Feature_Table == 'YES']\n",
    "ref_seqs_annot.reset_index(drop = True , inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Strain</th>\n",
       "      <th>Decision</th>\n",
       "      <th>Evaluation</th>\n",
       "      <th>Strain name</th>\n",
       "      <th>Has_Feature_Table</th>\n",
       "      <th>Feature_Table_filename</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GCA_000008585.1_ASM858v1_genomic</td>\n",
       "      <td>lineage4,lineage4.1,lineage4.1.1,lineage4.1.1.3</td>\n",
       "      <td>4</td>\n",
       "      <td>CDC1551</td>\n",
       "      <td>YES</td>\n",
       "      <td>GCF_000008585.1_ASM858v1_feature_table.txt.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GCA_000016145.1_ASM1614v1_genomic</td>\n",
       "      <td>lineage4,lineage4.9</td>\n",
       "      <td>4</td>\n",
       "      <td>H37Ra</td>\n",
       "      <td>YES</td>\n",
       "      <td>GCF_000016145.1_ASM1614v1_feature_table.txt.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GCA_000016925.1_ASM1692v1_genomic</td>\n",
       "      <td>lineage4,lineage4.3,lineage4.3.2,lineage4.3.2.1</td>\n",
       "      <td>4</td>\n",
       "      <td>F11</td>\n",
       "      <td>YES</td>\n",
       "      <td>GCF_000016925.1_ASM1692v1_feature_table.txt.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GCA_000023625.1_ASM2362v1_genomic</td>\n",
       "      <td>lineage4,lineage4.3,lineage4.3.3</td>\n",
       "      <td>4</td>\n",
       "      <td>KZN 1435</td>\n",
       "      <td>YES</td>\n",
       "      <td>GCF_000023625.1_ASM2362v1_feature_table.txt.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GCA_000153685.2_ASM15368v2_genomic</td>\n",
       "      <td>lineage4,lineage4.1,lineage4.1.2,lineage4.1.2.1</td>\n",
       "      <td>4</td>\n",
       "      <td>str. Haarlem</td>\n",
       "      <td>YES</td>\n",
       "      <td>GCF_000153685.2_ASM15368v2_feature_table.txt.gz</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                               Strain  \\\n",
       "0    GCA_000008585.1_ASM858v1_genomic   \n",
       "1   GCA_000016145.1_ASM1614v1_genomic   \n",
       "2   GCA_000016925.1_ASM1692v1_genomic   \n",
       "3   GCA_000023625.1_ASM2362v1_genomic   \n",
       "4  GCA_000153685.2_ASM15368v2_genomic   \n",
       "\n",
       "                                          Decision Evaluation   Strain name  \\\n",
       "0  lineage4,lineage4.1,lineage4.1.1,lineage4.1.1.3          4       CDC1551   \n",
       "1                              lineage4,lineage4.9          4         H37Ra   \n",
       "2  lineage4,lineage4.3,lineage4.3.2,lineage4.3.2.1          4           F11   \n",
       "3                 lineage4,lineage4.3,lineage4.3.3          4      KZN 1435   \n",
       "4  lineage4,lineage4.1,lineage4.1.2,lineage4.1.2.1          4  str. Haarlem   \n",
       "\n",
       "  Has_Feature_Table                           Feature_Table_filename  \n",
       "0               YES    GCF_000008585.1_ASM858v1_feature_table.txt.gz  \n",
       "1               YES   GCF_000016145.1_ASM1614v1_feature_table.txt.gz  \n",
       "2               YES   GCF_000016925.1_ASM1692v1_feature_table.txt.gz  \n",
       "3               YES   GCF_000023625.1_ASM2362v1_feature_table.txt.gz  \n",
       "4               YES  GCF_000153685.2_ASM15368v2_feature_table.txt.gz  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ref_seqs_annot.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Store the Mapping DF for each RefGenome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "#directory where CSV files are stored (1 for each RefGenome)\n",
    "directory_for_tables = '/n/data1/hms/dbmi/farhat/Roger/inhost_TB_dynamics_project/simulated_reads/reference_sequences/strains_typed/genome_plasticity_window_analysis_csv_files_10_gene_windows_rough_match/'\n",
    "os.chdir(directory_for_tables)\n",
    "\n",
    "#get list of all CSV file names\n",
    "RefGenome_map_filenames = os.listdir(directory_for_tables)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "RefGenome_Mappings_DF_dict = {}\n",
    "\n",
    "for RefGenome_Mapping in RefGenome_map_filenames:\n",
    "    \n",
    "    RefGenome_ID = RefGenome_Mapping.split('_genomic_')[0]\n",
    "    \n",
    "    RefGenome_Mapping_sliding_window_info = pd.read_csv(directory_for_tables + RefGenome_Mapping).set_index('Unnamed: 0')\n",
    "\n",
    "    RefGenome_Mappings_DF_dict[RefGenome_ID] = RefGenome_Mapping_sliding_window_info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>symbol</th>\n",
       "      <th>H37Rv_start</th>\n",
       "      <th>H37Rv_end</th>\n",
       "      <th>RefGenome_start</th>\n",
       "      <th>RefGenome_end</th>\n",
       "      <th>strand</th>\n",
       "      <th>Highest_window_score</th>\n",
       "      <th>second_Highest_window_score</th>\n",
       "      <th>window_index</th>\n",
       "      <th>Score_Diff</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Rv0001</th>\n",
       "      <td>dnaA</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1524.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1524.0</td>\n",
       "      <td>+</td>\n",
       "      <td>6.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rv0002</th>\n",
       "      <td>dnaN</td>\n",
       "      <td>2051.0</td>\n",
       "      <td>3260.0</td>\n",
       "      <td>2051.0</td>\n",
       "      <td>3260.0</td>\n",
       "      <td>+</td>\n",
       "      <td>7.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rv0003</th>\n",
       "      <td>recF</td>\n",
       "      <td>3279.0</td>\n",
       "      <td>4437.0</td>\n",
       "      <td>3279.0</td>\n",
       "      <td>4437.0</td>\n",
       "      <td>+</td>\n",
       "      <td>8.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rv0004</th>\n",
       "      <td>NaN</td>\n",
       "      <td>4433.0</td>\n",
       "      <td>4997.0</td>\n",
       "      <td>4433.0</td>\n",
       "      <td>4997.0</td>\n",
       "      <td>+</td>\n",
       "      <td>9.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rv0005</th>\n",
       "      <td>gyrB</td>\n",
       "      <td>5239.0</td>\n",
       "      <td>7267.0</td>\n",
       "      <td>5239.0</td>\n",
       "      <td>7267.0</td>\n",
       "      <td>+</td>\n",
       "      <td>10.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           symbol  H37Rv_start  H37Rv_end  RefGenome_start  RefGenome_end  \\\n",
       "Unnamed: 0                                                                  \n",
       "Rv0001       dnaA          0.0     1524.0              0.0         1524.0   \n",
       "Rv0002       dnaN       2051.0     3260.0           2051.0         3260.0   \n",
       "Rv0003       recF       3279.0     4437.0           3279.0         4437.0   \n",
       "Rv0004        NaN       4433.0     4997.0           4433.0         4997.0   \n",
       "Rv0005       gyrB       5239.0     7267.0           5239.0         7267.0   \n",
       "\n",
       "           strand  Highest_window_score  second_Highest_window_score  \\\n",
       "Unnamed: 0                                                             \n",
       "Rv0001          +                   6.0                          5.0   \n",
       "Rv0002          +                   7.0                          6.0   \n",
       "Rv0003          +                   8.0                          7.0   \n",
       "Rv0004          +                   9.0                          8.0   \n",
       "Rv0005          +                  10.0                          9.0   \n",
       "\n",
       "            window_index  Score_Diff  \n",
       "Unnamed: 0                            \n",
       "Rv0001               0.0         1.0  \n",
       "Rv0002               1.0         1.0  \n",
       "Rv0003               2.0         1.0  \n",
       "Rv0004               3.0         1.0  \n",
       "Rv0005               4.0         1.0  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "RefGenome_Mapping_sliding_window_info.head() #example of mapping for 1 of the Reference Genomes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.style.use('ggplot')\n",
    "plt.rcParams['lines.linewidth']=3\n",
    "plt.rcParams['axes.facecolor']='white'\n",
    "plt.rcParams['xtick.color']='black'\n",
    "plt.rcParams['axes.grid']=True\n",
    "plt.rcParams['axes.edgecolor']='black'\n",
    "plt.rcParams['grid.color']= '0.85'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Calculate the abolute difference between the Starting Positions for all of the Mapped Genes between each Reference Genome and H37Rv to asses the well the genes mapped to eachother."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Given the clonality of TB, the genes that end up being mapped between a given Reference Genome and H37Rv should end up being in the same neighborhood with respect to both genomes. We look to see how 'far', on average, genes mapped from H37Rv to the Reference Genome to give us an indication if there were any problematic Reference Genomes that were full of bad mappings."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "#create dictionary to store Average Absolute Difference in Gene Mapping Starting Positions\n",
    "ALL_RefGenome_Avg_StartPos_Diffs = {} #keys: Reference Genome , #values: absolute difference in starting values for gene mappings\n",
    "\n",
    "for RefGenome_ID in RefGenome_Mappings_DF_dict.keys():\n",
    "    \n",
    "    #for a specific Reference Genome, get the absolute difference for all of the gene-gene mappings between the Reference Genome & H37Rv\n",
    "    RefGenome_H37Rv_Mapping_StartPos_Diffs = abs( RefGenome_Mappings_DF_dict[RefGenome_ID].H37Rv_start.values - RefGenome_Mappings_DF_dict[RefGenome_ID].RefGenome_start.values )\n",
    "    \n",
    "    #store the average of these differences (ignoring Nan values which are placeholders for unnsuccessfully mapped genes)\n",
    "    ALL_RefGenome_Avg_StartPos_Diffs[RefGenome_ID] = np.nanmean( RefGenome_H37Rv_Mapping_StartPos_Diffs )\n",
    "\n",
    "ALL_RefGenome_Avg_StartPos_Diffs = pd.Series(ALL_RefGenome_Avg_StartPos_Diffs , name = 'RefGenome')  \n",
    "ALL_RefGenome_Avg_StartPos_Diffs.sort_values(ascending = True , inplace = True)\n",
    "\n",
    "#convert to DataFrame\n",
    "ALL_RefGenome_Avg_StartPos_Diffs_DF = pd.DataFrame()\n",
    "ALL_RefGenome_Avg_StartPos_Diffs_DF['avg_StartPos_diff'] = ALL_RefGenome_Avg_StartPos_Diffs\n",
    "ALL_RefGenome_Avg_StartPos_Diffs_DF['index_num'] = range(1 , len(ALL_RefGenome_Avg_StartPos_Diffs)+1 )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>avg_StartPos_diff</th>\n",
       "      <th>index_num</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GCA_000195955.2_ASM19595v2</th>\n",
       "      <td>3.222193e+00</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_001922485.1_ASM192248v1</th>\n",
       "      <td>1.590713e+01</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000738475.1_ASM73847v1</th>\n",
       "      <td>1.799506e+01</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000738445.1_ASM73844v1</th>\n",
       "      <td>2.142289e+01</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000277735.2_ASM27773v2</th>\n",
       "      <td>3.375160e+01</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000364825.1_ASM36482v1</th>\n",
       "      <td>2.034171e+02</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000827085.1_ASM82708v1</th>\n",
       "      <td>3.409074e+02</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_001545015.1_ASM154501v1</th>\n",
       "      <td>3.639588e+03</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000400615.1_ASM40061v1</th>\n",
       "      <td>4.103268e+03</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000786505.1_MT49-02</th>\n",
       "      <td>4.196670e+03</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_001702435.1_ASM170243v1</th>\n",
       "      <td>4.324937e+03</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000224435.1_ASM22443v1</th>\n",
       "      <td>4.436756e+03</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000828995.1_ASM82899v1</th>\n",
       "      <td>4.555685e+03</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000756545.1_ASM75654v1</th>\n",
       "      <td>4.988120e+03</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_002116815.1_ASM211681v1</th>\n",
       "      <td>4.995460e+03</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000572195.1_ASM57219v1</th>\n",
       "      <td>5.012973e+03</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_001544705.1_ASM154470v1</th>\n",
       "      <td>5.044254e+03</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_001275565.2_ASM127556v2</th>\n",
       "      <td>5.069076e+03</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000831245.1_ASM83124v1</th>\n",
       "      <td>5.370744e+03</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_001544955.1_ASM154495v1</th>\n",
       "      <td>5.601121e+03</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000572125.1_ASM57212v1</th>\n",
       "      <td>5.806339e+03</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_001708265.1_ASM170826v1</th>\n",
       "      <td>6.156259e+03</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000954155.1_ASM95415v1</th>\n",
       "      <td>6.198484e+03</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000016145.1_ASM1614v1</th>\n",
       "      <td>6.344706e+03</td>\n",
       "      <td>24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_001544985.1_ASM154498v1</th>\n",
       "      <td>6.757254e+03</td>\n",
       "      <td>25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000008585.1_ASM858v1</th>\n",
       "      <td>6.877468e+03</td>\n",
       "      <td>26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_002072775.2_ASM207277v2</th>\n",
       "      <td>7.413662e+03</td>\n",
       "      <td>27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_002116795.1_ASM211679v1</th>\n",
       "      <td>7.439999e+03</td>\n",
       "      <td>28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000572175.1_ASM57217v1</th>\n",
       "      <td>7.470953e+03</td>\n",
       "      <td>29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_001938725.1_ASM193872v1</th>\n",
       "      <td>7.683533e+03</td>\n",
       "      <td>30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000422125.1_ASM42212v1</th>\n",
       "      <td>7.788020e+03</td>\n",
       "      <td>31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000389945.1_ASM38994v1</th>\n",
       "      <td>8.164887e+03</td>\n",
       "      <td>32</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000270365.1_ASM27036v1</th>\n",
       "      <td>8.203806e+03</td>\n",
       "      <td>33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000153685.2_ASM15368v2</th>\n",
       "      <td>8.291415e+03</td>\n",
       "      <td>34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000706665.1_ASM70666v1</th>\n",
       "      <td>8.373933e+03</td>\n",
       "      <td>35</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000331445.1_ASM33144v1</th>\n",
       "      <td>8.462492e+03</td>\n",
       "      <td>36</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000016925.1_ASM1692v1</th>\n",
       "      <td>8.588649e+03</td>\n",
       "      <td>37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_002116855.1_ASM211685v1</th>\n",
       "      <td>8.608781e+03</td>\n",
       "      <td>38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_002116755.1_ASM211675v1</th>\n",
       "      <td>9.198138e+03</td>\n",
       "      <td>39</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_001895865.1_ASM189586v1</th>\n",
       "      <td>9.795429e+03</td>\n",
       "      <td>40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_001855255.1_ASM185525v1</th>\n",
       "      <td>1.043145e+04</td>\n",
       "      <td>41</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000572155.1_ASM57215v1</th>\n",
       "      <td>1.057631e+04</td>\n",
       "      <td>42</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000270345.1_ASM27034v1</th>\n",
       "      <td>1.117395e+04</td>\n",
       "      <td>43</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_001895765.1_ASM189576v1</th>\n",
       "      <td>1.150069e+04</td>\n",
       "      <td>44</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_001895845.1_ASM189584v1</th>\n",
       "      <td>1.223566e+04</td>\n",
       "      <td>45</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_001545055.1_ASM154505v1</th>\n",
       "      <td>1.234128e+04</td>\n",
       "      <td>46</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000350205.1_ASM35020v1</th>\n",
       "      <td>1.307971e+04</td>\n",
       "      <td>47</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_001895805.1_ASM189580v1</th>\n",
       "      <td>1.342812e+04</td>\n",
       "      <td>48</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000756525.1_ASM75652v1</th>\n",
       "      <td>1.408453e+04</td>\n",
       "      <td>49</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_001895785.1_ASM189578v1</th>\n",
       "      <td>1.575030e+04</td>\n",
       "      <td>50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_002116775.1_ASM211677v1</th>\n",
       "      <td>1.579492e+04</td>\n",
       "      <td>51</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_001895825.1_ASM189582v1</th>\n",
       "      <td>1.598983e+04</td>\n",
       "      <td>52</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000698475.1_ASM69847v1</th>\n",
       "      <td>1.637949e+04</td>\n",
       "      <td>53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_002116835.1_ASM211683v1</th>\n",
       "      <td>2.718225e+04</td>\n",
       "      <td>54</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000193185.2_ASM19318v2</th>\n",
       "      <td>4.700708e+05</td>\n",
       "      <td>55</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000023625.1_ASM2362v1</th>\n",
       "      <td>5.520132e+05</td>\n",
       "      <td>56</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000154605.2_ASM15460v2</th>\n",
       "      <td>5.550821e+05</td>\n",
       "      <td>57</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000154585.2_ASM15458v2</th>\n",
       "      <td>5.552846e+05</td>\n",
       "      <td>58</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_001750865.1_ASM175086v1</th>\n",
       "      <td>1.413905e+06</td>\n",
       "      <td>59</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_001870145.1_ASM187014v1</th>\n",
       "      <td>1.427028e+06</td>\n",
       "      <td>60</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                             avg_StartPos_diff  index_num\n",
       "GCA_000195955.2_ASM19595v2        3.222193e+00          1\n",
       "GCA_001922485.1_ASM192248v1       1.590713e+01          2\n",
       "GCA_000738475.1_ASM73847v1        1.799506e+01          3\n",
       "GCA_000738445.1_ASM73844v1        2.142289e+01          4\n",
       "GCA_000277735.2_ASM27773v2        3.375160e+01          5\n",
       "GCA_000364825.1_ASM36482v1        2.034171e+02          6\n",
       "GCA_000827085.1_ASM82708v1        3.409074e+02          7\n",
       "GCA_001545015.1_ASM154501v1       3.639588e+03          8\n",
       "GCA_000400615.1_ASM40061v1        4.103268e+03          9\n",
       "GCA_000786505.1_MT49-02           4.196670e+03         10\n",
       "GCA_001702435.1_ASM170243v1       4.324937e+03         11\n",
       "GCA_000224435.1_ASM22443v1        4.436756e+03         12\n",
       "GCA_000828995.1_ASM82899v1        4.555685e+03         13\n",
       "GCA_000756545.1_ASM75654v1        4.988120e+03         14\n",
       "GCA_002116815.1_ASM211681v1       4.995460e+03         15\n",
       "GCA_000572195.1_ASM57219v1        5.012973e+03         16\n",
       "GCA_001544705.1_ASM154470v1       5.044254e+03         17\n",
       "GCA_001275565.2_ASM127556v2       5.069076e+03         18\n",
       "GCA_000831245.1_ASM83124v1        5.370744e+03         19\n",
       "GCA_001544955.1_ASM154495v1       5.601121e+03         20\n",
       "GCA_000572125.1_ASM57212v1        5.806339e+03         21\n",
       "GCA_001708265.1_ASM170826v1       6.156259e+03         22\n",
       "GCA_000954155.1_ASM95415v1        6.198484e+03         23\n",
       "GCA_000016145.1_ASM1614v1         6.344706e+03         24\n",
       "GCA_001544985.1_ASM154498v1       6.757254e+03         25\n",
       "GCA_000008585.1_ASM858v1          6.877468e+03         26\n",
       "GCA_002072775.2_ASM207277v2       7.413662e+03         27\n",
       "GCA_002116795.1_ASM211679v1       7.439999e+03         28\n",
       "GCA_000572175.1_ASM57217v1        7.470953e+03         29\n",
       "GCA_001938725.1_ASM193872v1       7.683533e+03         30\n",
       "GCA_000422125.1_ASM42212v1        7.788020e+03         31\n",
       "GCA_000389945.1_ASM38994v1        8.164887e+03         32\n",
       "GCA_000270365.1_ASM27036v1        8.203806e+03         33\n",
       "GCA_000153685.2_ASM15368v2        8.291415e+03         34\n",
       "GCA_000706665.1_ASM70666v1        8.373933e+03         35\n",
       "GCA_000331445.1_ASM33144v1        8.462492e+03         36\n",
       "GCA_000016925.1_ASM1692v1         8.588649e+03         37\n",
       "GCA_002116855.1_ASM211685v1       8.608781e+03         38\n",
       "GCA_002116755.1_ASM211675v1       9.198138e+03         39\n",
       "GCA_001895865.1_ASM189586v1       9.795429e+03         40\n",
       "GCA_001855255.1_ASM185525v1       1.043145e+04         41\n",
       "GCA_000572155.1_ASM57215v1        1.057631e+04         42\n",
       "GCA_000270345.1_ASM27034v1        1.117395e+04         43\n",
       "GCA_001895765.1_ASM189576v1       1.150069e+04         44\n",
       "GCA_001895845.1_ASM189584v1       1.223566e+04         45\n",
       "GCA_001545055.1_ASM154505v1       1.234128e+04         46\n",
       "GCA_000350205.1_ASM35020v1        1.307971e+04         47\n",
       "GCA_001895805.1_ASM189580v1       1.342812e+04         48\n",
       "GCA_000756525.1_ASM75652v1        1.408453e+04         49\n",
       "GCA_001895785.1_ASM189578v1       1.575030e+04         50\n",
       "GCA_002116775.1_ASM211677v1       1.579492e+04         51\n",
       "GCA_001895825.1_ASM189582v1       1.598983e+04         52\n",
       "GCA_000698475.1_ASM69847v1        1.637949e+04         53\n",
       "GCA_002116835.1_ASM211683v1       2.718225e+04         54\n",
       "GCA_000193185.2_ASM19318v2        4.700708e+05         55\n",
       "GCA_000023625.1_ASM2362v1         5.520132e+05         56\n",
       "GCA_000154605.2_ASM15460v2        5.550821e+05         57\n",
       "GCA_000154585.2_ASM15458v2        5.552846e+05         58\n",
       "GCA_001750865.1_ASM175086v1       1.413905e+06         59\n",
       "GCA_001870145.1_ASM187014v1       1.427028e+06         60"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ALL_RefGenome_Avg_StartPos_Diffs_DF.head(n=60)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAsgAAAHMCAYAAADI0l79AAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMi4zLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvIxREBQAAIABJREFUeJzs3XtY1GX+//HXiAIiaBoEaoSHysnV2nSSEsRSbNVqXc3wl7n1zSKtkFot17Y2s+1kp91EMw9thy3bVKrvt+xqw1NAFom6mdnYQUVUIMhKUJhJmN8fBEFy+AzMZ2bA5+O6uq7mMzczb7s39+Xd+75vi8vlcgkAAACAJKmDrwsAAAAA/AkBGQAAAKiDgAwAAADUQUAGAAAA6iAgAwAAAHUQkAEAAIA6CMgAAABAHQRkAAAAoA4CMgAAAFBHR18XYDaLxeLrEgAAAOAnjFwi3e4DsmTsH0ReXp5iYmK8UA3cxdz4J+bFfzE3/ol58V/Mjf/y9NwYXTilxQIAAACog4AMAAAA1EFABgAAAOogIAMAAAB1EJABAACAOgjIAAAAQB0EZAAAAKCOdnsOcm5urrZt2+brMgAAANDGtNsVZJvNphkzZvi6DAAAALQx7TYgAwAAAC1BQAYAAADqICADAADA7xQXF2vPnj0qLi72+ncTkAEAAOAVxcXF2rVrV5Oh12636w+JiTo3Olp3XH21zo2O1sQxY2S3271WJwEZAAAApqobeqfExTUaeu12uy6NjdXIjRt10OHQF2Vlync4lLBhgy6NjfVaSCYgAwAAwDS/Dr2fHz3aaOidl5KiP5eW6k8ul7r8/CxU0p9cLs0tLdU9s2Z5pWYCMgAAAExjNPQWFxfrg+xs3eJyNfg5t7hc2pyVpZKSEtNrJiADAADAFO6E3qKiIvUKCqoN0b8WKqlnYKAKCwtNq7cGARkAAACmcCf0RkZG6rDDoWONjC2TVOB0Kioqypxi6yAgAwAAwBTuhN6IiAiNjI/XcoulwbHLLRZdOmKEwsPDTau3BgEZAAAApnA39D62eLEWhoXpaYtFZT+PKZP0tMWix8PC9GhamlfqJiADAADANO6EXqvVqs05OcoaPVrRQUGyhoYqOihI2YmJ2pyTI6vV6pWaO3rlWzzg888/V3p6ulwul8aNG6dhw4b5uiQAAAA0oyb03jNrlv6WlaWegYEqcDp1WUKCNi9adFLotVqtejMjQyUlJdq+fbuGDBnilbaKutpEQHY6nXr77bf1l7/8RR07tomSAQAA8LO6obewsFBRUVHNht7w8HANGDDA6+FYaiMB+csvv1RgYKAWLlyowMBAJScn67TTTvN1WQAAAHBDeHi4TwKvu7wakN977z1t3rxZBw4cUFxcnG6//fba98rKyrR06VLt3LlTYWFhmjp1quLj4yVJP/zwg4qKivTwww9r586dWr16tW655RZvlg4AAIBThFc36XXv3l2TJk3SZZdddtJ7K1euVMeOHbVixQqlpqZqxYoVys/PlyR16dJFAwYMUMeOHTV48GAdOnTIm2UDAADgFOLVFeTY2FhJ0t69e/Xdd9/VPq+oqFBOTo6eeuopBQcHy2q1ymazKTMzU9ddd53OPvtsvfPOO3K5XNq/f78iIyOb/J7169dr/fr1ta/z8vKarc3pdBoaB+9jbvwT8+K/mBv/xLz4L+bGf/lqbvyiB7mgoEAdOnRQr169ap/FxMRo9+7dkqSwsDBddNFFeuCBB2SxWHTrrbc2+XmJiYlKTEyUJC1cuFAxMTHN1pCXl2doHLyPufFPzIv/Ym78E/Piv5gb/+WrufGLgFxRUaGQkJB6z0JCQlRRUVH7euzYsRo7dqy3SwMAAMApxi8CcnBwsMrLy+s9Ky8vV3BwcIs/Mzc3V9u2bWttaQAAADjFNBqQ33jjDbc/bNy4cercubPbP9ezZ09VVlaqoKBAPXv2lFS9pB4dHe32Z9Ww2Wyy2WyaOXNmiz8DAAAAp55GA/LkyZPd+iCLxaKvvvpK/fr1a3RMZWWlKisrVVVVpaqqKjmdTgUEBCg4OFixsbF6/fXXNXPmTO3fv19bt27VQw895FYNAAAAQGs12WJRWFioM844w9AHhYWFNTsmPT1da9eurX2dlZWlyZMnKykpSTfffLOeffZZJScnKzQ0VMnJya1aQQYAAABaotGAfMMNN7jVLjFt2jR17dq1yTFJSUlKSkpq8L3Q0FDNnTvX8Pc1hx5kAAAAtESjF4W88MILhlaFayxdutSvrg602WyaMWOGr8sAAABAG+PVm/QAAAAAf+cXx7yZgRYLAAAAtES7XUGmxQIAAAAt0W4DMgAAABpWXFysXbt2qbi42Kvj2goCMgAAwCnCbrfrD4mJOjc6WlPi4nRudLQmjhkju91u6ri2xnBA3rdvnzZt2qR3331XW7duVUVFhZl1AQAAwIPsdrsujY3VyI0bddDh0OdHjyrf4VDChg26NDa2NtR6elxb1OQmvf3792vp0qV67bXXdOjQIblcrtr3AgMDNWLECN1yyy26+uqr1aGDfy1Gs0kPAADgF/NSUvTn0lL9qU6eC5X0J5dLrtJS3TNrlt7MyPD4uLao0VR7xx136IILLtDevXv18MMPa/fu3frxxx/ldDpVWFiod999V/Hx8frrX/+q888/X1u3bvVm3c1ikx4AAEC14uJifZCdrVvqhNm6bnG5tDkrS3a73aPjSkpKPPZr8KZGV5ADAwP1zTffNHj5xxlnnKFRo0Zp1KhRmj9/vt59913l5eXpoosuMrVYAAAANKy4uFhFRUWKjIxUREREvfeKiorUKyhIXRyOBn82VFLPwEB98cUXHh1XWFjoVxfJGdXoCvITTzxh+Bc0fvx4TZ482WNFAQAAwBgjG+UiIyN12OHQsUY+o0xSgdOp8847z6PjoqKiWvzr8iVDjcNVVVWqqqqqfV1YWKiVK1fqww8/NK0wAAAANM3oRrmIiAiNjI/Xcoulwc9ZbrHo0hEjZLVaPTquLa4eSwZv0rviiis0duxY3XHHHSorK5PNZtOxY8dUVlam559/Xtdff73ZdbqNTXoAAKC9c2ej3GOLF+vS2Fi5Skt1i8ulUFWv9C63WPR4WJg2p6WZMq4tMrSCvG3bNo0aNUqS9MYbb6hr16769ttvtWLFCj355JOmFthSbNIDAADtmdGNdzUb5axWqzbn5Chr9GhFBwVpYFiYooOClJ2YqM05ObJaraaMa4sMrSCXlpbqtNNOkyS9//77mjhxojp16qRRo0bp9ttvN7VAAAAAnMzoxru6G+WsVqvezMhQSUmJCgsLFRUV1WAbhKfHtTWGAvJZZ52lDz/8UFdddZX+85//aM2aNZKkI0eOKCQkxNQCAQAAcLK6G++6NPB+UxvlwsPDDQVZT49rKwy1WMyePVt//OMfdeaZZ6p3795KSEiQJGVmZmrw4MGmFggAAICTGd14156Cq7cYWkGeMWOGhg4dqvz8fI0ZM6b21rz+/fvrb3/7m6kFAgAAoGHteaOcLxkKyFL1pjebzVbv2RVXXOHxgjyFUywAAEB7V7NR7p5Zs/S3rCz1DAxUgdOpyxIStHnRoja9Uc6XDLVYSNK6deuUkJCg8PDw6iX9kSP17rvvmllbq3CKBQAAOBXUbJT76uBBrd6yRV8dPKg33n+fcNwKhgLyypUrNXHiRPXv318LFy7UY489pr59+2rixIn65z//aXaNAAAAaEZ4eLgGDRpEz7EHGGqxWLhwoZ5++mmlpKTUPrvppps0dOhQPfbYY5o+fbppBQIAAADeZGgF+cCBAxo7duxJz8eNG6e8vDyPFwUAAAD4iqGAfNZZZynj52sK63r//fcVExPj8aIAAAAAXzHUYnHXXXdp1qxZ2r59u4YPHy6LxaLs7Gz961//UhrHhwAAAKAdMXwO8hlnnKGnnnpKb7zxhiTpvPPO0+rVqzVhwgRTCwQAAAC8yfA5yBMnTtTEiRPNrMWjOAcZAAAALdFsD/KaNWt03XXXKSkpScuXL/dGTR7BOcgAAABoiSZXkJcvX66ZM2fqnHPOUXBwsNLT07Vv3z49+uij3qoPAAAA8KomV5DT0tJ07733as+ePfr000/1/PPPa/Hixd6qDQAAAPC6JgPy3r17deONN9a+/uMf/yiHw6HCwkLTCwMAAAB8ocmAXF5ertDQ0NrXAQEBCgoK0vHjx00vDAAAAPCFZk+xeO655+qF5BMnTuj555/X6aefXvts9uzZ5lQHAAAAeFmTAfmss87SCy+8UO9ZVFSUVq1aVfvaYrEQkAEAANBuNBmQ9+/f76UyAAAAAP/Q7DnIkvTyyy/L4XCc9NzpdOrll1/2eFEAAACArxgKyDfeeKN+/PHHk56XlpbWO+UCAAAAaOsMBWSXyyWLxXLS8wMHDqhbt24eLwoAAADwlSZ7kAcPHiyLxSKLxaKRI0eqY8dfhldWViovL0/jx483vciWyM3N1bZt23xdBgAAANqYJgPy5MmTJUm7du3SFVdcUe+4t8DAQPXp00dXX321uRW2kM1mk81m08yZM31dCgAAANqQJgPy/PnzdeLECYWHh+sPf/iDevfu7a26AAAAAJ9otge5Y8eOmjNnjn766Sdv1AMAAAD4lKFNehdccIG+/vprs2sBAAAAfM5QQH7ggQc0Z84cvfXWW8rPz9eRI0fq/QUAAAC0F032INe44oorJEmTJk2qd9xbzfFvlZWV5lQHAAAAeJmhgLxp0yaz6wAAAAD8gqGAPHLkSLPrAAAAAPyCoYBc4/Dhwzpw4ICcTme95wkJCR4tCgAAAPAVQwH58OHDmjp1qjIzM2WxWE66epoeZAAAALQXhk6xuPPOOxUQEKDdu3crJCREWVlZWrNmjc477zy99957ZtcIAAAAeI2hFeQPPvhA69atk9VqlcViUUREhOLi4hQUFKS//vWvGjNmjNl1AgAAAF5hKCCXl5crPDxcktSjRw99++23OvfcczVw4EDt3LnT1AIl6dtvv9U999yj6OhoSdLs2bPVtWtX078XAAAApx5DAdlqtcput6tPnz767W9/q+eee07R0dFasmSJevfubXaNkqSBAwdqzpw5XvkuAAAAnLoMBeQ77rhDhYWFkqT7779fY8eO1WuvvaagoCC99NJLphZYY8+ePbr//vtltVp17bXX1tskCAAAAHiKoYB83XXX1f79kCFDtH//ftntdp111lm1rRdGvPfee9q8ebMOHDiguLg43X777bXvlZWVaenSpdq5c6fCwsI0depUxcfHS5K6d++uRYsWKSgoSMuWLVNOTo4uvvhiw98LAAAAGGXoFIsHH3xQx48fr30dEhKiIUOGqEuXLnrwwQcNf1n37t01adIkXXbZZSe9t3LlSnXs2FErVqxQamqqVqxYofz8fElSp06dFBwcLIvFomHDhikvL8/wdwIAAADuMBSQFyxYoLKyspOeHz9+XAsWLDD8ZbGxsRo2bJjCwsLqPa+oqFBOTo6mTJmi4OBgWa1W2Ww2ZWZm1n5PDbvdrqioKMPfCQAAALjDUIvFry8GqbFjxw716NGj1UUUFBSoQ4cO6tWrV+2zmJgY7d69W5K0e/durVmzRkFBQTrjjDM0ZcqUJj9v/fr1Wr9+fe1rIyvOTqeTlWk/xdz4J+bFfzE3/ol58V/Mjf/y1dw0GZDDwsJksVhksVjUr1+/k27Pq6io0MyZM1tdREVFhUJCQuo9CwkJUUVFhSTJZrPJZrMZ/rzExEQlJiZKkhYuXKiYmJhmfyYvL8/QOHgfc+OfmBf/xdz4J+bFfzE3/stXc9NkQF68eLFcLpemT5+uhx9+WN26dat9LzAwUH369NEll1zS6iKCg4NVXl5e71l5ebmCg4Nb/dkAAACAO5oMyDfccIMkqW/fvoqLi1PHjoY6MtzWs2dPVVZWqqCgQD179pRU/SeGmotBWiI3N1fbtm3zVIkAAAA4RRjapBcREaFvvvmm9nVGRoamTZumRx99VJWVlYa/rLKyUk6nU1VVVaqqqpLT6VRlZaWCg4MVGxur119/XRUVFbLb7dq6dasSEhLc/xX9zGazacaMGS3+eQAAAJyaDAXkm266STt27JAkHTx4UBMmTNCRI0e0ZMkS3XfffYa/LD09XdOmTdNbb72lrKwsTZs2Tenp6ZKkm2++WU6nU8nJyXrmmWeUnJzcqhVkAAAAoCUM9Ux88cUXGjJkiCRpzZo1io2N1bvvvqtNmzbpxhtv1KOPPmroy5KSkpSUlNTge6GhoZo7d67BsptHiwUAAABawtAKcmVlpQIDAyVJGzZs0Pjx4yVJ/fv3V1FRkXnVtQItFgAAAGgJQwF50KBBWrp0qbKysrRhwwaNHTtWknTo0CG3rpoGAAAA/J2hgLxw4UKtWLFCl156qa699loNHjxYkvR///d/GjZsmKkFAgAAAN5kqAc5ISFBxcXFOnr0qLp37177fMaMGSdd8OEv6EEGAABASxhaQZakgICAeuFYkvr06aMzzjjD40V5Aj3IAAAAaAnDARkAAAA4FRCQAQAAgDoIyAAAAEAdhjbptUVs0gMAAEBLGF5B/uyzz5SSkqJx48apoKBAkvTWW2/VXkHtb9ikBwAAgJYwFJDff/99XXTRRTp06JA2btyo8vJySdI333yjBQsWmFogAAAA4E2GAvJf//pXPf3003rzzTdrr5yWpEsvvVSffPKJacUBAAAA3mYoIH/++ecaP378Sc979OihI0eOeLwoAAAAwFcMbdLr3r27Dh06pD59+tR7vn37dp155plm1NVqbNIDAABASxhaQZ46daruvvtuHTx4UBaLRSdOnNAHH3ygu+66S9dff73ZNbYIm/QAAADQEoYC8kMPPaS+ffsqJiZGZWVlGjhwoEaNGqX4+Hjde++9ZtcIAAAAeI2hFotOnTrp1Vdf1YMPPqgdO3aoqqpKF154oc455xyz6wMAAAC8yq2LQvr376/+/fubVQsAAADgc4YD8ptvvqlNmzbp22+/VVVVVb33Vq9e7fHCAAAAAF8wFJDnzJmjtLQ0xcXFKTIyUgEBAWbX1WqcYgEAAICWMBSQX3rpJa1Zs0YTJkwwux6Psdlsstlsmjlzpq9LAQAAQBti6BSLkJAQWa1Ws2sBAAAAfM5QQJ43b54ef/xxnThxwux6AAAAAJ8y1GKRnJysd955R71799a5556rTp061Xt/48aNphQHAAAAeJuhgDxz5kxlZWVp7NixioyMlMViMbsuAAAAwCcMBeTVq1frjTfe0JgxY8yuBwAAAHUUFxerqKhIkZGRioiI8HU5pwRDPcjh4eHq3bu32bUAAADgZ3a7XYnjExXdL1px4+IU3S9aY64YI7vd7uvS2j1DAXnBggW6//77VVZWZnY9HpObm6tly5b5ugwAAAC32e12xcbHamPVRjlSHTp681E5Zjm0oXKDYuNjCckmM9Ri8cQTT2j//v2KjIzUWWedddImvZ07d5pSXGtwDjIAAGirUmanqNRWKtclrl8eBkmuS1wqValmzZmljHUZviuwnTMUkCdPnmx2HQAAAFB1z3F2VrZcqa4G33cNcSkrLUslJSUKDw/3cnWnBkMBef78+WbXAQAAAElFRUUKOi1IjkBHwwOCpMBugSosLCQgm8RQQK6xceNG7d69WxaLRb/5zW906aWXmlQWAADAqSkyMlKOHxySU1JgAwMckvNHp6Kiorxd2inDUEA+dOiQJk6cqG3btqlXr16SpMOHD8tms+nNN9+sfQYAAIDWiYiIUPyIeG3ctrF+D/LPLNstGpEwgtVjExk6xSI1NVUBAQH6+uuvlZ+fr/z8fH311VcKCAhQamqq2TUCAACcUhY/vVhhuWGyfGSRajotHJLlI4vCcsOU9lSaT+tr7wwF5IyMDC1ZskR9+/atfdavXz8tWrRIGRnsoAQAAPAkq9WqnOwcjQ4YraC0IIWtCFNQWpASOyYqJztHVqvV1yW2a271IP9ahw6G8jUAAADcZLValbEuQyUlJSosLFRUVBRtFV5iKOGOHj1aqampys/Pr3124MAB3XHHHRo9erRpxQEAAJzqwsPDNWjQIMKxFxkKyIsWLdLx48fVr18/xcTEqE+fPurfv7+OHz+uRYsWmV0jAAAA4DWGWiyio6O1fft2ZWRkyG63y+VyaeDAgUpMTDS7PgAAAMCr3OpBHjNmjMaMGWNWLR6Vm5urbdu2+boMAACABhUXF6uoqEiRkZGKiIjwdTmow3BAzsnJ0YYNG/Ttt9+qqqqq3nv+2GZhs9lks9k0c+ZMX5cCAAD8VHFxsfbs2aOQkJAmQ6o7Yba5sXa7XSmzU5SdlV19Y94PDo1IGKG0p9I4ncJPGOpBfvLJJ3XJJZfoxRdf1H//+1999tlntX/t2rXL7BoBAAA8ym63K3F8oqL7RevqG69WdL9ojblijOx2e6Pj4sbFNTrO6Fi73a7Y+FhtrNooR6pDR28+KscshzZUblBsfGyDnwvvs7hcrpOvaPmV6Oho/fnPf1ZKSoo3avIoi8UiA79E5eXlKSYmxgsVwV3MjX9iXvwXc+OfmBf/URNSS22lcg11VV/n7Ki+oS4sN6z2nGGj49z5zMTxidpY1cgNeR9ZNDpgtDLWccdEDU//e2M0FxpaQT569KjGjx/f6qIAAAC8obi4WLt27VJxcfFJ76XMTqkOspf8HGQlKUhyXeJSqa1Us+bMcmuc0bHFxcXKzsquDtANcA1xKSszSyUlJR76p4CWMhSQr732Wr333ntm1wIAANAqzbU5GA2pdrvdcJh15zODTgv6JUD/WpAU2C1QhYWFbv+64VmGj3mbP3++PvzwQ51//vnq1KlTvfdnz55tSnEAAAA1jGx+q21zSHXJEeiQHNKG7dX9vTnZOTpx4kT1xrhAR8Nf8nNI/eKLLwyNqwmzRsa6XC45fnBITjUckh2S80enoqKimv+HAVMZCsgrV65UaGiotmzZoi1bttR7z2KxEJABAECLeerUh3ptDjVq2hxU3eaw6sVVhkLqeeedZzjMGg2+AwcOVPyIeG3c1kgP8naLRiSM4MY8P2CoxWLfvn2N/rV3716zawQAAH6kqf5ed8Z58tQHo20OFotF8SPiZdlmaXBcTUi1Wq2GxoWHhysiIsLw2MVPL1ZYbpgsH1mkmgVnR/UGvbDcMKU9ldbgZ8C7DAXkuoqKik46BxkAALR/Ro8882TwNbpRrqioyHB/r9GQ6k6YNTrWarUqJztHowNGKygtSGErwhSUFqTEjon1TsWAbxkKyD/99JPmzp2rsLAw9e7dW/v375ck/fnPf9azzz5rZn0AAMALjKz2Ggm0ngy+7pz6EBkZ+UubQ0PqtET8OqSGLg9tMKS6E2bdHZuxLkMH9x3Ulve26OC+g3r/nfcJx37EUEBesGCB3n77bb3yyisKCgqqfT5s2DC9+OKLZtUGAABayRNtDpJnj0Yz49QHd9ocpPohNf2F9EZDqjth1t3gGx4erkGDBtFz7IcMBeTXXntNzz33nCZMmKAOHX75kUGDBunLL780rTgAANAyvujvNXo0mtHgW2/zW0N+depDS/p7w8PDNWDAgGZDqjthluDb9hkKyIcPH27wFpMTJ07oxIkTHi+qMdnZ2brpppu89n0AAHibJzbA+aq/t+ZoNE8F35pTH9xZFaa/F55gKCD/5je/UWZm5knPV69eraFDh3q8qIZUVVXp448/5k9jAIB2yZMb4HzV31vvaLQmxrkTfN1dFaa/F55g6Bzk+fPna9q0acrPz1dlZaXWrFkju92uVatWad26dWbXKKl69fjiiy/WO++845XvAwDAU4qLi7Vnzx6FhIS0+IILq9VqaNzpp59eHXxTmwi+ab+0ORi5CGPQoEGGzu+tORrNyDm/i59eXP1rUalcQ1xSkKqD7/afg292/VMfZs2Zpay0LAV2C5TzR6cSRiZoUfaiJvt7WVRDSxlaQb7qqqu0evVqvf/+++rQoYMWLFigr776Sm+//bYSExMNf9l7772nefPmaerUqVqyZEm998rKyvTEE0/oj3/8o2677TZlZ2fXvldVVaWPPvpIw4cPN/xdAACYzZ0NcFffeLVXNsAZbYcwq7/XjOPOWBWGtxk+B/l3v/udPvjgA5WVlen48ePKzs7W5Zdf7taXde/eXZMmTdJll1120nsrV65Ux44dtWLFCqWmpmrFihXKz8+XJGVmZuqSSy6pt0EQAABfackGuLJbyryyAa5jx44+7e81M/iy+Q3eYqjFosbBgwfrXRISEBCg3r17G/752NhYSdLevXv13Xff1T6vqKhQTk6OnnrqKQUHB8tqtcpmsykzM1PXXXedDh48qP379ysrK0sFBQX65z//qenTp7tTOgAAzV5pbGSc0XYII9ceZ6zLqF3xba7VoWYDXHPjTpw44fE2hxo1gbakpESFhYWKiopqMKwaHVeDdgj4myYDcm5urubOnauNGzdKks477zwdP35ckuRyuWSxWJSZmam4uLhWFVFQUKAOHTqoV69etc9iYmK0e/duSdK0adNqn8+bN6/ZcLx+/XqtX7++9nVeXl6zNTidTkPj4H3MjX9iXvyXv8/Nd999p5KSEoWHh+v000/3yrivv/5a8x+er62fbK3tYb1o2EVacO8CnX322W6Nu/m2m5sMvsm3J+uZJ55RVmaWXHc0vtqbuShTO3bskMvlUsX3FdUrvg21RTgkxw8OdevWzdC4n376SffMuUc51+TomI6dFHy7fNJF89bMU15enjp37qw3Xn9DDzzygD5Z9Entrzn24ljNf32+Onfu3Oj/lsLCwnTs2DEdO3aswffdHedL/v7vzKnMV3PTZEBevHixxo4dW+/ZG2+8oTPPPFMul0vPPfeclixZ0uqAXFFRoZCQkHrPQkJCVFFRcdLYxx57rNnPS0xMrO2NXrhwYYNH1P1aXl6eoXHwPubGPzEv/suTc2N0xdXIWLvdrpTZKcrOyq5eCf3BoREJI5T2VFq9/6RuxrhJUyZVh9o7XHIGOiWHtGX7Fk2aMqneBrjmxp1++unK3Zrb5Aa4rWlbdezYMQV3D67+jIYESUGnBalTp04aNGiQRiSMaHLFN2FkgkaNGmVo3IUXXihJ2rpla8Mb27YbamfBAAAgAElEQVTU39gWExOj0aNHG17tbY/4/cx/+WpummzqbWhj3ODBgzV06FDZbDbNmDGj3ma6lgoODlZ5eXm9Z+Xl5QoODm71ZwMATuap29WMjvX0NcVGx0mn5gY4if5eoDWaDMj5+fmKjo6uff3444/XWxWIjIxUUVFRq4vo2bOnKisrVVBQUPssLy+v3ne7Kzc3V8uWLWt1bQBgFqMXQrgztuY4sdYEX3fCp6cvpPD0uFN5A1wNgi/gviYDcufOnWtPkpCkW2+9VWFhYbWvDxw4oNDQUMNfVllZKafTqaqqKlVVVcnpdKqyslLBwcGKjY3V66+/roqKCtntdm3dulUJCQkt+CVVq1nhBgBv8/bq7K/HNXacmKfDrNGxZlxTbPSCC0/fAFezAc6XF1xw5BlgviYD8pAhQ/TWW281+v7atWtre52MSE9P17Rp0/TWW28pKytL06ZNU3p6uiTp5ptvltPpVHJysp555hklJye3agUZALzNV6uzRo8T8/Ttau4EX0+GVKPjCgsLPX4DXFRUVIvP+Q1dHmro2mOjK76sDAPmaXKT3u23365rrrlGffv21a233lp7DnFlZaXS0tK0ZMkSrV692vCXJSUlKSkpqcH3QkNDNXfuXDdKb1pubq62bdvmsc8D0H7549FfkgyPNTJu1YurPH67miRDY+v14zZy+sJJIdUD42o2m3n6Brjw8HDDN7vVPe5s+/btGjJkCIEWaAOaXEH+wx/+oLlz52rWrFnq3r27LrzwQl144YXq3r275syZozlz5mjixIneqtUttFgAaE5L2hcaG+fL1VlPruK6s7nM6Oqs0X7cmpDqqXE1QdRfNsANGDCAcAy0Ec1eTffwww/r448/1v/8z/+oZ8+e6tmzp6ZPn64tW7YYOnINAHzBSB+wp05L8HSrQWFhocd7Z40GX3c2l0VERHi8H9esMMsGOADuaDQgjxo1Sj/88IOk6v+DePzxx/Xuu+/q3Xff1T/+8Y/aW/EAwBPcOaXBExvg/P3oL0/3zroTfN0Jny3tx21tSHU3zLIBDoA7LC6Xq8Elj6CgIO3bt0+9evVSQECACgoKdMYZZ3i7vhar6UGeOXOmGvkl1sMh4f6LufFPRufFm5dH1OsDHvpzoK1zbW5NcCouLlZ0v2g5Uh2N9rAGpQXpv1v/q99e9FuPjTu476D+3/X/TxurGulz/cii0QGja3uQE8cnGhprdFy9fz4NXCtcN1ja7fbqHtvMX/XYPrnopLDozlhJhi+k8PQ4X+P3Mv/F3PgvT8+NxWIxlAsbDcgXXHCBLrzwQl122WW68cYbtWjRInXt2rXBD7n++utbV62JjP6D4F8O/8Xc+J/i4mLt2LFDF154YaOb2jwZaI2OMxoUd+3apbhxcTp689FGf41hK8L0UtpL+p/U/2l23Jb3tujOuXeaElKNjDUr+Eruhc+2ElR9hd/L/Bdz47/8LiBv2bJFd9xxh77++msdPXpUnTt3lsVy8n+as1gsOnq08f/z8DUCctvH3PiPuqG3U9dO+unoT42u9noy0BoZt+rFVYZWhQ/uOyiXy+XRFeSD+w6qpKTEp6uzdcfVzI2ngi88g9/L/Bdz47/8LiDX1aFDBxUWFrapFosaBOS2j7nxHsPHmDUReiVjrQFGA63RkJrxboaunHqlodXeQYMGebx9oeafka9XZzlOzH/xe5n/Ym78l68CcpPnINfYt29fk2eDAvANT5zfKxlrhzB6Jm/tiQ4eOm+35pQGT521GxUVJal6c1lsfKxK1fCKb1p2mlvjpPpn3hoJszVn6hphdCzHiQFA6zV7zJskxcTEaNeuXUpJSdG4ceNUUFAgSXrrrbe0Y8cOUwtsqdzcXC1btszXZeAU4KnTF9wZ58nzez15jJk7V/saPdHBjFMaJI7+AgA0zlBAfv/993XRRRfp0KFD2rhxo8rLyyVJ33zzjRYsWGBqgS3FRSEwmydDqrvjPHV+r+TZY8zcudrX05dHuHs8mcTRXwCAhhnqQY6NjdUNN9yg2267TWFhYfr000/Vr18/bdu2TVdddZUOHz7sjVpbhB7kts8Xc2OkJcGTpy94ur/X0xvb3NmsFh4e7vFjx8w8paE94vcz/8S8+C/mxn/5qgfZ0Ary559/rvHjx5/0vEePHjpy5Ij71QE+4o+XTLgzzlfXD584ccKUq33NaHNgtRcA0FqGNul1795dhw4dUp8+feo93759u84880wz6gLc4olNaPVWKVNd1ZvCHNKG7dUtCXUvmTC6Cc2T4+r29/piY5u7m9VysnOqV3LTfrWSm72owUDb3MY2MzfAAQBQl6EV5KlTp+ruu+/WwYMHZbFYdOLECX3wwQe66667/PaSEDbp+S+jm9Vqxu7Zs6dVK76e7MWVZLgftyakemqcO/29Zmxs+/UqbujyUI9c7VvD6MY2NsABAMxmKCA/9NBD6tu3r2JiYlRWVqaBAwdq1KhRio+P17333mt2jS3CJj3/Y7R94ddjr77xatODrzunNHg6pBodFxUVpYiICJ9ubKsbetNfSDfUvkCgBQC0NYYCcqdOnfTqq6/qq6++0urVq7Vq1SrZ7Xb961//UkBAgNk1oh0wGmYbGlt2S5npwddoL25hYaHHQ6o7YVYy3t/r6T7gujhrFwDQnhkKyDX69eunyZMnKykpSeecc44k6eOPPzalMLQvRtsXjI71dPA1eiZv3UsmPBlS3V3FZWMbAADmMRSQy8rKas8+rrFjxw5dccUViouLM6UwtD2N9Ra7077gq+Dr60sm3F3FNev8XtohAABo5hSLgwcPasqUKfr4448VEBCglJQUPfzww7rtttv06quv6ve//72ys7O9VSv8VHMnRBg9eaGwsLD6pQdPX6gJvhu3NXIm7696cY2e0iB5/vQFd09pkNy7fpjQCwCAMU0G5Hnz5qmsrEzPPPOM0tPT9cwzzygrK0uDBw/Wnj171LdvX2/V6bbc3Fxt27bN12W0e0aORqu3qa2Zo8R8GXzdOZqsLk+HVMIsAAC+1eRNer1799bq1asVFxenwsJC9erVS4888ojmzZvnzRpbhZv0zOXJW90y1mW49Zlm3q7mziruqYp/Z/wXc+OfmBf/xdz4L1/dpNdkQA4ICNChQ4dqNyZ16dJFubm5Ou+88zxWqNkIyOYpLi42dE3xwX0HVVJS4laYJfj6P/6d8V/MjX9iXvwXc+O/fBWQm71Jr+4xbh06dFBwcHDrKkO74U5v8aBBg9y6Wa3u2E5dO+mnoz+16ha2GrQvAACA5jQZkF0ul0aOHKmOHauHlZeXa9y4cQoMrL9cuHPnTvMqhN9yp7dYci/M1h27fft2DRkyhOALAAC8osmAPH/+/Hqvr776alOLQdtSc2GGkY1ydbkTZrmQAgAAeJtbARn4NXePRgMAAPB3bt2kB/xaS64pBgAA8GfNbtJrqzgH2XtacsEFAACAv2q3K8g2m00zZszwdRmnFK4pBgAA7UG7DcgAAABASxCQAQAAgDrcDshdu3bV3r17zagFAAAA8Dm3A7KR6/kAAACAtooWCwAAAKAOtwPytGnT1LVrVzNqAQAAAHzO7XOQly5dakYdAAAAgF+gxQIAAACog4AMAAAA1MFV0wAAAEAd7XYFmaumAQAA0BKGAnJxcbGKi4trX3/22We677779Nprr5lWGAAAAOALhgJyUlKS3n77bUlSSUmJEhIS9Oabb2rmzJl66qmnTC0QAAAA8CZDAXnnzp26+OKLJUlr167V2Wefrc8//1wvv/yyli1bZmqBAAAAgDcZCsjl5eUKDQ2VJK1fv16///3vJUlDhgxRfn6+edUBAAAAXmYoIJ9zzjl64403lJ+fr/fff1+XX365JKmoqEinnXaaqQUCAAAA3mQoIM+fP19//vOf1adPH1188cWKjY2VJP3nP//RhRdeaGqBAAAAgDcZOgd50qRJOnDggA4fPqwLLrig9nliYqKuvvpq04oDAAAAvM3wRSGRkZGKjIxUWVmZLBaLunTpUruSDAAAALQXhi8KWbJkic466yx169ZNXbt2VUxMjJ599lkzawMAAAC8ztAK8iOPPKJHH31Ud911l+Lj4yVJWVlZmjdvno4ePap58+aZWiQAAADgLYYC8nPPPafly5fr2muvrX02evRonXPOOfrLX/5CQAYAAEC7YajF4ttvv9VFF1100vNhw4apqKjI40UBAAAAvmJoBfncc8/VqlWrdP/999d7vmrVKg0YMMCUwur64Ycf9OSTTyogIEAdOnRQamqqunfvbvr3AgAA4NRjKCA/8MADSkpKUmZmpuLi4mSxWJSdna0PPvhAa9asMbtGde3aVQ8++KA6dOigzZs3a+PGjRwvBwAAAFMYPgc5JydHf//73/XOO+/I5XJp4MCB+uSTT7xyUUiHDr90gpSXlys6Otr07wQAAMCpyfA5yEOHDtUrr7zSqi977733tHnzZh04cEBxcXG6/fbba98rKyvT0qVLtXPnToWFhWnq1Km1J2ZI0v79+7V8+XIdO3ZM9913X6vqAAAAABrT5Ca9qqoqPfHEE4qLi9OwYcP0l7/8RRUVFS3+su7du2vSpEm67LLLTnpv5cqV6tixo1asWKHU1FStWLFC+fn5te/36dNHjzzyiKZMmaI333yzxTUAAAAATWkyIC9cuFDz5s1Tly5d1LNnTz399NNKTU1t8ZfFxsZq2LBhCgsLq/e8oqJCOTk5mjJlioKDg2W1WmWz2ZSZmSlJ+umnn2rHhoSEKCgoqMU1AAAAAE1pssXixRdfVFpamm677TZJ1S0Sf/jDH7Rs2TJZLBaPFVFQUKAOHTqoV69etc9iYmK0e/duSdK+ffv06quvqkOHDurUqZNuvfXWJj9v/fr1Wr9+fe3rvLy8ZmtwOp2GxsH7mBv/xLz4L+bGPzEv/ou58V++mpsmA3JeXp6uvPLK2te/+93v5HK5dPjwYfXu3dtjRVRUVCgkJKTes5CQkNp2jnPPPVcLFiww/HmJiYlKTEyUVL0KHhMT0+zP5OXlGRoH72Nu/BPz4r+YG//EvPgv5sZ/+WpummyxcDqd6ty5c+1ri8WiwMBAORwOjxYRHBys8vLyes/Ky8sVHBzs0e8BAAAAmtPsKRb33HNPvdVdp9OpBQsWqFu3brXPFi1a1KoievbsqcrKShUUFKhnz56Sqv/E0Jrj3HJzc7Vt27ZW1QUAAIBTT5MBOSEhQd988029Z8OHD9eBAwda9GWVlZWqrKxUVVWVqqqq5HQ6FRAQoODgYMXGxur111/XzJkztX//fm3dulUPPfRQi75Hkmw2m2w2m2bOnNnizwAAAMCpp8mAvHnzZo9+WXp6utauXVv7OisrS5MnT1ZSUpJuvvlmPfvss0pOTlZoaKiSk5O5EAQAAABeZ+iikAcffFB33XXXSRvpysvL9cQTT+j+++839GVJSUlKSkpq8L3Q0FDNnTvX0OcYQYsFAAAAWqLJTXo1FixYoLKyspOeHz9+3K3TJbzJZrNpxowZvi4DAAAAbYyhgOxyuRo893jHjh3q0aOHx4sCAAAAfKXJFouwsDBZLBZZLBb169evXkiurKxURUUFm+AAAADQrjQZkBcvXiyXy6Xp06fr4Ycfrne0W2BgoPr06aNLLrnE9CJbgh5kAAAAtESTAfmGG27QiRMndOzYMU2YMEFnnnmmt+pqNY55AwAAQEs024PcsWNHzZ07V5WVld6oBwAAAPApQ5v0Lr74YtoVAAAAcEowdA5ycnKy7rrrLh04cEBDhw5Vly5d6r0/ZMgQU4prDXqQAQAA0BKGAvLUqVMlSbNnzz7pPYvF4pftF/QgAwAAoCUMBeR9+/aZXQcAAADgFwwF5JiYGLPrAAAAAPyCoYAsSSdOnNAnn3yiAwcOyOl01nvv+uuv93hhAAAAgC8YCsh2u11XXXWV9u3bJ5fLpYCAAJ04cUKdOnVSUFAQARkAAADthqFj3u68804NHTpUP/74o0JCQvTFF18oNzdXv/3tb5Wenm52jS2Sm5urZcuW+boMAAAAtDGGAvLWrVt13333qUuXLurQoYNOnDihIUOG6PHHH9ecOXPMrrFFbDabZsyY4esyAAAA0MYYCsgul0shISGSpIiICB06dEiSdOaZZ+rrr782rzoAAADAywz1IA8aNEiffvqp+vXrp2HDhmnhwoUKCAjQihUrdPbZZ5tdIwAAAOA1hgLyvffeq2PHjkmSHnroIV155ZW67LLLFB4ertdff93UAgEAAABvMhSQf/e739X+fb9+/bR7924dOXJE3bt3l8ViMa04+Ifi4mIVFRUpMjJSERERvi4HAADAVIZ6kKdPn67S0tJ6z3r06KHjx49r+vTpphTWWpxi0Xp2u12J4xMV3S9acePiFN0vWmOuGCO73e7r0gAAAExjKCC/9NJLKi8vP+l5eXm5Xn75ZY8X5QmcYtE6drtdsfGx2li1UY5Uh47efFSOWQ5tqNyg2PhYQjIAAGi3mmyxOHLkiFwul1wul77//nt17PjL8MrKSq1bt06RkZGmFwnvS5mdolJbqVyXuH55GCS5LnGpVKWaNWeWMtZl+K5AAAAAkzQZkMPDw2WxWGSxWDRw4MCT3rdYLFqwYIFpxcE3iouLlZ2VLVeqq8H3XUNcykrLUklJicLDw71cHQAAgLmaDMibNm2Sy+XSqFGjlJ6erh49etS+FxgYqJiYGPXq1cv0IuFdRUVFCjotSI5AR8MDgqTAboEqLCwkIAMAgHanyYA8cuRISdK+fft01llncWLFKSIyMlKOHxySU1JgAwMckvNHp6KiorxdGgAAgOkMbdLbv3+/Pvnkk9rXL774ouLj4zVjxgyVlZWZVhx8IyIiQvEj4mXZ1vAfiCzbLRqRMILVYwAA0C4ZCsh33nmnCgsLJUl79uzRjBkzdP755+ujjz7S3XffbWqB8I3FTy9WWG6YLB9ZpJpOC4dk+ciisNwwpT2V5tP6AAAAzGIoIH/zzTcaPHiwJCk9PV1jxozRs88+qxUrVujtt982tcCW4hzk1rFarcrJztHogNEKSgtS2IowBaUFKbFjonKyc2S1Wn1dIgAAgCkM3aRnsVhUWVkpSdqwYYMmTpwoSYqKitJ3331nXnWtYLPZZLPZNHPmTF+X0mZZrVZlrMtQSUmJCgsLFRUVRVsFAABo9wwF5Isuukh/+9vfNGbMGGVlZWn58uWSqnuT2ajV/oWHhxOMAQDAKcNQi8U//vEP/fe//1VKSoruvfde9e/fX5K0Zs0aDR8+3NQCAQAAAG8ytII8aNAg7dy586TnTz75pAICAjxeFAAAAOArhgJyY4KDgz1VBwAAAOAXDLVYoH0qLi7Wrl27VFxc7OtSAAAA/AYB+RRkt9uVOD5R0f2iFTcuTtH9ojXmijGy2+2+Lg0AAMDnWtVigbbHbrcrNj5WpbZSuVJdcgQ6JIe0YfsGxcbHcsYxAAA45RlaQS4pKTG7DnhJyuyU6nB8iUsK/PlhkOS6xKVSW6lmzZnl0/oAAAB8zVBA7tWrl6666iqtXr1aFRUVZtcEkxQXFys7K1uuoa4G33cNcSkrM4s/EAEAgFOaoYD8zjvv6PTTT1dycrIiIyN14403auPGjXK5Gg5a/oCrpk9WVFSkoNOCflk5/rUgKbBboAoLC71aFwAAgD8xFJAvv/xyvfjiiyoqKtLy5ct15MgRjRs3TtHR0Zo7d67ZNbaIzWbTjBkzfF2GX4mMjJTjB4fkbGSAQ3L+6OR2RAAAcEpz6xSL4OBgTZkyRf/7v/+r//73v4qIiNBTTz1lVm3wsIiICMWPiJdlm6XB9y3bLRqRMIJrpQEAwCnNrYB87NgxvfLKKxo3bpwuuOAClZaW6r777jOrNphg8dOLFZYbJstHFsnx80OHZPnIorDcMKU9lebT+gAAAHzN0DFv69at06uvvqr/+7//U+fOnXXNNddo8+bNGj58uNn1wcOsVqtysnM0a84sZaVlKbBboJw/OpUwMkGLshdxxBsAADjlGQrI11xzja688kqtWrVK48ePV8eOHJ/cllmtVmWsy1BJSYkKCwsVFRVFWwUAAMDPDCXdwsJCde3a1exa4GXh4eEEYwAAgF8x1IPcWDjOzc3V2LFjPVoQAAAA4EvNBuSMjAzdfffd+stf/qK9e/dKkr788ktNmDBBsbGxphcIAAAAeFOTLRYvvfSSbrzxRvXo0UNHjhzR888/r2eeeUYzZszQpEmT9Omnn2rQoEHeqhUAAAAwXZMryH//+9/1yCOPqKSkRP/+979VXFysJ554Qtu3b9cLL7xAOAYAAEC702RA/uabbzRlyhRJ0uTJkxUQEKCnn35a/fv390pxAAAAgLc1GZCPHTumLl26VA/s0EHBwcGKjo72SmEAAACALzR7zNu6devUrVs3SVJVVZX+85//KDIyst6YSZMmmVMdAAAA4GXNBuSbbrqp3uvbb7+93muLxaLKykrPVtWAL7/8Ui+99JI6duyo7t27KyUlhQtLAAAA4HFNJsyqqipv1dGs8PBwzZ8/X4GBgXrttdeUm5uriy++2NdlAQAAoJ1pM0uwPXr0qP37gIAAWSwWH1YDAACA9qrJgJyenq5x48YpJCTEY1/43nvvafPmzTpw4IDi4uLqtWyUlZVp6dKl2rlzp8LCwjR16lTFx8fX+/lvv/1WO3bsoO8ZAAAApmjyFItrrrlGPXv21K233qodO3Z45Au7d++uSZMm6bLLLjvpvZUrV6pjx45asWKFUlNTtWLFCuXn59e+f/z4cS1ZskSzZs2i/xgAAACmaPaq6ZSUFG3atEk2m01DhgzRsmXLVFpa2uIvjI2N1bBhwxQWFlbveUVFhXJycjRlyhQFBwfLarXKZrMpMzNTklRZWalnnnlG11xzjXr16tXi7wcAAACa0uwy7B133KGHH35Ymzdv1vLly3XnnXdqzpw5mjJlipKTkz22Ua6goEAdOnSoF35jYmK0e/duSdKHH36or7/+WmvXrtXatWt1+eWXa/jw4Q1+1vr167V+/fra13l5ec1+v9PpNDQO3sfc+CfmxX8xN/6JefFfzI3/8tXcGO5TuPTSS3XppZfq+++/14svvqjnn39eL7zwgn7zm9/os88+a3UhFRUVJ/U6h4SEqKKiQpKUkJCghIQEQ5+VmJioxMRESdLChQsVExPT7M/k5eUZGgfvY278E/Piv5gb/8S8+C/mxn/5am6abLFo6KSI7t27609/+pN27dqlzMxMDR061COFBAcHq7y8vN6z8vJyBQcHe+TzAQAAACOaXEF2uVxN/nB8fPxJp0y0VM+ePVVZWamCggL17NlTUvWfGlp6tXVubq62bdvmkdoAAABw6mhyBXnTpk31zh/2hMrKSjmdTlVVVamqqkpOp1OVlZUKDg5WbGysXn/9dVVUVMhut2vr1q2G2yp+zWazacaMGR6tHQAAAO1fkyvII0eO9PgXpqena+3atbWvs7KyNHnyZCUlJenmm2/Ws88+q+TkZIWGhio5ObnFK8gAAABASzS7Sa+yslJ79+5Vnz591KlTJ1VUVCg9PV0Oh0NXXnmlzjjjDLe+MCkpSUlJSQ2+Fxoaqrlz57r1eY2hxQIAAAAt0WSLhd1uV58+fTRgwAANGDBAe/fu1fDhwzVjxgzNmjVL5513nr788ktv1eoWWiwAAADQEk0G5Hnz5mnIkCH69NNPNWHCBF155ZU666yz9P333+vIkSOKi4vT3/72N2/VCgAAAJiuyYC8ZcsWLViwQIMHD9ZDDz0ku92uu+66S506dVJQUJDmzZtXe9MdAAAA0B402YNcVlZWe4pFly5d1KVLl9oj2CQpOjpaRUVF5lbYQvQgAwAAoCWaXEHu1auX8vPza18//vjj9TblFRcX67TTTjOvulagBxkAAAAt0WRAHjVqlHbv3l37+tZbb1VYWFjt6/Xr12vIkCHmVQcAAAB4WZMtFsuXL2/yh6+++mpNnTrVowUBAAAAvtTsOchN6d+/v6fq8Dh6kAEAANASTQbkl19+2dCHXH/99R4pxpNsNptsNptmzpzp61IAAADQhjQZkG+//fZ6r48fP67OnTvLYrHUPrNYLH4ZkAEAAICWaDIgl5aW1nsdFhamTz/9VP369TO1KAAAAMBXmjzFAgAAADjVtGqTnj9jkx4AAABaot2uIHNRCAAAAFrCrYBssVjqbdADAAAA2psmWyzCwsLqBeKysjKdf/75J4Xko0ePmlMdAAAA4GVNBuTFixd7qw4AAADALzQZkG+44QZv1QEAAAD4hUZ7kF0ul1sf5O54AAAAwB81GpCtVqteeeUVVVRUNPkBX3zxhZKTk/XYY495vLjWyM3N1bJly3xdhtcVFxdr165dKi4u9nUpAAAAbVKjLRbLly/X3LlzlZKSosTERNlsNvXs2VPBwcH6/vvvtXv3bmVnZ+vLL79UamqqUlJSvFl3s2w2m2w2m2bOnOnrUrzCbrcrZXaKsrOyFXRakBw/ODQiYYTSnkqT1Wr1dXkAAABtRqMBeeTIkcrJydGWLVv02muv6d///rfy8vJUXl6u8PBwXXjhhZo+fbqmTZum0047zZs141fsdrti42NVaiuVK9UlR6BDckgbtm9QbHyscrJzCMkAAAAGNXuT3vDhwzV8+HBv1IIWSpmdUh2OL6nTBx4kuS5xqVSlmjVnljLWZfiuQAAAgDakxTfp/fTTT56sAy1UXFys7KxsuYY2vEnSNcSlrMwslZSUeLkyAACAtslQQF60aJHS09NrX0+fPl2dO3fWgAEDtGfPHtOKQ/OKiooUdFqQFNjIgCApsFugCgsLvVoXAABAW2U4IEdEREiSMjMztWbNGq1atUq//e1vNWfOHFMLRNMiIyPl+MEhORsZ4JCcPzoVFRXl1boAAADaqmZ7kCXp0KFD6tOnjyTp7bff1jXXXKOkpCQNHjxYI0aMMLM+NCRHHFIAAB6ASURBVCMiIkLxI+K1cdvG+j3IP7Nst2hEwgiFh4f7oDoAAIC2x9AKcteuXWvP1c3IyNDo0aMlSZ06dWr2nGRfOZXOQV789GKF5YbJ8pFFcvz80CFZPrIoLDdMaU+l+bQ+AACAtsRQQL788suVnJysm266SV9//bXGjRsnSfr888/Vt29fUwtsKZvNphkzZvi6DK+wWq3Kyc7R6IDRCkoLUtiKMAWlBSmxYyJHvAEAALjJUIvFkiVLdO+99+rAgQNau3atevToIUnavn27rr32WlMLhDFWq1UZ6zJUUlKiwsJCRUVF0VYBAADQAoYCcteuXZWWdvJ/pl+wYIHHC0LrhIeHE4wBAABawVBAlqSKigqtWrVKu3fvliQNHDhQ1157rTp37mxacfhFcXGxioqKFBkZWXuiCAAAADzPUA/y9u3b1bdvX82ZM0effPKJPvnkE911113q16+ftm/fbnaNpzS73a7E8YmK7hetuHFxiu4XrTFXjJHdbvd1aQAAAO2SoYB8yy23aMSIETp48KAyMzOVmZmp/Px8JSQk6JZbbjG7xlOW3W5XbHysNlZtlCPVoaM3H5VjlkMbKjcoNj6WkAwAAGACQwH5888/1wMPPKAuXbrUPuvSpYvuv/9+ff7556YVd6pLmZ2iUltp9fnGNTflBUmuS1wqtZVq1pxZPq0PAACgPTIUkK1Wqw4fPnzS84KCAp177rkeLwrVPcfZWdlyDT358g9Jcg1xKSszSyUlJV6uDAAAoH1rdJPekSNHav/+oYceUmpqqu6//35dfPHFkqSPP/5YDz30kB577DHzq2zHGtt8V1RUpKDTguQIdDT8g0FSYLdAFRYWcmoFAACABzUakMPDw2WxWGpfu1wuTZ06tfaZy1W9sjlhwgRVVlaaXGb7Y7fblTI7RdlZ2dVB+AeHRiSMUNpTabJarYqMjJTjB4fk1C/tFXU5JOePTkVFRXm7dAAAgHat0YC8adMmb9bhcbm5udq2bZuvy2hQzea7UlupXKmu6lVih7Rhe/Xmu5rb7+JHxGvjto3VPci/Ytlu0YiEEaweAwAAeFijAXnkyJHerMPjbDabbDabZs6c6etSTlJv812Nms13qt58l7EuQ4ufXlwdpPX/27v36JjP/A/g75HLjJmM/JBIkMgRERGURhhEhTYubVdt2RXXUtem7u2W5rh1y9bRqj3bqFuctVsiayUOjq5scVxilZqUIFcJk1BMgpKEXEQ+vz+6ZmeQmKSJmSTv1zk5x/f5fr/P9/OdzznTz3n6PM8UQoIEUAIo/aU41uq1iDrx9I+3EBEREdGvY/UPhQDA9evXkZubi7KyMov2AQMG1GpQDUFlc4tNi+/mVrH4LuqXxXcBAQE4feI05nw4B4lRiXB2dUbZvTIMCB2Ar058hYCAgBf1OkRERESNhlUF8vXr1zFu3DgcP34cCoUCImIxP5lzkP/neXOLq7v4LiAgAAe/PYhbt27h5s2b8PT05LQKIiIiojpk1TZv8+fPh4ODA1JTU6FWq5GYmIhdu3ahc+fOSEhIqOsY6w1rftjDYvHds1Sy+M7NzQ1du3ZlcUxERERUx6waQT527Bi+/fZbBAQEQKFQwN3dHSEhIVAqlVi6dCkGDx5c13HWC9bOLebiOyIiIiL7ZdUIcnFxsalga9GiBfLy8gAAgYGBOH/+fN1FV49U54c91q1dB61eC8X3CuDxTItSQPH9fxfffcnFd0RERES2YvUv6aWnpwMAevTogY0bNyInJwdff/012rZtW6cB1heP5xY/c89iwGJu8ePFd685vAZllBLaaC2UUUqEOYaZtngjIiIiItuwaorFvHnzcPPmTQDAsmXLMGzYMMTGxkKpVOLvf/97nQZYX1T3hz24+I6IiIjIPllVII8fP97076CgIBgMBqSnp6Ndu3Ys6v7L3d29RnOL3dzc+BkSERER2ZFq7YP8mFqtRlBQUG3HUu/xhz2IiIiI6j+r5iCTdTi3mIiIiKj+q9EIMlWOc4uJiIiI6jcWyHWEc4uJiIiI6qd6McXiwYMHiIyMxMSJE5Gbm2vrcIiIiIioAasXBbKzszMiIyPRp08fW4dCRERERA1cpQWyWq1Gfn6+1R21atUKBoOhNmJ6iqOjI5o1a1YnfRMRERERmat0DnJJSQn2798PV1dXqzq6f/8+KioqqrwmISEBR48eRW5uLkJCQjBr1izTuaKiImzYsAHnz5+HVqvFuHHj0L9/fytfg4iIiIiodlS5SG/q1Km1+rDmzZtj5MiRSE5ORllZmcW5LVu2wNHREdHR0TAYDFi1ahV8fHzg7e1dqzEQEREREVWl0ikWFRUV1f7z9fWt8mE6nQ69e/eGVqu1aC8pKcHp06cRHh4OlUqFgIAABAcH4/jx47XzlkREREREVrKLbd5u3LiBJk2aoE2bNqY2Hx8fpKammo5XrVoFg8GA69evY/DgwRg4cGCl/R06dAiHDh0yHefk5Dw3hrKyMquuoxePubFPzIv9Ym7sE/Niv5gb+2Wr3FhVIH/66afPbFcoFFCpVPDz88OwYcPQtGnTGgVRUlICtVpt0aZWq1FSUmI6joyMtLq/sLAwhIWFAQBWr14NHx+f596Tk5Nj1XX04jE39ol5sV/MjX1iXuwXc2O/bJUbqwrkXbt2ITc3F/fv3zeN8l6/fh0ajQbu7u64evUqWrVqhWPHjj13msWzqFQqFBcXW7QVFxdDpVJVuy8iIiIiol/Dqn2QP/zwQ/Tq1QsGgwG5ubnIzc2FwWCATqfDsmXLcP36dfj7+2PBggU1CqJ169Z49OgRbty4YWrLycn5VQv09Ho9Nm3aVOP7iYiIiKhxsqpA/uMf/4i1a9fCy8vL1Obl5YXPP/8cy5YtQ8uWLfGnP/0Jp06dqrKfR48eoayszLSor6ysDI8ePYJKpYJOp8POnTtRUlKC9PR0nDlzBgMGDKjxiwUHB2PmzJk1vp+IiIiIGierplgYjUaL+cCPlZaWIi8vDwDg4eGBBw8eVNlPfHw84uLiTMeJiYn43e9+h9GjR2PatGlYv349pk+fDhcXF0yfPp1bvBERERHRC2dVgRwWFoaZM2di8+bN6NmzJwAgKSkJERERGDx4MADgwoULaN++fZX9jB49GqNHj37mORcXFyxcuLA6sVdJr9cjKSmp1vojIiIiosbBqikWW7ZsgYeHB3Q6HZRKJZRKJfr06QMPDw9ER0cDALRaLdasWVOnwVYHp1gQERERUU1YNYLcqlUrJCQkICMjAxkZGRARdO7cGf7+/qZrBg0aVGdBEhERERG9KFYVyHv37sWbb76JTp06oVOnTnUdExERERGRzVg1xWLs2LHw9PREREQETp48Wdcx1Qpu80ZERERENWFVgWw0GvHFF18gKysLAwYMgK+vL5YuXYqMjIy6jq/GOAeZiIiIiGrCqgJZq9Xi3XffxcGDB3H16lXMnj0bBw4cQGBgIHr37l3XMRIRERERvTBWzUE217p1a8yePRs+Pj5YuXIlt1IjIiIiogalWgXykSNHEBMTg/j4eADA22+/jbVr19ZJYL8W90EmIiIiopqwaorFRx99BG9vbwwbNgx5eXnYtGkTbt68ib/+9a92u70b5yATERERUU1YNYL8n//8B5GRkRgzZgxatGhR1zEREREREdmMVQVyfdnajYiIiIjo17J6DnJ5eTl++OEH5ObmoqyszOLcO++8U+uBERERERHZglUFcnp6OoYPH44rV65ARODg4IDy8nI4OTlBqVTaZYHMRXpEREREVBNWLdKbP38+evbsiXv37kGtViMtLQ16vR49evQw7Whhb7hIj4iIiIhqwqoR5DNnzuDYsWPQaDRo0qQJysvLERQUhM8//xxz5szB+fPn6zpOIiIiIqIXwqoRZBGBWq0GALi7u+Onn34CAHh5eSErK6vuoiMiIiIiesGsGkHu2rUrkpOT4evri969e2P16tVwcHBAdHQ0/Pz86jpGIiIiIqIXxqoCefHixbh//z4AYOXKlfjNb36DQYMGwc3NDf/85z/rNEAiIiIiohfJqgJ56NChpn/7+voiNTUVd+7cQfPmzaFQKOosOCIiIiKiF83qfZCfZO+/qMdt3oiIiIioJqxapFcf1cU2b/n5+bh48SLy8/NrtV8iIiIish8NtkCuTenp6Qh7Iwzevt4IeT0E3r7eGPzmYKSnp9s6NCIiIiKqZTWeYtFYpKenQ9dfh8LgQshcQalzKVAKHP7xMHT9dTh94jQCAgJsHSYRERER1RKOID/H7A9m/1Ic9xXA+b+NSkD6CgqDCzHnwzk2jY+IiIiIahcL5Crk5+fjROIJSE955nkJEiQeT8StW7decGREREREVFdYIFfBaDRC+X/K/40cP0kJOLs64+bNmy80LiIiIiKqOyyQq+Dh4YHSu6VAWSUXlAJl98rg6en5QuMiIiIiorrTYAtkvV6PTZs2/ao+3N3d0f+V/lAkPfvHUBQ/KvDKgFfg5ub2q55DRERERPajwRbItbUP8rq166DVa6H4XgGU/rexFFB8r4BWr0XUl1G/+hlEREREZD8abIFcWwICAnD6xGm85vAalFFKaKO1UEYpEeYYxi3eiIiIiBog7oNshYCAABz89iBu3bqFmzdvwtPTk9MqiIiIiBooFsjV4ObmxsKYiIiIqIHjFAsiIiIiIjMskImIiIiIzLBAJiIiIiIywwKZiIiIiMgMC2QiIiIiIjMskImIiIiIzDTYbd70ej2SkpJsHQYRERER1TMNdgS5tn5qmoiIiIgalwZbIBMRERER1QQLZCIiIiIiMyyQiYiIiIjMNNhFeuYUCoWtQyAiIiKieqLBF8giYtV1wcHB0Ov1dRwN1QRzY5+YF/vF3Ngn5sV+MTf2y1a54RQLIiIiIiIzLJCJiIiIiMw4fPLJJ5/YOgh70bNnT1uHQJVgbuwT82K/mBv7xLzYL+bGftkiNwqxdpIuEREREVEjwCkWRERERERmWCATEREREZlhgUxEREREZIYFMhERERGRGRbIRERERERmGn2BfOfOHbz99tvQaDTw8fHBjh07bB1So7Ru3ToEBwdDqVRi8uTJFucOHz6MgIAAqNVqDBo0CDk5ObYJspEqLS3F1KlT4ePjA61Wi5dffhkHDhwwnWd+bGfChAlo3bo1mjVrBn9/f2zZssV0jnmxD5cuXYJKpcKECRNMbTt27ICPjw80Gg1++9vf4s6dOzaMsHEZOHAgVCoVXFxc4OLigk6dOpnOMS+2949//AOdO3eGRqNBhw4dkJiYCMA232eNvkCeNWsWnJ2dYTQaERMTg4iICKSkpNg6rEanTZs2WLJkCaZMmWLRfuvWLYwcORIrVqzAnTt3EBwcjPDwcBtF2TiVl5fD29sbx44dw71797BixQqMHj0aBoOB+bGxyMhIGAwGFBQUYN++fViyZAmSkpKYFzsya9Ys9OrVy3SckpKCmTNnYtu2bTAajVCr1Xj//fdtGGHjs27dOhQVFaGoqAgZGRkAmBd7cPDgQSxatAhbt25FYWEhjh8/Dl9fX9t9n0kjVlRUJE5OTpKRkWFqmzBhgixatMiGUTVuixcvlkmTJpmON23aJH379jUdFxUViUqlkrS0NBtER49169ZN4uLimB87kp6eLp6enrJz507mxU7ExsbK73//e1m+fLmMHz9eREQiIyNl7NixpmuysrLEyclJCgoKbBVmoxIaGirR0dFPtTMvtte3b1/ZsmXLU+22+j5r1CPImZmZcHBwgL+/v6mte/fuHEG2IykpKejevbvp+PH/dmGObMdoNCIzMxNdunRhfuzA+++/D7VajYCAALRu3RpvvPEG82IHCgoKsGzZMnz55ZcW7U/mpkOHDnB2dkZmZuaLDrHRioyMhJubG0JCQnD06FEAzIutPXr0CHq9Hvn5+fDz84OXlxdmz56N4uJim32fNeoCuaioCK6urhZtrq6uKCwstFFE9CTmyL48fPgQ48ePx6RJkxAQEMD82IH169ejsLAQiYmJGDlyJJRKJfNiB5YuXYqpU6fC29vbop25sa3Vq1fj8uXL+OmnnzBjxgwMHz4c2dnZzIuNGY1GPHz4EHFxcUhMTMS5c+dw9uxZrFy50ma5adQFsouLCwoKCizaCgoKoNVqbRQRPYk5sh8VFRWYOHEinJ2dsW7dOgDMj71wcHBA//79ce3aNWzYsIF5sbFz587h0KFDWLBgwVPnmBvb0ul00Gq1UCqVmDRpEkJCQvCvf/2LebGxpk2bAgDmzJmD1q1bw83NDR988IFNc9OoC2R/f3+Ul5fj0qVLprbk5GR06dLFhlGRuS5duiA5Odl0fP/+fWRnZzNHL5iIYOrUqTAajYiPj4eTkxMA5sfelJeXmz5/5sV2jh49CoPBgHbt2sHT0xNr1qxBfHw8goKCnsrN5cuXUVpaajHVj14chUIBEWFebKx58+bw8vKCQqF46pzNvs/qdIZzPRAeHi5jxoyRoqIiOXHihDRr1kwuXrxo67AanYcPH0pxcbF8/PHHMmHCBCkuLpaHDx9KXl6eNGvWTOLi4qS4uFgWLlwoOp3O1uE2OjNnzhSdTieFhYUW7cyP7RiNRomNjZXCwkIpLy+XhIQEUavVsmfPHubFxu7fvy83btww/X344YcyatQoycvLk4sXL4pWq5Xjx49LUVGRjB8/XsLDw20dcqPw888/S0JCgum/L9u3bxe1Wi3p6enMix1YunSpBAcHi9FolDt37kj//v1lyZIlNvs+a/QF8u3bt2XEiBGiVqvF29tbYmJibB1So7R8+XIBYPG3fPlyERE5ePCgdOrUSVQqlYSGhsqVK1dsGmtjYzAYBIAolUrRaDSmv+3bt4sI82MreXl5MmDAAHF1dRWtVitdu3aVzZs3m84zL/bDfBcLEZGYmBjx9vYWtVotb731lty+fduG0TUeeXl5EhwcLC4uLuLq6io6nU6+++4703nmxbbKysokIiJCXF1dxcPDQ+bMmSPFxcUiYpvvM4WISN2OURMRERER1R+Neg4yEREREdGTWCATEREREZlhgUxEREREZIYFMhERERGRGRbIRERERERmWCATEREREZlhgUxEDc7evXvRsWNHODo6YvLkybYOh+yUXq+HQqGAwWCwdShEZGdYIBORXZg8eTIUCgUUCgUcHR3Rrl07RERE4Oeff652X9OmTcOoUaOQk5ODv/zlL3UQrX0wGo1YsGABOnbsCJVKhVatWqFfv36IiopCUVGRrcOrMwaDAQqFAnq93tahEFED5WjrAIiIHgsLC8O2bdtQXl6O1NRUTJkyBXfv3kVsbKzVfdy9exe3bt3C0KFD0bZt2xrHUlZWBmdn5xrfX9cMBgNCQkLQrFkzrFixAi+99BIqKiqQmZmJb775Bi1btsS4ceNsHSYRUb3EEWQishtKpRKenp7w8vLCkCFDEB4eju+++87imnv37mHGjBlo1aoVtFotQkNDTSOJR48eRfPmzQEAr776KhQKBY4ePQoAOHnyJEJDQ6FWq9G2bVtERESgoKDA1O/AgQMRERGBP/zhD3B3d0dISMhznwcAf/vb3+Di4oLDhw+ja9eu0Gg0GDRoEK5cuWIR97fffgudToemTZuiZcuWGD58OEpKSgD8UowvWrQIXl5e0Gg06NWrF/79739X+VlFRESgSZMm0Ov1GDNmDAIDA9G1a1eMHDkSe/bswdixY636zKrzDps2bYKfnx+cnZ3h5+eH6Ohoi/MKhQIbNmzAiBEjoFar4e/vjyNHjuDatWsYOnQoNBoNevTogR9//NHivufl5nkejyjHx8dj8ODBUKvVCAwMxMGDBy2uS0hIQEBAAFQqFV555RVkZmY+1VdVseTn56N169b49NNPTdefP38eKpUKcXFxVsdLRPVAnf+YNRGRFSZNmiRvvvmm6Tg7O1sCAwPFw8PD1FZRUSEhISHyxhtvyOnTp+XSpUuyZMkS0Wq1cv36dSktLZWUlBQBIPHx8XLjxg0pLS2V8+fPi0ajkTVr1khmZqacOnVK+vTpI6NGjTL1HRoaKi4uLvLBBx9IWlqapKamPvd5IiJbt24VR0dHee211+T06dOSnJwsPXr0kCFDhpj6PnDggDg4OMjixYslJSVFkpOT5YsvvpD79++LiMi4ceNEp9PJsWPHJDs7W6KiosTJyUnOnTv3zM/q9u3bolAoZNWqVc/9XGvrHXbv3i2Ojo4SFRUlGRkZ8tVXX4mjo6Ps27fPdA0AadOmjezYsUMyMzNl7Nix4uHhIUOGDJE9e/ZIRkaGvP7669KtWzfTPdbk5klXrlwRAHLmzBmL406dOsm+ffskMzNT3nnnHWnRooUUFhaKiEhubq4olUqZPXu2pKWlyc6dO6Vt27YCQK5cuWJ1LAkJCeLk5CQnT56UBw8eSGBgoEyePPm5eSCi+oUFMhHZhUmTJomDg4NoNBpRqVQCQADI2rVrTdccPnxYNBqNPHjwwOLe7t27y+rVq0VEJD8/XwDIkSNHTOcnTpwoU6ZMsbjn7NmzAkCMRqOI/FIgmxdu1j5v69atAkDS09NN57dv3y5OTk7y6NEjERHp16+fhIeHP/O9s7KyRKFQSE5OjkX7iBEjJCIi4pn3nDp1SgDI7t27Ldrbtm0rGo1GNBqNzJw5s9bf4d1337XoY9KkSRISEmI6BiAff/yx6fjChQsCQL788ktT25EjRwSA5Ofni4h1uXlSZQXyxo0bTddcu3ZNAEhiYqKIiERGRkrHjh2loqLCdM2KFSssCmRrY5k3b560b99eJk+eLB06dDAV4UTUcHAOMhHZjQEDBmDz5s0oLi5GdHQ0srOzMXfuXNP5pKQkPHjwAO7u7hb3lZSUIDs7u9J+k5KSkJWVhZ07d5raRAQAkJ2djVatWgEAevbs+dR91jxPqVSiU6dOpuM2bdrg4cOHuHv3Llq0aIGzZ89WupvGjz/+CBFBYGCgRXtpaSleffXVSt/pWRITE/Ho0SPMmDHDNH2jtt4hLS0NU6ZMseijf//+2Ldvn0XbSy+9ZPq3h4cHAKBbt25PteXl5cHNzc3q3FjD/Nlt2rQxPQcA0tLS0KdPHygUCtM1ffv2tbjf2lhWr16NhIQEfPPNNzh58iRcXFysjpGI6gcWyERkN9RqNfz8/AAAX331FQYNGoQVK1bgk08+AQBUVFTAw8MDiYmJT93brFmzSvutqKjAtGnTsGDBgqfOmS/k02g0T91nzfMcHS2/Sh8XYRUVFZXGZP4MhUKBM2fOwMnJyeJc06ZNn3mPn58fFAoF0tPTLdrbt28P4JfPsS7ewby4rKzN/B0en3tW2+N+rc2NNap6zuNCtyrWxmIwGHD16lUoFApcvnwZOp2uWnESkf1jgUxEdmv58uV4/fXXMWPGDLRp0wZBQUEwGo1o0qQJfH19re4nKCgIKSkppuK7OvfV5HlPevnll3H48GFMnz79medEBDdv3sSgQYOs6q9ly5YYMmQI1q1bhzlz5lQ5gllb79C5c2ecOHHCYhT5xIkTT418V1dNc1NdgYGBiI+Ph4iYiudTp05VO5aHDx9i/PjxeOutt6DT6RAREYGQkBC0a9euTuMnoheLu1gQkd0aOHAgunTpgpUrVwL4ZRu4kJAQjBgxAgcOHMCVK1fw/fffY/ny5c8cIX1s0aJF+OGHH/Dee+/h7NmzyMrKwv79+zFz5swqn1/T5z1p8eLF2LVrF5YsWYLU1FSkpKTgz3/+Mx48eAB/f3+MHz8ekydPRlxcHC5fvgy9Xo81a9Zg9+7dlfa5fv16VFRUoGfPnoiNjUVqaioyMzMRGxuL5ORkODg41Oo7fPTRR9i2bRu+/vprXLp0CVFRUYiJicHChQut7uNZapqb6nrvvfdgMBgwf/58ZGRkIC4uDhs3bqx2LEuXLkVeXh42bNiAefPmoU+fPpg4caJV/7eAiOoRG85/JiIyeXIXi8diYmLE2dlZDAaDiIgUFBTI3LlzpW3btuLk5CReXl4SHh4uWVlZIvLsRXoiImfOnJGhQ4eKVqsVtVotXbt2laVLl5rOh4aGyqxZs556/vOet3XrVtFoNBb3PLkQTURk7969EhQUJM7OztKyZUsZPny4FBcXi4hIWVmZLF++XNq3by9OTk7i4eEhw4cPF71eX+VnduPGDZk7d6506NBBnJ2dRaPRSHBwsHz22WdSUFBQ6++wYcMG6dChgzg6OkqHDh1k8+bNFvcAkF27dpmOn5WLtLQ0ASAXLlwwtT0vN0+qbJHe4+PK4tm/f7/4+/uLUqmUfv36yfbt2y0W6T0vlqNHj4qjo6PF+9y4cUPc3d3ls88+qzReIqp/FCJWTMwiIiIiImokOMWCiIiIiMgMC2QiIiIiIjMskImIiIiIzLBAJiIiIiIywwKZiIiIiMgMC2QiIiIiIjMskImIiIiIzLBAJiIiIiIy8/9qLYyGYsGuDwAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 720x468 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "fig , ax = plt.subplots()\n",
    "\n",
    "#plot good Reference Genomes\n",
    "plt.scatter( ALL_RefGenome_Avg_StartPos_Diffs_DF[ALL_RefGenome_Avg_StartPos_Diffs_DF.avg_StartPos_diff < 50000].index_num , ALL_RefGenome_Avg_StartPos_Diffs_DF[ALL_RefGenome_Avg_StartPos_Diffs_DF.avg_StartPos_diff < 50000].avg_StartPos_diff , color = 'green' , edgecolor = 'k' , linewidth = 1.0 , s = 70)\n",
    "\n",
    "#plot bad reference Genomes\n",
    "plt.scatter( ALL_RefGenome_Avg_StartPos_Diffs_DF[ALL_RefGenome_Avg_StartPos_Diffs_DF.avg_StartPos_diff >= 50000].index_num , ALL_RefGenome_Avg_StartPos_Diffs_DF[ALL_RefGenome_Avg_StartPos_Diffs_DF.avg_StartPos_diff >= 50000].avg_StartPos_diff , color = 'red' , edgecolor = 'k' , linewidth = 1.0 , s = 70)\n",
    "\n",
    "ax.yaxis.set_major_formatter(plt.ScalarFormatter())\n",
    "ax.ticklabel_format(style='sci' , axis = 'y' , scilimits=(0,0))\n",
    "ax.tick_params(axis='y', which='major', labelsize=12 , pad = 5)\n",
    "ax.tick_params(axis='x', which='major', labelsize=12 , pad = 5)\n",
    "\n",
    "ax.set_xlim([-1 , 62])\n",
    "\n",
    "plt.ylabel('avg[ abs(H37Rv startPos - RefGenome startPos) ] ' , color = 'k' , fontsize = 14)\n",
    "plt.xlabel('Reference Genome Index' , color = 'k' , fontsize = 14)\n",
    "ax.set_yscale(\"log\", nonposy='clip')\n",
    "\n",
    "fig = plt.gcf()\n",
    "fig.set_size_inches(10, 6.5)\n",
    "fig.tight_layout()\n",
    "\n",
    "file_name = '/n/data1/hms/dbmi/farhat/Roger/inhost_TB_dynamics_project/figures/SNP_recall_simulations/H37Rv_RefGenome_StartPos_diffs_mapping_quality.pdf'\n",
    "plt.savefig(file_name, bbox_inches='tight', dpi = 300 , transparent = True)\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Drop Reference Genomes that mapped badly across the entire genome from use in simulations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Strain</th>\n",
       "      <th>Decision</th>\n",
       "      <th>Evaluation</th>\n",
       "      <th>Strain name</th>\n",
       "      <th>Has_Feature_Table</th>\n",
       "      <th>Feature_Table_filename</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GCA_000008585.1_ASM858v1_genomic</td>\n",
       "      <td>lineage4,lineage4.1,lineage4.1.1,lineage4.1.1.3</td>\n",
       "      <td>4</td>\n",
       "      <td>CDC1551</td>\n",
       "      <td>YES</td>\n",
       "      <td>GCF_000008585.1_ASM858v1_feature_table.txt.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GCA_000016145.1_ASM1614v1_genomic</td>\n",
       "      <td>lineage4,lineage4.9</td>\n",
       "      <td>4</td>\n",
       "      <td>H37Ra</td>\n",
       "      <td>YES</td>\n",
       "      <td>GCF_000016145.1_ASM1614v1_feature_table.txt.gz</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                              Strain  \\\n",
       "0   GCA_000008585.1_ASM858v1_genomic   \n",
       "1  GCA_000016145.1_ASM1614v1_genomic   \n",
       "\n",
       "                                          Decision Evaluation Strain name  \\\n",
       "0  lineage4,lineage4.1,lineage4.1.1,lineage4.1.1.3          4     CDC1551   \n",
       "1                              lineage4,lineage4.9          4       H37Ra   \n",
       "\n",
       "  Has_Feature_Table                          Feature_Table_filename  \n",
       "0               YES   GCF_000008585.1_ASM858v1_feature_table.txt.gz  \n",
       "1               YES  GCF_000016145.1_ASM1614v1_feature_table.txt.gz  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ref_seqs_annot.head(n=2) #all Reference Genomes that have a Feature Table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(60, 6)"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.shape(ref_seqs_annot)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "Bad_RefGenome_IDs = list( ALL_RefGenome_Avg_StartPos_Diffs_DF[ALL_RefGenome_Avg_StartPos_Diffs_DF.avg_StartPos_diff >= 50000].index )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['GCA_000193185.2_ASM19318v2',\n",
       " 'GCA_000023625.1_ASM2362v1',\n",
       " 'GCA_000154605.2_ASM15460v2',\n",
       " 'GCA_000154585.2_ASM15458v2',\n",
       " 'GCA_001750865.1_ASM175086v1',\n",
       " 'GCA_001870145.1_ASM187014v1']"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Bad_RefGenome_IDs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "Ref_Genome_indices_to_Drop = [] #store the index of each Badly Mapped Reference Genome\n",
    "\n",
    "for RefGenome_i in ref_seqs_annot.index:\n",
    "    \n",
    "    if ref_seqs_annot.loc[RefGenome_i , 'Strain'].split('_genomic')[0] in Bad_RefGenome_IDs:\n",
    "        \n",
    "        Ref_Genome_indices_to_Drop.append(RefGenome_i)\n",
    "    \n",
    "#drop Bad Indices from Table\n",
    "ref_seqs_annot_for_sims = ref_seqs_annot.drop(Ref_Genome_indices_to_Drop , axis = 0)\n",
    "\n",
    "#reset index\n",
    "ref_seqs_annot_for_sims.reset_index(inplace = True , drop = True)\n",
    "\n",
    "#drop unneccessary column\n",
    "ref_seqs_annot_for_sims.drop('Has_Feature_Table' , axis = 1 , inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Strain</th>\n",
       "      <th>Decision</th>\n",
       "      <th>Evaluation</th>\n",
       "      <th>Strain name</th>\n",
       "      <th>Feature_Table_filename</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GCA_000008585.1_ASM858v1_genomic</td>\n",
       "      <td>lineage4,lineage4.1,lineage4.1.1,lineage4.1.1.3</td>\n",
       "      <td>4</td>\n",
       "      <td>CDC1551</td>\n",
       "      <td>GCF_000008585.1_ASM858v1_feature_table.txt.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GCA_000016145.1_ASM1614v1_genomic</td>\n",
       "      <td>lineage4,lineage4.9</td>\n",
       "      <td>4</td>\n",
       "      <td>H37Ra</td>\n",
       "      <td>GCF_000016145.1_ASM1614v1_feature_table.txt.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GCA_000016925.1_ASM1692v1_genomic</td>\n",
       "      <td>lineage4,lineage4.3,lineage4.3.2,lineage4.3.2.1</td>\n",
       "      <td>4</td>\n",
       "      <td>F11</td>\n",
       "      <td>GCF_000016925.1_ASM1692v1_feature_table.txt.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GCA_000153685.2_ASM15368v2_genomic</td>\n",
       "      <td>lineage4,lineage4.1,lineage4.1.2,lineage4.1.2.1</td>\n",
       "      <td>4</td>\n",
       "      <td>str. Haarlem</td>\n",
       "      <td>GCF_000153685.2_ASM15368v2_feature_table.txt.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GCA_000195955.2_ASM19595v2_genomic</td>\n",
       "      <td>lineage4,lineage4.9</td>\n",
       "      <td>4</td>\n",
       "      <td>H37Rv complete genome</td>\n",
       "      <td>GCF_000195955.2_ASM19595v2_feature_table.txt.gz</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                               Strain  \\\n",
       "0    GCA_000008585.1_ASM858v1_genomic   \n",
       "1   GCA_000016145.1_ASM1614v1_genomic   \n",
       "2   GCA_000016925.1_ASM1692v1_genomic   \n",
       "3  GCA_000153685.2_ASM15368v2_genomic   \n",
       "4  GCA_000195955.2_ASM19595v2_genomic   \n",
       "\n",
       "                                          Decision Evaluation  \\\n",
       "0  lineage4,lineage4.1,lineage4.1.1,lineage4.1.1.3          4   \n",
       "1                              lineage4,lineage4.9          4   \n",
       "2  lineage4,lineage4.3,lineage4.3.2,lineage4.3.2.1          4   \n",
       "3  lineage4,lineage4.1,lineage4.1.2,lineage4.1.2.1          4   \n",
       "4                              lineage4,lineage4.9          4   \n",
       "\n",
       "             Strain name                           Feature_Table_filename  \n",
       "0                CDC1551    GCF_000008585.1_ASM858v1_feature_table.txt.gz  \n",
       "1                  H37Ra   GCF_000016145.1_ASM1614v1_feature_table.txt.gz  \n",
       "2                    F11   GCF_000016925.1_ASM1692v1_feature_table.txt.gz  \n",
       "3           str. Haarlem  GCF_000153685.2_ASM15368v2_feature_table.txt.gz  \n",
       "4  H37Rv complete genome  GCF_000195955.2_ASM19595v2_feature_table.txt.gz  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ref_seqs_annot_for_sims.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(54, 5)"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.shape(ref_seqs_annot_for_sims)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### There are 54 Good Reference Genomes (we'll use these in our simulations)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "#save Reference Genomes Table (each Reference Genome has a feature annotation table & was mapped well against H37Rv)\n",
    "ref_seqs_annot_for_sims.to_csv('/n/data1/hms/dbmi/farhat/Roger/inhost_TB_dynamics_project/simulated_reads/reference_sequences/strains_typed/Reference_Genomes_used_in_simulations.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "################################################################################################################################################################################################################################################"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "## [4] Package the Complete Genome > H37Rv Mappings for sharing"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "################################################################################################################################################################################################################################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "#import necessary packages\n",
    "import vcf\n",
    "import os\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "from itertools import compress\n",
    "import ast\n",
    "import itertools\n",
    "\n",
    "import time\n",
    "import sys\n",
    "\n",
    "import Bio\n",
    "from Bio.Alphabet import IUPAC\n",
    "from Bio.Blast.Applications import NcbiblastnCommandline\n",
    "from Bio.Blast import NCBIXML\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.SeqFeature import SeqFeature, FeatureLocation\n",
    "from Bio import pairwise2\n",
    "from Bio import SeqIO\n",
    "from Bio.Graphics import GenomeDiagram\n",
    "from Bio.SeqUtils import GC\n",
    "\n",
    "from Bio.Align.Applications import MuscleCommandline\n",
    "from StringIO import StringIO\n",
    "from Bio import AlignIO\n",
    "from Bio.Align import AlignInfo\n",
    "from Bio.Seq import MutableSeq\n",
    "from Bio import pairwise2\n",
    "from shutil import copy\n",
    "import subprocess\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import fastcluster\n",
    "from sklearn import cluster, datasets\n",
    "import scipy.cluster.hierarchy as hier\n",
    "import matplotlib as mpl\n",
    "\n",
    "#for exporting to Adobe Illustrator\n",
    "mpl.rcParams['pdf.fonttype'] = 42\n",
    "mpl.rcParams['ps.fonttype'] = 42"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Load list of genomes (that were well mapped) that we used in the simulations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "ref_seqs_annot_for_sims = pd.read_csv('/n/data1/hms/dbmi/farhat/Roger/inhost_TB_dynamics_project/simulated_reads/reference_sequences/strains_typed/Reference_Genomes_used_in_simulations.csv').set_index('Unnamed: 0')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Strain</th>\n",
       "      <th>Decision</th>\n",
       "      <th>Evaluation</th>\n",
       "      <th>Strain name</th>\n",
       "      <th>Feature_Table_filename</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GCA_000008585.1_ASM858v1_genomic</td>\n",
       "      <td>lineage4,lineage4.1,lineage4.1.1,lineage4.1.1.3</td>\n",
       "      <td>4</td>\n",
       "      <td>CDC1551</td>\n",
       "      <td>GCF_000008585.1_ASM858v1_feature_table.txt.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GCA_000016145.1_ASM1614v1_genomic</td>\n",
       "      <td>lineage4,lineage4.9</td>\n",
       "      <td>4</td>\n",
       "      <td>H37Ra</td>\n",
       "      <td>GCF_000016145.1_ASM1614v1_feature_table.txt.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GCA_000016925.1_ASM1692v1_genomic</td>\n",
       "      <td>lineage4,lineage4.3,lineage4.3.2,lineage4.3.2.1</td>\n",
       "      <td>4</td>\n",
       "      <td>F11</td>\n",
       "      <td>GCF_000016925.1_ASM1692v1_feature_table.txt.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GCA_000153685.2_ASM15368v2_genomic</td>\n",
       "      <td>lineage4,lineage4.1,lineage4.1.2,lineage4.1.2.1</td>\n",
       "      <td>4</td>\n",
       "      <td>str. Haarlem</td>\n",
       "      <td>GCF_000153685.2_ASM15368v2_feature_table.txt.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GCA_000195955.2_ASM19595v2_genomic</td>\n",
       "      <td>lineage4,lineage4.9</td>\n",
       "      <td>4</td>\n",
       "      <td>H37Rv complete genome</td>\n",
       "      <td>GCF_000195955.2_ASM19595v2_feature_table.txt.gz</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                        Strain  \\\n",
       "Unnamed: 0                                       \n",
       "0             GCA_000008585.1_ASM858v1_genomic   \n",
       "1            GCA_000016145.1_ASM1614v1_genomic   \n",
       "2            GCA_000016925.1_ASM1692v1_genomic   \n",
       "3           GCA_000153685.2_ASM15368v2_genomic   \n",
       "4           GCA_000195955.2_ASM19595v2_genomic   \n",
       "\n",
       "                                                   Decision Evaluation  \\\n",
       "Unnamed: 0                                                               \n",
       "0           lineage4,lineage4.1,lineage4.1.1,lineage4.1.1.3          4   \n",
       "1                                       lineage4,lineage4.9          4   \n",
       "2           lineage4,lineage4.3,lineage4.3.2,lineage4.3.2.1          4   \n",
       "3           lineage4,lineage4.1,lineage4.1.2,lineage4.1.2.1          4   \n",
       "4                                       lineage4,lineage4.9          4   \n",
       "\n",
       "                      Strain name  \\\n",
       "Unnamed: 0                          \n",
       "0                         CDC1551   \n",
       "1                           H37Ra   \n",
       "2                             F11   \n",
       "3                    str. Haarlem   \n",
       "4           H37Rv complete genome   \n",
       "\n",
       "                                     Feature_Table_filename  \n",
       "Unnamed: 0                                                   \n",
       "0             GCF_000008585.1_ASM858v1_feature_table.txt.gz  \n",
       "1            GCF_000016145.1_ASM1614v1_feature_table.txt.gz  \n",
       "2            GCF_000016925.1_ASM1692v1_feature_table.txt.gz  \n",
       "3           GCF_000153685.2_ASM15368v2_feature_table.txt.gz  \n",
       "4           GCF_000195955.2_ASM19595v2_feature_table.txt.gz  "
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ref_seqs_annot_for_sims.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(54, 5)"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.shape(ref_seqs_annot_for_sims)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "for strain_i in ref_seqs_annot_for_sims.index:\n",
    "    \n",
    "    #load mappings from simulations\n",
    "    complete_genome_H37Rv_mapping_table_i = pd.read_csv('/n/data1/hms/dbmi/farhat/Roger/inhost_TB_dynamics_project/simulated_reads/reference_sequences/strains_typed/genome_plasticity_window_analysis_csv_files_10_gene_windows_rough_match/{0}_H37Rv_mapping_window_analysis.csv'.format(ref_seqs_annot_for_sims.loc[strain_i , 'Strain']))\n",
    "\n",
    "    #clean table by dropping unneccessary columns\n",
    "    complete_genome_H37Rv_mapping_table_i = complete_genome_H37Rv_mapping_table_i.loc[: , ['Unnamed: 0' , 'symbol' , 'H37Rv_start' , 'H37Rv_end' , 'RefGenome_start' , 'RefGenome_end' , 'strand']]\n",
    "\n",
    "    #rename columns\n",
    "    complete_genome_H37Rv_mapping_table_i.rename(columns={\"Unnamed: 0\": \"H37Rv Tag\" , \"RefGenome_start\": \"CompleteGenome_start\" , \"RefGenome_end\": \"CompleteGenome_end\"} , inplace = True)\n",
    "\n",
    "    #set index to H37Rv locus tag\n",
    "    complete_genome_H37Rv_mapping_table_i.set_index('H37Rv Tag' , inplace = True , drop = True)\n",
    "\n",
    "    #output to directory with cleaned tables\n",
    "    complete_genome_H37Rv_mapping_table_i.to_csv('/n/data1/hms/dbmi/farhat/Roger/inhost_TB_dynamics_project/simulated_reads/reference_sequences/strains_typed/genome_plasticity_window_analysis_csv_files_10_gene_windows_rough_match_cleaned_tables/{0}_H37Rv_mappings.csv'.format(ref_seqs_annot_for_sims.loc[strain_i , 'Strain']))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
